--- a/src/main/java/org/bukkit/inventory/meta/AxolotlBucketMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/AxolotlBucketMeta.java
@@ -37,4 +_,43 @@
 
     @Override
     AxolotlBucketMeta clone();
+
+    // Kiterino start - Expose age to AxolotlBucketMeta
+    /**
+     * Gets the age of this axolotl
+     *
+     * @return the age of this axolotl
+     */
+    int getAge();
+
+    /**
+     * Sets the age of this axolotl
+     *
+     * @param age the age of this axolotl
+     */
+    void setAge(int age);
+
+    /**
+     * Checks whether this axolotl is an adult
+     *
+     * @return whether this axolotl is an adult
+     */
+    default boolean isAdult() {
+        return getAge() >= 0;
+    }
+
+    /**
+     * Makes this axolotl an adult
+     */
+    default void setAdult() {
+        setAge(0);
+    }
+
+    /**
+     * Makes this axolotl a baby
+     */
+    default void setBaby() {
+        setAge(-24000);
+    }
+    // Kiterino end - Expose age to AxolotlBucketMeta
 }
