--- a/src/main/java/io/papermc/paper/datacomponent/item/PaperItemLore.java
+++ b/src/main/java/io/papermc/paper/datacomponent/item/PaperItemLore.java
@@ -58,6 +_,20 @@
             return this;
         }
 
+        // Kiterino start - Data components API improvements
+        @Override
+        public List<Component> getLines() {
+            return this.lines;
+        }
+
+        @Override
+        public ItemLore.Builder addLine(int index, final ComponentLike line) {
+            validateLineCount(this.lines.size(), 1);
+            this.lines.add(index, line.asComponent());
+            return this;
+        }
+        // Kiterino end - Data components API improvements
+
         @Override
         public ItemLore.Builder addLines(final List<? extends ComponentLike> lines) {
             validateLineCount(this.lines.size(), lines.size());
