--- a/src/main/java/org/bukkit/block/Beacon.java
+++ b/src/main/java/org/bukkit/block/Beacon.java
@@ -34,6 +_,21 @@
      */
     int getTier();
 
+    // Kiterino start - Beacon level override API
+    /**
+     * Gets the overridden tier of the beacon pyramid (0-4), or -1 if not overridden.
+     *
+     * @see #getTier()
+     * @return the overridden beacon tier
+     */
+    int getTierOverride();
+
+    /**
+     * Sets the overridden tier of the beacon pyramid (0-4), or -1 if not overridden
+     */
+    void setTierOverride(@org.jetbrains.annotations.Range(from = -1L, to = 4L) int tier);
+    // Kiterino end - Beacon level override API
+
     /**
      * Returns the primary effect set on the beacon
      *
