--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -151,6 +_,19 @@
 
             // Paper start - eula system property
             boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
+            // Kiterino start - Accept the EULA
+            if (!eula.hasAgreedToEULA() && !eulaAgreed) {
+                Main.LOGGER.warn("Automatically accepting Mojang's EULA (https://aka.ms/MinecraftEULA).");
+                Main.LOGGER.warn("Please, stop using this Software immediately if you do not agree with this.");
+                Main.LOGGER.warn("Note: this is a persistent change and will be stored in eula.txt from now on.");
+                try (java.io.OutputStream outputStream = java.nio.file.Files.newOutputStream(eulaFile)) {
+                    var properties = new java.util.Properties();
+                    properties.setProperty("eula", "true");
+                    properties.store(outputStream, null);
+                }
+                eulaAgreed = true;
+            }
+            // Kiterino end - Accept the EULA
             if (eulaAgreed) {
                 LOGGER.warn("You have used the Paper command line EULA agreement flag.");
                 LOGGER.warn("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
