--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
@@ -70,7 +_,11 @@
                 choices.add(CraftItemStack.asBukkitCopy(i));
             }
 
-            return new RecipeChoice.ExactChoice(choices);
+            // Kiterino start - Add predicate support for the recipe book
+            var recipeChoice = new RecipeChoice.ExactChoice(choices);
+            if (ingredient.predicate != null) recipeChoice.setPredicate(ingredient.predicate);
+            return recipeChoice;
+            // Kiterino end - Add predicate support for the recipe book
         } else {
             final RegistryKeySet<ItemType> itemTypes = PaperRegistrySets.convertToApi(RegistryKey.ITEM, ingredient.values);
             return RecipeChoice.itemType(itemTypes);
