--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaAxolotlBucket.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaAxolotlBucket.java
@@ -221,4 +_,19 @@
 
         return builder;
     }
+
+    // Kiterino start - Expose age to AxolotlBucketMeta
+    @Override
+    public int getAge() {
+        return this.bucketEntityTag == null ? 0 : this.bucketEntityTag.getIntOr("Age", 0);
+    }
+
+    @Override
+    public void setAge(int age) {
+        if (this.bucketEntityTag == null) {
+            this.bucketEntityTag = new CompoundTag();
+        }
+        this.bucketEntityTag.putInt("Age", age);
+    }
+    // Kiterino end - Expose age to AxolotlBucketMeta
 }
