--- a/net/minecraft/network/chat/ComponentSerialization.java
+++ b/net/minecraft/network/chat/ComponentSerialization.java
@@ -52,13 +_,92 @@
 
             @Override
             public void encode(RegistryFriendlyByteBuf registryFriendlyByteBuf, Component object) {
+                // Kiterino start - Parse hover events
+                if (me.sosedik.kiterino.KiterinoConfig.parseItemHoversEverywhere && !DONT_RENDER_TRANSLATABLES.get()) {
+                    java.util.Locale locale = registryFriendlyByteBuf.adventure$locale;
+                    if (locale == null) {
+                        locale = java.util.Locale.US;
+                    }
+                    object = replaceHoverEvent(null, locale, object);
+                }
+                Component finalObject = object;
+                // Kiterino end - Parse hover events
                 RegistryOps<net.minecraft.nbt.Tag> registryOps = registryFriendlyByteBuf.registryAccess().createSerializationContext(net.minecraft.nbt.NbtOps.INSTANCE);
                 net.minecraft.nbt.Tag tag = (DONT_RENDER_TRANSLATABLES.get() ? CODEC : ComponentSerialization.localizedCodec(registryFriendlyByteBuf.adventure$locale))
-                    .encodeStart(registryOps, object).getOrThrow(error -> new io.netty.handler.codec.EncoderException("Failed to encode: " + error + " " + object));
+                    .encodeStart(registryOps, object).getOrThrow(error -> new io.netty.handler.codec.EncoderException("Failed to encode: " + error + " " + finalObject)); // Kiterino - Parse hover events
                 this.streamCodec.encode(registryFriendlyByteBuf, tag);
             }
         };
     }
+    // Kiterino start - Parse hover events
+    public static @org.jetbrains.annotations.Nullable Component replaceHoverEvent(@org.jetbrains.annotations.Nullable org.bukkit.craftbukkit.entity.CraftPlayer player, java.util.Locale locale, @org.jetbrains.annotations.Nullable Component component) {
+        if (component == null) {
+            return null;
+        }
+        // Go through children first
+        List<Component> children = component.getSiblings();
+        if (!children.isEmpty()) {
+            // Adventure components have immutable children/siblings
+            if (!(component instanceof MutableComponent)) {
+                component = component.copy();
+                children = component.getSiblings();
+            }
+            try {
+                children.replaceAll(child -> replaceHoverEvent(player, locale, child));
+                // Mutable component with immutable children?
+            } catch (UnsupportedOperationException ignored1) {
+                component = component.copy();
+                children = component.getSiblings();
+                try {
+                    children.replaceAll(child -> replaceHoverEvent(player, locale, child));
+                    // Shouldn't happen, but just in case
+                } catch (UnsupportedOperationException ignored2) {
+                    return component;
+                }
+            }
+        }
+
+        // Translatables may have component arguments
+        if (component.getContents() instanceof TranslatableContents translatable) {
+            Object[] args = translatable.getArgs();
+            for (int i = 0; i < args.length; i++) {
+                if (!(args[i] instanceof Component arg)) continue;
+
+                args[i] = replaceHoverEvent(player, locale, arg);
+            }
+        }
+
+        // Finally replace hover if needed
+        HoverEvent hoverEvent = component.getStyle().hoverEvent;
+        if (hoverEvent == null) {
+            return component;
+        }
+
+        if (hoverEvent instanceof HoverEvent.ShowItem(net.minecraft.world.item.ItemStack item)) {
+	        var context = me.sosedik.kiterino.modifier.item.context.ItemModifierContext.EMPTY;
+            var contextBox = new me.sosedik.kiterino.modifier.item.ItemContextBox(player, locale, me.sosedik.kiterino.modifier.item.context.ItemModifierContextType.CHAT_ITEM, context, item.asBukkitCopy());
+            var bukkitItem = me.sosedik.kiterino.modifier.item.ItemModifier.modifyItem(contextBox);
+            if (bukkitItem != null) {
+                Style style = component.getStyle().withHoverEvent(new HoverEvent.ShowItem(net.minecraft.world.item.ItemStack.fromBukkitCopy(bukkitItem)));
+                if (component instanceof MutableComponent mutableComponent) {
+                    mutableComponent.setStyle(style);
+                } else {
+                    component = component.copy().setStyle(style);
+                }
+            }
+        }
+
+        if (hoverEvent instanceof HoverEvent.ShowEntity(HoverEvent.EntityTooltipInfo entityTooltipInfo)) {
+            net.minecraft.world.entity.EntityType<?> entityType = entityTooltipInfo.type;
+            var bukkitType = org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(entityType);
+            if (bukkitType != null && !org.bukkit.NamespacedKey.MINECRAFT.equals(bukkitType.key().namespace())) {
+                component = component.copy().setStyle(component.getStyle().withHoverEvent(new HoverEvent.ShowText(Component.literal(bukkitType.key().asString()))));
+            }
+        }
+
+        return component;
+    }
+    // Kiterino end - Parse hover events
     // Paper end - adventure; use locale from bytebuf for translation
     public static final StreamCodec<RegistryFriendlyByteBuf, Optional<Component>> TRUSTED_OPTIONAL_STREAM_CODEC = TRUSTED_STREAM_CODEC.apply(
         ByteBufCodecs::optional
