--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -3341,7 +_,9 @@
     );
     public static final Block PACKED_ICE = register(
         "packed_ice",
+        me.sosedik.kiterino.world.level.block.PackedIceBlock::new, // Kiterino - Melt packed ice in Nether
         BlockBehaviour.Properties.of().mapColor(MapColor.ICE).instrument(NoteBlockInstrument.CHIME).friction(0.98F).strength(0.5F).sound(SoundType.GLASS)
+            .randomTicks() // Kiterino - Melt packed ice in Nether
     );
     public static final Block SUNFLOWER = register(
         "sunflower",
@@ -7076,5 +_,6 @@
                 blockState.initCache();
             }
         }
+        me.sosedik.kiterino.world.block.KiterinoBlockStateIdMapper.lastVanillaBlockStateId = Block.BLOCK_STATE_REGISTRY.size() - 1; // Kiterino - Data-driven blocks
     }
 }
