--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -11,7 +_,7 @@
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
 
-public class CauldronBlock extends AbstractCauldronBlock {
+public class CauldronBlock extends AbstractCauldronBlock implements LiquidBlockContainer { // Kiterino - Cauldrons can consume liquids
     public static final MapCodec<CauldronBlock> CODEC = simpleCodec(CauldronBlock::new);
     private static final float RAIN_FILL_CHANCE = 0.05F;
     private static final float POWDER_SNOW_FILL_CHANCE = 0.1F;
@@ -84,4 +_,30 @@
             level.levelEvent(1046, pos, 0);
         }
     }
+
+    // Kiterino start - Cauldrons can consume liquids
+    @Override
+    public boolean canPlaceLiquid(@javax.annotation.Nullable net.minecraft.world.entity.LivingEntity owner, net.minecraft.world.level.BlockGetter level, BlockPos pos, BlockState state, Fluid fluid) {
+        return (fluid == Fluids.FLOWING_WATER || fluid == Fluids.FLOWING_LAVA) && state.getFluidState().getAmount() == 0;
+    }
+
+    @Override
+    public boolean placeLiquid(net.minecraft.world.level.LevelAccessor level, BlockPos pos, BlockState state, net.minecraft.world.level.material.FluidState fluidState) {
+        if (fluidState.getAmount() != 8) {
+            return false;
+        }
+
+        BlockState newState;
+        if (fluidState.getType() == Fluids.WATER) {
+            newState = Blocks.WATER_CAULDRON.defaultBlockState().setValue(LayeredCauldronBlock.LEVEL, LayeredCauldronBlock.MAX_FILL_LEVEL);
+        } else if (fluidState.getType() == Fluids.LAVA) {
+            newState = Blocks.LAVA_CAULDRON.defaultBlockState();
+        } else {
+            return false;
+        }
+
+        level.setBlock(pos, newState, 3);
+        return true;
+    }
+    // Kiterino end - Cauldrons can consume liquids
 }
