--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -83,6 +_,13 @@
         ResourceKey<Level> resourceKey1 = flag ? respawnData.dimension() : Level.END;
         BlockPos blockPos = flag ? respawnData.pos() : ServerLevel.END_SPAWN_POINT;
         ServerLevel level1 = level.getServer().getLevel(resourceKey1);
+        // Kiterino start - Call EntityPortalReadyEvent for End Portal
+        io.papermc.paper.event.entity.EntityPortalReadyEvent portalReadyEvent = new io.papermc.paper.event.entity.EntityPortalReadyEvent(entity.getBukkitEntity(), level1 == null ? null : level1.getWorld(), org.bukkit.PortalType.ENDER);
+        if (!portalReadyEvent.callEvent()) {
+            return null;
+        }
+        level1 = portalReadyEvent.getTargetWorld() == null ? null : ((org.bukkit.craftbukkit.CraftWorld) portalReadyEvent.getTargetWorld()).getHandle();
+        // Kiterino end - Call EntityPortalReadyEvent for End Portal
         if (level1 == null) {
             return null;
         } else {
@@ -103,7 +_,7 @@
                 f1 = respawnData.pitch();
                 set = Relative.union(Relative.DELTA, Relative.ROTATION);
                 if (entity instanceof ServerPlayer serverPlayer) {
-                    return serverPlayer.findRespawnPositionAndUseSpawnBlock(false, TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.END_PORTAL); // CraftBukkit
+                    return serverPlayer.findRespawnPositionAndUseSpawnBlock(false, TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.END_PORTAL, level1); // CraftBukkit // Kiterino - Call EntityPortalReadyEvent for End Portal
                 }
 
                 bottomCenter = entity.adjustSpawnLocation(level1, blockPos).getBottomCenter();
