--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -39,10 +_,19 @@
     public abstract MapCodec<? extends LeavesBlock> codec();
 
     public LeavesBlock(float leafParticleChance, BlockBehaviour.Properties properties) {
-        super(properties);
+        super(properties(properties)); // Kiterino - No collision for leaves
         this.leafParticleChance = leafParticleChance;
         this.registerDefaultState(this.stateDefinition.any().setValue(DISTANCE, 7).setValue(PERSISTENT, false).setValue(WATERLOGGED, false));
     }
+
+    // Kiterino start - No collision for leaves
+    private static BlockBehaviour.Properties properties(BlockBehaviour.Properties properties) {
+        if (me.sosedik.kiterino.KiterinoConfig.noCollisionForLeaves) {
+            properties = properties.noCollision().forceSolidOff();
+        }
+        return properties;
+    }
+    // Kiterino end - No collision for leaves
 
     @Override
     protected boolean skipRendering(BlockState state, BlockState adjacentState, Direction direction) {
