--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -39,11 +_,20 @@
     public abstract MapCodec<? extends LeavesBlock> codec();
 
     public LeavesBlock(final float leafParticleChance, final BlockBehaviour.Properties properties) {
-        super(properties);
+        super(properties(properties)); // Kiterino - No collision for leaves
         this.leafParticleChance = leafParticleChance;
         this.registerDefaultState(this.stateDefinition.any().setValue(DISTANCE, 7).setValue(PERSISTENT, false).setValue(WATERLOGGED, false));
     }
 
+    // Kiterino start - No collision for leaves
+    private static BlockBehaviour.Properties properties(BlockBehaviour.Properties properties) {
+        if (me.sosedik.kiterino.KiterinoConfig.noCollisionForLeaves) {
+            properties = properties.noCollision().forceSolidOff();
+        }
+        return properties;
+    }
+    // Kiterino end - No collision for leaves
+
     @Override
     protected boolean skipRendering(final BlockState state, final BlockState neighborState, final Direction direction) {
         return !cutoutLeaves && neighborState.getBlock() instanceof LeavesBlock || super.skipRendering(state, neighborState, direction);
@@ -55,7 +_,7 @@
 
     @Override
     protected VoxelShape getBlockSupportShape(final BlockState state, final BlockGetter level, final BlockPos pos) {
-        return Shapes.empty();
+        return me.sosedik.kiterino.KiterinoConfig.noCollisionForLeaves ? Shapes.block() : Shapes.empty();
     }
 
     @Override
