--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -41,6 +_,8 @@
     public static final IntegerProperty NOTE = BlockStateProperties.NOTE;
     public static final int NOTE_VOLUME = 3;
 
+    private @Nullable Player lastPlayedBy = null; // Kiterino - Add player to NotePlayEvent
+
     @Override
     public MapCodec<NoteBlock> codec() {
         return CODEC;
@@ -108,12 +_,26 @@
     }
 
     private void playNote(final @Nullable Entity source, final BlockState state, final Level level, final BlockPos pos) {
-        if (level.purpurConfig.noteBlockIgnoreAbove || state.getValue(INSTRUMENT).worksAboveNoteBlock() || level.getBlockState(pos.above()).isAir()) { // Purpur - Config to allow Note Block sounds when blocked
+        if (level.purpurConfig.noteBlockIgnoreAbove || state.getValue(INSTRUMENT).worksAboveNoteBlock() || (level.kiterinoConfig.noteBlockAirOnAnySide && hasAir(level, pos)) || level.getBlockState(pos.above()).isAir()) { // Purpur - Config to allow Note Block sounds when blocked // Kiterino - Note blocks require air on any side
+            if (source instanceof Player player) this.lastPlayedBy = player; // Kiterino - Add player to NotePlayEvent
             level.blockEvent(pos, this, 0, 0);
             level.gameEvent(source, GameEvent.NOTE_BLOCK_PLAY, pos);
         }
     }
 
+    // Kiterino start - Note blocks require air on any side
+    private boolean hasAir(Level level, BlockPos pos) {
+        for (Direction.Plane plane : Direction.Plane.values()) {
+            for (Direction direction : plane) {
+                if (level.getBlockState(pos.relative(direction)).isAir()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    // Kiterino end - Note blocks require air on any side
+
     @Override
     protected InteractionResult useItemOn(
         final ItemStack itemStack,
@@ -132,7 +_,15 @@
     @Override
     protected InteractionResult useWithoutItem(BlockState state, final Level level, final BlockPos pos, final Player player, final BlockHitResult hitResult) {
         if (!level.isClientSide()) {
-            if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) state = state.cycle(NoteBlock.NOTE); // Paper - prevent noteblock note from updating
+            // Kiterino start - Reverse note block playing
+            if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) {
+                if (level.kiterinoConfig.reverseNoteBlockPlaying && player.isShiftKeyDown()) {
+                    state = state.cycleBackwards(NoteBlock.NOTE);
+                } else {
+                    state = state.cycle(NoteBlock.NOTE); // Paper - prevent noteblock note from updating
+                }
+            }
+            // Kiterino end - Reverse note block playing
             level.setBlock(pos, state, Block.UPDATE_ALL);
             this.playNote(player, state, level, pos);
             player.awardStat(Stats.TUNE_NOTEBLOCK);
@@ -157,7 +_,8 @@
     protected boolean triggerEvent(final BlockState state, final Level level, final BlockPos pos, final int b0, final int b1) {
         NoteBlockInstrument instrument = state.getValue(INSTRUMENT);
         // Paper start - move NotePlayEvent call to fix instrument/note changes
-        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(level, pos, instrument, state.getValue(NOTE));
+        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(level, pos, instrument, state.getValue(NOTE), this.lastPlayedBy); // Kiterino - Add player to NotePlayEvent
+        this.lastPlayedBy = null; // Kiterino - Add player to NotePlayEvent
         if (event.isCancelled()) return false;
         // Paper end - move NotePlayEvent call to fix instrument/note changes
         float pitch;
