--- a/net/minecraft/world/entity/projectile/ThrownSplashPotion.java
+++ b/net/minecraft/world/entity/projectile/ThrownSplashPotion.java
@@ -40,7 +_,7 @@
         PotionContents potionContents = stack.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY);
         float orDefault = stack.getOrDefault(DataComponents.POTION_DURATION_SCALE, 1.0F);
         Iterable<MobEffectInstance> allEffects = potionContents.getAllEffects();
-        AABB aabb = this.getBoundingBox().move(hitResult.getLocation().subtract(this.position()));
+        AABB aabb = hitResult == null ? this.getBoundingBox() : this.getBoundingBox().move(hitResult.getLocation().subtract(this.position())); // Kiterino - Fixup ThrownPotion#splash NPE
         AABB aabb1 = aabb.inflate(4.0, 2.0, 4.0);
         List<LivingEntity> entitiesOfClass = this.level().getEntitiesOfClass(LivingEntity.class, aabb1);
         java.util.Map<org.bukkit.entity.LivingEntity, Double> affected = new java.util.HashMap<>(); // CraftBukkit
