--- a/net/minecraft/world/entity/monster/zombie/Drowned.java
+++ b/net/minecraft/world/entity/monster/zombie/Drowned.java
@@ -99,6 +_,17 @@
     }
     // Purpur end - Ridables
 
+    // Kiterino start - Allow jumping in water as Drowned
+    @Override
+    public boolean onSpacebar() {
+        if (isInWater()) {
+            jumpFromGround();
+            return true;
+        }
+        return false;
+    }
+    // Kiterino end - Allow jumping in water as Drowned
+
     @Override
     protected PathNavigation createNavigation(final Level level) {
         return new AmphibiousPathNavigation(this, level);
