--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -493,6 +_,28 @@
         return true;
     }
 
+    // Kiterino start - Allow individual chunk refresh
+    @Override
+    public boolean refreshChunk(int x, int z, Collection<Player> players) {
+        ChunkHolder playerChunk = this.world.getChunkSource().chunkMap.getVisibleChunkIfPresent(ChunkPos.pack(x, z));
+        if (playerChunk == null) return false;
+
+        // Paper start - chunk system
+        net.minecraft.world.level.chunk.LevelChunk chunk = playerChunk.getChunkToSend();
+        if (chunk == null) {
+            return false;
+        }
+        // Paper end - chunk system
+                List<ServerPlayer> playersInRange = players.stream().map(player -> ((CraftPlayer) player).getHandle()).toList();
+                if (playersInRange.isEmpty()) return true; // Paper - chunk system
+
+                FeatureHooks.sendChunkRefreshPackets(playersInRange, chunk);
+        // Paper - chunk system
+
+        return true;
+    }
+    // Kiterino end - Allow individual chunk refresh
+
     @Override
     public Collection<Player> getPlayersSeeingChunk(Chunk chunk) {
         Preconditions.checkArgument(chunk != null, "chunk cannot be null");
@@ -2090,4 +_,11 @@
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    // Kiterino start - Allow checking for Halloween season
+    @Override
+    public boolean isHalloweenSeason() {
+        return getHandle().purpurConfig.forceHalloweenSeason || net.minecraft.util.SpecialDates.isExtendedHalloween();
+    }
+    // Kiterino end - Allow checking for Halloween season
 }
