--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -190,6 +_,14 @@
                         .defaultsTo("Unknown Server")
                         .describedAs("Name");
                 // Paper end
+
+                // Kiterino start - Kiterino Config
+                acceptsAll(asList("kiterino", "kiterino-settings"), "File for Kiterino settings")
+                    .withRequiredArg()
+                    .ofType(File.class)
+                    .defaultsTo(new File("kiterino.yml"))
+                    .describedAs("Yml file");
+// Kiterino end - Kiterino Config
             }
         };
 
