--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -1114,7 +_,7 @@
         float friction = 0.6F;
         float speedFactor = 1.0F;
         float jumpFactor = 1.0F;
-        private @Nullable ResourceKey<Block> id;
+        private @Nullable ResourceKey<Block> id; public @Nullable ResourceKey<Block> getId() { return this.id; } // Kiterino - Data-driven blocks
         private DependantName<Block, Optional<ResourceKey<LootTable>>> drops = key -> Optional.of(
             ResourceKey.create(Registries.LOOT_TABLE, key.identifier().withPrefix("blocks/"))
         );
