--- a/src/main/java/org/bukkit/scheduler/BukkitRunnable.java
+++ b/src/main/java/org/bukkit/scheduler/BukkitRunnable.java
@@ -10,6 +_,17 @@
 public abstract class BukkitRunnable implements Runnable {
     private BukkitTask task;
 
+    // Kiterino start - Allow checking task schedule
+    /**
+     * Checks whether this task has been scheduled
+     *
+     * @return true if the task has been scheduled
+     */
+    public synchronized boolean isScheduled() {
+        return task != null;
+    }
+    // Kiterino end - Allow checking task schedule
+
     /**
      * Returns true if this task has been cancelled.
      *
