--- a/net/minecraft/network/protocol/game/ClientboundSetBorderCenterPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetBorderCenterPacket.java
@@ -23,6 +_,13 @@
         this.newCenterZ = input.readDouble();
     }
 
+    // Kiterino start - Red tint API
+    public ClientboundSetBorderCenterPacket(double newCenterX, double newCenterZ) {
+        this.newCenterX = newCenterX;
+        this.newCenterZ = newCenterZ;
+    }
+    // Kiterino ens - Red tint API
+
     private void write(final FriendlyByteBuf output) {
         output.writeDouble(this.newCenterX);
         output.writeDouble(this.newCenterZ);
