--- a/net/minecraft/world/attribute/BedRule.java
+++ b/net/minecraft/world/attribute/BedRule.java
@@ -51,7 +_,7 @@
         public boolean test(final Level level) {
             return switch (this) {
                 case ALWAYS -> true;
-                case WHEN_DARK -> level.isDarkOutside();
+                case WHEN_DARK -> level.kiterinoConfig.dayDreaming || level.isDarkOutside(); // Kiterino - Sleeping options
                 case NEVER -> false;
             };
         }
