From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 12 Dec 2025 06:50:05 +0200
Subject: [PATCH] Fixup ridables controls


diff --git a/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java b/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java
index 65301b22a918b300f30d52838fc7c885f3f7b31c..7bfe89337bb169c98c37546135eb69186e1a5fd8 100644
--- a/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java
+++ b/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java
@@ -36,7 +36,7 @@ public class WaterMoveControllerWASD<T extends Mob> extends MoveControllerWASD<T
             vertical = 0.0F;
         }
 
-        if (rider.isJumping() && spacebarEvent(entity)) {
+        if (lastClientInput.jump() && spacebarEvent(entity)) { // Kiterino - Fixup ridables controls
             entity.onSpacebar();
         }
 
