From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 12 Dec 2025 06:50:05 +0200
Subject: [PATCH] Fixup ridables controls


diff --git a/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java b/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java
index 922e48799c43ca322a8f550c98a26e1e2959439c..d3ac67ded8188bd4e06eaea62600e09c69154389 100644
--- a/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java
+++ b/src/main/java/org/purpurmc/purpur/controller/WaterMoveControllerWASD.java
@@ -36,7 +36,7 @@ public class WaterMoveControllerWASD extends MoveControllerWASD {
             vertical = 0.0F;
         }
 
-        if (rider.jumping && spacebarEvent(entity)) {
+        if (lastClientInput.jump() && spacebarEvent(entity)) { // Kiterino - Fixup ridables controls
             entity.onSpacebar();
         }
 
