--- a/net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
@@ -48,6 +_,7 @@
             }
         }
 
+        if (!org.purpurmc.purpur.PurpurConfig.loggerSuppressUnrecognizedRecipeErrors) // Kiterino - Suppress unknown recipe warn
         LOGGER.warn("Couldn't smelt {} because there is no smelting recipe", itemStack);
         return itemStack;
     }
