--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -466,10 +_,24 @@
         net.minecraft.world.item.ItemStack nmsItem = CraftItemStack.asNMSCopy(item);
         boolean result = false;
 
+        // Kiterino start - Call BlockDestroyEvent for breakNaturally
+        boolean dropResources = true;
+        int expDrop = dropExperience ? block.getExpDrop(state, this.level.getMinecraftWorld(), this.position, nmsItem, true) : 0;
+        if (!state.isAir()) {
+            var blockDestroyEvent = new com.destroystokyo.paper.event.block.BlockDestroyEvent(this, state.getFluidState().createLegacyBlock().asBlockData(), state.asBlockData(), expDrop, true);
+            blockDestroyEvent.setPlayEffect(triggerEffect);
+            if (!blockDestroyEvent.callEvent()) return false;
+
+            triggerEffect = blockDestroyEvent.playEffect();
+            dropResources = blockDestroyEvent.willDrop();
+            expDrop = blockDestroyEvent.getExpToDrop();
+        }
+        // Kiterino end - Call BlockDestroyEvent for breakNaturally
+
         // Modelled off Player#hasCorrectToolForDrops
         if (!state.isAir() && (item == null || !state.requiresCorrectToolForDrops() || nmsItem.isCorrectToolForDrops(state))) {
-            net.minecraft.world.level.block.Block.dropResources(state, this.level.getMinecraftWorld(), this.position, this.level.getBlockEntity(this.position), null, nmsItem, false);
-            if (dropExperience) block.popExperience(this.level.getMinecraftWorld(), this.position, block.getExpDrop(state, this.level.getMinecraftWorld(), this.position, nmsItem, true));
+            if (dropResources) net.minecraft.world.level.block.Block.dropResources(state, this.level.getMinecraftWorld(), this.position, this.level.getBlockEntity(this.position), null, nmsItem, false); // Kiterino - Call BlockDestroyEvent for breakNaturally
+            if (expDrop > 0) block.popExperience(this.level.getMinecraftWorld(), this.position, expDrop); // Kiterino - Call BlockDestroyEvent for breakNaturally
             result = true;
         }
 
