--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -511,11 +_,25 @@
         net.minecraft.world.item.ItemStack nmsItem = CraftItemStack.asNMSCopy(item);
         boolean result = false;
 
+        // Kiterino start - Call BlockDestroyEvent for breakNaturally
+        boolean dropResources = true;
+        int expDrop = dropExperience ? block.getExpDrop(state, this.world.getMinecraftWorld(), this.position, nmsItem, true) : 0;
+        if (block != Blocks.AIR) {
+            var blockDestroyEvent = new com.destroystokyo.paper.event.block.BlockDestroyEvent(this, state.getFluidState().createLegacyBlock().createCraftBlockData(), state.createCraftBlockData(), expDrop, true);
+            blockDestroyEvent.setPlayEffect(triggerEffect);
+            if (!blockDestroyEvent.callEvent()) return false;
+
+            triggerEffect = blockDestroyEvent.playEffect();
+            dropResources = blockDestroyEvent.willDrop();
+            expDrop = blockDestroyEvent.getExpToDrop();
+        }
+        // Kiterino end - Call BlockDestroyEvent for breakNaturally
+
         // Modelled off Player#hasCorrectToolForDrops
         if (block != Blocks.AIR && (item == null || !state.requiresCorrectToolForDrops() || nmsItem.isCorrectToolForDrops(state))) {
-            net.minecraft.world.level.block.Block.dropResources(state, this.world.getMinecraftWorld(), this.position, this.world.getBlockEntity(this.position), null, nmsItem, false); // Paper - Properly handle xp dropping
+            if (dropResources) net.minecraft.world.level.block.Block.dropResources(state, this.world.getMinecraftWorld(), this.position, this.world.getBlockEntity(this.position), null, nmsItem, false); // Paper - Properly handle xp dropping // Kiterino - Call BlockDestroyEvent for breakNaturally
             // Paper start - improve Block#breakNaturally
-            if (dropExperience) block.popExperience(this.world.getMinecraftWorld(), this.position, block.getExpDrop(state, this.world.getMinecraftWorld(), this.position, nmsItem, true));
+            if (expDrop > 0) block.popExperience(this.world.getMinecraftWorld(), this.position, expDrop); // Kiterino - Call BlockDestroyEvent for breakNaturally
             // Paper end
             result = true;
         }
