From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 23 Jan 2025 13:04:56 +0200
Subject: [PATCH] Data-driven items


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index 01dbd27a3dea07a46962ab19b66d16a03ab64f99..d4dd11cccb16e46994d441745e37303622d525d6 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -95,7 +95,7 @@ public final class PaperRegistries {
             start(Registries.STRUCTURE_TYPE, RegistryKey.STRUCTURE_TYPE).craft(StructureType.class, CraftStructureType::new).build(),
             start(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT).craft(PotionEffectType.class, CraftPotionEffectType::new).build(),
             start(Registries.BLOCK, RegistryKey.BLOCK).craft(BlockType.class, CraftBlockType::new).build(),
-            start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).build(),
+            start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).writable(me.sosedik.kiterino.registry.data.KiterinoItemRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven items
             start(Registries.VILLAGER_PROFESSION, RegistryKey.VILLAGER_PROFESSION).craft(Villager.Profession.class, CraftVillager.CraftProfession::new).build(),
             start(Registries.VILLAGER_TYPE, RegistryKey.VILLAGER_TYPE).craft(Villager.Type.class, CraftVillager.CraftType::new).build(),
             start(Registries.MAP_DECORATION_TYPE, RegistryKey.MAP_DECORATION_TYPE).craft(MapCursor.Type.class, CraftMapCursor.CraftType::new).build(),
diff --git a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
index b1710da835cee3333bfb1c238b37ec0c4d16ba09..cadbf29c4667a74d79fdc538e0cca9cc61480d29 100644
--- a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
+++ b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
@@ -55,4 +55,24 @@ public class Conversions {
     public Component asAdventure(final net.minecraft.network.chat.@Nullable Component vanilla) {
         return vanilla == null ? Component.empty() : this.serializer.deserialize(vanilla);
     }
+
+    // Kiterino start - Data-driven items
+    public net.minecraft.world.item.Rarity asVanilla(final org.bukkit.inventory.ItemRarity rarity) {
+        return switch (rarity) {
+            case COMMON -> net.minecraft.world.item.Rarity.COMMON;
+            case UNCOMMON -> net.minecraft.world.item.Rarity.UNCOMMON;
+            case RARE -> net.minecraft.world.item.Rarity.RARE;
+            case EPIC -> net.minecraft.world.item.Rarity.EPIC;
+        };
+    }
+
+    public org.bukkit.inventory.ItemRarity asBukkit(final net.minecraft.world.item.Rarity rarity) {
+        return switch (rarity) {
+            case COMMON -> org.bukkit.inventory.ItemRarity.COMMON;
+            case UNCOMMON -> org.bukkit.inventory.ItemRarity.UNCOMMON;
+            case RARE -> org.bukkit.inventory.ItemRarity.RARE;
+            case EPIC -> org.bukkit.inventory.ItemRarity.EPIC;
+        };
+    }
+    // Kiterino end - Data-driven items
 }
diff --git a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
index 7fa18dd100f92cc866ba71270ead883373fda2d9..6377057ef2b5d832367f5d736fa5e6acd9da01ea 100644
--- a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
+++ b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
@@ -24,6 +24,7 @@ class RegistryBuilderTest {
         return PaperRegistries.REGISTRY_ENTRIES.stream()
             .map(RegistryEntry::meta)
             .filter(RegistryEntryMeta.Buildable.class::isInstance)
+            .filter(entry -> !entry.mcKey().location().equals(net.minecraft.core.registries.Registries.ITEM.location())) // Kiterino - Data-driven items
             .map(Arguments::arguments);
     }
 
