From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 28 Jan 2025 20:36:45 +0200
Subject: [PATCH] Data-driven mob effects


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index d7740899502c3ee71b3e652b33c078b364cc958f..ecac060b943b61d9e5761c2fd39a90fb506b49e4 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -114,7 +114,7 @@ public final class PaperRegistries {
             // built-in
             start(Registries.GAME_EVENT, RegistryKey.GAME_EVENT).craft(GameEvent.class, CraftGameEvent::new).writable(PaperGameEventRegistryEntry.PaperBuilder::new),
             start(Registries.STRUCTURE_TYPE, RegistryKey.STRUCTURE_TYPE).craft(StructureType.class, CraftStructureType::new).build(),
-            start(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT).craft(PotionEffectType.class, CraftPotionEffectType::new).build(),
+            start(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT).craft(PotionEffectType.class, CraftPotionEffectType::new).writable(me.sosedik.kiterino.registry.data.KiterinoMobEffectRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven mob effects
             start(Registries.BLOCK, RegistryKey.BLOCK).craft(BlockType.class, CraftBlockType::new).writable(me.sosedik.kiterino.registry.data.KiterinoBlockRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven blocks
             start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).writable(me.sosedik.kiterino.registry.data.KiterinoItemRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven items
             start(Registries.VILLAGER_PROFESSION, RegistryKey.VILLAGER_PROFESSION).craft(Villager.Profession.class, CraftVillager.CraftProfession::new).build(),
diff --git a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
index 41f9b61bcd66ed4cf2c03623838d052921417e31..037f1808be5f3e919a68f623bdccdeec29c4e951 100644
--- a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
+++ b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
@@ -118,4 +118,18 @@ public class Conversions {
         };
     }
     // Kiterino end - Data-driven items
+
+    // Kiterino start - Data-driven mob effects
+    public net.minecraft.world.effect.MobEffectCategory asVanilla(final org.bukkit.potion.PotionEffectType.Category category) {
+        return switch (category) {
+            case BENEFICIAL -> net.minecraft.world.effect.MobEffectCategory.BENEFICIAL;
+            case HARMFUL -> net.minecraft.world.effect.MobEffectCategory.HARMFUL;
+            case NEUTRAL -> net.minecraft.world.effect.MobEffectCategory.NEUTRAL;
+        };
+    }
+
+    public org.bukkit.potion.PotionEffectType.Category asBukkit(final net.minecraft.world.effect.MobEffectCategory category) {
+        return org.bukkit.craftbukkit.potion.CraftPotionEffectType.fromNMS(category);
+    }
+    // Kiterino end - Data-driven mob effects
 }
diff --git a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
index 3e1bbb5f92610ff2f15c839c340759c7783f36b6..e4728a3d07d31919ce6077621fd5f2e2bdcd57f8 100644
--- a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
+++ b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
@@ -25,6 +25,7 @@ class RegistryBuilderTest {
             .filter(RegistryEntryMeta.Buildable.class::isInstance)
             .filter(entry -> !entry.mcKey().identifier().equals(net.minecraft.core.registries.Registries.ITEM.identifier())) // Kiterino - Data-driven items
             .filter(entry -> !entry.mcKey().identifier().equals(net.minecraft.core.registries.Registries.BLOCK.identifier())) // Kiterino - Data-driven blocks
+            .filter(entry -> !entry.mcKey().identifier().equals(net.minecraft.core.registries.Registries.MOB_EFFECT.identifier())) // Kiterino - Data-driven mob effects
             .map(Arguments::arguments);
     }
 
