From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 11 Oct 2022 14:28:39 +0300
Subject: [PATCH] Paper PR - Fire ServerListPingEvent for secondary motd send

https://github.com/PaperMC/Paper/pull/8074

diff --git a/src/main/java/com/destroystokyo/paper/network/StandardPaperServerListPingEventImpl.java b/src/main/java/com/destroystokyo/paper/network/StandardPaperServerListPingEventImpl.java
index 5e89bad7231d0ec389f6be085943cc027485bd7a..46bf215b99068e26e53a61ccdbd6f95b3ab6975d 100644
--- a/src/main/java/com/destroystokyo/paper/network/StandardPaperServerListPingEventImpl.java
+++ b/src/main/java/com/destroystokyo/paper/network/StandardPaperServerListPingEventImpl.java
@@ -64,13 +64,23 @@ public final class StandardPaperServerListPingEventImpl extends PaperServerListP
     }
 
     public static void processRequest(MinecraftServer server, Connection connection) {
+	    ServerStatus ping = getEventResponse(server, connection);
+	    if (ping == null) {
+		    connection.disconnect((Component) null);
+		    return;
+	    }
+
+	    // Send response
+	    connection.send(new ClientboundStatusResponsePacket(ping));
+    }
+
+	public static ServerStatus getEventResponse(MinecraftServer server, Connection connection) {
         StandardPaperServerListPingEventImpl event = new StandardPaperServerListPingEventImpl(server, connection, server.getStatus());
         server.server.getPluginManager().callEvent(event);
 
         // Close connection immediately if event is cancelled
         if (event.isCancelled()) {
-            connection.disconnect((Component) null);
-            return;
+            return null;
         }
 
         // Setup response
@@ -98,8 +108,7 @@ public final class StandardPaperServerListPingEventImpl extends PaperServerListP
         }
         final ServerStatus ping = new ServerStatus(description, players, Optional.of(version), favicon, server.enforceSecureProfile());
 
-        // Send response
-        connection.send(new ClientboundStatusResponsePacket(ping));
+	    return ping;
     }
 
 }
