From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 29 Jan 2025 18:01:34 +0200
Subject: [PATCH] Keep rider when teleporting


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 200eead2c76602a341d221f2ba8ede307d735718..3d317f372367f157f6d295862184cab6c2837a1c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -293,11 +293,18 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     // Kiterino start - EMC - Vehicle passenger teleporting
     private static CraftEntity teleportingEntity;
     private CraftEntity storedRoot;
+    private CraftHumanEntity storedRider; // Kiterino - Keep rider when teleporting
     private com.google.common.collect.ImmutableList<CraftEntity> storedPassengers;
     void detachEntities() {
         if (teleportingEntity == null) {
             teleportingEntity = this;
             this.storedRoot = entity.getRootVehicle().getBukkitEntity();
+            // Kiterino start - Keep rider when teleporting
+            CraftEntity vehicle = (CraftEntity) getVehicle();
+            if (vehicle != null) {
+                this.storedRider = (CraftHumanEntity) vehicle.getRider();
+            }
+            // Kiterino end - Keep rider when teleporting
             this.storedRoot.detachEntities();
             teleportingEntity = null;
             return;
@@ -321,6 +328,15 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
                 teleportingEntity = this;
                 this.storedRoot.teleportAndReattachEntities(location);
                 this.storedRoot.reattachPassengers();
+                // Kiterino start - Keep rider when teleporting
+                if (this.storedRider != null) {
+                    CraftEntity vehicle = (CraftEntity) getVehicle();
+                    if (vehicle != null) {
+                        vehicle.getHandle().rider = this.storedRider.getHandle();
+                    }
+                    this.storedRider = null;
+                }
+                // Kiterino end - Keep rider when teleporting
                 this.storedRoot = null;
                 teleportingEntity = null;
             }
