From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 17 May 2025 00:33:39 +0300
Subject: [PATCH] Extra recipes API


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftBlastingRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftBlastingRecipe.java
index 3e39a5172c06dcd2a7964df73dac9e960e583794..e1d89fb41f9bb5a12e4ff718402cf2b838fe2612 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftBlastingRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftBlastingRecipe.java
@@ -19,6 +19,7 @@ public class CraftBlastingRecipe extends BlastingRecipe implements CraftRecipe {
             return (CraftBlastingRecipe) recipe;
         }
         CraftBlastingRecipe ret = new CraftBlastingRecipe(recipe.getKey(), recipe.getResult(), recipe.getInputChoice(), recipe.getExperience(), recipe.getCookingTime());
+        ret.setSpecial(recipe.isSpecial()); // Kiterino - Extra recipes API
         ret.setGroup(recipe.getGroup());
         ret.setCategory(recipe.getCategory());
         return ret;
@@ -34,6 +35,7 @@ public class CraftBlastingRecipe extends BlastingRecipe implements CraftRecipe {
             this.getExperience(),
             this.getCookingTime()
         );
+	    recipe.special = isSpecial(); // Kiterino - Extra recipes API
         MinecraftServer.getServer().getRecipeManager().addRecipe(new RecipeHolder<>(CraftNamespacedKey.toResourceKey(Registries.RECIPE, this.getKey()), recipe));
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftCampfireRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftCampfireRecipe.java
index 00a65754b649883b5af0c58aca759882c24ca1ea..e5e690f06a3e388a630abecba6a9cb1f703ccfb4 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftCampfireRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftCampfireRecipe.java
@@ -20,6 +20,7 @@ public class CraftCampfireRecipe extends CampfireRecipe implements CraftRecipe {
             return (CraftCampfireRecipe) recipe;
         }
         CraftCampfireRecipe ret = new CraftCampfireRecipe(recipe.getKey(), recipe.getResult(), recipe.getInputChoice(), recipe.getExperience(), recipe.getCookingTime());
+        ret.setSpecial(recipe.isSpecial()); // Kiterino - Extra recipes API
         ret.setGroup(recipe.getGroup());
         ret.setCategory(recipe.getCategory());
         return ret;
@@ -35,6 +36,7 @@ public class CraftCampfireRecipe extends CampfireRecipe implements CraftRecipe {
             this.getExperience(),
             this.getCookingTime()
         );
+	    recipe.special = isSpecial(); // Kiterino - Extra recipes API
         MinecraftServer.getServer().getRecipeManager().addRecipe(new RecipeHolder<>(CraftNamespacedKey.toResourceKey(Registries.RECIPE, this.getKey()), recipe));
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftFurnaceRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftFurnaceRecipe.java
index 658af5f3aeaa52cf70704497e0f1ec25df7670a6..c498b8d6feaea844b64fa0a1ac628833a81e3061 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftFurnaceRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftFurnaceRecipe.java
@@ -20,6 +20,7 @@ public class CraftFurnaceRecipe extends FurnaceRecipe implements CraftRecipe {
             return (CraftFurnaceRecipe) recipe;
         }
         CraftFurnaceRecipe ret = new CraftFurnaceRecipe(recipe.getKey(), recipe.getResult(), recipe.getInputChoice(), recipe.getExperience(), recipe.getCookingTime());
+        ret.setSpecial(recipe.isSpecial()); // Kiterino - Extra recipes API
         ret.setGroup(recipe.getGroup());
         ret.setCategory(recipe.getCategory());
         return ret;
@@ -35,6 +36,7 @@ public class CraftFurnaceRecipe extends FurnaceRecipe implements CraftRecipe {
             this.getExperience(),
             this.getCookingTime()
         );
+	    recipe.special = isSpecial(); // Kiterino - Extra recipes API
         MinecraftServer.getServer().getRecipeManager().addRecipe(new RecipeHolder<>(CraftNamespacedKey.toResourceKey(Registries.RECIPE, this.getKey()), recipe));
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapedRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapedRecipe.java
index bc509ebe9fdc16a0e3850a5c1b69c2e364cae1de..db77cf3e0d15a5544fedcdbf100158384670ee15 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapedRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapedRecipe.java
@@ -32,6 +32,7 @@ public class CraftShapedRecipe extends ShapedRecipe implements CraftRecipe {
             return (CraftShapedRecipe) recipe;
         }
         CraftShapedRecipe ret = new CraftShapedRecipe(recipe.getKey(), recipe.getResult());
+        ret.setSpecial(recipe.isSpecial()); // Kiterino - Extra recipes API
         ret.setGroup(recipe.getGroup());
         ret.setCategory(recipe.getCategory());
         String[] shape = recipe.getShape();
@@ -60,6 +61,7 @@ public class CraftShapedRecipe extends ShapedRecipe implements CraftRecipe {
             pattern,
             CraftItemStack.asTemplate(this.getResult())
         );
+	    recipe.special = isSpecial(); // Kiterino - Extra recipes API
         MinecraftServer.getServer().getRecipeManager().addRecipe(new RecipeHolder<>(CraftNamespacedKey.toResourceKey(Registries.RECIPE, this.getKey()), recipe));
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapelessRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapelessRecipe.java
index 6a7bb225f044c9f299d6a3e703de82a1093878b4..dc901881404254aef41cb61badb73f1a9ab27b52 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapelessRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftShapelessRecipe.java
@@ -30,6 +30,7 @@ public class CraftShapelessRecipe extends ShapelessRecipe implements CraftRecipe
             return (CraftShapelessRecipe) recipe;
         }
         CraftShapelessRecipe ret = new CraftShapelessRecipe(recipe.getKey(), recipe.getResult());
+        ret.setSpecial(recipe.isSpecial()); // Kiterino - Extra recipes API
         ret.setGroup(recipe.getGroup());
         ret.setCategory(recipe.getCategory());
         for (RecipeChoice ingredient : recipe.getChoiceList()) {
@@ -52,6 +53,7 @@ public class CraftShapelessRecipe extends ShapelessRecipe implements CraftRecipe
             CraftItemStack.asTemplate(this.getResult()),
             ingredients
         );
+	    recipe.special = isSpecial(); // Kiterino - Extra recipes API
         MinecraftServer.getServer().getRecipeManager().addRecipe(new RecipeHolder<>(CraftNamespacedKey.toResourceKey(Registries.RECIPE, this.getKey()), recipe));
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftSmokingRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftSmokingRecipe.java
index 25b5c019e63d1922802b7e501ffac5223fad52b6..20bbf4960ff5a1d90d55a1b8163c8286aa6606df 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftSmokingRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftSmokingRecipe.java
@@ -19,6 +19,7 @@ public class CraftSmokingRecipe extends SmokingRecipe implements CraftRecipe {
             return (CraftSmokingRecipe) recipe;
         }
         CraftSmokingRecipe ret = new CraftSmokingRecipe(recipe.getKey(), recipe.getResult(), recipe.getInputChoice(), recipe.getExperience(), recipe.getCookingTime());
+        ret.setSpecial(recipe.isSpecial()); // Kiterino - Extra recipes API
         ret.setGroup(recipe.getGroup());
         ret.setCategory(recipe.getCategory());
         return ret;
@@ -34,6 +35,7 @@ public class CraftSmokingRecipe extends SmokingRecipe implements CraftRecipe {
             this.getExperience(),
             this.getCookingTime()
         );
+	    recipe.special = isSpecial(); // Kiterino - Extra recipes API
         MinecraftServer.getServer().getRecipeManager().addRecipe(new RecipeHolder<>(CraftNamespacedKey.toResourceKey(Registries.RECIPE, this.getKey()), recipe));
     }
 }
