From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Isaac - The456 <the456@the456gamer.dev>
Date: Sun, 8 Sep 2024 23:57:54 +0100
Subject: [PATCH] Paper PR - Add Improved Lidded API

https://github.com/PaperMC/Paper/pull/11814

diff --git a/src/main/java/io/papermc/paper/block/PaperLidded.java b/src/main/java/io/papermc/paper/block/PaperLidded.java
new file mode 100644
index 0000000000000000000000000000000000000000..a3aab126cf2ccd04ebf32c80b530f43c4a688736
--- /dev/null
+++ b/src/main/java/io/papermc/paper/block/PaperLidded.java
@@ -0,0 +1,87 @@
+package io.papermc.paper.block;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public interface PaperLidded extends Lidded, org.bukkit.block.Lidded {
+
+    @Override
+    default LidMode setLidMode(final LidMode targetLidMode) {
+        final LidMode oldLidMode = getLidMode();
+        final LidMode newLidMode = getResultantLidMode(targetLidMode);
+
+        if (oldLidMode == newLidMode) {
+            // already in correct state
+            return newLidMode;
+        }
+
+        boolean wasForcedOpen =
+                oldLidMode == LidMode.FORCED_OPEN || oldLidMode == LidMode.OPEN_UNTIL_VIEWED;
+        boolean wasForcedClosed =
+                oldLidMode == LidMode.FORCED_CLOSED || oldLidMode == LidMode.CLOSED_UNTIL_NOT_VIEWED;
+        boolean isForcedOpen =
+                newLidMode == LidMode.FORCED_OPEN || newLidMode == LidMode.OPEN_UNTIL_VIEWED;
+        boolean isForcedClosed =
+                newLidMode == LidMode.FORCED_CLOSED || newLidMode == LidMode.CLOSED_UNTIL_NOT_VIEWED;
+
+        // stop any existing force open/close, if next state doesn't need it.
+        if (wasForcedOpen && !isForcedOpen) {
+            stopForceLiddedLidOpen();
+        } else if (wasForcedClosed && !isForcedClosed) {
+            stopForceLiddedLidClose();
+        }
+
+        // start new force open/close, if it wasn't previously.
+        if (isForcedOpen && !wasForcedOpen) {
+            startForceLiddedLidOpen();
+        } else if (isForcedClosed && !wasForcedClosed) {
+            startForceLiddedLidClose();
+        }
+
+        // return the new lid mode, so it can be stored by the implementation.
+        return newLidMode;
+    }
+
+    private LidMode getResultantLidMode(LidMode targetLidMode) {
+        final LidState trueLidState = getTrueLidState();
+
+        // check that target lid mode is valid for true lid state.
+        LidMode newLidMode;
+
+        if (targetLidMode == LidMode.CLOSED_UNTIL_NOT_VIEWED
+                && trueLidState == LidState.CLOSED) {
+            // insta-revert to default, as the lid is already closed.
+            newLidMode = LidMode.DEFAULT;
+        } else if (targetLidMode == LidMode.OPEN_UNTIL_VIEWED
+                && trueLidState == LidState.OPEN) {
+            // insta-revert to default, as the lid is already open.
+            newLidMode = LidMode.DEFAULT;
+        } else {
+            newLidMode = targetLidMode;
+        }
+        return newLidMode;
+    }
+
+    // these should be similar to the vanilla open/close behavior.
+    void startForceLiddedLidOpen();
+    void stopForceLiddedLidOpen();
+    void startForceLiddedLidClose();
+    void stopForceLiddedLidClose();
+
+    // bukkit lidded impl using the paper lidded api.
+
+    @Override
+    default boolean isOpen() {
+        return getLidMode() == LidMode.FORCED_OPEN || getLidMode() == LidMode.OPEN_UNTIL_VIEWED;
+    }
+
+    @Override
+    default void close() {
+        setLidMode(LidMode.DEFAULT);
+    }
+
+    @Override
+    default void open() {
+        setLidMode(LidMode.FORCED_OPEN);
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBarrel.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBarrel.java
index 6063f0e1fdc232d063105971359ae688168a2bc4..b777f0f8c0feaaeb331b56dfa0f20bf84fa611d3 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBarrel.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBarrel.java
@@ -10,7 +10,7 @@ import org.bukkit.block.Barrel;
 import org.bukkit.craftbukkit.inventory.CraftInventory;
 import org.bukkit.inventory.Inventory;
 
-public class CraftBarrel extends CraftLootable<BarrelBlockEntity> implements Barrel {
+public class CraftBarrel extends CraftLootable<BarrelBlockEntity> implements Barrel, io.papermc.paper.block.PaperLidded { // Paper - Add Improved Lidded Api
 
     public CraftBarrel(World world, BarrelBlockEntity tileEntity) {
         super(world, tileEntity);
@@ -34,35 +34,35 @@ public class CraftBarrel extends CraftLootable<BarrelBlockEntity> implements Bar
         return new CraftInventory(this.getTileEntity());
     }
 
-    @Override
-    public void open() {
-        this.requirePlaced();
-        if (!this.getTileEntity().openersCounter.opened) {
-            BlockState blockData = this.getTileEntity().getBlockState();
-            boolean open = blockData.getValue(BarrelBlock.OPEN);
-
-            if (!open) {
-                this.getTileEntity().updateBlockState(blockData, true);
-                if (this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-                    this.getTileEntity().playSound(blockData, SoundEvents.BARREL_OPEN);
-                }
-            }
-        }
-        this.getTileEntity().openersCounter.opened = true;
-    }
-
-    @Override
-    public void close() {
-        this.requirePlaced();
-        if (this.getTileEntity().openersCounter.opened) {
-            BlockState blockData = this.getTileEntity().getBlockState();
-            this.getTileEntity().updateBlockState(blockData, false);
-            if (this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-                this.getTileEntity().playSound(blockData, SoundEvents.BARREL_CLOSE);
-            }
-        }
-        this.getTileEntity().openersCounter.opened = false;
-    }
+//    @Override
+//    public void open() {
+//        this.requirePlaced();
+//        if (!this.getTileEntity().openersCounter.opened) {
+//            BlockState blockData = this.getTileEntity().getBlockState();
+//            boolean open = blockData.getValue(BarrelBlock.OPEN);
+//
+//            if (!open) {
+//                this.getTileEntity().updateBlockState(blockData, true);
+//                if (this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//                    this.getTileEntity().playSound(blockData, SoundEvents.BARREL_OPEN);
+//                }
+//            }
+//        }
+//        this.getTileEntity().openersCounter.opened = true;
+//    }
+//
+//    @Override
+//    public void close() {
+//        this.requirePlaced();
+//        if (this.getTileEntity().openersCounter.opened) {
+//            BlockState blockData = this.getTileEntity().getBlockState();
+//            this.getTileEntity().updateBlockState(blockData, false);
+//            if (this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//                this.getTileEntity().playSound(blockData, SoundEvents.BARREL_CLOSE);
+//            }
+//        }
+//        this.getTileEntity().openersCounter.opened = false;
+//    }
 
     @Override
     public CraftBarrel copy() {
@@ -74,10 +74,53 @@ public class CraftBarrel extends CraftLootable<BarrelBlockEntity> implements Bar
         return new CraftBarrel(this, location);
     }
 
-    // Paper start - More Lidded Block API
+    // Paper start - add Improved Lidded API
+//    // Paper start - More Lidded Block API
+//    @Override
+//    public boolean isOpen() {
+//        return getTileEntity().openersCounter.opened;
+//    }
+//    // Paper end - More Lidded Block API
+    public void startForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.startForceLiddedLidOpen(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
     @Override
-    public boolean isOpen() {
-        return getTileEntity().openersCounter.opened;
+    public void stopForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.stopForceLiddedLidOpen(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void startForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.startForceLiddedLidClose(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void stopForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.stopForceLiddedLidClose(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getEffectiveLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getTrueLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode getLidMode() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getLidMode();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode setLidMode(final io.papermc.paper.block.LidMode targetLidMode) {
+        this.requirePlaced();
+        io.papermc.paper.block.LidMode newEffectiveMode = io.papermc.paper.block.PaperLidded.super.setLidMode(targetLidMode);
+        this.getTileEntity().openersCounter.setLidMode(newEffectiveMode);
+        return newEffectiveMode;
     }
-    // Paper end - More Lidded Block API
+    // Paper end - add Improved Lidded API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java b/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java
index cc7bf4d39b834fba472bc163226a01a0cd4b6010..8062ecd51b4bc369be08a26c02ec593ef8ab01f6 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java
@@ -14,7 +14,7 @@ import org.bukkit.craftbukkit.inventory.CraftInventory;
 import org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest;
 import org.bukkit.inventory.Inventory;
 
-public class CraftChest extends CraftLootable<ChestBlockEntity> implements Chest {
+public class CraftChest extends CraftLootable<ChestBlockEntity> implements Chest, io.papermc.paper.block.PaperLidded { // Paper - Add Improved Lidded Api
 
     public CraftChest(World world, ChestBlockEntity tileEntity) {
         super(world, tileEntity);
@@ -57,31 +57,31 @@ public class CraftChest extends CraftLootable<ChestBlockEntity> implements Chest
         return inventory;
     }
 
-    @Override
-    public void open() {
-        this.requirePlaced();
-        if (!this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-            BlockState block = this.getTileEntity().getBlockState();
-            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
-
-            this.getTileEntity().openersCounter.onAPIOpen((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
-            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, openCount + 1);
-        }
-        this.getTileEntity().openersCounter.opened = true;
-    }
-
-    @Override
-    public void close() {
-        this.requirePlaced();
-        if (this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-            BlockState block = this.getTileEntity().getBlockState();
-            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
-
-            this.getTileEntity().openersCounter.onAPIClose((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
-            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, 0);
-        }
-        this.getTileEntity().openersCounter.opened = false;
-    }
+//    @Override
+//    public void open() {
+//        this.requirePlaced();
+//        if (!this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//            BlockState block = this.getTileEntity().getBlockState();
+//            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
+//
+//            this.getTileEntity().openersCounter.onAPIOpen((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
+//            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, openCount + 1);
+//        }
+//        this.getTileEntity().openersCounter.opened = true;
+//    }
+//
+//    @Override
+//    public void close() {
+//        this.requirePlaced();
+//        if (this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//            BlockState block = this.getTileEntity().getBlockState();
+//            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
+//
+//            this.getTileEntity().openersCounter.onAPIClose((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
+//            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, 0);
+//        }
+//        this.getTileEntity().openersCounter.opened = false;
+//    }
 
     @Override
     public CraftChest copy() {
@@ -93,12 +93,12 @@ public class CraftChest extends CraftLootable<ChestBlockEntity> implements Chest
         return new CraftChest(this, location);
     }
 
-    // Paper start - More Lidded Block API
-    @Override
-    public boolean isOpen() {
-        return getTileEntity().openersCounter.opened;
-    }
-    // Paper end - More Lidded Block API
+//    // Paper start - More Lidded Block API
+//    @Override
+//    public boolean isOpen() {
+//        return getTileEntity().openersCounter.opened;
+//    }
+//    // Paper end - More Lidded Block API
 
     // Paper start - More Chest Block API
     @Override
@@ -124,4 +124,49 @@ public class CraftChest extends CraftLootable<ChestBlockEntity> implements Chest
             && ChestBlock.isChestBlockedAt(world, neighbourBlockPos);
     }
     // Paper end - More Chest Block API
+
+    // Paper start - add Improved Lidded API
+    @Override
+    public void startForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.startForceLiddedLidOpen(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void stopForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.stopForceLiddedLidOpen(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void startForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.startForceLiddedLidClose(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void stopForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.stopForceLiddedLidClose(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getEffectiveLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getTrueLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode getLidMode() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getLidMode();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode setLidMode(final io.papermc.paper.block.LidMode targetLidMode) {
+        this.requirePlaced();
+        io.papermc.paper.block.LidMode newEffectiveMode = io.papermc.paper.block.PaperLidded.super.setLidMode(targetLidMode);
+        this.getTileEntity().openersCounter.setLidMode(newEffectiveMode);
+        return newEffectiveMode;
+    }
+    // Paper end - add Improved Lidded API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftEnderChest.java b/src/main/java/org/bukkit/craftbukkit/block/CraftEnderChest.java
index f45ee675a10729845bf376fa95e648b23b9aac12..90a00efd5178491b82dab273bde88efbdadab3a7 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftEnderChest.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftEnderChest.java
@@ -6,7 +6,7 @@ import org.bukkit.Location;
 import org.bukkit.World;
 import org.bukkit.block.EnderChest;
 
-public class CraftEnderChest extends CraftBlockEntityState<EnderChestBlockEntity> implements EnderChest {
+public class CraftEnderChest extends CraftBlockEntityState<EnderChestBlockEntity> implements EnderChest, io.papermc.paper.block.PaperLidded { // Paper - Add Improved Lidded Api
 
     public CraftEnderChest(World world, EnderChestBlockEntity tileEntity) {
         super(world, tileEntity);
@@ -16,31 +16,31 @@ public class CraftEnderChest extends CraftBlockEntityState<EnderChestBlockEntity
         super(state, location);
     }
 
-    @Override
-    public void open() {
-        this.requirePlaced();
-        if (!this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-            BlockState block = this.getTileEntity().getBlockState();
-            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
-
-            this.getTileEntity().openersCounter.onAPIOpen((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
-            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, openCount + 1);
-        }
-        this.getTileEntity().openersCounter.opened = true;
-    }
-
-    @Override
-    public void close() {
-        this.requirePlaced();
-        if (this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-            BlockState block = this.getTileEntity().getBlockState();
-            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
-
-            this.getTileEntity().openersCounter.onAPIClose((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
-            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, 0);
-        }
-        this.getTileEntity().openersCounter.opened = false;
-    }
+//    @Override
+//    public void open() {
+//        this.requirePlaced();
+//        if (!this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//            BlockState block = this.getTileEntity().getBlockState();
+//            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
+//
+//            this.getTileEntity().openersCounter.onAPIOpen((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
+//            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, openCount + 1);
+//        }
+//        this.getTileEntity().openersCounter.opened = true;
+//    }
+//
+//    @Override
+//    public void close() {
+//        this.requirePlaced();
+//        if (this.getTileEntity().openersCounter.opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//            BlockState block = this.getTileEntity().getBlockState();
+//            int openCount = this.getTileEntity().openersCounter.getOpenerCount();
+//
+//            this.getTileEntity().openersCounter.onAPIClose((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block);
+//            this.getTileEntity().openersCounter.openerAPICountChanged((net.minecraft.world.level.Level) this.getWorldHandle(), this.getPosition(), block, openCount, 0);
+//        }
+//        this.getTileEntity().openersCounter.opened = false;
+//    }
 
     @Override
     public CraftEnderChest copy() {
@@ -52,12 +52,12 @@ public class CraftEnderChest extends CraftBlockEntityState<EnderChestBlockEntity
         return new CraftEnderChest(this, location);
     }
 
-    // Paper start - More Lidded Block API
-    @Override
-    public boolean isOpen() {
-        return getTileEntity().openersCounter.opened;
-    }
-    // Paper end - More Lidded Block API
+//    // Paper start - More Lidded Block API
+//    @Override
+//    public boolean isOpen() {
+//        return getTileEntity().openersCounter.opened;
+//    }
+//    // Paper end - More Lidded Block API
 
     // Paper start - More Chest Block API
     @Override
@@ -67,4 +67,49 @@ public class CraftEnderChest extends CraftBlockEntityState<EnderChestBlockEntity
         return this.isPlaced() && this.getWorldHandle().getBlockState(abovePos).isRedstoneConductor(this.getWorldHandle(), abovePos);
     }
     // Paper end - More Chest Block API
+
+    // Paper start - add Improved Lidded API
+    @Override
+    public void startForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.startForceLiddedLidOpen(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void stopForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.stopForceLiddedLidOpen(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void startForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.startForceLiddedLidClose(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public void stopForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().openersCounter.stopForceLiddedLidClose(this.getTileEntity().getLevel(), this.getTileEntity().getBlockPos(), this.getTileEntity().getBlockState());
+    }
+    @Override
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getEffectiveLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getTrueLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode getLidMode() {
+        this.requirePlaced();
+        return this.getTileEntity().openersCounter.getLidMode();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode setLidMode(final io.papermc.paper.block.LidMode targetLidMode) {
+        this.requirePlaced();
+        io.papermc.paper.block.LidMode newEffectiveMode = io.papermc.paper.block.PaperLidded.super.setLidMode(targetLidMode);
+        this.getTileEntity().openersCounter.setLidMode(newEffectiveMode);
+        return newEffectiveMode;
+    }
+    // Paper end - add Improved Lidded API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftShulkerBox.java b/src/main/java/org/bukkit/craftbukkit/block/CraftShulkerBox.java
index f7b199fbc7a740de3ee6952ce12ef2c35f057d7a..8a9222ed563d11f5e77e1c796f098d67fc946518 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftShulkerBox.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftShulkerBox.java
@@ -11,7 +11,7 @@ import org.bukkit.block.ShulkerBox;
 import org.bukkit.craftbukkit.inventory.CraftInventory;
 import org.bukkit.inventory.Inventory;
 
-public class CraftShulkerBox extends CraftLootable<ShulkerBoxBlockEntity> implements ShulkerBox {
+public class CraftShulkerBox extends CraftLootable<ShulkerBoxBlockEntity> implements ShulkerBox, io.papermc.paper.block.PaperLidded { // Paper - Add Improved Lidded Api
 
     public CraftShulkerBox(World world, ShulkerBoxBlockEntity tileEntity) {
         super(world, tileEntity);
@@ -42,27 +42,27 @@ public class CraftShulkerBox extends CraftLootable<ShulkerBoxBlockEntity> implem
         return (color == null) ? null : DyeColor.getByWoolData((byte) color.getId());
     }
 
-    @Override
-    public void open() {
-        this.requirePlaced();
-        if (!this.getTileEntity().opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-            net.minecraft.world.level.Level world = this.getTileEntity().getLevel();
-            world.blockEvent(this.getPosition(), this.getTileEntity().getBlockState().getBlock(), 1, 1);
-            world.playSound(null, this.getPosition(), SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, world.random.nextFloat() * 0.1F + 0.9F);
-        }
-        this.getTileEntity().opened = true;
-    }
-
-    @Override
-    public void close() {
-        this.requirePlaced();
-        if (this.getTileEntity().opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
-            net.minecraft.world.level.Level world = this.getTileEntity().getLevel();
-            world.blockEvent(this.getPosition(), this.getTileEntity().getBlockState().getBlock(), 1, 0);
-            world.playSound(null, this.getPosition(), SoundEvents.SHULKER_BOX_CLOSE, SoundSource.BLOCKS, 0.5F, world.random.nextFloat() * 0.1F + 0.9F); // Paper - More Lidded Block API (Wrong sound)
-        }
-        this.getTileEntity().opened = false;
-    }
+//    @Override
+//    public void open() {
+//        this.requirePlaced();
+//        if (!this.getTileEntity().opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//            net.minecraft.world.level.Level world = this.getTileEntity().getLevel();
+//            world.blockEvent(this.getPosition(), this.getTileEntity().getBlockState().getBlock(), 1, 1);
+//            world.playSound(null, this.getPosition(), SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, world.random.nextFloat() * 0.1F + 0.9F);
+//        }
+//        this.getTileEntity().opened = true;
+//    }
+//
+//    @Override
+//    public void close() {
+//        this.requirePlaced();
+//        if (this.getTileEntity().opened && this.getWorldHandle() instanceof net.minecraft.world.level.Level) {
+//            net.minecraft.world.level.Level world = this.getTileEntity().getLevel();
+//            world.blockEvent(this.getPosition(), this.getTileEntity().getBlockState().getBlock(), 1, 0);
+//            world.playSound(null, this.getPosition(), SoundEvents.SHULKER_BOX_CLOSE, SoundSource.BLOCKS, 0.5F, world.random.nextFloat() * 0.1F + 0.9F); // Paper - More Lidded Block API (Wrong sound)
+//        }
+//        this.getTileEntity().opened = false;
+//    }
 
     @Override
     public CraftShulkerBox copy() {
@@ -74,10 +74,54 @@ public class CraftShulkerBox extends CraftLootable<ShulkerBoxBlockEntity> implem
         return new CraftShulkerBox(this, location);
     }
 
-    // Paper start - More Lidded Block API
+//    // Paper start - More Lidded Block API
+//    @Override
+//    public boolean isOpen() {
+//        return getTileEntity().opened;
+//    }
+//    // Paper end - More Lidded Block API
+
+    // Paper start - add Improved Lidded API
+    public void startForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().startForceLiddedLidOpen();
+    }
     @Override
-    public boolean isOpen() {
-        return getTileEntity().opened;
+    public void stopForceLiddedLidOpen() {
+        this.requirePlaced();
+        this.getTileEntity().stopForceLiddedLidOpen();
+    }
+    @Override
+    public void startForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().startForceLiddedLidClose();
+    }
+    @Override
+    public void stopForceLiddedLidClose() {
+        this.requirePlaced();
+        this.getTileEntity().stopForceLiddedLidClose();
+    }
+    @Override
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().getEffectiveLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        this.requirePlaced();
+        return this.getTileEntity().getTrueLidState();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode getLidMode() {
+        this.requirePlaced();
+        return this.getTileEntity().getLidMode();
+    }
+    @Override
+    public io.papermc.paper.block.LidMode setLidMode(final io.papermc.paper.block.LidMode targetLidMode) {
+        this.requirePlaced();
+        io.papermc.paper.block.LidMode newEffectiveMode = io.papermc.paper.block.PaperLidded.super.setLidMode(targetLidMode);
+        this.getTileEntity().setLidMode(newEffectiveMode);
+        return newEffectiveMode;
     }
-    // Paper end - More Lidded Block API
+    // Paper end - add Improved Lidded API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 0134cdf4f1fabefa55cecd973c3e237b6227c713..eb06dc183e59fa2be2c7a80f56f85cc2a57b704d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -2306,4 +2306,16 @@ public class CraftEventFactory {
         return event;
     }
     // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
+
+    // Kiterino start - Paper PR - New and Improved Lidded API
+    public static boolean callPlayerLiddedOpenEvent(net.minecraft.world.entity.player.Player who, final Level world, final BlockPos pos) {
+        Player player = (Player) who.getBukkitEntity();
+        Block block = CraftBlock.at(world, pos);
+        io.papermc.paper.block.PaperLidded blockState = (io.papermc.paper.block.PaperLidded) CraftBlockStates.getBlockState(block);
+
+        var event = new io.papermc.paper.event.player.PlayerLiddedOpenEvent(player, blockState, block);
+
+        return event.callEvent();
+    }
+    // Kiterino end - Paper PR - New and Improved Lidded API
 }
