From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 15 Nov 2025 19:37:04 +0200
Subject: [PATCH] Expose bucket in bucket events


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 22725f22a51773f96262416e852f6bbefe2fd289..6f75aae59fb33c5dc0829b41d625f4b40fb171ce 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -437,17 +437,22 @@ public class CraftEventFactory {
     }
 
     public static PlayerBucketEmptyEvent callPlayerBucketEmptyEvent(Level world, net.minecraft.world.entity.player.Player player, BlockPos changed, BlockPos clicked, Direction clickedFace, ItemStack itemInHand, InteractionHand hand) {
-        return (PlayerBucketEmptyEvent) CraftEventFactory.getPlayerBucketEvent(false, world, player, changed, clicked, clickedFace, itemInHand, Items.BUCKET, hand);
+        return (PlayerBucketEmptyEvent) CraftEventFactory.getPlayerBucketEvent(false, world, player, changed, clicked, clickedFace, itemInHand, new ItemStack(Items.BUCKET), hand); // Kiterino - Expose bucket in bucket events
     }
 
     public static PlayerBucketFillEvent callPlayerBucketFillEvent(Level world, net.minecraft.world.entity.player.Player player, BlockPos changed, BlockPos clicked, Direction clickedFace, ItemStack itemInHand, net.minecraft.world.item.Item bucket, InteractionHand hand) {
+        // Kiterino start - Expose bucket in bucket events
+        return callPlayerBucketFillEvent(world, player, changed, clicked, clickedFace, itemInHand, new ItemStack(bucket), hand);
+    }
+    public static PlayerBucketFillEvent callPlayerBucketFillEvent(Level world, net.minecraft.world.entity.player.Player player, BlockPos changed, BlockPos clicked, Direction clickedFace, ItemStack itemInHand, ItemStack bucket, InteractionHand hand) {
+        // Kiterino end - Expose bucket in bucket events
         return (PlayerBucketFillEvent) CraftEventFactory.getPlayerBucketEvent(true, world, player, clicked, changed, clickedFace, itemInHand, bucket, hand);
     }
 
-    private static PlayerEvent getPlayerBucketEvent(boolean isFilling, Level world, net.minecraft.world.entity.player.Player player, BlockPos changed, BlockPos clicked, Direction clickedFace, ItemStack bucket, net.minecraft.world.item.Item item, InteractionHand hand) {
+    private static PlayerEvent getPlayerBucketEvent(boolean isFilling, Level world, net.minecraft.world.entity.player.Player player, BlockPos changed, BlockPos clicked, Direction clickedFace, ItemStack usedItem, ItemStack resultingItem, InteractionHand hand) { // Kiterino - Expose bucket in bucket events
         Player cplayer = (Player) player.getBukkitEntity();
-        CraftItemStack itemInHand = CraftItemStack.asNewCraftStack(item);
-        Material bucketItem = CraftItemType.minecraftToBukkit(bucket.getItem());
+//        CraftItemStack itemInHand = CraftItemStack.asNewCraftStack(item); // Kiterino - Expose bucket in bucket events
+//        Material bucketItem = CraftItemType.minecraftToBukkit(bucket.getItem()); // Kiterino - Expose bucket in bucket events
 
         Block block = CraftBlock.at(world, changed);
         Block clickedBlock = CraftBlock.at(world, clicked);
@@ -456,10 +461,10 @@ public class CraftEventFactory {
 
         PlayerEvent event;
         if (isFilling) {
-            event = new PlayerBucketFillEvent(cplayer, block, clickedBlock, blockFace, bucketItem, itemInHand, handSlot);
+            event = new PlayerBucketFillEvent(cplayer, block, clickedBlock, blockFace, CraftItemStack.asBukkitCopy(usedItem), CraftItemStack.asBukkitCopy(resultingItem), handSlot); // Kiterino - Expose bucket in bucket events
             ((PlayerBucketFillEvent) event).setCancelled(!CraftEventFactory.canBuild(world, cplayer, changed.getX(), changed.getZ()));
         } else {
-            event = new PlayerBucketEmptyEvent(cplayer, block, clickedBlock, blockFace, bucketItem, itemInHand, handSlot);
+            event = new PlayerBucketEmptyEvent(cplayer, block, clickedBlock, blockFace, CraftItemStack.asBukkitCopy(usedItem), CraftItemStack.asBukkitCopy(resultingItem), handSlot); // Kiterino - Expose bucket in bucket events
             ((PlayerBucketEmptyEvent) event).setCancelled(!CraftEventFactory.canBuild(world, cplayer, changed.getX(), changed.getZ()));
         }
 
