From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Fri, 9 Jun 2023 22:49:45 -0400
Subject: [PATCH] Paper PR - Player standing on position API

https://github.com/PaperMC/Paper/pull/9298

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 3d317f372367f157f6d295862184cab6c2837a1c..b2372a35736e9e9e0bc5a8b822ea61ee06fd8ce6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1437,6 +1437,20 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return this.entity.get(io.papermc.paper.datacomponent.PaperDataComponentType.bukkitToMinecraft(type)) != null;
     }
 
+    // Kiterino start - Paper PR - Player standing on position API
+    @Override
+    public org.bukkit.block.Block getMovementAffectingBlock() {
+        return CraftBlock.at(this.getHandle().level(), this.getHandle().getBlockPosBelowThatAffectsMyMovement());
+    }
+
+    @Override
+    public org.bukkit.block.Block getSupportingBlock() {
+        return this.getHandle().mainSupportingBlockPos
+                .map((pos) -> CraftBlock.at(this.getHandle().level(), pos))
+                .orElse(null);
+    }
+    // Kiterino end - Paper PR - Player standing on position API
+
     // Purpur start - Ridables
     @Override
     public org.bukkit.entity.Player getRider() {
