From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 16 Nov 2025 14:24:56 +0200
Subject: [PATCH] Add highly questionable BlockDispenseEvent additions


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 2013a91e4a86d23a6fcc095d6e3b6d08a9f7ac1c..496a0e6eedff0f1f31512d5012855663b79e8d55 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -2266,13 +2266,18 @@ public class CraftEventFactory {
         return event.callEvent();
     }
 
-    public static @Nullable ItemStack handleBlockDispenseEvent(net.minecraft.core.dispenser.BlockSource pointer, BlockPos to, ItemStack dispensed, net.minecraft.core.dispenser.DispenseItemBehavior instance) {
+    public static DispenseResult handleBlockDispenseEvent(net.minecraft.core.dispenser.BlockSource pointer, BlockPos to, ItemStack dispensed, net.minecraft.core.dispenser.DispenseItemBehavior instance, java.util.function.Consumer<org.bukkit.event.block.BlockDispenseEvent> extra) { // Kiterino - More BlockDispenseEvent API
         org.bukkit.block.Block bukkitBlock = CraftBlock.at(pointer.level(), pointer.pos());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(dispensed.isDamageableItem() ? dispensed : dispensed.copyWithCount(1));
 
         org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), CraftVector.toBukkit(to));
+        // Kiterino start - More BlockDispenseEvent API
+        if (extra != null) {
+            extra.accept(event);
+        }
+        // Kiterino end - More BlockDispenseEvent API
         if (!event.callEvent()) {
-            return dispensed;
+	        return new DispenseResult(dispensed, event); // Kiterino - More BlockDispenseEvent API
         }
 
         if (!event.getItem().equals(craftItem)) {
@@ -2281,11 +2286,13 @@ public class CraftEventFactory {
             net.minecraft.core.dispenser.DispenseItemBehavior itemBehavior = net.minecraft.world.level.block.DispenserBlock.getDispenseBehavior(pointer, eventStack);
             if (itemBehavior != net.minecraft.core.dispenser.DispenseItemBehavior.NOOP && itemBehavior != instance) {
                 itemBehavior.dispense(pointer, eventStack);
-                return dispensed;
+	            dispensed.shrink(1); // Kiterino - More BlockDispenseEvent API
+	            return new DispenseResult(dispensed, event); // Kiterino - More BlockDispenseEvent API
             }
         }
-        return null;
+        return new DispenseResult(null, event); // Kiterino - More BlockDispenseEvent API
     }
+	public record DispenseResult(@Nullable ItemStack dispensedItem, org.bukkit.event.block.BlockDispenseEvent event) {}
 
     /**
      * Calls the {@link io.papermc.paper.event.entity.EntityFertilizeEggEvent}.
