From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 16 Nov 2025 14:24:56 +0200
Subject: [PATCH] Add highly questionable BlockDispenseEvent additions


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index a6670f04bb93f71d95f5e63e7737dcb166af9ca3..960fd0c6d028db4ddb97dcfdd8fc51f535992ce3 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -2263,14 +2263,19 @@ public class CraftEventFactory {
         return event.callEvent();
     }
 
-    @Nullable
-    public static ItemStack handleBlockDispenseEvent(net.minecraft.core.dispenser.BlockSource pointer, BlockPos to, ItemStack itemStack, net.minecraft.core.dispenser.DispenseItemBehavior instance) {
+//    @Nullable // Kiterino - More BlockDispenseEvent API
+    public static DispenseResult handleBlockDispenseEvent(net.minecraft.core.dispenser.BlockSource pointer, BlockPos to, ItemStack itemStack, net.minecraft.core.dispenser.DispenseItemBehavior instance, java.util.function.Consumer<org.bukkit.event.block.BlockDispenseEvent> extra) { // Kiterino - More BlockDispenseEvent API
         org.bukkit.block.Block bukkitBlock = CraftBlock.at(pointer.level(), pointer.pos());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemStack.isDamageableItem() ? itemStack : itemStack.copyWithCount(1));
 
         org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), CraftVector.toBukkit(to));
+        // Kiterino start - More BlockDispenseEvent API
+        if (extra != null) {
+            extra.accept(event);
+        }
+        // Kiterino end - More BlockDispenseEvent API
         if (!event.callEvent()) {
-            return itemStack;
+            return new DispenseResult(itemStack, event); // Kiterino - More BlockDispenseEvent API
         }
 
         if (!event.getItem().equals(craftItem)) {
@@ -2279,11 +2284,13 @@ public class CraftEventFactory {
             net.minecraft.core.dispenser.DispenseItemBehavior itemBehavior = net.minecraft.world.level.block.DispenserBlock.getDispenseBehavior(pointer, eventStack);
             if (itemBehavior != net.minecraft.core.dispenser.DispenseItemBehavior.NOOP && itemBehavior != instance) {
                 itemBehavior.dispense(pointer, eventStack);
-                return itemStack;
+                itemStack.shrink(1); // Kiterino - More BlockDispenseEvent API
+                return new DispenseResult(itemStack, event); // Kiterino - More BlockDispenseEvent API
             }
         }
-        return null;
+        return new DispenseResult(null, event); // Kiterino - More BlockDispenseEvent API
     }
+    public record DispenseResult(@Nullable ItemStack dispensedItem, org.bukkit.event.block.BlockDispenseEvent event) {}
 
     /**
      * Calls the {@link io.papermc.paper.event.entity.EntityFertilizeEggEvent}.
