From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Fri, 9 Jun 2023 22:49:45 -0400
Subject: [PATCH] Paper PR - Player standing on position API

https://github.com/PaperMC/Paper/pull/9298

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 39f13ca7be830ec7af1a5b8b6cbe1ae33e92e1bb..929b215f3d56ad43a7f453b97945f836381388ed 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1453,6 +1453,20 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     }
     // Paper end - broadcast hurt animation
 
+    // Paper start - Player standing on position API
+    @Override
+    public org.bukkit.block.Block getMovementAffectingBlock() {
+        return CraftBlock.at(this.getHandle().level(), this.getHandle().getBlockPosBelowThatAffectsMyMovement());
+    }
+
+    @Override
+    public org.bukkit.block.Block getSupportingBlock() {
+        return this.getHandle().mainSupportingBlockPos
+                .map((pos) -> CraftBlock.at(this.getHandle().level(), pos))
+                .orElse(null);
+    }
+    // Paper end - Player standing on position API
+
     // Purpur start - Ridables
     @Override
     public org.bukkit.entity.Player getRider() {
