From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 18 Nov 2025 17:03:54 +0200
Subject: [PATCH] Custom fuel rules


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index ca516f93e16294a1a98529a2be0a1dbaced5344a..a7012912a3fc7a7537cbc7705b4ab2e7391f8508 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1516,6 +1516,13 @@ public final class CraftServer implements Server {
     }
     // Purpur end - Added the ability to add combustible items
 
+    // Kiterino start - Custom fuel rules
+    @Override
+    public void addFuel(java.util.function.Function<ItemStack, Integer> rule) {
+        MinecraftServer.getServer().fuelValues().extraRules.add(rule);
+    }
+    // Kiterino end - Custom fuel rules
+
     // Purpur start - Debug Marker API
     @Override
     public void sendBlockHighlight(Location location, int duration) {
