From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 18 Jan 2025 19:04:25 +0200
Subject: [PATCH] Rebrand


diff --git a/net/minecraft/CrashReport.java b/net/minecraft/CrashReport.java
index c239edddec50792b14b2c8163f18cb3dc0c05e63..4a169ef8ed6bff426d2bc56c45b7dba37abe870f 100644
--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -30,7 +30,7 @@ public class CrashReport {
     private boolean trackingStackTrace = true;
     private StackTraceElement[] uncategorizedStackTrace = new StackTraceElement[0];
     private final SystemReport systemReport = new SystemReport();
-    private List<String> extraInfo = List.of("", "DO NOT REPORT THIS TO PAPER! REPORT TO PURPUR INSTEAD!", ""); // Purpur - Rebrand
+    private List<String> extraInfo = List.of("", "DO NOT REPORT THIS TO PAPER/PURPUR! REPORT TO KITERINO INSTEAD!", ""); // Kiterino - Rebrand
 
     public CrashReport(final String title, final Throwable t) {
         this.title = title;
