From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 19 Jun 2024 01:40:24 +0300
Subject: [PATCH] Temporary solution for respawn events parity until Paper PR
 11792 is merged


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 2d3c0cf56426d549fdf52ba8e38aa881eb3bf32e..810d98ba885e345a3af8566ee09964842e17d43d 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -745,6 +745,7 @@ public abstract class PlayerList {
         }
         // Paper start - Add PlayerPostRespawnEvent
         boolean isBedSpawn = false;
+        boolean isAnchorSpawn = false; // Kiterino - Parity for respawn events
         boolean isRespawn = false;
         // Paper end - Add PlayerPostRespawnEvent
 
@@ -810,6 +811,7 @@ public abstract class PlayerList {
                 BlockPos blockPos = respawnConfig.pos();
                 BlockState blockState = level1.getBlockState(blockPos);
                 if (blockState.is(Blocks.RESPAWN_ANCHOR)) {
+                    isAnchorSpawn = true; // Kiterino - Parity for respawn events
                     serverPlayer.connection
                         .send(
                             new ClientboundSoundPacket(
@@ -854,7 +856,13 @@ public abstract class PlayerList {
 
         // Paper start - Add PlayerPostRespawnEvent
         if (isRespawn) {
-            cserver.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerPostRespawnEvent(player.getBukkitEntity(), location, isBedSpawn));
+            // Kiterino start - Parity for respawn events
+            var flags = com.google.common.collect.ImmutableSet.<org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag>builder();
+            if (eventReason == org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.END_PORTAL) {
+                flags.add(org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag.END_PORTAL);
+            }
+            cserver.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerPostRespawnEvent(player.getBukkitEntity(), location, isBedSpawn && !isAnchorSpawn, isAnchorSpawn, eventReason, flags));
+            // Kiterino end - Parity for respawn events
         }
         // Paper end - Add PlayerPostRespawnEvent
 
