From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 11 Oct 2022 14:28:39 +0300
Subject: [PATCH] Paper PR - Fire ServerListPingEvent for secondary motd send

https://github.com/PaperMC/Paper/pull/8074

diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 3ad9b41704349838cc9d672b4ae6ac1b3a7cfa7f..620851ae25ef3245e9dfa763345fbfaa2a5d9ca2 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -214,10 +214,15 @@ public abstract class PlayerList {
         mutableComponent.withStyle(ChatFormatting.YELLOW);
         Component joinMessage = mutableComponent; // Paper - Adventure
         serverGamePacketListenerImpl.teleport(player.getX(), player.getY(), player.getZ(), player.getYRot(), player.getXRot());
-        ServerStatus status = this.server.getStatus();
-        if (status != null && !cookie.transferred()) {
-            player.sendServerStatus(status);
-        }
+        // Paper start - fire ServerListPingEvent
+        if (!cookie.transferred()) {
+            io.papermc.paper.util.MCUtil.scheduleAsyncTask(() -> {
+                if (player.hasDisconnected()) return;
+                net.minecraft.network.protocol.status.ServerStatus status = com.destroystokyo.paper.network.StandardPaperServerListPingEventImpl.getEventResponse(this.server, player.connection.connection);
+                if (status != null) player.sendServerStatus(status);
+            });
+        }
+        // Paper end - fire ServerListPingEvent
 
         // player.connection.send(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(this.players)); // CraftBukkit - replaced with loop below
         this.players.add(player);
