From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 20 Sep 2024 08:20:11 +0300
Subject: [PATCH] Extra recipes API


diff --git a/net/minecraft/world/item/crafting/BlastingRecipe.java b/net/minecraft/world/item/crafting/BlastingRecipe.java
index c09b5fe3fa709129919c93f41b82879bcc464d10..64e40d27185a0351282cb94cd56940c84a5bbc5d 100644
--- a/net/minecraft/world/item/crafting/BlastingRecipe.java
+++ b/net/minecraft/world/item/crafting/BlastingRecipe.java
@@ -54,6 +54,7 @@ public class BlastingRecipe extends AbstractCookingRecipe {
         org.bukkit.craftbukkit.inventory.CraftBlastingRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftBlastingRecipe(id, result, org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.input()), this.experience(), this.cookingTime());
         recipe.setGroup(this.group());
         recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        recipe.setSpecial(isSpecial());// Kiterino - Extra recipes API
 
         return recipe;
     }
diff --git a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
index a7d8a869cdd47b9c4780eff400e7615396b21f93..9c15f66bbe8fa49ec81835f14eb4059d94420028 100644
--- a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
@@ -51,6 +51,7 @@ public class CampfireCookingRecipe extends AbstractCookingRecipe {
         org.bukkit.craftbukkit.inventory.CraftCampfireRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftCampfireRecipe(id, result, org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.input()), this.experience(), this.cookingTime());
         recipe.setGroup(this.group());
         recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        recipe.setSpecial(isSpecial());// Kiterino - Extra recipes API
 
         return recipe;
     }
diff --git a/net/minecraft/world/item/crafting/CustomRecipe.java b/net/minecraft/world/item/crafting/CustomRecipe.java
index af19d0ba60f094fc2a6a23ffbef2f9125ae58bc7..f14c5093f499365126d78c4af3e8cc154c469f73 100644
--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -35,6 +35,7 @@ public abstract class CustomRecipe implements CraftingRecipe {
         org.bukkit.craftbukkit.inventory.CraftComplexRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftComplexRecipe(id, org.bukkit.inventory.ItemStack.empty(), this);
         recipe.setGroup(this.group());
         recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        recipe.setSpecial(isSpecial());// Kiterino - Extra recipes API
 
         return recipe;
     }
diff --git a/net/minecraft/world/item/crafting/ShapedRecipe.java b/net/minecraft/world/item/crafting/ShapedRecipe.java
index 7ebac76135580738e7d115f925ca9a1556b9ba13..85beef8aaea7fb1d4cf192fba05704825a85d27e 100644
--- a/net/minecraft/world/item/crafting/ShapedRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapedRecipe.java
@@ -39,6 +39,7 @@ public class ShapedRecipe extends NormalCraftingRecipe {
     public static final RecipeSerializer<ShapedRecipe> SERIALIZER = new RecipeSerializer<>(MAP_CODEC, STREAM_CODEC);
     public final ShapedRecipePattern pattern;
     public final ItemStackTemplate result;
+    public boolean special = false; // Kiterino - Extra recipes API
 
     public ShapedRecipe(
         final Recipe.CommonInfo commonInfo, final CraftingRecipe.CraftingBookInfo bookInfo, final ShapedRecipePattern pattern, final ItemStackTemplate result
@@ -94,6 +95,13 @@ public class ShapedRecipe extends NormalCraftingRecipe {
         );
     }
 
+    // Kiterino start - Extra recipes API
+    @Override
+    public boolean isSpecial() {
+        return this.special;
+    }
+    // Kiterino end - Extra recipes API
+
     // CraftBukkit start
     @Override
     public org.bukkit.inventory.ShapedRecipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
@@ -101,6 +109,7 @@ public class ShapedRecipe extends NormalCraftingRecipe {
         org.bukkit.craftbukkit.inventory.CraftShapedRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftShapedRecipe(id, result, this);
         recipe.setGroup(this.group());
         recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        recipe.setSpecial(isSpecial()); // Kiterino - Extra recipes API
 
         switch (this.pattern.height()) {
             case 1:
diff --git a/net/minecraft/world/item/crafting/ShapelessRecipe.java b/net/minecraft/world/item/crafting/ShapelessRecipe.java
index fb7eaa2a9c4b49b494e33ec7066b7850aa581842..fbe4a4b861c14c325b14d05b22b7e89fd68c6a91 100644
--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -38,6 +38,7 @@ public class ShapelessRecipe extends NormalCraftingRecipe {
     public static final RecipeSerializer<ShapelessRecipe> SERIALIZER = new RecipeSerializer<>(MAP_CODEC, STREAM_CODEC);
     public final ItemStackTemplate result;
     public final List<Ingredient> ingredients;
+    public boolean special = false; // Kiterino - Extra recipes API
 
     public ShapelessRecipe(
         final Recipe.CommonInfo commonInfo, final CraftingRecipe.CraftingBookInfo bookInfo, final ItemStackTemplate result, final List<Ingredient> ingredients
@@ -47,6 +48,13 @@ public class ShapelessRecipe extends NormalCraftingRecipe {
         this.ingredients = ingredients;
     }
 
+    // Kiterino start - Extra recipes API
+    @Override
+    public boolean isSpecial() {
+        return this.special;
+    }
+    // Kiterino end - Extra recipes API
+
     // CraftBukkit start
     @Override
     public org.bukkit.inventory.ShapelessRecipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
@@ -54,6 +62,7 @@ public class ShapelessRecipe extends NormalCraftingRecipe {
         org.bukkit.craftbukkit.inventory.CraftShapelessRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftShapelessRecipe(id, result, this);
         recipe.setGroup(this.group());
         recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        recipe.setSpecial(isSpecial()); // Kiterino - Extra recipes API
 
         for (Ingredient ingredient : this.ingredients) {
             recipe.addIngredient(org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(ingredient));
diff --git a/net/minecraft/world/item/crafting/SingleItemRecipe.java b/net/minecraft/world/item/crafting/SingleItemRecipe.java
index 443ab1e86cb8f551d224f36cd766dd59a8133201..77b8141561d0eb8b810dd06462dc283ab63be76d 100644
--- a/net/minecraft/world/item/crafting/SingleItemRecipe.java
+++ b/net/minecraft/world/item/crafting/SingleItemRecipe.java
@@ -14,6 +14,14 @@ public abstract class SingleItemRecipe implements Recipe<SingleRecipeInput> {
     private final Ingredient input;
     public final ItemStackTemplate result;
     private @Nullable PlacementInfo placementInfo;
+    // Kiterino start - Extra recipes API
+    public boolean special;
+
+    @Override
+    public boolean isSpecial() {
+        return this.special;
+    }
+    // Kiterino end - Extra recipes API
 
     public SingleItemRecipe(final Recipe.CommonInfo commonInfo, final Ingredient input, final ItemStackTemplate result) {
         this.commonInfo = commonInfo;
diff --git a/net/minecraft/world/item/crafting/SmeltingRecipe.java b/net/minecraft/world/item/crafting/SmeltingRecipe.java
index 6608fe9655e3dbe75bb6cedfda37ff7a86699795..732ded89f16b6fd27b0fefb019b82026bbde23ef 100644
--- a/net/minecraft/world/item/crafting/SmeltingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmeltingRecipe.java
@@ -55,6 +55,7 @@ public class SmeltingRecipe extends AbstractCookingRecipe {
         org.bukkit.craftbukkit.inventory.CraftFurnaceRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftFurnaceRecipe(id, result, org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.input()), this.experience(), this.cookingTime());
         recipe.setGroup(this.group());
         recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        recipe.setSpecial(isSpecial());// Kiterino - Extra recipes API
 
         return recipe;
     }
diff --git a/net/minecraft/world/item/crafting/SmokingRecipe.java b/net/minecraft/world/item/crafting/SmokingRecipe.java
index 40d72718798a577edcfec21dfce43dabdd81bd17..271c8694fd4eb4bf6550913bd3feeb03a186c2fe 100644
--- a/net/minecraft/world/item/crafting/SmokingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmokingRecipe.java
@@ -51,6 +51,7 @@ public class SmokingRecipe extends AbstractCookingRecipe {
         org.bukkit.craftbukkit.inventory.CraftSmokingRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftSmokingRecipe(id, result, org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.input()), this.experience(), this.cookingTime());
         recipe.setGroup(this.group());
         recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        recipe.setSpecial(isSpecial());// Kiterino - Extra recipes API
 
         return recipe;
     }
