From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Jun 2022 00:12:32 +0300
Subject: [PATCH] Add EntityLoadsProjectileEvent


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index c737cc4a1edcc720a756fa2c1f4124ccc2cf2949..9bb87a254074165456ad17263932df41bf05dc73 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4932,6 +4932,23 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return ItemStack.EMPTY;
     }
 
+    // Kiterino start - Add EntityLoadsProjectileEvent
+    public ItemStack getProjectile(ItemStack weapon, ItemStack projectile, Predicate<ItemStack> projectileValidator) {
+        var event = new me.sosedik.kiterino.event.entity.EntityLoadsProjectileEvent(
+            getBukkitLivingEntity(),
+            weapon.asBukkitMirror(),
+            projectile.asBukkitMirror(),
+            bukkitItem -> projectileValidator.test(CraftItemStack.asNMSCopy(bukkitItem))
+        );
+        event.callEvent();
+        var item = event.getProjectile();
+        if (item.isEmpty()) {
+            return ItemStack.EMPTY;
+        }
+        return item instanceof CraftItemStack craftItem && craftItem.handle != null ? craftItem.handle : CraftItemStack.asNMSCopy(item);
+    }
+    // Kiterino end - Add EntityLoadsProjectileEvent
+
     public static byte entityEventForEquipmentBreak(final EquipmentSlot equipmentSlot) {
         return switch (equipmentSlot) {
             case MAINHAND -> EntityEvent.MAINHAND_BREAK;
diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index 81f1d4faebc38aa07a2c76c2c163584c8e5b624c..24792d5a47361e90bda09cd14b244fced9d95199 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -157,7 +157,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
         if (heldWeapon.getItem() instanceof ProjectileWeaponItem) {
             Predicate<ItemStack> supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getSupportedHeldProjectiles();
             ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedProjectiles);
-            return heldProjectile.isEmpty() ? new ItemStack(Items.ARROW) : heldProjectile;
+            return getProjectile(heldWeapon, heldProjectile.isEmpty() ? new ItemStack(Items.ARROW) : heldProjectile, supportedProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
         } else {
             return ItemStack.EMPTY;
         }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 309fba8b0e8337cd53c0bde7e60e7a86771d9f00..50c69537e3b867563abf7aa563d0b35255bba301 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -2127,7 +2127,7 @@ public abstract class Player extends Avatar implements ContainerUser {
         Predicate<ItemStack> supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getSupportedHeldProjectiles().and(item -> this.tryReadyArrow(heldWeapon, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
         ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedProjectiles);
         if (!heldProjectile.isEmpty()) {
-            return heldProjectile;
+	        return getProjectile(heldWeapon, heldProjectile, supportedProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
         }
 
         supportedProjectiles = ((ProjectileWeaponItem)heldWeapon.getItem()).getAllSupportedProjectiles().and(item -> this.tryReadyArrow(heldWeapon, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
@@ -2135,12 +2135,12 @@ public abstract class Player extends Avatar implements ContainerUser {
         for (int i = 0; i < this.inventory.getContainerSize(); i++) {
             ItemStack itemStack = this.inventory.getItem(i);
             if (supportedProjectiles.test(itemStack)) {
-                return itemStack;
+	            return getProjectile(heldProjectile, itemStack, supportedProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
             }
         }
 
         if (anyEventCancelled.booleanValue() && !this.abilities.instabuild && this instanceof final ServerPlayer player) this.resyncUsingItem(player); // Paper - resync if no item matched the Predicate
-        return this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+	    return getProjectile(heldWeapon, this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY, supportedProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
     }
 
     @Override
