From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Jun 2022 00:12:32 +0300
Subject: [PATCH] Add EntityLoadsProjectileEvent


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 4ff84692186a250ba514ea79b4d0f82626ed9e6b..8e0600962fb20a38c349982fb72604a53e8a093e 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4793,6 +4793,23 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return ItemStack.EMPTY;
     }
 
+    // Kiterino start - Add EntityLoadsProjectileEvent
+    public ItemStack getProjectile(ItemStack weapon, ItemStack projectile, Predicate<ItemStack> projectileValidator) {
+        var event = new me.sosedik.kiterino.event.entity.EntityLoadsProjectileEvent(
+            getBukkitLivingEntity(),
+            weapon.asBukkitMirror(),
+            projectile.asBukkitMirror(),
+            bukkitItem -> projectileValidator.test(CraftItemStack.asNMSCopy(bukkitItem))
+        );
+        event.callEvent();
+        var item = event.getProjectile();
+        if (item.isEmpty()) {
+            return ItemStack.EMPTY;
+        }
+        return item instanceof CraftItemStack craftItem && craftItem.handle != null ? craftItem.handle : CraftItemStack.asNMSCopy(item);
+    }
+    // Kiterino end - Add EntityLoadsProjectileEvent
+
     public static byte entityEventForEquipmentBreak(EquipmentSlot slot) {
         return switch (slot) {
             case MAINHAND -> EntityEvent.MAINHAND_BREAK;
diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index 54033572e325f8486a438e31dde34f8407f9180a..6f844935432f8c512d16a71f6fa6151df1cfee5c 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -151,7 +151,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
         if (shootable.getItem() instanceof ProjectileWeaponItem) {
             Predicate<ItemStack> supportedHeldProjectiles = ((ProjectileWeaponItem)shootable.getItem()).getSupportedHeldProjectiles();
             ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedHeldProjectiles);
-            return heldProjectile.isEmpty() ? new ItemStack(Items.ARROW) : heldProjectile;
+            return getProjectile(shootable, heldProjectile.isEmpty() ? new ItemStack(Items.ARROW) : heldProjectile, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
         } else {
             return ItemStack.EMPTY;
         }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 6f2819f76fe9ffed838b6b6fdc426a834ea9129d..74110abb9687f1d7df3fe9ac95173d87ca553e9a 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -2117,19 +2117,19 @@ public abstract class Player extends Avatar implements ContainerUser {
             Predicate<ItemStack> supportedHeldProjectiles = ((ProjectileWeaponItem)shootable.getItem()).getSupportedHeldProjectiles().and(item -> this.tryReadyArrow(shootable, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
             ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedHeldProjectiles);
             if (!heldProjectile.isEmpty()) {
-                return heldProjectile;
+                return getProjectile(shootable, heldProjectile, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
             } else {
                 supportedHeldProjectiles = ((ProjectileWeaponItem)shootable.getItem()).getAllSupportedProjectiles().and(item -> this.tryReadyArrow(shootable, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
 
                 for (int i = 0; i < this.inventory.getContainerSize(); i++) {
                     ItemStack item = this.inventory.getItem(i);
                     if (supportedHeldProjectiles.test(item)) {
-                        return item;
+                        return getProjectile(shootable, item, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
                     }
                 }
 
                 if (anyEventCancelled.booleanValue() && !this.abilities.instabuild && this instanceof final ServerPlayer player) this.resyncUsingItem(player); // Paper - resync if no item matched the Predicate
-                return this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+                return getProjectile(shootable, this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
             }
         }
     }
