From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 22 Jun 2024 23:47:13 +0300
Subject: [PATCH] Add EntityStartUsingItemEvent


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 5913beef80612067b4fdb4afb8f124fa89a18082..676763af882416e0a3844e1b4b810404ff663652 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4384,8 +4384,16 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         ItemStack itemStack = this.getItemInHand(hand);
         if ((!itemStack.isEmpty() && !this.isUsingItem()) || forceUpdate) { // Paper - Prevent consuming the wrong itemstack
             this.useItem = itemStack;
+            // Kiterino start - Add EntityStartUsingItemEvent
+            var event = new me.sosedik.kiterino.event.entity.EntityStartUsingItemEvent(getBukkitLivingEntity(), itemStack.asBukkitMirror(), itemStack.getUseDuration(this));
+            if (!event.callEvent()) {
+                stopUsingItem();
+                return;
+            }
+            int useDuration = event.getUseDuration();
+            // Kiterino end - Add EntityStartUsingItemEvent
             // Paper start - lag compensate eating
-            this.useItemRemaining = this.totalEatTimeTicks = itemStack.getUseDuration(this);
+            this.useItemRemaining = this.totalEatTimeTicks = useDuration; // Kiterino - Add EntityStartUsingItemEvent
             this.eatStartTime = System.nanoTime();
             // Paper end - lag compensate eating
             if (!this.level().isClientSide()) {
