From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Jun 2022 00:12:32 +0300
Subject: [PATCH] Add EntityLoadsProjectileEvent


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 4c4ffdf7721c8f2c3be607aeacf2ca102954cff0..ad6bd5251ca73f01d53817c4c6100d9a1211cb9f 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4589,6 +4589,23 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return ItemStack.EMPTY;
     }
 
+    // Kiterino start - Add EntityLoadsProjectileEvent
+    public ItemStack getProjectile(ItemStack weapon, ItemStack projectile, Predicate<ItemStack> projectileValidator) {
+        var event = new me.sosedik.kiterino.event.entity.EntityLoadsProjectileEvent(
+                getBukkitLivingEntity(),
+                weapon.asBukkitMirror(),
+                projectile.asBukkitMirror(),
+                bukkitItem -> projectileValidator.test(CraftItemStack.asNMSCopy(bukkitItem))
+        );
+        event.callEvent();
+        var item = event.getProjectile();
+        if (item.isEmpty()) {
+            return ItemStack.EMPTY;
+        }
+        return item instanceof CraftItemStack craftItem && craftItem.handle != null ? craftItem.handle : CraftItemStack.asNMSCopy(item);
+    }
+    // Kiterino end - Add EntityLoadsProjectileEvent
+
     public static byte entityEventForEquipmentBreak(EquipmentSlot slot) {
         return switch (slot) {
             case MAINHAND -> 47;
diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index 04d5910d736dee2a88a2602f4a98495459277157..9251ff10f30d833f7b2ee6c8d771aab87e5c1443 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -147,7 +147,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
         if (shootable.getItem() instanceof ProjectileWeaponItem) {
             Predicate<ItemStack> supportedHeldProjectiles = ((ProjectileWeaponItem)shootable.getItem()).getSupportedHeldProjectiles();
             ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedHeldProjectiles);
-            return heldProjectile.isEmpty() ? new ItemStack(Items.ARROW) : heldProjectile;
+            return getProjectile(shootable, heldProjectile.isEmpty() ? new ItemStack(Items.ARROW) : heldProjectile, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
         } else {
             return ItemStack.EMPTY;
         }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 18c8538f5ac8e35f62f7f2e266fd63a4f45849d6..0f4bf06b8b770e1dcacd10c4f44495a31eb3d3bf 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -2213,19 +2213,19 @@ public abstract class Player extends LivingEntity {
             Predicate<ItemStack> supportedHeldProjectiles = ((ProjectileWeaponItem)shootable.getItem()).getSupportedHeldProjectiles().and(item -> this.tryReadyArrow(shootable, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
             ItemStack heldProjectile = ProjectileWeaponItem.getHeldProjectile(this, supportedHeldProjectiles);
             if (!heldProjectile.isEmpty()) {
-                return heldProjectile;
+                return getProjectile(shootable, heldProjectile, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
             } else {
                 supportedHeldProjectiles = ((ProjectileWeaponItem)shootable.getItem()).getAllSupportedProjectiles().and(item -> this.tryReadyArrow(shootable, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
 
                 for (int i = 0; i < this.inventory.getContainerSize(); i++) {
                     ItemStack item = this.inventory.getItem(i);
                     if (supportedHeldProjectiles.test(item)) {
-                        return item;
+                        return getProjectile(shootable, item, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
                     }
                 }
 
                 if (anyEventCancelled.booleanValue() && !this.abilities.instabuild && this instanceof final ServerPlayer player) this.resyncUsingItem(player); // Paper - resync if no item matched the Predicate
-                return this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+                return getProjectile(shootable, this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY, supportedHeldProjectiles); // Kiterino - Add EntityLoadsProjectileEvent
             }
         }
     }
