From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 22 Jun 2024 23:47:13 +0300
Subject: [PATCH] Add EntityStartUsingItemEvent


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index eff4a51ac1817489b1d590327865dcbc9e846773..e997556eb0c6934e409f682dca30d977416d5a7a 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4055,8 +4055,16 @@ public abstract class LivingEntity extends Entity implements Attackable {
         ItemStack itemInHand = this.getItemInHand(hand);
         if (!itemInHand.isEmpty() && !this.isUsingItem() || forceUpdate) { // Paper - Prevent consuming the wrong itemstack
             this.useItem = itemInHand;
+            // Kiterino start - Add EntityStartUsingItemEvent
+            var event = new me.sosedik.kiterino.event.entity.EntityStartUsingItemEvent(getBukkitLivingEntity(), itemInHand.asBukkitMirror(), itemInHand.getUseDuration(this));
+            if (!event.callEvent()) {
+                stopUsingItem();
+                return;
+            }
+            int useDuration = event.getUseDuration();
+            // Kiterino end - Add EntityStartUsingItemEvent
             // Paper start - lag compensate eating
-            this.useItemRemaining = this.totalEatTimeTicks = itemInHand.getUseDuration(this);
+            this.useItemRemaining = this.totalEatTimeTicks = useDuration; // Kiterino - Add EntityStartUsingItemEvent
             this.eatStartTime = System.nanoTime();
             // Paper end - lag compensate eating
             if (!this.level().isClientSide) {
