From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Isaac - The456 <the456@the456gamer.dev>
Date: Sun, 8 Sep 2024 23:57:54 +0100
Subject: [PATCH] Paper PR - Add Improved Lidded API

https://github.com/PaperMC/Paper/pull/11814

diff --git a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
index 52737172ccfa0c7c977e4f2fe1db242e0bb92d25..39502867a2e76580cd674570aacf5aabead53812 100644
--- a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -16,20 +16,7 @@ public abstract class ContainerOpenersCounter {
     private static final int CHECK_TICK_DELAY = 5;
     private int openCount;
     private double maxInteractionRange;
-    // CraftBukkit start
-    public boolean opened;
-    public void onOpenAPI(Level level, BlockPos pos, BlockState blockState) {
-        this.onOpen(level, pos, blockState);
-    }
-
-    public void onCloseAPI(Level level, BlockPos pos, BlockState blockState) {
-        this.onClose(level, pos, blockState);
-    }
-
-    public void openerCountChangedAPI(Level level, BlockPos pos, BlockState blockState, int previous, int current) {
-        this.openerCountChanged(level, pos, blockState, previous, current);
-    }
-    // CraftBukkit end
+    // public boolean opened; // Kiterino - Paper PR - Replace with new Lidded API
 
     protected abstract void onOpen(final Level level, final BlockPos pos, final BlockState blockState);
 
@@ -37,16 +24,67 @@ public abstract class ContainerOpenersCounter {
 
     protected abstract void openerCountChanged(final Level level, final BlockPos pos, final BlockState blockState, int previous, int current);
 
+    // Kiterino start - Paper PR - Add Improved Lidded API
+    public abstract boolean isOwnContainer(Player player);
+
+    private io.papermc.paper.block.LidMode apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    private final java.util.Set<Player> cancelledPlayers = new java.util.HashSet<>(); // Kiterino - Paper PR - Store players whose opening was cancelled by PlayerLiddedOpenEvent
+    public void startForceLiddedLidOpen(Level level, BlockPos pos, BlockState state) {
+        incrementOpeners(null, level, pos, state, 16);
+    }
+
+    public void stopForceLiddedLidOpen(Level level, BlockPos pos, BlockState state) {
+        decrementOpeners(null, level, pos, state);
+        apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    }
+
+    public void startForceLiddedLidClose(Level level, BlockPos pos, BlockState state) {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.onClose(level, pos, state);
+            level.gameEvent(null, GameEvent.CONTAINER_CLOSE, pos);
+        }
+        this.openerCountChanged(level, pos, state, this.openCount, 0);
+    }
+    public void stopForceLiddedLidClose(Level level, BlockPos pos, BlockState state) {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.onOpen(level, pos, state);
+            level.gameEvent(null, GameEvent.CONTAINER_OPEN, pos);
+            scheduleRecheck(level, pos, state);
+        }
+        this.openerCountChanged(level, pos, state, 0, this.openCount);
+        apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    }
+    public io.papermc.paper.block.LidMode getLidMode() {
+        return apiLidMode;
+    }
+    public void setLidMode(final io.papermc.paper.block.LidMode targetLidMode) {
+        apiLidMode = targetLidMode;
+    }
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        return switch (apiLidMode) {
+            case OPEN_UNTIL_VIEWED, FORCED_OPEN -> io.papermc.paper.block.LidState.OPEN;
+            case CLOSED_UNTIL_NOT_VIEWED, FORCED_CLOSED -> io.papermc.paper.block.LidState.CLOSED;
+            default -> getTrueLidState();
+        };
+    }
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        boolean virtualViewerPresent = (apiLidMode == io.papermc.paper.block.LidMode.FORCED_OPEN || apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED);
+        int trueOpenCount = this.openCount - (virtualViewerPresent ? 1 : 0);
+        if (trueOpenCount < 0) {
+            throw new IllegalStateException("trueOpenCount is negative: " + trueOpenCount + " openCount: " + openCount + " virtualViewerPresent: " + virtualViewerPresent);
+        }
+        return trueOpenCount > 0 ? io.papermc.paper.block.LidState.OPEN : io.papermc.paper.block.LidState.CLOSED;
+    }
+    // Kiterino end - Paper PR - Add Improved Lidded API
+
     // Paper start - delay open/close callbacks
     public boolean delayCallbacks() {
         return false;
     }
     // Paper end - delay open/close callbacks
 
-    public abstract boolean isOwnContainer(final Player player);
-
     public void incrementOpeners(
-        final LivingEntity entity, final Level level, final BlockPos pos, final BlockState blockState, final double maxInteractionRange
+        final @javax.annotation.Nullable LivingEntity entity, final Level level, final BlockPos pos, final BlockState blockState, final double maxInteractionRange // Kiterino - Paper PR - Make entity nullable for New Lidded API
     ) {
         // Paper start - delay open/close callbacks
         if (this.delayCallbacks()) {
@@ -54,8 +92,20 @@ public abstract class ContainerOpenersCounter {
             return;
         }
         // Paper end - delay open/close callbacks
+        // Paper start - Call PlayerLiddedOpenEvent
+        if (entity instanceof net.minecraft.server.level.ServerPlayer player && !org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerLiddedOpenEvent(player, level, pos)) {
+            this.cancelledPlayers.add(player);
+            return;
+        }
+        // Paper end - Call PlayerLiddedOpenEvent
+        if (this.openCount == 0 && this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+            this.apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+            stopForceLiddedLidClose(level, pos, blockState);
+        }
+        // Kiterino end - Paper PR - Add Improved Lidded Api
         int previous = this.openCount++;
 
+        if (this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) return; // Kiterino - Paper PR - Add improved Lidded API
         // Paper start - Call BlockRedstoneEvent
         if (blockState.is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
             int oldCurrent = Math.clamp(previous, 0, 15);
@@ -74,10 +124,18 @@ public abstract class ContainerOpenersCounter {
         }
 
         this.openerCountChanged(level, pos, blockState, previous, this.openCount);
+        if (entity != null) // Kiterino - Paper PR - Make player nullable for improved Lidded API
         this.maxInteractionRange = Math.max(maxInteractionRange, this.maxInteractionRange);
+        // Kiterino start - Paper PR - Add Improved Lidded API
+        if (entity != null && this.apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED) {
+            // reset to default
+            this.apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+            stopForceLiddedLidOpen(level, pos, blockState);
+        }
+        // Kiterino end - Paper PR - Add Improved Lidded API
     }
 
-    public void decrementOpeners(final LivingEntity entity, final Level level, final BlockPos pos, final BlockState blockState) {
+    public void decrementOpeners(final @javax.annotation.Nullable LivingEntity entity, final Level level, final BlockPos pos, final BlockState blockState) { // Kiterino - Paper PR - Make player nullable for New Lidded API
         // Paper start - delay open/close callbacks
         if (this.delayCallbacks()) {
             scheduleRecheck(level, pos, blockState, 1);
@@ -87,6 +145,16 @@ public abstract class ContainerOpenersCounter {
         if (this.openCount == 0) return; // Paper - Prevent ContainerOpenersCounter openCount from going negative
         int previous = this.openCount--;
 
+        // Kiterino start - Paper PR - Add Improved Lidded Api
+        if (this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+            if (this.openCount == 0 && this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                this.apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                stopForceLiddedLidClose(level, pos, blockState);
+            }
+            return;
+        }
+        // Kiterino end - Paper PR - Add Improved Lidded Api
+
         // Paper start - Call BlockRedstoneEvent
         if (blockState.is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
             int oldCurrent = Math.clamp(previous, 0, 15);
@@ -126,13 +194,22 @@ public abstract class ContainerOpenersCounter {
         List<ContainerUser> containerUsers = this.getEntitiesWithContainerOpen(level, pos);
         this.maxInteractionRange = 0.0;
 
+        // Kiterino start - Paper PR - Maintain cancelledPlayers, list of players with the chest open, but without the lid.
+        this.cancelledPlayers.removeIf(java.util.function.Predicate.not(containerUsers::contains));
+        containerUsers.removeIf(this.cancelledPlayers::contains);
+        // Kiterino end - Paper PR - Maintain cancelledPlayers, list of players with the chest open, but without the lid.
+
         for (ContainerUser containerUser : containerUsers) {
             this.maxInteractionRange = Math.max(containerUser.getContainerInteractionRange(), this.maxInteractionRange);
         }
 
-        int openCount = containerUsers.size();
-        if (this.opened) openCount++; // CraftBukkit - add dummy count from API
-        int prevCount = this.openCount;
+        // Paper Start - Replace with Add Improved Lidded API
+        boolean forceClosed = this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED || this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED;
+        boolean forceOpened = this.apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED || this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_OPEN;
+        int openCount = forceClosed ? 0 : containerUsers.size() + (forceOpened ? 1 : 0);
+        // if (this.opened) openCount++; // CraftBukkit - add dummy count from API
+        int prevCount = forceClosed ? 0 : this.openCount;
+        // Paper End - Replace with Add Improved Lidded API
         if (prevCount != openCount) {
             // Paper start - Call BlockRedstoneEvent
             if (blockState.is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
diff --git a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
index d1a2e671509b51f0bfad1633b1cdc2be671fca81..bb0fe2c93e6c5b16d36bfb8106fa4fb657ca0cbc 100644
--- a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
@@ -52,7 +52,7 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     // CraftBukkit start - add fields and methods
     public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
     private int maxStack = MAX_STACK;
-    public boolean opened;
+    // public boolean opened; // Kiterino - Paper PR - Replace with new Lidded API
 
     @Override
     public List<ItemStack> getContents() {
@@ -175,7 +175,7 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     @Override
     public boolean triggerEvent(final int b0, final int b1) {
         if (b0 == EVENT_SET_OPEN_COUNT) {
-            this.openCount = b1;
+            if (this.apiLidMode != io.papermc.paper.block.LidMode.FORCED_CLOSED && this.apiLidMode != io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) this.openCount = b1; // Kiterino - Paper PR - Skip Mutate when forced closed by lidded api
             if (b1 == 0) {
                 this.animationStatus = ShulkerBoxBlockEntity.AnimationStatus.CLOSING;
             }
@@ -199,6 +199,64 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     public void preRemoveSideEffects(final BlockPos pos, final BlockState state) {
     }
 
+    // Kiterino start - Paper PR - Add Improved Lidded API
+    private io.papermc.paper.block.LidMode apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    private final java.util.Set<ContainerUser> cancelledPlayers = new java.util.HashSet<>(); // Kiterino - Paper PR - Store players whose opening was cancelled by PlayerLiddedOpenEvent
+    public void startForceLiddedLidOpen() {
+        this.openCount++;
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
+        if (this.openCount == 1) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_OPEN, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, this.level.getRandom().nextFloat() * 0.1F + 0.9F);
+        }
+    }
+    public void stopForceLiddedLidOpen() {
+        this.openCount--;
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
+        if (this.openCount <= 0) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_CLOSE, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_CLOSE, SoundSource.BLOCKS, 0.5F, this.level.getRandom().nextFloat() * 0.1F + 0.9F);
+        }
+        apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    }
+    public void startForceLiddedLidClose() {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_CLOSE, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_CLOSE, SoundSource.BLOCKS, 0.5F, this.level.getRandom().nextFloat() * 0.1F + 0.9F);
+        }
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, 0);
+    }
+    public void stopForceLiddedLidClose() {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_OPEN, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, this.level.getRandom().nextFloat() * 0.1F + 0.9F);
+        }
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
+    }
+    public io.papermc.paper.block.LidMode getLidMode() {
+        return apiLidMode;
+    }
+    public void setLidMode(final io.papermc.paper.block.LidMode lidMode) {
+        this.apiLidMode = lidMode;
+    }
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        return switch (apiLidMode) {
+            case OPEN_UNTIL_VIEWED, FORCED_OPEN -> io.papermc.paper.block.LidState.OPEN;
+            case CLOSED_UNTIL_NOT_VIEWED, FORCED_CLOSED -> io.papermc.paper.block.LidState.CLOSED;
+            default -> getTrueLidState();
+        };
+    }
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        boolean virtualViewerPresent = (apiLidMode == io.papermc.paper.block.LidMode.FORCED_OPEN || apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED);
+        int trueOpenCount = this.openCount - (virtualViewerPresent ? 1 : 0);
+        // ensure trueOpenCount is never negative, throw
+        if (trueOpenCount < 0) {
+            throw new IllegalStateException("trueOpenCount is negative: " + trueOpenCount + " openCount: " + openCount + " virtualViewerPresent: " + virtualViewerPresent);
+        }
+        return trueOpenCount > 0 ? io.papermc.paper.block.LidState.OPEN : io.papermc.paper.block.LidState.CLOSED;
+    }
+    // Kiterino end - Paper PR - Add Improved Lidded API
+
     @Override
     public void startOpen(final ContainerUser containerUser) {
         if (!this.remove && !containerUser.getLivingEntity().isSpectator()) {
@@ -206,8 +264,26 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
                 this.openCount = 0;
             }
 
+            // Paper start - Call PlayerLiddedOpenEvent
+            if (containerUser instanceof net.minecraft.server.level.ServerPlayer player && !org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerLiddedOpenEvent(player, this.level, this.worldPosition)) {
+                cancelledPlayers.add(player);
+                return;
+            }
+            // Paper end - Call PlayerLiddedOpenEvent
+            // Kiterino start - Paper PR - Add Improved Lidded API
+            if (this.openCount == 0) {
+                if (apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                    apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                    stopForceLiddedLidClose();
+                }
+            }
+            // Kiterino end - Paper PR - Add Improved Lidded API
+
             this.openCount++;
-            if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call
+            // Kiterino start - Paper PR - Replace with Improved Lidded API
+            //if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
+            if (this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) return;
+            // Kiterino end - Paper PR - Replace with Improved Lidded API
             this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), EVENT_SET_OPEN_COUNT, this.openCount);
             if (this.openCount == 1) {
                 this.level.gameEvent(containerUser.getLivingEntity(), GameEvent.CONTAINER_OPEN, this.worldPosition);
@@ -216,14 +292,32 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
                         null, this.worldPosition, SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, this.level.getRandom().nextFloat() * 0.1F + 0.9F
                     );
             }
+            // Kiterino start - Paper PR - Add Improved Lidded API
+            if (this.apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED) {
+                // reset to default
+                this.apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                stopForceLiddedLidOpen();
+            }
+            // Kiterino end - Paper PR - Add Improved Lidded API
         }
     }
 
     @Override
     public void stopOpen(final ContainerUser containerUser) {
         if (!this.remove && !containerUser.getLivingEntity().isSpectator()) {
+            if (this.cancelledPlayers.remove(containerUser)) return; // Kiterino - Paper PR - Do not decrement if player's opening was cancelled by PlayerLiddedOpenEvent
             this.openCount--;
-            if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
+            // Kiterino start - Paper PR - Add Improved Lidded API
+            // if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
+            if (this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                if (this.openCount <= 0 && this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                    this.openCount = 0;
+                    this.apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                    this.stopForceLiddedLidClose();
+                }
+                return;
+            }
+            // Kiterino end - Paper PR - Add Improved Lidded API
             this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), EVENT_SET_OPEN_COUNT, this.openCount);
             if (this.openCount <= 0) {
                 this.level.gameEvent(containerUser.getLivingEntity(), GameEvent.CONTAINER_CLOSE, this.worldPosition);
