From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Isaac - The456 <the456@the456gamer.dev>
Date: Sun, 8 Sep 2024 23:57:54 +0100
Subject: [PATCH] Paper PR - Add Improved Lidded API

https://github.com/PaperMC/Paper/pull/11814

diff --git a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
index 4f3b6a32fe4e791aab05aa9e553ec935e3cd5e6b..a5f00d7fa03654024b059b2ab69ae478fbbc4157 100644
--- a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -16,20 +16,57 @@ public abstract class ContainerOpenersCounter {
     private static final int CHECK_TICK_DELAY = 5;
     private int openCount;
     private double maxInteractionRange;
-    // CraftBukkit start
-    public boolean opened;
-    public void onAPIOpen(Level level, BlockPos blockPos, BlockState blockState) {
-        this.onOpen(level, blockPos, blockState);
-    }
+    // Kiterino start - Paper PR - Add Improved Lidded API
+    // public boolean opened; // Kiterino - Paper PR - Replace with new Lidded API
+    public abstract boolean isOwnContainer(Player player);
 
-    public void onAPIClose(Level level, BlockPos blockPos, BlockState blockState) {
-        this.onClose(level, blockPos, blockState);
+    private io.papermc.paper.block.LidMode apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    private final java.util.Set<Player> cancelledPlayers = new java.util.HashSet<>(); // Kiterino - Paper PR - Store players whose opening was cancelled by PlayerLiddedOpenEvent
+    public void startForceLiddedLidOpen(Level level, BlockPos pos, BlockState state) {
+        incrementOpeners(null, level, pos, state, 16);
     }
 
-    public void openerAPICountChanged(Level level, BlockPos blockPos, BlockState blockState, int count, int openCount) {
-        this.openerCountChanged(level, blockPos, blockState, count, openCount);
+    public void stopForceLiddedLidOpen(Level level, BlockPos pos, BlockState state) {
+        decrementOpeners(null, level, pos, state);
+        apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    }
+    public void startForceLiddedLidClose(Level level, BlockPos pos, BlockState state) {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.onClose(level, pos, state);
+            level.gameEvent(null, GameEvent.CONTAINER_CLOSE, pos);
+        }
+        this.openerCountChanged(level, pos, state, this.openCount, 0);
+    }
+    public void stopForceLiddedLidClose(Level level, BlockPos pos, BlockState state) {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.onOpen(level, pos, state);
+            level.gameEvent(null, GameEvent.CONTAINER_OPEN, pos);
+            scheduleRecheck(level, pos, state);
+        }
+        this.openerCountChanged(level, pos, state, 0, this.openCount);
+        apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    }
+    public io.papermc.paper.block.LidMode getLidMode() {
+        return apiLidMode;
+    }
+    public void setLidMode(final io.papermc.paper.block.LidMode targetLidMode) {
+        apiLidMode = targetLidMode;
+    }
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        return switch (apiLidMode) {
+            case OPEN_UNTIL_VIEWED, FORCED_OPEN -> io.papermc.paper.block.LidState.OPEN;
+            case CLOSED_UNTIL_NOT_VIEWED, FORCED_CLOSED -> io.papermc.paper.block.LidState.CLOSED;
+            default -> getTrueLidState();
+        };
+    }
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        boolean virtualViewerPresent = (apiLidMode == io.papermc.paper.block.LidMode.FORCED_OPEN || apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED);
+        int trueOpenCount = this.openCount - (virtualViewerPresent ? 1 : 0);
+        if (trueOpenCount < 0) {
+            throw new IllegalStateException("trueOpenCount is negative: " + trueOpenCount + " openCount: " + openCount + " virtualViewerPresent: " + virtualViewerPresent);
+        }
+        return trueOpenCount > 0 ? io.papermc.paper.block.LidState.OPEN : io.papermc.paper.block.LidState.CLOSED;
     }
-    // CraftBukkit end
 
     protected abstract void onOpen(Level level, BlockPos pos, BlockState state);
 
@@ -37,12 +74,22 @@ public abstract class ContainerOpenersCounter {
 
     protected abstract void openerCountChanged(Level level, BlockPos pos, BlockState state, int count, int openCount);
 
-    public abstract boolean isOwnContainer(Player player);
-
-    public void incrementOpeners(LivingEntity entity, Level level, BlockPos pos, BlockState state, double interactionRange) {
+    public void incrementOpeners(@javax.annotation.Nullable LivingEntity entity, Level level, BlockPos pos, BlockState state, double interactionRange) { // Kiterino - Paper PR - Make entity nullable for New Lidded API
+        // Paper start - Call PlayerLiddedOpenEvent
+        if (entity instanceof net.minecraft.server.level.ServerPlayer player && !org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerLiddedOpenEvent(player, level, pos)) {
+            cancelledPlayers.add(player);
+            return;
+        }
+        // Paper end - Call PlayerLiddedOpenEvent
+        if (this.openCount == 0 && apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+            apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+            stopForceLiddedLidClose(level, pos, state);
+        }
+        // Paper end - add Improved Lidded API
         int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         int i = this.openCount++;
 
+        if (apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) return; // Kiterino - Paper PR - Add improved Lidded API
         // CraftBukkit start - Call redstone event
         if (level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
             int newPower = Math.max(0, Math.min(15, this.openCount));
@@ -60,14 +107,31 @@ public abstract class ContainerOpenersCounter {
         }
 
         this.openerCountChanged(level, pos, state, i, this.openCount);
+        if (entity != null) // Kiterino - Paper PR - Make player nullable for improved Lidded API
         this.maxInteractionRange = Math.max(interactionRange, this.maxInteractionRange);
+        // Kiterino start - Paper PR - Add Improved Lidded API
+        if (entity != null && apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED) {
+            // reset to default
+            apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+            stopForceLiddedLidOpen(level, pos, state);
+        }
+        // Kiterino end - Paper PR - Add Improved Lidded API
     }
 
-    public void decrementOpeners(LivingEntity entity, Level level, BlockPos pos, BlockState state) {
+    public void decrementOpeners(@javax.annotation.Nullable LivingEntity entity, Level level, BlockPos pos, BlockState state) { // Kiterino - Paper PR - Make player nullable for New Lidded API
         int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         if (this.openCount == 0) return; // Paper - Prevent ContainerOpenersCounter openCount from going negative
         int i = this.openCount--;
 
+        // Paper start - add Improved Lidded API
+        if (apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+            if (this.openCount == 0 && apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                stopForceLiddedLidClose(level, pos, state);
+            }
+            return;
+        }
+        // Paper end - add Improved Lidded API
         // CraftBukkit start - Call redstone event
         if (level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
             int newPower = Math.max(0, Math.min(15, this.openCount));
@@ -104,13 +168,22 @@ public abstract class ContainerOpenersCounter {
         List<ContainerUser> entitiesWithContainerOpen = this.getEntitiesWithContainerOpen(level, pos);
         this.maxInteractionRange = 0.0;
 
+        // Kiterino start - Paper PR - Maintain cancelledPlayers, list of players with the chest open, but without the lid.
+        cancelledPlayers.removeIf(java.util.function.Predicate.not(entitiesWithContainerOpen::contains));
+        entitiesWithContainerOpen.removeIf(cancelledPlayers::contains);
+        // Kiterino end - Paper PR - Maintain cancelledPlayers, list of players with the chest open, but without the lid.
+
         for (ContainerUser containerUser : entitiesWithContainerOpen) {
             this.maxInteractionRange = Math.max(containerUser.getContainerInteractionRange(), this.maxInteractionRange);
         }
 
-        int size = entitiesWithContainerOpen.size();
-        if (this.opened) size++; // CraftBukkit - add dummy count from API
-        int i = this.openCount;
+        // Paper Start - Replace with add Improved Lidded API
+        boolean forceClosed = apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED || apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED;
+        boolean forceOpened = apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED || apiLidMode == io.papermc.paper.block.LidMode.FORCED_OPEN;
+        int size = forceClosed ? 0 : entitiesWithContainerOpen.size() + (forceOpened ? 1 : 0);
+        // if (this.opened) size++; // CraftBukkit - add dummy count from API
+        int i = forceClosed ? 0 : this.openCount;
+        // Paper End - Replace with add Improved Lidded API
         if (i != size) {
             boolean flag = size != 0;
             boolean flag1 = i != 0;
diff --git a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
index 4d268864656ad67e6971b74534e25ac47d922cbd..5e93bfdff7e7521449dc037dcff5e47d5e9f1cc9 100644
--- a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
@@ -53,7 +53,7 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     // CraftBukkit start - add fields and methods
     public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
     private int maxStack = MAX_STACK;
-    public boolean opened;
+    // public boolean opened; // Kiterino - Paper PR - Replace with new Lidded API
 
     @Override
     public List<ItemStack> getContents() {
@@ -176,7 +176,7 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     @Override
     public boolean triggerEvent(int id, int type) {
         if (id == 1) {
-            this.openCount = type;
+            if (apiLidMode != io.papermc.paper.block.LidMode.FORCED_CLOSED && apiLidMode != io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) this.openCount = type; // Kiterino - Paper PR - Skip Mutate when forced closed by lidded api
             if (type == 0) {
                 this.animationStatus = ShulkerBoxBlockEntity.AnimationStatus.CLOSING;
             }
@@ -200,6 +200,64 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     public void preRemoveSideEffects(BlockPos pos, BlockState state) {
     }
 
+    // Kiterino start - Paper PR - add Improved Lidded API
+    private io.papermc.paper.block.LidMode apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    private final java.util.Set<ContainerUser> cancelledPlayers = new java.util.HashSet<>(); // Kiterino - Paper PR - Store players whose opening was cancelled by PlayerLiddedOpenEvent
+    public void startForceLiddedLidOpen() {
+        this.openCount++;
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
+        if (this.openCount == 1) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_OPEN, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
+        }
+    }
+    public void stopForceLiddedLidOpen() {
+        this.openCount--;
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
+        if (this.openCount <= 0) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_CLOSE, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_CLOSE, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
+        }
+        apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+    }
+    public void startForceLiddedLidClose() {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_CLOSE, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_CLOSE, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
+        }
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, 0);
+    }
+    public void stopForceLiddedLidClose() {
+        if (this.getTrueLidState() == io.papermc.paper.block.LidState.OPEN) {
+            this.level.gameEvent(null, GameEvent.CONTAINER_OPEN, this.worldPosition);
+            this.level.playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
+        }
+        this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
+    }
+    public io.papermc.paper.block.LidMode getLidMode() {
+        return apiLidMode;
+    }
+    public void setLidMode(final io.papermc.paper.block.LidMode lidMode) {
+        this.apiLidMode = lidMode;
+    }
+    public io.papermc.paper.block.LidState getEffectiveLidState() {
+        return switch (apiLidMode) {
+            case OPEN_UNTIL_VIEWED, FORCED_OPEN -> io.papermc.paper.block.LidState.OPEN;
+            case CLOSED_UNTIL_NOT_VIEWED, FORCED_CLOSED -> io.papermc.paper.block.LidState.CLOSED;
+            default -> getTrueLidState();
+        };
+    }
+    public io.papermc.paper.block.LidState getTrueLidState() {
+        boolean virtualViewerPresent = (apiLidMode == io.papermc.paper.block.LidMode.FORCED_OPEN || apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED);
+        int trueOpenCount = this.openCount - (virtualViewerPresent ? 1 : 0);
+        // ensure trueOpenCount is never negative, throw
+        if (trueOpenCount < 0) {
+            throw new IllegalStateException("trueOpenCount is negative: " + trueOpenCount + " openCount: " + openCount + " virtualViewerPresent: " + virtualViewerPresent);
+        }
+        return trueOpenCount > 0 ? io.papermc.paper.block.LidState.OPEN : io.papermc.paper.block.LidState.CLOSED;
+    }
+    // Kiterino end - Paper PR - Add Improved Lidded API
+
     @Override
     public void startOpen(ContainerUser user) {
         if (!this.remove && !user.getLivingEntity().isSpectator()) {
@@ -207,22 +265,58 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
                 this.openCount = 0;
             }
 
+            // Paper start - Call PlayerLiddedOpenEvent
+            if (user instanceof net.minecraft.server.level.ServerPlayer player && !org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerLiddedOpenEvent(player, this.level, this.worldPosition)) {
+                cancelledPlayers.add(player);
+                return;
+            }
+            // Paper end - Call PlayerLiddedOpenEvent
+            // Kiterino start - Paper PR - Add Improved Lidded API
+            if (this.openCount == 0) {
+                if (apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                    apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                    stopForceLiddedLidClose();
+                }
+            }
+            // Kiterino end - Paper PR - Add Improved Lidded API
+
             this.openCount++;
-            if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call
+            // Kiterino start - Paper PR - Replace with Improved Lidded API
+            //if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
+            if (this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) return;
+            // Kiterino end - Paper PR - Replace with Improved Lidded API
             this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
             if (this.openCount == 1) {
                 this.level.gameEvent(user.getLivingEntity(), GameEvent.CONTAINER_OPEN, this.worldPosition);
                 this.level
                     .playSound(null, this.worldPosition, SoundEvents.SHULKER_BOX_OPEN, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
             }
+            // Kiterino start - Paper PR - Add Improved Lidded API
+            if (apiLidMode == io.papermc.paper.block.LidMode.OPEN_UNTIL_VIEWED) {
+                // reset to default
+                apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                stopForceLiddedLidOpen();
+            }
+            // Kiterino end - Paper PR - Add Improved Lidded API
         }
     }
 
     @Override
     public void stopOpen(ContainerUser user) {
         if (!this.remove && !user.getLivingEntity().isSpectator()) {
+            if (cancelledPlayers.remove(user)) return; // Kiterino - Paper PR - Do not decrement if player's opening was cancelled by PlayerLiddedOpenEvent
             this.openCount--;
-            if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
+            // Kiterino start - Paper PR - Add Improved Lidded API
+            // if (this.opened) return; // CraftBukkit - only animate if the ShulkerBox hasn't been forced open already by an API call.
+            if (this.apiLidMode == io.papermc.paper.block.LidMode.FORCED_CLOSED || this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                if (this.openCount <= 0 && this.apiLidMode == io.papermc.paper.block.LidMode.CLOSED_UNTIL_NOT_VIEWED) {
+                    this.openCount = 0;
+                    this.apiLidMode = io.papermc.paper.block.LidMode.DEFAULT;
+                    this.stopForceLiddedLidClose();
+                }
+                return;
+            }
+            // Kiterino end - Paper PR - Add Improved Lidded API
             this.level.blockEvent(this.worldPosition, this.getBlockState().getBlock(), 1, this.openCount);
             if (this.openCount <= 0) {
                 this.level.gameEvent(user.getLivingEntity(), GameEvent.CONTAINER_CLOSE, this.worldPosition);
