From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 28 Sep 2024 06:09:37 +0300
Subject: [PATCH] Paper PR - Combat tracker API

https://github.com/PaperMC/Paper/pull/11853

diff --git a/net/minecraft/world/damagesource/CombatTracker.java b/net/minecraft/world/damagesource/CombatTracker.java
index 0924fd5203d63a008830d944e3651acd17021c22..ea97e517a6fcfd0374a35e7279e8d55719dc17d0 100644
--- a/net/minecraft/world/damagesource/CombatTracker.java
+++ b/net/minecraft/world/damagesource/CombatTracker.java
@@ -37,7 +37,14 @@ public class CombatTracker {
     public void recordDamage(DamageSource source, float damage) {
         this.recheckStatus();
         FallLocation currentFallLocation = FallLocation.getCurrentFallLocation(this.mob);
-        CombatEntry combatEntry = new CombatEntry(source, damage, currentFallLocation, (float)this.mob.fallDistance);
+        CombatEntry combatEntry = new CombatEntry(source, damage, currentFallLocation, (float) this.mob.fallDistance);
+        // Kiterino start - Paper PR - Combat tracker API
+        recordDamageAndCheckCombatState(combatEntry);
+    }
+
+    public void recordDamageAndCheckCombatState(CombatEntry combatEntry) {
+        DamageSource source = combatEntry.source();
+        // Kiterino end - Paper PR - Combat tracker API
         this.entries.add(combatEntry);
         this.lastDamageTime = this.mob.tickCount;
         this.takingDamage = true;
@@ -156,6 +163,13 @@ public class CombatTracker {
     public void recheckStatus() {
         int i = this.inCombat ? 300 : 100;
         if (this.takingDamage && (!this.mob.isAlive() || this.mob.tickCount - this.lastDamageTime > i)) {
+            // Kiterino start - Paper PR - Combat tracker API
+            resetCombatState();
+        }
+    }
+
+    public void resetCombatState() {
+            // Kiterino end - Paper PR - Combat tracker API
             boolean flag = this.inCombat;
             this.takingDamage = false;
             this.inCombat = false;
@@ -165,6 +179,5 @@ public class CombatTracker {
             }
 
             this.entries.clear();
-        }
     }
 }
