From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 29 Apr 2025 02:18:02 +0300
Subject: [PATCH] Support custom shears


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 37fe408eb87ad04b344a7b96c2dcdae60123d775..d97c68fcd726d42c815a3bcb69ffe877555cd7b3 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1854,7 +1854,7 @@ public class ServerGamePacketListenerImpl
 
     // Purpur start - Dont run with scissors!
     public boolean isScissors(ItemStack stack) {
-        if (!stack.is(Items.SHEARS)) return false;
+        if (!(stack.getItem() instanceof net.minecraft.world.item.ShearsItem)) return false; // Kiterino - Support custom shears
 
         Identifier itemModelReference = stack.get(net.minecraft.core.component.DataComponents.ITEM_MODEL);
         if (itemModelReference != null && itemModelReference.equals(this.player.level().purpurConfig.dontRunWithScissorsItemModelReference)) return true;
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 80300c304fe86090600e3d182679172b5b8180e6..6e2c0b2568a532047bc41f27126ace7f97fb3826 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3148,11 +3148,11 @@ public abstract class Entity
         }
 
         ItemStack heldItem = player.getItemInHand(hand);
-        if (heldItem.is(Items.SHEARS) && this.shearOffAllLeashConnections(player, hand)) { // Paper - PlayerUnleashEntityEvent - pass used hand
+        if (heldItem.getItem() instanceof net.minecraft.world.item.ShearsItem && this.shearOffAllLeashConnections(player, hand)) { // Paper - PlayerUnleashEntityEvent - pass used hand // Kiterino - Support custom shears
             heldItem.hurtAndBreak(1, player, hand);
             return InteractionResult.SUCCESS;
         } else if (this instanceof Mob target
-            && heldItem.is(Items.SHEARS)
+            && heldItem.getItem() instanceof net.minecraft.world.item.ShearsItem // Kiterino - Support custom shears
             && target.canShearEquipment(player)
             && !player.isSecondaryUseActive()
             && this.attemptToShearEquipment(player, hand, heldItem, target)) {
diff --git a/net/minecraft/world/entity/animal/cow/MushroomCow.java b/net/minecraft/world/entity/animal/cow/MushroomCow.java
index 39595cc7d2a74e067beb9c9f9c07c941d34b10ed..8ed6325516442625f8c29d8355e1e4266db0a702 100644
--- a/net/minecraft/world/entity/animal/cow/MushroomCow.java
+++ b/net/minecraft/world/entity/animal/cow/MushroomCow.java
@@ -159,7 +159,7 @@ public class MushroomCow extends AbstractCow implements Shearable {
 
             this.playSound(milkSound, 1.0F, 1.0F);
             return InteractionResult.SUCCESS;
-        } else if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
+        } else if (itemStack.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel level) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/animal/golem/CopperGolem.java b/net/minecraft/world/entity/animal/golem/CopperGolem.java
index eb93463395c8a580faa7c5d0aaa8bf11cdd05f2d..fda7b9dbb73164f66b4402f1e477320927440697 100644
--- a/net/minecraft/world/entity/animal/golem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolem.java
@@ -277,7 +277,7 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
         }
 
         Level level = this.level();
-        if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemStack.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (level instanceof ServerLevel serverLevel) {
                 // Paper start
                 java.util.List<ItemStack> drops = this.generateDefaultDrops(serverLevel, itemStack);
diff --git a/net/minecraft/world/entity/animal/golem/SnowGolem.java b/net/minecraft/world/entity/animal/golem/SnowGolem.java
index ffb0e9cc4f8604fd03cb3a2412001182fb47bce8..91a5b43d08d87561e11916fd6d8941f1940839d4 100644
--- a/net/minecraft/world/entity/animal/golem/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/golem/SnowGolem.java
@@ -184,7 +184,7 @@ public class SnowGolem extends AbstractGolem implements RangedAttackMob, Shearab
     @Override
     protected InteractionResult mobInteract(final Player player, final InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemStack.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel level) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/animal/sheep/Sheep.java b/net/minecraft/world/entity/animal/sheep/Sheep.java
index f0ff3484c73b2348e7b2239517c8984422cd1bdd..c157ca2c7d6ec0f9a73a6cc87e25157173d2896f 100644
--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -185,7 +185,7 @@ public class Sheep extends Animal implements Shearable {
     @Override
     public InteractionResult mobInteract(final Player player, final InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.SHEARS)) {
+        if (itemStack.getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel level && this.readyForShearing()) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
index 818682cc7a14bbb4587dd686a6e590ba7100fe32..8ca82621822f5689bef6d80aa9d9a64adfc23fdb 100644
--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -76,7 +76,7 @@ public class LeashFenceKnotEntity extends BlockAttachedEntity {
             return InteractionResult.SUCCESS;
         }
 
-        if (player.getItemInHand(hand).is(Items.SHEARS)) {
+        if (player.getItemInHand(hand).getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
             InteractionResult result = super.interact(player, hand, location);
             if (result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                 return result;
diff --git a/net/minecraft/world/entity/monster/skeleton/Bogged.java b/net/minecraft/world/entity/monster/skeleton/Bogged.java
index 410c5f8e1571970e8043eb5d976cd6d8b95689eb..211d94c3e7eda6346ab0c45c51a50a056a385b95 100644
--- a/net/minecraft/world/entity/monster/skeleton/Bogged.java
+++ b/net/minecraft/world/entity/monster/skeleton/Bogged.java
@@ -102,7 +102,7 @@ public class Bogged extends AbstractSkeleton implements Shearable {
     @Override
     protected InteractionResult mobInteract(final Player player, final InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemStack.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel level) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/level/block/BeehiveBlock.java b/net/minecraft/world/level/block/BeehiveBlock.java
index 4cc8721b2c81a14286c427bbb33760ff7460a3c5..f8db94ce9d67e5bc90f17dae25ce7828a8b06e92 100644
--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -161,7 +161,7 @@ public class BeehiveBlock extends BaseEntityBlock {
         boolean hiveEmptied = false;
         if (honeyLevel >= 5) {
             Item item = itemStack.getItem();
-            if (level instanceof ServerLevel serverLevel && itemStack.is(Items.SHEARS)) {
+            if (level instanceof ServerLevel serverLevel && itemStack.getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
                 // Paper start - Add PlayerShearBlockEvent
                 List<org.bukkit.inventory.ItemStack> drops = new java.util.ArrayList<>();
                 dropFromBlockInteractLootTable(
diff --git a/net/minecraft/world/level/block/PumpkinBlock.java b/net/minecraft/world/level/block/PumpkinBlock.java
index fcca3424d23cd972a260b93826600b854acbc16c..e60a083534d9317b4037f3f87a0e5232dc03acf1 100644
--- a/net/minecraft/world/level/block/PumpkinBlock.java
+++ b/net/minecraft/world/level/block/PumpkinBlock.java
@@ -43,7 +43,7 @@ public class PumpkinBlock extends Block {
         final InteractionHand hand,
         final BlockHitResult hitResult
     ) {
-        if (!itemStack.is(Items.SHEARS)) {
+        if (!(itemStack.getItem() instanceof net.minecraft.world.item.ShearsItem)) { // Kiterino - Support custom shears
             return super.useItemOn(itemStack, state, level, pos, player, hand, hitResult);
         } else if (level instanceof ServerLevel serverLevel) {
             Direction clickedDirection = hitResult.getDirection();
diff --git a/net/minecraft/world/level/block/TripWireBlock.java b/net/minecraft/world/level/block/TripWireBlock.java
index 83823edc4a943e006e3237ecb9198a571a681873..4d779d18de177bd7639670335396c7cfe18eb9fd 100644
--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -118,7 +118,7 @@ public class TripWireBlock extends Block {
     @Override
     public BlockState playerWillDestroy(final Level level, final BlockPos pos, final BlockState state, final Player player) {
         if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return state; // Paper - prevent disarming tripwires
-        if (!level.isClientSide() && !player.getMainHandItem().isEmpty() && player.getMainHandItem().is(Items.SHEARS)) {
+        if (!level.isClientSide() && !player.getMainHandItem().isEmpty() && player.getMainHandItem().getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
             level.setBlock(pos, state.setValue(DISARMED, true), Block.UPDATE_NONE);
             level.gameEvent(player, GameEvent.SHEAR, pos);
         }
