From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 29 Apr 2025 02:18:02 +0300
Subject: [PATCH] Support custom shears


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 9e233c1aef2d8496365b8bbb375e62036ea76073..e30807eb7bb9a78e7b0616166d61c9aed4fa4bfa 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1793,7 +1793,7 @@ public class ServerGamePacketListenerImpl
 
     // Purpur start - Dont run with scissors!
     public boolean isScissors(ItemStack stack) {
-        if (!stack.is(Items.SHEARS)) return false;
+        if (!(stack.getItem() instanceof net.minecraft.world.item.ShearsItem)) return false; // Kiterino - Support custom shears
 
         ResourceLocation itemModelReference = stack.get(net.minecraft.core.component.DataComponents.ITEM_MODEL);
         if (itemModelReference != null && itemModelReference.equals(this.player.level().purpurConfig.dontRunWithScissorsItemModelReference)) return true;
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 01288f43532258277ca15a6c6c6b1d8a4f6e7122..5691a241505881f4d3c23bfa9426bcb51cceb070 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3073,11 +3073,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
 
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS) && this.shearOffAllLeashConnections(player, hand)) { // Paper - PlayerUnleashEntityEvent - pass used hand
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.shearOffAllLeashConnections(player, hand)) { // Paper - PlayerUnleashEntityEvent - pass used hand // Kiterino - Support custom shears
             itemInHand.hurtAndBreak(1, player, hand);
             return InteractionResult.SUCCESS;
         } else if (this instanceof Mob mob
-            && itemInHand.is(Items.SHEARS)
+            && itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem // Kiterino - Support custom shears
             && mob.canShearEquipment(player)
             && !player.isSecondaryUseActive()
             && this.attemptToShearEquipment(player, hand, itemInHand, mob)) {
diff --git a/net/minecraft/world/entity/animal/MushroomCow.java b/net/minecraft/world/entity/animal/MushroomCow.java
index 3b18ea446cc7e19295a9fa37b74c3a02be421e55..ab0e0ba9c99b5d069613de0d8923891f7d4a3c2e 100644
--- a/net/minecraft/world/entity/animal/MushroomCow.java
+++ b/net/minecraft/world/entity/animal/MushroomCow.java
@@ -161,7 +161,7 @@ public class MushroomCow extends AbstractCow implements Shearable {
 
             this.playSound(soundEvent, 1.0F, 1.0F);
             return InteractionResult.SUCCESS;
-        } else if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        } else if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/animal/SnowGolem.java b/net/minecraft/world/entity/animal/SnowGolem.java
index 03071650204193f1130e8e17f280ae6ad2cfad38..ce5478470c73149870b66b91ee6bd1b743d696ca 100644
--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -182,7 +182,7 @@ public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackM
     @Override
     protected InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java b/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
index 739deece6ae077b8fa1694eb6cc6ed9996dff09b..341cdd38997d9ce2d7320af10049172d59b3929e 100644
--- a/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
@@ -281,7 +281,7 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
         }
 
         Level level = this.level();
-        if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (level instanceof ServerLevel serverLevel) {
                 // Paper start
                 java.util.List<ItemStack> drops = this.generateDefaultDrops(serverLevel, itemInHand);
diff --git a/net/minecraft/world/entity/animal/sheep/Sheep.java b/net/minecraft/world/entity/animal/sheep/Sheep.java
index 2f90f1a5e4d2aad7fc1fe84bac6cf18ed8f35f4d..8d121f24f39e50a12826fb3d7ee4671ea1087b94 100644
--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -185,7 +185,7 @@ public class Sheep extends Animal implements Shearable {
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS)) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel && this.readyForShearing()) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
index 94f965ef441ec8a36bc0f088b39a1a4d52a029ab..20e332c6d4de562e3e0442301d08c39e04769fe0 100644
--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -72,7 +72,7 @@ public class LeashFenceKnotEntity extends BlockAttachedEntity {
         if (this.level().isClientSide()) {
             return InteractionResult.SUCCESS;
         } else {
-            if (player.getItemInHand(hand).is(Items.SHEARS)) {
+            if (player.getItemInHand(hand).getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
                 InteractionResult interactionResult = super.interact(player, hand);
                 if (interactionResult instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                     return interactionResult;
diff --git a/net/minecraft/world/entity/monster/Bogged.java b/net/minecraft/world/entity/monster/Bogged.java
index 2ec4e1cc570ebdbaa8a0f3fc681026a4b89f53ed..aaa7163bde8a934887222444137b87bf59faba37 100644
--- a/net/minecraft/world/entity/monster/Bogged.java
+++ b/net/minecraft/world/entity/monster/Bogged.java
@@ -97,7 +97,7 @@ public class Bogged extends AbstractSkeleton implements Shearable {
     @Override
     protected InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/level/block/BeehiveBlock.java b/net/minecraft/world/level/block/BeehiveBlock.java
index 9e8ae7931af5954c688c7be82b5877a0ef396ec3..98287032229acd5d420bdbbb7978530b8e83ed03 100644
--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -140,7 +140,7 @@ public class BeehiveBlock extends BaseEntityBlock {
         boolean flag = false;
         if (honeyLevelValue >= 5) {
             Item item = stack.getItem();
-            if (level instanceof ServerLevel serverLevel && stack.is(Items.SHEARS)) {
+            if (level instanceof ServerLevel serverLevel && stack.getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
                 // Paper start - Add PlayerShearBlockEvent
                 List<org.bukkit.inventory.ItemStack> drops = new java.util.ArrayList<>();
                 dropFromBlockInteractLootTable(
diff --git a/net/minecraft/world/level/block/PumpkinBlock.java b/net/minecraft/world/level/block/PumpkinBlock.java
index cb244c822de515040ab87d9469ad1bf390337ca4..89abeee8478560679ef8be3490b82fc47961ca0c 100644
--- a/net/minecraft/world/level/block/PumpkinBlock.java
+++ b/net/minecraft/world/level/block/PumpkinBlock.java
@@ -36,7 +36,7 @@ public class PumpkinBlock extends Block {
     protected InteractionResult useItemOn(
         ItemStack stack, BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult
     ) {
-        if (!stack.is(Items.SHEARS)) {
+        if (!(stack.getItem() instanceof net.minecraft.world.item.ShearsItem)) { // Kiterino - Support custom shears
             return super.useItemOn(stack, state, level, pos, player, hand, hitResult);
         } else if (level instanceof ServerLevel serverLevel) {
             Direction direction = hitResult.getDirection();
diff --git a/net/minecraft/world/level/block/TripWireBlock.java b/net/minecraft/world/level/block/TripWireBlock.java
index 9d005c3db8f0c784cdd7217e7b110daa703a9cf3..57ebf23e727befb0147b420d7a032fff413de8f5 100644
--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -119,7 +119,7 @@ public class TripWireBlock extends Block {
     @Override
     public BlockState playerWillDestroy(Level level, BlockPos pos, BlockState state, Player player) {
         if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return state; // Paper - prevent disarming tripwires
-        if (!level.isClientSide() && !player.getMainHandItem().isEmpty() && player.getMainHandItem().is(Items.SHEARS)) {
+        if (!level.isClientSide() && !player.getMainHandItem().isEmpty() && player.getMainHandItem().getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
             level.setBlock(pos, state.setValue(DISARMED, true), Block.UPDATE_NONE);
             level.gameEvent(player, GameEvent.SHEAR, pos);
         }
@@ -225,7 +225,7 @@ public class TripWireBlock extends Block {
 
         if (flag != poweredValue) {
             blockState = blockState.setValue(POWERED, flag);
-            level.setBlock(pos, blockState, Block.UPDATE_ALL);
+            level.setBlock(pos, blockState, 3);
             this.updateSource(level, pos, blockState);
         }
 
