From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 29 Apr 2025 02:18:02 +0300
Subject: [PATCH] Support custom shears


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index e381cbc7d3af6731acdb0e371f1c439750603e69..096d945b7743f367acfeee5c913e49eb31fd24dd 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1792,7 +1792,7 @@ public class ServerGamePacketListenerImpl
 
     // Purpur start - Dont run with scissors!
     public boolean isScissors(ItemStack stack) {
-        if (!stack.is(Items.SHEARS)) return false;
+        if (!(stack.getItem() instanceof net.minecraft.world.item.ShearsItem)) return false; // Kiterino - Support custom shears
 
         ResourceLocation itemModelReference = stack.get(net.minecraft.core.component.DataComponents.ITEM_MODEL);
         if (itemModelReference != null && itemModelReference.equals(this.player.level().purpurConfig.dontRunWithScissorsItemModelReference)) return true;
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index abba7c7c5ab4dc498e2d85bc02df39e95e50477b..b37ba8f7835678a685ba7376af3cde59193dd829 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3073,11 +3073,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
 
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS) && this.shearOffAllLeashConnections(player, hand)) { // Paper - PlayerUnleashEntityEvent - pass used hand
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.shearOffAllLeashConnections(player, hand)) { // Paper - PlayerUnleashEntityEvent - pass used hand // Kiterino - Support custom shears
             itemInHand.hurtAndBreak(1, player, hand);
             return InteractionResult.SUCCESS;
         } else if (this instanceof Mob mob
-            && itemInHand.is(Items.SHEARS)
+            && itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem // Kiterino - Support custom shears
             && mob.canShearEquipment(player)
             && !player.isSecondaryUseActive()
             && this.attemptToShearEquipment(player, hand, itemInHand, mob)) {
diff --git a/net/minecraft/world/entity/animal/MushroomCow.java b/net/minecraft/world/entity/animal/MushroomCow.java
index 3b18ea446cc7e19295a9fa37b74c3a02be421e55..ab0e0ba9c99b5d069613de0d8923891f7d4a3c2e 100644
--- a/net/minecraft/world/entity/animal/MushroomCow.java
+++ b/net/minecraft/world/entity/animal/MushroomCow.java
@@ -161,7 +161,7 @@ public class MushroomCow extends AbstractCow implements Shearable {
 
             this.playSound(soundEvent, 1.0F, 1.0F);
             return InteractionResult.SUCCESS;
-        } else if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        } else if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/animal/SnowGolem.java b/net/minecraft/world/entity/animal/SnowGolem.java
index 03071650204193f1130e8e17f280ae6ad2cfad38..ce5478470c73149870b66b91ee6bd1b743d696ca 100644
--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -182,7 +182,7 @@ public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackM
     @Override
     protected InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java b/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
index 088b2cf28f56d60167b03bcce5fbd2c47fa2de9a..2638a74286ec81d50b200dfaaf380b39153c768e 100644
--- a/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
@@ -279,7 +279,7 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
         }
 
         Level level = this.level();
-        if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (level instanceof ServerLevel serverLevel) {
                 // Paper start
                 java.util.List<ItemStack> drops = this.generateDefaultDrops(serverLevel, itemInHand);
diff --git a/net/minecraft/world/entity/animal/sheep/Sheep.java b/net/minecraft/world/entity/animal/sheep/Sheep.java
index 21d66dd46650afae2091e236dfec760c096c34f8..8178d5c3e8ad934c61caf55a41985ae5fb494a38 100644
--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -184,7 +184,7 @@ public class Sheep extends Animal implements Shearable {
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS)) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel && this.readyForShearing()) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
index 94f965ef441ec8a36bc0f088b39a1a4d52a029ab..20e332c6d4de562e3e0442301d08c39e04769fe0 100644
--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -72,7 +72,7 @@ public class LeashFenceKnotEntity extends BlockAttachedEntity {
         if (this.level().isClientSide()) {
             return InteractionResult.SUCCESS;
         } else {
-            if (player.getItemInHand(hand).is(Items.SHEARS)) {
+            if (player.getItemInHand(hand).getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
                 InteractionResult interactionResult = super.interact(player, hand);
                 if (interactionResult instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                     return interactionResult;
diff --git a/net/minecraft/world/entity/monster/Bogged.java b/net/minecraft/world/entity/monster/Bogged.java
index 2ec4e1cc570ebdbaa8a0f3fc681026a4b89f53ed..aaa7163bde8a934887222444137b87bf59faba37 100644
--- a/net/minecraft/world/entity/monster/Bogged.java
+++ b/net/minecraft/world/entity/monster/Bogged.java
@@ -97,7 +97,7 @@ public class Bogged extends AbstractSkeleton implements Shearable {
     @Override
     protected InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemInHand.getItem() instanceof net.minecraft.world.item.ShearsItem && this.readyForShearing()) { // Kiterino - Support custom shears
             if (this.level() instanceof ServerLevel serverLevel) {
                 // CraftBukkit start
                 // Paper start - custom shear drops
diff --git a/net/minecraft/world/level/block/BeehiveBlock.java b/net/minecraft/world/level/block/BeehiveBlock.java
index 978ebd1c72ad897115e2953fd7b716febff9c168..16efb0c23b18dc71d685ee3d3b6bb1c7d1833038 100644
--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -140,7 +140,7 @@ public class BeehiveBlock extends BaseEntityBlock {
         boolean flag = false;
         if (honeyLevelValue >= 5) {
             Item item = stack.getItem();
-            if (level instanceof ServerLevel serverLevel && stack.is(Items.SHEARS)) {
+            if (level instanceof ServerLevel serverLevel && stack.getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
                 // Paper start - Add PlayerShearBlockEvent
                 List<org.bukkit.inventory.ItemStack> drops = new java.util.ArrayList<>();
                 dropFromBlockInteractLootTable(
diff --git a/net/minecraft/world/level/block/PumpkinBlock.java b/net/minecraft/world/level/block/PumpkinBlock.java
index 2e2963ad6441522554fa58cf6116abde3cf5dea7..7aafad6ea5003b51149eeb36af071dff0599a671 100644
--- a/net/minecraft/world/level/block/PumpkinBlock.java
+++ b/net/minecraft/world/level/block/PumpkinBlock.java
@@ -36,7 +36,7 @@ public class PumpkinBlock extends Block {
     protected InteractionResult useItemOn(
         ItemStack stack, BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult
     ) {
-        if (!stack.is(Items.SHEARS)) {
+        if (!(stack.getItem() instanceof net.minecraft.world.item.ShearsItem)) { // Kiterino - Support custom shears
             return super.useItemOn(stack, state, level, pos, player, hand, hitResult);
         } else if (level instanceof ServerLevel serverLevel) {
             Direction direction = hitResult.getDirection();
diff --git a/net/minecraft/world/level/block/TripWireBlock.java b/net/minecraft/world/level/block/TripWireBlock.java
index 9fcb791740e9df2ef7828cd38344a7c0105c792b..e6e3ddd425fdb7a4d9d246d28af7116253041084 100644
--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -119,7 +119,7 @@ public class TripWireBlock extends Block {
     @Override
     public BlockState playerWillDestroy(Level level, BlockPos pos, BlockState state, Player player) {
         if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return state; // Paper - prevent disarming tripwires
-        if (!level.isClientSide() && !player.getMainHandItem().isEmpty() && player.getMainHandItem().is(Items.SHEARS)) {
+        if (!level.isClientSide() && !player.getMainHandItem().isEmpty() && player.getMainHandItem().getItem() instanceof net.minecraft.world.item.ShearsItem) { // Kiterino - Support custom shears
             level.setBlock(pos, state.setValue(DISARMED, true), 260);
             level.gameEvent(player, GameEvent.SHEAR, pos);
         }
