From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 15 Jul 2025 14:06:00 +0300
Subject: [PATCH] Default pickup loot in summon


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 79cecd8a00198fa5d04d4535d0800fcd727aa0fe..62dd7bc180dfc2faf6c98a84be6463d22db0ed52 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4844,6 +4844,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return false;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    public boolean getCanPickUpLootDefault() {
+        return false;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public final EntityDimensions getDimensions(final Pose pose) {
         return pose == Pose.SLEEPING ? SLEEPING_DIMENSIONS : this.getDefaultDimensions(pose).scale(this.getScale());
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 9335166a301493f9e2470b0c0d6eb8d26ece79c6..15611bbd35b1ce319fa90ee64151a2cc9f58c026 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -479,7 +479,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected void readAdditionalSaveData(final ValueInput input) {
         super.readAdditionalSaveData(input);
         // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
-        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", false);
+        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault()); // Kiterino - Default pickup loot in summon
         if (isLevelAtLeast(input, 1) || canPickUpLoot) {
             this.setCanPickUpLoot(canPickUpLoot);
         }
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 489e09f26aa3595317e70a26e14ff3018636911e..a3f78816fa42075534bb1bf562d416592143b26f 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -323,6 +323,13 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         return !this.isOnPickupCooldown() && this.hasItemInHand();
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.canPickUpLoot();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public boolean hasItemInHand() {
         return !this.getItemInHand(InteractionHand.MAIN_HAND).isEmpty();
     }
diff --git a/net/minecraft/world/entity/animal/dolphin/Dolphin.java b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
index d363e9da76a292c5d388f4b8a81e54f1322f5bd7..6d46b5b85a5699ff3f71a86680dbd3a148498aa2 100644
--- a/net/minecraft/world/entity/animal/dolphin/Dolphin.java
+++ b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
@@ -125,6 +125,13 @@ public class Dolphin extends AgeableWaterCreature {
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/fox/Fox.java b/net/minecraft/world/entity/animal/fox/Fox.java
index cc2b5316cf1aa06fc5586e0032a8f8e069a982fe..15928824469ce86be1f65227aba89dc1416ff610 100644
--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -161,6 +161,13 @@ public class Fox extends Animal {
         this.getNavigation().setRequiredPathLength(32.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/panda/Panda.java b/net/minecraft/world/entity/animal/panda/Panda.java
index b1f680ffb815d57e717c59b0c44a2e0b27f5a94e..865ff642cd5cbb563e0effbc218aacfd22e232a9 100644
--- a/net/minecraft/world/entity/animal/panda/Panda.java
+++ b/net/minecraft/world/entity/animal/panda/Panda.java
@@ -112,6 +112,13 @@ public class Panda extends Animal {
         }
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return !this.isBaby();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/monster/illager/Pillager.java b/net/minecraft/world/entity/monster/illager/Pillager.java
index ed9052bfa3b9c19ffcc54d74a7a24c36ba476812..8901c619aa9d2700b5e1d2c12fc9d179596baf92 100644
--- a/net/minecraft/world/entity/monster/illager/Pillager.java
+++ b/net/minecraft/world/entity/monster/illager/Pillager.java
@@ -191,6 +191,13 @@ public class Pillager extends AbstractIllager implements CrossbowAttackMob, Inve
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public float getWalkTargetValue(final BlockPos pos, final LevelReader level) {
         return 0.0F;
diff --git a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
index c651423836c8e65021e3e385351c8e702056af39..07387ba711ca6403bb0abccf1560ab53b778c4b1 100644
--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -40,6 +40,13 @@ public abstract class AbstractPiglin extends Monster {
         this.setPathfindingMalus(PathType.FIRE, -1.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     private void applyOpenDoorsAbility() {
         if (GoalUtils.hasGroundPathNavigation(this)) {
             this.getNavigation().setCanOpenDoors(true);
@@ -72,7 +79,7 @@ public abstract class AbstractPiglin extends Monster {
     @Override
     protected void readAdditionalSaveData(final ValueInput input) {
         super.readAdditionalSaveData(input);
-        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", true));
+        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault())); // Kiterino - Default pickup loot in summon
         this.setImmuneToZombification(input.getBooleanOr("IsImmuneToZombification", false));
         this.timeInOverworld = input.getIntOr("TimeInOverworld", 0);
     }
diff --git a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
index 8f30f9f0f12e5fd6ddcb1809195f0823ea3ebd51..46dc95841037f135449ef09ba51a325f75c4601e 100644
--- a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
@@ -148,6 +148,13 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         return groupData;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.skeletons;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public void reassessWeaponGoal() {
         if (this.level() != null && !this.level().isClientSide()) {
             this.goalSelector.removeGoal(this.meleeGoal);
diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index 6335b03c39452c2e8e46084e61cefcf8185d2580..32db74cde95c79d3c9095a401cf1522097768b84 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -114,6 +114,13 @@ public class Zombie extends Monster {
         this(EntityTypes.ZOMBIE, level);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.zombies;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index a1ff5e589516bb46236ef1408f0f4f2f1387e72a..af1f025b21b0b084a78ea4936e64c4843750f167 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -192,6 +192,13 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Allow leashing villagers
     @Override
     public boolean canBeLeashed() {
