From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 15 Jul 2025 14:06:00 +0300
Subject: [PATCH] Default pickup loot in summon


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 93e3bad5d597154ed778470bfd028b8d2eb40456..11b5d9c4130e9fafe5915f625613f0851a1afa3b 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4706,6 +4706,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return false;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    public boolean getCanPickUpLootDefault() {
+        return false;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public final EntityDimensions getDimensions(Pose pose) {
         return pose == Pose.SLEEPING ? SLEEPING_DIMENSIONS : this.getDefaultDimensions(pose).scale(this.getScale());
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 642c939ac007369b603819e8d74c1fdf18b09cf3..69257ee8a3347e5e7a824ee7816ce351288e506c 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -468,7 +468,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
         // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
-        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", false);
+        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault()); // Kiterino - Default pickup loot in summon
         if (isLevelAtLeast(input, 1) || canPickUpLoot) {
             this.setCanPickUpLoot(canPickUpLoot);
         }
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 8635ecfa49e2504c981dd48acbfbca5303a37611..135c687ec205a593c3dd7b9e98d546d205f82cee 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -340,6 +340,13 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         return !this.isOnPickupCooldown() && this.hasItemInHand();
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.canPickUpLoot();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public boolean hasItemInHand() {
         return !this.getItemInHand(InteractionHand.MAIN_HAND).isEmpty();
     }
diff --git a/net/minecraft/world/entity/animal/dolphin/Dolphin.java b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
index ea3ccc1f39f70c8ce407ed94ab19a3caa3b90dc4..f140602be66b1ce1da6d45664e55ce93791e0111 100644
--- a/net/minecraft/world/entity/animal/dolphin/Dolphin.java
+++ b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
@@ -116,6 +116,13 @@ public class Dolphin extends AgeableWaterCreature {
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/fox/Fox.java b/net/minecraft/world/entity/animal/fox/Fox.java
index f15ce0c57e0983df437bdf5c863776e8534843e5..ce1b35dea6a4802be51ba4049eca50a0f7068839 100644
--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -151,6 +151,13 @@ public class Fox extends Animal {
         this.getNavigation().setRequiredPathLength(32.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/panda/Panda.java b/net/minecraft/world/entity/animal/panda/Panda.java
index 5116e9e16070e23d13b526d21facb7b5ad0988ba..df782b6bafc369b3e1c34b4612f22118bc67e674 100644
--- a/net/minecraft/world/entity/animal/panda/Panda.java
+++ b/net/minecraft/world/entity/animal/panda/Panda.java
@@ -108,6 +108,13 @@ public class Panda extends Animal {
         }
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return !this.isBaby();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/monster/illager/Pillager.java b/net/minecraft/world/entity/monster/illager/Pillager.java
index 1a283185ac03ba44bae531f9b234703fa213df1a..39dacbf73f4daa8a139301695ae3db36630b9763 100644
--- a/net/minecraft/world/entity/monster/illager/Pillager.java
+++ b/net/minecraft/world/entity/monster/illager/Pillager.java
@@ -191,6 +191,13 @@ public class Pillager extends AbstractIllager implements CrossbowAttackMob, Inve
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public float getWalkTargetValue(BlockPos pos, LevelReader level) {
         return 0.0F;
diff --git a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
index 16afad411aa86c0439be169958926519a287cc24..97c95b2404b6e41840710b75d39e1f1fb7deac5c 100644
--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -38,6 +38,13 @@ public abstract class AbstractPiglin extends Monster {
         this.setPathfindingMalus(PathType.DAMAGE_FIRE, -1.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     private void applyOpenDoorsAbility() {
         if (GoalUtils.hasGroundPathNavigation(this)) {
             this.getNavigation().setCanOpenDoors(true);
@@ -70,7 +77,7 @@ public abstract class AbstractPiglin extends Monster {
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
-        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", true));
+        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault())); // Kiterino - Default pickup loot in summon
         this.setImmuneToZombification(input.getBooleanOr("IsImmuneToZombification", false));
         this.timeInOverworld = input.getIntOr("TimeInOverworld", 0);
     }
diff --git a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
index 60afd81d3bf671889fbff5d4a3fabb38a8e2d461..d380663708c63584b5ae01758b985a75e9531d78 100644
--- a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
@@ -148,6 +148,13 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         return spawnGroupData;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.skeletons;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public void reassessWeaponGoal() {
         if (this.level() != null && !this.level().isClientSide()) {
             this.goalSelector.removeGoal(this.meleeGoal);
diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index f9435eee22c85f8a62611d6742d145f4eeeff525..8c9d3da9fad8b5887c19cde42bf18a0ce00143af 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -104,6 +104,13 @@ public class Zombie extends Monster {
         this(EntityType.ZOMBIE, level);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.zombies;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 0c110421074d57e7283168c1133cd4ee314b57f6..81c774d69aab7d44e9bad0dd8d5629036cee0f4c 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -199,6 +199,13 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         this.setVillagerData(this.getVillagerData().withType(villagerType).withProfession(level.registryAccess(), VillagerProfession.NONE));
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Allow leashing villagers
     @Override
     public boolean canBeLeashed() {
