From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 15 Jul 2025 14:06:00 +0300
Subject: [PATCH] Default pickup loot in summon


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 195bc559c9a400744c9cef0f36a6afed82c1f51a..871839d823785f32f4d43dc003ca6fd7c86f6c3a 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4839,6 +4839,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return false;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    public boolean getCanPickUpLootDefault() {
+        return false;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public final EntityDimensions getDimensions(final Pose pose) {
         return pose == Pose.SLEEPING ? SLEEPING_DIMENSIONS : this.getDefaultDimensions(pose).scale(this.getScale());
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index c2459b062aefe2b99d4bba783eb7ce7ac63be703..49cc0ec83cb7adc03bb05053f3a1d85ef5e39781 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -480,7 +480,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected void readAdditionalSaveData(final ValueInput input) {
         super.readAdditionalSaveData(input);
         // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
-        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", false);
+        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault()); // Kiterino - Default pickup loot in summon
         if (isLevelAtLeast(input, 1) || canPickUpLoot) {
             this.setCanPickUpLoot(canPickUpLoot);
         }
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 9e31aaff00f76cfbba9539d7e2bb3283afdaae65..7e10d1920c867861a7a9e2e3c481e85833e26a89 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -322,6 +322,13 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         return !this.isOnPickupCooldown() && this.hasItemInHand();
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.canPickUpLoot();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public boolean hasItemInHand() {
         return !this.getItemInHand(InteractionHand.MAIN_HAND).isEmpty();
     }
diff --git a/net/minecraft/world/entity/animal/dolphin/Dolphin.java b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
index 048deece8e45f51d3bad36ef0afd39847086aa9f..319de886224ad08d6d5f81ea390929a865a972d6 100644
--- a/net/minecraft/world/entity/animal/dolphin/Dolphin.java
+++ b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
@@ -118,6 +118,13 @@ public class Dolphin extends AgeableWaterCreature {
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/fox/Fox.java b/net/minecraft/world/entity/animal/fox/Fox.java
index afbfa283a489d3d6e02c9724d338798e99490557..48bf3a6fab6536aa93cab6667ef0a6cff6ea858f 100644
--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -154,6 +154,13 @@ public class Fox extends Animal {
         this.getNavigation().setRequiredPathLength(32.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/panda/Panda.java b/net/minecraft/world/entity/animal/panda/Panda.java
index f509b3a3a8f1021d5292ec6ed0a6a91e09787754..3f8ada9058c1e50044ed764672789f3245bf4938 100644
--- a/net/minecraft/world/entity/animal/panda/Panda.java
+++ b/net/minecraft/world/entity/animal/panda/Panda.java
@@ -110,6 +110,13 @@ public class Panda extends Animal {
         }
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return !this.isBaby();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/monster/illager/Pillager.java b/net/minecraft/world/entity/monster/illager/Pillager.java
index 20c0b12b56416bc532de4de827c88850bfd1f0f3..84de99923227864760c761d78f69b1a2332c8da4 100644
--- a/net/minecraft/world/entity/monster/illager/Pillager.java
+++ b/net/minecraft/world/entity/monster/illager/Pillager.java
@@ -191,6 +191,13 @@ public class Pillager extends AbstractIllager implements CrossbowAttackMob, Inve
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public float getWalkTargetValue(final BlockPos pos, final LevelReader level) {
         return 0.0F;
diff --git a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
index 65090418480c1bb4d30c7228db8c804802e3b500..9e911949017f288b624dc7a16f6d78c4b2099ccf 100644
--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -38,6 +38,13 @@ public abstract class AbstractPiglin extends Monster {
         this.setPathfindingMalus(PathType.FIRE, -1.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     private void applyOpenDoorsAbility() {
         if (GoalUtils.hasGroundPathNavigation(this)) {
             this.getNavigation().setCanOpenDoors(true);
@@ -70,7 +77,7 @@ public abstract class AbstractPiglin extends Monster {
     @Override
     protected void readAdditionalSaveData(final ValueInput input) {
         super.readAdditionalSaveData(input);
-        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", true));
+        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault())); // Kiterino - Default pickup loot in summon
         this.setImmuneToZombification(input.getBooleanOr("IsImmuneToZombification", false));
         this.timeInOverworld = input.getIntOr("TimeInOverworld", 0);
     }
diff --git a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
index d02869b5c17378abf18b4ba3514e8916c62b904c..79b0356ac48188e12ae9368bf125bb4da50c728a 100644
--- a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
@@ -148,6 +148,13 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         return groupData;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.skeletons;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public void reassessWeaponGoal() {
         if (this.level() != null && !this.level().isClientSide()) {
             this.goalSelector.removeGoal(this.meleeGoal);
diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index 04ba29b8864804899e19b08b1952df781c96a23e..91c65edb4c0acd56be8cfcebf6dcadcf8beabab5 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -113,6 +113,13 @@ public class Zombie extends Monster {
         this(EntityType.ZOMBIE, level);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.zombies;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 0503f9aa01099ebb807c3b6ccfe88f9610cc36b5..b41278fa5fa442081496b699addd86391d698f50 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -201,6 +201,13 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
         this.setVillagerData(this.getVillagerData().withType(type).withProfession(level.registryAccess(), VillagerProfession.NONE));
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Allow leashing villagers
     @Override
     public boolean canBeLeashed() {
