From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 15 Jul 2025 14:06:00 +0300
Subject: [PATCH] Default pickup loot in summon


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 4780b829a54ce18288df57267c5cf5d64fb81b3c..1687197bf89b58009a5b8efd9bea96ef1a00f5ad 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4557,6 +4557,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return false;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    public boolean getCanPickUpLootDefault() {
+        return false;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public final EntityDimensions getDimensions(Pose pose) {
         return pose == Pose.SLEEPING ? SLEEPING_DIMENSIONS : this.getDefaultDimensions(pose).scale(this.getScale());
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 6fdc56f7a949d2b1bd0f561adbfe1e501a4cb60e..23468734bebf184389c1662d88fff952d649f719 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -472,7 +472,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
         // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
-        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", false);
+        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault()); // Kiterino - Default pickup loot in summon
         if (isLevelAtLeast(input, 1) || canPickUpLoot) {
             this.setCanPickUpLoot(canPickUpLoot);
         }
diff --git a/net/minecraft/world/entity/animal/Dolphin.java b/net/minecraft/world/entity/animal/Dolphin.java
index 9bf53221bf6652fa8e0a2ef47ac1524065ffe784..e321c0f15b0d9bab8c6b137e07e5f3f97c22b31d 100644
--- a/net/minecraft/world/entity/animal/Dolphin.java
+++ b/net/minecraft/world/entity/animal/Dolphin.java
@@ -116,6 +116,13 @@ public class Dolphin extends AgeableWaterCreature {
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/Fox.java b/net/minecraft/world/entity/animal/Fox.java
index c4fae5fb7dd0da2a5d850b1513c9dfab1d05d727..43f4e1a49e0ab0672f0870a7bd123dbb296d6e95 100644
--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -144,6 +144,13 @@ public class Fox extends Animal {
         this.getNavigation().setRequiredPathLength(32.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/Panda.java b/net/minecraft/world/entity/animal/Panda.java
index 9fae0d54fce37ef1b3511339ba05673e80ef6d01..e0c991f6b799f186be8c1882144a5f96a2c01db9 100644
--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -107,6 +107,13 @@ public class Panda extends Animal {
         }
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return !this.isBaby();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index bbe59d6ab90f1b1274289ef20811f914afa55fd3..9333d3911363ab9d5a134f84a9013e11732ccb4c 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -334,6 +334,13 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         return !this.isOnPickupCooldown() && this.hasItemInHand();
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.canPickUpLoot();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public boolean hasItemInHand() {
         return !this.getItemInHand(InteractionHand.MAIN_HAND).isEmpty();
     }
diff --git a/net/minecraft/world/entity/monster/AbstractSkeleton.java b/net/minecraft/world/entity/monster/AbstractSkeleton.java
index 14730f53561dec218b0fa6648c8de2b2157165fc..fe9f69a1d3d4801139bdacf32e7996821bf54f79 100644
--- a/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -72,6 +72,13 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         this.setShouldBurnInDay(true); // Purpur - API for any mob to burn daylight
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.skeletons;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     protected void registerGoals() {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
diff --git a/net/minecraft/world/entity/monster/Pillager.java b/net/minecraft/world/entity/monster/Pillager.java
index 8bc07bbf713ebd3afc8c0e62c92b5e11fb424307..22e44ecd18bdbf7014b0728c750b676b1c1056cc 100644
--- a/net/minecraft/world/entity/monster/Pillager.java
+++ b/net/minecraft/world/entity/monster/Pillager.java
@@ -183,6 +183,13 @@ public class Pillager extends AbstractIllager implements CrossbowAttackMob, Inve
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public float getWalkTargetValue(BlockPos pos, LevelReader level) {
         return 0.0F;
diff --git a/net/minecraft/world/entity/monster/Zombie.java b/net/minecraft/world/entity/monster/Zombie.java
index 2dd08e5ef413c37a682046071b83f595c8499cf2..e931ec977cbf9370e638c8b20d038a8637c41ebc 100644
--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -104,6 +104,13 @@ public class Zombie extends Monster {
         this(EntityType.ZOMBIE, level);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.zombies;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
index 9b76d5156309981127adb2ccd98fb666480618bb..db0171607f5497903187dd0e9275369ec7dcae48 100644
--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -37,6 +37,13 @@ public abstract class AbstractPiglin extends Monster {
         this.setPathfindingMalus(PathType.DAMAGE_FIRE, -1.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     private void applyOpenDoorsAbility() {
         if (GoalUtils.hasGroundPathNavigation(this)) {
             this.getNavigation().setCanOpenDoors(true);
@@ -69,7 +76,7 @@ public abstract class AbstractPiglin extends Monster {
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
-        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", true));
+        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault())); // Kiterino - Default pickup loot in summon
         this.setImmuneToZombification(input.getBooleanOr("IsImmuneToZombification", false));
         this.timeInOverworld = input.getIntOr("TimeInOverworld", 0);
     }
diff --git a/net/minecraft/world/entity/npc/Villager.java b/net/minecraft/world/entity/npc/Villager.java
index bfff91e5e252141f887d8965ecf1dd9ff1cebfe3..61ac4cc1af553add9c1f45bb38f182ff49839249 100644
--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -199,6 +199,13 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         this.setVillagerData(this.getVillagerData().withType(villagerType).withProfession(level.registryAccess(), VillagerProfession.NONE));
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Allow leashing villagers
     @Override
     public boolean canBeLeashed() {
