From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 5 Aug 2025 02:38:09 +0300
Subject: [PATCH] Injecting custom items extras


diff --git a/net/minecraft/core/MappedRegistry.java b/net/minecraft/core/MappedRegistry.java
index ad1558693564ebd18f5e16b85ae552d2af95efef..169b1227cdf26c9eaf0d672492f8d2644b5f9ae9 100644
--- a/net/minecraft/core/MappedRegistry.java
+++ b/net/minecraft/core/MappedRegistry.java
@@ -333,6 +333,13 @@ public class MappedRegistry<T> implements WritableRegistry<T> {
         }
     }
 
+    // Kiterino start - Injecting custom items extras
+    @Override
+    public boolean isFrozen() {
+        return this.frozen;
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public Holder.Reference<T> createIntrusiveHolder(T value) {
         if (this.unregisteredIntrusiveHolders == null) {
diff --git a/net/minecraft/core/Registry.java b/net/minecraft/core/Registry.java
index af9de8a3c46231159973632fcfc798a4988b9ef5..ef125597f95f0bf108720a4369f62b8630cb477e 100644
--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -128,6 +128,10 @@ public interface Registry<T> extends Keyable, HolderLookup.RegistryLookup<T>, Id
 
     Registry<T> freeze();
 
+    // Kiterino start - Injecting custom items extras
+    boolean isFrozen();
+    // Kiterino end - Injecting custom items extras
+
     Holder.Reference<T> createIntrusiveHolder(T value);
 
     Optional<Holder.Reference<T>> get(int index);
diff --git a/net/minecraft/world/item/AxeItem.java b/net/minecraft/world/item/AxeItem.java
index 443410efadc9b8e919a23e475ff9457b93e8f156..35b2bb4643c9f33ca23a3fe7e8b8ba4bc3ace56c 100644
--- a/net/minecraft/world/item/AxeItem.java
+++ b/net/minecraft/world/item/AxeItem.java
@@ -57,6 +57,12 @@ public class AxeItem extends Item {
         super(properties.axe(material, attackDamage, attackSpeed));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public AxeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/HoeItem.java b/net/minecraft/world/item/HoeItem.java
index db539e794dc2df0ca859a429ece5a6c75344d5bf..a7ff6cb0f1fca34f6efee38d204e46c64572688b 100644
--- a/net/minecraft/world/item/HoeItem.java
+++ b/net/minecraft/world/item/HoeItem.java
@@ -40,6 +40,12 @@ public class HoeItem extends Item {
         super(properties.hoe(material, attackDamage, attackSpeed));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public HoeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ShearsItem.java b/net/minecraft/world/item/ShearsItem.java
index 5e25ddf78f72ada318ef28d29199de9b93fe667e..df5d528cea828841cfac3ed61f91bac85f157d26 100644
--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -30,7 +30,7 @@ public class ShearsItem extends Item {
     }
 
     public static Tool createToolProperties() {
-        HolderGetter<Block> holderGetter = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> holderGetter = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return new Tool(
             List.of(
                 Tool.Rule.minesAndDrops(HolderSet.direct(Blocks.COBWEB.builtInRegistryHolder()), 15.0F),
diff --git a/net/minecraft/world/item/ShovelItem.java b/net/minecraft/world/item/ShovelItem.java
index 7910c96cc47985acc8b64ba376933c84c51022d4..0075865e62395078c64e54868597f356af6bce1f 100644
--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -34,6 +34,12 @@ public class ShovelItem extends Item {
         super(properties.shovel(material, attackDamage, attackSpeed));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public ShovelItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ToolMaterial.java b/net/minecraft/world/item/ToolMaterial.java
index 5a37b3a4aca3f0e0976fa6ef2b4be0c921db1dad..118392f46c6227295ee45872fe216fbe58235483 100644
--- a/net/minecraft/world/item/ToolMaterial.java
+++ b/net/minecraft/world/item/ToolMaterial.java
@@ -31,13 +31,18 @@ public record ToolMaterial(
     );
 
     private Item.Properties applyCommonProperties(Item.Properties properties) {
+        // Kiterino start - Injecting custom items extras
+        if (this.repairItems == ItemTags.ANVIL) {
+            return properties;
+        }
+        // Kiterino end - Injecting custom items extras
         return properties.durability(this.durability).repairable(this.repairItems).enchantable(this.enchantmentValue);
     }
 
     public Item.Properties applyToolProperties(
         Item.Properties properties, TagKey<Block> mineableBlocks, float attackDamage, float attackSpeed, float disableBlockingForSeconds
     ) {
-        HolderGetter<Block> holderGetter = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> holderGetter = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
@@ -71,7 +76,7 @@ public record ToolMaterial(
     }
 
     public Item.Properties applySwordProperties(Item.Properties properties, float attackDamage, float attackSpeed) {
-        HolderGetter<Block> holderGetter = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> holderGetter = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
