From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 5 Aug 2025 02:38:09 +0300
Subject: [PATCH] Injecting custom items extras


diff --git a/net/minecraft/core/MappedRegistry.java b/net/minecraft/core/MappedRegistry.java
index 470d7c770ae9d045b97e2df145cfe3cf9d101deb..3463296f5e3656a7af24d17c4d09c7dd6f192258 100644
--- a/net/minecraft/core/MappedRegistry.java
+++ b/net/minecraft/core/MappedRegistry.java
@@ -341,6 +341,13 @@ public class MappedRegistry<T> implements WritableRegistry<T> {
         }
     }
 
+    // Kiterino start - Injecting custom items extras
+    @Override
+    public boolean isFrozen() {
+        return this.frozen;
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public Holder.Reference<T> createIntrusiveHolder(final T value) {
         if (this.unregisteredIntrusiveHolders == null) {
diff --git a/net/minecraft/core/Registry.java b/net/minecraft/core/Registry.java
index d54db1a3fc89222663495ad02365d044a15c9063..b2794374c3293898f61a5c324979d47137e46baa 100644
--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -128,6 +128,10 @@ public interface Registry<T> extends IdMap<T>, Keyable, HolderLookup.RegistryLoo
 
     Registry<T> freeze();
 
+    // Kiterino start - Injecting custom items extras
+    boolean isFrozen();
+    // Kiterino end - Injecting custom items extras
+
     Holder.Reference<T> createIntrusiveHolder(T value);
 
     Optional<Holder.Reference<T>> get(int id);
diff --git a/net/minecraft/world/item/AxeItem.java b/net/minecraft/world/item/AxeItem.java
index 1ff4f2f3a864efe4b9fcfb4881cafc64b0f3b8df..39b77ca4c9d54f2df947b0178213d6c1c4442700 100644
--- a/net/minecraft/world/item/AxeItem.java
+++ b/net/minecraft/world/item/AxeItem.java
@@ -57,6 +57,12 @@ public class AxeItem extends Item {
         super(properties.axe(material, attackDamageBaseline, attackSpeedBaseline));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public AxeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(final UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/HoeItem.java b/net/minecraft/world/item/HoeItem.java
index 78255f85968481e9cede2b1e47f99d64c86a9263..38beb9b3df0eb87bec1687fbb0261ebe77d886b9 100644
--- a/net/minecraft/world/item/HoeItem.java
+++ b/net/minecraft/world/item/HoeItem.java
@@ -40,6 +40,12 @@ public class HoeItem extends Item {
         super(properties.hoe(material, attackDamageBaseline, attackSpeedBaseline));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public HoeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(final UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ShearsItem.java b/net/minecraft/world/item/ShearsItem.java
index 837f8c56bf5d68dd8bfbeca8ae1b66be8316ddea..b3374fe902600e1a85a153f75a8514668c692d04 100644
--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -30,7 +30,7 @@ public class ShearsItem extends Item {
     }
 
     public static Tool createToolProperties() {
-        HolderGetter<Block> registrationLookup = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> registrationLookup = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return new Tool(
             List.of(
                 Tool.Rule.minesAndDrops(HolderSet.direct(Blocks.COBWEB.builtInRegistryHolder()), 15.0F),
diff --git a/net/minecraft/world/item/ShovelItem.java b/net/minecraft/world/item/ShovelItem.java
index 05650ae28ecca550a7ece09f9d4e631044c933e0..d3cee51b16a695b4589d0625a68294ee41e70660 100644
--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -34,6 +34,12 @@ public class ShovelItem extends Item {
         super(properties.shovel(material, attackDamageBaseline, attackSpeedBaseline));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public ShovelItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(final UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ToolMaterial.java b/net/minecraft/world/item/ToolMaterial.java
index b4dcac7a0d4e598a10017ff370fa1b4071c74737..0fc90d4e8ef36425d282ce51525b2da120dcef2d 100644
--- a/net/minecraft/world/item/ToolMaterial.java
+++ b/net/minecraft/world/item/ToolMaterial.java
@@ -31,6 +31,11 @@ public record ToolMaterial(
     );
 
     private Item.Properties applyCommonProperties(final Item.Properties properties) {
+        // Kiterino start - Injecting custom items extras
+        if (this.repairItems == ItemTags.ANVIL) {
+            return properties;
+        }
+        // Kiterino end - Injecting custom items extras
         return properties.durability(this.durability).repairable(this.repairItems).enchantable(this.enchantmentValue);
     }
 
@@ -41,7 +46,7 @@ public record ToolMaterial(
         final float attackSpeedBaseline,
         final float disableBlockingSeconds
     ) {
-        HolderGetter<Block> registrationLookup = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> registrationLookup = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
@@ -75,7 +80,7 @@ public record ToolMaterial(
     }
 
     public Item.Properties applySwordProperties(final Item.Properties properties, final float attackDamageBaseline, final float attackSpeedBaseline) {
-        HolderGetter<Block> registrationLookup = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> registrationLookup = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
