From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 5 Aug 2025 02:38:09 +0300
Subject: [PATCH] Injecting custom items extras


diff --git a/net/minecraft/core/MappedRegistry.java b/net/minecraft/core/MappedRegistry.java
index c79206e0eb455081f63cf4de2c97413638fd6596..181448d0a83c420a3cfb725ca307ecdff2a51c6d 100644
--- a/net/minecraft/core/MappedRegistry.java
+++ b/net/minecraft/core/MappedRegistry.java
@@ -343,6 +343,13 @@ public class MappedRegistry<T> implements WritableRegistry<T> {
         return this;
     }
 
+    // Kiterino start - Injecting custom items extras
+    @Override
+    public boolean isFrozen() {
+        return this.frozen;
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public Holder.Reference<T> createIntrusiveHolder(final T value) {
         if (this.unregisteredIntrusiveHolders == null) {
diff --git a/net/minecraft/core/Registry.java b/net/minecraft/core/Registry.java
index 38a2f7604ce2b603ea66339675574ea9c80f46ac..364f1cb86c45b9e3957fdf75a6cea19e7f336a24 100644
--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -127,6 +127,10 @@ public interface Registry<T> extends IdMap<T>, Keyable, HolderLookup.RegistryLoo
 
     Registry<T> freeze();
 
+    // Kiterino start - Injecting custom items extras
+    boolean isFrozen();
+    // Kiterino end - Injecting custom items extras
+
     Holder.Reference<T> createIntrusiveHolder(T value);
 
     Optional<Holder.Reference<T>> get(int id);
diff --git a/net/minecraft/world/item/AxeItem.java b/net/minecraft/world/item/AxeItem.java
index 1f225fa6dd016ecd8db13ac2588b90582900103b..cdde9151209412f65f242b161a1005a3a07acc89 100644
--- a/net/minecraft/world/item/AxeItem.java
+++ b/net/minecraft/world/item/AxeItem.java
@@ -57,6 +57,12 @@ public class AxeItem extends Item {
         super(properties.axe(material, attackDamageBaseline, attackSpeedBaseline));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public AxeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(final UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/HoeItem.java b/net/minecraft/world/item/HoeItem.java
index 08632170e32d9c33dcf49d1cbc51461b8ef7db2f..349c4641bab3929cc80572cc6c543487f2ec39ff 100644
--- a/net/minecraft/world/item/HoeItem.java
+++ b/net/minecraft/world/item/HoeItem.java
@@ -40,6 +40,12 @@ public class HoeItem extends Item {
         super(properties.hoe(material, attackDamageBaseline, attackSpeedBaseline));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public HoeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(final UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ShearsItem.java b/net/minecraft/world/item/ShearsItem.java
index c9ecf7347f568b455c31e0a0f6d679e1a460e9bf..f565c732ce564607da4712caad5c9c96cc821a3b 100644
--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -30,7 +30,7 @@ public class ShearsItem extends Item {
     }
 
     public static Tool createToolProperties() {
-        HolderGetter<Block> registrationLookup = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> registrationLookup = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return new Tool(
             List.of(
                 Tool.Rule.minesAndDrops(HolderSet.direct(Blocks.COBWEB.builtInRegistryHolder()), 15.0F),
diff --git a/net/minecraft/world/item/ShovelItem.java b/net/minecraft/world/item/ShovelItem.java
index 46ef115c5ad92e3bfb0b5a8a3eedeadcd20bdef4..ee191c95604a312c1fd2cd731e7dabdaf347d5f6 100644
--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -34,6 +34,12 @@ public class ShovelItem extends Item {
         super(properties.shovel(material, attackDamageBaseline, attackSpeedBaseline));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public ShovelItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(final UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ToolMaterial.java b/net/minecraft/world/item/ToolMaterial.java
index b4dcac7a0d4e598a10017ff370fa1b4071c74737..0fc90d4e8ef36425d282ce51525b2da120dcef2d 100644
--- a/net/minecraft/world/item/ToolMaterial.java
+++ b/net/minecraft/world/item/ToolMaterial.java
@@ -31,6 +31,11 @@ public record ToolMaterial(
     );
 
     private Item.Properties applyCommonProperties(final Item.Properties properties) {
+        // Kiterino start - Injecting custom items extras
+        if (this.repairItems == ItemTags.ANVIL) {
+            return properties;
+        }
+        // Kiterino end - Injecting custom items extras
         return properties.durability(this.durability).repairable(this.repairItems).enchantable(this.enchantmentValue);
     }
 
@@ -41,7 +46,7 @@ public record ToolMaterial(
         final float attackSpeedBaseline,
         final float disableBlockingSeconds
     ) {
-        HolderGetter<Block> registrationLookup = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> registrationLookup = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
@@ -75,7 +80,7 @@ public record ToolMaterial(
     }
 
     public Item.Properties applySwordProperties(final Item.Properties properties, final float attackDamageBaseline, final float attackSpeedBaseline) {
-        HolderGetter<Block> registrationLookup = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> registrationLookup = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
