From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 15 Jul 2025 14:06:00 +0300
Subject: [PATCH] Default pickup loot in summon


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 444c03ca4c285951ceb62643df15ac0f67bdf263..f58864368fcae1bc45f5e44ec6cac35f1b1627a8 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4544,6 +4544,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return false;
     }
 
+    // Kiterino start - Default pickup loot in summon
+    public boolean getCanPickUpLootDefault() {
+        return false;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public final EntityDimensions getDimensions(Pose pose) {
         return pose == Pose.SLEEPING ? SLEEPING_DIMENSIONS : this.getDefaultDimensions(pose).scale(this.getScale());
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 57b2b25a5fe47741325b28fef25a55d8d113b7f0..86351463835c2bcdc4a66cdd468eaf0122d1b9b3 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -472,7 +472,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
         // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
-        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", false);
+        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault()); // Kiterino - Default pickup loot in summon
         if (isLevelAtLeast(input, 1) || canPickUpLoot) {
             this.setCanPickUpLoot(canPickUpLoot);
         }
diff --git a/net/minecraft/world/entity/animal/Dolphin.java b/net/minecraft/world/entity/animal/Dolphin.java
index c790233d6f254a77a42ec5628441345a506c37b0..6410c52b868d4570f51334e13cf498a87f0c04d4 100644
--- a/net/minecraft/world/entity/animal/Dolphin.java
+++ b/net/minecraft/world/entity/animal/Dolphin.java
@@ -115,6 +115,13 @@ public class Dolphin extends AgeableWaterCreature {
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/Fox.java b/net/minecraft/world/entity/animal/Fox.java
index b571ae4fd284806478df5c311a19b2c87cbe903a..41737ad05b80c54fff8b16a279f901715f56e675 100644
--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -142,6 +142,13 @@ public class Fox extends Animal {
         this.getNavigation().setRequiredPathLength(32.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/Panda.java b/net/minecraft/world/entity/animal/Panda.java
index c97ec93732d10197ee978477dc07c097d45bbe15..792a1061d2ea450baf7520be4bac1738f0c761bf 100644
--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -107,6 +107,13 @@ public class Panda extends Animal {
         }
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return !this.isBaby();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index d842d0158f1c685a9350c61158a6eaa36ea9e897..8d0e4590d329171b67fc8186a2a7c1589f060055 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -333,6 +333,13 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         return !this.isOnPickupCooldown() && this.hasItemInHand();
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.canPickUpLoot();
+    }
+    // Kiterino end - Default pickup loot in summon
+
     public boolean hasItemInHand() {
         return !this.getItemInHand(InteractionHand.MAIN_HAND).isEmpty();
     }
diff --git a/net/minecraft/world/entity/monster/AbstractSkeleton.java b/net/minecraft/world/entity/monster/AbstractSkeleton.java
index 14730f53561dec218b0fa6648c8de2b2157165fc..fe9f69a1d3d4801139bdacf32e7996821bf54f79 100644
--- a/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -72,6 +72,13 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         this.setShouldBurnInDay(true); // Purpur - API for any mob to burn daylight
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.skeletons;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     protected void registerGoals() {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
diff --git a/net/minecraft/world/entity/monster/Pillager.java b/net/minecraft/world/entity/monster/Pillager.java
index 8bc07bbf713ebd3afc8c0e62c92b5e11fb424307..22e44ecd18bdbf7014b0728c750b676b1c1056cc 100644
--- a/net/minecraft/world/entity/monster/Pillager.java
+++ b/net/minecraft/world/entity/monster/Pillager.java
@@ -183,6 +183,13 @@ public class Pillager extends AbstractIllager implements CrossbowAttackMob, Inve
         this.setCanPickUpLoot(true);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     @Override
     public float getWalkTargetValue(BlockPos pos, LevelReader level) {
         return 0.0F;
diff --git a/net/minecraft/world/entity/monster/Zombie.java b/net/minecraft/world/entity/monster/Zombie.java
index 7e13f2958f3935077fe25570567d2576f435028c..a6773d16f76796ee8f065b0badf125bdc2246619 100644
--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -103,6 +103,13 @@ public class Zombie extends Monster {
         this(EntityType.ZOMBIE, level);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return this.level() != null && this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.zombies;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
index 9b76d5156309981127adb2ccd98fb666480618bb..db0171607f5497903187dd0e9275369ec7dcae48 100644
--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -37,6 +37,13 @@ public abstract class AbstractPiglin extends Monster {
         this.setPathfindingMalus(PathType.DAMAGE_FIRE, -1.0F);
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     private void applyOpenDoorsAbility() {
         if (GoalUtils.hasGroundPathNavigation(this)) {
             this.getNavigation().setCanOpenDoors(true);
@@ -69,7 +76,7 @@ public abstract class AbstractPiglin extends Monster {
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
-        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", true));
+        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", getCanPickUpLootDefault())); // Kiterino - Default pickup loot in summon
         this.setImmuneToZombification(input.getBooleanOr("IsImmuneToZombification", false));
         this.timeInOverworld = input.getIntOr("TimeInOverworld", 0);
     }
diff --git a/net/minecraft/world/entity/npc/Villager.java b/net/minecraft/world/entity/npc/Villager.java
index 0ab9ac120703285baa563391653323c17d05b20b..34727b69a811f16798002a77ef000f202a9707ae 100644
--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -198,6 +198,13 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         this.setVillagerData(this.getVillagerData().withType(villagerType).withProfession(level.registryAccess(), VillagerProfession.NONE));
     }
 
+    // Kiterino start - Default pickup loot in summon
+    @Override
+    public boolean getCanPickUpLootDefault() {
+        return true;
+    }
+    // Kiterino end - Default pickup loot in summon
+
     // Purpur start - Allow leashing villagers
     @Override
     public boolean canBeLeashed() {
