From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 5 Aug 2025 02:38:09 +0300
Subject: [PATCH] Injecting custom items extras


diff --git a/net/minecraft/core/MappedRegistry.java b/net/minecraft/core/MappedRegistry.java
index 1610198bd96e0d79debdeddc360500e98cd28504..29e1fda36bd68587f17e2a17be69a406f0a63d00 100644
--- a/net/minecraft/core/MappedRegistry.java
+++ b/net/minecraft/core/MappedRegistry.java
@@ -339,6 +339,13 @@ public class MappedRegistry<T> implements WritableRegistry<T> {
         }
     }
 
+    // Kiterino start - Injecting custom items extras
+    @Override
+    public boolean isFrozen() {
+        return this.frozen;
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public Holder.Reference<T> createIntrusiveHolder(T value) {
         if (this.unregisteredIntrusiveHolders == null) {
diff --git a/net/minecraft/core/Registry.java b/net/minecraft/core/Registry.java
index a032a2ce71707e1d1e46a1acf0a63a3bdb1c3db8..3307642782aaa49a4a05a07d61d299ecc860e862 100644
--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -131,6 +131,10 @@ public interface Registry<T> extends Keyable, HolderLookup.RegistryLookup<T>, Id
 
     Registry<T> freeze();
 
+    // Kiterino start - Injecting custom items extras
+    boolean isFrozen();
+    // Kiterino end - Injecting custom items extras
+
     Holder.Reference<T> createIntrusiveHolder(T value);
 
     Optional<Holder.Reference<T>> get(int index);
diff --git a/net/minecraft/world/item/AxeItem.java b/net/minecraft/world/item/AxeItem.java
index b0b96dd52747e6e3e63f0ad475db55357e6d4916..3b8ff764862b4bb33fcfd32519d8cbccbd4c5b29 100644
--- a/net/minecraft/world/item/AxeItem.java
+++ b/net/minecraft/world/item/AxeItem.java
@@ -56,6 +56,12 @@ public class AxeItem extends Item {
         super(properties.axe(material, attackDamage, attackSpeed));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public AxeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/HoeItem.java b/net/minecraft/world/item/HoeItem.java
index 9b5cf74989f41f95476d583079c778eebd707dff..42081d243a85fbf60fb8add720d523a235f60503 100644
--- a/net/minecraft/world/item/HoeItem.java
+++ b/net/minecraft/world/item/HoeItem.java
@@ -40,6 +40,12 @@ public class HoeItem extends Item {
         super(properties.hoe(material, attackDamage, attackSpeed));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public HoeItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ShearsItem.java b/net/minecraft/world/item/ShearsItem.java
index 5e25ddf78f72ada318ef28d29199de9b93fe667e..df5d528cea828841cfac3ed61f91bac85f157d26 100644
--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -30,7 +30,7 @@ public class ShearsItem extends Item {
     }
 
     public static Tool createToolProperties() {
-        HolderGetter<Block> holderGetter = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> holderGetter = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return new Tool(
             List.of(
                 Tool.Rule.minesAndDrops(HolderSet.direct(Blocks.COBWEB.builtInRegistryHolder()), 15.0F),
diff --git a/net/minecraft/world/item/ShovelItem.java b/net/minecraft/world/item/ShovelItem.java
index 013f155677ab4a7588bd7ad1d612ce582bbd7927..169dc74642613b34f42352d3688fbdc0e53cfb8a 100644
--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -33,6 +33,12 @@ public class ShovelItem extends Item {
         super(properties.shovel(material, attackDamage, attackSpeed));
     }
 
+    // Kiterino start - Injecting custom items extras
+    public ShovelItem(Item.Properties properties) {
+        super(properties);
+    }
+    // Kiterino end - Injecting custom items extras
+
     @Override
     public InteractionResult useOn(UseOnContext context) {
         Level level = context.getLevel();
diff --git a/net/minecraft/world/item/ToolMaterial.java b/net/minecraft/world/item/ToolMaterial.java
index 5a37b3a4aca3f0e0976fa6ef2b4be0c921db1dad..118392f46c6227295ee45872fe216fbe58235483 100644
--- a/net/minecraft/world/item/ToolMaterial.java
+++ b/net/minecraft/world/item/ToolMaterial.java
@@ -31,13 +31,18 @@ public record ToolMaterial(
     );
 
     private Item.Properties applyCommonProperties(Item.Properties properties) {
+        // Kiterino start - Injecting custom items extras
+        if (this.repairItems == ItemTags.ANVIL) {
+            return properties;
+        }
+        // Kiterino end - Injecting custom items extras
         return properties.durability(this.durability).repairable(this.repairItems).enchantable(this.enchantmentValue);
     }
 
     public Item.Properties applyToolProperties(
         Item.Properties properties, TagKey<Block> mineableBlocks, float attackDamage, float attackSpeed, float disableBlockingForSeconds
     ) {
-        HolderGetter<Block> holderGetter = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> holderGetter = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
@@ -71,7 +76,7 @@ public record ToolMaterial(
     }
 
     public Item.Properties applySwordProperties(Item.Properties properties, float attackDamage, float attackSpeed) {
-        HolderGetter<Block> holderGetter = BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK);
+        HolderGetter<Block> holderGetter = BuiltInRegistries.BLOCK.isFrozen() ? BuiltInRegistries.BLOCK : BuiltInRegistries.acquireBootstrapRegistrationLookup(BuiltInRegistries.BLOCK); // Kiterino - Injecting custom items extras
         return this.applyCommonProperties(properties)
             .component(
                 DataComponents.TOOL,
