From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 22 Oct 2025 14:53:13 +0300
Subject: [PATCH] Correctly handle insta break destroy progress


diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 81aec62cfb7a9689667030bc1b0f392fe02a76c8..9d61cfef0227876b5da7345a8677027ca6dbe2c1 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -364,6 +364,11 @@ public abstract class BlockBehaviour implements FeatureElement {
 
     protected float getDestroyProgress(final BlockState state, final Player player, final BlockGetter level, final BlockPos pos) {
         float destroySpeed = state.getDestroySpeed(level, pos);
+        // Kiterino start - Correctly handle insta break destroy progress
+        if (destroySpeed == 0F) {
+            return 1F;
+        }
+        // Kiterino end - Correctly handle insta break destroy progress
         if (destroySpeed == -1.0F) {
             return 0.0F;
         } else {
