From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 31 Oct 2025 21:03:45 +0200
Subject: [PATCH] Allow AI opening fence gates


diff --git a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
index 3d7026568c6ab3ae090f0f3fc4bc03fe8c883470..9a25862be99ee089338668ddab0264f73e7a7741 100644
--- a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -67,6 +67,22 @@ public class InteractWithDoor {
 
                                 doors = rememberDoorToClose(doorsMemory, doors, level, fromPos);
                             }
+                            // Kiterino start - Allow AI opening fence gates
+                            else if (level.kiterinoConfig.doorOpenAIAffectsGates && fromState.getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock) {
+                                net.minecraft.world.level.block.FenceGateBlock doorBlock = (net.minecraft.world.level.block.FenceGateBlock) fromState.getBlock();
+                                if (!fromState.getValue(net.minecraft.world.level.block.FenceGateBlock.OPEN)) {
+                                    // CraftBukkit start - entities opening doors
+                                    org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(body.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(body.level(), fromPos));
+                                    if (!event.callEvent()) {
+                                        return false;
+                                    }
+                                    // CraftBukkit end - entities opening doors
+                                    doorBlock.setOpen(body, level, fromState, fromPos, true);
+                                }
+
+                                doors = rememberDoorToClose(doorsMemory, doors, level, fromPos);
+                            }
+                            // Kiterino end - Allow AI opening fence gates
 
                             BlockPos toPos = toNode.asBlockPos();
                             BlockState toState = level.getBlockState(toPos);
@@ -83,6 +99,21 @@ public class InteractWithDoor {
                                     doors = rememberDoorToClose(doorsMemory, doors, level, toPos);
                                 }
                             }
+                            // Kiterino start - Allow AI opening fence gates
+                            else if (level.kiterinoConfig.doorOpenAIAffectsGates && toState.getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock) {
+                                net.minecraft.world.level.block.FenceGateBlock doorBlock = (net.minecraft.world.level.block.FenceGateBlock) toState.getBlock();
+                                if (!toState.getValue(net.minecraft.world.level.block.FenceGateBlock.OPEN)) {
+                                    // CraftBukkit start - entities opening doors
+                                    org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(body.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(body.level(), toPos));
+                                    if (!event.callEvent()) {
+                                        return false;
+                                    }
+                                    // CraftBukkit end - entities opening doors
+                                    doorBlock.setOpen(body, level, toState, toPos, true);
+                                    doors = rememberDoorToClose(doorsMemory, doors, level, toPos);
+                                }
+                            }
+                            // Kiterino end - Allow AI opening fence gates
 
                             doors.ifPresent(
                                 doorSet -> closeDoorsThatIHaveOpenedOrPassedThrough(
@@ -117,6 +148,19 @@ public class InteractWithDoor {
                     iterator.remove();
                 } else {
                     BlockState state = level.getBlockState(doorPos);
+                    // Kiterino start - Allow AI opening fence gates
+                    if (level.kiterinoConfig.doorOpenAIAffectsGates && state.getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock) {
+                        if (!state.getValue(net.minecraft.world.level.block.FenceGateBlock.OPEN)) {
+                            iterator.remove();
+                        } else if (areOtherMobsComingThroughDoor(body, doorPos, nearestEntities)) {
+                            iterator.remove();
+                        } else {
+                            ((net.minecraft.world.level.block.FenceGateBlock) state.getBlock()).setOpen(body, level, state, doorPos, false);
+                            iterator.remove();
+                        }
+                        continue;
+                    }
+                    // Kiterino end - Allow AI opening fence gates
                     if (!state.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock) || DoorBlock.requiresRedstone(body.level(), state, doorPos)) { // Purpur - Option to make doors require redstone
                         iterator.remove();
                     } else {
diff --git a/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java b/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
index 0b383b16145d44e5dc29b8de123546656b4793dd..2ebf75064a2b1fb088e0f358e4833ccf968ed9f4 100644
--- a/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
+++ b/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
@@ -29,6 +29,11 @@ public abstract class DoorInteractGoal extends Goal {
         } else {
             BlockState blockState = this.mob.level().getBlockState(this.doorPos);
             if (!(blockState.getBlock() instanceof DoorBlock)) {
+                // Kiterino start - Allow AI opening fence gates
+                if (this.mob.level().kiterinoConfig.doorOpenAIAffectsGates && blockState.getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock) {
+                    return blockState.getValue(net.minecraft.world.level.block.FenceGateBlock.OPEN);
+                }
+                // Kiterino end - Allow AI opening fence gates
                 this.hasDoor = false;
                 return false;
             } else {
@@ -43,6 +48,11 @@ public abstract class DoorInteractGoal extends Goal {
             if (blockState.getBlock() instanceof DoorBlock) {
                 ((DoorBlock)blockState.getBlock()).setOpen(this.mob, this.mob.level(), blockState, this.doorPos, open);
             }
+            // Kiterino start - Allow AI opening fence gates
+            else if (this.mob.level().kiterinoConfig.doorOpenAIAffectsGates && blockState.getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock) {
+                ((net.minecraft.world.level.block.FenceGateBlock) blockState.getBlock()).setOpen(this.mob, this.mob.level(), blockState, this.doorPos, open);
+            }
+            // Kiterino end - Allow AI opening fence gates
         }
     }
 
@@ -63,6 +73,11 @@ public abstract class DoorInteractGoal extends Goal {
                 this.doorPos = new BlockPos(node.x, node.y + 1, node.z);
                 if (!(this.mob.distanceToSqr(this.doorPos.getX(), this.mob.getY(), this.doorPos.getZ()) > 2.25)) {
                     this.hasDoor = DoorBlock.isWoodenDoor(this.mob.level(), this.doorPos);
+                    // Kiterino start - Allow AI opening fence gates
+                    if (this.mob.level().kiterinoConfig.doorOpenAIAffectsGates && !this.hasDoor) {
+                        this.hasDoor = this.mob.level().getBlockState(this.doorPos).getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock;
+                    }
+                    // Kiterino end - Allow AI opening fence gates
                     if (this.hasDoor) {
                         return true;
                     }
@@ -71,6 +86,11 @@ public abstract class DoorInteractGoal extends Goal {
 
             this.doorPos = this.mob.blockPosition().above();
             this.hasDoor = DoorBlock.isWoodenDoor(this.mob.level(), this.doorPos);
+            // Kiterino start - Allow AI opening fence gates
+            if (this.mob.level().kiterinoConfig.doorOpenAIAffectsGates && !this.hasDoor) {
+                this.hasDoor = this.mob.level().getBlockState(this.doorPos).getBlock() instanceof net.minecraft.world.level.block.FenceGateBlock;
+            }
+            // Kiterino end - Allow AI opening fence gates
             return this.hasDoor;
         } else {
             return false;
diff --git a/net/minecraft/world/level/block/FenceGateBlock.java b/net/minecraft/world/level/block/FenceGateBlock.java
index 9ca0a8ae74a0a6a1c52d8ce70fd28a01ffe6cffb..f3635f74d3457bbf7b6a5bbce17c4106a9462667 100644
--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -214,4 +214,21 @@ public class FenceGateBlock extends HorizontalDirectionalBlock {
     public static boolean connectsToDirection(final BlockState state, final Direction direction) {
         return state.getValue(FACING).getAxis() == direction.getClockWise().getAxis();
     }
+
+    // Kiterino start - Allow AI opening fence gates
+    // Copy-paste from DoorBlock
+    public void setOpen(net.minecraft.world.entity.@Nullable Entity entity, Level level, BlockState state, BlockPos pos, boolean _open) {
+        if (state.is(this) && state.getValue(OPEN) != _open) {
+            level.setBlock(pos, state.setValue(OPEN, _open), Block.UPDATE_CLIENTS | Block.UPDATE_IMMEDIATE);
+            this.playSound(entity, level, pos, _open);
+            level.gameEvent(entity, _open ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
+        }
+    }
+
+    private void playSound(net.minecraft.world.entity.@Nullable Entity source, Level level, BlockPos pos, boolean isOpening) {
+        level.playSound(
+            source, pos, isOpening ? this.type.fenceGateOpen() : this.type.fenceGateClose(), SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
+        );
+    }
+    // Kiterino end - Allow AI opening fence gates
 }
diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index 29da2af5d4a09f3ff21fa2d0653c05d45c475aa6..13dcc4a503872ff8b5d8f4c4f6ab258e9e713ecc 100644
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -562,6 +562,11 @@ public class WalkNodeEvaluator extends NodeEvaluator {
                     return door.type().canOpenByHand() ? PathType.DOOR_WOOD_CLOSED : PathType.DOOR_IRON_CLOSED;
                 }
             } else {
+                // Kiterino start - Allow AI opening fence gates
+                if (level instanceof net.minecraft.server.level.ServerLevel serverLevel && serverLevel.kiterinoConfig.doorOpenAIAffectsGates && block instanceof FenceGateBlock && !blockState.getValue(FenceGateBlock.OPEN)) {
+                    return PathType.DOOR_WOOD_CLOSED;
+                }
+                // Kiterino end - Allow AI opening fence gates
                 if (block instanceof BaseRailBlock) {
                     return PathType.RAIL;
                 }
