From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 16 Nov 2025 14:25:15 +0200
Subject: [PATCH] Add highly questionable BlockDispenseEvent additions


diff --git a/net/minecraft/core/dispenser/DefaultDispenseItemBehavior.java b/net/minecraft/core/dispenser/DefaultDispenseItemBehavior.java
index bcd998d6e9d56929b50563d79838d7b9065c4a0e..68994962c6845f2cf59b282c1f03abfae1645f04 100644
--- a/net/minecraft/core/dispenser/DefaultDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DefaultDispenseItemBehavior.java
@@ -27,7 +27,12 @@ public class DefaultDispenseItemBehavior implements DispenseItemBehavior {
         Position position = DispenserBlock.getDispensePosition(source);
         ItemStack itemStack = dispensed.split(1);
         // CraftBukkit start
-        if (!DefaultDispenseItemBehavior.spawnItem(source.level(), itemStack, 6, this.direction, position, source)) {
+        // Kiterino start - More BlockDispenseEvent API
+        ItemStack result = DefaultDispenseItemBehavior.spawnItem(source.level(), itemStack, itemStack, 6, this.direction, position, source);
+        if (result != null) {
+            return result;
+        } else {
+            // Kiterino end - More BlockDispenseEvent API
             dispensed.grow(1);
         }
         // CraftBukkit end
@@ -63,8 +68,8 @@ public class DefaultDispenseItemBehavior implements DispenseItemBehavior {
     }
 
     // CraftBukkit start - void -> boolean return
-    public static boolean spawnItem(Level level, ItemStack itemStack, int accuracy, Direction direction, Position position, BlockSource source) {
-        if (itemStack.isEmpty()) return true;
+    public static @org.jspecify.annotations.Nullable ItemStack spawnItem(Level level, ItemStack original, ItemStack itemStack, int accuracy, Direction direction, Position position, BlockSource source) { // Kiterino - Change return type, expose original - More BlockDispenseEvent API
+        if (itemStack.isEmpty()) return itemStack;
         ItemEntity itemEntity = DefaultDispenseItemBehavior.prepareItem(level, itemStack, accuracy, direction, position);
 
         org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(level, source.pos());
@@ -74,27 +79,37 @@ public class DefaultDispenseItemBehavior implements DispenseItemBehavior {
         level.getCraftServer().getPluginManager().callEvent(event);
 
         if (event.isCancelled()) {
-            return false;
+            return null; // Kiterino - More BlockDispenseEvent API
         }
 
-        itemEntity.setItem(org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem()));
+        // Kiterino start - More BlockDispenseEvent API
+        if (source.state().is(net.minecraft.world.level.block.Blocks.DISPENSER) && event.getLeftoverItem() != null) {
+            if (original.isEmpty()) {
+                return ItemStack.fromBukkitCopy(event.getLeftoverItem());
+            }
+            addToInventoryOrDispenseS(source, ItemStack.fromBukkitCopy(event.getLeftoverItem()));
+            return itemStack;
+        }
+        // Kiterino end - More BlockDispenseEvent API
+
+        itemEntity.setItem(org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDispensedItem())); // Kiterino - More BlockDispenseEvent API
         itemEntity.setDeltaMovement(org.bukkit.craftbukkit.util.CraftVector.toVec3(event.getVelocity()));
 
         if (source.state().is(net.minecraft.world.level.block.Blocks.DISPENSER) && !event.getItem().getType().equals(craftItem.getType())) {
             // Chain to handler for new item
-            ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+            ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDispensedItem()); // Kiterino - More BlockDispenseEvent API
             DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(source, eventStack);
             if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior.getClass() != DefaultDispenseItemBehavior.class) {
                 dispenseBehavior.dispense(source, eventStack);
             } else {
                 level.addFreshEntity(itemEntity);
             }
-            return false;
+            return null; // Kiterino - More BlockDispenseEvent API
         }
 
         level.addFreshEntity(itemEntity);
 
-        return true;
+        return original; // Kiterino - More BlockDispenseEvent API
         // CraftBukkit end
     }
 
@@ -124,7 +139,12 @@ public class DefaultDispenseItemBehavior implements DispenseItemBehavior {
         return dispensed;
     }
 
+    // Kiterino start - More BlockDispenseEvent API
     private void addToInventoryOrDispense(final BlockSource source, final ItemStack itemStack) {
+        addToInventoryOrDispenseS(source, itemStack);
+    }
+    private static void addToInventoryOrDispenseS(BlockSource source, ItemStack itemStack) {
+        // Kiterino end - More BlockDispenseEvent API
         ItemStack remainder = source.blockEntity().insertItem(itemStack);
         if (!remainder.isEmpty()) {
             Direction direction = source.state().getValue(DispenserBlock.FACING);
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index acc5a81c0193cac07959210f232a1564875c0ed5..354eb92c45cb384ba6b98c7226bbcebe0287d239 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -103,7 +103,7 @@ public interface DispenseItemBehavior {
                     if (!event.getItem().equals(craftItem)) {
                         shrink = false;
                         // Chain to handler for new item
-                        ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                        ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDispensedItem()); // Kiterino - More BlockDispenseEvent API
                         DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(source, eventStack);
                         if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
                             dispenseBehavior.dispense(source, eventStack);
@@ -152,7 +152,7 @@ public interface DispenseItemBehavior {
                                 if (!event.getItem().equals(craftItem)) {
                                     shrink = false;
                                     // Chain to handler for new item
-                                    ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                                    ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDispensedItem()); // Kiterino - More BlockDispenseEvent API
                                     DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(source, eventStack);
                                     if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
                                         dispenseBehavior.dispense(source, eventStack);
@@ -209,21 +209,28 @@ public interface DispenseItemBehavior {
                 boolean willEmptyContentsSolidBucketItem = bucket instanceof net.minecraft.world.item.SolidBucketItem && level.isInWorldBounds(target) && state.isAir();
                 /* Taken from BucketItem#emptyContents */
                 boolean willEmptyBucketItem = bucket instanceof final net.minecraft.world.item.BucketItem bucketItem && bucketItem.content instanceof net.minecraft.world.level.material.FlowingFluid && (state.isAir() || state.canBeReplaced(bucketItem.content) || (state.getBlock() instanceof net.minecraft.world.level.block.LiquidBlockContainer liquidBlockContainer && liquidBlockContainer.canPlaceLiquid(null, level, target, state, bucketItem.content)));
+                ItemStack leftover = new ItemStack(Items.BUCKET); // Kiterino - More BlockDispenseEvent API
                 if (willEmptyContentsSolidBucketItem || willEmptyBucketItem) {
                 // Paper end - correctly check if the bucket place will succeed
                     org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(level, source.pos());
                     org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(dispensed.copyWithCount(1)); // Paper - single item in event
 
                     org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), org.bukkit.craftbukkit.util.CraftVector.toBukkit(target));
+                    event.setLeftoverItem(leftover.asBukkitCopy()); // Kiterino - More BlockDispenseEvent API
                     level.getCraftServer().getPluginManager().callEvent(event);
 
                     if (event.isCancelled()) {
                         return dispensed;
                     }
+                    // Kiterino start - More BlockDispenseEvent API
+                    if (event.getLeftoverItem() != null) {
+                        leftover = ItemStack.fromBukkitCopy(event.getLeftoverItem());
+                    }
+                    // Kiterino end - More BlockDispenseEvent API
 
                     if (!event.getItem().equals(craftItem)) {
                         // Chain to handler for new item
-                        ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                        ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDispensedItem()); // Kiterino - More BlockDispenseEvent API
                         DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(source, eventStack);
                         if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
                             dispenseBehavior.dispense(source, eventStack);
@@ -240,7 +247,7 @@ public interface DispenseItemBehavior {
 
                 if (bucket.emptyContents(null, level, target, null)) {
                     bucket.checkExtraContent(null, level, dispensedItem, target); // Paper - track changed item from dispense event
-                    return this.consumeWithRemainder(source, dispensed, new ItemStack(Items.BUCKET));
+                    return this.consumeWithRemainder(source, dispensed, leftover); // Kiterino - More BlockDispenseEvent API
                 } else {
                     return this.defaultDispenseItemBehavior.dispense(source, dispensed);
                 }
@@ -270,13 +277,17 @@ public interface DispenseItemBehavior {
                     level.gameEvent(null, GameEvent.FLUID_PICKUP, target);
                     Item targetType = pickup.getItem();
                     // Paper start - Call BlockDispenseEvent
-                    ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this);
-                    if (result != null) {
-                        return result;
+                    // Kiterino start - More BlockDispenseEvent API
+                    var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this, event -> {
+                        event.setLeftoverItem(pickup.asBukkitCopy());
+                    });
+                    if (result.dispensedItem() != null) {
+                        return result.dispensedItem();
                     }
                     // Paper end - Call BlockDispenseEvent
-                    pickup = bucket.pickupBlock(null, level, target, blockState); // CraftBukkit - from above
-                    return this.consumeWithRemainder(source, dispensed, new ItemStack(targetType));
+                    bucket.pickupBlock(null, level, target, blockState); // CraftBukkit - from above
+                    return result.event().getLeftoverItem() != null ? this.consumeWithRemainder(source, dispensed, ItemStack.fromBukkitCopy(result.event().getLeftoverItem())) : pickup;
+                    // Kiterino end - More BlockDispenseEvent API
                 } else {
                     return super.execute(source, dispensed);
                 }
@@ -290,11 +301,13 @@ public interface DispenseItemBehavior {
                 Direction facing = source.state().getValue(DispenserBlock.FACING);
                 BlockPos targetPos = source.pos().relative(facing);
                 // Paper start - Call BlockDispenseEvent
-                ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, targetPos, dispensed, this);
-                if (result != null) {
+                // Kiterino start - More BlockDispenseEvent API
+                var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, targetPos, dispensed, this, null);
+                if (result.dispensedItem() != null) {
                     this.setSuccess(false);
-                    return result;
+                    return result.dispensedItem();
                 }
+                // Kiterino end - More BlockDispenseEvent API
                 // Paper end - Call BlockDispenseEvent
                 BlockState target = level.getBlockState(targetPos);
                 if (BaseFireBlock.canBePlacedAt(level, targetPos, facing)) {
@@ -331,11 +344,13 @@ public interface DispenseItemBehavior {
                 Level level = source.level();
                 BlockPos target = source.pos().relative(source.state().getValue(DispenserBlock.FACING));
                 // Paper start - Call BlockDispenseEvent
-                ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this);
-                if (result != null) {
+                // Kiterino start - More BlockDispenseEvent API
+                var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this, null);
+                if (result.dispensedItem() != null) {
                     this.setSuccess(false);
-                    return result;
+                    return result.dispensedItem();
                 }
+                // Kiterino end - More BlockDispenseEvent API
                 // Paper end - Call BlockDispenseEvent
                 level.captureTreeGeneration = true; // CraftBukkit
                 if (!BoneMealItem.growCrop(dispensed, level, target) && !BoneMealItem.growWaterPlant(dispensed, level, target, null)) {
@@ -399,7 +414,7 @@ public interface DispenseItemBehavior {
                     if (!event.getItem().equals(craftItem)) {
                         shrink = false;
                         // Chain to handler for new item
-                        ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                        ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDispensedItem()); // Kiterino - More BlockDispenseEvent API
                         DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(source, eventStack);
                         if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
                             dispenseBehavior.dispense(source, eventStack);
@@ -427,11 +442,13 @@ public interface DispenseItemBehavior {
                     Direction direction = source.state().getValue(DispenserBlock.FACING);
                     BlockPos target = source.pos().relative(direction);
                     // Paper start - Call BlockDispenseEvent
-                    ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this);
-                    if (result != null) {
+                    // Kiterino start - More BlockDispenseEvent API
+                    var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this, null);
+                    if (result.dispensedItem() != null) {
                         this.setSuccess(false);
-                        return result;
+                        return result.dispensedItem();
                     }
+                    // Kiterino end - More BlockDispenseEvent API
                     // Paper end - Call BlockDispenseEvent
                     if (level.isEmptyBlock(target) && WitherSkullBlock.canSpawnMob(level, target, dispensed)) {
                         level.setBlock(
@@ -462,11 +479,13 @@ public interface DispenseItemBehavior {
                 BlockPos target = source.pos().relative(source.state().getValue(DispenserBlock.FACING));
                 CarvedPumpkinBlock pumpkinBlock = (CarvedPumpkinBlock)Blocks.CARVED_PUMPKIN;
                 // Paper start - Call BlockDispenseEvent
-                ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this);
-                if (result != null) {
+                // Kiterino start - More BlockDispenseEvent API
+                var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this, null);
+                if (result.dispensedItem() != null) {
                     this.setSuccess(false);
-                    return result;
+                    return result.dispensedItem();
                 }
+                // Kiterino end - More BlockDispenseEvent API
                 // Paper end - Call BlockDispenseEvent
                 if (level.isEmptyBlock(target) && pumpkinBlock.canSpawnGolem(level, target)) {
                     if (!level.isClientSide()) {
@@ -516,10 +535,12 @@ public interface DispenseItemBehavior {
                     BlockPos target = source.pos().relative(source.state().getValue(DispenserBlock.FACING));
                     BlockState state = level.getBlockState(target);
                     // Paper start - Call BlockDispenseEvent
-                    ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this);
-                    if (result != null) {
-                        return result;
+                    // Kiterino start - More BlockDispenseEvent API
+                    var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this, null);
+                    if (result.dispensedItem() != null) {
+                        return result.dispensedItem();
                     }
+                    // Kiterino end - More BlockDispenseEvent API
                     // Paper end - Call BlockDispenseEvent
                     if (state.is(BlockTags.BEEHIVES, s -> s.hasProperty(BeehiveBlock.HONEY_LEVEL) && s.getBlock() instanceof BeehiveBlock)
                         && state.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
@@ -559,11 +580,13 @@ public interface DispenseItemBehavior {
                 if (blockState.is(Blocks.RESPAWN_ANCHOR)) {
                     if (blockState.getValue(RespawnAnchorBlock.CHARGE) != 4) {
                         // Paper start - Call BlockDispenseEvent
-                        ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, pos, dispensed, this);
-                        if (result != null) {
+                        // Kiterino start - More BlockDispenseEvent API
+                        var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, pos, dispensed, this, null);
+                        if (result.dispensedItem() != null) {
                             this.setSuccess(false);
-                            return result;
+                            return result.dispensedItem();
                         }
+                        // Kiterino end - More BlockDispenseEvent API
                         // Paper end - Call BlockDispenseEvent
                         RespawnAnchorBlock.charge(null, level, pos, blockState);
                         dispensed.shrink(1);
@@ -603,7 +626,7 @@ public interface DispenseItemBehavior {
 
                 if (!event.getItem().equals(craftItem)) {
                     // Chain to handler for new item
-                    ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDispensedItem()); // Kiterino - More BlockDispenseEvent API
                     DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(source, eventStack);
                     if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
                         dispenseBehavior.dispense(source, eventStack);
@@ -631,11 +654,13 @@ public interface DispenseItemBehavior {
                 Optional<BlockState> maybeWaxed = HoneycombItem.getWaxed(blockState);
                 if (maybeWaxed.isPresent()) {
                     // Paper start - Call BlockDispenseEvent
-                    ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, pos, dispensed, this);
-                    if (result != null) {
+                    // Kiterino start - More BlockDispenseEvent API
+                    var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, pos, dispensed, this, null);
+                    if (result.dispensedItem() != null) {
                         this.setSuccess(false);
-                        return result;
+                        return result.dispensedItem();
                     }
+                    // Kiterino end - More BlockDispenseEvent API
                     // Paper end - Call BlockDispenseEvent
                     level.setBlockAndUpdate(pos, maybeWaxed.get());
                     level.levelEvent(LevelEvent.PARTICLES_AND_SOUND_WAX_ON, pos, 0);
@@ -667,10 +692,14 @@ public interface DispenseItemBehavior {
                     }
 
                     // Paper start - Call BlockDispenseEvent
-                    ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this);
-                    if (result != null) {
-                        return result;
+                    // Kiterino start - More BlockDispenseEvent API
+                    var result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, target, dispensed, this, event -> {
+                        event.setLeftoverItem(new ItemStack(Items.GLASS_BOTTLE).asBukkitMirror());
+                    });
+                    if (result.dispensedItem() != null) {
+                        return result.dispensedItem();
                     }
+                    // Kiterino end - More BlockDispenseEvent API
                     // Paper end - Call BlockDispenseEvent
                     if (!level.isClientSide()) {
                         RandomSource random = level.getRandom();
@@ -685,7 +714,7 @@ public interface DispenseItemBehavior {
                     level.playSound(null, pos, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                     level.gameEvent(null, GameEvent.FLUID_PLACE, pos);
                     level.setBlockAndUpdate(target, Blocks.MUD.defaultBlockState());
-                    return this.consumeWithRemainder(source, dispensed, new ItemStack(Items.GLASS_BOTTLE));
+                    return result.event().getLeftoverItem() != null ? this.consumeWithRemainder(source, dispensed, ItemStack.fromBukkitCopy(result.event().getLeftoverItem())) : new ItemStack(Items.GLASS_BOTTLE); // Kiterino - More BlockDispenseEvent API
                 }
             }
         );
