From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 16 Nov 2025 23:45:31 +0200
Subject: [PATCH] Extra bucketable mobs


diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index f2f47f02355f3db911a7137bc347ae512473d9c8..b9ebbd50a94f774b208b12d654d10586a34f5f7e 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -209,8 +209,9 @@ public interface DispenseItemBehavior {
                 boolean willEmptyContentsSolidBucketItem = bucket instanceof net.minecraft.world.item.SolidBucketItem && level.isInWorldBounds(target) && state.isAir();
                 /* Taken from BucketItem#emptyContents */
                 boolean willEmptyBucketItem = bucket instanceof final net.minecraft.world.item.BucketItem bucketItem && bucketItem.content instanceof net.minecraft.world.level.material.FlowingFluid && (state.isAir() || state.canBeReplaced(bucketItem.content) || (state.getBlock() instanceof net.minecraft.world.level.block.LiquidBlockContainer liquidBlockContainer && liquidBlockContainer.canPlaceLiquid(null, level, target, state, bucketItem.content)));
+                boolean willPlaceCustomMob = bucket instanceof net.minecraft.world.item.MobBucketItem bucketItem && bucketItem.content == net.minecraft.world.level.material.Fluids.EMPTY; // Kiterino - Extra bucketable mobs
                 ItemStack leftover = new ItemStack(Items.BUCKET); // Kiterino - More BlockDispenseEvent API
-                if (willEmptyContentsSolidBucketItem || willEmptyBucketItem) {
+                if (willEmptyContentsSolidBucketItem || willEmptyBucketItem || willPlaceCustomMob) { // Kiterino - Extra bucketable mobs
                 // Paper end - correctly check if the bucket place will succeed
                     org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(level, source.pos());
                     org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(dispensed.copyWithCount(1)); // Paper - single item in event
@@ -253,6 +254,7 @@ public interface DispenseItemBehavior {
                 }
             }
         };
+        net.minecraft.world.item.MobBucketItem.BUCKET_DISPENSE_BEHAVIOR = filledBucketBehavior; // Kiterino - Extra bucketable mobs
         DispenserBlock.registerBehavior(Items.LAVA_BUCKET, filledBucketBehavior);
         DispenserBlock.registerBehavior(Items.WATER_BUCKET, filledBucketBehavior);
         DispenserBlock.registerBehavior(Items.POWDER_SNOW_BUCKET, filledBucketBehavior);
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 263f9859df1794d29f28f233585af6250cebea1a..05788d6a623596096377aa5640e88a2b8d037491 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1366,6 +1366,12 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     }
 
     protected InteractionResult mobInteract(final Player player, final InteractionHand hand) {
+        // Kiterino start - Extra bucketable mobs // TODO Client sends block interact after, which auto releases the mob; currently worked around in a plugin
+        // Optional<InteractionResult> result = net.minecraft.world.entity.animal.Bucketable.bucketMobPickup(player, hand, this);
+        // if (result.isPresent()) {
+        //     return result.get();
+        // }
+        // Kiterino end - Extra bucketable mobs
         return tryRide(player, hand); // Purpur - Ridables
     }
 
diff --git a/net/minecraft/world/entity/animal/Bucketable.java b/net/minecraft/world/entity/animal/Bucketable.java
index d1ab91fafdc3971324dc17983504d3c3430f7160..53ebdd37e39a1ad91f2ebc29bc169abba8a48757 100644
--- a/net/minecraft/world/entity/animal/Bucketable.java
+++ b/net/minecraft/world/entity/animal/Bucketable.java
@@ -30,6 +30,24 @@ public interface Bucketable {
 
     SoundEvent getPickupSound();
 
+    // Kiterino start - Extra bucketable mobs
+    default net.minecraft.world.item.Item getPickBucket() {
+        return Items.WATER_BUCKET;
+    }
+
+    default boolean canPickup(LivingEntity entity) {
+        return true;
+    }
+
+    default void saveToBucketTag(LivingEntity entity, ItemStack stack) {
+        saveToBucketTag(stack);
+    }
+
+    default void loadFromBucketTag(LivingEntity entity, CompoundTag tag) {
+        loadFromBucketTag(tag);
+    }
+    // Kiterino end - Extra bucketable mobs
+
     @Deprecated
     static void saveDefaultDataToBucketTag(final Mob entity, final ItemStack bucket) {
         bucket.copyFrom(DataComponents.CUSTOM_NAME, entity);
@@ -68,15 +86,34 @@ public interface Bucketable {
         tag.getFloat("Health").ifPresent(entity::setHealth);
     }
 
-    static <T extends LivingEntity & Bucketable> Optional<InteractionResult> bucketMobPickup(
+    static <T extends LivingEntity> Optional<InteractionResult> bucketMobPickup( // Kiterino - Extra bucketable mobs
         final Player player, final InteractionHand hand, final T pickupEntity
     ) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.getItem() == Items.WATER_BUCKET && pickupEntity.isAlive()) {
+        // Kiterino start - Extra bucketable mobs
+        java.util.List<Bucketable> bucketables = net.minecraft.world.item.MobBucketItem.EXTRA_BUCKETABLE_RULES.get(pickupEntity.getType());
+        if (bucketables == null) {
+            if (pickupEntity instanceof Bucketable bucketable) {
+                return bucketMobPickup(player, itemStack, hand, pickupEntity, bucketable);
+            }
+            return Optional.empty();
+        }
+        for (Bucketable bucketable : bucketables) {
+            Optional<InteractionResult> result = bucketMobPickup(player, itemStack, hand, pickupEntity, bucketable);
+            if (result.isPresent()) {
+                return result;
+            }
+        }
+        return Optional.empty();
+    }
+
+    static <T extends LivingEntity> Optional<InteractionResult> bucketMobPickup(Player player, ItemStack itemStack, InteractionHand hand, T pickupEntity, Bucketable bucketable) {
+        // Kiterino end - Extra bucketable mobs
+        if (itemStack.getItem() == bucketable.getPickBucket() && pickupEntity.isAlive() && bucketable.canPickup(pickupEntity)) { // Kiterino - Extra bucketable mobs
             // CraftBukkit start
             // pickupEntity.playSound(pickupEntity.getPickupSound(), 1.0F, 1.0F); // CraftBukkit - moved down
-            ItemStack bucket = pickupEntity.getBucketItemStack();
-            pickupEntity.saveToBucketTag(bucket);
+            ItemStack bucket = bucketable.getBucketItemStack(); // Kiterino - Extra bucketable mobs
+            bucketable.saveToBucketTag(pickupEntity, bucket); // Kiterino - Extra bucketable mobs
             org.bukkit.event.player.PlayerBucketEntityEvent playerBucketFishEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFishBucketEvent(pickupEntity, player, itemStack, bucket, hand);
             bucket = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(playerBucketFishEvent.getEntityBucket());
             if (playerBucketFishEvent.isCancelled()) {
@@ -90,7 +127,7 @@ public interface Bucketable {
                 pickupEntity.resendPossiblyDesyncedEntityData((ServerPlayer) player); // Paper
                 return Optional.of(InteractionResult.FAIL);
             }
-            pickupEntity.playSound(pickupEntity.getPickupSound(), 1.0F, 1.0F);
+            pickupEntity.playSound(bucketable.getPickupSound(), 1.0F, 1.0F); // Kiterino - Extra bucketable mobs
             // CraftBukkit end
             ItemStack result = ItemUtils.createFilledResult(itemStack, player, bucket, false);
             player.setItemInHand(hand, result);
diff --git a/net/minecraft/world/item/BucketItem.java b/net/minecraft/world/item/BucketItem.java
index 6bac133cdc35aff865624525821b36e33996d7c3..ed47a39a7735871d8e00a270f7899aef4c3686a0 100644
--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -54,7 +54,7 @@ public class BucketItem extends Item implements DispensibleContainerItem {
             BlockPos directionOffsetPos = pos.relative(direction);
             if (!level.mayInteract(player, pos) || !player.mayUseItemAt(directionOffsetPos, direction, itemStack)) {
                 return InteractionResult.FAIL;
-            } else if (this.content == Fluids.EMPTY) {
+            } else if (this.content == Fluids.EMPTY && !(this instanceof MobBucketItem)) { // Kiterino - Extra bucketable mobs
                 BlockState blockState = level.getBlockState(pos);
                 if (blockState.getBlock() instanceof BucketPickup bucketPickupBlock) {
                     // CraftBukkit start
@@ -87,10 +87,11 @@ public class BucketItem extends Item implements DispensibleContainerItem {
                 BlockPos placePos = clicked.getBlock() instanceof LiquidBlockContainer && this.content == Fluids.WATER ? pos : directionOffsetPos;
                 if (this.emptyContents(player, level, placePos, hitResult, hitResult.getDirection(), pos, itemStack, hand)) { // CraftBukkit
                     this.checkExtraContent(player, level, itemStack, placePos);
-                    if (player instanceof ServerPlayer) {
+                    if (player instanceof ServerPlayer && this.content != Fluids.EMPTY) { // Kiterino - Extra bucketable mobs
                         CriteriaTriggers.PLACED_BLOCK.trigger((ServerPlayer)player, placePos, itemStack);
                     }
 
+                    player.swing(hand, true); // Kiterino - Extra bucketable mobs
                     player.awardStat(Stats.ITEM_USED.get(this));
                     ItemStack emptyResult = ItemUtils.createFilledResult(itemStack, player, getEmptySuccessItem(itemStack, player));
                     return InteractionResult.SUCCESS.heldItemTransformedTo(emptyResult);
@@ -124,7 +125,7 @@ public class BucketItem extends Item implements DispensibleContainerItem {
 
     public boolean emptyContents(final @Nullable LivingEntity user, final Level level, final BlockPos pos, final @Nullable BlockHitResult hitResult, final Direction direction, final BlockPos clicked, final ItemStack itemStack, final InteractionHand hand) {
         // CraftBukkit end
-        if (!(this.content instanceof FlowingFluid flowingFluid)) {
+        if (!(this.content instanceof FlowingFluid flowingFluid) && !(this instanceof MobBucketItem && this.content == Fluids.EMPTY)) { // Kiterino - Extra bucketable mobs
             return false;
         } else {
             BlockState blockState = level.getBlockState(pos);
@@ -158,16 +159,16 @@ public class BucketItem extends Item implements DispensibleContainerItem {
                 }
 
                 return true;
-            } else if (block instanceof LiquidBlockContainer containerx && this.content == Fluids.WATER) {
+            } else if (block instanceof LiquidBlockContainer containerx && this.content == Fluids.WATER && this.content instanceof FlowingFluid flowingFluid) { // Kiterino - Extra bucketable mobs
                 containerx.placeLiquid(level, pos, blockState, flowingFluid.getSource(false));
                 this.playEmptySound(user, level, pos);
                 return true;
             } else {
-                if (!level.isClientSide() && mayReplace && !blockState.liquid()) {
+                if (!(this instanceof MobBucketItem && this.content == Fluids.EMPTY) && !level.isClientSide() && mayReplace && !blockState.liquid()) { // Kiterino - Extra bucketable mobs
                     level.destroyBlock(pos, true);
                 }
 
-                if (!level.setBlock(pos, this.content.defaultFluidState().createLegacyBlock(), Block.UPDATE_ALL_IMMEDIATE)
+                if (!(this instanceof MobBucketItem && this.content == Fluids.EMPTY) && !level.setBlock(pos, this.content.defaultFluidState().createLegacyBlock(), Block.UPDATE_ALL_IMMEDIATE) // Kiterino - Extra bucketable mobs
                     && !blockState.getFluidState().isSource()) {
                     return false;
                 } else {
diff --git a/net/minecraft/world/item/MobBucketItem.java b/net/minecraft/world/item/MobBucketItem.java
index c2574ea7a67259268187aa607d52fd21a569b05b..89399718091fd49b13b0fb802b961747ec06ccbc 100644
--- a/net/minecraft/world/item/MobBucketItem.java
+++ b/net/minecraft/world/item/MobBucketItem.java
@@ -18,6 +18,14 @@ import net.minecraft.world.level.material.Fluid;
 import org.jspecify.annotations.Nullable;
 
 public class MobBucketItem extends BucketItem {
+    // Kiterino start - Extra bucketable mobs
+    public static net.minecraft.core.dispenser.DispenseItemBehavior BUCKET_DISPENSE_BEHAVIOR;
+    public static final java.util.Map<EntityType<? extends Mob>, java.util.List<Bucketable>> EXTRA_BUCKETABLE_RULES = new java.util.HashMap<>();
+    public MobBucketItem bucketable(Bucketable bucketable) {
+        EXTRA_BUCKETABLE_RULES.computeIfAbsent(this.type, k -> new java.util.ArrayList<>()).add(bucketable);
+        return this;
+    }
+    // Kiterino end - Extra bucketable mobs
     private final EntityType<? extends Mob> type;
     private final SoundEvent emptySound;
 
@@ -42,9 +50,30 @@ public class MobBucketItem extends BucketItem {
 
     private void spawn(final ServerLevel level, final ItemStack itemStack, final BlockPos spawnPos) {
         Mob mob = this.type.create(level, EntityType.createDefaultStackConfig(level, itemStack, null), spawnPos, EntitySpawnReason.BUCKET, true, false);
-        if (mob instanceof Bucketable bucketable) {
+        // Kiterino start - Extra bucketable mobs
+        Bucketable bucketable = null;
+        java.util.List<Bucketable> bucketables = EXTRA_BUCKETABLE_RULES.get(this.type);
+        if (bucketables != null) {
+            for (Bucketable b : bucketables) {
+                if (b.getBucketItemStack().is(itemStack.getItem())) {
+                    bucketable = b;
+                    break;
+                }
+            }
+        }
+        if (bucketable == null && mob instanceof Bucketable b) {
+            bucketable = b;
+        }
+        if (bucketable != null) {
+        // Kiterino end - Extra bucketable mobs
             CustomData entityData = itemStack.getOrDefault(DataComponents.BUCKET_ENTITY_DATA, CustomData.EMPTY);
-            bucketable.loadFromBucketTag(entityData.copyTag());
+            // Kiterino start - Extra bucketable mobs
+            if (mob == null) {
+                bucketable.loadFromBucketTag(entityData.copyTag());
+            } else {
+                bucketable.loadFromBucketTag(mob, entityData.copyTag());
+            }
+            // Kiterino end - Extra bucketable mobs
             bucketable.setFromBucket(true);
         }
 
