From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 16 Nov 2025 23:45:31 +0200
Subject: [PATCH] Extra bucketable mobs


diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 2a03093f8fe465076747fddc1b615d3d7154e801..02910a75b45e2d778174039698f928ebaca44aa8 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -204,7 +204,8 @@ public interface DispenseItemBehavior {
                 /* Taken from BucketItem#emptyContents */
                 boolean willEmptyBucketItem = bucket instanceof final net.minecraft.world.item.BucketItem bucketItem && bucketItem.getContent() instanceof net.minecraft.world.level.material.FlowingFluid && (state.isAir() || state.canBeReplaced(bucketItem.getContent()) || (state.getBlock() instanceof net.minecraft.world.level.block.LiquidBlockContainer liquidBlockContainer && liquidBlockContainer.canPlaceLiquid(null, level, target, state, bucketItem.getContent())));
                 ItemStack leftover = new ItemStack(Items.BUCKET); // Kiterino - More BlockDispenseEvent API
-                if (willEmptyContentsSolidBucketItem || willEmptyBucketItem) {
+                boolean willPlaceCustomMob = bucket instanceof net.minecraft.world.item.MobBucketItem bucketItem && bucketItem.getContent() == net.minecraft.world.level.material.Fluids.EMPTY; // Kiterino - Extra bucketable mobs
+                if (willEmptyContentsSolidBucketItem || willEmptyBucketItem || willPlaceCustomMob) { // Kiterino - Extra bucketable mobs
                 // Paper end - correctly check if the bucket place will succeed
                     org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(level, source.pos());
                     org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(dispensed.copyWithCount(1)); // Paper - single item in event
@@ -247,6 +248,7 @@ public interface DispenseItemBehavior {
                 }
             }
         };
+        net.minecraft.world.item.MobBucketItem.BUCKET_DISPENSE_BEHAVIOR = filledBucketBehavior; // Kiterino - Extra bucketable mobs
         DispenserBlock.registerBehavior(Items.LAVA_BUCKET, filledBucketBehavior);
         DispenserBlock.registerBehavior(Items.WATER_BUCKET, filledBucketBehavior);
         DispenserBlock.registerBehavior(Items.POWDER_SNOW_BUCKET, filledBucketBehavior);
diff --git a/net/minecraft/world/entity/Bucketable.java b/net/minecraft/world/entity/Bucketable.java
index c0a931526bf663da06949fbe0a4a19ba5ecc5838..1c471a7266b0146d6a85fa8995fd638325dd7ddb 100644
--- a/net/minecraft/world/entity/Bucketable.java
+++ b/net/minecraft/world/entity/Bucketable.java
@@ -28,6 +28,24 @@ public interface Bucketable {
 
     SoundEvent getPickupSound();
 
+    // Kiterino start - Extra bucketable mobs
+    default net.minecraft.world.item.Item getPickBucket() {
+        return Items.WATER_BUCKET;
+    }
+
+    default boolean canPickup(LivingEntity entity) {
+        return true;
+    }
+
+    default void saveToBucketTag(LivingEntity entity, ItemStack stack) {
+        saveToBucketTag(stack);
+    }
+
+    default void loadFromBucketTag(LivingEntity entity, CompoundTag tag) {
+        loadFromBucketTag(tag);
+    }
+    // Kiterino end - Extra bucketable mobs
+
     @Deprecated
     static void saveDefaultDataToBucketTag(final Mob entity, final ItemStack bucket) {
         bucket.copyFrom(DataComponents.CUSTOM_NAME, entity);
@@ -75,15 +93,34 @@ public interface Bucketable {
         tag.getFloat("Health").ifPresent(entity::setHealth);
     }
 
-    static <T extends LivingEntity & Bucketable> Optional<InteractionResult> bucketMobPickup(
+    static <T extends LivingEntity> Optional<InteractionResult> bucketMobPickup( // Kiterino - Extra bucketable mobs
         final Player player, final InteractionHand hand, final T pickupEntity
     ) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (pickupEntity.canBePickedUpWithBucket(itemStack) && pickupEntity.isAlive()) {
+        // Kiterino start - Extra bucketable mobs
+        java.util.List<Bucketable> bucketables = net.minecraft.world.item.MobBucketItem.EXTRA_BUCKETABLE_RULES.get(pickupEntity.getType());
+        if (bucketables == null) {
+            if (pickupEntity instanceof Bucketable bucketable) {
+                return bucketMobPickup(player, itemStack, hand, pickupEntity, bucketable);
+            }
+            return Optional.empty();
+        }
+        for (Bucketable bucketable : bucketables) {
+            Optional<InteractionResult> result = bucketMobPickup(player, itemStack, hand, pickupEntity, bucketable);
+            if (result.isPresent()) {
+                return result;
+            }
+        }
+        return Optional.empty();
+    }
+
+    static <T extends LivingEntity> Optional<InteractionResult> bucketMobPickup(Player player, ItemStack itemStack, InteractionHand hand, T pickupEntity, Bucketable bucketable) {
+        // Kiterino end - Extra bucketable mobs
+        if (itemStack.getItem() == bucketable.getPickBucket() && pickupEntity.isAlive() && bucketable.canPickup(pickupEntity)) { // Kiterino - Extra bucketable mobs
             // CraftBukkit start
             // pickupEntity.playSound(pickupEntity.getPickupSound(), 1.0F, 1.0F); // CraftBukkit - moved down
-            ItemStack bucket = pickupEntity.getBucketItemStack();
-            pickupEntity.saveToBucketTag(bucket);
+            ItemStack bucket = bucketable.getBucketItemStack(); // Kiterino - Extra bucketable mobs
+            bucketable.saveToBucketTag(pickupEntity, bucket); // Kiterino - Extra bucketable mobs
             org.bukkit.event.player.PlayerBucketEntityEvent playerBucketFishEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFishBucketEvent(pickupEntity, player, itemStack, bucket, hand);
             bucket = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(playerBucketFishEvent.getEntityBucket());
             if (playerBucketFishEvent.isCancelled()) {
@@ -97,7 +134,7 @@ public interface Bucketable {
                 pickupEntity.resendPossiblyDesyncedEntityData((ServerPlayer) player); // Paper
                 return Optional.of(InteractionResult.FAIL);
             }
-            pickupEntity.playSound(pickupEntity.getPickupSound(), 1.0F, 1.0F);
+            pickupEntity.playSound(bucketable.getPickupSound(), 1.0F, 1.0F); // Kiterino - Extra bucketable mobs
             // CraftBukkit end
             ItemStack result = ItemUtils.createFilledResult(itemStack, player, bucket, false);
             player.setItemInHand(hand, result);
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 2d3fab8bda2335aad7b0dad4f5dacb3392cdf8bf..2bc16bdb23103473df0ad1e627e09a94dc03a720 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1397,6 +1397,12 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     }
 
     protected InteractionResult mobInteract(final Player player, final InteractionHand hand) {
+        // Kiterino start - Extra bucketable mobs // TODO Client sends block interact after, which auto releases the mob; currently worked around in a plugin
+        // Optional<InteractionResult> result = net.minecraft.world.entity.animal.Bucketable.bucketMobPickup(player, hand, this);
+        // if (result.isPresent()) {
+        //     return result.get();
+        // }
+        // Kiterino end - Extra bucketable mobs
         return tryRide(player, hand); // Purpur - Ridables
     }
 
diff --git a/net/minecraft/world/item/BucketItem.java b/net/minecraft/world/item/BucketItem.java
index 332666a29ceba68daf280414a6fc698057e09ead..5b264a05575a7c3189a7a9ba010cc42b2d045f8d 100644
--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -64,11 +64,12 @@ public class BucketItem extends Item implements DispensibleContainerItem {
                     CriteriaTriggers.PLACED_BLOCK.trigger(serverPlayer, placePos, itemStack);
                 }
 
+                player.swing(hand, true); // Kiterino - Extra bucketable mobs
                 player.awardStat(Stats.ITEM_USED.get(this));
                 ItemStack emptyResult = ItemUtils.createFilledResult(itemStack, player, getEmptySuccessItem(itemStack, player));
                 return InteractionResult.SUCCESS.heldItemTransformedTo(emptyResult);
             } else {
-                if (this.content == Fluids.EMPTY) {
+                if (this.content == Fluids.EMPTY && !(this instanceof MobBucketItem)) { // Kiterino - Extra bucketable mobs
                     BlockState blockState = level.getBlockState(pos);
                     if (blockState.getBlock() instanceof BucketPickup bucketPickupBlock) {
                         // CraftBukkit start
@@ -130,7 +131,7 @@ public class BucketItem extends Item implements DispensibleContainerItem {
 
     public boolean emptyContents(final @Nullable LivingEntity user, final Level level, final BlockPos pos, final @Nullable BlockHitResult hitResult, final Direction direction, final BlockPos clicked, final ItemStack itemStack, final InteractionHand hand) {
         // CraftBukkit end
-        if (!(this.content instanceof FlowingFluid flowingFluid)) {
+        if (!(this.content instanceof FlowingFluid flowingFluid) && !(this instanceof MobBucketItem && this.content == Fluids.EMPTY)) { // Kiterino - Extra bucketable mobs
             return false;
         } else {
             BlockState blockState = level.getBlockState(pos);
@@ -166,16 +167,16 @@ public class BucketItem extends Item implements DispensibleContainerItem {
                 }
 
                 return true;
-            } else if (block instanceof LiquidBlockContainer container && this.content == Fluids.WATER) {
+            } else if (block instanceof LiquidBlockContainer container && this.content == Fluids.WATER && this.content instanceof FlowingFluid flowingFluid) { // Kiterino - Extra bucketable mobs
                 container.placeLiquid(level, pos, blockState, flowingFluid.getSource(false));
                 this.playEmptySound(user, level, pos);
                 return true;
             } else {
-                if (!level.isClientSide() && mayReplace && !blockState.liquid()) {
+                if (!(this instanceof MobBucketItem && this.content == Fluids.EMPTY) && !level.isClientSide() && mayReplace && !blockState.liquid()) { // Kiterino - Extra bucketable mobs
                     level.destroyBlock(pos, true);
                 }
 
-                if (!level.setBlock(pos, this.content.defaultFluidState().createLegacyBlock(), Block.UPDATE_ALL_IMMEDIATE)
+                if (!(this instanceof MobBucketItem && this.content == Fluids.EMPTY) && !level.setBlock(pos, this.content.defaultFluidState().createLegacyBlock(), Block.UPDATE_ALL_IMMEDIATE) // Kiterino - Extra bucketable mobs
                     && !blockState.getFluidState().isSource()) {
                     return false;
                 }
diff --git a/net/minecraft/world/item/MobBucketItem.java b/net/minecraft/world/item/MobBucketItem.java
index 73ebbb4e5403db903a9153a7852b3bfcea118fc0..aa1ed1b723f6adabf3d796a7120ea218a99f6069 100644
--- a/net/minecraft/world/item/MobBucketItem.java
+++ b/net/minecraft/world/item/MobBucketItem.java
@@ -21,6 +21,14 @@ import net.minecraft.world.phys.BlockHitResult;
 import org.jspecify.annotations.Nullable;
 
 public class MobBucketItem extends BucketItem {
+    // Kiterino start - Extra bucketable mobs
+    public static net.minecraft.core.dispenser.DispenseItemBehavior BUCKET_DISPENSE_BEHAVIOR;
+    public static final java.util.Map<EntityType<? extends Mob>, java.util.List<Bucketable>> EXTRA_BUCKETABLE_RULES = new java.util.HashMap<>();
+    public MobBucketItem bucketable(Bucketable bucketable) {
+        EXTRA_BUCKETABLE_RULES.computeIfAbsent(this.type, k -> new java.util.ArrayList<>()).add(bucketable);
+        return this;
+    }
+    // Kiterino end - Extra bucketable mobs
     private final EntityType<? extends Mob> type;
     private final SoundEvent emptySound;
 
@@ -45,9 +53,30 @@ public class MobBucketItem extends BucketItem {
 
     private void spawn(final ServerLevel level, final ItemStack itemStack, final BlockPos spawnPos) {
         Mob mob = this.type.create(level, EntityType.createDefaultStackConfig(level, itemStack, null), spawnPos, EntitySpawnReason.BUCKET, true, false);
-        if (mob instanceof Bucketable bucketable) {
+        // Kiterino start - Extra bucketable mobs
+        Bucketable bucketable = null;
+        java.util.List<Bucketable> bucketables = EXTRA_BUCKETABLE_RULES.get(this.type);
+        if (bucketables != null) {
+            for (Bucketable b : bucketables) {
+                if (b.getBucketItemStack().is(itemStack.getItem())) {
+                    bucketable = b;
+                    break;
+                }
+            }
+        }
+        if (bucketable == null && mob instanceof Bucketable b) {
+            bucketable = b;
+        }
+        if (bucketable != null) {
+            // Kiterino end - Extra bucketable mobs
             CustomData entityData = itemStack.getOrDefault(DataComponents.BUCKET_ENTITY_DATA, CustomData.EMPTY);
-            bucketable.loadFromBucketTag(entityData.copyTag());
+            // Kiterino start - Extra bucketable mobs
+            if (mob == null) {
+                bucketable.loadFromBucketTag(entityData.copyTag());
+            } else {
+                bucketable.loadFromBucketTag(mob, entityData.copyTag());
+            }
+            // Kiterino end - Extra bucketable mobs
             bucketable.setFromBucket(true);
         }
 
