From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 18 Nov 2025 16:27:06 +0200
Subject: [PATCH] Expose leftover items in events


diff --git a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 993e6aabfd05a74523705b3dc2c4f8f2f88b2de3..356dadc60b4500634a72d490982e9a8470578039 100644
--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -224,6 +224,12 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
                                 org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(fuel),
                                 newLitTime
                             );
+                            // Kiterino start - Expose leftover items in events
+                            ItemStackTemplate craftingRemainder = fuel.getItem().getCraftingRemainder();
+                            if (craftingRemainder != null) {
+                                burnEvent.setRemainingItem(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(craftingRemainder.create()));
+                            }
+                            // Kiterino end - Expose leftover items in events
                             if (!burnEvent.callEvent()) return;
                             newLitTime = burnEvent.getBurnTime();
                             // CraftBukkit end
@@ -231,7 +237,17 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
                             entity.litTotalTime = newLitTime;
                             if (newLitTime > 0 && burnEvent.isBurning()) { // CraftBukkit - respect event output
                                 if (burnEvent.willConsumeFuel()) { // Paper - add consumeFuel to FurnaceBurnEvent
-                                consumeFuel(entity.items, fuel);
+                                    // Kiterino start - Expose leftover items in events
+                                    org.bukkit.inventory.ItemStack remainingItem = burnEvent.getRemainingItem();
+                                    ItemStack craftingRemainderItem = remainingItem == null ? ItemStack.EMPTY : ItemStack.fromBukkitCopy(remainingItem);
+                                    fuel.shrink(1);
+                                    if (fuel.isEmpty()) {
+                                        entity.items.set(1, craftingRemainderItem); // Kiterino - Expose leftover items in events
+                                    }
+                                    else if (!craftingRemainderItem.isEmpty()) {
+                                        net.minecraft.world.Containers.dropItemStack(level, pos.getX(), pos.getY(), pos.getZ(), craftingRemainderItem);
+                                    }
+                                    // Kiterino end - Expose leftover items in events
                                 } // Paper - add consumeFuel to FurnaceBurnEvent
                                 isLit = true;
                                 changed = true;
diff --git a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
index 5d26efb1feb1fd7d0f3bcc0fd7678b303518312a..00c6bed3678ed2c29f1a443886973ac1e0480173 100644
--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -245,6 +245,7 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
             brewResults.add(dest, org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(potionBrewing.mix(ingredient, items.get(dest))));
         }
 
+        ItemStackTemplate remainder = ingredient.getItem().getCraftingRemainder(); // Kiterino - Expose leftover items in events
         if (owner != null) {
             org.bukkit.event.inventory.BrewEvent event = new org.bukkit.event.inventory.BrewEvent(
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
@@ -252,9 +253,18 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
                 brewResults,
                 entity.fuel
             );
+            // Kiterino start - Expose leftover items in events
+            if (remainder != null) {
+                event.setRemainingItem(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(remainder.create()));
+            }
+            // Kiterino end - Expose leftover items in events
             if (!event.callEvent()) {
                 return;
             }
+            // Kiterino start - Expose leftover items in events
+            org.bukkit.inventory.ItemStack remainingItem = event.getRemainingItem();
+            remainder = remainingItem == null ? null : ItemStackTemplate.fromNonEmptyStack(ItemStack.fromBukkitCopy(remainingItem));
+            // Kiterino end - Expose leftover items in events
 
             for (int dest = 0; dest < 3; dest++) {
                 if (dest < brewResults.size()) {
@@ -267,7 +277,7 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
         // CraftBukkit end
 
         ingredient.shrink(1);
-        ItemStackTemplate remainder = ingredient.getItem().getCraftingRemainder();
+        // ItemStackTemplate remainder = ingredient.getItem().getCraftingRemainder(); // Kiterino - Moved up - Expose leftover items in events
         if (remainder != null) {
             if (ingredient.isEmpty()) {
                 ingredient = remainder.create();
