From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 18 Nov 2025 16:27:06 +0200
Subject: [PATCH] Expose leftover items in events


diff --git a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 26d6fa5b9ab8d75eb5a300b399a81daf59a409d4..08dac7994cdf9399bbc9592606e35035ba9e664a 100644
--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -227,6 +227,12 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
                             org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(fuel),
                             entity.getBurnDuration(level.fuelValues(), fuel)
                         );
+                        // Kiterino start - Expose leftover items in events
+                        ItemStackTemplate craftingRemainder = fuel.getItem().getCraftingRemainder();
+                        if (craftingRemainder != null) {
+                            furnaceBurnEvent.setRemainingItem(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(craftingRemainder.create()));
+                        }
+                        // Kiterino end - Expose leftover items in events
                         if (!furnaceBurnEvent.callEvent()) return;
                         // CraftBukkit end
                         if (!isLit) {
@@ -235,7 +241,17 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
                             entity.litTotalTime = newLitTime;
                             if (newLitTime > 0 && furnaceBurnEvent.isBurning()) { // CraftBukkit - respect event output // TODO - snapshot - double check surrounding diff post squash
                                 if (furnaceBurnEvent.willConsumeFuel()) { // Paper - add consumeFuel to FurnaceBurnEvent
-                                consumeFuel(entity.items, fuel);
+                                    // Kiterino start - Expose leftover items in events
+                                    org.bukkit.inventory.ItemStack remainingItem = furnaceBurnEvent.getRemainingItem();
+                                    ItemStack craftingRemainderItem = remainingItem == null ? ItemStack.EMPTY : ItemStack.fromBukkitCopy(remainingItem);
+                                    fuel.shrink(1);
+                                    if (fuel.isEmpty()) {
+                                        entity.items.set(1, craftingRemainderItem); // Kiterino - Expose leftover items in events
+                                    }
+                                    else if (!craftingRemainderItem.isEmpty()) {
+                                        net.minecraft.world.Containers.dropItemStack(level, pos.getX(), pos.getY(), pos.getZ(), craftingRemainderItem);
+                                    }
+                                    // Kiterino end - Expose leftover items in events
                                 } // Paper - add consumeFuel to FurnaceBurnEvent
                                 isLit = true;
                                 changed = true;
diff --git a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
index 60a0a2c466cc688377815b9f17ff09c59b72eb64..adb043e6daf79d3911a24eeaaba8068826a8d388 100644
--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -248,6 +248,7 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
             brewResults.add(dest, org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(potionBrewing.mix(ingredient, items.get(dest))));
         }
 
+        ItemStackTemplate remainder = ingredient.getItem().getCraftingRemainder(); // Kiterino - Expose leftover items in events
         if (owner != null) {
             org.bukkit.event.inventory.BrewEvent event = new org.bukkit.event.inventory.BrewEvent(
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
@@ -255,9 +256,18 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
                 brewResults,
                 entity.fuel
             );
+            // Kiterino start - Expose leftover items in events
+            if (remainder != null) {
+                event.setRemainingItem(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(remainder.create()));
+            }
+            // Kiterino end - Expose leftover items in events
             if (!event.callEvent()) {
                 return;
             }
+            // Kiterino start - Expose leftover items in events
+            org.bukkit.inventory.ItemStack remainingItem = event.getRemainingItem();
+            remainder = remainingItem == null ? null : ItemStackTemplate.fromNonEmptyStack(ItemStack.fromBukkitCopy(remainingItem));
+            // Kiterino end - Expose leftover items in events
 
             for (int dest = 0; dest < 3; dest++) {
                 if (dest < brewResults.size()) {
@@ -270,7 +280,7 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
         // CraftBukkit end
 
         ingredient.shrink(1);
-        ItemStackTemplate remainder = ingredient.getItem().getCraftingRemainder();
+        // ItemStackTemplate remainder = ingredient.getItem().getCraftingRemainder(); // Kiterino - Moved up - Expose leftover items in events
         if (remainder != null) {
             if (ingredient.isEmpty()) {
                 ingredient = remainder.create();
