From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 18 Nov 2025 17:03:33 +0200
Subject: [PATCH] Custom fuel rules


diff --git a/net/minecraft/world/level/block/entity/FuelValues.java b/net/minecraft/world/level/block/entity/FuelValues.java
index 9af4edf9fea87325957008e2e464469d1cb8366c..37f7a552a82f6e7582ec99f036b4f585c6182f9c 100644
--- a/net/minecraft/world/level/block/entity/FuelValues.java
+++ b/net/minecraft/world/level/block/entity/FuelValues.java
@@ -18,13 +18,28 @@ import net.minecraft.world.level.block.Blocks;
 
 public class FuelValues {
     public final Object2IntSortedMap<Item> values;
+    public final java.util.List<java.util.function.Function<org.bukkit.inventory.ItemStack, @org.jspecify.annotations.Nullable Integer>> extraRules = new java.util.ArrayList<>(); // Kiterino - Custom fuel rules
 
     private FuelValues(final Object2IntSortedMap<Item> values) {
         this.values = values;
     }
 
     public boolean isFuel(final ItemStack itemStack) {
-        return this.values.containsKey(itemStack.getItem());
+        // Kiterino - Custom fuel rules
+        if (this.values.containsKey(itemStack.getItem())) {
+            return true;
+        }
+        if (itemStack.isEmpty()) {
+            return false;
+        }
+        var bukkitItem = itemStack.asBukkitMirror();
+        for (var extraCheck : this.extraRules) {
+            if (extraCheck.apply(bukkitItem) != null) {
+                return true;
+            }
+        }
+        return false;
+        // Kiterino - Custom fuel rules
     }
 
     public SequencedSet<Item> fuelItems() {
@@ -32,7 +47,19 @@ public class FuelValues {
     }
 
     public int burnDuration(final ItemStack itemStack) {
-        return itemStack.isEmpty() ? 0 : this.values.getInt(itemStack.getItem());
+        // Kiterino start - Custom fuel rules
+        if (itemStack.isEmpty()) {
+            return 0;
+        }
+        var bukkitItem = itemStack.asBukkitMirror();
+        for (var extraCheck : this.extraRules) {
+            Integer duration = extraCheck.apply(bukkitItem);
+            if (duration != null) {
+                return duration;
+            }
+        }
+        return this.values.getInt(itemStack.getItem());
+        // Kiterino end - Custom fuel rules
     }
 
     public static FuelValues vanillaBurnTimes(final HolderLookup.Provider registries, final FeatureFlagSet enabledFeatures) {
