From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 19 Nov 2025 03:27:23 +0200
Subject: [PATCH] Add CrafterCraftPreviewEvent


diff --git a/net/minecraft/world/inventory/CrafterMenu.java b/net/minecraft/world/inventory/CrafterMenu.java
index eab7ab8a272b97b8073825875c6e1e8030db0bb7..f04df8a643f86ca0e26cfca2e270fa377e1b45f8 100644
--- a/net/minecraft/world/inventory/CrafterMenu.java
+++ b/net/minecraft/world/inventory/CrafterMenu.java
@@ -122,7 +122,15 @@ public class CrafterMenu extends AbstractContainerMenu implements ContainerListe
         if (this.player instanceof ServerPlayer serverPlayer) {
             ServerLevel level = serverPlayer.level();
             CraftingInput craftInput = this.container.asCraftInput();
-            ItemStack result = CrafterBlock.getPotentialResults(level, craftInput).map(recipe -> recipe.value().assemble(craftInput)).orElse(ItemStack.EMPTY);
+            // Kiterino start - Add CrafterCraftPreviewEvent
+            var potentialResults = CrafterBlock.getPotentialResults(level, craftInput);
+            ItemStack result = potentialResults.map(recipe -> recipe.value().assemble(craftInput)).orElse(ItemStack.EMPTY);
+            if (!result.isEmpty() && potentialResults.isPresent()) {
+                var event = new me.sosedik.kiterino.event.inventory.CrafterCraftPreviewEvent(serverPlayer.getBukkitEntity(), getBukkitView(), potentialResults.get().toBukkitRecipe(), result.asBukkitMirror());
+                event.callEvent();
+                result = ItemStack.fromBukkitCopy(event.getResult());
+            }
+            // Kiterino end - Add CrafterCraftPreviewEvent
             this.resultContainer.setItem(0, result);
         }
     }
