From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 12 Dec 2025 06:49:23 +0200
Subject: [PATCH] Fixup ridables controls


diff --git a/net/minecraft/world/entity/animal/polarbear/PolarBear.java b/net/minecraft/world/entity/animal/polarbear/PolarBear.java
index 46276e32dcc7884758e333c45f65ce03133c04fe..127c6baa388872fe8242b86199d94f90b2421191 100644
--- a/net/minecraft/world/entity/animal/polarbear/PolarBear.java
+++ b/net/minecraft/world/entity/animal/polarbear/PolarBear.java
@@ -115,7 +115,7 @@ public class PolarBear extends Animal implements NeutralMob {
     @Override
     public boolean onSpacebar() {
         if (!isStanding()) {
-            if (getRider() != null && getRider().getForwardMot() == 0 && getRider().getStrafeMot() == 0) {
+            if (getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && serverPlayer.getLastClientInput().forward() == serverPlayer.getLastClientInput().backward()) { // Kiterino - Fixup ridables controls
                 setStanding(true);
                 playSound(SoundEvents.POLAR_BEAR_WARNING, 1.0F, 1.0F);
             }
diff --git a/net/minecraft/world/entity/animal/squid/Squid.java b/net/minecraft/world/entity/animal/squid/Squid.java
index dcb597337bc69be21f0924dda89a5299aa8d0f21..f0dfc1e1e11df3a5406aae105dcaf4a564bcaebf 100644
--- a/net/minecraft/world/entity/animal/squid/Squid.java
+++ b/net/minecraft/world/entity/animal/squid/Squid.java
@@ -380,12 +380,13 @@ public class Squid extends AgeableWaterCreature {
         public void tick() {
             // Purpur start - Ridables
             net.minecraft.world.entity.player.Player rider = squid.getRider();
-            if (rider != null && squid.isControllable()) {
-                if (rider.isJumping()) {
+            if (rider instanceof net.minecraft.server.level.ServerPlayer serverPlayer && squid.isControllable()) {
+                net.minecraft.world.entity.player.Input lastClientInput = serverPlayer.getLastClientInput(); // Kiterino - Fixup ridables controls
+                if (lastClientInput.jump()) { // Kiterino - Fixup ridables controls
                     squid.onSpacebar();
                 }
-                float forward = rider.getForwardMot();
-                float strafe = rider.getStrafeMot();
+                float forward = (lastClientInput.forward() == lastClientInput.backward() ? 0.0F : lastClientInput.forward() ? 1.0F : -1.0F); // Kiterino - Fixup ridables controls
+                float strafe = (lastClientInput.left() == lastClientInput.right() ? 0.0F : lastClientInput.left() ? 1.0F : -1.0F) * 0.5F; // strafe slower by default // Kiterino - Fixup ridables controls
                 float speed = (float) squid.getAttributeValue(Attributes.MOVEMENT_SPEED) * 10F;
                 if (forward < 0.0F) {
                     speed *= -0.5;
diff --git a/net/minecraft/world/entity/monster/Creeper.java b/net/minecraft/world/entity/monster/Creeper.java
index 9d8b9e3e21ee721f5ea9339143839eeaebd341ae..b409f0f4232a6c32f7da33e06759cf2bb74943df 100644
--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -88,8 +88,8 @@ public class Creeper extends Monster {
         if (powerToggleDelay > 0) {
             powerToggleDelay--;
         }
-        if (getRider() != null && this.isControllable()) {
-            if (getRider().getForwardMot() != 0 || getRider().getStrafeMot() != 0) {
+        if (getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && this.isControllable()) { // Kiterino - Fixup ridables controls
+            if (serverPlayer.getLastClientInput().forward() != serverPlayer.getLastClientInput().backward()) { // Kiterino - Fixup ridables controls
                 spacebarCharge = 0;
                 setIgnited(false);
                 setSwellDir(-1);
@@ -126,7 +126,7 @@ public class Creeper extends Monster {
             }
         }
         if (!isIgnited()) {
-            if (getRider() != null && getRider().getForwardMot() == 0 && getRider().getStrafeMot() == 0 &&
+            if (getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && serverPlayer.getLastClientInput().forward() == serverPlayer.getLastClientInput().backward() && // Kiterino - Fixup ridables controls
                     getRider().getBukkitEntity().hasPermission("allow.special.creeper")) {
                 setIgnited(true);
                 setSwellDir(1);
diff --git a/net/minecraft/world/entity/monster/cubemob/AbstractCubeMob.java b/net/minecraft/world/entity/monster/cubemob/AbstractCubeMob.java
index 9d4410c84c0ebb9d9e0a9eead27b235992ed2df8..b4a69663b3e4486bbb417c2cf2da36916ade2fff 100644
--- a/net/minecraft/world/entity/monster/cubemob/AbstractCubeMob.java
+++ b/net/minecraft/world/entity/monster/cubemob/AbstractCubeMob.java
@@ -558,7 +558,7 @@ public abstract class AbstractCubeMob extends AgeableMob {
             } else {
                 this.operation = MoveControl.Operation.WAIT;
                 if (this.mob.onGround()) {
-                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (this.mob.getRider() != null && this.mob.isControllable() && (this.mob.getRider().getForwardMot() != 0 || this.mob.getRider().getStrafeMot() != 0) ? 2.0D : 1.0D))); // Purpur - Ridables
+                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (this.mob.getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && this.mob.isControllable() && serverPlayer.getLastClientInput().forward() != serverPlayer.getLastClientInput().backward() ? 2.0D : 1.0D))); // Purpur - Ridables // Kiterino - Fixup ridables controls
                     if (this.jumpDelay-- <= 0) {
                         this.jumpDelay = this.mob.getJumpDelay();
                         if (this.isAggressive) {
@@ -575,7 +575,7 @@ public abstract class AbstractCubeMob extends AgeableMob {
                         this.mob.setSpeed(0.0F);
                     }
                 } else {
-                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (this.mob.getRider() != null && this.mob.isControllable() && (this.mob.getRider().getForwardMot() != 0 || this.mob.getRider().getStrafeMot() != 0) ? 2.0D : 1.0D))); // Purpur - Ridables
+                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (this.mob.getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && this.mob.isControllable() && serverPlayer.getLastClientInput().forward() != serverPlayer.getLastClientInput().backward() ? 2.0D : 1.0D))); // Purpur - Ridables // Kiterino - Fixup ridables controls
                 }
             }
         }
diff --git a/net/minecraft/world/entity/monster/cubemob/Slime.java b/net/minecraft/world/entity/monster/cubemob/Slime.java
index 1faab89a0c6febdbbb15c8f1345fbd8e71f3af6e..e5121dce0b6d3e4bc042457b96475334b388b643 100644
--- a/net/minecraft/world/entity/monster/cubemob/Slime.java
+++ b/net/minecraft/world/entity/monster/cubemob/Slime.java
@@ -69,9 +69,10 @@ public class Slime extends AbstractCubeMob implements Enemy {
 
     @Override
     public boolean onSpacebar() {
-        if (onGround && getRider() != null && this.isControllable()) {
+        if (onGround && getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && this.isControllable()) { // Kiterino - Fixup ridables controls
             actualJump = true;
-            if (getRider().getForwardMot() == 0 || getRider().getStrafeMot() == 0) {
+            net.minecraft.world.entity.player.Input lastClientInput = serverPlayer.getLastClientInput(); // Kiterino - Fixup ridables controls
+            if (lastClientInput.forward() == lastClientInput.backward()) { // Kiterino - Fixup ridables controls
                 jumpFromGround(); // jump() here if not moving
             }
         }
