From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 12 Dec 2025 06:49:23 +0200
Subject: [PATCH] Fixup ridables controls


diff --git a/net/minecraft/world/entity/animal/polarbear/PolarBear.java b/net/minecraft/world/entity/animal/polarbear/PolarBear.java
index 592711e205ee959cc66de80933ad5a2dcec3092f..97c999b0b5c72d86e332938ec68bc923d5bdd8e1 100644
--- a/net/minecraft/world/entity/animal/polarbear/PolarBear.java
+++ b/net/minecraft/world/entity/animal/polarbear/PolarBear.java
@@ -109,7 +109,7 @@ public class PolarBear extends Animal implements NeutralMob {
     @Override
     public boolean onSpacebar() {
         if (!isStanding()) {
-            if (getRider() != null && getRider().getForwardMot() == 0 && getRider().getStrafeMot() == 0) {
+            if (getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && serverPlayer.getLastClientInput().forward() == serverPlayer.getLastClientInput().backward()) { // Kiterino - Fixup ridables controls
                 setStanding(true);
                 playSound(SoundEvents.POLAR_BEAR_WARNING, 1.0F, 1.0F);
             }
diff --git a/net/minecraft/world/entity/animal/squid/Squid.java b/net/minecraft/world/entity/animal/squid/Squid.java
index 6b5be8e58aa7e3acd14ff3cb1ba25d15f8d37410..96c7d43dfbf1bbc5dd8f621819852a753fd50fe6 100644
--- a/net/minecraft/world/entity/animal/squid/Squid.java
+++ b/net/minecraft/world/entity/animal/squid/Squid.java
@@ -372,12 +372,13 @@ public class Squid extends AgeableWaterCreature {
         public void tick() {
             // Purpur start - Ridables
             net.minecraft.world.entity.player.Player rider = squid.getRider();
-            if (rider != null && squid.isControllable()) {
-                if (rider.jumping) {
+            if (rider instanceof net.minecraft.server.level.ServerPlayer serverPlayer && squid.isControllable()) {
+                net.minecraft.world.entity.player.Input lastClientInput = serverPlayer.getLastClientInput(); // Kiterino - Fixup ridables controls
+                if (lastClientInput.jump()) { // Kiterino - Fixup ridables controls
                     squid.onSpacebar();
                 }
-                float forward = rider.getForwardMot();
-                float strafe = rider.getStrafeMot();
+                float forward = (lastClientInput.forward() == lastClientInput.backward() ? 0.0F : lastClientInput.forward() ? 1.0F : -1.0F); // Kiterino - Fixup ridables controls
+                float strafe = (lastClientInput.left() == lastClientInput.right() ? 0.0F : lastClientInput.left() ? 1.0F : -1.0F) * 0.5F; // strafe slower by default // Kiterino - Fixup ridables controls
                 float speed = (float) squid.getAttributeValue(Attributes.MOVEMENT_SPEED) * 10F;
                 if (forward < 0.0F) {
                     speed *= -0.5;
diff --git a/net/minecraft/world/entity/monster/Creeper.java b/net/minecraft/world/entity/monster/Creeper.java
index 608d787b73dce1726ccece9e0372acf995f261d4..956721ceeab199ff48ab699b393aa32cb4c7ec53 100644
--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -88,8 +88,8 @@ public class Creeper extends Monster {
         if (powerToggleDelay > 0) {
             powerToggleDelay--;
         }
-        if (getRider() != null && this.isControllable()) {
-            if (getRider().getForwardMot() != 0 || getRider().getStrafeMot() != 0) {
+        if (getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && this.isControllable()) { // Kiterino - Fixup ridables controls
+            if (serverPlayer.getLastClientInput().forward() != serverPlayer.getLastClientInput().backward()) { // Kiterino - Fixup ridables controls
                 spacebarCharge = 0;
                 setIgnited(false);
                 setSwellDir(-1);
@@ -126,7 +126,7 @@ public class Creeper extends Monster {
             }
         }
         if (!isIgnited()) {
-            if (getRider() != null && getRider().getForwardMot() == 0 && getRider().getStrafeMot() == 0 &&
+            if (getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && serverPlayer.getLastClientInput().forward() == serverPlayer.getLastClientInput().backward() && // Kiterino - Fixup ridables controls
                     getRider().getBukkitEntity().hasPermission("allow.special.creeper")) {
                 setIgnited(true);
                 setSwellDir(1);
diff --git a/net/minecraft/world/entity/monster/Slime.java b/net/minecraft/world/entity/monster/Slime.java
index 29d6d0c9687b835d24fe51c3824deeb67af1ba4f..42dab50b1e843f7a79faf5793e5133a7ac18d597 100644
--- a/net/minecraft/world/entity/monster/Slime.java
+++ b/net/minecraft/world/entity/monster/Slime.java
@@ -92,9 +92,10 @@ public class Slime extends Mob implements Enemy {
 
     @Override
     public boolean onSpacebar() {
-        if (onGround && getRider() != null && this.isControllable()) {
+        if (onGround && getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && this.isControllable()) { // Kiterino - Fixup ridables controls
             actualJump = true;
-            if (getRider().getForwardMot() == 0 || getRider().getStrafeMot() == 0) {
+            net.minecraft.world.entity.player.Input lastClientInput = serverPlayer.getLastClientInput(); // Kiterino - Fixup ridables controls
+            if (lastClientInput.forward() == lastClientInput.backward()) { // Kiterino - Fixup ridables controls
                 jumpFromGround(); // jump() here if not moving
             }
         }
@@ -651,7 +652,7 @@ public class Slime extends Mob implements Enemy {
             } else {
                 this.operation = MoveControl.Operation.WAIT;
                 if (this.mob.onGround()) {
-                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (slime.getRider() != null && slime.isControllable() && (slime.getRider().getForwardMot() != 0 || slime.getRider().getStrafeMot() != 0) ? 2.0D : 1.0D))); // Purpur - Ridables
+                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (slime.getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && slime.isControllable() && serverPlayer.getLastClientInput().forward() != serverPlayer.getLastClientInput().backward() ? 2.0D : 1.0D))); // Purpur - Ridables // Kiterino - Fixup ridables controls
                     if (this.jumpDelay-- <= 0) {
                         this.jumpDelay = this.slime.getJumpDelay();
                         if (this.isAggressive) {
@@ -668,7 +669,7 @@ public class Slime extends Mob implements Enemy {
                         this.mob.setSpeed(0.0F);
                     }
                 } else {
-                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (slime.getRider() != null && slime.isControllable() && (slime.getRider().getForwardMot() != 0 || slime.getRider().getStrafeMot() != 0) ? 2.0D : 1.0D))); // Purpur - Ridables
+                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (slime.getRider() instanceof net.minecraft.server.level.ServerPlayer serverPlayer && slime.isControllable() && serverPlayer.getLastClientInput().forward() != serverPlayer.getLastClientInput().backward() ? 2.0D : 1.0D))); // Purpur - Ridables // Kiterino - Fixup ridables controls
                 }
             }
         }
