From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 7 Feb 2026 14:48:26 +0200
Subject: [PATCH] Floating paintings


diff --git a/net/minecraft/world/entity/decoration/painting/Painting.java b/net/minecraft/world/entity/decoration/painting/Painting.java
index a9998d3e8d79a147e21085bff41964050e1b6e7c..b3aaab7b42eb77ba1c0dd2b36b35c7732ae37c37 100644
--- a/net/minecraft/world/entity/decoration/painting/Painting.java
+++ b/net/minecraft/world/entity/decoration/painting/Painting.java
@@ -47,6 +47,30 @@ public class Painting extends HangingEntity {
         super(type, level);
     }
 
+    // Kiterino start - Floating paintings
+    @Override
+    public boolean survives() {
+        if (level().kiterinoConfig.floatingPaintings) {
+            return canCoexist(true);
+        }
+        return super.survives();
+    }
+
+    public boolean canBePlaced(boolean requireSupport) {
+        if (!level().kiterinoConfig.floatingPaintings) {
+            return survives();
+        }
+        if (hasLevelCollision(getPopBox())) {
+            return false;
+        }
+        boolean flag = !requireSupport || BlockPos.betweenClosedStream(this.calculateSupportBox()).allMatch(blockPos -> {
+	        net.minecraft.world.level.block.state.BlockState blockState = this.level().getBlockState(blockPos);
+	        return !blockState.isAir();
+        });
+        return flag && canCoexist(true);
+    }
+    // Kiterino end - Floating paintings
+
     @Override
     protected void defineSynchedData(final SynchedEntityData.Builder entityData) {
         super.defineSynchedData(entityData);
@@ -101,7 +125,7 @@ public class Painting extends HangingEntity {
         candidate.setDirection(direction);
         potentialVariants.removeIf(variant -> {
             candidate.setVariant((Holder<PaintingVariant>)variant);
-            return !candidate.survives();
+                return !candidate.canBePlaced(true); // Kiterino - Floating paintings
         });
         if (potentialVariants.isEmpty()) {
             return Optional.empty();
diff --git a/net/minecraft/world/item/HangingEntityItem.java b/net/minecraft/world/item/HangingEntityItem.java
index e3d81dc449545d6ce25c2f2cb46d521cc6a2a422..547d322ea3018b543d88646bcff0492af358925e 100644
--- a/net/minecraft/world/item/HangingEntityItem.java
+++ b/net/minecraft/world/item/HangingEntityItem.java
@@ -84,6 +84,11 @@ public class HangingEntityItem extends Item {
             }
 
             itemInHand.shrink(1);
+                // Kiterino start - Floating paintings
+                if (player != null && context.getLevel().kiterinoConfig.floatingPaintings) {
+                    player.swing(context.getHand(), true);
+                }
+                // Kiterino end - Floating paintings
             return InteractionResult.SUCCESS;
         } else {
             return InteractionResult.CONSUME;
