From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 13 Dec 2025 02:57:31 +0200
Subject: [PATCH] Make villager trades public


diff --git a/net/minecraft/world/entity/npc/villager/VillagerTrades.java b/net/minecraft/world/entity/npc/villager/VillagerTrades.java
index 32a3b26ff9f654329052e3333d3b43bfec836401..50f7f4ca79a219486302bbbf959a62dc24fdd706 100644
--- a/net/minecraft/world/entity/npc/villager/VillagerTrades.java
+++ b/net/minecraft/world/entity/npc/villager/VillagerTrades.java
@@ -1308,7 +1308,7 @@ public class VillagerTrades {
         return PotionContents.createItemStack(Items.POTION, potion);
     }
 
-    static class DyedArmorForEmeralds implements VillagerTrades.ItemListing {
+    public static class DyedArmorForEmeralds implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final Item item;
         private final int value;
         private final int maxUses;
@@ -1351,7 +1351,7 @@ public class VillagerTrades {
         }
     }
 
-    static class EmeraldForItems implements VillagerTrades.ItemListing {
+    public static class EmeraldForItems implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final ItemCost itemStack;
         private final int maxUses;
         private final int villagerXp;
@@ -1380,7 +1380,7 @@ public class VillagerTrades {
         }
     }
 
-    static class EmeraldsForVillagerTypeItem implements VillagerTrades.ItemListing {
+    public static class EmeraldsForVillagerTypeItem implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final Map<ResourceKey<VillagerType>, Item> trades;
         private final int cost;
         private final int maxUses;
@@ -1412,7 +1412,7 @@ public class VillagerTrades {
         }
     }
 
-    static class EnchantBookForEmeralds implements VillagerTrades.ItemListing {
+    public static class EnchantBookForEmeralds implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final int villagerXp;
         private final TagKey<Enchantment> tradeableEnchantments;
         private final int minLevel;
@@ -1460,7 +1460,7 @@ public class VillagerTrades {
         }
     }
 
-    static class EnchantedItemForEmeralds implements VillagerTrades.ItemListing {
+    public static class EnchantedItemForEmeralds implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final ItemStack itemStack;
         private final int baseEmeraldCost;
         private final int maxUses;
@@ -1491,7 +1491,7 @@ public class VillagerTrades {
         }
     }
 
-    static class FailureItemListing implements VillagerTrades.ItemListing {
+    public static class FailureItemListing implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private FailureItemListing() {
         }
 
@@ -1505,7 +1505,7 @@ public class VillagerTrades {
         @Nullable MerchantOffer getOffer(ServerLevel level, Entity trader, RandomSource random);
     }
 
-    static class ItemsAndEmeraldsToItems implements VillagerTrades.ItemListing {
+    public static class ItemsAndEmeraldsToItems implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final ItemCost fromItem;
         private final int emeraldCost;
         private final ItemStack toItem;
@@ -1581,7 +1581,7 @@ public class VillagerTrades {
         }
     }
 
-    static class ItemsForEmeralds implements VillagerTrades.ItemListing {
+    public static class ItemsForEmeralds implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final ItemStack itemStack;
         private final int emeraldCost;
         private final int maxUses;
@@ -1656,7 +1656,7 @@ public class VillagerTrades {
         }
     }
 
-    static class SuspiciousStewForEmerald implements VillagerTrades.ItemListing {
+    public static class SuspiciousStewForEmerald implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final SuspiciousStewEffects effects;
         private final int xp;
         private final float priceMultiplier;
@@ -1679,7 +1679,7 @@ public class VillagerTrades {
         }
     }
 
-    static class TippedArrowForItemsAndEmeralds implements VillagerTrades.ItemListing {
+    public static class TippedArrowForItemsAndEmeralds implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final ItemStack toItem;
         private final int toCount;
         private final int emeraldCost;
@@ -1716,7 +1716,7 @@ public class VillagerTrades {
         }
     }
 
-    static class TreasureMapForEmeralds implements VillagerTrades.ItemListing {
+    public static class TreasureMapForEmeralds implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         private final int emeraldCost;
         private final TagKey<Structure> destination;
         private final String displayName;
@@ -1753,7 +1753,7 @@ public class VillagerTrades {
         }
     }
 
-    record TypeSpecificTrade(Map<ResourceKey<VillagerType>, VillagerTrades.ItemListing> trades) implements VillagerTrades.ItemListing {
+    public record TypeSpecificTrade(Map<ResourceKey<VillagerType>, VillagerTrades.ItemListing> trades) implements VillagerTrades.ItemListing { // Kiterino - Make villager trades public
         @SafeVarargs
         public static VillagerTrades.TypeSpecificTrade oneTradeInBiomes(VillagerTrades.ItemListing listing, ResourceKey<VillagerType>... villagerTypes) {
             return new VillagerTrades.TypeSpecificTrade(
