--- a/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java
+++ b/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java
@@ -82,6 +_,31 @@
         return this.items;
     }
 
+    // Kiterino start - Extra block drops API
+    /**
+     * Adds a new item to drop
+     *
+     * @param item item
+     * @return Item entity created as a result of this method
+     */
+    public @NotNull Item addDrop(org.bukkit.inventory.@NotNull ItemStack item) {
+        return addDrop(item, null);
+    }
+
+    /**
+     * Adds a new item to drop
+     *
+     * @param item item
+     * @param function the function to be run before the entity is spawned
+     * @return Item entity created as a result of this method
+     */
+    public @NotNull Item addDrop(org.bukkit.inventory.@NotNull ItemStack item, java.util.function.@org.jetbrains.annotations.Nullable Consumer<? super Item> function) {
+        Item drop = this.blockState.getWorld().dropItemNaturally(this.blockState.getLocation().toCenterLocation(), item, function);
+        this.items.add(drop);
+        return drop;
+    }
+    // Kiterino end - Extra block drops API
+
     @Override
     public boolean isCancelled() {
         return this.cancelled;
