--- a/src/main/java/io/papermc/paper/datacomponent/item/ItemLore.java
+++ b/src/main/java/io/papermc/paper/datacomponent/item/ItemLore.java
@@ -71,6 +_,19 @@
         @Contract(value = "_ -> this", mutates = "this")
         Builder addLine(ComponentLike line);
 
+        // Kiterino start - Data components API improvements
+        /**
+         * Adds a component to the lore.
+         *
+         * @param index component index
+         * @param line component
+         * @return the builder for chaining
+         * @see #lines()
+         */
+        @Contract(value = "_, _ -> this", mutates = "this")
+        Builder addLine(int index, ComponentLike line);
+        // Kiterino end - Data components API improvements
+
         /**
          * Adds components to the lore.
          *
