--- a/net/minecraft/world/level/ExplosionDamageCalculator.java
+++ b/net/minecraft/world/level/ExplosionDamageCalculator.java
@@ -9,6 +_,14 @@
 
 public class ExplosionDamageCalculator {
     public Optional<Float> getBlockExplosionResistance(Explosion explosion, BlockGetter level, BlockPos pos, BlockState state, FluidState fluid) {
+        // Kiterino start - Allow explosions in fluids
+        if (level instanceof net.minecraft.server.level.ServerLevel serverLevel && serverLevel.kiterinoConfig.explosionsInFluids) {
+            if (state.isAir() || !fluid.isEmpty()) {
+                return Optional.empty();
+            }
+            return Optional.of(state.getBlock().getExplosionResistance());
+        }
+        // Kiterino end - Allow explosions in fluids
         return state.isAir() && fluid.isEmpty()
             ? Optional.empty()
             : Optional.of(Math.max(state.getBlock().getExplosionResistance(), fluid.getExplosionResistance()));
