--- a/net/minecraft/world/level/ExplosionDamageCalculator.java
+++ b/net/minecraft/world/level/ExplosionDamageCalculator.java
@@ -11,6 +_,14 @@
     public Optional<Float> getBlockExplosionResistance(
         final Explosion explosion, final BlockGetter level, final BlockPos pos, final BlockState block, final FluidState fluid
     ) {
+        // Kiterino start - Allow explosions in fluids
+        if (level instanceof net.minecraft.server.level.ServerLevel serverLevel && serverLevel.kiterinoConfig.explosionsInFluids) {
+            if (block.isAir() || !fluid.isEmpty()) {
+                return Optional.empty();
+            }
+            return Optional.of(block.getBlock().getExplosionResistance());
+        }
+        // Kiterino end - Allow explosions in fluids
         return block.isAir() && fluid.isEmpty()
             ? Optional.empty()
             : Optional.of(Math.max(block.getBlock().getExplosionResistance(), fluid.getExplosionResistance()));
