--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -337,8 +_,9 @@
      * being on fire).
      *
      * @param ticks Current ticks remaining
+     * @return whether the combust even succeeded, if fired
      */
-    public void setFireTicks(int ticks);
+    public boolean setFireTicks(int ticks); // Kiterino - void -> boolean - Call EntityCombustEvent on setFireTicks
 
     /**
      * Sets if the entity has visual fire (it will always appear to be on fire).
@@ -1361,4 +_,134 @@
      */
     void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end - Fire Immunity API
+
+    // Kiterino start - More Sound API
+	/**
+	 * Gets the sound category of this entity
+	 *
+	 * @return sound category
+	 */
+	default @NotNull org.bukkit.SoundCategory getSoundCategory() {
+		return switch (this) {
+			case HumanEntity humanEntity -> org.bukkit.SoundCategory.PLAYERS;
+			case Ambient ambient -> org.bukkit.SoundCategory.AMBIENT;
+			case Monster monster -> org.bukkit.SoundCategory.HOSTILE;
+			default -> org.bukkit.SoundCategory.NEUTRAL;
+		};
+	}
+
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param sound sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        emitSound(sound, getSoundCategory(), volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param sound sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(this, sound, category, volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param soundKey sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull net.kyori.adventure.key.Keyed soundKey, float volume, float pitch) {
+        emitSound(soundKey, getSoundCategory(), volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param soundKey sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull net.kyori.adventure.key.Keyed soundKey, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(net.kyori.adventure.sound.Sound.sound(soundKey.key(), category.soundSource(), volume, pitch), this);
+    }
+
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param sound sound to play
+     */
+    default void emitSound(@NotNull net.kyori.adventure.sound.Sound sound) {
+        getWorld().playSound(sound, this);
+    }
+    // Kiterino end - More Sound API
+
+    // Kiterino start - Forced panic mode API
+    /**
+     * Gets the number of forced panic ticks left
+     *
+     * @return the number of panic ticks
+     */
+    int getPanicTicks();
+
+    /**
+     * Sets the number of forced panic ticks
+     *
+     * @param ticks ticks
+     */
+    void setPanicTicks(int ticks);
+    // Kiterino end - Forced panic mode API
+
+    // Kiterino start - Allow setting entity's rider
+    /**
+     * Set the riding player. The player has to be the first passenger.
+     * <br> Note: this will call {@link org.bukkit.event.entity.EntityMountEvent}.
+     *
+     * @param rider rider
+     * @return whether the rider was set
+     */
+    default boolean setRider(@Nullable Player rider) {
+        return setRider(rider, true);
+    }
+
+    /**
+     * Set the riding player and optionally dismount the previous one.
+     * <br> Note: this will call {@link org.bukkit.event.entity.EntityMountEvent}.
+     *
+     * @param rider rider
+     * @param dismount whether to dismount the current rider
+     * @return whether rider was set
+     */
+    boolean setRider(@Nullable Player rider, boolean dismount);
+    // Kiterino end - Allow setting entity's rider
+
+    // Kiterino start - Allow resending entity metadata
+    /**
+     * Makes entity resend metadata on the next tick
+     *
+     * @param index metadata index
+     */
+    void resendMetadata(int index);
+    // Kiterino end - Allow resending entity metadata
+
+    // Kiterino start - Wet mobs API
+    /**
+     * Checks whether this entity is wet
+     *
+     * @return whether this entity is wet
+     */
+    default boolean isWet() {
+        return isInWater() || isInRain();
+    }
+    // Kiterino end - Wet mobs API
 }
