--- a/src/main/java/org/bukkit/entity/EntityFactory.java
+++ b/src/main/java/org/bukkit/entity/EntityFactory.java
@@ -16,4 +_,28 @@
      */
     @NotNull
     EntitySnapshot createEntitySnapshot(@NotNull String input) throws IllegalArgumentException;
+
+    // Kiterino start - Add extra EntityFactory methods
+    /**
+     * Create a new EntitySnapshot with the supplied input.<br>
+     * Accepts strings in the format output by {@link EntitySnapshot#getAsString()}.
+     *
+     * @param entityType entity type
+     * @return the created EntitySnapshot
+     */
+    default @NotNull EntitySnapshot createEntitySnapshot(@NotNull EntityType entityType) {
+        return createEntitySnapshot(entityType, null);
+    }
+
+    /**
+     * Create a new EntitySnapshot with the supplied input.<br>
+     * Accepts strings in the format output by {@link EntitySnapshot#getAsString()}.
+     *
+     * @param entityType entity type
+     * @param input the input string
+     * @return the created EntitySnapshot
+     * @throws IllegalArgumentException if the input string was provided in an invalid or unsupported format
+     */
+    @NotNull EntitySnapshot createEntitySnapshot(@NotNull EntityType entityType, @org.jetbrains.annotations.Nullable String input) throws IllegalArgumentException;
+    // Kiterino end - Add extra EntityFactory methods
 }
