--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1510,4 +_,43 @@
      */
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end - API for any mob to burn daylight
+
+    // Kiterino start - Expose sleeping loc
+    /**
+     * Checks whether this entity's bed exists as a physical bed
+     *
+     * @return whether this entity's bed exists
+     */
+    boolean bedExists();
+
+	/**
+	 * Gets the sleeping location
+	 *
+	 * @return sleeping location
+	 * @see #isSleeping()
+	 */
+	@Nullable Location getSleepingLocation();
+
+	/**
+	 * Sets the entity's sleeping pos
+	 *
+	 * @param loc sleeping location
+	 * @see #getSleepingLocation()
+	 * @see #isSleeping()
+	 */
+	void setSleepingLocation(@Nullable Location loc);
+    // Kiterino end - Expose sleeping loc
+
+    // Kiterino start - Allow laying at will
+    /**
+     * Lays the entity at its current location with
+     * the bed at the provided location
+     * <p>
+     * Note: bed is not required to be a bed
+     *
+     * @param sleepLocation actual sleeping location
+     * @param bedLocation location of the bed
+     */
+    void lay(@NotNull Location sleepLocation, @NotNull Location bedLocation);
+    // Kiterino end - Allow laying at will
 }
