--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -4029,4 +_,101 @@
         sendDeathScreen(message);
     }
     // Purpur end
+
+    // Kiterino start - Red tint API
+    /**
+     * Abuses world border packets to apply red tint on screen
+     *
+     * @param percentage tint percentage
+     */
+    void sendRedTint(double percentage);
+    // Kiterino end - Red tint API
+
+    // Kiterino start - Send ItemStack API
+    /**
+     * Sends set item packet. This will not actually set the item in the provided slot.
+     *
+     * @param slot item slot
+     * @param item item
+     */
+    void sendItem(int slot, @Nullable ItemStack item);
+
+    /**
+     * Sends set item packet for item on the cursor. This will not actually set the item on the cursor.
+     *
+     * @param item item
+     */
+    default void sendCursorItem(@Nullable ItemStack item) {
+        sendItem(me.sosedik.kiterino.inventory.InventorySlotHelper.CURSOR, item);
+    }
+    // Kiterino end - Send ItemStack API
+
+    // Kiterino start - Anger nearby piglins API
+    /**
+     * Triggers piglins angrying AI
+     *
+     * @param blockOpen whether the cause of anger is an opened block
+     */
+    void angerNearbyPiglins(boolean blockOpen);
+    // Kiterino end - Anger nearby piglins API
+
+    // Kiterino start - Open ender chest API
+    /**
+     * Opens ender chest for the player
+     *
+     * @param enderChest ender chest
+     * @return whether the chest was opened
+     */
+    boolean openEnderChest(org.bukkit.block.EnderChest enderChest);
+    // Kiterino end - Open ender chest API
+
+    // Kiterino start - Block place API
+    /**
+     * Place block as if this player placed it
+     * <p>Item in hand will be consumed if placed successfully
+     *
+     * @param hand hand to use for placing
+     * @param location location to place
+     * @param blockFace facing against which block will be placed
+     * @return true, if placed successfully
+     */
+    default boolean placeBlock(EquipmentSlot hand, Location location, org.bukkit.block.BlockFace blockFace) {
+        return placeBlock(hand, location, blockFace, true);
+    }
+
+    /**
+     * Place block as if this player placed it
+     *
+     * @param hand hand to use for placing
+     * @param location location to place
+     * @param blockFace facing against which block will be placed
+     * @param consume whether item should be consumed or not upon successful placement
+     * @return true, if placed successfully
+     */
+    boolean placeBlock(EquipmentSlot hand, Location location, org.bukkit.block.BlockFace blockFace, boolean consume);
+
+    /**
+     * Place block as if this player placed it
+     * <p>Item in hand will be consumed if placed successfully
+     *
+     * @param block block item to use for placing
+     * @param location location to place
+     * @param blockFace facing against which block will be placed
+     * @return true, if placed successfully
+     */
+    default boolean placeBlock(ItemStack block, Location location, org.bukkit.block.BlockFace blockFace) {
+        return placeBlock(block, location, blockFace, true);
+    }
+
+    /**
+     * Place block as if this player placed it
+     *
+     * @param block block item to use for placing
+     * @param location location to place
+     * @param blockFace facing against which block will be placed
+     * @param consume whether item should be consumed or not upon successful placement
+     * @return true, if placed successfully
+     */
+    boolean placeBlock(ItemStack block, Location location, org.bukkit.block.BlockFace blockFace, boolean consume);
+    // Kiterino end - Block place API
 }
