--- a/src/main/java/org/bukkit/entity/Raider.java
+++ b/src/main/java/org/bukkit/entity/Raider.java
@@ -127,4 +_,13 @@
      */
     @NotNull
     Sound getCelebrationSound();
+
+    // Kiterino start = More Raider API
+    /**
+     * Checks whether the raider is patrolling
+     *
+     * @return whether the raider is patrolling
+     */
+    boolean isPatrolling();
+    // Kiterino end = More Raider API
 }
