--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1111,4 +_,40 @@
         this.getHandle().setShouldBurnInDay(shouldBurnInDay);
     }
     // Purpur end - API for any mob to burn daylight
+
+    // Kiterino start - Expose sleeping loc
+    @Override
+    public boolean bedExists() {
+        return this.getHandle().checkBedExists();
+    }
+
+	@Override
+	public Location getSleepingLocation() {
+		return this.getHandle().getSleepingPos().map(pos -> new Location(this.getWorld(), pos.getX(), pos.getY(), pos.getZ())).orElse(null);
+	}
+
+	@Override
+	public void setSleepingLocation(Location loc) {
+		this.getHandle().setSleepingPos(loc == null ? null : org.bukkit.craftbukkit.util.CraftLocation.toBlockPosition(loc));
+	}
+    // Kiterino end - Expose sleeping loc
+
+    // Kiterino start - Allow laying at will
+    @Override
+    public void lay(Location sleepLocation, Location bedLocation) {
+        this.getHandle().setPose(net.minecraft.world.entity.Pose.SLEEPING);
+        this.getHandle().setSleepingPos(org.bukkit.craftbukkit.util.CraftLocation.toBlockPosition(bedLocation));
+        this.getHandle().setPos(sleepLocation.getX(), sleepLocation.getY(), sleepLocation.getZ());
+        this.getHandle().setDeltaMovement(Vec3.ZERO);
+        this.getHandle().hasImpulse = true;
+        if (this.getHandle() instanceof net.minecraft.server.level.ServerPlayer player && player.connection != null) {
+            // Sync player from bed to actual sleep loc
+            this.taskScheduler.schedule(entity -> {
+                if (player.connection != null) {
+                    player.connection.teleport(sleepLocation);
+                }
+            }, entity -> {}, 4L);
+        }
+    }
+    // Kiterino end - Allow laying at will
 }
