--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -378,8 +_,20 @@
     }
 
     @Override
-    public void setFireTicks(int ticks) {
-        this.getHandle().setRemainingFireTicks(ticks);
+    public boolean setFireTicks(int ticks) { // Kiterino start - Call EntityCombustEvent on setFireTicks
+        if (ticks <= 0 || getFireTicks() > 0) {
+            this.getHandle().setRemainingFireTicks(ticks);
+            return true;
+        }
+
+        var event = new org.bukkit.event.entity.EntityCombustEvent(this, ticks / 20F);
+        if (!event.callEvent()) {
+            return false;
+        }
+
+        this.getHandle().setRemainingFireTicks((int) (event.getDuration() * 20));
+        return true;
+        // Kiterino end - Call EntityCombustEvent on setFireTicks
     }
 
     @Override
@@ -1335,4 +_,37 @@
         return !getHandle().dismountsUnderwater();
     }
     // Purpur end - Ridables
+
+    // Kiterino start - Forced panic mode API
+    @Override
+    public int getPanicTicks() {
+        return getHandle().forcePanicTicks;
+    }
+
+    @Override
+    public void setPanicTicks(int ticks) {
+        getHandle().forcePanicTicks = Math.max(0, ticks);
+    }
+    // Kiterino end - Forced panic mode API
+
+    // Kiterino start - Allow setting entity's rider
+    @Override
+    public boolean setRider(Player rider, boolean dismount) {
+        Player currentRider = getRider();
+        if (rider == currentRider) {
+            return true;
+        }
+        if (currentRider != null && (!dismount || !currentRider.leaveVehicle())) {
+            return false;
+        }
+        return getHandle().setRider(((CraftPlayer) rider).getHandle());
+    }
+    // Kiterino end - Allow setting entity's rider
+
+    // Kiterino start - Allow resending entity metadata
+    @Override
+    public void resendMetadata(int index) {
+        getHandle().getEntityData().markDirty(index);
+    }
+    // Kiterino end - Allow resending entity metadata
 }
