--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -29,6 +_,17 @@
         } else if (target instanceof CraftLivingEntity) {
             entity.setTarget(((CraftLivingEntity) target).getHandle(), null, false);
         }
+        // Kiterino start - Update goal target when changing target
+        entity.targetSelector.getAvailableGoals().forEach(goal -> {
+            if (goal.getGoal() instanceof net.minecraft.world.entity.ai.goal.target.TargetGoal targetGoal) {
+                if (target == null) {
+                    targetGoal.setTargetMob(null);
+                } else if (target instanceof CraftLivingEntity craftLivingEntity) {
+                    targetGoal.setTargetMob(craftLivingEntity.getHandle());
+                }
+            }
+        });
+        // Kiterino end - Update goal target when changing target
     }
 
     @Override
