--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1391,7 +_,7 @@
         }
         // Minecraft does not currently support teleporting players between worlds with passengers.
         // It causes them to be dismounted, and causes weird behavior.
-        if (location.getWorld() != this.getWorld() && this.getHandle().isVehicle()) {
+        if (false && location.getWorld() != this.getWorld() && this.getHandle().isVehicle()) { // Kiterino - Do not block between-world teleports
             return false;
         }
 
@@ -3466,4 +_,122 @@
         super.knockback(strength, directionX, directionZ);
         this.entity.hurtMarked = true;
     }
+
+    // Kiterino start - Red tint API
+    @Override
+    public void sendRedTint(double percentage) {
+        percentage = 1D - Math.clamp(percentage, 0D, 1D);
+        Location loc = this.getLocation();
+        ServerGamePacketListenerImpl connection = this.getHandle().connection;
+        connection.send(new ClientboundSetBorderCenterPacket(loc.getX(), loc.getZ()));
+
+        double distance = 13_000_000 - 10_000_000 * percentage;
+        connection.send(new ClientboundSetBorderLerpSizePacket(20_000_00D, distance, 14_000L));
+    }
+    // Kiterino end - Red tint API
+
+    // Kiterino start - Send ItemStack API
+    @Override
+    public void sendItem(int slot, ItemStack item) {
+        int stateId = this.getHandle().containerMenu.incrementStateId();
+        this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket(this.getHandle().containerMenu.containerId, stateId, slot, CraftItemStack.asNMSCopy(item)));
+    }
+
+    @Override
+    public void sendCursorItem(ItemStack item) {
+        this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundSetCursorItemPacket(CraftItemStack.asNMSCopy(item)));
+    }
+    // Kiterino end - Send ItemStack API
+
+    // Kiterino start - Anger nearby piglins API
+    @Override
+    public void angerNearbyPiglins(boolean blockOpen) {
+        net.minecraft.world.entity.monster.piglin.PiglinAi.angerNearbyPiglins(this.getHandle().level(), this.getHandle(), blockOpen);
+    }
+    // Kiterino end - Anger nearby piglins API
+
+    // Kiterino start - Open ender chest API
+    @Override
+    public boolean openEnderChest(org.bukkit.block.EnderChest enderChest) {
+        net.minecraft.world.inventory.PlayerEnderChestContainer playerEnderChestContainer = this.getHandle().getEnderChestInventory();
+        net.minecraft.world.level.block.entity.EnderChestBlockEntity blockEntity = ((org.bukkit.craftbukkit.block.CraftEnderChest) enderChest).getBlockEntity();
+        playerEnderChestContainer.setActiveChest(blockEntity);
+        return this.getHandle().openMenu(new net.minecraft.world.SimpleMenuProvider((i, inventory, playerx) -> org.purpurmc.purpur.PurpurConfig.enderChestSixRows ? net.minecraft.world.level.block.EnderChestBlock.getEnderChestSixRows(i, inventory, this.getHandle(), playerEnderChestContainer) : net.minecraft.world.inventory.ChestMenu.threeRows(i, inventory, playerEnderChestContainer), net.minecraft.world.level.block.EnderChestBlock.CONTAINER_TITLE)).isPresent();
+    }
+    // Kiterino end - Open ender chest API
+
+    // Kiterino start - Block place API
+    @Override
+    public boolean canPlaceBlock(EquipmentSlot hand, ItemStack block, Location location, org.bukkit.block.BlockFace blockFace) {
+        return getPlacedState0(hand, block, location, blockFace) != null;
+    }
+
+    @Override
+    public BlockData getPlacedState(EquipmentSlot hand, ItemStack block, Location location, org.bukkit.block.BlockFace blockFace) {
+        net.minecraft.world.level.block.state.BlockState placedState = getPlacedState0(hand, block, location, blockFace);
+        return placedState == null ? null : placedState.asBlockData();
+    }
+
+    private net.minecraft.world.level.block.state.BlockState getPlacedState0(EquipmentSlot hand, ItemStack block, Location location, org.bukkit.block.BlockFace blockFace) {
+        if (block == null) {
+            return null;
+        }
+        if (!block.getType().isBlock()) {
+            return null;
+        }
+        if (!(location.getWorld() instanceof CraftWorld craftWorld)) {
+            return null;
+        }
+
+        var nmsItem = CraftItemStack.asNMSCopy(block);
+        if (!(nmsItem.getItem() instanceof net.minecraft.world.item.BlockItem blockItem)) {
+            return null;
+        }
+
+        BlockPos blockPos = CraftLocation.toBlockPos(location);
+        var mop = new net.minecraft.world.phys.BlockHitResult(CraftLocation.toVec3(location), com.google.common.base.Enums.getIfPresent(net.minecraft.core.Direction.class, blockFace.toString()).or(net.minecraft.core.Direction.UP), blockPos, false);
+        var context = new net.minecraft.world.item.context.BlockPlaceContext(getHandle(), CraftEquipmentSlot.getHand(hand), nmsItem, mop);
+
+        if (!craftWorld.getHandle().getBlockState(blockPos).canBeReplaced(context)) {
+            return null;
+        }
+
+        return blockItem.getPlacementState(context);
+    }
+
+    @Override
+    public boolean placeBlock(EquipmentSlot hand, Location location, org.bukkit.block.BlockFace blockFace, boolean consume) {
+        return placeBlock(hand, getInventory().getItem(hand), location, blockFace, consume);
+    }
+
+    @Override
+    public boolean placeBlock(ItemStack block, Location location, org.bukkit.block.BlockFace blockFace, boolean consume) {
+        return placeBlock(EquipmentSlot.HAND, block, location, blockFace, consume);
+    }
+
+    private boolean placeBlock(EquipmentSlot hand, ItemStack block, Location location, org.bukkit.block.BlockFace blockFace, boolean consume) {
+        if (block == null) {
+            return false;
+        }
+        if (!block.getType().isBlock()) {
+            return false;
+        }
+
+        var nmsItem = CraftItemStack.asNMSCopy(block);
+        var mop = new net.minecraft.world.phys.BlockHitResult(CraftLocation.toVec3(location), com.google.common.base.Enums.getIfPresent(net.minecraft.core.Direction.class, blockFace.toString()).or(net.minecraft.core.Direction.UP), CraftLocation.toBlockPos(location), false);
+        var context = new net.minecraft.world.item.context.BlockPlaceContext(getHandle(), CraftEquipmentSlot.getHand(hand), nmsItem, mop);
+        var result = nmsItem.useOn(context);
+        if (!(result instanceof net.minecraft.world.InteractionResult.Success)) {
+            return false;
+        }
+
+        swingHand(hand);
+        getWorld().playSound(location, location.getBlock().getBlockData().getSoundGroup().getPlaceSound(), 1F, 1F);
+        if (consume && getGameMode() != GameMode.CREATIVE) {
+            block.subtract();
+        }
+
+        return true;
+    }
+    // Kiterino end - Block place API
 }
