--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -3083,6 +_,11 @@
         if (filtered.startsWith(NamespacedKey.MINECRAFT + ":")) {
             filtered = filtered.substring((NamespacedKey.MINECRAFT + ":").length());
         }
+        // Kiterino start - Injecting custom Materials
+        else {
+            filtered = filtered.replace(':', '_');
+        }
+        // Kiterino end - Injecting custom Materials
 
         filtered = filtered.toUpperCase(Locale.ROOT);
 
@@ -3739,4 +_,16 @@
         }
     }
     // Purpur end - ItemStack convenience methods
+
+    // Kiterino start - Injecting custom Materials
+    /**
+     * Checks whether this item is injected
+     * (i.e., is a non-vanilla one)
+     *
+     * @return whether this item is injected
+     */
+    public boolean isInjected() {
+        return !NamespacedKey.MINECRAFT.equals(this.key.namespace());
+    }
+    // Kiterino end - Injecting custom Materials
 }
