--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -2900,6 +_,12 @@
      * @return Maximum stack size for this material
      */
     public int getMaxStackSize() {
+        // Kiterino start - Support changing default item stack sizes
+        ItemType type = asItemType();
+        if (type != null) {
+            return type.getMaxStackSize();
+        }
+        // Kiterino end - Support changing default item stack sizes
         return maxStack;
     }
 
@@ -3089,6 +_,11 @@
         if (filtered.startsWith(NamespacedKey.MINECRAFT + ":")) {
             filtered = filtered.substring((NamespacedKey.MINECRAFT + ":").length());
         }
+        // Kiterino start - Injecting custom Materials
+        else {
+            filtered = filtered.replace(':', '_');
+        }
+        // Kiterino end - Injecting custom Materials
 
         filtered = filtered.toUpperCase(Locale.ROOT);
 
@@ -3744,4 +_,16 @@
         }
     }
     // Purpur end - ItemStack convenience methods
+
+    // Kiterino start - Injecting custom Materials
+    /**
+     * Checks whether this item is injected
+     * (i.e., is a non-vanilla one)
+     *
+     * @return whether this item is injected
+     */
+    public boolean isInjected() {
+        return !this.key.namespace().equals(NamespacedKey.MINECRAFT);
+    }
+    // Kiterino end - Injecting custom Materials
 }
