--- a/src/main/java/org/bukkit/Note.java
+++ b/src/main/java/org/bukkit/Note.java
@@ -127,6 +_,7 @@
     }
 
     private final byte note;
+    private final net.kyori.adventure.text.format.TextColor color; // Kiterino - Note color API
 
     /**
      * Creates a new note.
@@ -138,6 +_,7 @@
         Preconditions.checkArgument(note >= 0 && note <= 24, "The note value has to be between 0 and 24.");
 
         this.note = (byte) note;
+        this.color = getColor(note); // Kiterino - Note color API
     }
 
     /**
@@ -158,6 +_,7 @@
         }
 
         this.note = (byte) (octave * Tone.TONES_COUNT + tone.getId(sharped));
+        this.color = getColor(note); // Kiterino - Note color API
     }
 
     /**
@@ -298,4 +_,50 @@
     public String toString() {
         return "Note{" + getTone().toString() + (isSharped() ? "#" : "") + "}";
     }
+
+    // Kiterino start - Note color API
+    /**
+     * Get played note color
+     *
+     * @return note color
+     */
+    public @NotNull net.kyori.adventure.text.format.TextColor getColor() {
+        return this.color;
+    }
+
+    private static @NotNull net.kyori.adventure.text.format.TextColor getColor(int note) {
+        return switch (note) {
+            case 0 -> fromHex("#77D700");
+            case 1 -> fromHex("#95C000");
+            case 2 -> fromHex("#B2A500");
+            case 3 -> fromHex("#CC8600");
+            case 4 -> fromHex("#E26500");
+            case 5 -> fromHex("#F34100");
+            case 6 -> fromHex("#FC1E00");
+            case 7 -> fromHex("#FE000F");
+            case 8 -> fromHex("#F70033");
+            case 9 -> fromHex("#E8005A");
+            case 10 -> fromHex("#CF0083");
+            case 11 -> fromHex("#AE00A9");
+            case 12 -> fromHex("#8600CC");
+            case 13 -> fromHex("#5B00E7");
+            case 14 -> fromHex("#2D00F9");
+            case 15 -> fromHex("#020AFE");
+            case 16 -> fromHex("#0037F6");
+            case 17 -> fromHex("#0068E0");
+            case 18 -> fromHex("#009ABC");
+            case 19 -> fromHex("#00C68D");
+            case 20 -> fromHex("#00E958");
+            case 21 -> fromHex("#00FC21");
+            case 22 -> fromHex("#1FFC00");
+            case 23 -> fromHex("#59E800");
+            case 24 -> fromHex("#94C100");
+            default -> throw new IllegalArgumentException("Unknown note: " + note);
+        };
+    }
+
+    private static @NotNull net.kyori.adventure.text.format.TextColor fromHex(@NotNull String hex) {
+        return java.util.Objects.requireNonNull(net.kyori.adventure.text.format.TextColor.fromHexString(hex));
+    }
+    // Kiterino end - Note color API
 }
