--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -513,6 +_,19 @@
      */
     public boolean refreshChunk(int x, int z);
 
+    // Kiterino start - Allow individual chunk refresh
+    /**
+     * Resends the {@link Chunk} to all clients
+     *
+     * @param x X-coordinate of the chunk
+     * @param z Z-coordinate of the chunk
+     * @param players players
+     * @return Whether the chunk was actually refreshed
+     *
+     */
+    boolean refreshChunk(int x, int z, @NotNull Collection<Player> players);
+    // Kiterino end - Allow individual chunk refresh
+
     /**
      * Get a list of all players who are can view the specified chunk from their
      * client
@@ -4739,4 +_,13 @@
             }
         }
     }
+
+    // Kiterino start - Allow checking for Halloween season
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    boolean isHalloweenSeason();
+    // Kiterino end - Allow checking for Halloween season
 }
