--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
@@ -382,7 +_,7 @@
     }
 
     public void overrideEntityToSpawn(EntityType<?> entityType, Level level) {
-        this.data.reset();
+        this.data.reset(this); // Paper - Fix TrialSpawner forgets assigned mob; MC-273635
         this.normalConfig = Holder.direct(this.normalConfig.value().withSpawning(entityType));
         this.ominousConfig = Holder.direct(this.ominousConfig.value().withSpawning(entityType));
         this.setState(level, TrialSpawnerState.INACTIVE);
