--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
@@ -241,7 +_,7 @@
         BlockFace.EAST, BlockFace.EAST_SOUTH_EAST, BlockFace.SOUTH_EAST, BlockFace.SOUTH_SOUTH_EAST
     };
 
-    private static final Map<Class<? extends Block>, Function<BlockState, CraftBlockData>> INSTANCE_CREATOR = new HashMap<>();
+    public static final Map<Class<? extends Block>, Function<BlockState, CraftBlockData>> INSTANCE_CREATOR = new HashMap<>(); // Kiterino - private -> public - Data-driven blocks
 
     static {
         //<editor-fold desc="CraftBlockData Registration" defaultstate="collapsed">
@@ -438,7 +_,7 @@
         //</editor-fold>
     }
 
-    private static void register(Class<? extends Block> blockClass, Function<BlockState, CraftBlockData> newInstance) {
+    public static void register(Class<? extends Block> blockClass, Function<BlockState, CraftBlockData> newInstance) { // Kiterino - private -> public - Data-driven blocks
         Preconditions.checkState(INSTANCE_CREATOR.put(blockClass, newInstance) == null, "Duplicate mapping %s->%s", blockClass, newInstance);
     }
 
