--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -59,6 +_,18 @@
         final CampfireBlockEntity entity,
         final RecipeManager.CachedCheck<SingleRecipeInput, CampfireCookingRecipe> recipeCache
     ) {
+        // Kiterino start - Campfire options
+        if (entity.checkGravity(level, pos, state)) {
+            return;
+        }
+        if (level.kiterinoConfig.campfireBurnOutInRainChance > 0 && level.getRandom().nextDouble() < level.kiterinoConfig.campfireBurnOutInRainChance) {
+            if (state.getValue(CampfireBlock.WATERLOGGED) || level.isRainingAt(pos.above())) {
+                level.setBlock(pos, state.setValue(CampfireBlock.LIT, false), 3);
+                level.playSound(null, pos.getX() + 0.5, pos.getY() + 1D, pos.getZ() + 0.5, net.minecraft.sounds.SoundEvents.FIRE_EXTINGUISH, net.minecraft.sounds.SoundSource.BLOCKS, 1F, level.getRandom().nextFloat() * 0.4F + 0.8F);
+                setChanged(level, pos, state);
+            }
+        }
+        // Kiterino end - Campfire options
         boolean changed = false;
 
         for (int slot = 0; slot < entity.items.size(); slot++) {
@@ -115,6 +_,9 @@
     }
 
     public static void cooldownTick(final Level level, final BlockPos pos, final BlockState state, final CampfireBlockEntity entity) {
+        if (entity.checkGravity(level, pos, state)) {
+            return;
+        }
         boolean changed = false;
 
         for (int slot = 0; slot < entity.items.size(); slot++) {
@@ -129,6 +_,23 @@
         }
     }
 
+    // Kiterino start - Campfire options
+    private boolean checkGravity(Level level, BlockPos pos, BlockState state) {
+        if (!level.kiterinoConfig.campfireAffectedByGravity) {
+            return false;
+        }
+
+        BlockPos below = pos.below();
+        if (level.isOutsideBuildHeight(below.getY()) || net.minecraft.world.level.block.FallingBlock.isFree(level.getBlockState(below))) {
+            level.setBlock(pos, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(), Block.UPDATE_CLIENTS);
+            level.addFreshEntity(new net.minecraft.world.entity.item.FallingBlockEntity(level, pos.getX() + 0.5D, pos.getY(), pos.getZ() + 0.5D, state));
+            return true;
+        }
+
+        return false;
+    }
+    // Kiterino end - Campfire options
+
     public static void particleTick(final Level level, final BlockPos pos, final BlockState state, final CampfireBlockEntity entity) {
         RandomSource random = level.getRandom();
         if (random.nextFloat() < 0.11F) {
@@ -210,7 +_,21 @@
         CompoundTag var4;
         try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(this.problemPath(), LOGGER)) {
             TagValueOutput output = TagValueOutput.createWithContext(reporter, registries);
-            ContainerHelper.saveAllItems(output, this.items, true);
+            // Kiterino start - Spoof sent items on block entities
+            NonNullList<ItemStack> items = NonNullList.withSize(this.items.size(), ItemStack.EMPTY);
+            for (int i = 0; i < this.items.size(); i++) {
+                ItemStack original = this.items.get(i);
+                var context = new me.sosedik.kiterino.modifier.item.context.packet.BaseItemContext(me.sosedik.kiterino.modifier.item.context.ItemModifierContextType.CAMPFIRE, null);
+                var contextBox = new me.sosedik.kiterino.modifier.item.ItemContextBox(null, context, original.asBukkitCopy());
+                var bukkitItem = me.sosedik.kiterino.modifier.item.ItemModifier.modifyItem(contextBox);
+                if (bukkitItem == null) {
+                    items.set(i, original);
+                } else {
+                    items.set(i, ItemStack.fromBukkitCopy(bukkitItem));
+                }
+            }
+            ContainerHelper.saveAllItems(output, items, true);
+            // Kiterino end - Spoof sent items on block entities
             var4 = output.buildResult();
         }
 
