--- a/net/minecraft/core/component/DataComponentMap.java
+++ b/net/minecraft/core/component/DataComponentMap.java
@@ -175,7 +_,7 @@
             }
         }
 
-        private record SimpleMap(Reference2ObjectMap<DataComponentType<?>, Object> map) implements DataComponentMap {
+        public record SimpleMap(Reference2ObjectMap<DataComponentType<?>, Object> map) implements DataComponentMap { // Kiterino - Make public
             @Override
             public <T> @Nullable T get(final DataComponentType<? extends T> type) {
                 return (T)this.map.get(type);
