--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -29,6 +_,7 @@
             //  Purpur start - Infinity bow settings
             if (level.purpurConfig.infinityWorksWithoutArrows && projectile.isEmpty() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.INFINITY, itemStack) > 0) {
                 projectile = new ItemStack(Items.ARROW);
+                projectile.set(net.minecraft.core.component.DataComponents.INTANGIBLE_PROJECTILE, net.minecraft.util.Unit.INSTANCE); // Kiterino - Infinite projectiles
             }
             // Purpur end - Infinity bow settings
             if (projectile.isEmpty()) {
@@ -41,7 +_,7 @@
                 return false;
             }
 
-            List<ItemStack> firedProjectiles = draw(itemStack, projectile, player);
+            List<ItemStack> firedProjectiles = draw(itemStack, projectile, player, !projectile.has(net.minecraft.core.component.DataComponents.INTANGIBLE_PROJECTILE)); // Kiterino - Infinite projectiles
             if (level instanceof ServerLevel serverLevel && !firedProjectiles.isEmpty()) {
                 this.shoot(serverLevel, player, player.getUsedItemHand(), itemStack, firedProjectiles, pow * 3.0F, (float) serverLevel.purpurConfig.bowProjectileOffset, pow == 1.0F, null, pow); // Paper - Pass draw strength // Purpur - Projectile offset config
             }
