--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -31,6 +_,7 @@
             //  Purpur start - Infinity bow settings
             if (level.purpurConfig.infinityWorksWithoutArrows && projectile.isEmpty() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.INFINITY, stack) > 0) {
                 projectile = new ItemStack(Items.ARROW);
+                projectile.set(net.minecraft.core.component.DataComponents.INTANGIBLE_PROJECTILE, net.minecraft.util.Unit.INSTANCE); // Kiterino - Infinite projectiles
             }
             // Purpur end - Infinity bow settings
             if (projectile.isEmpty()) {
@@ -41,7 +_,7 @@
                 if (powerForTime < 0.1) {
                     return false;
                 } else {
-                    List<ItemStack> list = draw(stack, projectile, player);
+                    List<ItemStack> list = draw(stack, projectile, player, !projectile.has(net.minecraft.core.component.DataComponents.INTANGIBLE_PROJECTILE)); // Kiterino - Infinite projectiles
                     if (level instanceof ServerLevel serverLevel && !list.isEmpty()) {
                         this.shoot(serverLevel, player, player.getUsedItemHand(), stack, list, powerForTime * 3.0F, (float) serverLevel.purpurConfig.bowProjectileOffset, powerForTime == 1.0F, null, powerForTime); // Paper - Pass draw strength // Purpur - Projectile offset config
                     }
