--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -196,7 +_,25 @@
                 } else {
                     Holder<Item> holder = Item.STREAM_CODEC.decode(buffer);
                     DataComponentPatch dataComponentPatch = codec.decode(buffer);
-                    return new ItemStack(holder, varInt, dataComponentPatch);
+                    // Kiterino start - Prevent creative from overriding items
+                    var itemStack = new ItemStack(holder, varInt, dataComponentPatch);
+                    if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride) {
+                        net.minecraft.world.item.component.CustomData customData = itemStack.components.get(DataComponents.CUSTOM_DATA);
+                        if (customData != null && customData.contains("kiterino_og_item")) {
+                            net.minecraft.nbt.Tag itemTag = customData.getUnsafe().get("kiterino_og_item");
+                            assert itemTag != null;
+                            if (itemTag instanceof net.minecraft.nbt.StringTag) {
+                                itemStack = ItemStack.EMPTY;
+                            } else {
+                                itemStack = net.minecraft.world.item.ItemStack.CODEC.parse(
+                                    me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.SERIALIZATION_CONTEXT,
+                                    itemTag
+                                ).getOrThrow();
+                            }
+                        }
+                    }
+                    return itemStack;
+                    // Kiterino end - Prevent creative from overriding items
                 }
             }
 
@@ -207,6 +_,23 @@
                 } else {
                     buffer.writeVarInt(io.papermc.paper.util.sanitizer.ItemComponentSanitizer.sanitizeCount(io.papermc.paper.util.sanitizer.ItemObfuscationSession.currentSession(), value, value.getCount())); // Paper - potentially sanitize count
                     Item.STREAM_CODEC.encode(buffer, value.getItemHolder());
+                    // Kiterino start - Prevent creative from overriding items
+                    net.minecraft.world.item.component.CustomData customData = null;
+                    if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride && !value.isEmpty() && me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.modifiers.isEmpty()) {
+                        customData = value.get(DataComponents.CUSTOM_DATA);
+                        net.minecraft.nbt.Tag itemData = net.minecraft.world.item.ItemStack.CODEC.encodeStart(
+                            me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.SERIALIZATION_CONTEXT,
+                            value
+                        ).getOrThrow();
+                        if (customData == null) {
+                            customData = net.minecraft.world.item.component.CustomData.of(new net.minecraft.nbt.CompoundTag());
+                            customData.getUnsafe().put("kiterino_og_item", itemData);
+                            value.set(DataComponents.CUSTOM_DATA, customData);
+                        } else {
+                            customData.getUnsafe().put("kiterino_og_item", itemData);
+                        }
+                    }
+                    // Kiterino end - Prevent creative from overriding items
                     // Paper start - adventure; conditionally render translatable components
                     boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
                     try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(value))) { // pass the itemstack as context to the obfuscation session
@@ -214,6 +_,15 @@
                     codec.encode(buffer, value.components.asPatch());
                     } finally {
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
+                        // Kiterino start - Prevent creative from overriding items
+                        if (customData != null) {
+                            if (customData.getUnsafe().size() == 1) {
+                                value.remove(DataComponents.CUSTOM_DATA);
+                            } else {
+                                customData.getUnsafe().remove("kiterino_og_item");
+                            }
+                        }
+                        // Kiterino end - Prevent creative from overriding items
                     }
                     // Paper end - adventure; conditionally render translatable components
                 }
