--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -181,8 +_,53 @@
 
                 Holder<Item> item = Item.STREAM_CODEC.decode(input);
                 DataComponentPatch patch = patchCodec.decode(input);
-                return new ItemStack(item, count, patch);
-            }
+                // Kiterino start - Prevent creative from overriding items
+                var itemStack = new ItemStack(item, count, patch);
+                if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride) {
+                    net.minecraft.world.item.component.CustomData customData = itemStack.components.get(DataComponents.CUSTOM_DATA);
+                    if (customData != null && customData.contains("kiterino_og_item")) {
+                        net.minecraft.nbt.Tag itemTag = customData.getUnsafe().get("kiterino_og_item");
+                        assert itemTag != null;
+                        itemStack = decodeItem(input, itemTag);
+                    }
+                }
+                return itemStack;
+                // Kiterino end - Prevent creative from overriding items
+            }
+
+            // Kiterino start - Prevent creative from overriding items
+            private ItemStack decodeItem(RegistryFriendlyByteBuf input, net.minecraft.nbt.Tag itemTag) {
+                if (!(itemTag instanceof net.minecraft.nbt.ByteArrayTag tag)) {
+                    return ItemStack.EMPTY;
+                }
+
+                byte[] serializedItem = tag.asByteArray().orElseGet(() -> new byte[0]);
+                if (serializedItem.length == 0) {
+                    return ItemStack.EMPTY;
+                }
+
+                net.minecraft.nbt.CompoundTag compound;
+                try {
+                    compound = net.minecraft.nbt.NbtIo.readCompressed(
+                        new java.io.ByteArrayInputStream(serializedItem), net.minecraft.nbt.NbtAccounter.unlimitedHeap()
+                    );
+                } catch (java.io.IOException _) {
+                    return ItemStack.EMPTY;
+                }
+                if (compound.getStringOr("id", "minecraft:air").equals("minecraft:air")) {
+                    return ItemStack.EMPTY;
+                }
+
+                DataResult<ItemStack> dataResult = ItemStack.CODEC.parse(
+                    input.registryAccess().createSerializationContext(net.minecraft.nbt.NbtOps.INSTANCE), compound
+                );
+                if (dataResult.isError()) {
+                    return ItemStack.EMPTY;
+                }
+
+                return dataResult.getOrThrow();
+            }
+            // Kiterino end - Prevent creative from overriding items
 
             @Override
             public void encode(final RegistryFriendlyByteBuf output, final ItemStack itemStack) {
@@ -191,6 +_,32 @@
                 } else {
                     output.writeVarInt(io.papermc.paper.util.sanitizer.ItemComponentSanitizer.sanitizeCount(io.papermc.paper.util.sanitizer.ItemObfuscationSession.currentSession(), itemStack, itemStack.getCount())); // Paper - potentially sanitize count
                     Item.STREAM_CODEC.encode(output, itemStack.typeHolder());
+                    // Kiterino start - Prevent creative from overriding items
+                    net.minecraft.world.item.component.CustomData customData = null;
+                    if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride && !itemStack.isEmpty() && me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.modifiers.isEmpty()) {
+                        try {
+                            customData = itemStack.get(DataComponents.CUSTOM_DATA);
+                            net.minecraft.nbt.CompoundTag itemData = (net.minecraft.nbt.CompoundTag) net.minecraft.world.item.ItemStack.CODEC.encodeStart(
+                                me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.SERIALIZATION_CONTEXT,
+                                itemStack
+                            ).getOrThrow();
+                            var outputStream = new java.io.ByteArrayOutputStream();
+                            net.minecraft.nbt.NbtIo.writeCompressed(
+                                itemData,
+                                outputStream
+                            );
+                            byte[] serializedItem = outputStream.toByteArray();
+                            if (customData == null) {
+                                customData = net.minecraft.world.item.component.CustomData.of(new net.minecraft.nbt.CompoundTag());
+                                customData.getUnsafe().putByteArray("kiterino_og_item", serializedItem);
+                                itemStack.set(DataComponents.CUSTOM_DATA, customData);
+                            } else {
+                                customData.getUnsafe().putByteArray("kiterino_og_item", serializedItem);
+                            }
+                        } catch (java.lang.IllegalStateException | java.io.IOException _) {
+                        }
+                    }
+                    // Kiterino end - Prevent creative from overriding items
                     // Paper start - adventure; conditionally render translatable components
                     boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
                     try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(itemStack))) { // pass the itemstack as context to the obfuscation session
@@ -198,6 +_,15 @@
                     patchCodec.encode(output, itemStack.components.asPatch());
                     } finally {
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
+                        // Kiterino start - Prevent creative from overriding items
+                        if (customData != null) {
+                            if (customData.getUnsafe().size() == 1) {
+                                itemStack.remove(DataComponents.CUSTOM_DATA);
+                            } else {
+                                customData.getUnsafe().remove("kiterino_og_item");
+                            }
+                        }
+                        // Kiterino end - Prevent creative from overriding items
                     }
                     // Paper end - adventure; conditionally render translatable components
                 }
@@ -282,7 +_,7 @@
     }
 
     private static DataResult<?> validateComponents(final DataComponentMap components) {
-        if (components.has(DataComponents.MAX_DAMAGE) && components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1) > 1) {
+        if (false && components.has(DataComponents.MAX_DAMAGE) && components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1) > 1) { // Kiterino - Allow stackable damageable items
             return DataResult.error(() -> "Item cannot be both damageable and stackable");
         }
 
