--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -193,6 +_,15 @@
         }
         org.purpurmc.purpur.PurpurConfig.registerCommands();
         // Purpur end - Purpur config files
+        // Kiterino start - Kiterino Config
+        try {
+            me.sosedik.kiterino.KiterinoConfig.init((java.io.File) options.valueOf("kiterino-settings"));
+            me.sosedik.kiterino.KiterinoInitializer.initPostConfig();
+        } catch (Exception e) {
+            DedicatedServer.LOGGER.error("Unable to load Kiterino server configuration", e);
+            return false;
+        }
+        // Kiterino end - Kiterino Config
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
 
         this.setPvpAllowed(properties.pvp);
@@ -264,6 +_,13 @@
             }
         }
         // Purpur end - UPnP Port Forwarding
+
+        // Kiterino start - Init/destruct injectors
+        if (!Boolean.getBoolean("Kiterino.testingBase")) {
+            me.sosedik.kiterino.util.KiterinoBootstrapMaterialInjectorImpl.destruct(); // Kiterino - Injecting custom Materials
+            me.sosedik.kiterino.util.KiterinoBootstrapEntityTypeInjectorImpl.destruct(); // Kiterino - Injecting custom entities
+        }
+        // Kiterino end - Init/destruct injectors
 
         // CraftBukkit start
         this.server.loadPlugins();
