--- a/src/main/java/io/papermc/paper/event/player/AsyncPlayerSpawnLocationEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/AsyncPlayerSpawnLocationEvent.java
@@ -25,10 +_,22 @@
     private final PlayerConfigurationConnection connection;
     private final boolean newPlayer;
     private Location spawnLocation;
+    // Kiterino start - Expose invalid world to PlayerSpawnLocationEvent
+    private final Location initialLocation;
+    private final org.bukkit.@org.jspecify.annotations.Nullable NamespacedKey initialDimensionId;
+    // Kiterino end - Expose invalid world to PlayerSpawnLocationEvent
 
     @ApiStatus.Internal
     public AsyncPlayerSpawnLocationEvent(final PlayerConfigurationConnection connection, final Location spawnLocation, final boolean newPlayer) {
+        // Kiterino start - Expose invalid world to PlayerSpawnLocationEvent
+        this(connection, spawnLocation, newPlayer, null, null);
+    }
+
+    public AsyncPlayerSpawnLocationEvent(final PlayerConfigurationConnection connection, final Location spawnLocation, final boolean newPlayer, @org.jspecify.annotations.Nullable Location initialLocation, @org.jspecify.annotations.Nullable String initialDimensionId) {
         super(true);
+        this.initialLocation = initialLocation == null ? spawnLocation.clone() : initialLocation;
+        this.initialDimensionId = initialDimensionId == null ? null : org.bukkit.NamespacedKey.fromString(initialDimensionId);
+        // Kiterino end - Expose invalid world to PlayerSpawnLocationEvent
         this.connection = connection;
         this.spawnLocation = spawnLocation;
         this.newPlayer = newPlayer;
@@ -75,6 +_,36 @@
     public boolean isNewPlayer() {
         return this.newPlayer;
     }
+
+    // Kiterino start - Expose invalid world to PlayerSpawnLocationEvent
+    /**
+     * Gets the initial spawn location the player had.
+     * <br>The initial location might have a nullable world in case the world isn't loaded.
+     *
+     * @return initial spawn location
+     */
+    public Location getInitialLocation() {
+        return this.initialLocation.clone();
+    }
+
+    /**
+     * Checks whether the player is initially in an unloaded world
+     *
+     * @return whether the player is initially in an unloaded world
+     */
+    public boolean isInitiallyInUnloadedWorld() {
+        return this.initialLocation.getWorld() == null;
+    }
+
+    /**
+     * Gets the initial dimension id, if present
+     *
+     * @return the initial dimension id
+     */
+    public org.bukkit.@org.jspecify.annotations.Nullable NamespacedKey getInitialDimensionId() {
+        return this.initialDimensionId;
+    }
+    // Kiterino end - Expose invalid world to PlayerSpawnLocationEvent
 
     @Override
     public HandlerList getHandlers() {
