--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1352,13 +_,19 @@
 
     // CraftBukkit start
     public @Nullable TeleportTransition findRespawnPositionAndUseSpawnBlock(boolean useCharge, TeleportTransition.PostTeleportTransition postTeleportTransition, @Nullable org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason) {
+        // Kiterino start - Call EntityPortalReadyEvent for End Portal
+        return findRespawnPositionAndUseSpawnBlock(useCharge, postTeleportTransition, respawnReason, null);
+    }
+
+    public @Nullable TeleportTransition findRespawnPositionAndUseSpawnBlock(boolean useCharge, TeleportTransition.PostTeleportTransition postTeleportTransition, @Nullable org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason, @Nullable ServerLevel levelTo) {
+        // Kiterino end - Call EntityPortalReadyEvent for End Portal
         TeleportTransition teleportTransition;
         boolean isBedSpawn = false;
         boolean isAnchorSpawn = false;
         Runnable consumeAnchorCharge = null;
         // CraftBukkit end
         ServerPlayer.RespawnConfig respawnConfig = this.getRespawnConfig();
-        ServerLevel level = this.server.getLevel(ServerPlayer.RespawnConfig.getDimensionOrDefault(respawnConfig));
+        ServerLevel level = levelTo == null ? this.server.getLevel(ServerPlayer.RespawnConfig.getDimensionOrDefault(respawnConfig)) : levelTo; // Kiterino - Call EntityPortalReadyEvent for End Portal
         if (level != null && respawnConfig != null) {
             Optional<ServerPlayer.RespawnPosAngle> optional = findRespawnAndUseSpawnBlock(level, respawnConfig, useCharge);
             if (optional.isPresent()) {
@@ -1687,7 +_,7 @@
             this.setRespawnPosition(
                 new ServerPlayer.RespawnConfig(LevelData.RespawnData.of(this.level().dimension(), at, this.getYRot(), this.getXRot()), false), true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.BED // Paper - Add PlayerSetSpawnEvent
             );
-            if (this.level().isBrightOutside()) {
+            if (!this.level().kiterinoConfig.dayDreaming && this.level().isBrightOutside()) { // Kiterino - Sleeping options
                 return Either.left(Player.BedSleepingProblem.NOT_POSSIBLE_NOW);
             } else {
                 if (!this.isCreative()) {
@@ -2040,8 +_,13 @@
                     }
                 }
             } else if (this.isFallFlying()) {
-                int rounded = Math.round((float)Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
-                this.awardStat(Stats.AVIATE_ONE_CM, rounded);
+                // Kiterino start - Require wearing an item for elytra distance statistics
+                ItemStack chestItem = getItemBySlot(net.minecraft.world.entity.EquipmentSlot.CHEST);
+                if (!me.sosedik.kiterino.KiterinoConfig.aviateRequireItem || (chestItem.has(DataComponents.GLIDER) && !chestItem.isBroken())) {
+                    int rounded = Math.round((float)Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
+                    this.awardStat(Stats.AVIATE_ONE_CM, rounded);
+                }
+                // Kiterino end - Require wearing an item for elytra distance statistics
             } else {
                 int rounded = Math.round((float)Math.sqrt(dx * dx + dz * dz) * 100.0F);
                 if (rounded > 25) {
@@ -2055,6 +_,11 @@
         if (this.isPassenger() && !didNotMove(dx, dy, dz)) {
             int rounded = Math.round((float)Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
             Entity vehicle = this.getVehicle();
+            // Kiterino start - Don't count Purpur's riding towards distance statistics
+            if (me.sosedik.kiterino.KiterinoConfig.vehicleIgnoreRider && vehicle != null && vehicle.getRider() == this) {
+                return;
+            }
+            // Kiterino end - Don't count Purpur's riding towards distance statistics
             if (vehicle instanceof AbstractMinecart) {
                 this.awardStat(Stats.MINECART_ONE_CM, rounded);
             } else if (vehicle instanceof AbstractBoat) {
