--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -827,7 +_,7 @@
                 this.foodData.tick(this);
                 this.awardStat(Stats.PLAY_TIME);
                 this.awardStat(Stats.TOTAL_WORLD_TIME);
-                if (this.isAlive()) {
+                if (this.isAlive() && (!me.sosedik.kiterino.KiterinoConfig.surviveIgnoreInvulnerability || !isInvulnerable())) { // Kiterino - Ignore invulnerable player for time since death
                     this.awardStat(Stats.TIME_SINCE_DEATH);
                 }
 
@@ -1359,7 +_,13 @@
         final boolean consumeSpawnBlock, final TeleportTransition.PostTeleportTransition postTeleportTransition,
         final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason
     ) {
-        RespawnResult result = this.findRespawnPositionAndUseSpawnBlock0(consumeSpawnBlock, postTeleportTransition, respawnReason);
+        // Kiterino start - Call EntityPortalReadyEvent for End Portal
+        return findRespawnPositionAndUseSpawnBlock(consumeSpawnBlock, postTeleportTransition, respawnReason, null);
+    }
+
+    public @Nullable TeleportTransition findRespawnPositionAndUseSpawnBlock(boolean consumeSpawnBlock, TeleportTransition.PostTeleportTransition postTeleportTransition, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason, @Nullable ServerLevel levelTo) {
+        // Kiterino end - Call EntityPortalReadyEvent for End Portal
+        RespawnResult result = this.findRespawnPositionAndUseSpawnBlock0(consumeSpawnBlock, postTeleportTransition, respawnReason, levelTo); // Kiterino - Call EntityPortalReadyEvent for End Portal
         return result == null ? null : result.transition();
     }
 
@@ -1367,13 +_,19 @@
         final boolean consumeSpawnBlock, final TeleportTransition.PostTeleportTransition postTeleportTransition,
         final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason
     ) {
+        // Kiterino start - Call EntityPortalReadyEvent for End Portal
+        return findRespawnPositionAndUseSpawnBlock0(consumeSpawnBlock, postTeleportTransition, respawnReason, null);
+    }
+
+    public @Nullable RespawnResult findRespawnPositionAndUseSpawnBlock0(boolean consumeSpawnBlock, TeleportTransition.PostTeleportTransition postTeleportTransition, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason, @Nullable ServerLevel levelTo) {
+        // Kiterino end - Call EntityPortalReadyEvent for End Portal
         TeleportTransition teleportTransition;
         boolean isBedSpawn = false;
         boolean isAnchorSpawn = false;
         Runnable consumeAnchorCharge = null;
         // Paper end
         ServerPlayer.RespawnConfig respawnConfig = this.getRespawnConfig();
-        ServerLevel respawnLevel = this.server.getLevel(ServerPlayer.RespawnConfig.getDimensionOrDefault(respawnConfig));
+        ServerLevel respawnLevel = levelTo == null ? this.server.getLevel(ServerPlayer.RespawnConfig.getDimensionOrDefault(respawnConfig)) : levelTo; // Kiterino - Call EntityPortalReadyEvent for End Portal
         if (respawnLevel != null && respawnConfig != null) {
             Optional<ServerPlayer.RespawnPosAngle> respawn = findRespawnAndUseSpawnBlock(respawnLevel, respawnConfig, consumeSpawnBlock);
             if (respawn.isPresent()) {
@@ -2073,8 +_,13 @@
                     }
                 }
             } else if (this.isFallFlying()) {
-                int distance = Math.round((float)Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
-                this.awardStat(Stats.AVIATE_ONE_CM, distance);
+                // Kiterino start - Require wearing an item for elytra distance statistics
+                ItemStack chestItem = getItemBySlot(net.minecraft.world.entity.EquipmentSlot.CHEST);
+                if (!me.sosedik.kiterino.KiterinoConfig.aviateRequireItem || (chestItem.has(DataComponents.GLIDER) && !chestItem.isBroken())) {
+                    int distance = Math.round((float) Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
+                    this.awardStat(Stats.AVIATE_ONE_CM, distance);
+                }
+                // Kiterino end - Require wearing an item for elytra distance statistics
             } else {
                 int horizontalDistance = Math.round((float)Math.sqrt(dx * dx + dz * dz) * 100.0F);
                 if (horizontalDistance > 25) {
@@ -2088,6 +_,11 @@
         if (this.isPassenger() && !didNotMove(dx, dy, dz)) {
             int distance = Math.round((float)Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
             Entity vehicle = this.getVehicle();
+            // Kiterino start - Don't count Purpur's riding towards distance statistics
+            if (me.sosedik.kiterino.KiterinoConfig.vehicleIgnoreRider && vehicle != null && vehicle.getRider() == this) {
+                return;
+            }
+            // Kiterino end - Don't count Purpur's riding towards distance statistics
             if (vehicle instanceof AbstractMinecart) {
                 this.awardStat(Stats.MINECART_ONE_CM, distance);
             } else if (vehicle instanceof AbstractBoat) {
