--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -102,6 +_,11 @@
     protected void setGameModeForPlayer(GameType gameMode, @Nullable GameType previousGameMode) {
         this.previousGameModeForPlayer = previousGameMode;
         this.gameModeForPlayer = gameMode;
+        // Kiterino start - Prevent creative from overriding items
+        if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride && gameMode == GameType.CREATIVE) {
+            this.player.containerMenu.sendAllDataToRemote();
+        }
+        // Kiterino end - Prevent creative from overriding items
         Abilities abilities = this.player.getAbilities();
         gameMode.updatePlayerAbilities(abilities);
     }
