--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -91,6 +_,11 @@
     protected void setGameModeForPlayer(GameType gameModeForPlayer, @Nullable GameType previousGameModeForPlayer) {
         this.previousGameModeForPlayer = previousGameModeForPlayer;
         this.gameModeForPlayer = gameModeForPlayer;
+        // Kiterino start - Prevent creative from overriding items
+        if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride && gameModeForPlayer == GameType.CREATIVE) {
+            this.player.containerMenu.sendAllDataToRemote();
+        }
+        // Kiterino end - Prevent creative from overriding items
         gameModeForPlayer.updatePlayerAbilities(this.player.getAbilities());
     }
 
