--- a/net/minecraft/world/entity/monster/Drowned.java
+++ b/net/minecraft/world/entity/monster/Drowned.java
@@ -92,6 +_,17 @@
     }
     // Purpur end - Ridables
 
+    // Kiterino start - Allow jumping in water as Drowned
+    @Override
+    public boolean onSpacebar() {
+        if (isInWater()) {
+            jumpFromGround();
+            return true;
+        }
+        return false;
+    }
+    // Kiterino end - Allow jumping in water as Drowned
+
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
