--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -1267,7 +_,7 @@
 
         @Override
         public boolean shouldPanic() {
-            return !Fox.this.isDefending() && super.shouldPanic();
+            return (!Fox.this.isDefending() && super.shouldPanic()) || this.mob.isForcePanicking(); // Kiterino - Forced panic mode API
         }
     }
 
