--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1228,6 +_,16 @@
     private final FeatureFlagSet requiredFeatures;
     private final boolean allowedInPeaceful;
 
+    // Kiterino start - Injecting custom entities
+    static {
+        me.sosedik.kiterino.util.KiterinoBootstrapEntityTypeInjectorImpl.setup();
+    }
+
+    public static <T extends Entity> EntityType<T> register(net.kyori.adventure.key.Key key, EntityType.Builder<T> builder) {
+        return register(ResourceKey.create(Registries.ENTITY_TYPE, Identifier.fromNamespaceAndPath(key.namespace(), key.value())), builder);
+    }
+    // Kiterino end - Injecting custom entities
+
     private static <T extends Entity> EntityType<T> register(ResourceKey<EntityType<?>> key, EntityType.Builder<T> builder) {
         return Registry.register(BuiltInRegistries.ENTITY_TYPE, key, builder.build(key));
     }
