--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -303,6 +_,7 @@
     public final RandomSource random; // Paper - Share random for entities to make them more random // Purpur - Add toggle for RNG manipulation
     public int tickCount;
     private int remainingFireTicks;
+    public int forcePanicTicks; public boolean isForcePanicking() { return forcePanicTicks > 0; } // Kiterino - Forced panic mode API
     private final EntityFluidInteraction fluidInteraction = new EntityFluidInteraction(Set.of(FluidTags.WATER, FluidTags.LAVA));
     protected boolean wasTouchingWater;
     protected boolean wasEyeInWater;
@@ -944,6 +_,12 @@
             this.clearFire();
         }
 
+        // Kiterino start - Forced panic mode API
+        if (this.forcePanicTicks > 0) {
+            this.forcePanicTicks--;
+        }
+        // Kiterino end - Forced panic mode API
+
         if (this.isInLava()) {
             this.fallDistance *= 0.5;
             // CraftBukkit start
@@ -1297,6 +_,30 @@
                 this.setDeltaMovement(this.getDeltaMovement().multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
                 profiler.pop();
             }
+            // Kiterino start - Observers detect moving entities
+            if (this.level.kiterinoConfig.observersDetectMovingEntities && this instanceof net.minecraft.world.entity.LivingEntity && movement.lengthSqr() > 0.00615D) {
+                profiler.push("move");
+                profiler.push("kiterino:observerCheck");
+                double height = Math.min(getBoundingBox().getYsize(), 3D);
+                BlockPos base = blockPosition();
+                do {
+                    for (Direction direction : Direction.Plane.HORIZONTAL) {
+                        BlockPos relative = base.relative(direction);
+                        BlockState relativeBlockState = this.level.getBlockState(relative);
+                        if (relativeBlockState.getBlock() == Blocks.OBSERVER && !this.level.getBlockState(base).isSolid()
+                            && !relativeBlockState.getValue(net.minecraft.world.level.block.ObserverBlock.POWERED)
+                            && relativeBlockState.getValue(net.minecraft.world.level.block.ObserverBlock.FACING) == direction.getOpposite()
+                            && this.level instanceof ServerLevel serverLevel) {
+                            relativeBlockState.tick(serverLevel, relative.immutable(), this.level.getRandom());
+                        }
+                    }
+                    height--;
+                    base = base.above();
+                } while (height > 0);
+                profiler.pop();
+                profiler.pop();
+            }
+            // Kiterino end - Observers detect moving entities
         }
         // Paper start - detailed watchdog information
         } finally {
@@ -2160,6 +_,18 @@
         return this.isInWater() || this.isInLava();
     }
 
+    // Kiterino start - MC-145311, Treat entities in water cauldrons as wet
+    public boolean isInCauldronFilledWithWater() {
+        BlockState state = getInBlockState();
+        if (!state.is(Blocks.WATER_CAULDRON)) {
+            return false;
+        }
+        BlockPos pos = blockPosition();
+        double contentHeight = (6D + state.getValue(net.minecraft.world.level.block.LayeredCauldronBlock.LEVEL) * 3D) / 16D; // Copypaste from LayeredCauldronBlock#getContentHeight
+        return getY() < pos.getY() + contentHeight && getBoundingBox().maxY > pos.getY() + 0.25D;
+    }
+    // Kiterino end - MC-145311, Treat entities in water cauldrons as wet
+
     public boolean isUnderWater() {
         return this.wasEyeInWater && this.isInWater();
     }
@@ -2194,7 +_,7 @@
 
     protected boolean updateFluidInteraction() {
         this.fluidInteraction.update(this, !this.isPushedByFluid());
-        boolean inWater = this.fluidInteraction.isInFluid(FluidTags.WATER);
+        boolean inWater = this.fluidInteraction.isInFluid(FluidTags.WATER) || isInCauldronFilledWithWater(); // Kiterino - MC-145311, Treat entities in water cauldrons as wet
         boolean inLava = this.fluidInteraction.isInFluid(FluidTags.LAVA);
         if (inWater) {
             this.resetFallDistance();
@@ -2612,6 +_,18 @@
         if (victim instanceof ServerPlayer serverPlayer) {
             CriteriaTriggers.ENTITY_KILLED_PLAYER.trigger(serverPlayer, this, killingBlow);
         }
+        // Kiterino start - Award stats to rider
+        if (me.sosedik.kiterino.KiterinoConfig.awardStatsToRider && victim != this && getRider() instanceof ServerPlayer riderPlayer) {
+            DamageSource killBlow = killingBlow.copy();
+            if (killBlow.directEntity == this) {
+                killBlow.directEntity = riderPlayer;
+            }
+            if (killBlow.causingEntity == this) {
+                killBlow.causingEntity = riderPlayer;
+            }
+            riderPlayer.awardKillScore(victim, killBlow);
+        }
+        // Kiterino end - Award stats to rider
     }
 
     public boolean shouldRender(final double camX, final double camY, final double camZ) {
@@ -3428,10 +_,30 @@
     }
 
     public void ejectPassengers() {
+        // Kiterino start - Properly pass dismount cancellation for cross-dimension teleport
+        ejectPassengers(false);
+    }
+    public void ejectPassengers(boolean suppressCancellation) {
+        // Kiterino end - Properly pass dismount cancellation for cross-dimension teleport
         for (int i = this.passengers.size() - 1; i >= 0; i--) {
-            this.passengers.get(i).stopRiding();
+            this.passengers.get(i).stopRiding(suppressCancellation); // Kiterino - Properly pass dismount cancellation for cross-dimension teleport
         }
     }
+
+    // Kiterino start - Allow setting entity's rider
+    public boolean setRider(Player rider) {
+        if (!new org.bukkit.event.entity.EntityMountEvent(rider.getBukkitEntity(), this.getBukkitEntity()).callEvent()) return false;
+        if (!rider.startRiding(this, true, true)) return false;
+
+        onMount(rider);
+        this.rider = rider;
+        return true;
+    }
+
+    public boolean hasRider() {
+        return getRider() != null;
+    }
+    // Kiterino end - Allow setting entity's rider
 
     public void removeVehicle() {
         // Paper start - Force entity dismount during teleportation
@@ -4235,12 +_,18 @@
     private @Nullable Entity teleportCrossDimension(final ServerLevel oldLevel, final ServerLevel newLevel, final TeleportTransition transition) {
         List<Entity> oldPassengers = this.getPassengers();
         List<Entity> newPassengers = new ArrayList<>(oldPassengers.size());
-        this.ejectPassengers();
+        Player preservedRider = this.getRider(); // Kiterino - Preserve rider on cross-dimension teleport
+        this.ejectPassengers(true); // Kiterino - Properly pass dismount cancellation for cross-dimension teleport
 
         for (Entity passenger : oldPassengers) {
             Entity newPassenger = passenger.teleport(this.calculatePassengerTransition(transition, passenger));
             if (newPassenger != null) {
                 newPassengers.add(newPassenger);
+                // Kiterino start - Preserve rider on cross-dimension teleport
+                if (passenger == preservedRider && newPassenger instanceof Player newRider) {
+                    preservedRider = newRider;
+                }
+                // Kiterino end - Preserve rider on cross-dimension teleport
             }
         }
 
@@ -4262,9 +_,16 @@
         newEntity.teleportSetPosition(PositionMoveRotation.of(this), PositionMoveRotation.of(transition), transition.relatives());
         if (this.inWorld) newLevel.addDuringTeleport(newEntity); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned
 
+        newEntity.rider = preservedRider; // Kiterino - Preserve rider on cross-dimension teleport
         for (Entity newPassenger : newPassengers) {
             newPassenger.startRiding(newEntity, true, false);
         }
+
+        // Kiterino start - Preserve rider on cross-dimension teleport
+        if (newEntity.rider != null && !newEntity.getPassengers().contains(newEntity.rider)) {
+            newEntity.rider = null;
+        }
+        // Kiterino end - Preserve rider on cross-dimension teleport
 
         newLevel.resetEmptyTime();
         transition.postTeleportTransition().onTransition(newEntity);
