--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -973,10 +_,20 @@
                         }
                         // CraftBukkit end
                         iterator.remove();
+                        // Kiterino start - Expand custom mob effects API
+                        if (mobEffectInstance.getEffect().value() instanceof me.sosedik.kiterino.world.effect.KiterinoMobEffectWrapper wrapper) {
+                            wrapper.getWrapper().onEffectExpired(getBukkitLivingEntity(), mobEffectInstance.getAmplifier());
+                        }
+                        // Kiterino end - Expand custom mob effects API
                         this.onEffectsRemoved(List.of(mobEffectInstance));
                     } else if (mobEffectInstance.getDuration() % 600 == 0) {
                         this.onEffectUpdated(mobEffectInstance, false, null);
                     }
+                    // Kiterino start - Reset fire ticks when having fire resistance
+                    if (level().kiterinoConfig.fireResistanceResetsFireTicks && mobEffectInstance.getEffect() == MobEffects.FIRE_RESISTANCE && getRemainingFireTicks() > 0) {
+                        setRemainingFireTicks(0);
+                    }
+                    // Kiterino end - Reset fire ticks when having fire resistance
                 }
             } catch (ConcurrentModificationException var6) {
             }
@@ -1340,6 +_,11 @@
                         serverPlayer.connection.send(new ClientboundRemoveMobEffectPacket(this.getId(), mobEffectInstance.getEffect()));
                     }
                 }
+                // Kiterino start - Expand custom mob effects API
+                if (mobEffectInstance.getEffect().value() instanceof me.sosedik.kiterino.world.effect.KiterinoMobEffectWrapper wrapper) {
+                    wrapper.getWrapper().onEffectRemoved(getBukkitLivingEntity(), mobEffectInstance.getAmplifier());
+                }
+                // Kiterino end - Expand custom mob effects API
             }
 
             this.refreshDirtyAttributes();
@@ -2693,6 +_,7 @@
                 );
                 ServerChunkCache chunkSource = ((ServerLevel)this.level()).getChunkSource();
                 if (updateSelf) {
+                    if (this instanceof ServerPlayer serverPlayer && new me.sosedik.kiterino.event.player.PlayerArmSwingFromServerEvent(serverPlayer.getBukkitEntity(), hand == InteractionHand.MAIN_HAND ? org.bukkit.inventory.EquipmentSlot.HAND : org.bukkit.inventory.EquipmentSlot.OFF_HAND).callEvent()) // Kiterino - Add PlayerArmSwingFromServerEvent
                     chunkSource.sendToTrackingPlayersAndSelf(this, clientboundAnimatePacket);
                 } else {
                     chunkSource.sendToTrackingPlayers(this, clientboundAnimatePacket);
@@ -3452,7 +_,7 @@
                 this.getCombatTracker().recheckStatus();
             }
 
-            if (this.isSleeping() && (!this.canInteractWithLevel() || !this.checkBedExists())) {
+            if ((!level().kiterinoConfig.sleepWithNoBed || !(this instanceof ServerPlayer)) && this.isSleeping() && (!this.canInteractWithLevel() || !this.checkBedExists())) { // Kiterino - Sleeping options
                 this.stopSleeping();
             }
         }
@@ -3873,9 +_,17 @@
                     flag = false;
                 }
 
-                if (flag) {
-                    if (getRider() == null || !this.isControllable()) // Purpur - ignore mobs which are uncontrollable or without rider - API for any mob to burn daylight
-                        this.igniteForSeconds(8.0F);
+                if (true || flag) { // Kiterino - Add EntitySunburnEvent
+                    if (getRider() == null || !this.isControllable() || getRider().isInvisible()) // Purpur - ignore mobs which are uncontrollable or without rider - API for any mob to burn daylight // Kiterino - Allow burning ridables with invisible rider
+                    // Kiterino start - Add EntitySunburnEvent
+                    {
+                        var event = new me.sosedik.kiterino.event.entity.EntitySunburnEvent(getBukkitLivingEntity(), 8F);
+                        event.setCancelled(!flag);
+                        if (event.callEvent()) {
+                            this.igniteForSeconds(event.getBurnDuration());
+                        }
+                    }
+                    // Kiterino end - Add EntitySunburnEvent
                 }
             }
         }
