--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -994,7 +_,7 @@
     }
 
     public boolean isSaddled() {
-        return this.hasValidEquippableItemForSlot(EquipmentSlot.SADDLE);
+        return this.hasValidEquippableItemForSlot(EquipmentSlot.SADDLE) || hasRider(); // Kiterino - Allow riders to control mobs without a saddle
     }
 
     public boolean isWearingBodyArmor() {
@@ -1668,6 +_,11 @@
 
         this.postPiercingAttack();
         if (target instanceof net.minecraft.server.level.ServerPlayer) this.ticksSinceLastInteraction = 0; // Purpur - Entity lifespan
+        // Kiterino start - Award stats to rider
+        if (me.sosedik.kiterino.KiterinoConfig.awardStatsToRider && wasHurt && getRider() instanceof net.minecraft.server.level.ServerPlayer riderPlayer) {
+            riderPlayer.awardStat(net.minecraft.stats.Stats.DAMAGE_DEALT, Math.round(dmg * 10.0F));
+        }
+        // Kiterino end - Award stats to rider
         return wasHurt;
     }
 
