From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 8 Jun 2024 17:05:42 +0300
Subject: [PATCH] Rework some deprecations


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 598353f5624d404dc7266ccf33a34e073040b26a..0da29365086a1a5da7676b1657e6b8353120cc7f 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2632,7 +2632,7 @@ public final class Bukkit {
      * @return the unsafe values instance
      * @see UnsafeValues
      */
-    @Deprecated(since = "1.7.2")
+    //@Deprecated(since = "1.7.2") // Kiterino - Rework some deprecations
     @NotNull
     public static UnsafeValues getUnsafe() {
         return server.getUnsafe();
diff --git a/src/main/java/org/bukkit/OfflinePlayer.java b/src/main/java/org/bukkit/OfflinePlayer.java
index c597c298795fb9893447bc822d941c1748dcb9c5..745c6b73f82949f7cd64d26b6771507784f1a598 100644
--- a/src/main/java/org/bukkit/OfflinePlayer.java
+++ b/src/main/java/org/bukkit/OfflinePlayer.java
@@ -233,9 +233,11 @@ public interface OfflinePlayer extends ServerOperator, AnimalTamer, Configuratio
      * UTC.
      *
      * @return Date of last log-in for this player, or 0
-     * @deprecated The API contract is ambiguous and the implementation may or may not return the correct value given this API ambiguity. It is instead recommended use {@link #getLastLogin()} or {@link #getLastSeen()} depending on your needs.
+     * @see #hasPlayedBefore()
+     * @see #getLastLogin()
+     * @see #getLastSeen()
      */
-    @Deprecated
+    //@Deprecated // Kiterino - Rework some deprecations
     public long getLastPlayed();
 
     /**
diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 451ba8f49357846bfda9adf042522eecfa437fe8..37528146e01219700a95e9ff53f29bdd9945d858 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -32,7 +32,7 @@ import org.jetbrains.annotations.Nullable;
  * may be poorly named, throw exceptions, have misleading parameters, or any
  * other bad programming practice.
  */
-@Deprecated(since = "1.7.2")
+//@Deprecated(since = "1.7.2") // Kiterino - Rework some deprecations
 public interface UnsafeValues {
     // Paper start
     net.kyori.adventure.text.flattener.ComponentFlattener componentFlattener();
diff --git a/src/main/java/org/bukkit/entity/Damageable.java b/src/main/java/org/bukkit/entity/Damageable.java
index 4b94f1ed620c57bebca7cf0a5a0d478a93952d36..32ce88d2d8ebc5c15c9a10bda06215d18a381294 100644
--- a/src/main/java/org/bukkit/entity/Damageable.java
+++ b/src/main/java/org/bukkit/entity/Damageable.java
@@ -96,9 +96,9 @@ public interface Damageable extends Entity {
      * Gets the maximum health this entity has.
      *
      * @return Maximum health
-     * @deprecated use {@link Attribute#MAX_HEALTH}.
+     * @see Attribute#MAX_HEALTH
      */
-    @Deprecated(since = "1.11")
+    //@Deprecated(since = "1.11") // Kiterino - Rework some deprecations
     double getMaxHealth();
 
     /**
@@ -111,15 +111,15 @@ public interface Damageable extends Entity {
      * {@link Wither}, etc...} will have their bar scaled accordingly.
      *
      * @param health amount of health to set the maximum to
-     * @deprecated use {@link Attribute#MAX_HEALTH}.
+     * @see Attribute#MAX_HEALTH
      */
-    @Deprecated(since = "1.11")
+    //@Deprecated(since = "1.11") // Kiterino - Rework some deprecations
     void setMaxHealth(double health);
 
     /**
      * Resets the max health to the original amount.
-     * @deprecated use {@link Attribute#MAX_HEALTH}.
+     * @see Attribute#MAX_HEALTH
      */
-    @Deprecated(since = "1.11")
+    //@Deprecated(since = "1.11") // Kiterino - Rework some deprecations
     void resetMaxHealth();
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 65196f51d2ec50be9513123aeb5a548f573ffdac..58cabc02de3d3c4d3b34b4f6b53ef06fc8bc3cfc 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3324,9 +3324,9 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * they wish.
      *
      * @return the player's locale
-     * @deprecated in favour of {@link #locale()}
+     * @see #locale()
      */
-    @Deprecated // Paper
+    //@Deprecated // Paper // Kiterino - Rework some deprecations
     public String getLocale();
 
     // Paper start
diff --git a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
index 759aaa9351027394d4393e8c3ef2847d82b9ff82..a19d5900e60453b7dd209f0e91216b52113a7773 100644
--- a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
@@ -235,11 +235,11 @@ public class PlayerInteractEvent extends PlayerEvent implements Cancellable {
      * <p>
      * All vector components are between 0.0 and 1.0 inclusive.
      *
-     * @deprecated misleading, use {@link #getInteractionPoint()}
+     * @see #getInteractionPoint()
      * @return the clicked position. May be null.
      */
     @Nullable
-    @Deprecated // Paper
+    //@Deprecated // Paper // Kiterino - Rework some deprecations
     public Vector getClickedPosition() {
         if (this.clickedPosistion == null) {
             return null;
