From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 8 Jun 2024 17:05:42 +0300
Subject: [PATCH] Rework some deprecations/javadocs


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 3267b864ccdf7f1435cd79eaea486658c5a7ee5b..1cb089a00e9ecd8e345177e342d821ad0478f584 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2643,7 +2643,7 @@ public final class Bukkit {
      * @return the unsafe values instance
      * @see UnsafeValues
      */
-    @Deprecated(since = "1.7.2")
+    // @Deprecated(since = "1.7.2") // Kiterino - Rework some deprecations
     @NotNull
     public static UnsafeValues getUnsafe() {
         return server.getUnsafe();
diff --git a/src/main/java/org/bukkit/OfflinePlayer.java b/src/main/java/org/bukkit/OfflinePlayer.java
index 31ce15fa8cc6b4c79f6be4cbb4be5d9e6a9d7536..8806f151ade9f5ab7633ec7e32f198530a59bc37 100644
--- a/src/main/java/org/bukkit/OfflinePlayer.java
+++ b/src/main/java/org/bukkit/OfflinePlayer.java
@@ -234,9 +234,11 @@ public interface OfflinePlayer extends ServerOperator, AnimalTamer, Configuratio
      * UTC.
      *
      * @return Date of last log-in for this player, or 0
-     * @deprecated The API contract is ambiguous and the implementation may or may not return the correct value given this API ambiguity. It is instead recommended use {@link #getLastLogin()} or {@link #getLastSeen()} depending on your needs.
+     * @see #hasPlayedBefore()
+     * @see #getLastLogin()
+     * @see #getLastSeen()
      */
-    @Deprecated
+    // @Deprecated // Kiterino - Rework some deprecations
     public long getLastPlayed();
 
     /**
diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index cff5d4681e6eaf0769112e05a83880109d6463ed..4237ebfe7d1fe12558668565b46e869687e31054 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -33,7 +33,7 @@ import java.util.Map;
  * may be poorly named, throw exceptions, have misleading parameters, or any
  * other bad programming practice.
  */
-@Deprecated(since = "1.7.2")
+// @Deprecated(since = "1.7.2") // Kiterino - Rework some deprecations
 public interface UnsafeValues {
     // Paper start
     net.kyori.adventure.text.flattener.ComponentFlattener componentFlattener();
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index e332fb2d6d5ddbe31356a139ae7ff86e8e0c7034..7078a4a58246f9656f4f790769c3aa75f9b4402a 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3389,9 +3389,9 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * they wish.
      *
      * @return the player's locale
-     * @deprecated in favour of {@link #locale()}
+     * @see #locale()
      */
-    @Deprecated // Paper
+    // @Deprecated // Paper // Kiterino - Rework some deprecations
     public String getLocale();
 
     // Paper start
diff --git a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
index b3c8ba52eb9cc707e626fc88d5cde34e739f5f1e..b170f76e8b457e32caf2996d42d7046c53e5b20e 100644
--- a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
@@ -199,11 +199,11 @@ public class PlayerInteractEvent extends PlayerEvent implements Cancellable {
      * <p>
      * All vector components are between 0.0 and 1.0 inclusive.
      *
-     * @deprecated misleading, use {@link #getInteractionPoint()}
+     * @see #getInteractionPoint()
      * @return the clicked position. May be {@code null}.
      */
     @Nullable
-    @Deprecated
+    // @Deprecated // Kiterino - Rework some deprecations
     public Vector getClickedPosition() {
         if (this.clickedPosition == null) {
             return null;
diff --git a/src/main/java/org/bukkit/inventory/ShapedRecipe.java b/src/main/java/org/bukkit/inventory/ShapedRecipe.java
index 56c2820aa18dcbe6583a177eff357db06d2af7c5..873fb03633470b4eda1373c566eaf05e70cfcbdb 100644
--- a/src/main/java/org/bukkit/inventory/ShapedRecipe.java
+++ b/src/main/java/org/bukkit/inventory/ShapedRecipe.java
@@ -191,7 +191,7 @@ public class ShapedRecipe extends CraftingRecipe {
      */
     @Deprecated // Paper
     @NotNull
-    public Map<Character, ItemStack> getIngredientMap() {
+    public Map<Character, @org.jetbrains.annotations.Nullable ItemStack> getIngredientMap() {
         HashMap<Character, ItemStack> result = new HashMap<>();
         for (Map.Entry<Character, RecipeChoice> ingredient : this.ingredients.entrySet()) {
             if (ingredient.getValue() == null) {
@@ -209,7 +209,7 @@ public class ShapedRecipe extends CraftingRecipe {
      * @return The mapping of character to ingredients.
      */
     @NotNull
-    public Map<Character, RecipeChoice> getChoiceMap() {
+    public Map<Character, @org.jetbrains.annotations.Nullable RecipeChoice> getChoiceMap() {
         Map<Character, RecipeChoice> result = new HashMap<>();
         for (Map.Entry<Character, RecipeChoice> ingredient : this.ingredients.entrySet()) {
             if (ingredient.getValue() == null) {
