From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Fri, 9 Jun 2023 22:49:45 -0400
Subject: [PATCH] Paper PR - Player standing on position API

https://github.com/PaperMC/Paper/pull/9298

diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 7cd517899b880d90afc44cb536e499caed779915..c3b9494cb713f79d73dc6476a6e9d43863ba02bc 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1253,6 +1253,33 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     void broadcastHurtAnimation(@NotNull java.util.Collection<Player> players);
     // Paper end - broadcast hurt animation
 
+    // Paper start - Player standing on position API
+    /**
+     * Gets the block that is currently being used to calculate movement effects (i.e. friction) on the entity.
+     * <p>
+     * Examples:
+     * - When standing on a slab under an ice block, this will return the position of the ice block.
+     * - When flying on top of a flower placed on a grass block, this will return the position of the grass block (although this technically will not affect your movement).
+     *
+     * @return block used for movement effects
+     */
+    @NotNull
+    org.bukkit.block.Block getMovementAffectingBlock();
+
+    /**
+     * Gets the block that is currently supporting the entity.
+     * This takes into account the collisions of blocks under the entity.
+     * <p>
+     * Examples:
+     * - When standing on a slab under an ice block, this will return the position of the slab.
+     * - When flying on top of a flower placed on a grass block, this will return null, as no block is supporting the entity.
+     *
+     * @return block currently supporting the entity, or null if no block is currently supporting the entity
+     */
+    @Nullable
+    org.bukkit.block.Block getSupportingBlock();
+    // Paper end
+
     // Purpur start - Ridables
     /**
      * Get the riding player
