From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 18 Nov 2025 17:03:43 +0200
Subject: [PATCH] Custom fuel rules


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 1cb089a00e9ecd8e345177e342d821ad0478f584..73680d8b2e046c3a8c1e7d21ffd1b676d99b9e31 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3050,6 +3050,17 @@ public final class Bukkit {
     }
     // Purpur end - Added the ability to add combustible items
 
+    // Kiterino start - Custom fuel rules
+    /**
+     * Add an extra fuel rule
+     *
+     * @param rule burning rule
+     */
+    public static void addFuel(@NotNull java.util.function.Function<ItemStack, Integer> rule) {
+        server.addFuel(rule);
+    }
+    // Kiterino end - Custom fuel rules
+
     // Purpur start - Debug Marker API
     /**
      * Creates debug block highlight on specified block location and show it to all players on the server.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 81428ad31c7726eb9f91ddc8d44fb2f34f3df729..656fab7ad2cdeb3f511277919a661790fbfc7855 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2834,6 +2834,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     public void removeFuel(@NotNull Material material);
     // Purpur end - Added the ability to add combustible items
 
+    // Kiterino start - Custom fuel rules
+    /**
+     * Add an extra fuel rule
+     *
+     * @param rule burning rule
+     */
+    void addFuel(@NotNull java.util.function.Function<ItemStack, Integer> rule);
+    // Kiterino end - Custom fuel rules
+
     // Purpur start - Debug Marker API
     /**
      * Creates debug block highlight on specified block location and show it to all players on the server.
