From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 18 Nov 2025 17:03:43 +0200
Subject: [PATCH] Custom fuel rules


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index e7285a912117505332425ffe98e170929b283926..bbcfe03ca481478819ba953790d2c8d0197d29b8 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3054,6 +3054,17 @@ public final class Bukkit {
     }
     // Purpur end - Added the ability to add combustible items
 
+    // Kiterino start - Custom fuel rules
+    /**
+     * Add an extra fuel rule
+     *
+     * @param rule burning rule
+     */
+    public static void addFuel(@NotNull java.util.function.Function<ItemStack, Integer> rule) {
+        server.addFuel(rule);
+    }
+    // Kiterino end - Custom fuel rules
+
     // Purpur start - Debug Marker API
     /**
      * Creates debug block highlight on specified block location and show it to all players on the server.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 69ac7e96c1a2f796fa7326a40d10c5ba46d6ab77..33c2d74eb8a7b9c63343a1b9a4e1b1a1f58662f7 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2830,6 +2830,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     public void removeFuel(@NotNull Material material);
     // Purpur end - Added the ability to add combustible items
 
+    // Kiterino start - Custom fuel rules
+    /**
+     * Add an extra fuel rule
+     *
+     * @param rule burning rule
+     */
+    void addFuel(@NotNull java.util.function.Function<ItemStack, Integer> rule);
+    // Kiterino end - Custom fuel rules
+
     // Purpur start - Debug Marker API
     /**
      * Creates debug block highlight on specified block location and show it to all players on the server.
