From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 12 Feb 2026 15:12:25 +0200
Subject: [PATCH] Injecting custom items extras


diff --git a/src/main/java/org/bukkit/Registry.java b/src/main/java/org/bukkit/Registry.java
index bdf1957e5b72b3d680c7756aadc9dde063b4b167..edad99142cb568b87c8675f6817138d8330f2e5f 100644
--- a/src/main/java/org/bukkit/Registry.java
+++ b/src/main/java/org/bukkit/Registry.java
@@ -608,7 +608,7 @@ public interface Registry<T extends Keyed> extends Iterable<T> {
                 }
             }
 
-            this.map = builder.build();
+            this.map = new java.util.HashMap<>(builder.build()); // Kiterino - Use modifiable HashMap - Injecting custom items
             this.type = type;
         }
 
