--- a/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -127,8 +_,8 @@
 
     protected AbstractHorse(EntityType<? extends AbstractHorse> entityType, Level level) {
         super(entityType, level);
-        this.moveControl = new net.minecraft.world.entity.ai.control.MoveControl(this); // Purpur - use vanilla controller
-        this.lookControl = new net.minecraft.world.entity.ai.control.LookControl(this); // Purpur - use vanilla controller
+        // this.moveControl = new net.minecraft.world.entity.ai.control.MoveControl(this); // Purpur - use vanilla controller // Kiterino - Allow riders to control horses without a saddle
+        // this.lookControl = new net.minecraft.world.entity.ai.control.LookControl(this); // Purpur - use vanilla controller // Kiterino - Allow riders to control horses without a saddle
         this.createInventory();
     }
 
@@ -464,7 +_,7 @@
 
     @Override
     public void openCustomInventoryScreen(Player player) {
-        if (!this.level().isClientSide() && (!this.isVehicle() || this.hasPassenger(player)) && this.isTamed()) {
+        if (!this.level().isClientSide() && (!this.isVehicle() || this.hasPassenger(player)) && (this.isTamed() || getRider() == player)) { // Kiterino - Allow riders to control horses without a saddle
             player.openHorseInventory(this, this.inventory);
         }
     }
