--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -182,7 +_,7 @@
     }
 
     public static boolean isUsableForCrafting(ItemStack stack) {
-        return !stack.isDamaged() && !stack.isEnchanted() && !stack.has(DataComponents.CUSTOM_NAME);
+        return (me.sosedik.kiterino.KiterinoConfig.allowDamagedItemsInRecipeBook || !stack.isDamaged()) && (me.sosedik.kiterino.KiterinoConfig.allowEnchantedItemsInRecipeBook || !stack.isEnchanted()) && (me.sosedik.kiterino.KiterinoConfig.allowRenamedItemsInRecipeBook || !stack.has(DataComponents.CUSTOM_NAME)); // Kiterino - Less limited recipe matcher
     }
 
     public int findSlotMatchingCraftingIngredient(io.papermc.paper.inventory.recipe.ItemOrExact item, ItemStack stack) { // Paper - Improve exact choice recipe ingredients
