--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -328,7 +_,7 @@
                 this.sleepCounter = 100;
             }
 
-            if (!this.level().isClientSide && this.level().isBrightOutside()) {
+            if (!this.level().kiterinoConfig.noAutoWakeUp && !this.level().isClientSide && this.level().isBrightOutside()) { // Kiterino - Sleeping options
                 this.stopSleepInBed(false, true);
             }
         } else if (this.sleepCounter > 0) {
