--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -294,7 +_,7 @@
                 this.sleepCounter = 100;
             }
 
-            if (!this.level().isClientSide()
+            if (!this.level().kiterinoConfig.noAutoWakeUp && !this.level().isClientSide() // Kiterino - Sleeping options
                 && !this.level().environmentAttributes().getValue(EnvironmentAttributes.BED_RULE, this.position()).canSleep(this.level())) {
                 this.stopSleepInBed(false, true);
             }
