--- a/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -183,6 +_,14 @@
     ) {
         MutableFloat result = new MutableFloat(damage);
         runIterationOnItem(itemStack, (enchantment, level) -> enchantment.value().modifyDamage(serverLevel, level, itemStack, victim, damageSource, result));
+        // Kiterino start - Apply Impaling on wet mobs
+        if (serverLevel.kiterinoConfig.applyImpalingOnWetMobs && victim instanceof LivingEntity livingEntity && !livingEntity.is(net.minecraft.tags.EntityTypeTags.SENSITIVE_TO_IMPALING) && livingEntity.isInWaterOrRain()) {
+            int impalingLevel = getItemEnchantmentLevel(Enchantments.IMPALING, itemStack);
+            if (impalingLevel > 0) {
+                result.add(impalingLevel * 2.5F);
+            }
+        }
+        // Kiterino end - Apply Impaling on wet mobs
         return result.floatValue();
     }
 
