--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -34,7 +_,7 @@
     public static final float TICKS_PER_FLAP = 10.0F;
     private static final EntityDataAccessor<Byte> DATA_ID_FLAGS = SynchedEntityData.defineId(Bat.class, EntityDataSerializers.BYTE);
     private static final int FLAG_RESTING = 1;
-    private static final TargetingConditions BAT_RESTING_TARGETING = TargetingConditions.forNonCombat().range(4.0);
+    private static final TargetingConditions BAT_RESTING_TARGETING = TargetingConditions.forNonCombat().range(4.0).selector((entity, level) -> !level.kiterinoConfig.batsIgnoreInvisiblePlayers || !entity.isInvisible()); // Kiterino - Bats ignore invisible players
     private static final byte DEFAULT_FLAGS = 0;
     public final AnimationState flyAnimationState = new AnimationState();
     public final AnimationState restAnimationState = new AnimationState();
@@ -306,6 +_,23 @@
     public static boolean checkBatSpawnRules(
         final EntityType<Bat> type, final LevelAccessor level, final EntitySpawnReason spawnReason, final BlockPos pos, final RandomSource random
     ) {
+        // Kiterino start - Extra bats on Halloween
+        if (level instanceof ServerLevel serverLevel && serverLevel.kiterinoConfig.extraBatsOnHalloween) {
+            if (pos.getY() >= level.getHeightmapPos(Heightmap.Types.WORLD_SURFACE, pos).getY()) {
+                return false;
+            }
+
+            int i = level.getMaxLocalRawBrightness(pos);
+            int j = 4;
+            if (serverLevel.purpurConfig.forceHalloweenSeason || net.minecraft.util.SpecialDates.isExtendedHalloween()) {
+                j = 7;
+            } else if (random.nextBoolean()) {
+                return false;
+            }
+
+            return i > random.nextInt(j) && level.getBlockState(pos.below()).is(BlockTags.BATS_SPAWNABLE_ON) && checkMobSpawnRules(type, level, spawnReason, pos, random);
+        }
+        // Kiterino end - Extra bats on Halloween
         return pos.getY() < level.getHeightmapPos(Heightmap.Types.WORLD_SURFACE, pos).getY()
             && !random.nextBoolean()
             && level.getMaxLocalRawBrightness(pos) <= random.nextInt(4)
