From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 31 Jul 2021 16:01:50 +0300
Subject: [PATCH] Kiterino Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a7e2fab3e174fe2c1d79d904f7eaff88dd0c3c82..a21d6c276a7a162bef937de5167c674de73f54ea 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1981,6 +1981,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Kiterino start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKiterinoConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kiterino end
+
         /**
          * Sends the component to the player
          *
