From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 31 Jul 2021 16:01:50 +0300
Subject: [PATCH] Kiterino Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d73b7d72d00fa0edd6542226348af14c5203cf8d..68a9c7add98f016934377d021b05b93ea8fe24ee 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1949,6 +1949,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Kiterino start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKiterinoConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kiterino end
+
         /**
          * Sends the component to the player
          *
