From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 31 Jul 2021 16:01:50 +0300
Subject: [PATCH] Kiterino Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 90f109af3c83a1bc12cb27497d107de5029d0ce6..1f773fd688e19a5714ed3fe8e5d3303008f33d4b 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2241,6 +2241,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Kiterino start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKiterinoConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kiterino end
+
         /**
          * Sends the component to the player
          *
