From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 31 Jul 2021 16:01:50 +0300
Subject: [PATCH] Kiterino Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index be0e4c789a02d67160287c39cf6bdf8ef2229afc..8bec56ee5df2861bde50e6a04211ee993be22161 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2226,6 +2226,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Kiterino start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKiterinoConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kiterino end
+
         /**
          * Sends the component to the player
          *
