From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 29 Nov 2022 01:15:55 +0100
Subject: [PATCH] Hide irrelevant compilation warnings


diff --git a/build.gradle.kts b/build.gradle.kts
index 6aef83558a5ef7e84873b127c3bb43a6468c9a24..a5a059f6fd2d3c4cf70bb5bba6a521a54b17c8d3 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -222,3 +222,12 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+
+// Kiterino start - Hide irrelevant compilation warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// Kiterino end - Hide irrelevant compilation warnings
