From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 16 Jun 2024 18:51:46 +0300
Subject: [PATCH] More Sound API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 175164f7b20ae31afa4c33819ab0c03c4a10dd73..f4db637dd524184dc0b5acfc02cff7343dd7e0de 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -828,4 +828,29 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
         return new de.tr7zw.changeme.nbtapi.NBTBlock(this).getData();
     }
     // Kiterino end - Add NBT-API as a first-class lib
+
+    // Kiterino start - More Sound API
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        emitSound(sound, org.bukkit.SoundCategory.BLOCKS, volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation().toCenterLocation(), sound, category, volume, pitch);
+    }
+    // Kiterino end - More Sound API
 }
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 71f479f5ce41de5a3da7e0fa16046b301ac66946..65751e7293b1bce076b44f0ce69bb33c158ff9d8 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1283,4 +1283,35 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
         de.tr7zw.changeme.nbtapi.NBT.modifyPersistentData(this, consumer);
     }
     // Kiterino end - Add NBT-API as a first-class lib
+
+    // Kiterino start - More Sound API
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param sound sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        org.bukkit.SoundCategory soundGroup = switch (this) {
+	        case HumanEntity humanEntity -> org.bukkit.SoundCategory.PLAYERS;
+	        case Ambient ambient -> org.bukkit.SoundCategory.AMBIENT;
+	        case Monster monster -> org.bukkit.SoundCategory.HOSTILE;
+	        default -> org.bukkit.SoundCategory.NEUTRAL;
+        };
+	    emitSound(sound, soundGroup, volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(this, sound, category, volume, pitch);
+    }
+    // Kiterino end - More Sound API
 }
