From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 16 Mar 2022 12:43:34 +0200
Subject: [PATCH] More Sound API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index a13c8ddd4a1222e7a16debb61769af3758502d7c..84a2e44fe8a6ff534230f7407b948e70d2dcb0cc 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -816,4 +816,29 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
         return this.getBlockData().getDestroySpeed(itemStack, considerEnchants);
     }
     // Paper end - destroy speed API
+
+    // Kiterino start - More Sound API
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        emitSound(sound, org.bukkit.SoundCategory.BLOCKS, volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation().toCenterLocation(), sound, category, volume, pitch);
+    }
+    // Kiterino end - More Sound API
 }
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 2d12c70b0baa50504619c8e37881a11a56d7df2c..bb1d64df05e9c0727344e35176c00c28aa3cc1a6 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1210,4 +1210,35 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end
+
+    // Kiterino start - More Sound API
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param sound sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        org.bukkit.SoundCategory soundGroup = switch (this) {
+            case HumanEntity humanEntity -> org.bukkit.SoundCategory.PLAYERS;
+            case Ambient ambient -> org.bukkit.SoundCategory.AMBIENT;
+            case Monster monster -> org.bukkit.SoundCategory.HOSTILE;
+            default -> org.bukkit.SoundCategory.NEUTRAL;
+        };
+        emitSound(sound, soundGroup, volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(this, sound, category, volume, pitch);
+    }
+    // Kiterino end - More Sound API
 }
