From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 18:47:01 -0400
Subject: [PATCH] Parchment - Expose NMS ordinals


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index de36dab37d3af249cf51573d163731c12346066a..ed3462a957a3fb6b4b573603df95aaa93521141b 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -252,4 +252,13 @@ public interface UnsafeValues {
 
     String getStatisticCriteriaKey(@NotNull org.bukkit.Statistic statistic);
     // Paper end
+
+    // Parchment start
+    /**
+     * Gets the ID/ordinal of an entity according to NMS.
+     * @param type entity to get
+     * @return entity's ID
+     */
+    int entityID(org.bukkit.entity.EntityType type);
+    // Parchment end
 }
