From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 18:47:01 -0400
Subject: [PATCH] Parchment - Expose NMS ordinals


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 9a65c4f614a6c358d74491794d7b25172a00bc11..8a1330e998729024cf57d1247b7d4e6d181ef283 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -300,4 +300,13 @@ public interface UnsafeValues {
     // Paper end - lifecycle event API
 
     @NotNull java.util.List<net.kyori.adventure.text.Component> computeTooltipLines(@NotNull ItemStack itemStack, @NotNull io.papermc.paper.inventory.tooltip.TooltipContext tooltipContext, @Nullable org.bukkit.entity.Player player); // Paper - expose itemstack tooltip lines
+
+    // Parchment start
+    /**
+     * Gets the ID/ordinal of an entity according to NMS.
+     * @param type entity to get
+     * @return entity's ID
+     */
+    int entityID(org.bukkit.entity.EntityType type);
+    // Parchment end
 }
