From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yannick Lamprecht <yannicklamprecht@live.de>
Date: Tue, 12 Oct 2021 21:44:34 +0200
Subject: [PATCH] Weeper - Implement take and place


diff --git a/src/main/java/com/github/weepingmc/inventory/InventoryItemStackMoveStrategy.java b/src/main/java/com/github/weepingmc/inventory/InventoryItemStackMoveStrategy.java
new file mode 100644
index 0000000000000000000000000000000000000000..49151b6c9b413f22c8f11a10f6a72639d2744eed
--- /dev/null
+++ b/src/main/java/com/github/weepingmc/inventory/InventoryItemStackMoveStrategy.java
@@ -0,0 +1,13 @@
+package com.github.weepingmc.inventory;
+
+import org.jetbrains.annotations.NotNull;
+
+public interface InventoryItemStackMoveStrategy {
+    default boolean canPlace(@NotNull SlotItemStackMovementStrategyItem slotItemStackMovementStrategyItem) {
+        return true;
+    }
+
+    default boolean canTake(@NotNull SlotItemStackMovementStrategyItem slotItemStackMovementStrategyItem) {
+        return true;
+    }
+}
diff --git a/src/main/java/com/github/weepingmc/inventory/SlotItemStackMovementStrategyItem.java b/src/main/java/com/github/weepingmc/inventory/SlotItemStackMovementStrategyItem.java
new file mode 100644
index 0000000000000000000000000000000000000000..be4156c71f9dd45d5fd97559952d819131cb823e
--- /dev/null
+++ b/src/main/java/com/github/weepingmc/inventory/SlotItemStackMovementStrategyItem.java
@@ -0,0 +1,11 @@
+package com.github.weepingmc.inventory;
+
+import org.bukkit.inventory.InventoryHolder;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public record SlotItemStackMovementStrategyItem(
+        @NotNull InventoryHolder inventoryHolder,
+        @NotNull ItemStack itemStack,
+        int slot
+) { }
diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 3b204144e2d245098b3dc23b8779f9ed817bb6d1..dd6db7819af5c4c8b3508cc63229035a727eccfe 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -91,6 +91,37 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Nullable
     public InventoryView openInventory(@NotNull Inventory inventory);
 
+    // Weeper start - inventoryItemStackMoveStrategy
+    /**
+     * Opens an inventory window with the specified inventory on the top and
+     * the player's inventory on the bottom.
+     *
+     * @param inventory The inventory to open
+     * @param mayPlaceStrategy place conditions
+     * @return The newly opened inventory view
+     */
+    @Nullable
+    default InventoryView openInventory(@NotNull Inventory inventory, @Nullable java.util.function.BiFunction<InventoryHolder, ItemStack, Boolean> mayPlaceStrategy){
+        return openInventory(inventory, new com.github.weepingmc.inventory.InventoryItemStackMoveStrategy() {
+            @Override
+            public boolean canPlace(@NotNull com.github.weepingmc.inventory.SlotItemStackMovementStrategyItem slotItemStackMovementStrategyItem) {
+                return mayPlaceStrategy != null ? mayPlaceStrategy.apply(slotItemStackMovementStrategyItem.inventoryHolder(), slotItemStackMovementStrategyItem.itemStack()) : true;
+            }
+        });
+    }
+
+    /**
+     * Opens an inventory window with the specified inventory on the top and
+     * the player's inventory on the bottom.
+     *
+     * @param inventory The inventory to open
+     * @param inventoryItemStackMoveStrategy place conditions
+     * @return The newly opened inventory view
+     */
+    @Nullable
+    InventoryView openInventory(@NotNull Inventory inventory, @Nullable com.github.weepingmc.inventory.InventoryItemStackMoveStrategy inventoryItemStackMoveStrategy);
+    // Weeper end - inventoryItemStackMoveStrategy
+
     /**
      * Opens an empty workbench inventory window with the player's inventory
      * on the bottom.
