From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 8 Jun 2024 17:05:42 +0300
Subject: [PATCH] Rework some deprecations


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index db453d04efb00baaeabb904a7bd1b99dd0a50735..fb190f104fa2bb714a12894c92e2c350d9e1932f 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2616,7 +2616,7 @@ public final class Bukkit {
      * @return the unsafe values instance
      * @see UnsafeValues
      */
-    @Deprecated(since = "1.7.2")
+    //@Deprecated(since = "1.7.2") // Kiterino - Rework some deprecations
     @NotNull
     public static UnsafeValues getUnsafe() {
         return server.getUnsafe();
diff --git a/src/main/java/org/bukkit/OfflinePlayer.java b/src/main/java/org/bukkit/OfflinePlayer.java
index 6d31b561d915180fcd473b317721064feed28f37..459997be6968568b99c04436d430939eb06aac5c 100644
--- a/src/main/java/org/bukkit/OfflinePlayer.java
+++ b/src/main/java/org/bukkit/OfflinePlayer.java
@@ -239,9 +239,11 @@ public interface OfflinePlayer extends ServerOperator, AnimalTamer, Configuratio
      * UTC.
      *
      * @return Date of last log-in for this player, or 0
-     * @deprecated The API contract is ambiguous and the implementation may or may not return the correct value given this API ambiguity. It is instead recommended use {@link #getLastLogin()} or {@link #getLastSeen()} depending on your needs.
+     * @see #hasPlayedBefore()
+     * @see #getLastLogin()
+     * @see #getLastSeen()
      */
-    @Deprecated
+    //@Deprecated // Kiterino - Rework some deprecations
     public long getLastPlayed();
 
     /**
diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index d0de7ce3c3a73863fada71d9e8d7b3ccee98ad8e..8942b93c2a3fb35a4d223710939a758da89b98fb 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -29,7 +29,7 @@ import org.jetbrains.annotations.Nullable;
  * may be poorly named, throw exceptions, have misleading parameters, or any
  * other bad programming practice.
  */
-@Deprecated(since = "1.7.2")
+//@Deprecated(since = "1.7.2") // Kiterino - Rework some deprecations
 public interface UnsafeValues {
     // Paper start
     net.kyori.adventure.text.flattener.ComponentFlattener componentFlattener();
diff --git a/src/main/java/org/bukkit/entity/Damageable.java b/src/main/java/org/bukkit/entity/Damageable.java
index 4b94f1ed620c57bebca7cf0a5a0d478a93952d36..32ce88d2d8ebc5c15c9a10bda06215d18a381294 100644
--- a/src/main/java/org/bukkit/entity/Damageable.java
+++ b/src/main/java/org/bukkit/entity/Damageable.java
@@ -96,9 +96,9 @@ public interface Damageable extends Entity {
      * Gets the maximum health this entity has.
      *
      * @return Maximum health
-     * @deprecated use {@link Attribute#MAX_HEALTH}.
+     * @see Attribute#MAX_HEALTH
      */
-    @Deprecated(since = "1.11")
+    //@Deprecated(since = "1.11") // Kiterino - Rework some deprecations
     double getMaxHealth();
 
     /**
@@ -111,15 +111,15 @@ public interface Damageable extends Entity {
      * {@link Wither}, etc...} will have their bar scaled accordingly.
      *
      * @param health amount of health to set the maximum to
-     * @deprecated use {@link Attribute#MAX_HEALTH}.
+     * @see Attribute#MAX_HEALTH
      */
-    @Deprecated(since = "1.11")
+    //@Deprecated(since = "1.11") // Kiterino - Rework some deprecations
     void setMaxHealth(double health);
 
     /**
      * Resets the max health to the original amount.
-     * @deprecated use {@link Attribute#MAX_HEALTH}.
+     * @see Attribute#MAX_HEALTH
      */
-    @Deprecated(since = "1.11")
+    //@Deprecated(since = "1.11") // Kiterino - Rework some deprecations
     void resetMaxHealth();
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 6f7f1fc3db0237021fa1bd0f11fe56b2d6d4f84a..10c9820dcc638b353848d28086dddc995e966f9c 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -470,14 +470,15 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * Returns true if the entity is supported by a block.
      *
      * This value is a state updated by the client after each movement.
-     *
-     * @return True if entity is on ground.
-     * @deprecated This value is controlled only by the client and is therefore
+     * <p>
+     * Note: This value is controlled only by the client and is therefore
      * unreliable and vulnerable to spoofing and/or desync depending on the
      * context/time which it is accessed
+     *
+     * @return True if entity is on ground.
      */
     @Override
-    @Deprecated(since = "1.16.1")
+    //@Deprecated(since = "1.16.1") // Kiterino - Rework some deprecations
     public boolean isOnGround();
 
     /**
@@ -3342,10 +3343,10 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * they wish.
      *
      * @return the player's locale
-     * @deprecated in favour of {@link #locale()}
+     * @see #locale()
      */
     @NotNull
-    @Deprecated // Paper
+    //@Deprecated // Paper // Kiterino - Rework some deprecations
     public String getLocale();
 
     // Paper start
diff --git a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
index 69c800d367d36c8730781b89e3530a75487d000a..c581a536bba362020c05d2cd0f356dc56d238950 100644
--- a/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerInteractEvent.java
@@ -235,11 +235,11 @@ public class PlayerInteractEvent extends PlayerEvent implements Cancellable {
      * <p>
      * All vector components are between 0.0 and 1.0 inclusive.
      *
-     * @deprecated misleading, use {@link #getInteractionPoint()}
+     * @see #getInteractionPoint()
      * @return the clicked position. May be null.
      */
     @Nullable
-    @Deprecated // Paper
+    //@Deprecated // Paper // Kiterino - Rework some deprecations
     public Vector getClickedPosition() {
         return clickedPosistion;
     }
diff --git a/src/main/java/org/bukkit/inventory/meta/MapMeta.java b/src/main/java/org/bukkit/inventory/meta/MapMeta.java
index 721164c26c5e219f854f65ab150a46670f4d33bd..fd8376b4748f5be7faf09f95a84ee3beb3d139e3 100644
--- a/src/main/java/org/bukkit/inventory/meta/MapMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/MapMeta.java
@@ -35,16 +35,17 @@ public interface MapMeta extends ItemMeta {
 
     /**
      * Sets the map ID. This is used to determine what map is displayed.
-     *
-     * @param id the map id to set
-     * @see #setMapView(org.bukkit.map.MapView)
-     * @deprecated These methods are poor API: They rely on the caller to pass
+     * <p>
+     * Note: These methods are poor API: They rely on the caller to pass
      * in an only an integer property, and have poorly defined implementation
      * behavior if that integer is not a valid map (the current implementation
      * for example will generate a new map with a different ID). The xxxMapView
      * family of methods should be used instead.
+     *
+     * @param id the map id to set
+     * @see #setMapView(org.bukkit.map.MapView)
      */
-    @Deprecated(since = "1.13.2")
+    //@Deprecated(since = "1.13.2") // Kiterino - Rework some deprecations
     void setMapId(int id);
 
     /**
