From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 24 May 2023 17:13:26 +0300
Subject: [PATCH] Forced panic mode API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 93145e96760d9e0a65fa1c0a210890bafc21d7b5..0c7f1481a0a69fc3d9823bf32582b4c5f488a595 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1241,4 +1241,20 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
         getWorld().playSound(this, sound, category, volume, pitch);
     }
     // Kiterino end - More Sound API
+
+    // Kiterino start - Forced panic mode API
+    /**
+     * Gets the amount of forced panic ticks left
+     *
+     * @return the amount of panic ticks
+     */
+    int getPanicTicks();
+
+    /**
+     * Sets the amount of forced panic ticks
+     *
+     * @param ticks ticks
+     */
+    void setPanicTicks(int ticks);
+    // Kiterino end - Forced panic mode API
 }
