From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 16 Jun 2024 21:13:08 +0300
Subject: [PATCH] Red tint API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 389005ed6fd4eb82ffad371142531975501e80a9..dcf02cda8444f7d2ed4ac64bedc02b399f034688 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -4031,4 +4031,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         sendDeathScreen(message);
     }
     // Purpur end
+
+    // Kiterino start - Red tint API
+    /**
+     * Abuses world border packets to apply red tint on screen
+     *
+     * @param percentage tint percentage
+     */
+    void sendRedTint(double percentage);
+    // Kiterino end - Red tint API
 }
