From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 17 Jun 2024 15:25:04 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 82ca25a345d52f007d47e4fa29a843c38dbc6ae9..76e1faf78a1f60534e3d607c5d78086455a4d80a 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1270,4 +1270,27 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     void setPanicTicks(int ticks);
     // Kiterino end - Forced panic mode API
+
+    // Kiterino start - Allow setting entity's rider
+    /**
+     * Set the riding player. The player has to be the first passenger.
+     * <br> Note: this will call {@link org.bukkit.event.entity.EntityMountEvent}.
+     *
+     * @param rider rider
+     * @return whether the rider was set
+     */
+    default boolean setRider(@Nullable Player rider) {
+        return setRider(rider, true);
+    }
+
+    /**
+     * Set the riding player and optionally dismount the previous one.
+     * <br> Note: this will call {@link org.bukkit.event.entity.EntityMountEvent}.
+     *
+     * @param rider rider
+     * @param dismount whether to dismount the current rider
+     * @return whether rider was set
+     */
+    boolean setRider(@Nullable Player rider, boolean dismount);
+    // Kiterino end - Allow setting entity's rider
 }
