From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 24 May 2023 17:13:26 +0300
Subject: [PATCH] Forced panic mode API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 624cbc6fe7500192d7c2ffad3b44d6aa6b3bca99..82ca25a345d52f007d47e4fa29a843c38dbc6ae9 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1254,4 +1254,20 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
         getWorld().playSound(this, sound, category, volume, pitch);
     }
     // Kiterino end - More Sound API
+
+    // Kiterino start - Forced panic mode API
+    /**
+     * Gets the amount of forced panic ticks left
+     *
+     * @return the amount of panic ticks
+     */
+    int getPanicTicks();
+
+    /**
+     * Sets the amount of forced panic ticks
+     *
+     * @param ticks ticks
+     */
+    void setPanicTicks(int ticks);
+    // Kiterino end - Forced panic mode API
 }
