From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 11 Dec 2021 12:18:47 -0500
Subject: [PATCH] Paper PR - Add Movement Direction API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index c730b7e091b897fac7faa2d756b90ea5f0cef44d..5b187e1045faa42e0397cb4032df5bbe1027df11 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1015,6 +1015,44 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @param directionZ The relative z position of the knockback source direction
      */
     void knockback(double strength, double directionX, double directionZ);
+
+    /**
+     * This number represents how much the entity is trying to move sideways.
+     * <p>
+     * This number is ranges from -1 to 1,
+     * where positive 1 represents the left and -1 represents the right.
+     * <p>
+     * Note that for {@link Player} entities only update this value when riding an entity, which may cause it
+     * to be inaccurate when dismounted.
+     *
+     * @return direction repesented as -1 to 1
+     */
+    float getSidewaysMotion();
+
+    /**
+     * This number represents how much the entity is trying to move
+     * in the upwards direction.
+     * <p>
+     * This number is ranges from -1 to 1,
+     * where positive 1 represents the up and -1 represents the down.
+     * <p>
+     * Note that for {@link Player} entities this value is never updated.
+     *
+     * @return direction repesented as -1 to 1
+     */
+    float getUpwardsMotion();
+
+    /**
+     * This number represents how much the entity is trying to move forward.
+     * <p>
+     * This number is ranges from -1 to 1,
+     * where positive 1 represents the up and -1 represents the down.
+     * <p>
+     * Note that for {@link Player} entities only update this value when riding an entity, so it may be inaccurate when dismounted.
+     *
+     * @return direction repesented as -1 to 1
+     */
+    float getForwardMotion();
     // Paper end
 
     // Purpur start
