From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sun, 17 Oct 2021 14:53:28 -0400
Subject: [PATCH] Paper PR - BoneMeal API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 565ce4b0c5eb300953aaefa786b9f8938c771f2a..6c0fe85588d223d92b6ec15afaf35db92f08f0a3 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4447,4 +4447,19 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
             }
         }
     }
+    // Paper start
+    /**
+     * Applies bonemeal at a specific location
+     * <p>
+     * Modifications done in the predicate are respected
+     * and will not be applied if false is returned.
+     *
+     * @param location Location to apply
+     * @param face Blockface to apply as
+     * @param showParticles If particles should be shown on success
+     * @param predicate blockstate predicate
+     * @return true if the bonemeal was applied, false if not
+     */
+    boolean applyBoneMeal(@NotNull Location location, @NotNull org.bukkit.block.BlockFace face, boolean showParticles, @Nullable Predicate<org.bukkit.block.BlockState> predicate);
+    // Paper end
 }
