From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NeumimTo <neumimto@users.noreply.github.com>
Date: Tue, 13 Jul 2021 15:31:43 +0300
Subject: [PATCH] Paper PR - Damage with any cause


diff --git a/src/main/java/org/bukkit/entity/Damageable.java b/src/main/java/org/bukkit/entity/Damageable.java
index 7b9394fcbfe608a6d05b89763bb0770f864752f3..64d5727504b0250ee3e4863c2b0e27f1e4aeccd3 100644
--- a/src/main/java/org/bukkit/entity/Damageable.java
+++ b/src/main/java/org/bukkit/entity/Damageable.java
@@ -1,6 +1,8 @@
 package org.bukkit.entity;
 
 import org.bukkit.attribute.Attribute;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 /**
@@ -23,6 +25,35 @@ public interface Damageable extends Entity {
      */
     void damage(double amount, @Nullable Entity source);
 
+    // Paper start
+    /**
+     * Deals the given amount of damage to this entity with specific {@code DamageCause}.
+     *
+     * Damage types {@code DamageCause.ENTITY_ATTACK}, {@code DamageCause.ENTITY_SWEEP_ATTACK},
+     * {@code DamageCause.PROJECTILE}, {@code DamageCause.THORNS} require to specify an entity,
+     * which triggered the damage, for such cases use {@link Damageable#damage(double, EntityDamageEvent.DamageCause, Entity)},
+     * otherwise they will fallback to Generic damage type.
+     *
+     * @param amount Amount of damage to deal
+     * @param damageCause Type of damage
+     */
+    default void damage(double amount, @NotNull EntityDamageEvent.DamageCause damageCause) {
+        damage(amount, damageCause, null);
+    }
+
+    /**
+     * Deals the given amount of damage to this entity, with specific {@code DamageCause}, from a specified
+     * entity.
+     *
+     * {@code DamageCause.PROJECTILE} needs any {@code Projectile} instance to be passed as the source
+     *
+     * @param amount Amount of damage to deal
+     * @param damageCause Type of damage
+     * @param source The entity who caused this damage
+     */
+    void damage(double amount, @NotNull EntityDamageEvent.DamageCause damageCause, @Nullable Entity source);
+    // Paper end
+
     /**
      * Gets the entity's health from 0 to {@link #getMaxHealth()}, where 0 is dead.
      *
