From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 14 May 2024 04:23:20 +0300
Subject: [PATCH] Injecting custom Materials

Injects custom Material enums for data-driven items

diff --git a/src/main/java/me/sosedik/kiterino/util/IKiterinoBootstrapMaterialInjector.java b/src/main/java/me/sosedik/kiterino/util/IKiterinoBootstrapMaterialInjector.java
new file mode 100644
index 0000000000000000000000000000000000000000..4fadea41afa71dd8deca52a33e40797f69be6607
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/IKiterinoBootstrapMaterialInjector.java
@@ -0,0 +1,15 @@
+package me.sosedik.kiterino.util;
+
+import net.kyori.adventure.key.Key;
+import org.bukkit.Material;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@ApiStatus.Internal
+interface IKiterinoBootstrapMaterialInjector {
+
+    void injectMaterials(@NotNull Class<?> materialsClass);
+
+    @NotNull Material allocateMaterial(@NotNull Key key);
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjector.java b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjector.java
new file mode 100644
index 0000000000000000000000000000000000000000..d506e6bb0b2c88ffb4e9274672bf8ee4dc339db1
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjector.java
@@ -0,0 +1,50 @@
+package me.sosedik.kiterino.util;
+
+import net.kyori.adventure.key.Key;
+import org.bukkit.Material;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Used for injecting custom Materials.
+ *
+ * <p><b>Can be used only during bootstrap phase.
+ */
+public class KiterinoBootstrapMaterialInjector {
+
+    private KiterinoBootstrapMaterialInjector() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    static IKiterinoBootstrapMaterialInjector injector;
+
+    /**
+     * Inject materials in the class using the plugin's name as namespace.
+     * Keys will be fetched from Material field names.
+     *
+     * <p><b>Can be used only during bootstrap phase.
+     *
+     * @param materialsClass class with public static final Material instances
+     * @throws IllegalStateException if not in bootstrap phase
+     */
+    public static void injectMaterials(@NotNull Class<?> materialsClass) {
+        if (injector == null) throw new IllegalStateException("Can't inject Materials outside bootstrap phase!");
+        injector.injectMaterials(materialsClass);
+    }
+
+    /**
+     * Injects a custom material.
+     * If the key is null, it will be calculated from Plugin's provided namespace
+     * and Material field's name as a key.
+     *
+     * <p><b>Can be used only during bootstrap phase.
+     *
+     * @param key material key
+     * @return injected Material
+     * @throws IllegalStateException if not in bootstrap phase
+     */
+    public static @NotNull Material injectMaterial(@NotNull Key key) {
+        if (injector == null) throw new IllegalStateException("Can't inject Materials outside bootstrap phase!");
+        return injector.allocateMaterial(key);
+    }
+
+}
diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 3ec2154956087a370799f0fa485f3b6cae894c69..e7b44dca183a461ebbcc2a259ad1b961e40bdf69 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4627,6 +4627,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     private final NamespacedKey key;
     private final Supplier<ItemType> itemType;
     private final Supplier<BlockType> blockType;
+    private final boolean injected = false; // Kiterino - Injecting custom Materials
 
     private Material(final int id) {
         this(id, 64);
@@ -4798,6 +4799,12 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
      * @return Maximum stack size for this material
      */
     public int getMaxStackSize() {
+        // Kiterino start - Injecting custom Materials
+        ItemType type = asItemType();
+        if (type != null) {
+            return type.getMaxStackSize();
+        }
+        // Kiterino end - Injecting custom Materials
         return maxStack;
     }
 
@@ -4807,6 +4814,12 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
      * @return Maximum durability for this material
      */
     public short getMaxDurability() {
+        // Kiterino start - Injecting custom Materials
+        ItemType type = asItemType();
+        if (type != null) {
+            return type.getMaxDurability();
+        }
+        // Kiterino end - Injecting custom Materials
         return durability;
     }
 
@@ -4985,6 +4998,11 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         if (filtered.startsWith(NamespacedKey.MINECRAFT + ":")) {
             filtered = filtered.substring((NamespacedKey.MINECRAFT + ":").length());
         }
+        // Kiterino start - Injecting custom Materials
+        else {
+            filtered = filtered.replace(':', '_');
+        }
+        // Kiterino end - Injecting custom Materials
 
         filtered = filtered.toUpperCase(Locale.ROOT);
 
@@ -5672,4 +5690,16 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         }
     }
     // Purpur end
+
+    // Kiterino start - Injecting custom Materials
+    /**
+     * Checks whether this item is injected
+     * (i.e. non-vanilla one)
+     *
+     * @return whether this item is injected
+     */
+    public boolean isInjected() {
+        return this.injected;
+    }
+    // Kiterino end - Injecting custom Materials
 }
