From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 11 Feb 2022 00:30:55 +0100
Subject: [PATCH] Improve GameMode enum


diff --git a/src/main/java/org/bukkit/GameMode.java b/src/main/java/org/bukkit/GameMode.java
index fdc42a79c5af30fdade41ee99245e6641f353571..9412b7a961cc98435b0c16f64a5311fda8aa68e6 100644
--- a/src/main/java/org/bukkit/GameMode.java
+++ b/src/main/java/org/bukkit/GameMode.java
@@ -48,6 +48,18 @@ public enum GameMode implements net.kyori.adventure.translation.Translatable { /
         this.value = value;
         this.translationKey = "gameMode." +  this.name().toLowerCase(java.util.Locale.ENGLISH); // Paper
     }
+    
+    // Kiterino start
+    /**
+     * Checks whether this game mode is
+     * considered "survivable"
+     *
+     * @return true, if survival or adventure
+     */
+    public boolean isSurvival() {
+        return this == SURVIVAL || this == ADVENTURE;
+    }
+    // Kiterino end
 
     /**
      * Gets the mode value associated with this GameMode
