From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 30 Jan 2021 18:55:39 +0200
Subject: [PATCH] Allow modifying vanilla block drops


diff --git a/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java b/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java
index 4c5ee91de162b202c2db8bf68259ad41a430125d..bc79866b3b96a3f754ed0d2dbc16f1d5abeefae8 100644
--- a/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java
+++ b/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java
@@ -35,6 +35,7 @@ public class BlockDropItemEvent extends BlockEvent implements Cancellable {
     private boolean cancel;
     private final BlockState blockState;
     private final List<Item> items;
+    private List<org.bukkit.inventory.ItemStack> drops; // Kiterino
 
     public BlockDropItemEvent(@NotNull Block block, @NotNull BlockState blockState, @NotNull Player player, @NotNull List<Item> items) {
         super(block);
@@ -77,6 +78,54 @@ public class BlockDropItemEvent extends BlockEvent implements Cancellable {
         return items;
     }
 
+    // Kiterino start
+    private void vanillaDrops() {
+        drops = new java.util.ArrayList<>();
+        items.forEach(item -> drops.add(item.getItemStack()));
+    }
+
+    /**
+     * Whether vanilla drops were overwritten or not
+     *
+     * @return True, if drops were changed
+     */
+    public boolean changeDrops() {
+        return drops != null;
+    }
+
+    /**
+     * Makes block to not drop any items
+     */
+    public void clearDrops() {
+        if (drops == null) drops = new java.util.ArrayList<>();
+        else drops.clear();
+    }
+
+    /**
+     * Adds items to be dropped
+     *
+     * @param drops Items to drop
+     */
+    public void addDrops(@NotNull org.bukkit.inventory.ItemStack... drops) {
+        if (this.drops == null) vanillaDrops();
+        this.drops.addAll(List.of(drops));
+    }
+
+    /**
+     * Gets list of the ItemStack drops caused by the block break.
+     * <p>
+     * This list is mutable - removing an item from it will cause it to not
+     * drop. Adding to the list is allowed.
+     *
+     * @return The Item the block caused to drop
+     */
+    @NotNull
+    public List<org.bukkit.inventory.ItemStack> getDrops() {
+        if (drops == null) vanillaDrops();
+        return drops;
+    }
+    // Kiterino end
+
     @Override
     public boolean isCancelled() {
         return cancel;
