From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 15 Jun 2024 16:14:19 +0300
Subject: [PATCH] Temporary exclude UnsafeValues from annotation test


diff --git a/src/test/java/org/bukkit/AnnotationTest.java b/src/test/java/org/bukkit/AnnotationTest.java
index 6369124775f8b51b10476a6819969fd075cb61e1..75cfb817c4523640d6173a6e57daab8fe023fd7c 100644
--- a/src/test/java/org/bukkit/AnnotationTest.java
+++ b/src/test/java/org/bukkit/AnnotationTest.java
@@ -39,6 +39,7 @@ public class AnnotationTest {
     };
 
     private static final String[] EXCLUDED_CLASSES = {
+        "org/bukkit/UnsafeValues", // Kiterino - Temporary exclude UnsafeValues from annotation test
         // Internal technical classes
         "org/bukkit/plugin/java/JavaPluginLoader",
         "org/bukkit/util/io/BukkitObjectInputStream",
