From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 12 Jul 2021 19:26:22 +0300
Subject: [PATCH] Add ability to get active container id


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index a3c59027f2a90eade112afc8dd24fbec6b944967..3ac7f0a134b706b9b50b7ff76c658968d692c5fe 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -836,4 +836,19 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Nullable
     public Firework fireworkBoost(@NotNull ItemStack fireworkItemStack);
 
+    // Kiterino start
+    /**
+     * Get current active container id
+     *
+     * @return current active container id
+     */
+    int getActiveContainerId();
+
+    /**
+     * Get current active container state id
+     *
+     * @return current active container state id
+     */
+    int getActiveContainerStateId();
+    // Kiterino start
 }
