From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 19 Jun 2024 01:40:24 +0300
Subject: [PATCH] Temporary solution for respawn events parity until Paper PR
 8705 is merged


diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerPostRespawnEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerPostRespawnEvent.java
index 16961aac061e78fb84029f8435ab5f7c493b1362..c8e081b49f3f0f5e8ce21f01eaa8c78149af9798 100644
--- a/src/main/java/com/destroystokyo/paper/event/player/PlayerPostRespawnEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerPostRespawnEvent.java
@@ -16,13 +16,34 @@ public class PlayerPostRespawnEvent extends PlayerEvent {
 
     private final Location respawnedLocation;
     private final boolean isBedSpawn;
+    // Kiterino start - Parity for respawn events
+    private final boolean isAnchorSpawn;
+    private final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason;
+    private final java.util.Set<org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag> respawnFlags;
+    // Kiterino end - Parity for respawn events
 
     @ApiStatus.Internal
+    @Deprecated // Kiterino - Parity for respawn events
     public PlayerPostRespawnEvent(@NotNull final Player respawnPlayer, @NotNull final Location respawnedLocation, final boolean isBedSpawn) {
+        // Kiterino start - Parity for respawn events
+        this(respawnPlayer, respawnedLocation, isBedSpawn, false, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.PLUGIN, com.google.common.collect.ImmutableSet.builder());
+    }
+
+    @ApiStatus.Internal
+    public PlayerPostRespawnEvent(@NotNull final Player respawnPlayer, @NotNull final Location respawnedLocation, final boolean isBedSpawn, final boolean isAnchorSpawn, @NotNull final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason, @NotNull final com.google.common.collect.ImmutableSet.Builder<org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag> respawnFlags) {
+        // Paper end
         super(respawnPlayer);
         this.respawnedLocation = respawnedLocation;
         this.isBedSpawn = isBedSpawn;
+        this.isAnchorSpawn = isAnchorSpawn;
+        this.respawnReason = respawnReason;
+        // Paper start
+        if (this.isBedSpawn) { respawnFlags.add(org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag.BED_SPAWN); }
+        if (this.isAnchorSpawn) { respawnFlags.add(org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag.ANCHOR_SPAWN); }
+        this.respawnFlags = respawnFlags.build();
+        // Paper end
     }
+    // Kiterino end - Parity for respawn events
 
     /**
      * Returns the location of the respawned player
@@ -43,6 +64,37 @@ public class PlayerPostRespawnEvent extends PlayerEvent {
         return this.isBedSpawn;
     }
 
+    // Kiterino start - Parity for respawn events
+    /**
+     * Gets whether the respawn location is the player's respawn anchor.
+     *
+     * @return true if the respawn location is the player's respawn anchor.
+     */
+    public boolean isAnchorSpawn() {
+        return isAnchorSpawn;
+    }
+
+    /**
+     * Gets the reason this respawn event was called.
+     *
+     * @return the reason the event was called.
+     */
+    @NotNull
+    public org.bukkit.event.player.PlayerRespawnEvent.RespawnReason getRespawnReason() {
+        return respawnReason;
+    }
+
+    /**
+     * Get the set of flags that apply to this respawn.
+     *
+     * @return an immutable set of the flags that apply to this respawn
+     */
+    @NotNull
+    public java.util.Set<org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag> getRespawnFlags() {
+        return respawnFlags;
+    }
+    // Kiterino end - Parity for respawn events
+
     @Override
     @NotNull
     public HandlerList getHandlers() {
