From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 12 Jul 2021 19:26:22 +0300
Subject: [PATCH] Add ability to get active container id


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 15fdde01fd9d4318ff229e0226c1c5baeed5b6ca..b66b11939716f676b1deb7a16d629d7f3b18ef06 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -826,4 +826,19 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Nullable
     public Firework fireworkBoost(@NotNull ItemStack fireworkItemStack);
 
+    // Kiterino start
+    /**
+     * Get current active container id
+     *
+     * @return current active container id
+     */
+    int getActiveContainerId();
+
+    /**
+     * Get current active container state id
+     *
+     * @return current active container state id
+     */
+    int getActiveContainerStateId();
+    // Kiterino start
 }
