From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 14:32:56 +0300
Subject: [PATCH] Allow force sleeping


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index b2bf94e981a83ff28c351e99941ce29e3cc8ea62..0d8280a455a9ed7cc3cdc435975513e3d9f75373 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3061,4 +3061,11 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     void sendDeathScreen(@NotNull Component message, @Nullable Entity killer);
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Forces player to sleep
+     */
+    void forceSleep();
+    // Kiterino end
 }
