From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 21 Feb 2022 11:44:15 +0200
Subject: [PATCH] Add extra MaterialTags


diff --git a/src/main/java/com/destroystokyo/paper/MaterialTags.java b/src/main/java/com/destroystokyo/paper/MaterialTags.java
index f59151cec84fc0e300a8d759c6d74a804cbd047f..4f988ef060bef2c868d9ef1a3786adc47c17bc8a 100644
--- a/src/main/java/com/destroystokyo/paper/MaterialTags.java
+++ b/src/main/java/com/destroystokyo/paper/MaterialTags.java
@@ -705,4 +705,21 @@ public class MaterialTags {
         .add(Material.NETHERITE_AXE, Material.NETHERITE_HOE, Material.NETHERITE_PICKAXE, Material.NETHERITE_SHOVEL, Material.NETHERITE_SWORD)
         .ensureSize("NETHERITE_TOOLS", 5).lock();
 
+    // Kiterino start
+    /**
+     * Covers the variants of items that can be equipped in any of the armor slots.
+     */
+    public static final MaterialSetTag ALL_EQUIPPABLE = new MaterialSetTag(keyFor("all_equippable"))
+        .add(HEAD_EQUIPPABLE)
+        .add(CHEST_EQUIPPABLE)
+        .add(LEGGINGS)
+        .add(BOOTS)
+        .lock();
+
+    /**
+     * Covers the minecarts
+     */
+    public static final MaterialSetTag MINECARTS = new MaterialSetTag(keyFor("minecarts"))
+        .endsWith("MINECART").ensureSize("MINECARTS", 6).lock();
+    // Kiterino end
 }
