From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 15 Jun 2024 16:14:19 +0300
Subject: [PATCH] Temporary exclude UnsafeValues from annotation test


diff --git a/src/test/java/org/bukkit/AnnotationTest.java b/src/test/java/org/bukkit/AnnotationTest.java
index b82f07a2879412f6b30643ca93a97439aa49a98a..952350c39c6afcffa6219b752aceac7c07504baa 100644
--- a/src/test/java/org/bukkit/AnnotationTest.java
+++ b/src/test/java/org/bukkit/AnnotationTest.java
@@ -39,6 +39,7 @@ public class AnnotationTest {
     };
 
     private static final String[] EXCLUDED_CLASSES = {
+        "org/bukkit/UnsafeValues", // Kiterino - Temporary exclude UnsafeValues from annotation test
         // Internal technical classes
         "org/bukkit/plugin/java/JavaPluginLoader",
         "org/bukkit/util/io/BukkitObjectInputStream",
