From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 12 Jul 2021 19:26:22 +0300
Subject: [PATCH] Add ability to get active container id


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index dd6db7819af5c4c8b3508cc63229035a727eccfe..8ef807d85b9ab3d213a384dd21d1dc2bfa8119c0 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -793,4 +793,20 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      */
     @Nullable
     public Firework fireworkBoost(@NotNull ItemStack fireworkItemStack);
+
+    // Kiterino start
+    /**
+     * Get current active container id
+     *
+     * @return current active container id
+     */
+    int getActiveContainerId();
+
+    /**
+     * Get current active container state id
+     *
+     * @return current active container state id
+     */
+    int getActiveContainerStateId();
+    // Kiterino start
 }
