From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 14:32:56 +0300
Subject: [PATCH] Allow force sleeping


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ee86564be4c6d85195317b5f4c3b0bf0c5a256be..f5e6974cfd3ad12a99189d314d7e124edeafbb56 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2990,4 +2990,11 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     void sendDeathScreen(@NotNull Component message, @Nullable Entity killer);
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Forces player to sleep
+     */
+    void forceSleep();
+    // Kiterino end
 }
