From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 20 Jun 2024 04:35:50 +0300
Subject: [PATCH] Expose canUseSlot


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 05123a4833b8908d8ceee3e72d42f2289e33999a..ab6c73387eb9b5952a07d28266352fc079a36221 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1463,4 +1463,14 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end - API for any mob to burn daylight
+
+    // Kiterino start - Expose canUseSlot
+    /**
+     * Checks whether this entity can use the equipment slot
+     *
+     * @param slot equipment slot
+     * @return whether this entity can use the equipment slot
+     */
+    boolean canUseSlot(org.bukkit.inventory.@NotNull EquipmentSlot slot);
+    // Kiterino end - Expose canUseSlot
 }
