From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 14:32:56 +0300
Subject: [PATCH] Allow force sleeping


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 7b2bb67896db9d498b8cf441b8dd14941e314781..54fdfe7de065e57451b4a7cd0bcbef16eab00dd8 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2820,4 +2820,11 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public boolean hasFlyingFallDamage();
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Forces player to sleep
+     */
+    void forceSleep();
+    // Kiterino end
 }
