From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 20 Jun 2024 20:02:03 +0300
Subject: [PATCH] Expose invalid world to PlayerSpawnLocationEvent


diff --git a/src/main/java/org/spigotmc/event/player/PlayerSpawnLocationEvent.java b/src/main/java/org/spigotmc/event/player/PlayerSpawnLocationEvent.java
index 2515887c20738b5add74eff02d2e9672080623d2..d6b92ba730b787676e3ead176c30f94106c7afe1 100644
--- a/src/main/java/org/spigotmc/event/player/PlayerSpawnLocationEvent.java
+++ b/src/main/java/org/spigotmc/event/player/PlayerSpawnLocationEvent.java
@@ -12,9 +12,21 @@ import org.jetbrains.annotations.NotNull;
 public class PlayerSpawnLocationEvent extends PlayerEvent {
     private static final HandlerList handlers = new HandlerList();
     private Location spawnLocation;
+    // Kiterino start - Expose invalid world to PlayerSpawnLocationEvent
+    private final Location initialLocation;
+    private final org.bukkit.NamespacedKey initialDimensionId;
+    // Kiterino end - Expose invalid world to PlayerSpawnLocationEvent
 
     public PlayerSpawnLocationEvent(@NotNull final Player who, @NotNull Location spawnLocation) {
+        // Kiterino start - Expose invalid world to PlayerSpawnLocationEvent
+        this(who, spawnLocation, null, null);
+    }
+
+    public PlayerSpawnLocationEvent(@NotNull final Player who, @NotNull Location spawnLocation, @org.jetbrains.annotations.Nullable Location initialLocation, @org.jetbrains.annotations.Nullable String initialDimensionId) {
         super(who);
+        this.initialLocation = initialLocation == null ? spawnLocation.clone() : initialLocation;
+        this.initialDimensionId = initialDimensionId == null ? null : org.bukkit.NamespacedKey.fromString(initialDimensionId);
+        // Kiterino end - Expose invalid world to PlayerSpawnLocationEvent
         this.spawnLocation = spawnLocation;
     }
 
@@ -40,6 +52,36 @@ public class PlayerSpawnLocationEvent extends PlayerEvent {
         this.spawnLocation = location;
     }
 
+    // Kiterino start - Expose invalid world to PlayerSpawnLocationEvent
+    /**
+     * Gets the initial spawn location the player had.
+     * <br>The initial location might have a nullable world in case the world isn't loaded.
+     *
+     * @return initial spawn location
+     */
+    public @NotNull Location getInitialLocation() {
+        return this.initialLocation.clone();
+    }
+
+    /**
+     * Checks whether the player is initially in an unloaded world
+     *
+     * @return whether the player is initially in an unloaded world
+     */
+    public boolean isInitiallyInUnloadedWorld() {
+        return this.initialLocation.getWorld() == null;
+    }
+
+    /**
+     * Gets the initial dimension id, if present
+     *
+     * @return the initial dimension id
+     */
+    public @org.jetbrains.annotations.Nullable org.bukkit.NamespacedKey getInitialDimensionId() {
+        return this.initialDimensionId;
+    }
+    // Kiterino end - Expose invalid world to PlayerSpawnLocationEvent
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
