From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 23 Jun 2024 15:28:31 +0300
Subject: [PATCH] Allow resending entity metadata


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index e3d666cf6271a67df63d7bd2606ae6b9aac0bed8..6e00b1729245be9ceed4cd5ae07ef7c118fef9c0 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1280,4 +1280,13 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean setRider(@Nullable Player rider, boolean dismount);
     // Kiterino end - Allow setting entity's rider
+
+    // Kiterino start - Allow resending entity metadata
+    /**
+     * Makes entity resend metadata on the next tick
+     *
+     * @param index metadata index
+     */
+    void resendMetadata(int index);
+    // Kiterino end - Allow resending entity metadata
 }
