From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 23 Jun 2024 15:28:31 +0300
Subject: [PATCH] Allow resending entity metadata


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 76e1faf78a1f60534e3d607c5d78086455a4d80a..6ae57e5006dabc7c253cd0ce1386fe2bd08e2ba1 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1293,4 +1293,13 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean setRider(@Nullable Player rider, boolean dismount);
     // Kiterino end - Allow setting entity's rider
+
+    // Kiterino start - Allow resending entity metadata
+    /**
+     * Makes entity resend metadata on the next tick
+     *
+     * @param index metadata index
+     */
+    void resendMetadata(int index);
+    // Kiterino end - Allow resending entity metadata
 }
