From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 21 Feb 2022 11:40:25 +0200
Subject: [PATCH] Non-solid coral fans


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index df992c41d736ee4e89773c2621d261d6afcb2148..bcfbd67f157d31096eaa7d45ef17a21866328b14 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -6133,26 +6133,28 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case DARK_PRISMARINE_SLAB:
             case DARK_PRISMARINE_STAIRS:
             case DAYLIGHT_DETECTOR:
-            case DEAD_BRAIN_CORAL:
+            // Kiterino start - dead coral fans, even despite being made of stone, are not solid
+            //case DEAD_BRAIN_CORAL:
             case DEAD_BRAIN_CORAL_BLOCK:
-            case DEAD_BRAIN_CORAL_FAN:
-            case DEAD_BRAIN_CORAL_WALL_FAN:
-            case DEAD_BUBBLE_CORAL:
+            //case DEAD_BRAIN_CORAL_FAN:
+            //case DEAD_BRAIN_CORAL_WALL_FAN:
+            //case DEAD_BUBBLE_CORAL:
             case DEAD_BUBBLE_CORAL_BLOCK:
-            case DEAD_BUBBLE_CORAL_FAN:
-            case DEAD_BUBBLE_CORAL_WALL_FAN:
-            case DEAD_FIRE_CORAL:
+            //case DEAD_BUBBLE_CORAL_FAN:
+            //case DEAD_BUBBLE_CORAL_WALL_FAN:
+            //case DEAD_FIRE_CORAL:
             case DEAD_FIRE_CORAL_BLOCK:
-            case DEAD_FIRE_CORAL_FAN:
-            case DEAD_FIRE_CORAL_WALL_FAN:
-            case DEAD_HORN_CORAL:
+            //case DEAD_FIRE_CORAL_FAN:
+            //case DEAD_FIRE_CORAL_WALL_FAN:
+            //case DEAD_HORN_CORAL:
             case DEAD_HORN_CORAL_BLOCK:
-            case DEAD_HORN_CORAL_FAN:
-            case DEAD_HORN_CORAL_WALL_FAN:
-            case DEAD_TUBE_CORAL:
+            //case DEAD_HORN_CORAL_FAN:
+            //case DEAD_HORN_CORAL_WALL_FAN:
+            //case DEAD_TUBE_CORAL:
             case DEAD_TUBE_CORAL_BLOCK:
-            case DEAD_TUBE_CORAL_FAN:
-            case DEAD_TUBE_CORAL_WALL_FAN:
+            //case DEAD_TUBE_CORAL_FAN:
+            //case DEAD_TUBE_CORAL_WALL_FAN:
+            // Kiterino end
             case DECORATED_POT:
             case DEEPSLATE:
             case DEEPSLATE_BRICKS:
