From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 23 Jun 2024 15:28:31 +0300
Subject: [PATCH] Allow resending entity metadata


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index ae048744121bf1554a119acc7da472a45d4b577a..48788493944ae3c5a6a1a5be5bd50dcf337e4723 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1280,4 +1280,13 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean setRider(@Nullable Player rider, boolean dismount);
     // Kiterino end - Allow setting entity's rider
+
+    // Kiterino start - Allow resending entity metadata
+    /**
+     * Makes entity resend metadata on the next tick
+     *
+     * @param index metadata index
+     */
+    void resendMetadata(int index);
+    // Kiterino end - Allow resending entity metadata
 }
