From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 19 Feb 2021 23:28:54 +0200
Subject: [PATCH] Flammable cobwebs


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 4b9c4d623d481e2a009fb163e8af226987d6f445..7268a548b625534e1dc143c59593c1bdbfaa70e4 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -7496,6 +7496,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             return false;
         }
         switch (this) {
+            case COBWEB: // Kiterino - isFlammable
             //<editor-fold defaultstate="collapsed" desc="isFlammable">
             case ACACIA_DOOR:
             case ACACIA_FENCE:
@@ -7844,6 +7845,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             return false;
         }
         switch (this) {
+            case COBWEB: // Kiterino - isBurnable
             //<editor-fold defaultstate="collapsed" desc="isBurnable">
             case ACACIA_FENCE:
             case ACACIA_FENCE_GATE:
