From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 19 Feb 2021 23:28:54 +0200
Subject: [PATCH] Flammable cobwebs


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index fe60ce94a7bb0349dafd1488bdc60990717f713f..d4d0df1624fef8f05cf068572b362a08887381c9 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -7498,6 +7498,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             return false;
         }
         switch (this) {
+            case COBWEB: // Kiterino - isFlammable
             //<editor-fold defaultstate="collapsed" desc="isFlammable">
             case ACACIA_DOOR:
             case ACACIA_FENCE:
@@ -7846,6 +7847,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             return false;
         }
         switch (this) {
+            case COBWEB: // Kiterino - isBurnable
             //<editor-fold defaultstate="collapsed" desc="isBurnable">
             case ACACIA_FENCE:
             case ACACIA_FENCE_GATE:
