From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 19 Feb 2021 23:28:54 +0200
Subject: [PATCH] Flammable cobwebs


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 00a009cacee6c27060b433ec58ff903ed5e97de5..78f4c71c9d4fac9b01bf346762325f203944c6f5 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -6915,6 +6915,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
             return false;
         }
         switch (this) {
+            case COBWEB: // Kiterino - isFlammable
             //<editor-fold defaultstate="collapsed" desc="isFlammable">
             case ACACIA_DOOR:
             case ACACIA_FENCE:
@@ -7248,6 +7249,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
             return false;
         }
         switch (this) {
+            case COBWEB: // Kiterino - isBurnable
             //<editor-fold defaultstate="collapsed" desc="isBurnable">
             case ACACIA_FENCE:
             case ACACIA_FENCE_GATE:
