From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 27 Jun 2024 06:36:56 +0300
Subject: [PATCH] More resource pack API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index fb190f104fa2bb714a12894c92e2c350d9e1932f..36856ce2f4123be76f1edc369d3695dfd2ae7c6d 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3091,4 +3091,30 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end
+
+    // Kiterino start - More resource pack API
+    /**
+     * Creates server resource pack
+     *
+     * @param uuid pack id
+     * @param url pack url
+     * @param hash pack sha-1 hash
+     * @param required whether the resource pack is required
+     * @param prompt pack prompt
+     * @return resource pack data
+     */
+    public static @NotNull ResourcePack createResourcePack(@NotNull UUID uuid, @NotNull String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt) {
+        return server.createResourcePack(uuid, url, hash, required, prompt);
+    }
+
+    /**
+     * Sets (or removes if null) the server resource pack
+     *
+     * @param resourcePack resource pack
+     * @see #createResourcePack(UUID, String, String, boolean, net.kyori.adventure.text.Component)
+     */
+    public static void setServerResourcePack(@Nullable ResourcePack resourcePack) {
+        server.setServerResourcePack(resourcePack);
+    }
+    // Kiterino end - More resource pack API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 48356b6df24e67fbe7e37a11a081db57ce44415a..3b68511c0e2851b53c933f737ac3839167835553 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2727,4 +2727,26 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end
+
+    // Kiterino start - More resource pack API
+    /**
+     * Creates server resource pack
+     *
+     * @param uuid pack id
+     * @param url pack url
+     * @param hash pack sha-1 hash
+     * @param required whether the resource pack is required
+     * @param prompt pack prompt
+     * @return resource pack data
+     */
+    @NotNull ResourcePack createResourcePack(@NotNull UUID uuid, @NotNull String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt);
+
+    /**
+     * Sets (or removes if null) the server resource pack
+     *
+     * @param resourcePack resource pack
+     * @see #createResourcePack(UUID, String, String, boolean, net.kyori.adventure.text.Component)
+     */
+    void setServerResourcePack(@Nullable ResourcePack resourcePack);
+    // Kiterino end - More resource pack API
 }
