From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 16 Sep 2024 18:02:03 +0300
Subject: [PATCH] Send ItemStack API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 18b30ee77fe42d7d543699eb0b5da43fcb732990..ecab002941f027a3eb168544e2e1592eaa8c3e74 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -4040,4 +4040,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     void sendRedTint(double percentage);
     // Kiterino end - Red tint API
+
+    // Kiterino start - Send ItemStack API
+    /**
+     * Sends set item packet. This will not actually set the item in the provided slot.
+     *
+     * @param slot item slot
+     * @param item item
+     */
+    void sendItem(int slot, @Nullable ItemStack item);
+
+    /**
+     * Sends set item packet for item on the cursor. This will not actually set the item on the cursor.
+     *
+     * @param item item
+     */
+    default void sendCursorItem(@Nullable ItemStack item) {
+        sendItem(me.sosedik.kiterino.inventory.InventorySlotHelper.CURSOR, item);
+    }
+    // Kiterino end - Send ItemStack API
 }
