From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 15 Dec 2021 19:24:31 +0100
Subject: [PATCH] Note color API


diff --git a/src/main/java/org/bukkit/Note.java b/src/main/java/org/bukkit/Note.java
index fc3da7ce6f7948aeab0962d9472e8f3a126834cf..f0152668bc699aa917079a3ce4cb6674d8e85b04 100644
--- a/src/main/java/org/bukkit/Note.java
+++ b/src/main/java/org/bukkit/Note.java
@@ -119,6 +119,7 @@ public class Note {
     }
 
     private final byte note;
+    private final net.kyori.adventure.text.format.TextColor color; // Kiterino
 
     /**
      * Creates a new note.
@@ -130,6 +131,7 @@ public class Note {
         Preconditions.checkArgument(note >= 0 && note <= 24, "The note value has to be between 0 and 24.");
 
         this.note = (byte) note;
+        this.color = getColor(note); // Kiterino
     }
 
     /**
@@ -150,6 +152,7 @@ public class Note {
         }
 
         this.note = (byte) (octave * Tone.TONES_COUNT + tone.getId(sharped));
+        this.color = getColor(note); // Kiterino
     }
 
     /**
@@ -280,4 +283,46 @@ public class Note {
     public String toString() {
         return "Note{" + getTone().toString() + (isSharped() ? "#" : "") + "}";
     }
+
+    // Kiterino start
+    /**
+     * Get played note color
+     *
+     * @return note color
+     */
+    @NotNull
+    public net.kyori.adventure.text.format.TextColor getColor() {
+        return color;
+    }
+
+    private static @NotNull net.kyori.adventure.text.format.TextColor getColor(int note) {
+        return switch (note) {
+            case 0 -> net.kyori.adventure.text.format.TextColor.fromHexString("#77D700");
+            case 1 -> net.kyori.adventure.text.format.TextColor.fromHexString("#95C000");
+            case 2 -> net.kyori.adventure.text.format.TextColor.fromHexString("#B2A500");
+            case 3 -> net.kyori.adventure.text.format.TextColor.fromHexString("#CC8600");
+            case 4 -> net.kyori.adventure.text.format.TextColor.fromHexString("#E26500");
+            case 5 -> net.kyori.adventure.text.format.TextColor.fromHexString("#F34100");
+            case 6 -> net.kyori.adventure.text.format.TextColor.fromHexString("#FC1E00");
+            case 7 -> net.kyori.adventure.text.format.TextColor.fromHexString("#FE000F");
+            case 8 -> net.kyori.adventure.text.format.TextColor.fromHexString("#F70033");
+            case 9 -> net.kyori.adventure.text.format.TextColor.fromHexString("#E8005A");
+            case 10 -> net.kyori.adventure.text.format.TextColor.fromHexString("#CF0083");
+            case 11 -> net.kyori.adventure.text.format.TextColor.fromHexString("#AE00A9");
+            case 12 -> net.kyori.adventure.text.format.TextColor.fromHexString("#8600CC");
+            case 13 -> net.kyori.adventure.text.format.TextColor.fromHexString("#5B00E7");
+            case 14 -> net.kyori.adventure.text.format.TextColor.fromHexString("#2D00F9");
+            case 15 -> net.kyori.adventure.text.format.TextColor.fromHexString("#020AFE");
+            case 16 -> net.kyori.adventure.text.format.TextColor.fromHexString("#0037F6");
+            case 17 -> net.kyori.adventure.text.format.TextColor.fromHexString("#0068E0");
+            case 18 -> net.kyori.adventure.text.format.TextColor.fromHexString("#009ABC");
+            case 19 -> net.kyori.adventure.text.format.TextColor.fromHexString("#00C68D");
+            case 20 -> net.kyori.adventure.text.format.TextColor.fromHexString("#00E958");
+            case 21 -> net.kyori.adventure.text.format.TextColor.fromHexString("#00FC21");
+            case 22 -> net.kyori.adventure.text.format.TextColor.fromHexString("#1FFC00");
+            case 23 -> net.kyori.adventure.text.format.TextColor.fromHexString("#59E800");
+            default -> net.kyori.adventure.text.format.TextColor.fromHexString("#94C100");
+        };
+    }
+    // Kiterino end
 }
