From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 16 Mar 2022 12:43:34 +0200
Subject: [PATCH] More Sound API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 73a3d41ab7e8cf9772cf1fc6688bd3b6fc93d687..869ee723e0fbb9403cfa57812f34150e05e8b0b3 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -840,4 +840,33 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
         return getPersistentData().getOrCreateCompound(org.bukkit.Bukkit.getName());
     }
     // Kiterino end - Kiterino NBT additions
+
+    // Kiterino start - More Sound API
+    /**
+     * Play a sound at the location of the block.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        playSound(sound, org.bukkit.SoundCategory.BLOCKS, volume, pitch);
+    }
+
+    /**
+     * Play a sound at the location of the block.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param category The category of the sound
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation().toCenterLocation(), sound, category, volume, pitch);
+    }
+    // Kiterino end - More Sound API
 }
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index e6eb00347d30b10dfbbb171ee1ebed965467724b..6365c6d3bbbc0e47b2edaf0073d17ad5f9d72c98 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1322,4 +1322,38 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
         });
     }
     // Kiterino end - Kiterino NBT additions
+
+    // Kiterino start - More Sound API
+    /**
+     * Play a sound at the location of the entity.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        org.bukkit.SoundCategory soundGroup;
+        if (this instanceof HumanEntity) soundGroup = org.bukkit.SoundCategory.PLAYERS;
+        else if (this instanceof Ambient) soundGroup = org.bukkit.SoundCategory.AMBIENT;
+        else if (this instanceof Monster) soundGroup = org.bukkit.SoundCategory.HOSTILE;
+        else soundGroup = org.bukkit.SoundCategory.NEUTRAL;
+        playSound(sound, soundGroup, volume, pitch);
+    }
+
+    /**
+     * Play a sound at the location of the entity.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param category The category of the sound
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation(), sound, category, volume, pitch);
+    }
+    // Kiterino end - More Sound API
 }
