From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 17 Mar 2022 19:05:46 +0200
Subject: [PATCH] More item drop API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 869ee723e0fbb9403cfa57812f34150e05e8b0b3..4157e2072db6c197afd69bb6312d1129db4c3df8 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -869,4 +869,54 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
         getWorld().playSound(getLocation().toCenterLocation(), sound, category, volume, pitch);
     }
     // Kiterino end - More Sound API
+
+    // Kiterino start - More item drop API
+    /**
+     * Drops an item at the location of the block.
+     *
+     * @param item ItemStack to drop
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default org.bukkit.entity.Item dropItem(@NotNull ItemStack item) {
+        return getWorld().dropItem(getLocation().toCenterLocation(), item);
+    }
+
+    /**
+     * Drops an item at the location of the block.
+     * Note that functions will run before the entity is spawned
+     *
+     * @param item ItemStack to drop
+     * @param function the function to be run before the entity is spawned.
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default org.bukkit.entity.Item dropItem(@NotNull ItemStack item, @Nullable org.bukkit.util.Consumer<org.bukkit.entity.Item> function) {
+        return getWorld().dropItem(getLocation().toCenterLocation(), item, function);
+    }
+
+    /**
+     * Drops an item at the location of the block with a random offset
+     *
+     * @param item ItemStack to drop
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default org.bukkit.entity.Item dropItemNaturally(@NotNull ItemStack item) {
+        return getWorld().dropItemNaturally(getLocation().toCenterLocation(), item);
+    }
+
+    /**
+     * Drops an item at the location of the block with a random offset
+     * Note that functions will run before the entity is spawned
+     *
+     * @param item ItemStack to drop
+     * @param function the function to be run before the entity is spawned.
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default org.bukkit.entity.Item dropItemNaturally(@NotNull ItemStack item, @Nullable org.bukkit.util.Consumer<org.bukkit.entity.Item> function) {
+        return getWorld().dropItem(getLocation().toCenterLocation(), item, function);
+    }
+    // Kiterino end - More item drop API
 }
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 6365c6d3bbbc0e47b2edaf0073d17ad5f9d72c98..cb01bdeb74974c02aac0b693e59137c7104e3de7 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1356,4 +1356,54 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
         getWorld().playSound(getLocation(), sound, category, volume, pitch);
     }
     // Kiterino end - More Sound API
+
+    // Kiterino start - More item drop API
+    /**
+     * Drops an item at the location of the entity.
+     *
+     * @param item ItemStack to drop
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default Item dropItem(@NotNull org.bukkit.inventory.ItemStack item) {
+        return getWorld().dropItem(getLocation(), item);
+    }
+
+    /**
+     * Drops an item at the location of the entity.
+     * Note that functions will run before the entity is spawned
+     *
+     * @param item ItemStack to drop
+     * @param function the function to be run before the entity is spawned.
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default Item dropItem(@NotNull org.bukkit.inventory.ItemStack item, @Nullable org.bukkit.util.Consumer<Item> function) {
+        return getWorld().dropItem(getLocation(), item, function);
+    }
+
+    /**
+     * Drops an item at the location of the entity with a random offset
+     *
+     * @param item ItemStack to drop
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default Item dropItemNaturally(@NotNull org.bukkit.inventory.ItemStack item) {
+        return getWorld().dropItemNaturally(getLocation(), item);
+    }
+
+    /**
+     * Drops an item at the location of the entity with a random offset
+     * Note that functions will run before the entity is spawned
+     *
+     * @param item ItemStack to drop
+     * @param function the function to be run before the entity is spawned.
+     * @return ItemDrop entity created as a result of this method
+     */
+    @NotNull
+    default Item dropItemNaturally(@NotNull org.bukkit.inventory.ItemStack item, @Nullable org.bukkit.util.Consumer<Item> function) {
+        return getWorld().dropItem(getLocation(), item, function);
+    }
+    // Kiterino end - More item drop API
 }
