From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 16 Mar 2022 12:43:34 +0200
Subject: [PATCH] More Sound API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 392a28a6ad12bb321e4c8ff6a5947a454324118f..aef157fdef1355ee0c0392b1b7837fd595034619 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -817,5 +817,32 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
     default de.tr7zw.changeme.nbtapi.NBTCompound getKNBT() {
         return getNBTC().getOrCreateCompound(org.bukkit.Bukkit.getName());
     }
+
+    /**
+     * Play a sound at the location of the block.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        playSound(sound, org.bukkit.SoundCategory.BLOCKS, volume, pitch);
+    }
+
+    /**
+     * Play a sound at the location of the block.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param category The category of the sound
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation().toCenterLocation(), sound, category, volume, pitch);
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 89c7d105423216cd6ae8439f0db77c6e5813fea5..9a4fe24a9c9380f3cc4cdbeb44e960d8d9e297b3 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -989,5 +989,37 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     default de.tr7zw.changeme.nbtapi.NBTEntity getNBT() {
         return new de.tr7zw.changeme.nbtapi.NBTEntity(this);
     }
+
+    /**
+     * Play a sound at the location of the entity.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        org.bukkit.SoundCategory soundGroup;
+        if (this instanceof HumanEntity) soundGroup = org.bukkit.SoundCategory.PLAYERS;
+        else if (this instanceof Ambient) soundGroup = org.bukkit.SoundCategory.AMBIENT;
+        else if (this instanceof Monster) soundGroup = org.bukkit.SoundCategory.HOSTILE;
+        else soundGroup = org.bukkit.SoundCategory.NEUTRAL;
+        playSound(sound, soundGroup, volume, pitch);
+    }
+
+    /**
+     * Play a sound at the location of the entity.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param category The category of the sound
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation(), sound, category, volume, pitch);
+    }
     // Kiterino end
 }
