From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 8 Jul 2021 22:46:50 +0300
Subject: [PATCH] Elder Guardian appearance API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 54fdfe7de065e57451b4a7cd0bcbef16eab00dd8..a0a6e956b1b9603f77b028d2ae010585426a5eb8 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2826,5 +2826,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * Forces player to sleep
      */
     void forceSleep();
+
+    /**
+     * Shows elder guardian effect with a sound
+     *
+     * @see #showElderGuardian(boolean)
+     */
+    default void showElderGuardian() {
+        showElderGuardian(true);
+    }
+
+    /**
+     * Shows elder guardian effect and optionally plays the sound
+     *
+     * @param playSound whether sound should be played
+     */
+    void showElderGuardian(boolean playSound);
     // Kiterino end
 }
