From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 28 Sep 2024 01:03:13 +0300
Subject: [PATCH] Anger nearby piglins API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 05697c0668aafc58c71351dc23eb6f49e592e981..cb5992f7e186be13a0b9e1e72e6a75aefa47086a 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -4059,4 +4059,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         sendItem(me.sosedik.kiterino.inventory.InventorySlotHelper.CURSOR, item);
     }
     // Kiterino end - Send ItemStack API
+
+    // Kiterino start - Anger nearby piglins API
+    /**
+     * Triggers piglins angrying AI
+     *
+     * @param blockOpen whether the cause of anger is an opened block
+     */
+    void angerNearbyPiglins(boolean blockOpen);
+    // Kiterino end - Anger nearby piglins API
 }
