From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 8 Jul 2021 22:46:50 +0300
Subject: [PATCH] Elder Guardian appearance API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index b590398de3d4a5f8c933515252d57b4f2dbff450..00fecf5e9f75a850db613d0484616faebcde4ec8 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2819,5 +2819,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * Forces player to sleep
      */
     void forceSleep();
+
+    /**
+     * Shows elder guardian effect with a sound
+     *
+     * @see #showElderGuardian(boolean)
+     */
+    default void showElderGuardian() {
+        showElderGuardian(true);
+    }
+
+    /**
+     * Shows elder guardian effect and optionally plays the sound
+     *
+     * @param playSound whether sound should be played
+     */
+    void showElderGuardian(boolean playSound);
     // Kiterino end
 }
