From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 7 Jun 2022 15:13:53 +0300
Subject: [PATCH] Max stack size API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index be0aa0e45c66a9e1f00d4a7f705395b7486a14ca..2f6aed7d7e543ab4c6a84751255e42a6a8287f9a 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2964,4 +2964,16 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    public static void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        server.setMaxStackSize(material, maxStackSize);
+    }
+    // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 5ee589e58e2a0d9939abec4c4814337e89fb0781..27f88bb7d11370f2ecd5517c46f95c33191b1082 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4392,7 +4392,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     private final int id;
     private final Constructor<? extends MaterialData> ctor;
     private static final Map<String, Material> BY_NAME = Maps.newHashMap();
-    private final int maxStack;
+    int maxStack; // Kiterino - remove final, make package-private
     private final short durability;
     public final Class<?> data;
     private final boolean legacy;
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d8aa4629838fd0aaa36835c99fbd284f4b034aee..6cb4dd4f3c0f1a4a065207283dcdc10f6eed3745 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2600,4 +2600,16 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    default void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        material.maxStack = maxStackSize;
+    }
+    // Kiterino end
 }
