From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 19 Feb 2021 23:28:54 +0200
Subject: [PATCH] Flammable cobwebs


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 5e59be9351f5ac95999e018f9fc9587a02247003..a3d4feb26c0213c1c4e72e6b6227f3d2d1696865 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -6690,6 +6690,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
             case CAMPFIRE:
             case CARTOGRAPHY_TABLE:
             case CHEST:
+            case COBWEB: // Kiterino
             case COMPOSTER:
             case CRAFTING_TABLE:
             case CYAN_BANNER:
@@ -6972,6 +6973,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
             case CAVE_VINES:
             case CAVE_VINES_PLANT:
             case COAL_BLOCK:
+            case COBWEB: // Kiterino
             case COMPOSTER:
             case CORNFLOWER:
             case CYAN_CARPET:
