From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 7 Jun 2022 15:13:53 +0300
Subject: [PATCH] Max stack size API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index d5027348cf3243b1c9d5ff8a3001df6c7eb3ae24..2c177a5144a1cb3a9a5f5fdb753df35869d8c088 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2888,4 +2888,16 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    public static void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        server.setMaxStackSize(material, maxStackSize);
+    }
+    // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 57beb955d4153128a0b43278b313c38b4d2801c9..a58d8d458ef207d74dd05542b413bcf3623548ca 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4389,7 +4389,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     private final int id;
     private final Constructor<? extends MaterialData> ctor;
     private static final Map<String, Material> BY_NAME = Maps.newHashMap();
-    private final int maxStack;
+    int maxStack; // Kiterino - remove final, make package-private
     private final short durability;
     public final Class<?> data;
     private final boolean legacy;
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a0dcf12758ed0b0474028161c0faf9157b9e635d..d8291fff17b80d2d5819297f76de62d136197d00 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2531,4 +2531,16 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    default void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        material.maxStack = maxStackSize;
+    }
+    // Kiterino end
 }
