From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 7 Jun 2022 15:13:53 +0300
Subject: [PATCH] Max stack size API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 7312c9c556c237af3d4d6de872a9355fbc018f5a..2921887e2e4302f50ddef65dd01e0ffc0d973910 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2596,4 +2596,16 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    public static void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        server.setMaxStackSize(material, maxStackSize);
+    }
+    // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 1e12fd9a219db496d3c746811cc53fcd1c8bd4fc..d42579395a3cd375dab27ac86291f1b9ca8e8195 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4434,7 +4434,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     private final int id;
     private final Constructor<? extends MaterialData> ctor;
     private static final Map<String, Material> BY_NAME = Maps.newHashMap();
-    private final int maxStack;
+    int maxStack; // Kiterino - remove final, make package-private
     private final short durability;
     public final Class<?> data;
     private final boolean legacy;
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 62319d068ecbb0f357f34bf69518f985035c6670..652c5a08bcf9010f2b98b4f5b77093459d777dab 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2266,4 +2266,16 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    default void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        material.maxStack = maxStackSize;
+    }
+    // Kiterino end
 }
