From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 19 Feb 2021 23:28:54 +0200
Subject: [PATCH] Flammable cobwebs


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 08830e1a143e14ed3da165d3bcf5a8301374ffd9..067005a9f6f85f6d8807d9570145f4ece766812a 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -6689,6 +6689,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
             case CAMPFIRE:
             case CARTOGRAPHY_TABLE:
             case CHEST:
+            case COBWEB: // Kiterino
             case COMPOSTER:
             case CRAFTING_TABLE:
             case CYAN_BANNER:
@@ -6971,6 +6972,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
             case CAVE_VINES:
             case CAVE_VINES_PLANT:
             case COAL_BLOCK:
+            case COBWEB: // Kiterino
             case COMPOSTER:
             case CORNFLOWER:
             case CYAN_CARPET:
