From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 28 Sep 2024 06:09:37 +0300
Subject: [PATCH] Add combat tracker API


diff --git a/src/main/java/me/sosedik/kiterino/world/damagesource/CombatEntry.java b/src/main/java/me/sosedik/kiterino/world/damagesource/CombatEntry.java
new file mode 100644
index 0000000000000000000000000000000000000000..b966423643f66b1a34d9d372007e8e37d6491db6
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/damagesource/CombatEntry.java
@@ -0,0 +1,27 @@
+package me.sosedik.kiterino.world.damagesource;
+
+import org.bukkit.damage.DamageSource;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Wrapper around combat entry
+ */
+@ApiStatus.Experimental
+public interface CombatEntry {
+
+    /**
+     * Gets the damage source
+     *
+     * @return the damage source
+     */
+    @NotNull DamageSource getDamageSource();
+
+    /**
+     * Gets the damage caused
+     *
+     * @return the damage caused
+     */
+    float getDamage();
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/damagesource/CombatTracker.java b/src/main/java/me/sosedik/kiterino/world/damagesource/CombatTracker.java
new file mode 100644
index 0000000000000000000000000000000000000000..d920f608b9913c267cf55903413c47050cf4539a
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/damagesource/CombatTracker.java
@@ -0,0 +1,80 @@
+package me.sosedik.kiterino.world.damagesource;
+
+import net.kyori.adventure.text.Component;
+import org.bukkit.damage.DamageSource;
+import org.bukkit.entity.LivingEntity;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+/**
+ * Wrapper around the combat tracker
+ */
+@ApiStatus.Experimental
+public interface CombatTracker {
+
+    /**
+     * Gets the entity behind this combat tracker
+     *
+     * @return the entity behind this combat tracker
+     */
+    @NotNull LivingEntity getEntity();
+
+    /**
+     * Gets the mutable list of combat entries
+     *
+     * @return the mutable list of combat entries
+     */
+    @NotNull List<CombatEntry> getCombatEntries();
+
+    /**
+     * Creates a new combat entry
+     *
+     * @param damageSource damage source
+     * @param damage damage amount
+     * @return a new combat entry
+     */
+    @NotNull CombatEntry createCombatEntry(@NotNull DamageSource damageSource, float damage);
+
+    /**
+     * Calculates the most significant damage entry for a fall
+     *
+     * @return the most significant damage entry for a fall
+     */
+    @Nullable CombatEntry getMostSignificantFall();
+
+    /**
+     * Checks whether the entity is in combat
+     *
+     * @return whether the entity is in combat
+     */
+    boolean isInCombat();
+
+    /**
+     * Gets the last or current combat duration
+     *
+     * @return the combat duration
+     * @see #isInCombat()
+     */
+    int getCombatDuration();
+
+    /**
+     * Records a new damage.
+     * <p>
+     * Note: this will not actually cause the damage, just store it into the combat history.
+     *
+     * @param damageSource damage source
+     * @param damage damage amount
+     */
+    void recordDamage(@NotNull DamageSource damageSource, float damage);
+
+    /**
+     * Constructs a death message
+     *
+     * @return a death message
+     */
+    @NotNull Component getDeathMessage();
+
+}
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index a22909446b7193b1de121bdd6d7b6fbf81dc3be4..6e74b2adacf6c8ac8efbd8d01b82c93708fcc676 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1484,4 +1484,14 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end - API for any mob to burn daylight
+
+    // Kiterino start - Add combat tracker API
+    /**
+     * Gets the entity's combat tracker
+     *
+     * @return the entity's combat tracker
+     */
+    @org.jetbrains.annotations.ApiStatus.Experimental
+    @NotNull me.sosedik.kiterino.world.damagesource.CombatTracker getCombatTracker();
+    // Kiterino end - Add combat tracker API
 }
