From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 16 Mar 2022 12:43:34 +0200
Subject: [PATCH] More Sound API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index f90d24409cd749fdc27cb8959f641f2bb9b4d2d4..22b7713f54b7babffd707e642fa65796d34cc206 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -792,5 +792,32 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
     default de.tr7zw.changeme.nbtapi.NBTCompound getKNBT() {
         return getNBTC().getOrCreateCompound(org.bukkit.Bukkit.getName());
     }
+
+    /**
+     * Play a sound at the location of the block.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        playSound(sound, org.bukkit.SoundCategory.BLOCKS, volume, pitch);
+    }
+
+    /**
+     * Play a sound at the location of the block.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param category The category of the sound
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation().toCenterLocation(), sound, category, volume, pitch);
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index c61941e90b393ac4999a1361d6dc85ad83d8a335..4c1562eea640a3c9f271756050bf077fbb759fdf 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -968,5 +968,37 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     default de.tr7zw.changeme.nbtapi.NBTEntity getNBT() {
         return new de.tr7zw.changeme.nbtapi.NBTEntity(this);
     }
+
+    /**
+     * Play a sound at the location of the entity.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        org.bukkit.SoundCategory soundGroup;
+        if (this instanceof HumanEntity) soundGroup = org.bukkit.SoundCategory.PLAYERS;
+        else if (this instanceof Ambient) soundGroup = org.bukkit.SoundCategory.AMBIENT;
+        else if (this instanceof Monster) soundGroup = org.bukkit.SoundCategory.HOSTILE;
+        else soundGroup = org.bukkit.SoundCategory.NEUTRAL;
+        playSound(sound, soundGroup, volume, pitch);
+    }
+
+    /**
+     * Play a sound at the location of the entity.
+     * <p>
+     * This function will fail silently if Sound is null.
+     *
+     * @param sound The sound to play
+     * @param category The category of the sound
+     * @param volume The volume of the sound
+     * @param pitch The pitch of the sound
+     */
+    default void playSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation(), sound, category, volume, pitch);
+    }
     // Kiterino end
 }
