From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 14 Oct 2022 09:41:48 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index cb01bdeb74974c02aac0b693e59137c7104e3de7..97ff71da733b1f622ee2036bb323f075b1fcb60d 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1154,6 +1154,29 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     @Nullable
     Player getRider();
 
+    // Kiterino start
+    /**
+     * Set the riding player. The player has to be
+     * the first passenger.
+     *
+     * @param rider rider
+     * @return whether rider was set
+     */
+    default boolean setRider(@Nullable Player rider) {
+        return setRider(rider, true);
+    }
+
+    /**
+     * Set the riding player and optionally dismount
+     * the previous one
+     *
+     * @param rider rider
+     * @param dismount whether to dismount the current rider
+     * @return whether rider was set
+     */
+    boolean setRider(@Nullable Player rider, boolean dismount);
+    // Kiterino end
+
     /**
      * Check if entity is being ridden
      *
