From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 24 Oct 2022 18:51:56 +0300
Subject: [PATCH] Get destroy speed from BlockData


diff --git a/src/main/java/org/bukkit/block/data/BlockData.java b/src/main/java/org/bukkit/block/data/BlockData.java
index 40ddf8909636ab3cdd78f4e66123358c6a3850f7..ebe5eaf7d923017306492a0abd3bedeccb8c7920 100644
--- a/src/main/java/org/bukkit/block/data/BlockData.java
+++ b/src/main/java/org/bukkit/block/data/BlockData.java
@@ -245,4 +245,28 @@ public interface BlockData extends Cloneable {
      */
     boolean isRandomlyTicked();
     // Paper end
+
+    // Kiterino start
+    /**
+     * Gets the speed at which this block will be destroyed by a given {@link org.bukkit.inventory.ItemStack}
+     * <p>Default value is 1.0
+     *
+     * @param item {@link org.bukkit.inventory.ItemStack} used to mine this Block
+     * @see #getDestroySpeed(org.bukkit.inventory.ItemStack, boolean)
+     * @return the speed that this Block will be mined by the given {@link org.bukkit.inventory.ItemStack}
+     */
+    default float getDestroySpeed(@NotNull org.bukkit.inventory.ItemStack item) {
+        return getDestroySpeed(item, false);
+    }
+
+    /**
+     * Gets the speed at which this block will be destroyed by a given {@link org.bukkit.inventory.ItemStack}
+     * <p>Default value is 1.0
+     *
+     * @param item {@link org.bukkit.inventory.ItemStack} used to mine this Block
+     * @param considerEnchants true to look at enchants on the item
+     * @return the speed that this Block will be mined by the given {@link org.bukkit.inventory.ItemStack}
+     */
+    float getDestroySpeed(@NotNull org.bukkit.inventory.ItemStack item, boolean considerEnchants);
+    // Kiterino end
 }
