From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 27 Oct 2022 15:47:27 +0300
Subject: [PATCH] Spawning player API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 4a83b02a7c072463248dd4fdd96bbb08194a8a3d..0e0b6a351c6eff5b3dd96d4ed474ffc1b077e8e4 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1382,5 +1382,16 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     default Item dropItemNaturally(@NotNull org.bukkit.inventory.ItemStack item, @Nullable org.bukkit.util.Consumer<Item> function) {
         return getWorld().dropItem(getLocation(), item, function);
     }
+
+    /**
+     * Gets the {@link UUID} of the player for whom
+     * this entity was spawned.
+     * <p>Entity must be spawned by Paper's "per-player-mob-spawns"
+     * feature for this to work.
+     *
+     * @return unique id of a spawning player, if any
+     */
+    @Nullable
+    UUID getSpawningPlayerUuid();
     // Kiterino end
 }
