From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 14 Oct 2022 09:41:48 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 0d454be0fcb809d784f9b0f8cd965865fbe95397..341995cf68cd45e4b45f3f6b318103239d70c546 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -980,6 +980,29 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     @Nullable
     Player getRider();
 
+    // Kiterino start
+    /**
+     * Set the riding player. The player has to be
+     * the first passenger.
+     *
+     * @param rider rider
+     * @return whether rider was set
+     */
+    default boolean setRider(@Nullable Player rider) {
+        return setRider(rider, true);
+    }
+
+    /**
+     * Set the riding player and optionally dismount
+     * the previous one
+     *
+     * @param rider rider
+     * @param dismount whether to dismount the current rider
+     * @return whether rider was set
+     */
+    boolean setRider(@Nullable Player rider, boolean dismount);
+    // Kiterino end
+
     /**
      * Check if entity is being ridden
      *
