From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 7 Jun 2022 15:13:53 +0300
Subject: [PATCH] Max stack size API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index cbcc6eddbb732c9450baf39805bcbb0f8f4f427f..c42882628d9bd6a63ad280319c6bbd8e3a876721 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2477,4 +2477,16 @@ public final class Bukkit {
         server.removeFuel(material);
     }
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    public static void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        server.setMaxStackSize(material, maxStackSize);
+    }
+    // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 067005a9f6f85f6d8807d9570145f4ece766812a..452e2c35b8ef6c5fc792ba8070b6a733d24011b5 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4066,7 +4066,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
     private final int id;
     private final Constructor<? extends MaterialData> ctor;
     private static final Map<String, Material> BY_NAME = Maps.newHashMap();
-    private final int maxStack;
+    int maxStack; // Kiterino - remove final, make package-private
     private final short durability;
     public final Class<?> data;
     private final boolean legacy;
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a8179cb6a2ca9c2ebb5b6c756f4002aa92d46d9c..8be13cb71b0cc6b6f871f44a56b0de63315661cd 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2162,4 +2162,16 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public void removeFuel(@NotNull Material material);
     // Purpur end
+
+    // Kiterino start
+    /**
+     * Set max stack size of an item
+     *
+     * @param material item type
+     * @param maxStackSize max stack size
+     */
+    default void setMaxStackSize(@NotNull Material material, int maxStackSize) {
+        material.maxStack = maxStackSize;
+    }
+    // Kiterino end
 }
