From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 14 Oct 2022 09:41:48 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 04868e6c1c79681188cf4adda1e36155e2f61cab..35aeb0010cf7ff3e3f75c3f82515a278ab31eba1 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -961,6 +961,29 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     @Nullable
     Player getRider();
 
+    // Kiterino start
+    /**
+     * Set the riding player. The player has to be
+     * the first passenger.
+     *
+     * @param rider rider
+     * @return whether rider was set
+     */
+    default boolean setRider(@Nullable Player rider) {
+        return setRider(rider, true);
+    }
+
+    /**
+     * Set the riding player and optionally dismount
+     * the previous one
+     *
+     * @param rider rider
+     * @param dismount whether to dismount the current rider
+     * @return whether rider was set
+     */
+    boolean setRider(@Nullable Player rider, boolean dismount);
+    // Kiterino end
+
     /**
      * Check if entity is being ridden
      *
