From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 21 Jan 2023 20:26:45 +0200
Subject: [PATCH] More FishHook API


diff --git a/src/main/java/org/bukkit/entity/FishHook.java b/src/main/java/org/bukkit/entity/FishHook.java
index 6ed83d3e4d23e0dc0e1b156a1ee221aaba5c7210..0981cfd1e4b3f0b07432bc61418ef997b921fc9a 100644
--- a/src/main/java/org/bukkit/entity/FishHook.java
+++ b/src/main/java/org/bukkit/entity/FishHook.java
@@ -338,4 +338,20 @@ public interface FishHook extends Projectile {
      */
     void setWaitTime(int ticks);
     // Paper end
+
+    // Kiterino start
+    /**
+     * Get the number of ticks the fish has to swim until biting the hook
+     *
+     * @return the number of ticks the fish has to swim until biting the hook
+     */
+    int getTimeUntilHooked();
+
+    /**
+     * Sets the number of ticks the fish has to swim until biting the hook
+     *
+     * @param ticks ticks
+     */
+    void setTimeUntilHooked(int ticks);
+    // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/event/player/PlayerFishEvent.java b/src/main/java/org/bukkit/event/player/PlayerFishEvent.java
index 45342030ad0f46632d3ee9a6d0348251f8ee375f..afb682631631f855490446808f3ec403a1cc9551 100644
--- a/src/main/java/org/bukkit/event/player/PlayerFishEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerFishEvent.java
@@ -163,6 +163,12 @@ public class PlayerFishEvent extends PlayerEvent implements Cancellable {
          * Called when there is a bite on the hook and it is ready to be reeled
          * in.
          */
-        BITE
+        BITE,
+        // Kiterino start
+        /**
+         * Called when a bobber was lured, and is now waiting to be hooked
+         */
+        LURED
+        // Kiterino end
     }
 }
