From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 Feb 2023 04:21:43 +0200
Subject: [PATCH] Additional drop method


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index b66b11939716f676b1deb7a16d629d7f3b18ef06..39eff3107328089b80320d3d9118c737816f79da 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -840,5 +840,13 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      * @return current active container state id
      */
     int getActiveContainerStateId();
+
+    /**
+     * Make the entity drop the item.
+     * <p>This will call the {@link org.bukkit.event.player.PlayerDropItemEvent} event.
+     *
+     * @param item item to drop
+     */
+    @Nullable Item drop(@NotNull ItemStack item);
     // Kiterino start
 }
