From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 Feb 2023 04:21:43 +0200
Subject: [PATCH] Additional drop method


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 3ac7f0a134b706b9b50b7ff76c658968d692c5fe..df7a17dd08e10eba886f6f3b51226ffe2bacd1b0 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -850,5 +850,13 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      * @return current active container state id
      */
     int getActiveContainerStateId();
+
+    /**
+     * Make the entity drop the item.
+     * <p>This will call the {@link org.bukkit.event.player.PlayerDropItemEvent} event.
+     *
+     * @param item item to drop
+     */
+    @Nullable Item drop(@NotNull ItemStack item);
     // Kiterino start
 }
