From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 3 Mar 2023 20:03:51 +0200
Subject: [PATCH] Additional pathfinding API


diff --git a/src/main/java/com/destroystokyo/paper/entity/Pathfinder.java b/src/main/java/com/destroystokyo/paper/entity/Pathfinder.java
index 43f062257472a06e9e64c2feef6c3b1012aee00e..07f6cdc2068a9a04861dc234e14ce7dc728e5899 100644
--- a/src/main/java/com/destroystokyo/paper/entity/Pathfinder.java
+++ b/src/main/java/com/destroystokyo/paper/entity/Pathfinder.java
@@ -208,5 +208,14 @@ public interface Pathfinder {
          * @return The closest point the path can get to the target location
          */
         @Nullable Location getFinalPoint();
+
+        // Kiterino start
+        /**
+         * Checks whether the final point of this path can be reached
+         *
+         * @return whether the final point of this path can be reached
+         */
+        boolean canReachFinalPoint();
+        // Kiterino end
     }
 }
