From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 24 May 2023 17:13:26 +0300
Subject: [PATCH] Add manual panic mode


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 361c932e6ccadb602ca8c02ce9f6d48bd91f0bec..a96ee45b91f0491f1ddefc0953b74a4d8daf4660 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1364,5 +1364,19 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     @Nullable
     UUID getSpawningPlayerUuid();
+
+    /**
+     * Gets the amount of manually set panic ticks left
+     *
+     * @return the amount of panic ticks
+     */
+    int getPanicTicks();
+
+    /**
+     * Sets the amount of panic ticks
+     *
+     * @param ticks ticks
+     */
+    void setPanicTicks(int ticks);
     // Kiterino end
 }
