From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 24 May 2023 17:13:26 +0300
Subject: [PATCH] Add manual panic mode


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index c4e6eb672beac664276f0e4e25c90029045830c3..a6495b9a096cc7e727db5641a202335670a2231f 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1396,5 +1396,19 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     @Nullable
     UUID getSpawningPlayerUuid();
+
+    /**
+     * Gets the amount of manually set panic ticks left
+     *
+     * @return the amount of panic ticks
+     */
+    int getPanicTicks();
+
+    /**
+     * Sets the amount of panic ticks
+     *
+     * @param ticks ticks
+     */
+    void setPanicTicks(int ticks);
     // Kiterino end
 }
