From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 Feb 2023 04:21:43 +0200
Subject: [PATCH] Additional drop method


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 624e7180e7645ae103a1d7519b37c517d78ddff4..da82bf09bee8bce33203f0677b8fb1f63cda1250 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -838,5 +838,13 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      * @return current active container state id
      */
     int getActiveContainerStateId();
+
+    /**
+     * Make the entity drop the item.
+     * <p>This will call the {@link org.bukkit.event.player.PlayerDropItemEvent} event.
+     *
+     * @param item item to drop
+     */
+    @Nullable Item drop(@NotNull ItemStack item);
     // Kiterino start
 }
