From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 22 Jun 2023 09:10:02 +0300
Subject: [PATCH] Additional Sign API


diff --git a/src/main/java/org/bukkit/block/Sign.java b/src/main/java/org/bukkit/block/Sign.java
index 7983ccb54f5f358dea1ffb530b9cc5bd716fb9b1..60f6316a82f4c14f2827d9b503ad6925fad08dd8 100644
--- a/src/main/java/org/bukkit/block/Sign.java
+++ b/src/main/java/org/bukkit/block/Sign.java
@@ -217,4 +217,20 @@ public interface Sign extends TileState, Colorable {
      */
     @NotNull Side getInteractableSideFor(double x, double z);
     // Paper end
+
+    // Kiterino start
+    /**
+     * Gets the {@link java.util.UUID} of the player who may edit this sign
+     *
+     * @return uuid
+     */
+    @org.jetbrains.annotations.Nullable java.util.UUID getWhoMayEditUuid();
+
+    /**
+     * Sets the player as the one who may edit this sign
+     *
+     * @param uuid player's uuid
+     */
+    void setWhoMayEditUuid(@NotNull java.util.UUID uuid);
+    // Kiterino end
 }
