From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 22 Jun 2023 09:10:02 +0300
Subject: [PATCH] Additional Sign API


diff --git a/src/main/java/org/bukkit/block/Sign.java b/src/main/java/org/bukkit/block/Sign.java
index 94e8f319fe0413a10496d7e5bba70633054765f4..14c69235f9b154707a0db934c2b59493e9781a4d 100644
--- a/src/main/java/org/bukkit/block/Sign.java
+++ b/src/main/java/org/bukkit/block/Sign.java
@@ -198,4 +198,20 @@ public interface Sign extends TileState, Colorable {
      */
     @NotNull Side getInteractableSideFor(double x, double z);
     // Paper end
+
+    // Kiterino start
+    /**
+     * Gets the {@link java.util.UUID} of the player who may edit this sign
+     *
+     * @return uuid
+     */
+    @org.jetbrains.annotations.Nullable java.util.UUID getWhoMayEditUuid();
+
+    /**
+     * Sets the player as the one who may edit this sign
+     *
+     * @param uuid player's uuid
+     */
+    void setWhoMayEditUuid(@NotNull java.util.UUID uuid);
+    // Kiterino end
 }
