From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 24 May 2023 17:13:26 +0300
Subject: [PATCH] Add manual panic mode


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 7db0e8b12fe93802400a3b8f0bd08b88c3ff304b..6f1754be340b757abd6478a014ff22a60f4ff05a 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1180,5 +1180,19 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @return whether the entity has a fixed {@link Pose}
      */
     boolean hasFixedPose();
+
+    /**
+     * Gets the amount of manually set panic ticks left
+     *
+     * @return the amount of panic ticks
+     */
+    int getPanicTicks();
+
+    /**
+     * Sets the amount of panic ticks
+     *
+     * @param ticks ticks
+     */
+    void setPanicTicks(int ticks);
     // Kiterino end
 }
