From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 22 Jun 2023 09:10:02 +0300
Subject: [PATCH] Additional Sign API


diff --git a/src/main/java/org/bukkit/block/Sign.java b/src/main/java/org/bukkit/block/Sign.java
index e6b87ad5b2099800a61e7f585a5404e9e672228a..65e03e05b136c3e41f862ddc85a5901147813e2b 100644
--- a/src/main/java/org/bukkit/block/Sign.java
+++ b/src/main/java/org/bukkit/block/Sign.java
@@ -166,4 +166,28 @@ public interface Sign extends TileState, Colorable {
      */
     @NotNull
     public SignSide getSide(@NotNull Side side);
+
+    // Kiterino start
+    /**
+     * Get the side of the sing that the player is facing
+     *
+     * @param player player
+     * @return sign side
+     */
+    @NotNull Side getFacingSide(@NotNull org.bukkit.entity.Player player);
+
+    /**
+     * Gets the {@link java.util.UUID} of the player who may edit this sign
+     *
+     * @return uuid
+     */
+    @org.jetbrains.annotations.Nullable java.util.UUID getWhoMayEditUuid();
+
+    /**
+     * Sets the player as the one who may edit this sign
+     *
+     * @param uuid player's uuid
+     */
+    void setWhoMayEditUuid(@NotNull java.util.UUID uuid);
+    // Kiterino end
 }
