From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 27 Jun 2023 16:52:18 +0300
Subject: [PATCH] Expose checkBedExists


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 7d8ce1dc557d3ffed9a512d9533f56b5b05e7e25..5feef236146e3019238df023911ebdd1fa17c16c 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1267,4 +1267,14 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end
+
+    // Kiterino start
+
+    /**
+     * Checks whether this entity's bed exists as a physical bed
+     *
+     * @return whether this entity's bed exists
+     */
+    boolean bedExists();
+    // Kiterino end
 }
