From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 21 Oct 2023 00:20:10 +0300
Subject: [PATCH] Allow checking for Halloween season


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 6e5f0e3d5cefea2fb63e36bb144dd2151ee1b2a8..5328759d6401ff1ad00bb333aeea66e23be0634d 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3043,5 +3043,14 @@ public final class Bukkit {
     public static void setMaxStackSize(@NotNull Material material, int maxStackSize) {
         server.setMaxStackSize(material, maxStackSize);
     }
+
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    public static boolean isHalloweenSeason() {
+        return server.isHalloweenSeason();
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d4dce4f1375892ddb42c0be0fb085282d7bcde92..1ec8119e98700ddcba197d7fd6fdfc7db0abe90a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2670,5 +2670,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     default void setMaxStackSize(@NotNull Material material, int maxStackSize) {
         material.maxStack = maxStackSize;
     }
+
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    default boolean isHalloweenSeason() {
+        return getWorlds().get(0).isHalloweenSeason();
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 6c0fe85588d223d92b6ec15afaf35db92f08f0a3..448a2b6cff49ded7bf69a920410ed04d3fb1abe6 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4462,4 +4462,13 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
      */
     boolean applyBoneMeal(@NotNull Location location, @NotNull org.bukkit.block.BlockFace face, boolean showParticles, @Nullable Predicate<org.bukkit.block.BlockState> predicate);
     // Paper end
+
+    // Kiterino start
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    boolean isHalloweenSeason();
+    // Kiterino end
 }
