From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 21 Oct 2023 00:20:10 +0300
Subject: [PATCH] Allow checking for Halloween season


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 6e2a9ba6e09778bd77998264e625413e58af0ded..7391b66f3367fc98d9837b173304e8b3010c700c 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3027,5 +3027,14 @@ public final class Bukkit {
     public static void setMaxStackSize(@NotNull Material material, int maxStackSize) {
         server.setMaxStackSize(material, maxStackSize);
     }
+
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    public static boolean isHalloweenSeason() {
+        return server.isHalloweenSeason();
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8a84fcb34f16f1a56ff458a33498229f119f9fc9..5773d99c6cbe0220d6508f718e8dc7328fbacfff 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2655,5 +2655,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     default void setMaxStackSize(@NotNull Material material, int maxStackSize) {
         material.maxStack = maxStackSize;
     }
+
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    default boolean isHalloweenSeason() {
+        return getWorlds().get(0).isHalloweenSeason();
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 6c0fe85588d223d92b6ec15afaf35db92f08f0a3..448a2b6cff49ded7bf69a920410ed04d3fb1abe6 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4462,4 +4462,13 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
      */
     boolean applyBoneMeal(@NotNull Location location, @NotNull org.bukkit.block.BlockFace face, boolean showParticles, @Nullable Predicate<org.bukkit.block.BlockState> predicate);
     // Paper end
+
+    // Kiterino start
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    boolean isHalloweenSeason();
+    // Kiterino end
 }
