From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 1 Nov 2023 16:53:27 +0200
Subject: [PATCH] Simple water biome cache


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 4157e2072db6c197afd69bb6312d1129db4c3df8..78d2a39c2f041aad037694727b7caea3a01da68c 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -919,4 +919,20 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
         return getWorld().dropItem(getLocation().toCenterLocation(), item, function);
     }
     // Kiterino end - More item drop API
+
+    // Kiterino start - Simple water biome cache
+    /**
+     * Checks whether this chunk has infinite water
+     *
+     * @return whether this chunk has infinite water
+     */
+    boolean hasInfiniteWater();
+
+    /**
+     * Gets the nearest cached water biome
+     *
+     * @return the nearest cached water biome, if present
+     */
+    @org.jetbrains.annotations.Nullable Biome getCachedWaterBiome();
+    // Kiterino end - Simple water biome cache
 }
