From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 8 Nov 2023 03:56:07 +0200
Subject: [PATCH] Set resource pack API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 7024883682539ca1e410bafc9ae588f5544d1692..c6d806400455cbd2e38224c55387e5fca420ecb4 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3058,5 +3058,17 @@ public final class Bukkit {
     public static boolean isHalloweenSeason() {
         return server.isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    public static void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt) {
+        server.setResourcePack(url, hash, required, prompt);
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ef5530d12031bc7e666b3e3d9bed8fce43b6bd0c..37dd1e418711d11c8dc373e35ca3834e227c1b0c 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2685,5 +2685,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     default boolean isHalloweenSeason() {
         return getWorlds().get(0).isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt);
     // Kiterino end
 }
