From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 8 Nov 2023 03:56:07 +0200
Subject: [PATCH] Set resource pack API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 5328759d6401ff1ad00bb333aeea66e23be0634d..0d8a511aa6936b18dc145b6a91067caf0aa24c3e 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3052,5 +3052,17 @@ public final class Bukkit {
     public static boolean isHalloweenSeason() {
         return server.isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    public static void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt) {
+        server.setResourcePack(url, hash, required, prompt);
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1ec8119e98700ddcba197d7fd6fdfc7db0abe90a..4367409e8000769aed6dc03d44c38d4376e16244 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2679,5 +2679,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     default boolean isHalloweenSeason() {
         return getWorlds().get(0).isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt);
     // Kiterino end
 }
