From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 8 Nov 2023 03:56:07 +0200
Subject: [PATCH] Set resource pack API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 7391b66f3367fc98d9837b173304e8b3010c700c..39ba2515493ac40d6bb82b666c03e8209bd7aa26 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3036,5 +3036,17 @@ public final class Bukkit {
     public static boolean isHalloweenSeason() {
         return server.isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    public static void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt) {
+        server.setResourcePack(url, hash, required, prompt);
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 5773d99c6cbe0220d6508f718e8dc7328fbacfff..c3ef953a2ef5bbf5e67753cca100224de440c85b 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2664,5 +2664,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     default boolean isHalloweenSeason() {
         return getWorlds().get(0).isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt);
     // Kiterino end
 }
