From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 8 Nov 2023 03:56:07 +0200
Subject: [PATCH] Set resource pack API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index a5ebb2eef334449e9568137952bfb73aff0aa0b4..70ceee7dd0aaa814162b03b8add13952dad991f5 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3016,5 +3016,17 @@ public final class Bukkit {
     public static boolean isHalloweenSeason() {
         return server.isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    public static void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt) {
+        server.setResourcePack(url, hash, required, prompt);
+    }
     // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 5fd0965051190c52d244a85c2ff80eb63b0ef983..7f87a316aa91905986c92caf78c925e6519b8c88 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2648,5 +2648,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     default boolean isHalloweenSeason() {
         return getWorlds().get(0).isHalloweenSeason();
     }
+
+    /**
+     * Sets the server resource pack. The sha1 hash will be calculated automatically.
+     *
+     * @param url url
+     * @param hash sha1 hash
+     * @param required whether resource pack is required
+     * @param prompt prompt
+     */
+    void setResourcePack(@Nullable String url, @Nullable String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt);
     // Kiterino end
 }
