From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Apr 2024 22:14:56 +0300
Subject: [PATCH] Init


diff --git a/build.gradle.kts b/build.gradle.kts
index c673a2cf682378aac38d047b999d8a1731f7569d..691b475a1d5577170e11a69700c2d98dd8192274 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -25,7 +25,7 @@ abstract class MockitoAgentProvider : CommandLineArgumentProvider {
 // Paper end - configure mockito agent that is needed in newer java versions
 
 dependencies {
-    implementation(project(":purpur-api")) // Pufferfish // Paper // Purpur
+    implementation(project(":kiterino-api")) // Kiterino - Init
     // Paper start
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
@@ -103,14 +103,14 @@ tasks.jar {
         val gitBranch = git("rev-parse", "--abbrev-ref", "HEAD").getText().trim() // Paper
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Purpur", // Pufferfish // Purpur
+            "Implementation-Title" to "Kiterino", // Kiterino - Init
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date, // Paper
-            "Specification-Title" to "Purpur", // Pufferfish // Purpur
+            "Specification-Title" to "Kiterino", // Kiterino - Init
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Purpur Team", // Pufferfish // Purpur
-            "Brand-Id" to "purpurmc:purpur", // Pufferfish // Purpur
-            "Brand-Name" to "Purpur", // Pufferfish // Purpur
+            "Specification-Vendor" to "SoSeDiK", // Kiterino - Init
+            "Brand-Id" to "kiterino:kiterino", // Kiterino - Init
+            "Brand-Name" to "Kiterino", // Kiterino - Init
             "Build-Number" to (build ?: ""),
             "Build-Time" to Instant.now().toString(),
             "Git-Branch" to gitBranch, // Paper
diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 99597258e8e88cd9e2c901c4ac3ff7faeeabee2b..df65a68740089f3367d9084b75f938aaa53041f5 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/org.purpurmc.purpur/purpur-api/pom.properties"); // Pufferfish // Purpur
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/me.sosedik.kiterino/kiterino-api/pom.properties"); // Kiterino - Init
         Properties properties = new Properties();
 
         if (stream != null) {
