From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 31 Jul 2021 20:15:05 +0300
Subject: [PATCH] Add NBT API as a first-class lib


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 21da8b6b0df40b6b88a325ced5d6553ad5912d31..86033a7ece7f83f280b38ff6845bca9d45bc22bb 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -305,6 +305,11 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
         // Purpur end
 
+        // Kiterino start
+        de.tr7zw.changeme.nbtapi.utils.MinecraftVersion.disablePackageWarning();
+        de.tr7zw.changeme.nbtapi.utils.MinecraftVersion.disableUpdateCheck();
+        // Kiterino end
+
         // CraftBukkit start
         // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage)); // Spigot - moved up
         server.loadPlugins();
