From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Apr 2024 22:14:56 +0300
Subject: [PATCH] Hide irrelevant compilation warnings


diff --git a/build.gradle.kts b/build.gradle.kts
index 1bd15526c70be8e68265a7b89fb5190f96f0ad87..0e86e7962be82ccc6f77d651f2ce6aa70806b789 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -257,3 +257,12 @@ tasks.registerRunTask("runReobfPaperclip") {
     classpath(rootProject.tasks.named<io.papermc.paperweight.tasks.CreatePaperclipJar>("createReobfPaperclipJar").flatMap { it.outputZip })
     mainClass.set(null as String?)
 }
+
+// Kiterino start - Hide irrelevant compilation warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// Kiterino end - Hide irrelevant compilation warnings
