From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 17 May 2024 08:38:33 +0300
Subject: [PATCH] Add Kiterino configs


diff --git a/src/main/java/me/sosedik/kiterino/KiterinoConfig.java b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..6c77d9c7583ea95e0ad7ffba4aecccb3ccc56d29
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
@@ -0,0 +1,121 @@
+package me.sosedik.kiterino;
+
+import com.google.common.base.Throwables;
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.List;
+import java.util.logging.Level;
+
+@SuppressWarnings("unused")
+public final class KiterinoConfig {
+
+    private KiterinoConfig() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    private static final String HEADER = """
+            This is the main configuration file for Kiterino
+            
+            GitHub: https://github.com/SoSeDiKs-Universe/Kiterino
+            """;
+
+    private static File configFile;
+    public static YamlConfiguration config;
+    private static int version;
+    static boolean verbose;
+
+    public static void init(@NotNull File configFile) {
+        KiterinoConfig.configFile = configFile;
+        config = new YamlConfiguration();
+        try {
+            config.load(configFile);
+        } catch (IOException ignored) {
+        } catch (InvalidConfigurationException ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not load kiterino.yml, please correct your syntax errors", ex);
+            throw Throwables.propagate(ex);
+        }
+        config.options().header(HEADER);
+        config.options().copyDefaults(true);
+        verbose = getBoolean(config, "verbose", false);
+
+        version = getInt(config, "config-version", 1);
+        set(config, "config-version", 1);
+
+        readConfig(KiterinoConfig.class, null);
+    }
+
+    static void readConfig(@NotNull Class<?> clazz, @Nullable Object instance) {
+        readConfig(configFile, config, clazz, instance);
+    }
+
+    public static void readConfig(@NotNull File configFile, @NotNull YamlConfiguration config, @NotNull Class<?> clazz, @Nullable Object instance) {
+        for (Method method : clazz.getDeclaredMethods()) {
+            if (!Modifier.isPrivate(method.getModifiers())) continue;
+            if (method.getParameterTypes().length != 0) continue;
+            if (method.getReturnType() != Void.TYPE) continue;
+
+            try {
+                method.setAccessible(true);
+                method.invoke(instance);
+            } catch (InvocationTargetException ex) {
+                throw Throwables.propagate(ex.getCause());
+            } catch (Exception ex) {
+                Bukkit.getLogger().log(Level.SEVERE, "Error invoking " + method, ex);
+            }
+        }
+
+        try {
+            config.save(configFile);
+        } catch (IOException ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not save " + configFile, ex);
+        }
+    }
+
+    public static void log(@NotNull String s) {
+        if (verbose) {
+            log(Level.INFO, s);
+        }
+    }
+
+    public static void log(@NotNull Level level, @NotNull String s) {
+        Bukkit.getLogger().log(level, s);
+    }
+
+    public static void set(@NotNull YamlConfiguration config, @NotNull String path, @Nullable Object val, @NotNull String... comments) {
+        config.addDefault(path, val);
+        config.set(path, val);
+        if (comments.length > 0) config.setComments(path, List.of(comments));
+    }
+
+    public static boolean getBoolean(@NotNull YamlConfiguration config, @NotNull String path, boolean def, @NotNull String... comments) {
+        config.addDefault(path, def);
+        if (comments.length > 0) config.setComments(path, List.of(comments));
+        return config.getBoolean(path, def);
+    }
+
+    public static int getInt(@NotNull YamlConfiguration config, @NotNull String path, int def, @NotNull String... comments) {
+        config.addDefault(path, def);
+        if (comments.length > 0) config.setComments(path, List.of(comments));
+        return config.getInt(path, def);
+    }
+
+    public static <T> List<T> getList(@NotNull YamlConfiguration config, @NotNull String path, @Nullable List<T> def, @NotNull String... comments) {
+        config.addDefault(path, def);
+        if (comments.length > 0) config.setComments(path, List.of(comments));
+        return (List<T>) config.getList(path, def);
+    }
+
+    public static int getVersion() {
+        return version;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/KiterinoWorldConfig.java b/src/main/java/me/sosedik/kiterino/KiterinoWorldConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..0e430c0915012eb55ad9f656565499a9071b2064
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/KiterinoWorldConfig.java
@@ -0,0 +1,19 @@
+package me.sosedik.kiterino;
+
+import org.bukkit.World;
+import org.jetbrains.annotations.NotNull;
+
+public final class KiterinoWorldConfig extends WorldConfig {
+
+    public KiterinoWorldConfig(@NotNull String worldName, @NotNull World.Environment environment) {
+        super(KiterinoConfig.config, worldName, environment);
+        init();
+    }
+
+    @Override
+    public void init() {
+        log("-------- World Settings For [" + worldName + "] --------");
+        KiterinoConfig.readConfig(KiterinoWorldConfig.class, this);
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/WorldConfig.java b/src/main/java/me/sosedik/kiterino/WorldConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..589f413cb6d6b0ff8acb4ca7bb7a727d5f59a140
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/WorldConfig.java
@@ -0,0 +1,52 @@
+package me.sosedik.kiterino;
+
+import org.bukkit.World;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+public abstract class WorldConfig {
+
+    protected final YamlConfiguration config;
+    protected final String worldName;
+    protected final World.Environment environment;
+
+    protected WorldConfig(@NotNull YamlConfiguration config, @NotNull String worldName, @NotNull World.Environment environment) {
+        this.config = config;
+        this.worldName = worldName;
+        this.environment = environment;
+    }
+
+    public abstract void init();
+
+    public void log(@NotNull String s) {
+        KiterinoConfig.log(s);
+    }
+
+    public @Nullable String getString(@NotNull String path, @NotNull String def, @NotNull String... comments) {
+        config.addDefault("world-settings.default." + path, def);
+        if (comments.length > 0) config.setComments("world-settings.default." + path, List.of(comments));
+        return config.getString("world-settings." + worldName + "." + path, config.getString("world-settings.default." + path));
+    }
+
+    public boolean getBoolean(@NotNull String path, boolean def, @NotNull String... comments) {
+        config.addDefault("world-settings.default." + path, def);
+        if (comments.length > 0) config.setComments("world-settings.default." + path, List.of(comments));
+        return config.getBoolean("world-settings." + worldName + "." + path, config.getBoolean("world-settings.default." + path));
+    }
+
+    public double getDouble(@NotNull String path, double def, @NotNull String... comments) {
+        config.addDefault("world-settings.default." + path, def);
+        if (comments.length > 0) config.setComments("world-settings.default." + path, List.of(comments));
+        return config.getDouble("world-settings." + worldName + "." + path, config.getDouble("world-settings.default." + path));
+    }
+
+    public int getInt(@NotNull String path, int def, @NotNull String... comments) {
+        config.addDefault("world-settings.default." + path, def);
+        if (comments.length > 0) config.setComments("world-settings.default." + path, List.of(comments));
+        return config.getInt("world-settings." + worldName + "." + path, config.getInt("world-settings.default." + path));
+    }
+
+}
