From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 31 Jul 2021 20:15:05 +0300
Subject: [PATCH] Add NBT API as a first-class lib


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index c307eecba7319bc0c9acc2a636276cbc972c17e8..4ee376e023c621e98f19c6cd1015c00d453702d6 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -330,6 +330,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
         // Purpur end
 
+        // Kiterino start - Add NBT API as a first-class lib
+        de.tr7zw.changeme.nbtapi.utils.MinecraftVersion.disablePackageWarning();
+        de.tr7zw.changeme.nbtapi.utils.MinecraftVersion.disableUpdateCheck();
+        de.tr7zw.changeme.nbtapi.utils.MinecraftVersion.disableBStats(); // bStats need plugin instance
+        de.tr7zw.changeme.nbtapi.utils.MinecraftVersion.getVersion(); // Init
+        de.tr7zw.changeme.nbtapi.utils.nmsmappings.ReflectionMethod.COMPOUND_SET_FLOAT.getMethodName(); // Cache reflections
+        // Kiterino end - Add NBT API as a first-class lib
+
         // CraftBukkit start
         // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage)); // Spigot - moved up
         this.server.loadPlugins();
