From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 23:15:44 -0400
Subject: [PATCH] Parchment - Expose EntityType Ordinals


diff --git a/src/main/java/net/minecraft/world/entity/EntityType.java b/src/main/java/net/minecraft/world/entity/EntityType.java
index f22ceffdc95224f2de8d19f501b43b266de196d4..898df45a543e88b629072d80766c10b4bd78b7e6 100644
--- a/src/main/java/net/minecraft/world/entity/EntityType.java
+++ b/src/main/java/net/minecraft/world/entity/EntityType.java
@@ -161,6 +161,8 @@ import org.slf4j.Logger;
 
 public class EntityType<T extends Entity> implements FeatureElement, EntityTypeTest<Entity, T> {
 
+    private static int ordinal = 0; // Parchment
+    private static final java.util.Map<EntityType<?>, Integer> ordinals = new java.util.HashMap<>(); // Parchment
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final String ENTITY_TAG = "EntityTag";
     private final Holder.Reference<EntityType<?>> builtInRegistryHolder;
@@ -299,6 +301,7 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private String descriptionId;
     @Nullable
     private Component description;
+    private String registeredId; // Kiterino
     @Nullable
     private ResourceLocation lootTable;
     private EntityDimensions dimensions; // Purpur - remove final
@@ -306,9 +309,31 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private final FeatureFlagSet requiredFeatures;
 
     private static <T extends Entity> EntityType<T> register(String id, EntityType.Builder type) { // CraftBukkit - decompile error
-        return (EntityType) Registry.register(BuiltInRegistries.ENTITY_TYPE, id, (EntityType<T>) type.build(id)); // CraftBukkit - decompile error
+        // Parchment start
+        EntityType<T> build = type.build(id);
+        build.registeredId = id; // Kiterino - use registered id for comparing
+        ordinals.put(build, ordinal++);
+        return (EntityType) Registry.register(BuiltInRegistries.ENTITY_TYPE, id, (EntityType<T>) build); // CraftBukkit - decompile error
     }
 
+    public int ordinal() {
+        return ordinals.get(this);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        EntityType<?> that = (EntityType<?>) o;
+        return registeredId.equals(that.registeredId);
+    }
+
+    @Override
+    public int hashCode() {
+        return java.util.Objects.hash(registeredId);
+    }
+    // Parchment end
+
     // Purpur start
     public static EntityType<?> getFromBukkitType(org.bukkit.entity.EntityType bukkitType) {
         return getFromKey(new ResourceLocation(bukkitType.getKey().toString()));
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 22296596ab3ecba0d318135991cb296f96f595cf..a6c694752448cca846e6f2a89c801c9015f02d37 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -620,6 +620,19 @@ public final class CraftMagicNumbers implements UnsafeValues {
     }
     // Paper end
 
+    // Parchment start
+    @Override
+    public int entityID(org.bukkit.entity.EntityType type) {
+        String name = type.getName();
+        if (name == null) {
+            throw new IllegalArgumentException("Entity name is undefined");
+        }
+        return net.minecraft.world.entity.EntityType.byString(name)
+                .orElseThrow(() -> new IllegalArgumentException("Could not find corresponding NMS entity"))
+                .ordinal();
+    }
+    // Parchment end
+
     /**
      * This helper class represents the different NBT Tags.
      * <p>
