From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 25 Apr 2014 00:38:10 -0400
Subject: [PATCH] EMC - Cancel Vehicle/Passenger if spawn is cancelled

Otherwise lots of chickens

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index c47f6434d8c52d0524b0ce065ace5c45d589867f..238fdc847d9545c7f41ef426f64ccb16f99d7da2 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -747,6 +747,16 @@ public class CraftEventFactory {
 
         CreatureSpawnEvent event = new CreatureSpawnEvent(entity, spawnReason);
         craftServer.getPluginManager().callEvent(event);
+        // EMC start
+        if (event.isCancelled()) {
+            if (entity.getVehicle() != null) {
+                entity.getVehicle().remove();
+            }
+            if (entity.getPassenger() != null) {
+                entity.getPassenger().remove();
+            }
+        }
+        // EMC end
         return event;
     }
 
