From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 27 Jun 2022 19:17:06 +0300
Subject: [PATCH] Accept the EULA

We at SoSeDiK's Universe agree to Mojang's EULA

diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index a880f4e5cf712654649ad043e58e073e9a87c0fe..530fbaf81e124d61b048d4b45fcf3ae743de5d84 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -171,6 +171,21 @@ public class Main {
                 System.err.println( "If you do not agree to the above EULA please stop your server and remove this flag immediately." );
             }
             // Spigot End
+
+            // Kiterino start - Accept the EULA
+            if (!eula.hasAgreedToEULA()) {
+                Main.LOGGER.warn("Automatically accepting Mojang's EULA (https://aka.ms/MinecraftEULA).");
+                Main.LOGGER.warn("Please, stop using this Software immediately if you do not agree with this.");
+                Main.LOGGER.warn("Note: this is a persistent change and will be stored in eula.txt from now on.");
+                try (java.io.OutputStream outputStream = java.nio.file.Files.newOutputStream(path2)) {
+                    var properties = new java.util.Properties();
+                    properties.setProperty("eula", "true");
+                    properties.store(outputStream, null);
+                }
+                eulaAgreed = true;
+            }
+            // Kiterino end - Accept the EULA
+
             if (!eula.hasAgreedToEULA() && !eulaAgreed) { // Spigot
                 Main.LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
                 return;
