From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 23:15:44 -0400
Subject: [PATCH] Parchment - Expose EntityType Ordinals


diff --git a/src/main/java/net/minecraft/world/entity/EntityType.java b/src/main/java/net/minecraft/world/entity/EntityType.java
index a9f69e4f880711c21ef28bd079960ef8f12da25c..2eed75b3c0b1f083e8dd885c53c28b4a3e6416ce 100644
--- a/src/main/java/net/minecraft/world/entity/EntityType.java
+++ b/src/main/java/net/minecraft/world/entity/EntityType.java
@@ -163,6 +163,8 @@ import org.slf4j.Logger;
 
 public class EntityType<T extends Entity> implements FeatureElement, EntityTypeTest<Entity, T> {
 
+    private static int ordinal = 0; // Parchment
+    private static final java.util.Map<EntityType<?>, Integer> ordinals = new java.util.HashMap<>(); // Parchment
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final String ENTITY_TAG = "EntityTag";
     private final Holder.Reference<EntityType<?>> builtInRegistryHolder;
@@ -305,6 +307,7 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private String descriptionId;
     @Nullable
     private Component description;
+    private String registeredId; // Kiterino
     @Nullable
     private ResourceLocation lootTable;
     private EntityDimensions dimensions; // Purpur - remove final
@@ -312,9 +315,31 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private final FeatureFlagSet requiredFeatures;
 
     private static <T extends Entity> EntityType<T> register(String id, EntityType.Builder type) { // CraftBukkit - decompile error
-        return (EntityType) Registry.register(BuiltInRegistries.ENTITY_TYPE, id, (EntityType<T>) type.build(id)); // CraftBukkit - decompile error
+        // Parchment start
+        EntityType<T> build = type.build(id);
+        build.registeredId = id; // Kiterino - use registered id for comparing
+        ordinals.put(build, ordinal++);
+        return (EntityType) Registry.register(BuiltInRegistries.ENTITY_TYPE, id, (EntityType<T>) build); // CraftBukkit - decompile error
     }
 
+    public int ordinal() {
+        return ordinals.get(this);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        EntityType<?> that = (EntityType<?>) o;
+        return registeredId.equals(that.registeredId);
+    }
+
+    @Override
+    public int hashCode() {
+        return java.util.Objects.hash(registeredId);
+    }
+    // Parchment end
+
     // Purpur start
     public static EntityType<?> getFromBukkitType(org.bukkit.entity.EntityType bukkitType) {
         return getFromKey(new ResourceLocation(bukkitType.getKey().toString()));
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 5ecd3bd49c1c36a5b234d43803d1d4009d7f34fc..00089c1bccc74b547e5b5382a7e6e155d416a7f9 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -624,6 +624,19 @@ public final class CraftMagicNumbers implements UnsafeValues {
     }
     // Paper end
 
+    // Parchment start
+    @Override
+    public int entityID(org.bukkit.entity.EntityType type) {
+        String name = type.getName();
+        if (name == null) {
+            throw new IllegalArgumentException("Entity name is undefined");
+        }
+        return net.minecraft.world.entity.EntityType.byString(name)
+                .orElseThrow(() -> new IllegalArgumentException("Could not find corresponding NMS entity"))
+                .ordinal();
+    }
+    // Parchment end
+
     /**
      * This helper class represents the different NBT Tags.
      * <p>
