From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 14 Oct 2022 09:41:48 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c3d25a06e95cde5a88d70c258cea02e838f4897a..241cdc86f3a7c5c8502a904591d699f79e646e86 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3391,6 +3391,21 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
+    // Kiterino start - Allow setting entity's rider
+    public boolean setRider(Player rider) {
+        if (!new org.bukkit.event.entity.EntityMountEvent(rider.getBukkitEntity(), this.getBukkitEntity()).callEvent()) return false;
+        if (!rider.startRiding(this, true)) return false;
+
+        onMount(rider);
+        this.rider = rider;
+        return true;
+    }
+
+    public boolean hasRider() {
+        return getRider() != null;
+    }
+    // Kiterino end - Allow setting entity's rider
+
     // Paper start - Force entity dismount during teleportation
     protected boolean removePassenger(Entity entity) { return removePassenger(entity, false);}
     protected boolean removePassenger(Entity entity, boolean suppressCancellation) { // CraftBukkit
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index d7668cdaa2718968e4b9c5e487a6752c5cef4836..18a4b993a7f5f895b1f9d768abd84578304f1e10 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1359,4 +1359,18 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         getHandle().forcePanicTicks = Math.max(0, ticks);
     }
     // Kiterino end - Forced panic mode API
+
+    // Kiterino start - Allow setting entity's rider
+    @Override
+    public boolean setRider(Player rider, boolean dismount) {
+        Player currentRider = getRider();
+        if (rider == currentRider) {
+            return true;
+        }
+        if (currentRider != null && (!dismount || !currentRider.leaveVehicle())) {
+            return false;
+        }
+        return getHandle().setRider(((CraftPlayer) rider).getHandle());
+    }
+    // Kiterino end - Allow setting entity's rider
 }
