From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yannick Lamprecht <yannicklamprecht@live.de>
Date: Tue, 12 Oct 2021 21:10:47 +0200
Subject: [PATCH] Weeper - Implement take and place


diff --git a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
index 31c9bb73b73be99b95a7dde55d276964522a6d7d..bc7fd6480e784c43e236bfbbe7a60d8eb204c127 100644
--- a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.inventory;
 
+import com.github.weepingmc.inventory.InventoryItemStackMoveStrategy;
 import com.google.common.base.Suppliers;
 import com.google.common.collect.HashBasedTable;
 import com.google.common.collect.Lists;
@@ -157,6 +158,12 @@ public abstract class AbstractContainerMenu {
         return slot;
     }
 
+    // Weeper start - inventoryItemStackMoveStrategy
+    public void setItemStackMoveStrategy(InventoryItemStackMoveStrategy inventoryItemStackMoveStrategy){
+        this.slots.forEach(slot -> slot.setInventoryItemstackMoveStrategy(inventoryItemStackMoveStrategy));
+    }
+    // Weeper end - inventoryItemStackMoveStrategy
+
     protected DataSlot addDataSlot(DataSlot property) {
         this.dataSlots.add(property);
         this.remoteDataSlots.add(0);
diff --git a/src/main/java/net/minecraft/world/inventory/Slot.java b/src/main/java/net/minecraft/world/inventory/Slot.java
index 52d555af415277d8d1f30d8b46855d4b05867597..c3bc0c196e771a10bb990a16ebcc0e7ee2da2761 100644
--- a/src/main/java/net/minecraft/world/inventory/Slot.java
+++ b/src/main/java/net/minecraft/world/inventory/Slot.java
@@ -1,5 +1,7 @@
 package net.minecraft.world.inventory;
 
+import com.github.weepingmc.inventory.InventoryItemStackMoveStrategy;
+import com.github.weepingmc.inventory.SlotItemStackMovementStrategyItem;
 import com.mojang.datafixers.util.Pair;
 import java.util.Optional;
 import javax.annotation.Nullable;
@@ -7,6 +9,7 @@ import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.Container;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 
 public class Slot {
     public final int slot;
@@ -14,6 +17,13 @@ public class Slot {
     public int index;
     public final int x;
     public final int y;
+    private @Nullable InventoryItemStackMoveStrategy inventoryItemStackMoveStrategy;
+
+    // Weeper start - inventoryItemStackMoveStrategy
+    public void setInventoryItemstackMoveStrategy(@Nullable InventoryItemStackMoveStrategy inventoryItemStackMoveStrategy){
+        this.inventoryItemStackMoveStrategy = inventoryItemStackMoveStrategy;
+    }
+    // Weeper end - inventoryItemStackMoveStrategy
 
     public Slot(Container inventory, int index, int x, int y) {
         this.container = inventory;
@@ -44,7 +54,7 @@ public class Slot {
     }
 
     public boolean mayPlace(ItemStack stack) {
-        return true;
+        return this.inventoryItemStackMoveStrategy == null || inventoryItemStackMoveStrategy.canPlace(new SlotItemStackMovementStrategyItem(container.getOwner(), CraftItemStack.asBukkitCopy(stack), slot));
     }
 
     public ItemStack getItem() {
@@ -87,7 +97,7 @@ public class Slot {
     }
 
     public boolean mayPickup(Player playerEntity) {
-        return true;
+        return this.inventoryItemStackMoveStrategy == null || inventoryItemStackMoveStrategy.canTake(new SlotItemStackMovementStrategyItem(container.getOwner(), CraftItemStack.asBukkitCopy(this.getItem()), slot));
     }
 
     public boolean isActive() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index dabbb41c9a7b0a825a241143ae7c196fceff3d60..e9ca6f7c521bb2966eebdb520fece12fa0c94ede 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit.entity;
 
+import com.github.weepingmc.inventory.InventoryItemStackMoveStrategy;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableSet;
 import java.util.ArrayList;
@@ -46,13 +47,13 @@ import org.bukkit.craftbukkit.inventory.CraftInventoryPlayer;
 import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.craftbukkit.inventory.CraftMerchantCustom;
-import org.bukkit.craftbukkit.util.CraftChatMessage;
 import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 import org.bukkit.craftbukkit.util.CraftNamespacedKey;
 import org.bukkit.entity.HumanEntity;
 import org.bukkit.entity.Villager;
 import org.bukkit.inventory.EntityEquipment;
 import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.InventoryHolder;
 import org.bukkit.inventory.InventoryView;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.MainHand;
@@ -63,6 +64,7 @@ import org.bukkit.permissions.Permission;
 import org.bukkit.permissions.PermissionAttachment;
 import org.bukkit.permissions.PermissionAttachmentInfo;
 import org.bukkit.plugin.Plugin;
+import org.checkerframework.checker.nullness.qual.Nullable;
 
 public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     private CraftInventoryPlayer inventory;
@@ -305,6 +307,11 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
 
     @Override
     public InventoryView openInventory(Inventory inventory) {
+        return openInventory(inventory, (InventoryItemStackMoveStrategy) null);
+    }
+
+    @Override
+    public InventoryView openInventory(Inventory inventory, InventoryItemStackMoveStrategy inventoryItemStackMoveStrategy) {
         if (!(this.getHandle() instanceof ServerPlayer)) return null;
         ServerPlayer player = (ServerPlayer) this.getHandle();
         AbstractContainerMenu formerContainer = this.getHandle().containerMenu;
@@ -334,7 +341,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         if (iinventory instanceof MenuProvider) {
             this.getHandle().openMenu(iinventory);
         } else {
-            CraftHumanEntity.openCustomInventory(inventory, player, container);
+            CraftHumanEntity.openCustomInventory(inventory, player, container, inventoryItemStackMoveStrategy);
         }
 
         if (this.getHandle().containerMenu == formerContainer) {
@@ -345,9 +352,20 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     }
 
     private static void openCustomInventory(Inventory inventory, ServerPlayer player, MenuType<?> windowType) {
+    // Weeper start - InventoryItemStackMoveStrategy
+        openCustomInventory( inventory, player, windowType, null);
+    }
+
+    private static void openCustomInventory(Inventory inventory, ServerPlayer player, MenuType<?> windowType, @Nullable InventoryItemStackMoveStrategy InventoryItemStackMoveStrategy) {
+    // Weeper end - InventoryItemStackMoveStrategy
         if (player.connection == null) return;
         Preconditions.checkArgument(windowType != null, "Unknown windowType");
         AbstractContainerMenu container = new CraftContainer(inventory, player, player.nextContainerCounter());
+        // Weeper start - InventoryItemStackMoveStrategy
+        if(InventoryItemStackMoveStrategy != null){
+            container.setItemStackMoveStrategy(InventoryItemStackMoveStrategy);
+        }
+        // Weeper end - InventoryItemStackMoveStrategy
 
         container = CraftEventFactory.callInventoryOpenEvent(player, container);
         if (container == null) return;
