From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yannick Lamprecht <yannicklamprecht@live.de>
Date: Sat, 19 Jun 2021 22:31:23 +0200
Subject: [PATCH] Weeper - Add mechanix to generate comments to vanilla mob
 goals


diff --git a/src/test/java/com/destroystokyo/paper/entity/ai/VanillaMobGoalTest.java b/src/test/java/com/destroystokyo/paper/entity/ai/VanillaMobGoalTest.java
index 6453b1d4182a47367e89500ed4180ce75556b47d..6c632e1bb56094bebe43593544a40cc6782d01b7 100644
--- a/src/test/java/com/destroystokyo/paper/entity/ai/VanillaMobGoalTest.java
+++ b/src/test/java/com/destroystokyo/paper/entity/ai/VanillaMobGoalTest.java
@@ -57,6 +57,11 @@ public class VanillaMobGoalTest {
         if (missingFromAPI.size() != 0) {
             System.out.println("Missing from API: ");
             for (GoalKey<?> key : missingFromAPI) {
+                System.out.printf("""
+                        /**
+                        * MobGoal %s applicable to any entity that is instance of %s
+                        */
+                        """, key.getNamespacedKey().getKey(), key.getEntityClass().getSimpleName());
                 System.out.println("GoalKey<" + key.getEntityClass().getSimpleName() + "> " + key.getNamespacedKey().getKey().toUpperCase() +
                     " = GoalKey.of(" + key.getEntityClass().getSimpleName() + ".class, NamespacedKey.minecraft(\"" + key.getNamespacedKey().getKey() + "\"));");
             }
