From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Thu, 8 Jul 2021 21:17:22 +0300
Subject: [PATCH] Origami - Keep weather clear after sleep


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 69ba84de4880a579c400b6daefdb98b44e9e28e8..7ff7a436dd7526a89ff00c44a7b2bbedb358f8c0 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1184,6 +1184,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
 
     private void resetWeatherCycle() {
+        this.serverLevelData.setClearWeatherTime(6000 + this.randomTickRandom.nextInt(24000)); // Origami - keep weather clear for at least 5 minutes after sleep
         // CraftBukkit start
         if (this.purpurConfig.rainStopsAfterSleep) // Purpur
         this.serverLevelData.setRaining(false, org.bukkit.event.weather.WeatherChangeEvent.Cause.SLEEP); // Paper - when passing the night
