From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Thu, 8 Jul 2021 21:17:22 +0300
Subject: [PATCH] Origami - Keep weather clear after sleep


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index c888e654ab9449bfdc7dfe16078eb0786ae6c15e..61f00f57ebdbd28b2710ab26e5cbfd3bd79cd444 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1180,6 +1180,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
 
     private void resetWeatherCycle() {
+        this.serverLevelData.setClearWeatherTime(6000 + this.randomTickRandom.nextInt(24000)); // Origami - keep weather clear for at least 5 minutes after sleep
         // CraftBukkit start
         if (this.purpurConfig.rainStopsAfterSleep) // Purpur
         this.serverLevelData.setRaining(false, org.bukkit.event.weather.WeatherChangeEvent.Cause.SLEEP); // Paper - when passing the night
