From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Thu, 8 Jul 2021 21:17:22 +0300
Subject: [PATCH] Origami - Keep weather clear after sleep


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index b396a329e753f42b3878736b6b14bafda02d3744..19f3598ed1786a2b6ac58e509f8fe7746b81b5b9 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1201,6 +1201,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
 
     private void resetWeatherCycle() {
+        this.serverLevelData.setClearWeatherTime(6000 + this.randomTickRandom.nextInt(24000)); // Origami - keep weather clear for at least 5 minutes after sleep
         // CraftBukkit start
         if (this.purpurConfig.rainStopsAfterSleep) // Purpur
         this.serverLevelData.setRaining(false, org.bukkit.event.weather.WeatherChangeEvent.Cause.SLEEP); // Paper - when passing the night
