From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Thu, 8 Jul 2021 21:17:22 +0300
Subject: [PATCH] Origami - Keep weather clear after sleep


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index c50dc34a731477c76315d8ef6b59b8539b76c7e3..3c01361e731e48674c2bc478c98ba2f9dccd1961 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1143,6 +1143,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
 
     private void resetWeatherCycle() {
+        this.serverLevelData.setClearWeatherTime(6000 + this.randomTickRandom.nextInt(24000)); // Origami - keep weather clear for at least 5 minutes after sleep
         // CraftBukkit start
         if (this.purpurConfig.rainStopsAfterSleep) // Purpur
         this.serverLevelData.setRaining(false, org.bukkit.event.weather.WeatherChangeEvent.Cause.SLEEP); // Paper - when passing the night
