From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 16 May 2024 11:37:52 +0300
Subject: [PATCH] Parse items in show_item hover event


diff --git a/src/main/java/me/sosedik/kiterino/KiterinoConfig.java b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
index 25eafe3b3cce238ebd556ac894bb59c549c27e48..48f6ee0a3ac1f3276d90229aafe42e162f757f04 100644
--- a/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
+++ b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
@@ -129,6 +129,13 @@ public final class KiterinoConfig {
     }
     // Kiterino end - Item Modifiers API
 
+    // Kiterino start - Parse items in show_item hover event
+    public static boolean parseItemHoversEverywhere;
+    private static void parseItemHoversEverywhere() {
+        parseItemHoversEverywhere = getBoolean(config, "item-modifiers.apply-modifiers-on-all-hover", false, "Apply item modifiers on all show_item hover events in text components.", "By default, they are applied only to system chat messages.", "Only enable if your client is getting kicked due to injected items.");
+    }
+    // Kiterino end - Parse items in show_item hover event
+
     // Kiterino start - Prevent creative from overriding items
     public static boolean preventCreativeItemOverride;
     private static void preventCreativeItemsOverride() {
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
index c6d52745ec4b99191c8bd366d8360316eca9faa2..ad238f0e17c70bcc0c637ff03ae4a7140ef29dc7 100644
--- a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
@@ -22,6 +22,8 @@ import net.minecraft.advancements.AdvancementHolder;
 import net.minecraft.advancements.DisplayInfo;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.ComponentSerialization;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientboundBundlePacket;
 import net.minecraft.network.protocol.game.ClientboundContainerSetContentPacket;
@@ -33,6 +35,7 @@ import net.minecraft.network.protocol.game.ClientboundSetCursorItemPacket;
 import net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket;
 import net.minecraft.network.protocol.game.ClientboundSetEquipmentPacket;
 import net.minecraft.network.protocol.game.ClientboundSetPlayerInventoryPacket;
+import net.minecraft.network.protocol.game.ClientboundSystemChatPacket;
 import net.minecraft.network.protocol.game.ClientboundUpdateAdvancementsPacket;
 import net.minecraft.network.protocol.game.ClientboundUpdateRecipesPacket;
 import net.minecraft.network.syncher.EntityDataSerializers;
@@ -238,10 +241,24 @@ public class ItemModifiersHandlerImpl extends ItemModifiersHandler {
 			case ClientboundRecipeBookAddPacket packet -> handle(player, packet);
 			case ClientboundUpdateAdvancementsPacket packet -> handle(player, packet);
 			case ClientboundMerchantOffersPacket packet -> handle(player, packet);
+			case ClientboundSystemChatPacket packet -> handle(player, packet);
 			default -> initialPacket;
 		};
 	}
 
+	private static @NotNull Packet<?> handle(@NotNull CraftPlayer player, @NotNull ClientboundSystemChatPacket initialPacket) {
+		if (initialPacket.overlay()) {
+			return initialPacket;
+		}
+
+		boolean prev = ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
+		ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(true);
+		Component component = initialPacket.content();
+		component = ComponentSerialization.replaceHoverItems(player, player.locale(), component);
+		ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
+		return new ClientboundSystemChatPacket(component, false);
+	}
+
 	@SuppressWarnings({"unchecked"})
 	private static @NotNull Packet<?> handle(@NotNull CraftPlayer player, @NotNull ClientboundBundlePacket initialPacket) {
 		if (!(initialPacket.subPackets() instanceof ArrayList list)) {
diff --git a/src/main/java/net/minecraft/network/chat/ComponentSerialization.java b/src/main/java/net/minecraft/network/chat/ComponentSerialization.java
index 22c6a324dedeb315eac2d3d3f55d2f3a9eebb0ad..a820af5033d69d3b130569e33454b89330d3258b 100644
--- a/src/main/java/net/minecraft/network/chat/ComponentSerialization.java
+++ b/src/main/java/net/minecraft/network/chat/ComponentSerialization.java
@@ -54,13 +54,88 @@ public class ComponentSerialization {
 
             @Override
             public void encode(RegistryFriendlyByteBuf registryFriendlyByteBuf, Component object) {
+                // Kiterino start - Parse items in show_item hover event
+                if (me.sosedik.kiterino.KiterinoConfig.parseItemHoversEverywhere && !DONT_RENDER_TRANSLATABLES.get()) {
+                    java.util.Locale locale = registryFriendlyByteBuf.adventure$locale;
+                    if (locale == null) {
+                        locale = java.util.Locale.ENGLISH;
+                    }
+                    object = replaceHoverItems(null, locale, object);
+                }
+                Component finalObject = object;
+                // Kiterino end - Parse items in show_item hover event
                 RegistryOps<net.minecraft.nbt.Tag> registryOps = registryFriendlyByteBuf.registryAccess().createSerializationContext(net.minecraft.nbt.NbtOps.INSTANCE);
                 net.minecraft.nbt.Tag tag = (DONT_RENDER_TRANSLATABLES.get() ? CODEC : ComponentSerialization.localizedCodec(registryFriendlyByteBuf.adventure$locale))
-                    .encodeStart(registryOps, object).getOrThrow(error -> new io.netty.handler.codec.EncoderException("Failed to encode: " + error + " " + object));
+                    .encodeStart(registryOps, object).getOrThrow(error -> new io.netty.handler.codec.EncoderException("Failed to encode: " + error + " " + finalObject)); // Kiterino - Parse items in show_item hover event
                 this.streamCodec.encode(registryFriendlyByteBuf, tag);
             }
         };
     }
+    // Kiterino start - Parse items in show_item hover event
+    public static Component replaceHoverItems(@org.jetbrains.annotations.Nullable org.bukkit.craftbukkit.entity.CraftPlayer player, java.util.Locale locale, @org.jetbrains.annotations.Nullable Component component) {
+        if (component == null) {
+            return component;
+        }
+        // Go through children first
+        List<Component> children = component.getSiblings();
+        if (!children.isEmpty()) {
+            // Adventure components have immutable children/siblings
+            if (!(component instanceof MutableComponent)) {
+                component = component.copy();
+                children = component.getSiblings();
+            }
+            try {
+                children.replaceAll(child -> replaceHoverItems(player, locale, child));
+                // Mutable component with immutable children?
+            } catch (UnsupportedOperationException ignored1) {
+                component = component.copy();
+                children = component.getSiblings();
+                try {
+                    children.replaceAll(child -> replaceHoverItems(player, locale, child));
+                    // Shouldn't happen, but just in case
+                } catch (UnsupportedOperationException ignored2) {
+                    return component;
+                }
+            }
+        }
+
+        // Translatables may have component arguments
+        if (component.getContents() instanceof TranslatableContents translatable) {
+            Object[] args = translatable.getArgs();
+            for (int i = 0; i < args.length; i++) {
+                if (!(args[i] instanceof Component arg)) continue;
+
+                args[i] = replaceHoverItems(player, locale, arg);
+            }
+        }
+
+        // Finally replace hover item if found
+        HoverEvent hoverEvent = component.getStyle().hoverEvent;
+        if (hoverEvent == null) {
+            return component;
+        }
+
+        HoverEvent.ItemStackInfo value = hoverEvent.getValue(HoverEvent.Action.SHOW_ITEM);
+        if (value == null) {
+            return component;
+        }
+
+        var original = value.getItemStack();
+        var context = me.sosedik.kiterino.modifier.item.context.ItemModifierContext.EMPTY;
+        var contextBox = new me.sosedik.kiterino.modifier.item.ItemContextBox(player, locale, me.sosedik.kiterino.modifier.item.context.ItemModifierContextType.CHAT_ITEM, context, original.asBukkitCopy());
+        var bukkitItem = me.sosedik.kiterino.modifier.item.ItemModifier.modifyItem(contextBox);
+        if (bukkitItem != null) {
+            Style style = component.getStyle().withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_ITEM, new HoverEvent.ItemStackInfo(net.minecraft.world.item.ItemStack.fromBukkitCopy(bukkitItem))));
+            if (component instanceof MutableComponent mutableComponent) {
+                mutableComponent.setStyle(style);
+            } else {
+                component = component.copy().setStyle(style);
+            }
+        }
+
+        return component;
+    }
+    // Kiterino end - Parse items in show_item hover event
     // Paper end - adventure; use locale from bytebuf for translation
     public static final StreamCodec<RegistryFriendlyByteBuf, Optional<Component>> TRUSTED_OPTIONAL_STREAM_CODEC = TRUSTED_STREAM_CODEC.apply(
         ByteBufCodecs::optional
