From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 16 May 2024 03:42:33 +0300
Subject: [PATCH] Real custom items


diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorImpl.java b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..5682ec993c63d9adaccc005fad0b02bbd06ec496
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorImpl.java
@@ -0,0 +1,216 @@
+package me.sosedik.kiterino.util;
+
+import com.destroystokyo.paper.event.executor.asm.SafeClassDefiner;
+import io.papermc.paper.adventure.PaperAdventure;
+import me.sosedik.kiterino.modifier.item.ItemContextBox;
+import me.sosedik.kiterino.world.item.KiterinoItem;
+import me.sosedik.kiterino.world.item.KiterinoItemProperties;
+import me.sosedik.kiterino.world.item.KiterinoNMSItem;
+import net.minecraft.core.component.DataComponents;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.Items;
+import org.bukkit.Material;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.FieldVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static me.sosedik.kiterino.util.KiterinoUnsafeUtil.getField;
+
+public class KiterinoDataInjectorImpl extends KiterinoDataInjector {
+
+	private KiterinoDataInjectorImpl() {}
+
+	private final Field itemMaterialField = getField(CraftMagicNumbers.class, "ITEM_MATERIAL");
+	private final Field materialItemField = getField(CraftMagicNumbers.class, "MATERIAL_ITEM");
+	private final Field maxStackField = getField(Material.class, "maxStack");
+	private final Field durabilityField = getField(Material.class, "durability");
+
+	public void injectItem(@NotNull Material material, @NotNull KiterinoNMSItem item) {
+		ResourceLocation id = PaperAdventure.asVanilla(material.getKey());
+		try {
+			if (item instanceof Item nmsItem) {
+				registerNMSItem(material, id, nmsItem);
+				applyExtraItemProperties(nmsItem, item.getItemProperties());
+			} else if (item instanceof KiterinoItem kiterinoItem) {
+				Item nmsItem = createNmsItemClass(id, kiterinoItem);
+				registerNMSItem(material, id, nmsItem);
+				applyExtraItemProperties(nmsItem, kiterinoItem.getItemProperties());
+			} else {
+				throw new KiterinoDataInjectorRuntimeException("Unknown Kiterino item: " + id);
+			}
+		} catch (Exception e) {
+			throw new KiterinoDataInjectorRuntimeException("Couldn't inject item with id " + id, e);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private void registerNMSItem(@NotNull Material material, @NotNull ResourceLocation id, @NotNull Item nmsItem) throws IllegalArgumentException, IllegalAccessException {
+		Items.registerItem(id, nmsItem);
+		applyBukkitData(material, nmsItem);
+
+		((Map<Item, Material>) itemMaterialField.get(null)).put(nmsItem, material);
+		((Map<Material, Item>) materialItemField.get(null)).put(material, nmsItem);
+	}
+
+	@SuppressWarnings("java:S3011")
+	private void applyBukkitData(@NotNull Material material, @NotNull Item item) throws IllegalArgumentException, IllegalAccessException {
+		maxStackField.set(material, item.getDefaultMaxStackSize());
+		durabilityField.set(material, (short) (int) item.components().getOrDefault(DataComponents.MAX_DAMAGE, 0));
+	}
+
+	private void applyExtraItemProperties(@NotNull Item nmsItem, @Nullable KiterinoItemProperties properties) {
+		if (properties == null) return;
+
+	}
+
+	private @NotNull Item createNmsItemClass(@NotNull ResourceLocation id, @NotNull KiterinoNMSItem item) throws Exception {
+		var classWriter = new ClassWriter(0);
+
+		String classInternalName = generateClassName(id, "Item");
+		String parentClass = Type.getInternalName(Item.class);
+
+		// Define class
+		List<Class<?>> interfaces = new ArrayList<>();
+		interfaces.add(KiterinoNMSItem.class);
+		classWriter.visit(Opcodes.V21, Opcodes.ACC_PUBLIC + Opcodes.ACC_SUPER, classInternalName, null, parentClass, interfaces.stream().map(Type::getInternalName).toArray(String[]::new));
+
+		// Add "private final wrapper KiterinoNMSItem" field
+		FieldVisitor fieldVisitor = classWriter.visitField(Opcodes.ACC_PRIVATE + Opcodes.ACC_FINAL, "wrapper", Type.getDescriptor(KiterinoNMSItem.class), null, null);
+		fieldVisitor.visitEnd();
+
+		// Create the default constructor
+		String constDesc = Type.getMethodDescriptor(Type.VOID_TYPE, Type.getType(KiterinoNMSItem.class), Type.getType(Item.Properties.class));
+		MethodVisitor constructor = classWriter.visitMethod(Opcodes.ACC_PUBLIC, "<init>", constDesc, null, null);
+		constructor.visitCode();
+		constructor.visitVarInsn(Opcodes.ALOAD, 0);
+		constructor.visitVarInsn(Opcodes.ALOAD, 2);
+		constructor.visitMethodInsn(Opcodes.INVOKESPECIAL, parentClass, "<init>", Type.getConstructorDescriptor(Item.class.getDeclaredConstructor(Item.Properties.class)), false);
+		constructor.visitVarInsn(Opcodes.ALOAD, 0);
+		constructor.visitVarInsn(Opcodes.ALOAD, 1);
+		constructor.visitFieldInsn(Opcodes.PUTFIELD, classInternalName, "wrapper", Type.getDescriptor(KiterinoNMSItem.class));
+		constructor.visitInsn(Opcodes.RETURN);
+		constructor.visitMaxs(2, 3);
+		constructor.visitEnd();
+
+		// Implement interface methods
+		for (Class<?> interf : interfaces) {
+			switch (interf.getSimpleName()) {
+				case "KiterinoNMSItem" ->
+					implementMethod(classWriter, KiterinoNMSItem.class.getDeclaredMethod("modify", ItemContextBox.class), (method, visitor) -> {
+						visitor.visitVarInsn(Opcodes.ALOAD, 0);
+						visitor.visitFieldInsn(Opcodes.GETFIELD, classInternalName, "wrapper", Type.getDescriptor(KiterinoNMSItem.class));
+						visitor.visitVarInsn(Opcodes.ALOAD, 1);
+						visitor.visitMethodInsn(Opcodes.INVOKEINTERFACE, Type.getInternalName(KiterinoNMSItem.class), method.getName(), Type.getMethodDescriptor(method), true);
+						visitor.visitInsn(Opcodes.RETURN); // Return void
+						visitor.visitMaxs(2, 2);
+					});
+			}
+		}
+
+		classWriter.visitEnd();
+
+		byte[] classData = classWriter.toByteArray();
+
+		Class<?> nmsItemClass = SafeClassDefiner.INSTANCE.defineClass(this.getClass().getClassLoader(), classInternalName.replace('/', '.'), classData);
+		return (Item) nmsItemClass.getDeclaredConstructor(KiterinoNMSItem.class, Item.Properties.class).newInstance(item, getFromBukkit(item.getItemProperties()));
+	}
+
+	@Override
+	protected @NotNull KiterinoNMSItem modifyNmsItemClass0(@NotNull Class<? extends KiterinoNMSItem> itemClass, @Nullable Object @NotNull ... constructorArgs) {
+		try {
+			if (itemClass.getInterfaces().length > 1) {
+				List<Class<?>> extenders = new ArrayList<>();
+				if (!extenders.isEmpty()) {
+					var classReader = new ClassReader(Type.getInternalName(itemClass));
+					var classWriter = new ClassWriter(ClassReader.EXPAND_FRAMES);
+					classReader.accept(new ClassExtender(classWriter, extenders), 0);
+					byte[] modifiedClassData = classWriter.toByteArray();
+					itemClass = (Class<? extends KiterinoNMSItem>) SafeClassDefiner.INSTANCE.defineClass(this.getClass().getClassLoader(), Type.getInternalName(itemClass).replace('/', '.'), modifiedClassData);
+				}
+			}
+
+			return (KiterinoNMSItem) itemClass.getDeclaredConstructors()[0].newInstance(constructorArgs);
+		} catch (Exception e) {
+			throw new KiterinoDataInjectorRuntimeException("Couldn't create nms item " + itemClass.getName(), e);
+		}
+	}
+
+	private @NotNull String generateClassName(@NotNull ResourceLocation id, @NotNull String suffix) {
+		var classNameB = new StringBuilder("me/sosedik/kiterino/");
+		for (String word : id.toString().replace(":", "_").split("_")) {
+			if (!word.isEmpty()) {
+				classNameB.append(Character.toUpperCase(word.charAt(0))).append(word.substring(1));
+			}
+		}
+		if (!classNameB.toString().endsWith(suffix)) {
+			classNameB.append(suffix);
+		}
+		return classNameB.toString();
+	}
+
+	private void implementMethod(@NotNull ClassWriter classWriter, @NotNull Method method, @NotNull BiConsumer<Method, MethodVisitor> impl) {
+		MethodVisitor methodVisitor = classWriter.visitMethod(Opcodes.ACC_PUBLIC, method.getName(), Type.getMethodDescriptor(method), null, null);
+		methodVisitor.visitCode();
+		impl.accept(method, methodVisitor);
+		methodVisitor.visitEnd();
+	}
+
+	private static class ClassExtender extends ClassVisitor {
+
+		private final List<Class<?>> extenders;
+
+		public ClassExtender(@NotNull ClassVisitor cv, @NotNull List<Class<?>> extenders) {
+			super(Opcodes.ASM9, cv);
+			this.extenders = extenders;
+		}
+
+		@Override
+		public void visitEnd() {
+			for (Class<?> extender : extenders) {
+				switch (extender.getSimpleName()) {
+				}
+			}
+			cv.visitEnd();
+		}
+
+	}
+
+	public static @NotNull Item.Properties getFromBukkit(@Nullable KiterinoItemProperties properties) {
+		var settings = new Item.Properties();
+		if (properties == null) return settings;
+
+		settings.stacksTo(properties.getMaxStackSize());
+		if (properties.isFireResistant()) settings.fireResistant();
+
+		return settings;
+	}
+
+	public static @UnknownNullability KiterinoDataInjectorImpl getInstance() {
+		return (KiterinoDataInjectorImpl) KiterinoDataInjector.injector;
+	}
+
+	public static void init() {
+		KiterinoDataInjector.injector = new KiterinoDataInjectorImpl();
+	}
+
+	public static void destruct() {
+		KiterinoDataInjector.injector = null;
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorRuntimeException.java b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorRuntimeException.java
new file mode 100644
index 0000000000000000000000000000000000000000..3e7018eb4112dfc31621dc436dd1bd64eb56ca34
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorRuntimeException.java
@@ -0,0 +1,16 @@
+package me.sosedik.kiterino.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class KiterinoDataInjectorRuntimeException extends RuntimeException {
+
+	public KiterinoDataInjectorRuntimeException(@NotNull String message) {
+		super(message);
+	}
+
+	public KiterinoDataInjectorRuntimeException(@NotNull String message, @Nullable Throwable cause) {
+		super(message, cause);
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoUnsafeUtil.java b/src/main/java/me/sosedik/kiterino/util/KiterinoUnsafeUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..f56f500cd78e042bc1359a797825c7c680bf9284
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoUnsafeUtil.java
@@ -0,0 +1,52 @@
+package me.sosedik.kiterino.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import sun.misc.Unsafe;
+
+import java.lang.reflect.Field;
+
+final class KiterinoUnsafeUtil {
+
+	private KiterinoUnsafeUtil() {
+		throw new IllegalStateException("Utility class");
+	}
+
+	private static Unsafe unsafe;
+
+	static {
+		try {
+			Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
+			unsafeField.setAccessible(true);
+			unsafe = (Unsafe) unsafeField.get(null);
+		} catch(Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	static @NotNull Field getField(@NotNull Class<?> clazz, @NotNull String fieldName) {
+		try {
+			Field field = clazz.getDeclaredField(fieldName);
+			field.setAccessible(true);
+			return field;
+		} catch (NoSuchFieldException e) {
+			throw new KiterinoDataInjectorRuntimeException("Couldn't get field", e);
+		}
+	}
+
+	static void setField(@NotNull Field field, @Nullable Object value) {
+		Object fieldBase = unsafe.staticFieldBase(field);
+		long fieldOffset = unsafe.staticFieldOffset(field);
+
+		unsafe.putObject(fieldBase, fieldOffset, value);
+	}
+
+	static <T> T allocateInstance(@NotNull Class<T> tClass) {
+		try {
+			return (T) unsafe.allocateInstance(tClass);
+		} catch (Exception e) {
+			throw new KiterinoDataInjectorRuntimeException("Couldn't allocate new instance", e);
+		}
+	}
+
+}
