From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 11 Dec 2021 12:18:42 -0500
Subject: [PATCH] Paper PR - Add Movement Direction API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index ce333ed345420d11102776cab36462adff5f2b8c..257156505d76b9577d19ad2a2441a4f18c66fb9d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -979,6 +979,23 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().hurtDir = hurtDirection;
     }
 
+    // Paper start - Movement Direction API
+    @Override
+    public float getSidewaysMotion() {
+        return this.getHandle().xxa;
+    }
+
+    @Override
+    public float getUpwardsMotion() {
+        return this.getHandle().yya;
+    }
+
+    @Override
+    public float getForwardMotion() {
+        return this.getHandle().zza;
+    }
+    // Paper end - Movement Direction API
+
     public static MobType fromBukkitEntityCategory(EntityCategory entityCategory) {
         switch (entityCategory) {
             case NONE:
