From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 11 Dec 2021 12:18:42 -0500
Subject: [PATCH] Paper PR - Add Movement Direction API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index ddec0b7470e5a925ad6247970269f3654792ab25..8079a5f0db670cdb55fc5ab6b6090d0cb0084c97 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -973,6 +973,23 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().hurtDir = hurtDirection;
     }
 
+    // Paper start - Movement Direction API
+    @Override
+    public float getSidewaysMotion() {
+        return this.getHandle().xxa;
+    }
+
+    @Override
+    public float getUpwardsMotion() {
+        return this.getHandle().yya;
+    }
+
+    @Override
+    public float getForwardMotion() {
+        return this.getHandle().zza;
+    }
+    // Paper end - Movement Direction API
+
     public static MobType fromBukkitEntityCategory(EntityCategory entityCategory) {
         switch (entityCategory) {
             case NONE:
