From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 16 May 2024 03:42:33 +0300
Subject: [PATCH] Customizing extra item properties


diff --git a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
index 809a820dd8eec3e48dd3263335c62fbea4cd4f2c..7561ae3037e2144580b14d19092c2875549eea7c 100644
--- a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
@@ -187,7 +187,7 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
         ComposterBlock.add(1.0F, Items.PUMPKIN_PIE);
     }
 
-    private static void add(float levelIncreaseChance, ItemLike item) {
+    public static void add(float levelIncreaseChance, ItemLike item) { // Kiterino - Customizing extra item properties
         ComposterBlock.COMPOSTABLES.put(item.asItem(), levelIncreaseChance);
     }
 
