From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 30 Nov 2021 00:04:10 +0100
Subject: [PATCH] Store field mappings


diff --git a/src/main/java/io/papermc/paper/util/ObfHelper.java b/src/main/java/io/papermc/paper/util/ObfHelper.java
index e8ff684d8bd994c64ff34f20e1e0601b678244c1..28576d2c5fcd28d6fe909c697dfa1b840cdc189c 100644
--- a/src/main/java/io/papermc/paper/util/ObfHelper.java
+++ b/src/main/java/io/papermc/paper/util/ObfHelper.java
@@ -111,10 +111,24 @@ public enum ObfHelper {
                     );
                 }
 
+                // Kiterino start
+                final Map<String, String> fieldMappings = new HashMap<>();
+                for (final MappingTree.FieldMapping fieldMapping : cls.getFields()) {
+                    fieldMappings.put(
+                        pool.string(methodKey(
+                            fieldMapping.getName(SPIGOT_NAMESPACE),
+                            fieldMapping.getDesc(SPIGOT_NAMESPACE)
+                        )),
+                        pool.string(fieldMapping.getName(MOJANG_PLUS_YARN_NAMESPACE))
+                    );
+                }
+                // Kiterino end
+
                 final ClassMapping map = new ClassMapping(
                     Objects.requireNonNull(cls.getName(SPIGOT_NAMESPACE)).replace('/', '.'),
                     Objects.requireNonNull(cls.getName(MOJANG_PLUS_YARN_NAMESPACE)).replace('/', '.'),
-                    Map.copyOf(methods)
+                    Map.copyOf(methods),
+                    Map.copyOf(fieldMappings) // Kiterino
                 );
                 classes.add(map);
             }
@@ -142,6 +156,7 @@ public enum ObfHelper {
     public record ClassMapping(
         String obfName,
         String mojangName,
-        Map<String, String> methodsByObf
+        Map<String, String> methodsByObf,
+        Map<String, String> fieldsByObf // Kiterino
     ) {}
 }
