From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 11 Dec 2021 12:18:42 -0500
Subject: [PATCH] Paper PR - Add Movement Direction API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 37bcc72ce132f7901460f25d700dba1bac6c210f..4e26fb4ef3f32321d9d7370de7472a6715806621 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -962,6 +962,23 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().hurtDir = hurtDirection;
     }
 
+    // Paper start - Movement Direction API
+    @Override
+    public float getSidewaysMotion() {
+        return this.getHandle().xxa;
+    }
+
+    @Override
+    public float getUpwardsMotion() {
+        return this.getHandle().yya;
+    }
+
+    @Override
+    public float getForwardMotion() {
+        return this.getHandle().zza;
+    }
+    // Paper end - Movement Direction API
+
     public static MobType fromBukkitEntityCategory(EntityCategory entityCategory) {
         switch (entityCategory) {
             case NONE:
