From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sun, 17 Oct 2021 14:53:35 -0400
Subject: [PATCH] Paper PR - BoneMeal API


diff --git a/src/main/java/net/minecraft/world/item/BoneMealItem.java b/src/main/java/net/minecraft/world/item/BoneMealItem.java
index 4bc5c7a68f5d2122864d30eee97d2550a73398f9..2cc68515addfc96cdf296a22f9d9896a8072108e 100644
--- a/src/main/java/net/minecraft/world/item/BoneMealItem.java
+++ b/src/main/java/net/minecraft/world/item/BoneMealItem.java
@@ -39,6 +39,11 @@ public class BoneMealItem extends Item {
     }
 
     public static InteractionResult applyBonemeal(UseOnContext itemactioncontext) {
+        // Paper start - BoneMeal API
+        return applyBonemeal(itemactioncontext, true);
+    }
+    public static InteractionResult applyBonemeal(UseOnContext itemactioncontext, boolean showParticles) {
+        // Paper end - BoneMeal API
         // CraftBukkit end
         Level world = itemactioncontext.getLevel();
         BlockPos blockposition = itemactioncontext.getClickedPos();
@@ -47,7 +52,7 @@ public class BoneMealItem extends Item {
         if (BoneMealItem.growCrop(itemactioncontext.getItemInHand(), world, blockposition)) {
             if (!world.isClientSide) {
                 if (itemactioncontext.getPlayer() != null) itemactioncontext.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
-                world.levelEvent(1505, blockposition, 0);
+                if (showParticles) world.levelEvent(1505, blockposition, 0); // Paper - BoneMeal API
             }
 
             return InteractionResult.sidedSuccess(world.isClientSide);
@@ -58,7 +63,7 @@ public class BoneMealItem extends Item {
             if (flag && BoneMealItem.growWaterPlant(itemactioncontext.getItemInHand(), world, blockposition1, itemactioncontext.getClickedFace())) {
                 if (!world.isClientSide) {
                     if (itemactioncontext.getPlayer() != null) itemactioncontext.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
-                    world.levelEvent(1505, blockposition1, 0);
+                    if (showParticles) world.levelEvent(1505, blockposition1, 0); // Paper - BoneMeal API
                 }
 
                 return InteractionResult.sidedSuccess(world.isClientSide);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index bc9a049df648fa5a169219bb009c92a213df9ffd..4b9758ced7beb6890a33960879a14f5262fa6366 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2540,5 +2540,43 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
         return this.adventure$pointers;
     }
+
+    @Override
+    public boolean applyBoneMeal(@org.jetbrains.annotations.NotNull Location location, org.bukkit.block.BlockFace face, boolean showParticles, @org.jetbrains.annotations.Nullable Predicate<BlockState> predicate) {
+        BlockPos pos = new BlockPos(location.getBlockX(), location.getBlockY(), location.getBlockZ());
+        var context = new net.minecraft.world.item.context.UseOnContext(this.getHandle(), null, net.minecraft.world.InteractionHand.MAIN_HAND, net.minecraft.world.item.Items.BONE_MEAL.getDefaultInstance(), new net.minecraft.world.phys.BlockHitResult(Vec3.ZERO, CraftBlock.blockFaceToNotch(face), pos, false));
+
+        // Save old capturing state
+        boolean wasCapturingTrees = world.captureTreeGeneration;
+        boolean wasCapturingBlockStates = world.captureBlockStates;
+
+        Map<BlockPos, org.bukkit.craftbukkit.block.CraftBlockState> capturedBlocks = world.capturedBlockStates;
+        Map<BlockPos, net.minecraft.world.level.block.entity.BlockEntity> capturedTileEntities = world.capturedTileEntities;
+
+        // Create new state, capture everything
+        world.capturedBlockStates = new java.util.LinkedHashMap<>();
+        world.capturedTileEntities = new java.util.LinkedHashMap<>();
+
+        world.captureTreeGeneration = true;
+        world.captureBlockStates = true;
+
+        net.minecraft.world.InteractionResult result = net.minecraft.world.item.BoneMealItem.applyBonemeal(context, showParticles);
+
+        // Revert back booleans
+        world.captureTreeGeneration = wasCapturingTrees;
+        world.captureBlockStates = wasCapturingBlockStates;
+
+        for (BlockState blockState : world.capturedBlockStates.values()) {
+            if (predicate != null && predicate.test(blockState)) {
+                blockState.update(true);
+            }
+        }
+
+        // Revertback maps
+        world.capturedBlockStates = capturedBlocks;
+        world.capturedTileEntities = capturedTileEntities;
+
+        return result == net.minecraft.world.InteractionResult.CONSUME;
+    }
     // Paper end
 }
