From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Sep 2024 21:35:48 +0300
Subject: [PATCH] Data-driven items


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index 6ec9d9b9acf557aa2ebf39d38a14225b0205fae1..7603395aace52f16453de854b95f2f8ca691c573 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -75,7 +75,7 @@ public final class PaperRegistries {
             entry(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT, PotionEffectType.class, CraftPotionEffectType::new),
             entry(Registries.STRUCTURE_TYPE, RegistryKey.STRUCTURE_TYPE, StructureType.class, CraftStructureType::new),
             entry(Registries.BLOCK, RegistryKey.BLOCK, BlockType.class, CraftBlockType::new),
-            entry(Registries.ITEM, RegistryKey.ITEM, ItemType.class, CraftItemType::new),
+            writable(Registries.ITEM, RegistryKey.ITEM, ItemType.class, CraftItemType::new, me.sosedik.kiterino.registry.data.KiterinoItemRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven items
             entry(Registries.CAT_VARIANT, RegistryKey.CAT_VARIANT, Cat.Type.class, CraftCat.CraftType::new),
             entry(Registries.FROG_VARIANT, RegistryKey.FROG_VARIANT, Frog.Variant.class, CraftFrog.CraftVariant::new),
             entry(Registries.VILLAGER_PROFESSION, RegistryKey.VILLAGER_PROFESSION, Villager.Profession.class, CraftVillager.CraftProfession::new),
diff --git a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
index eda5cc7d45ef59ccc1c9c7e027c1f044f1dcc86b..48de73884cdece86ff19e36d930e3990066de3bf 100644
--- a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
+++ b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
@@ -33,4 +33,24 @@ public class Conversions {
     public Component asAdventure(final net.minecraft.network.chat.@Nullable Component vanilla) {
         return vanilla == null ? Component.empty() : this.serializer.deserialize(vanilla);
     }
+
+    // Kiterino start - Data-driven items
+    public net.minecraft.world.item.Rarity asVanilla(final org.bukkit.inventory.ItemRarity rarity) {
+        return switch (rarity) {
+            case COMMON -> net.minecraft.world.item.Rarity.COMMON;
+            case UNCOMMON -> net.minecraft.world.item.Rarity.UNCOMMON;
+            case RARE -> net.minecraft.world.item.Rarity.RARE;
+            case EPIC -> net.minecraft.world.item.Rarity.EPIC;
+        };
+    }
+
+    public org.bukkit.inventory.ItemRarity asBukkit(final net.minecraft.world.item.Rarity rarity) {
+        return switch (rarity) {
+            case COMMON -> org.bukkit.inventory.ItemRarity.COMMON;
+            case UNCOMMON -> org.bukkit.inventory.ItemRarity.UNCOMMON;
+            case RARE -> org.bukkit.inventory.ItemRarity.RARE;
+            case EPIC -> org.bukkit.inventory.ItemRarity.EPIC;
+        };
+    }
+    // Kiterino end - Data-driven items
 }
diff --git a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..82b232742bd50e51d28dfac2258c60f77b0dd95e
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
@@ -0,0 +1,323 @@
+package me.sosedik.kiterino.registry.data;
+
+import net.kyori.adventure.key.Key;
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
+import net.minecraft.core.component.DataComponentMap;
+import net.minecraft.core.component.DataComponents;
+import org.bukkit.inventory.ItemRarity;
+import io.papermc.paper.adventure.PaperAdventure;
+import io.papermc.paper.registry.PaperRegistryBuilder;
+import io.papermc.paper.registry.TypedKey;
+import io.papermc.paper.registry.data.util.Conversions;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.block.ComposterBlock;
+import org.bukkit.inventory.ItemType;
+import org.bukkit.potion.PotionEffect;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.jetbrains.annotations.Range;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
+
+    protected int maxStackSize;
+    protected @Nullable Integer durability;
+    protected boolean fireResistant;
+    protected @Nullable ItemRarity rarity;
+    protected @Nullable Key craftRemainingItemKey;
+    protected FoodProperties.@Nullable Builder foodProperties;
+    protected @Nullable Float compostChance;
+    protected @Nullable Item nmsItem;
+    protected @Nullable Function<Object, Object> nmsItemFunction;
+
+    protected final Conversions conversions;
+    protected final TypedKey<ItemType> itemKey;
+
+    public KiterinoItemRegistryEntity(
+            final @NonNull Conversions conversions,
+            final @NonNull TypedKey<ItemType> ignoredKey,
+            final @Nullable Item internal
+    ) {
+        this.conversions = conversions;
+        this.itemKey = ignoredKey;
+        if (internal == null) {
+            this.maxStackSize = 64;
+            return;
+        }
+
+        parseValues(internal);
+    }
+
+    @Override
+    public int maxStackSize() {
+        return this.maxStackSize;
+    }
+
+    @Override
+    public @Nullable Integer durability() {
+        return this.durability;
+    }
+
+    @Override
+    public boolean fireResistant() {
+        return this.fireResistant;
+    }
+
+    @Override
+    public @Nullable ItemRarity rarity() {
+        return this.rarity;
+    }
+
+    @Override
+    public @Nullable Key craftRemainingItem() {
+        return this.craftRemainingItemKey;
+    }
+
+    @Override
+    public @Nullable FoodProperties foodProperties() {
+        return this.foodProperties;
+    }
+
+    @Override
+    public @Nullable Float compostChance() {
+        return this.compostChance;
+    }
+
+    @Override
+    public @Nullable Object nmsItem() {
+        return this.nmsItem;
+    }
+
+    @Override
+    public @Nullable Function<Object, Object> nmsItemFunction() {
+        return this.nmsItemFunction;
+    }
+
+    protected void parseValues(@NonNull Item item) {
+        DataComponentMap components = item.components();
+        this.maxStackSize = item.getDefaultMaxStackSize();
+        this.durability = components.get(DataComponents.MAX_DAMAGE);
+        this.fireResistant = components.has(DataComponents.FIRE_RESISTANT);
+        this.rarity = components.has(DataComponents.RARITY) ? conversions.asBukkit(Objects.requireNonNull(components.get(DataComponents.RARITY))) : null;
+        if (components.has(DataComponents.FOOD)) {
+            net.minecraft.world.food.FoodProperties properties = components.get(DataComponents.FOOD);
+            assert properties != null;
+            this.foodProperties = new KiterinoFoodProperties.KiterinoBuilder();
+            this.foodProperties.nutrition(properties.nutrition());
+            this.foodProperties.saturation(properties.saturation());
+            this.foodProperties.canAlwaysEat(properties.canAlwaysEat());
+            this.foodProperties.eatSeconds(properties.eatSeconds());
+            this.foodProperties.usingConvertsTo(properties.usingConvertsTo().map(itemStack -> Key.key(itemStack.getItem().toString())).orElse(null)); // TODO internally ItemStack, not just Item
+            this.foodProperties.possibleEffects(
+                properties.effects().isEmpty()
+                    ? List.of()
+                    : properties.effects().stream()
+                        .map(possibleEffect -> Map.entry(CraftPotionUtil.toBukkit(possibleEffect.effect()), possibleEffect.probability()))
+                        .toList()
+            );
+        }
+    }
+
+    public static class KiterinoFoodProperties implements FoodProperties {
+
+        protected int nutrition = 0;
+        protected float saturation = 0F;
+        protected boolean canAlwaysEat = false;
+        protected float eatSeconds = 1.6F;
+        protected @Nullable Key usingConvertsTo;
+        protected @Nullable List<Map.Entry<PotionEffect, Float>> possibleEffects;
+
+        @Override
+        public int nutrition() {
+            return this.nutrition;
+        }
+
+        @Override
+        public float saturation() {
+            return this.saturation;
+        }
+
+        @Override
+        public boolean canAlwaysEat() {
+            return this.canAlwaysEat;
+        }
+
+        @Override
+        public float eatSeconds() {
+            return this.eatSeconds;
+        }
+
+        @Override
+        public @Nullable Key usingConvertsTo() {
+            return this.usingConvertsTo;
+        }
+
+        @Override
+        public @Nullable List<Map.Entry<PotionEffect, Float>> possibleEffects() {
+            return this.possibleEffects;
+        }
+
+        public static final class KiterinoBuilder extends KiterinoFoodProperties implements Builder {
+
+            @Override
+            public @NonNull Builder nutrition(int nutrition) {
+                this.nutrition = nutrition;
+                return this;
+            }
+
+            @Override
+            public @NonNull Builder saturation(float saturation) {
+                this.saturation = saturation;
+                return this;
+            }
+
+            @Override
+            public @NonNull Builder canAlwaysEat(boolean alwaysEdible) {
+                this.canAlwaysEat = alwaysEdible;
+                return this;
+            }
+
+            @Override
+            public @NonNull Builder eatSeconds(float seconds) {
+                this.eatSeconds = seconds;
+                return this;
+            }
+
+            @Override
+            public @NonNull Builder usingConvertsTo(@Nullable Key key) {
+                this.usingConvertsTo = key;
+                return this;
+            }
+
+            @Override
+            public @NonNull Builder possibleEffects(@Nullable List<Map.Entry<PotionEffect, Float>> possibleEffects) {
+                this.possibleEffects = possibleEffects;
+                return this;
+            }
+
+        }
+
+    }
+
+    public static final class KiterinoBuilder extends KiterinoItemRegistryEntity implements Builder,
+            PaperRegistryBuilder<Item, ItemType> {
+
+        public KiterinoBuilder(final @NonNull Conversions conversions, final @NonNull TypedKey<ItemType> key, final @Nullable Item internal) {
+            super(conversions, key, internal);
+        }
+
+        @Override
+        public @NonNull Builder maxStackSize(int maxStackSize) {
+            this.maxStackSize = maxStackSize;
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder durability(@Range(from = 1L, to = Integer.MAX_VALUE) @Nullable Integer durability) {
+            this.durability = durability;
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder fireResistant(boolean fireResistant) {
+            this.fireResistant = fireResistant;
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder rarity(@Nullable ItemRarity rarity) {
+            this.rarity = rarity;
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder craftRemainingItem(@Nullable Key key) {
+            this.craftRemainingItemKey = key;
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder foodProperties(@Nullable Function<FoodProperties.Builder, FoodProperties.Builder> foodProperties) {
+            this.foodProperties = foodProperties == null
+                ? null
+                : foodProperties.apply(
+                    this.foodProperties != null
+                        ? this.foodProperties
+                        : new KiterinoFoodProperties.KiterinoBuilder()
+                );
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder compostChance(@Nullable Float levelIncreaseChance) {
+            this.compostChance = levelIncreaseChance;
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder nmsItem(@Nullable Object nmsItem) {
+            if (!(nmsItem instanceof Item item)) throw new IllegalArgumentException("NMS item must extend Item");
+            this.nmsItem = item;
+            parseValues(item);
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder nmsItemFunction(@Nullable Function<Object, Object> nmsItemFunction) {
+            this.nmsItemFunction = nmsItemFunction;
+            return this;
+        }
+
+        @Override
+        public @NonNull Item build() {
+            Item item = this.nmsItem;
+            if (item == null) {
+                var properties = new Item.Properties();
+                if (this.durability != null) properties.durability(this.durability);
+                properties.stacksTo(this.maxStackSize);
+                if (this.fireResistant) properties.fireResistant();
+                if (this.rarity != null) properties.rarity(conversions.asVanilla(this.rarity));
+                if (this.craftRemainingItemKey != null) properties.craftRemainder(BuiltInRegistries.ITEM.getOptional(PaperAdventure.asVanilla(this.craftRemainingItemKey)).orElseThrow());
+                if (this.foodProperties != null) properties.food(buildFoodProperties());
+                if (this.nmsItemFunction == null) {
+                    item = new Item(properties);
+                } else {
+                    item = (Item) this.nmsItemFunction.apply(properties);
+                }
+            }
+            if (this.compostChance != null) ComposterBlock.add(this.compostChance, item);
+            return item;
+        }
+
+        private net.minecraft.world.food.@NonNull FoodProperties buildFoodProperties() {
+            FoodProperties foodProperties = this.foodProperties;
+            assert foodProperties != null;
+            Optional<ItemStack> leftover = foodProperties.usingConvertsTo() == null
+                    ? Optional.empty()
+                    : BuiltInRegistries.ITEM.getOptional(PaperAdventure.asVanilla(Objects.requireNonNull(foodProperties.usingConvertsTo()))).map(ItemStack::new);
+            List<net.minecraft.world.food.FoodProperties.PossibleEffect> possibleEffects = foodProperties.possibleEffects() == null
+                    ? List.of()
+                    : Objects.requireNonNull(foodProperties.possibleEffects())
+                    .stream()
+                    .map(entry -> new net.minecraft.world.food.FoodProperties.PossibleEffect(CraftPotionUtil.fromBukkit(entry.getKey()), Math.clamp(entry.getValue(), 0F, 1F)))
+                    .toList();
+            return new net.minecraft.world.food.FoodProperties(
+                foodProperties.nutrition(),
+                foodProperties.saturation(),
+                foodProperties.canAlwaysEat(),
+                foodProperties.eatSeconds(),
+                leftover,
+                possibleEffects
+            );
+        }
+
+    }
+
+}
