From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Sep 2024 21:35:48 +0300
Subject: [PATCH] Data-driven items


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index 3ebc3dbc86f3f448f0c555366b55a77876ef5d6f..dbe3ef9ed46d10076da2ce7158e5b73460165a89 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -85,7 +85,7 @@ public final class PaperRegistries {
             start(Registries.STRUCTURE_TYPE, RegistryKey.STRUCTURE_TYPE).craft(StructureType.class, CraftStructureType::new).build(),
             start(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT).craft(PotionEffectType.class, CraftPotionEffectType::new).build(),
             start(Registries.BLOCK, RegistryKey.BLOCK).craft(BlockType.class, CraftBlockType::new).build(),
-            start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).build(),
+            start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).writable(me.sosedik.kiterino.registry.data.KiterinoItemRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven items
             start(Registries.CAT_VARIANT, RegistryKey.CAT_VARIANT).craft(Cat.Type.class, CraftCat.CraftType::new).build(),
             start(Registries.FROG_VARIANT, RegistryKey.FROG_VARIANT).craft(Frog.Variant.class, CraftFrog.CraftVariant::new).build(),
             start(Registries.VILLAGER_PROFESSION, RegistryKey.VILLAGER_PROFESSION).craft(Villager.Profession.class, CraftVillager.CraftProfession::new).build(),
diff --git a/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java b/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java
index f201f142505db8f5a87c20346f6e2998263372fd..f4d75526ddf84a68403ed58db2e4b85e50cc02a0 100644
--- a/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java
+++ b/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java
@@ -42,6 +42,11 @@ public class WritableCraftRegistry<M, T extends Keyed, B extends PaperRegistryBu
         final ResourceKey<M> resourceKey = PaperRegistries.toNms(key);
         this.registry.validateWrite(resourceKey);
         final B builder = this.newBuilder(conversions);
+        // Kiterino start - Data-driven items
+        if (builder instanceof me.sosedik.kiterino.registry.data.KiterinoItemRegistryEntity.KiterinoBuilder b) {
+            b.itemKey = net.minecraft.resources.ResourceLocation.fromNamespaceAndPath(key.key().namespace(), key.key().value());
+        }
+        // Kiterino end - Data-driven items
         value.accept(builder);
         PaperRegistryListenerManager.INSTANCE.registerWithListeners(
             this.registry,
diff --git a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
index b1710da835cee3333bfb1c238b37ec0c4d16ba09..cadbf29c4667a74d79fdc538e0cca9cc61480d29 100644
--- a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
+++ b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
@@ -55,4 +55,24 @@ public class Conversions {
     public Component asAdventure(final net.minecraft.network.chat.@Nullable Component vanilla) {
         return vanilla == null ? Component.empty() : this.serializer.deserialize(vanilla);
     }
+
+    // Kiterino start - Data-driven items
+    public net.minecraft.world.item.Rarity asVanilla(final org.bukkit.inventory.ItemRarity rarity) {
+        return switch (rarity) {
+            case COMMON -> net.minecraft.world.item.Rarity.COMMON;
+            case UNCOMMON -> net.minecraft.world.item.Rarity.UNCOMMON;
+            case RARE -> net.minecraft.world.item.Rarity.RARE;
+            case EPIC -> net.minecraft.world.item.Rarity.EPIC;
+        };
+    }
+
+    public org.bukkit.inventory.ItemRarity asBukkit(final net.minecraft.world.item.Rarity rarity) {
+        return switch (rarity) {
+            case COMMON -> org.bukkit.inventory.ItemRarity.COMMON;
+            case UNCOMMON -> org.bukkit.inventory.ItemRarity.UNCOMMON;
+            case RARE -> org.bukkit.inventory.ItemRarity.RARE;
+            case EPIC -> org.bukkit.inventory.ItemRarity.EPIC;
+        };
+    }
+    // Kiterino end - Data-driven items
 }
diff --git a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..ba8cf283dd9e240d2ceafbb260ea3b987cde937d
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
@@ -0,0 +1,142 @@
+package me.sosedik.kiterino.registry.data;
+
+import io.papermc.paper.datacomponent.DataComponentType;
+import io.papermc.paper.datacomponent.PaperDataComponentType;
+import io.papermc.paper.registry.PaperRegistryBuilder;
+import io.papermc.paper.registry.data.util.Conversions;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.level.block.ComposterBlock;
+import org.bukkit.inventory.ItemType;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
+
+	protected Map<DataComponentType, Object> components = new HashMap<>();
+	protected @Nullable Float compostChance;
+	protected @Nullable Item nmsItem;
+	protected @Nullable Function<Object, Object> nmsItemFunction;
+
+	protected final Conversions conversions;
+	public ResourceLocation itemKey;
+
+	public KiterinoItemRegistryEntity(
+		final @NonNull Conversions conversions,
+		final @Nullable Item internal
+	) {
+		this.conversions = conversions;
+		if (internal == null) {
+			return;
+		}
+
+		parseValues(internal);
+	}
+
+	@Override
+	public boolean hasData(@NonNull DataComponentType type) {
+		return this.components.containsKey(type);
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public <T> @Nullable T getData(DataComponentType.@NonNull Valued<T> type) {
+		return (T) this.components.get(type);
+	}
+
+	@Override
+	public @Nullable Float compostChance() {
+		return this.compostChance;
+	}
+
+	@Override
+	public @Nullable Object nmsItem() {
+		return this.nmsItem;
+	}
+
+	@Override
+	public @Nullable Function<Object, Object> nmsItemFunction() {
+		return this.nmsItemFunction;
+	}
+
+	protected void parseValues(@NonNull Item item) {
+		// Does it make sense to parse?
+		// DataComponentMap components = item.components();
+	}
+
+	public static final class KiterinoBuilder extends KiterinoItemRegistryEntity implements Builder,
+			PaperRegistryBuilder<Item, ItemType> {
+
+		public KiterinoBuilder(final @NonNull Conversions conversions, final @Nullable Item internal) {
+			super(conversions, internal);
+		}
+
+		@Override
+		public <T> @NonNull Builder setData(DataComponentType.@NonNull Valued<T> type, @Nullable T value) {
+			this.components.put(type, value);
+			return this;
+		}
+
+		@Override
+		public @NonNull Builder setData(DataComponentType.@NonNull NonValued type) {
+			this.components.put(type, null);
+			return this;
+		}
+
+		@Override
+		public @NonNull Builder compostChance(@Nullable Float levelIncreaseChance) {
+			this.compostChance = levelIncreaseChance;
+			return this;
+		}
+
+		@Override
+		public @NonNull Builder nmsItem(@Nullable Object nmsItem) {
+			if (!(nmsItem instanceof Item item)) throw new IllegalArgumentException("NMS item must extend Item");
+			this.nmsItem = item;
+			parseValues(item);
+			return this;
+		}
+
+		@Override
+		public @NonNull Builder nmsItemFunction(@Nullable Function<Object, Object> nmsItemFunction) {
+			this.nmsItemFunction = nmsItemFunction;
+			return this;
+		}
+
+		@SuppressWarnings("unchecked")
+		@Override
+		public @NonNull Item build() {
+			Item item = this.nmsItem;
+			if (item == null) {
+				var properties = new Item.Properties();
+				properties.setId(ResourceKey.create(Registries.ITEM, this.itemKey));
+				this.components.forEach((type, value) -> {
+					if (type instanceof PaperDataComponentType.ValuedImpl<?, ?> valued) {
+						setData(properties, (PaperDataComponentType<Object, Object>) valued, value);
+					} else {
+						setData(properties, (PaperDataComponentType.NonValuedImpl<Object, Object>) type, null);
+					}
+				});
+				if (this.nmsItemFunction == null) {
+					item = new Item(properties);
+				} else {
+					item = (Item) this.nmsItemFunction.apply(properties);
+				}
+			}
+			if (this.compostChance != null) ComposterBlock.add(this.compostChance, item);
+			return item;
+		}
+
+		private void setData(Item.@NonNull Properties properties, @NonNull PaperDataComponentType<Object, Object> type, @Nullable Object value) {
+			properties.component(type.getHandle(), type.getAdapter().toVanilla(value));
+		}
+
+	}
+
+}
diff --git a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
index 55f5fc45cfd5be16b4d8e6d083ff10a744196a5e..917182868cdfdd868748a26d9eeebd7ef772e628 100644
--- a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
+++ b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
@@ -25,6 +25,7 @@ class RegistryBuilderTest {
         return PaperRegistries.REGISTRY_ENTRIES.stream()
             .map(RegistryBuilderTest::possiblyUnwrap)
             .filter(RegistryEntry.BuilderHolder.class::isInstance)
+            .filter(entry -> !entry.mcKey().location().equals(net.minecraft.core.registries.Registries.ITEM.location())) // Kiterino - Data-driven items
             .map(Arguments::arguments);
     }
 
