From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 11 Dec 2021 12:18:42 -0500
Subject: [PATCH] Paper PR - Add Movement Direction API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 76a4803decea19f8bed8cb11e73bd7895675a3f4..3b088100dfd1f97177d6f67254bbcb7bb741b66f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -904,6 +904,23 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().hurtDir = hurtDirection;
     }
 
+    // Paper start - Movement Direction API
+    @Override
+    public float getSidewaysMotion() {
+        return this.getHandle().xxa;
+    }
+
+    @Override
+    public float getUpwardsMotion() {
+        return this.getHandle().yya;
+    }
+
+    @Override
+    public float getForwardMotion() {
+        return this.getHandle().zza;
+    }
+    // Paper end - Movement Direction API
+
     public static MobType fromBukkitEntityCategory(EntityCategory entityCategory) {
         switch (entityCategory) {
             case NONE:
