From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 14 Dec 2024 10:10:12 +0200
Subject: [PATCH] Data-driven blocks


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index dbe3ef9ed46d10076da2ce7158e5b73460165a89..b59169e47e163029848226749fdbcc4acd3cc045 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -84,7 +84,7 @@ public final class PaperRegistries {
             start(Registries.GAME_EVENT, RegistryKey.GAME_EVENT).craft(GameEvent.class, CraftGameEvent::new).writable(PaperGameEventRegistryEntry.PaperBuilder::new),
             start(Registries.STRUCTURE_TYPE, RegistryKey.STRUCTURE_TYPE).craft(StructureType.class, CraftStructureType::new).build(),
             start(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT).craft(PotionEffectType.class, CraftPotionEffectType::new).build(),
-            start(Registries.BLOCK, RegistryKey.BLOCK).craft(BlockType.class, CraftBlockType::new).build(),
+            start(Registries.BLOCK, RegistryKey.BLOCK).craft(BlockType.class, CraftBlockType::new).writable(me.sosedik.kiterino.registry.data.KiterinoBlockRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven blocks
             start(Registries.ITEM, RegistryKey.ITEM).craft(ItemType.class, CraftItemType::new).writable(me.sosedik.kiterino.registry.data.KiterinoItemRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven items
             start(Registries.CAT_VARIANT, RegistryKey.CAT_VARIANT).craft(Cat.Type.class, CraftCat.CraftType::new).build(),
             start(Registries.FROG_VARIANT, RegistryKey.FROG_VARIANT).craft(Frog.Variant.class, CraftFrog.CraftVariant::new).build(),
diff --git a/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java b/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java
index 4918b43e691e6c5d5f289a32671197a1bdda413d..42ce7f618f69f04791e23e76c325c2d3267f81a0 100644
--- a/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java
+++ b/src/main/java/io/papermc/paper/registry/WritableCraftRegistry.java
@@ -47,6 +47,11 @@ public class WritableCraftRegistry<M, T extends Keyed, B extends PaperRegistryBu
             b.itemKey = io.papermc.paper.adventure.PaperAdventure.asVanilla(key.key());
         }
         // Kiterino end - Data-driven items
+        // Kiterino start - Data-driven blocks
+        if (builder instanceof me.sosedik.kiterino.registry.data.KiterinoBlockRegistryEntity.KiterinoBuilder b) {
+            b.blockKey = io.papermc.paper.adventure.PaperAdventure.asVanilla(key.key());
+        }
+        // Kiterino end - Data-driven blocks
         value.accept(builder);
         PaperRegistryListenerManager.INSTANCE.registerWithListeners(
             this.registry,
diff --git a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoBlockRegistryEntity.java b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoBlockRegistryEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..4f9365a06cd4d61218b48aba812763ca647caeea
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoBlockRegistryEntity.java
@@ -0,0 +1,79 @@
+package me.sosedik.kiterino.registry.data;
+
+import io.papermc.paper.registry.PaperRegistryBuilder;
+import io.papermc.paper.registry.data.util.Conversions;
+import me.sosedik.kiterino.util.KiterinoDataInjectorRuntimeException;
+import me.sosedik.kiterino.world.block.KiterinoBlock;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockBehaviour;
+import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.block.BlockType;
+import org.bukkit.craftbukkit.block.data.CraftBlockData;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+import java.lang.reflect.InvocationTargetException;
+
+@NullMarked
+public class KiterinoBlockRegistryEntity implements BlockRegistryEntity {
+
+    protected @Nullable KiterinoBlock nmsBlock;
+
+    protected final Conversions conversions;
+    public ResourceLocation blockKey;
+
+    public KiterinoBlockRegistryEntity(
+        Conversions conversions,
+        @Nullable Block internal
+    ) {
+        this.conversions = conversions;
+    }
+
+    @Override
+    public @Nullable KiterinoBlock nmsBlock() {
+        return this.nmsBlock;
+    }
+
+    @Override
+    public Object constructBlockProperties() {
+        return BlockBehaviour.Properties.of().setId(ResourceKey.create(Registries.BLOCK, this.blockKey));
+    }
+
+    public static final class KiterinoBuilder extends KiterinoBlockRegistryEntity implements Builder,
+            PaperRegistryBuilder<Block, BlockType> {
+
+        public KiterinoBuilder(Conversions conversions, @Nullable Block internal) {
+            super(conversions, internal);
+        }
+
+        @Override
+        public Block build() {
+            Block block = (Block) this.nmsBlock;
+            assert block != null;
+            if (this.nmsBlock.getBlockDataClasses() != null) {
+                CraftBlockData.register(block.getClass(), state -> {
+                    try {
+                        return (CraftBlockData) this.nmsBlock.getBlockDataClasses().second().getDeclaredConstructor(BlockState.class).newInstance(state);
+                    } catch (NoSuchMethodException e) {
+                        throw new KiterinoDataInjectorRuntimeException("Invalid block data constructor", e);
+                    } catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
+                        throw new KiterinoDataInjectorRuntimeException("Something went wrong constructing block data", e);
+                    }
+                });
+            }
+            return block;
+        }
+
+        @Override
+        public Builder nmsBlock(KiterinoBlock nmsBlock) {
+            if (!(nmsBlock instanceof Block)) throw new IllegalArgumentException("NMS block must extend Block");
+            this.nmsBlock = nmsBlock;
+            return this;
+        }
+
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
index b570e05a15274f7582f3363990d077a9e927e3c6..01b2058573b0104a3618b42586b27bfb1a528154 100644
--- a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
+++ b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoItemRegistryEntity.java
@@ -1,5 +1,7 @@
 package me.sosedik.kiterino.registry.data;
 
+import net.minecraft.core.registries.BuiltInRegistries;
+
 import io.papermc.paper.datacomponent.DataComponentType;
 import io.papermc.paper.datacomponent.PaperDataComponentType;
 import io.papermc.paper.registry.PaperRegistryBuilder;
@@ -63,6 +65,16 @@ public class KiterinoItemRegistryEntity implements ItemRegistryEntity {
         return new Item.Properties().setId(ResourceKey.create(Registries.ITEM, this.itemKey));
     }
 
+    @Override
+    public boolean hasAttachedBlock() {
+        return BuiltInRegistries.BLOCK.containsKey(ResourceKey.create(Registries.BLOCK, this.itemKey));
+    }
+
+    @Override
+    public @NonNull Object asBlockOrThrow() {
+        return BuiltInRegistries.BLOCK.getValueOrThrow(ResourceKey.create(Registries.BLOCK, this.itemKey));
+    }
+
     protected void parseValues(@NonNull Item item) {
         // Does it make sense to parse?
         // DataComponentMap components = item.components();
diff --git a/src/main/java/me/sosedik/kiterino/world/block/KiterinoBlockStateIdMapper.java b/src/main/java/me/sosedik/kiterino/world/block/KiterinoBlockStateIdMapper.java
new file mode 100644
index 0000000000000000000000000000000000000000..a7ec3a95517da666907dccd50c01e2928fb388c0
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/block/KiterinoBlockStateIdMapper.java
@@ -0,0 +1,27 @@
+package me.sosedik.kiterino.world.block;
+
+import net.minecraft.core.IdMapper;
+import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class KiterinoBlockStateIdMapper extends IdMapper<BlockState> {
+
+    public static int lastVanillaBlockStateId = 0;
+
+    public KiterinoBlockStateIdMapper() {
+        super();
+    }
+
+    @Override
+    public int getId(BlockState value) {
+        if (!(value.getBlock() instanceof KiterinoBlock kiterinoBlock)) {
+            return super.getId(value);
+        }
+
+        org.bukkit.block.BlockState bukkitState = kiterinoBlock.serializeBlockToClient();
+        return bukkitState == null ? super.getId(value) : super.getId(((CraftBlockState) bukkitState).getHandle());
+    }
+
+}
diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index f00526753a83f95689fad2a132bef79f4479eec6..548811c2dabc104b43d50459221e8ca89a354a78 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -66,7 +66,7 @@ public class Block extends BlockBehaviour implements ItemLike {
     public static final MapCodec<Block> CODEC = simpleCodec(Block::new);
     private static final Logger LOGGER = LogUtils.getLogger();
     private final Holder.Reference<Block> builtInRegistryHolder;
-    public static final IdMapper<BlockState> BLOCK_STATE_REGISTRY = new IdMapper<>();
+    public static final IdMapper<BlockState> BLOCK_STATE_REGISTRY = /*new IdMapper<>()*/ new me.sosedik.kiterino.world.block.KiterinoBlockStateIdMapper(); // Kiterino - Data-driven blocks
     private static final LoadingCache<VoxelShape, Boolean> SHAPE_FULL_BLOCK_CACHE = CacheBuilder.newBuilder().maximumSize(512L).weakKeys().build(new CacheLoader<VoxelShape, Boolean>() {
         public Boolean load(VoxelShape voxelshape) {
             return !Shapes.joinIsNotEmpty(Shapes.block(), voxelshape, BooleanOp.NOT_SAME);
diff --git a/src/main/java/net/minecraft/world/level/block/Blocks.java b/src/main/java/net/minecraft/world/level/block/Blocks.java
index c5f8227cd9631d98cc8404e3f6d6109a55c617aa..b5a8aace580079354ff171e9595147077f40da4e 100644
--- a/src/main/java/net/minecraft/world/level/block/Blocks.java
+++ b/src/main/java/net/minecraft/world/level/block/Blocks.java
@@ -6920,5 +6920,6 @@ public class Blocks {
                 blockState.initCache();
             }
         }
+        me.sosedik.kiterino.world.block.KiterinoBlockStateIdMapper.lastVanillaBlockStateId = Block.BLOCK_STATE_REGISTRY.size() - 1;
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
index c8ae6e4cd74549f753ec04def5d882de1ab72308..d8bd826a8707237488906fe9db56f9e0c1e329e2 100644
--- a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -769,6 +769,12 @@ public abstract class BlockBehaviour implements FeatureElement {
             return this;
         }
 
+        // Kiterino start - Data-driven blocks
+        public @Nullable ResourceKey<Block> getId() {
+            return this.id;
+        }
+        // Kiterino end - Data-driven blocks
+
         public BlockBehaviour.Properties overrideDescription(String translationKey) {
             this.descriptionId = DependantName.fixed(translationKey);
             return this;
diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
index 66e40149a59c4bd5d5ca3b5a69c733eea49fb86b..e8fec76a821206b1156cf4b9518bc50ac9d66abe 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
@@ -544,7 +544,7 @@ public class CraftBlockData implements BlockData {
         //</editor-fold>
     }
 
-    private static void register(Class<? extends Block> nms, Function<net.minecraft.world.level.block.state.BlockState, CraftBlockData> bukkit) {
+    public static void register(Class<? extends Block> nms, Function<net.minecraft.world.level.block.state.BlockState, CraftBlockData> bukkit) { // Kiterino - private -> public - Data-driven blocks
         Preconditions.checkState(CraftBlockData.MAP.put(nms, bukkit) == null, "Duplicate mapping %s->%s", nms, bukkit);
     }
 
diff --git a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
index 917182868cdfdd868748a26d9eeebd7ef772e628..de503146ae81e540dfcfcc9812d7a68452442c5f 100644
--- a/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
+++ b/src/test/java/io/papermc/paper/registry/RegistryBuilderTest.java
@@ -26,6 +26,7 @@ class RegistryBuilderTest {
             .map(RegistryBuilderTest::possiblyUnwrap)
             .filter(RegistryEntry.BuilderHolder.class::isInstance)
             .filter(entry -> !entry.mcKey().location().equals(net.minecraft.core.registries.Registries.ITEM.location())) // Kiterino - Data-driven items
+            .filter(entry -> !entry.mcKey().location().equals(net.minecraft.core.registries.Registries.BLOCK.location())) // Kiterino - Data-driven blocks
             .map(Arguments::arguments);
     }
 
