From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 16 May 2024 03:42:33 +0300
Subject: [PATCH] Injecting custom Materials


diff --git a/build.gradle.kts b/build.gradle.kts
index 02934d5d212c9a74b58aa4a173d93d4351370d4d..7b8773c5d9c5f0f009ce0217792fb65b23239b51 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -125,6 +125,7 @@ tasks.jar {
         val gitBranch = git("rev-parse", "--abbrev-ref", "HEAD").getText().trim() // Paper
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
+            "Add-Opens" to "java.base/java.lang", // Kiterino - Injecting custom Materials
             "Implementation-Title" to "Kiterino", // Kiterino - Init
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date, // Paper
@@ -209,6 +210,7 @@ fun TaskContainer.registerRunTask(
         vendor.set(JvmVendorSpec.JETBRAINS)
     })
     jvmArgs("-XX:+AllowEnhancedClassRedefinition", "-XX:+AllowRedefinitionToAddDeleteMethods")
+    jvmArgs("--add-opens", "java.base/java.lang=ALL-UNNAMED") // Kiterino - Injecting custom Materials
 
     if (rootProject.childProjects["test-plugin"] != null) {
         val testPluginJar = rootProject.project(":test-plugin").tasks.jar.flatMap { it.archiveFile }
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjectorImpl.java b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjectorImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..a35b5265681c6c535fc5a72a548b4dbeb5d6afca
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjectorImpl.java
@@ -0,0 +1,161 @@
+package me.sosedik.kiterino.util;
+
+import com.google.common.base.Suppliers;
+import io.papermc.paper.adventure.PaperAdventure;
+import me.sosedik.kiterino.world.block.KiterinoBlock;
+import net.kyori.adventure.key.Key;
+import net.minecraft.core.component.DataComponents;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.level.block.Block;
+import org.bukkit.Material;
+import org.bukkit.NamespacedKey;
+import org.bukkit.Registry;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static me.sosedik.kiterino.util.KiterinoUnsafeUtil.getField;
+
+/**
+ * Abomination to workaround Material enum
+ * This accessibility hack should never be used, ever
+ */
+public class KiterinoBootstrapMaterialInjectorImpl implements IKiterinoBootstrapMaterialInjector {
+
+    private final int originalMaterialCount = Material.values().length;
+    private final List<Material> allocated = new ArrayList<>();
+    private final Field materialItemField = getField(CraftMagicNumbers.class, "MATERIAL_ITEM");
+    private final Field itemMaterialField = getField(CraftMagicNumbers.class, "ITEM_MATERIAL");
+    private final Field materialBlockField = getField(CraftMagicNumbers.class, "MATERIAL_BLOCK");
+    private final Field blockMaterialField = getField(CraftMagicNumbers.class, "BLOCK_MATERIAL");
+    private final Field nameField = getField(Enum.class, "name");
+    private final Field ordinalField = getField(Enum.class, "ordinal");
+    private final Field byNameField = getField(Material.class, "BY_NAME");
+    private final Field maxStackField = getField(Material.class, "maxStack");
+    private final Field durabilityField = getField(Material.class, "durability");
+    private final Field idField = getField(Material.class, "id");
+    private final Field ctorField = getField(Material.class, "ctor");
+    private final Field dataField = getField(Material.class, "data");
+    private final Field legacyField = getField(Material.class, "legacy");
+    private final Field keyField = getField(Material.class, "key");
+    private final Field itemTypeField = getField(Material.class, "itemType");
+    private final Field blockTypeField = getField(Material.class, "blockType");
+    private final Field injectedField = getField(Material.class, "injected");
+
+    @SuppressWarnings("java:S3011")
+    @Override
+    public void injectMaterials(@NotNull Class<?> materialsClass) {
+        try {
+            for (Field field : materialsClass.getDeclaredFields()) {
+                if (field.getType() != Material.class) continue;
+
+                int modifiers = field.getModifiers();
+                if (!Modifier.isPublic(modifiers)) continue;
+                if (!Modifier.isStatic(modifiers)) continue;
+                if (!Modifier.isFinal(modifiers)) continue;
+
+                Material material = (Material) field.get(null);
+                NamespacedKey materialKey = material.getKey();
+                Item item = BuiltInRegistries.ITEM.getValueOrThrow(ResourceKey.create(Registries.ITEM, PaperAdventure.asVanilla(materialKey)));
+
+                maxStackField.set(material, item.getDefaultMaxStackSize());
+                durabilityField.set(material, (short) (int) item.components().getOrDefault(DataComponents.MAX_DAMAGE, 0));
+
+                idField.set(material, 1);
+                legacyField.set(material, false);
+                if (item instanceof BlockItem blockItem && blockItem.getBlock() instanceof KiterinoBlock kiterinoBlock && kiterinoBlock.getBlockDataClasses() != null) {
+                    Class<?> dataClass = kiterinoBlock.getBlockDataClasses().first();
+                    ctorField.set(material, dataClass.getConstructor(Material.class, byte.class));
+                    dataField.set(material, dataClass);
+                } else {
+                    ctorField.set(material, org.bukkit.material.MaterialData.class.getConstructor(Material.class, byte.class));
+                    dataField.set(material, org.bukkit.material.MaterialData.class);
+                }
+
+                itemTypeField.set(material, Suppliers.memoize(() -> Registry.ITEM.get(materialKey)));
+                blockTypeField.set(material, Suppliers.memoize(() -> Registry.BLOCK.get(materialKey)));
+
+                ((Map<Material, Item>) materialItemField.get(null)).put(material, item);
+                ((Map<Item, Material>) itemMaterialField.get(null)).put(item, material);
+                if (item instanceof BlockItem blockItem) {
+                    ((Map<Material, Block>) materialBlockField.get(null)).put(material, blockItem.getBlock());
+                    ((Map<Block, Material>) blockMaterialField.get(null)).put(blockItem.getBlock(), material);
+                }
+
+                allocated.add(material);
+            }
+        } catch (Exception e) {
+            throw new KiterinoMaterialExtenderRuntimeException("Couldn't extend Material in " + materialsClass, e);
+        }
+    }
+
+    @SuppressWarnings({"unchecked", "java:S3011"})
+    @Override
+    public @NotNull Material allocateMaterial(@NotNull Key key) {
+        try {
+            Material material = KiterinoUnsafeUtil.allocateInstance(Material.class);
+            injectedField.set(material, true);
+            keyField.set(material, new NamespacedKey(key.namespace(), key.value()));
+
+            String enumName = toMaterialName(key);
+            nameField.set(material, enumName);
+            ((Map<String, Material>) byNameField.get(null)).put(enumName, material);
+
+            return material;
+        } catch (Exception e) {
+            throw new KiterinoMaterialExtenderRuntimeException("Couldn't extend Material: " + key, e);
+        }
+    }
+
+    @SuppressWarnings("java:S3011")
+    private void finalizeInject() {
+        try {
+            if (allocated.isEmpty()) return;
+
+            Material[] newValues = new Material[originalMaterialCount + allocated.size()];
+            for (int i = 0; i < allocated.size(); i++) {
+                Material material = allocated.get(i);
+                int ordinal = originalMaterialCount + i;
+                ordinalField.set(material, ordinal);
+                newValues[ordinal] = material;
+            }
+
+            System.arraycopy(Material.values(), 0, newValues, 0, originalMaterialCount);
+            Field valuesField = Material.class.getDeclaredField("$VALUES");
+            KiterinoUnsafeUtil.setField(valuesField, newValues);
+
+            Field enumConstants = Class.class.getDeclaredField("enumConstants");
+            enumConstants.setAccessible(true);
+            enumConstants.set(Material.class, null);
+            Field enumConstantDirectory = Class.class.getDeclaredField("enumConstantDirectory");
+            enumConstantDirectory.setAccessible(true);
+            enumConstantDirectory.set(Material.class, null);
+        } catch (Exception e) {
+            throw new KiterinoMaterialExtenderRuntimeException("Couldn't extend Material", e);
+        }
+    }
+
+    private static @NotNull String toMaterialName(@NotNull Key key) {
+        return key.asString().toUpperCase(Locale.ROOT).replace(':', '_');
+    }
+
+    public static void init() {
+        KiterinoBootstrapMaterialInjector.injector = new KiterinoBootstrapMaterialInjectorImpl();
+    }
+
+    public static void destruct() {
+        ((KiterinoBootstrapMaterialInjectorImpl) KiterinoBootstrapMaterialInjector.injector).finalizeInject();
+        KiterinoBootstrapMaterialInjector.injector = null;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorRuntimeException.java b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorRuntimeException.java
new file mode 100644
index 0000000000000000000000000000000000000000..aac1eca9d78a220f5f7895c2799c1f2d1bbfe1b3
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjectorRuntimeException.java
@@ -0,0 +1,16 @@
+package me.sosedik.kiterino.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class KiterinoDataInjectorRuntimeException extends RuntimeException {
+
+    public KiterinoDataInjectorRuntimeException(@NotNull String message) {
+        super(message);
+    }
+
+    public KiterinoDataInjectorRuntimeException(@NotNull String message, @Nullable Throwable cause) {
+        super(message, cause);
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoMaterialExtenderRuntimeException.java b/src/main/java/me/sosedik/kiterino/util/KiterinoMaterialExtenderRuntimeException.java
new file mode 100644
index 0000000000000000000000000000000000000000..f36af092ca5f7211c57a78f7305a1f6bcd112e80
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoMaterialExtenderRuntimeException.java
@@ -0,0 +1,12 @@
+package me.sosedik.kiterino.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class KiterinoMaterialExtenderRuntimeException extends RuntimeException {
+
+    public KiterinoMaterialExtenderRuntimeException(@NotNull String message, @Nullable Throwable cause) {
+        super(message, cause);
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoUnsafeUtil.java b/src/main/java/me/sosedik/kiterino/util/KiterinoUnsafeUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..5372d7336d677f6b89b7a7205807a227999fbf39
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoUnsafeUtil.java
@@ -0,0 +1,52 @@
+package me.sosedik.kiterino.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import sun.misc.Unsafe;
+
+import java.lang.reflect.Field;
+
+final class KiterinoUnsafeUtil {
+
+    private KiterinoUnsafeUtil() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    private static Unsafe unsafe;
+
+    static {
+        try {
+            Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
+            unsafeField.setAccessible(true);
+            unsafe = (Unsafe) unsafeField.get(null);
+        } catch(Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    static @NotNull Field getField(@NotNull Class<?> clazz, @NotNull String fieldName) {
+        try {
+            Field field = clazz.getDeclaredField(fieldName);
+            field.setAccessible(true);
+            return field;
+        } catch (NoSuchFieldException e) {
+            throw new KiterinoDataInjectorRuntimeException("Couldn't get field", e);
+        }
+    }
+
+    static void setField(@NotNull Field field, @Nullable Object value) {
+        Object fieldBase = unsafe.staticFieldBase(field);
+        long fieldOffset = unsafe.staticFieldOffset(field);
+
+        unsafe.putObject(fieldBase, fieldOffset, value);
+    }
+
+    static <T> T allocateInstance(@NotNull Class<T> tClass) {
+        try {
+            return (T) unsafe.allocateInstance(tClass);
+        } catch (Exception e) {
+            throw new KiterinoDataInjectorRuntimeException("Couldn't allocate new instance", e);
+        }
+    }
+
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 8979956334ad5802d221e45eba91248e7f048d13..f81f4a80d7a4b4f59410acbac935cdc548d585c6 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -154,7 +154,7 @@ public final class CraftMagicNumbers implements UnsafeValues {
         }
 
         for (Item item : BuiltInRegistries.ITEM) {
-            ITEM_MATERIAL.put(item, Material.getMaterial(BuiltInRegistries.ITEM.getKey(item).getPath().toUpperCase(Locale.ROOT)));
+            ITEM_MATERIAL.put(item, Material.matchMaterial(BuiltInRegistries.ITEM.getKey(item).toString())); // Kiterino - Injecting custom Materials
         }
 
         for (Material material : Material.values()) {
