From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 30 Jan 2021 18:55:38 +0200
Subject: [PATCH] Allow modifying vanilla block drops


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index b82a6143526bd1d4ecd4591c1253cdb0b913fe09..72fd97bf27ee0f19b88f4bcc584a76b9cf979823 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -462,6 +462,21 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
 
         if (!event.isCancelled()) {
+            // Kiterino start
+            if (event.changeDrops()) {
+                for (Item bukkit : list) {
+                    if (bukkit.isValid()) {
+                        bukkit.remove();
+                    }
+                }
+                org.bukkit.World world = block.getWorld();
+                org.bukkit.Location loc = block.getLocation();
+                for (org.bukkit.inventory.ItemStack drop : event.getDrops()) {
+                    world.dropItemNaturally(loc, drop);
+                }
+                return;
+            }
+            // Kiterino end
             // Paper start - Allow adding items to BlockDropItemEvent
             for (Item bukkit : list) {
                 if (!bukkit.isValid()) {
