From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 15 Jun 2024 03:26:19 +0300
Subject: [PATCH] Exclude some classes from Material rerouting


diff --git a/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java b/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java
index 946cd46f3389a4d4ceda86e0115c59c5725a8a0a..6703d2563219717c782cd268860e864da7f86102 100644
--- a/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java
+++ b/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java
@@ -58,6 +58,12 @@ public class MaterialReroutingTest extends AbstractTestingBase {
                 // Paper start - types that cannot be translated to ItemType/BlockType
                 .filter(entry -> !entry.getName().equals("com/destroystokyo/paper/MaterialSetTag.class"))
                 // Paper end - types that cannot be translated to ItemType/BlockType
+                // Kiterino start - Exclude some classes from Material rerouting
+                .filter(entry -> !entry.getName().equals("me/sosedik/kiterino/modifier/item/ItemContextBox.class"))
+                .filter(entry -> !entry.getName().equals("me/sosedik/kiterino/util/IKiterinoBootstrapMaterialInjector.class"))
+                .filter(entry -> !entry.getName().equals("me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjector.class"))
+                .filter(entry -> !entry.getName().equals("me/sosedik/kiterino/world/item/KiterinoItem.class"))
+                // Kiterino end - Exclude some classes from Material rerouting
                 .map(entry -> {
                     try {
                         return MaterialReroutingTest.jarFile.getInputStream(entry);
