From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 18:16:41 +0300
Subject: [PATCH] All mobs have default attack damage


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 2b24f8ba091b708eaa21cd0722ef2b5af1947b9a..cc544956282e97a99d8947f621270dab285a0b70 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -341,7 +341,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public static AttributeSupplier.Builder createLivingAttributes() {
-        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS);
+        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS).add(Attributes.ATTACK_DAMAGE); // Kiterino - add ATTACK_DAMAGE
     }
     public boolean shouldSendAttribute(Attribute attribute) { return true; } // Purpur
 
@@ -3054,7 +3054,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
     public boolean doHurtTarget(Entity target) {
         this.setLastHurtMob(target);
-        return false;
+        // Kiterino start - make all mobs able to attack
+        boolean flag = target.hurt(this.damageSources().mobAttack(this), (float) ((int) getAttribute(Attributes.ATTACK_DAMAGE).getValue()));
+        if (flag)
+            doEnchantDamageEffects(this, target);
+        return flag;
+        // Kiterino end
     }
 
     @Override
