From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 18:16:41 +0300
Subject: [PATCH] All mobs have default attack damage


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 88f2ed316c56a15993b412685287b35ab40188c8..6d598d5cadbe6f964f0867e17d8fb8f7e132ba82 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -340,7 +340,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public static AttributeSupplier.Builder createLivingAttributes() {
-        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS);
+        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS).add(Attributes.ATTACK_DAMAGE); // Kiterino - add ATTACK_DAMAGE
     }
     public boolean shouldSendAttribute(Attribute attribute) { return true; } // Purpur
 
@@ -3044,7 +3044,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
     public boolean doHurtTarget(Entity target) {
         this.setLastHurtMob(target);
-        return false;
+        // Kiterino start - make all mobs able to attack
+        boolean flag = target.hurt(this.damageSources().mobAttack(this), (float) ((int) getAttribute(Attributes.ATTACK_DAMAGE).getValue()));
+        if (flag)
+            doEnchantDamageEffects(this, target);
+        return flag;
+        // Kiterino end
     }
 
     @Override
