From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 30 Jan 2021 18:55:38 +0200
Subject: [PATCH] Allow modifying vanilla block drops


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 6ed0735e9b98b80ed45116907e7563f5e8b993d4..32a95d06ccc8b5624d5d3f9f0b623c75da615d41 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -434,6 +434,21 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
 
         if (!event.isCancelled()) {
+            // Kiterino start
+            if (event.changeDrops()) {
+                for (Item bukkit : list) {
+                    if (bukkit.isValid()) {
+                        bukkit.remove();
+                    }
+                }
+                org.bukkit.World world = block.getWorld();
+                org.bukkit.Location loc = block.getLocation();
+                for (org.bukkit.inventory.ItemStack drop : event.getDrops()) {
+                    world.dropItemNaturally(loc, drop);
+                }
+                return;
+            }
+            // Kiterino end
             // Paper start
             for (Item bukkit : list) {
                 if (!bukkit.isValid()) {
