From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 9 Jul 2024 19:12:41 +0300
Subject: [PATCH] Injecting custom mob effects


diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMobEffectInjectorImpl.java b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMobEffectInjectorImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..dece80403ed809fcb6f8144b6f3dbc2777b37412
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMobEffectInjectorImpl.java
@@ -0,0 +1,108 @@
+package me.sosedik.kiterino.util;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import me.sosedik.kiterino.world.effect.KiterinoMobEffect;
+import me.sosedik.kiterino.world.effect.KiterinoMobEffectWrapper;
+import me.sosedik.kiterino.world.effect.KiterinoNMSMobEffect;
+import net.minecraft.core.MappedRegistry;
+import net.minecraft.core.Registry;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.world.effect.MobEffect;
+import org.bukkit.NamespacedKey;
+import org.bukkit.potion.PotionEffectType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Supplier;
+
+public class KiterinoBootstrapMobEffectInjectorImpl implements IKiterinoBootstrapMobEffectInjector {
+
+	private final List<MobEffectInitData> mobEffectInitDatas = new ArrayList<>();
+	private final List<MobEffectData> cachedMobEffectDatas = new ArrayList<>();
+
+	@Override
+	public void injectMobEffects(@NotNull Class<?> mobEffectsClass, @NotNull String namespace) {
+		try {
+			Field[] fields = mobEffectsClass.getDeclaredFields();
+			if (fields.length > 0) fields[0].get(null); // Init
+			for (int i = 0; i < fields.length; i++) {
+				Field field = fields[i];
+				if (field.getType() != PotionEffectType.class) continue;
+
+				int modifiers = field.getModifiers();
+				if (!Modifier.isPublic(modifiers)) continue;
+				if (!Modifier.isStatic(modifiers)) continue;
+				if (!Modifier.isFinal(modifiers)) continue;
+
+				MobEffectData mobEffectData = cachedMobEffectDatas.get(i);
+				NamespacedKey key = mobEffectData.key();
+				if (key == null)
+					key = new NamespacedKey(namespace, field.getName().toLowerCase(Locale.ROOT));
+				injectToRegistry(field, key, mobEffectData.mobEffectSupplier());
+			}
+		} catch (Exception e) {
+			throw new KiterinoDataInjectorRuntimeException("Couldn't inject new mob effects", e);
+		} finally {
+			this.cachedMobEffectDatas.clear();
+		}
+	}
+
+	@Override
+	public @Nullable PotionEffectType injectMobEffect(@Nullable NamespacedKey key, @NotNull Supplier<? extends KiterinoNMSMobEffect> mobEffectSupplier) {
+		this.cachedMobEffectDatas.add(new MobEffectData(key, mobEffectSupplier));
+		return null;
+	}
+
+	private void injectToRegistry(@NotNull Field field, @NotNull NamespacedKey key, @NotNull Supplier<? extends KiterinoNMSMobEffect> effectSupplier) {
+		this.mobEffectInitDatas.add(new MobEffectInitData(field, key, () -> {
+			KiterinoNMSMobEffect effect = effectSupplier.get();
+			if (effect instanceof MobEffect mobEffect)
+				return mobEffect;
+			if (effect instanceof KiterinoMobEffect mobEffect)
+				return new KiterinoMobEffectWrapper(mobEffect);
+			throw new KiterinoDataInjectorRuntimeException(String.format("Unsupported mob effect implementation: must be %s or %s", MobEffect.class.getSimpleName(), KiterinoMobEffect.class.getSimpleName()));
+		}));
+	}
+
+	private record MobEffectData(
+		@Nullable NamespacedKey key,
+		@NotNull Supplier<? extends KiterinoNMSMobEffect> mobEffectSupplier
+	) {}
+
+	private record MobEffectInitData(
+		@NotNull Field field,
+		@NotNull NamespacedKey key,
+		@NotNull Supplier<MobEffect> mobEffectSupplier
+	) {}
+
+	public static void init() {
+		KiterinoBootstrapMobEffectInjector.injector = new KiterinoBootstrapMobEffectInjectorImpl();
+	}
+
+	public static void destruct() {
+		try {
+			// Put mob effects into nms registry
+			((MappedRegistry<MobEffect>) BuiltInRegistries.MOB_EFFECT).unfreeze();
+			for (MobEffectInitData mobEffectInitData : ((KiterinoBootstrapMobEffectInjectorImpl) KiterinoBootstrapMobEffectInjector.injector).mobEffectInitDatas) {
+				MobEffect mobEffect = mobEffectInitData.mobEffectSupplier().get();
+				Registry.registerForHolder(BuiltInRegistries.MOB_EFFECT, PaperAdventure.asVanilla(mobEffectInitData.key()), mobEffect);
+			}
+			// Construct bukkit wrappers and set fields to correct values
+			((MappedRegistry<MobEffect>) BuiltInRegistries.MOB_EFFECT).removeDefaultIntrusiveAndFreeze();
+			for (MobEffectInitData mobEffectInitData : ((KiterinoBootstrapMobEffectInjectorImpl) KiterinoBootstrapMobEffectInjector.injector).mobEffectInitDatas) {
+				PotionEffectType potionEffectType = org.bukkit.Registry.POTION_EFFECT_TYPE.get(mobEffectInitData.key());
+				KiterinoUnsafeUtil.setField(mobEffectInitData.field(), potionEffectType);
+			}
+		} catch (Exception e) {
+			throw new KiterinoDataInjectorRuntimeException("Couldn't inject new mob effects", e);
+		} finally {
+			KiterinoBootstrapMobEffectInjector.injector = null;
+		}
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffectWrapper.java b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffectWrapper.java
new file mode 100644
index 0000000000000000000000000000000000000000..5affa5379f1fb2121c095ad08378cf1bbba77de3
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffectWrapper.java
@@ -0,0 +1,68 @@
+package me.sosedik.kiterino.world.effect;
+
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.effect.MobEffect;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import org.bukkit.craftbukkit.damage.CraftDamageSource;
+import org.bukkit.craftbukkit.potion.CraftPotionEffectTypeCategory;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class KiterinoMobEffectWrapper extends MobEffect {
+
+	private final KiterinoMobEffect mobEffect;
+
+	public KiterinoMobEffectWrapper(@NotNull KiterinoMobEffect mobEffect) {
+		super(CraftPotionEffectTypeCategory.bukkitToMinecraft(mobEffect.getCategory()), mobEffect.getColor());
+		this.mobEffect = mobEffect;
+	}
+
+	@Override
+	public int getBlendDurationTicks() {
+		return this.mobEffect.getBlendDurationTicks();
+	}
+
+	@Override
+	public boolean applyEffectTick(@NotNull LivingEntity entity, int amplifier) {
+		return this.mobEffect.applyEffectTick(entity.getBukkitLivingEntity(), amplifier);
+	}
+
+	@Override
+	public void applyInstantenousEffect(@Nullable Entity source, @Nullable Entity attacker, @NotNull LivingEntity target, int amplifier, double proximity) {
+		this.mobEffect.applyInstantaneousEffect(source == null ? null : source.getBukkitEntity(), attacker == null ? null : attacker.getBukkitEntity(), target.getBukkitLivingEntity(), amplifier, proximity);
+	}
+
+	@Override
+	public boolean shouldApplyEffectTickThisTick(int duration, int amplifier) {
+		return this.mobEffect.shouldApplyEffectTickThisTick(duration, amplifier);
+	}
+
+	@Override
+	public void onEffectStarted(@NotNull LivingEntity entity, int amplifier) {
+		this.mobEffect.onEffectStarted(entity.getBukkitLivingEntity(), amplifier);
+	}
+
+	@Override
+	public void onEffectAdded(@NotNull LivingEntity entity, int amplifier) {
+		this.mobEffect.onEffectAdded(entity.getBukkitLivingEntity(), amplifier);
+	}
+
+	@Override
+	public void onMobRemoved(@NotNull LivingEntity entity, int amplifier, @NotNull Entity.RemovalReason reason) {
+		if (reason == Entity.RemovalReason.KILLED) {
+			this.mobEffect.onMobKilled(entity.getBukkitLivingEntity(), amplifier);
+		}
+	}
+
+	@Override
+	public void onMobHurt(@NotNull LivingEntity entity, int amplifier, @NotNull DamageSource source, float amount) {
+		this.mobEffect.onMobHurt(entity.getBukkitLivingEntity(), amplifier, new CraftDamageSource(source), amount);
+	}
+
+	@Override
+	public boolean isInstantenous() {
+		return this.mobEffect.isInstantaneous();
+	}
+
+}
