From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 14:32:55 +0300
Subject: [PATCH] Allow force sleeping


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 99cf8ce63316e127a5ee84cdd96df80d440dc0c3..94b8d44af25379f4cf9f310335c02947a143beb4 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -539,7 +539,7 @@ public class ServerPlayer extends Player {
         }
         this.getBukkitEntity().readExtraData(nbt); // CraftBukkit
 
-        if (this.isSleeping()) {
+        if (this.isSleeping() && this.checkBedExists()) { // Kiterino - wake up only if bed exists
             this.stopSleeping();
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index fbba9404d85c104e23062ad20f526c1b1379be09..f550d3d6265f2bfb267836c8b194a7d4aed996e7 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3102,7 +3102,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 this.getCombatTracker().recheckStatus();
             }
 
-            if (this.isSleeping() && !this.checkBedExists()) {
+            if (false && this.isSleeping() && !this.checkBedExists()) { // Kiterino - do not wake up automatically even if no bed underneath
                 this.stopSleeping();
             }
         }
@@ -4418,7 +4418,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         this.setPos((double) pos.getX() + 0.5D, (double) pos.getY() + 0.6875D, (double) pos.getZ() + 0.5D);
     }
 
-    private boolean checkBedExists() {
+    public boolean checkBedExists() { // Kiterino - private -> public
         return (Boolean) this.getSleepingPos().map((blockposition) -> {
             return this.level().getBlockState(blockposition).getBlock() instanceof BedBlock;
         }).orElse(false);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 28910943e5efcdf24865b961fba161b7d1e4a8f5..feb12c9e7322e1474ce0778660b5a154049ea111 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1381,7 +1381,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         if (dismount) entity.stopRiding(); // Paper - Teleport API
 
         // SPIGOT-5509: Wakeup, similar to riding
-        if (this.isSleeping()) {
+        if (this.isSleeping() && getHandle().checkBedExists()) { // Kiterino - wake up only if bed exists
             this.wakeup(false);
         }
 
