From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 14:32:55 +0300
Subject: [PATCH] Allow force sleeping


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 305b90d10a499e9731f5178433fb10207e428091..fa1ebfe3270ab138ae3598af67931b51f055075f 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -547,7 +547,7 @@ public class ServerPlayer extends Player {
         }
         this.getBukkitEntity().readExtraData(nbt); // CraftBukkit
 
-        if (this.isSleeping()) {
+        if (this.isSleeping() && this.checkBedExists()) { // Kiterino - wake up only if bed exists
             this.stopSleeping();
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index b8aa36986aac691f3334bb57c3a353f4cea77f6d..93fa6f20fae144dc36a8bbe3e4a06a1bda8ccae2 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3153,7 +3153,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 this.getCombatTracker().recheckStatus();
             }
 
-            if (this.isSleeping() && !this.checkBedExists()) {
+            if (false && this.isSleeping() && !this.checkBedExists()) { // Kiterino - do not wake up automatically even if no bed underneath
                 this.stopSleeping();
             }
         }
@@ -4482,7 +4482,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         this.setPos((double) pos.getX() + 0.5D, (double) pos.getY() + 0.6875D, (double) pos.getZ() + 0.5D);
     }
 
-    private boolean checkBedExists() {
+    public boolean checkBedExists() { // Kiterino - private -> public
         return (Boolean) this.getSleepingPos().map((blockposition) -> {
             return this.level().getBlockState(blockposition).getBlock() instanceof BedBlock;
         }).orElse(false);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index fb2d05e43df3bfb72b1f6e325736dd3cbc6c3096..557332a12d5fbac17369a5475177822b998ddbc9 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1382,7 +1382,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         if (dismount) entity.stopRiding(); // Paper - Teleport API
 
         // SPIGOT-5509: Wakeup, similar to riding
-        if (this.isSleeping()) {
+        if (this.isSleeping() && getHandle().checkBedExists()) { // Kiterino - wake up only if bed exists
             this.wakeup(false);
         }
 
