From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 18 Jul 2020 17:46:23 +0300
Subject: [PATCH] Make Pathfinders public


diff --git a/src/main/java/net/minecraft/world/entity/animal/Rabbit.java b/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
index f158cb5a8a04b9d9fa00c35774d2104a8d4e6416..c4ee232286206545dfc0119d81de96690f7125ec 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
@@ -595,7 +595,7 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
         }
     }
 
-    private static class RabbitPanicGoal extends PanicGoal {
+    public static class RabbitPanicGoal extends PanicGoal {
 
         private final Rabbit rabbit;
 
@@ -611,7 +611,7 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
         }
     }
 
-    private static class RabbitAvoidEntityGoal<T extends LivingEntity> extends AvoidEntityGoal<T> {
+    public static class RabbitAvoidEntityGoal<T extends LivingEntity> extends AvoidEntityGoal<T> {
 
         private final Rabbit rabbit;
 
@@ -626,7 +626,7 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
         }
     }
 
-    private static class RaidGardenGoal extends MoveToBlockGoal {
+    public static class RaidGardenGoal extends MoveToBlockGoal {
 
         private final Rabbit rabbit;
         private boolean wantsToRaid;
@@ -744,7 +744,7 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
         }
     }
 
-    private static class EvilRabbitAttackGoal extends MeleeAttackGoal {
+    public static class EvilRabbitAttackGoal extends MeleeAttackGoal {
 
         public EvilRabbitAttackGoal(Rabbit rabbit) {
             super(rabbit, 1.4D, true);
diff --git a/src/main/java/net/minecraft/world/entity/monster/Slime.java b/src/main/java/net/minecraft/world/entity/monster/Slime.java
index 360ad0ffd25c5d42d8d50060be40cab304c8fb32..9f7579cf27e80cf2f1b4d2184c175b7725005bdf 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Slime.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Slime.java
@@ -554,7 +554,7 @@ public class Slime extends Mob implements Enemy {
         }
     }
 
-    private static class SlimeFloatGoal extends Goal {
+    public static class SlimeFloatGoal extends Goal {
 
         private final Slime slime;
 
@@ -584,7 +584,7 @@ public class Slime extends Mob implements Enemy {
         }
     }
 
-    private static class SlimeAttackGoal extends Goal {
+    public static class SlimeAttackGoal extends Goal {
 
         private final Slime slime;
         private int growTiredTimer;
@@ -654,7 +654,7 @@ public class Slime extends Mob implements Enemy {
         // Paper end
     }
 
-    private static class SlimeRandomDirectionGoal extends Goal {
+    public static class SlimeRandomDirectionGoal extends Goal {
 
         private final Slime slime;
         private float chosenDegrees;
@@ -686,7 +686,7 @@ public class Slime extends Mob implements Enemy {
         }
     }
 
-    private static class SlimeKeepOnJumpingGoal extends Goal {
+    public static class SlimeKeepOnJumpingGoal extends Goal {
 
         private final Slime slime;
 
diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index 62996f5e65cc04744852897ecca0c83a8a2e13f2..cce0305b1d66647ed2ef94f3f4e2e8d6aae81bde 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -225,7 +225,7 @@ public class Spider extends Monster {
         return 0.65F;
     }
 
-    private static class SpiderAttackGoal extends MeleeAttackGoal {
+    public static class SpiderAttackGoal extends MeleeAttackGoal {
 
         public SpiderAttackGoal(Spider spider) {
             super(spider, 1.0D, true);
@@ -254,7 +254,7 @@ public class Spider extends Monster {
         }
     }
 
-    private static class SpiderTargetGoal<T extends LivingEntity> extends NearestAttackableTargetGoal<T> {
+    public static class SpiderTargetGoal<T extends LivingEntity> extends NearestAttackableTargetGoal<T> {
 
         public SpiderTargetGoal(Spider spider, Class<T> targetEntityClass) {
             super(spider, targetEntityClass, true);
