From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 9 Jul 2024 19:12:41 +0300
Subject: [PATCH] Data-driven mob effects


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index b2b3d8013ba6bdbd30bf2441efd069e04705cba9..469c8ce4b4ea0a9d2efc8237c7ce6c3c200dac69 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -75,7 +75,7 @@ public final class PaperRegistries {
             writable(Registries.GAME_EVENT, RegistryKey.GAME_EVENT, GameEvent.class, CraftGameEvent::new, PaperGameEventRegistryEntry.PaperBuilder::new),
             entry(Registries.STRUCTURE_TYPE, RegistryKey.STRUCTURE_TYPE, StructureType.class, CraftStructureType::new),
             entry(Registries.INSTRUMENT, RegistryKey.INSTRUMENT, MusicInstrument.class, CraftMusicInstrument::new),
-            entry(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT, PotionEffectType.class, CraftPotionEffectType::new),
+            writable(Registries.MOB_EFFECT, RegistryKey.MOB_EFFECT, PotionEffectType.class, CraftPotionEffectType::new, me.sosedik.kiterino.registry.data.KiterinoMobEffectRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven mob effects
             entry(Registries.BLOCK, RegistryKey.BLOCK, BlockType.class, CraftBlockType::new),
             writable(Registries.ITEM, RegistryKey.ITEM, ItemType.class, CraftItemType::new, me.sosedik.kiterino.registry.data.KiterinoItemRegistryEntity.KiterinoBuilder::new).delayed(), // Kiterino - Data-driven items
             entry(Registries.CAT_VARIANT, RegistryKey.CAT_VARIANT, Cat.Type.class, CraftCat.CraftType::new),
diff --git a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
index 48de73884cdece86ff19e36d930e3990066de3bf..b5d6fe97a9a9a1c7e923ec37828fda2389895a7b 100644
--- a/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
+++ b/src/main/java/io/papermc/paper/registry/data/util/Conversions.java
@@ -53,4 +53,18 @@ public class Conversions {
         };
     }
     // Kiterino end - Data-driven items
+
+    // Kiterino start - Data-driven mob effects
+    public net.minecraft.world.effect.MobEffectCategory asVanilla(final org.bukkit.potion.PotionEffectType.Category category) {
+        return switch (category) {
+            case BENEFICIAL -> net.minecraft.world.effect.MobEffectCategory.BENEFICIAL;
+            case HARMFUL -> net.minecraft.world.effect.MobEffectCategory.HARMFUL;
+            case NEUTRAL -> net.minecraft.world.effect.MobEffectCategory.NEUTRAL;
+        };
+    }
+
+    public org.bukkit.potion.PotionEffectType.Category asBukkit(final net.minecraft.world.effect.MobEffectCategory category) {
+        return org.bukkit.craftbukkit.potion.CraftPotionEffectType.fromNMS(category);
+    }
+    // Kiterino end - Data-driven mob effects
 }
diff --git a/src/main/java/me/sosedik/kiterino/registry/data/KiterinoMobEffectRegistryEntity.java b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoMobEffectRegistryEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..a83c4e4cf29e2232dedefaa02546df8d1e2e2b8c
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/registry/data/KiterinoMobEffectRegistryEntity.java
@@ -0,0 +1,86 @@
+package me.sosedik.kiterino.registry.data;
+
+import io.papermc.paper.registry.PaperRegistryBuilder;
+import io.papermc.paper.registry.TypedKey;
+import io.papermc.paper.registry.data.util.Conversions;
+import me.sosedik.kiterino.registry.wrapper.KiterinoMobEffectBehaviourWrapper;
+import me.sosedik.kiterino.world.effect.KiterinoMobEffectWrapper;
+import net.minecraft.world.effect.MobEffect;
+import net.minecraft.world.effect.MobEffectCategory;
+import org.bukkit.potion.PotionEffectType;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.jetbrains.annotations.NotNull;
+
+public class KiterinoMobEffectRegistryEntity implements MobEffectRegistryEntity {
+
+    protected KiterinoMobEffectBehaviourWrapper wrapper;
+    protected MobEffectCategory category;
+    protected int color;
+
+    protected final Conversions conversions;
+
+    public KiterinoMobEffectRegistryEntity(
+        final @NonNull Conversions conversions,
+        final @NonNull TypedKey<PotionEffectType> ignoredKey,
+        final @Nullable MobEffect internal
+    ) {
+        this.conversions = conversions;
+        if (internal == null) {
+            this.wrapper = new KiterinoMobEffectBehaviourWrapper() { };
+            this.category = MobEffectCategory.NEUTRAL;
+            return;
+        }
+
+        this.category = internal.getCategory();
+        this.color = internal.getColor();
+    }
+
+    @Override
+    public @NotNull KiterinoMobEffectBehaviourWrapper wrapper() {
+        return this.wrapper;
+    }
+
+    @Override
+    public PotionEffectType.@NonNull Category category() {
+        return conversions.asBukkit(this.category);
+    }
+
+    @Override
+    public int color() {
+        return this.color;
+    }
+
+    public static final class KiterinoBuilder extends KiterinoMobEffectRegistryEntity implements Builder,
+            PaperRegistryBuilder<MobEffect, PotionEffectType> {
+
+        public KiterinoBuilder(final @NonNull Conversions conversions, final @NonNull TypedKey<PotionEffectType> key, final @Nullable MobEffect internal) {
+            super(conversions, key, internal);
+        }
+
+        @Override
+        public @NonNull Builder wrapper(@NotNull KiterinoMobEffectBehaviourWrapper wrapper) {
+            this.wrapper = wrapper;
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder category(PotionEffectType.@NonNull Category category) {
+            this.category = conversions.asVanilla(category);
+            return this;
+        }
+
+        @Override
+        public @NonNull Builder color(int color) {
+            this.color = color;
+            return this;
+        }
+
+        @Override
+        public @NonNull MobEffect build() {
+            return new KiterinoMobEffectWrapper(this.category, this.color, this.wrapper);
+        }
+
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffectWrapper.java b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffectWrapper.java
new file mode 100644
index 0000000000000000000000000000000000000000..aa46cf2da05591f8bdcc14b4d75ddf634b5fc825
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffectWrapper.java
@@ -0,0 +1,69 @@
+package me.sosedik.kiterino.world.effect;
+
+import me.sosedik.kiterino.registry.wrapper.KiterinoMobEffectBehaviourWrapper;
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.effect.MobEffect;
+import net.minecraft.world.effect.MobEffectCategory;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import org.bukkit.craftbukkit.damage.CraftDamageSource;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class KiterinoMobEffectWrapper extends MobEffect {
+
+    private final KiterinoMobEffectBehaviourWrapper mobEffect;
+
+    public KiterinoMobEffectWrapper(@NotNull MobEffectCategory category, int color, @NotNull KiterinoMobEffectBehaviourWrapper mobEffect) {
+        super(category, color);
+        this.mobEffect = mobEffect;
+    }
+
+    @Override
+    public int getBlendDurationTicks() {
+        return this.mobEffect.getBlendDurationTicks();
+    }
+
+    @Override
+    public boolean applyEffectTick(@NotNull LivingEntity entity, int amplifier) {
+        return this.mobEffect.applyEffectTick(entity.getBukkitLivingEntity(), amplifier);
+    }
+
+    @Override
+    public void applyInstantenousEffect(@Nullable Entity source, @Nullable Entity attacker, @NotNull LivingEntity target, int amplifier, double proximity) {
+        this.mobEffect.applyInstantaneousEffect(source == null ? null : source.getBukkitEntity(), attacker == null ? null : attacker.getBukkitEntity(), target.getBukkitLivingEntity(), amplifier, proximity);
+    }
+
+    @Override
+    public boolean shouldApplyEffectTickThisTick(int duration, int amplifier) {
+        return this.mobEffect.shouldApplyEffectTickThisTick(duration, amplifier);
+    }
+
+    @Override
+    public void onEffectStarted(@NotNull LivingEntity entity, int amplifier) {
+        this.mobEffect.onEffectStarted(entity.getBukkitLivingEntity(), amplifier);
+    }
+
+    @Override
+    public void onEffectAdded(@NotNull LivingEntity entity, int amplifier) {
+        this.mobEffect.onEffectAdded(entity.getBukkitLivingEntity(), amplifier);
+    }
+
+    @Override
+    public void onMobRemoved(@NotNull LivingEntity entity, int amplifier, @NotNull Entity.RemovalReason reason) {
+        if (reason == Entity.RemovalReason.KILLED) {
+            this.mobEffect.onMobKilled(entity.getBukkitLivingEntity(), amplifier);
+        }
+    }
+
+    @Override
+    public void onMobHurt(@NotNull LivingEntity entity, int amplifier, @NotNull DamageSource source, float amount) {
+        this.mobEffect.onMobHurt(entity.getBukkitLivingEntity(), amplifier, new CraftDamageSource(source), amount);
+    }
+
+    @Override
+    public boolean isInstantenous() {
+        return this.mobEffect.isInstantaneous();
+    }
+
+}
