From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 14:32:55 +0300
Subject: [PATCH] Allow force sleeping


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index c9a8556d0aa832f05068c8e14343450d3be159bd..1403db4f224007382c3c81a0791cb4278a7ef55f 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -537,7 +537,7 @@ public class ServerPlayer extends Player {
         }
         this.getBukkitEntity().readExtraData(nbt); // CraftBukkit
 
-        if (this.isSleeping()) {
+        if (this.isSleeping() && this.checkBedExists()) { // Kiterino - wake up only if bed exists
             this.stopSleeping();
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 6d598d5cadbe6f964f0867e17d8fb8f7e132ba82..3d71c9ed59dde2f91ded8706fb82d14dd76636ac 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3089,7 +3089,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 this.getCombatTracker().recheckStatus();
             }
 
-            if (this.isSleeping() && !this.checkBedExists()) {
+            if (false && this.isSleeping() && !this.checkBedExists()) { // Kiterino - do not wake up automatically even if no bed underneath
                 this.stopSleeping();
             }
         }
@@ -4380,7 +4380,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         this.setPos((double) pos.getX() + 0.5D, (double) pos.getY() + 0.6875D, (double) pos.getZ() + 0.5D);
     }
 
-    private boolean checkBedExists() {
+    public boolean checkBedExists() { // Kiterino - private -> public
         return (Boolean) this.getSleepingPos().map((blockposition) -> {
             return this.level().getBlockState(blockposition).getBlock() instanceof BedBlock;
         }).orElse(false);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index e21cd8c4015cd726bb10d51545d52c3010381848..76053a68bffeaaad36d71e559a8c7c060f533115 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1370,7 +1370,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         if (dismount) entity.stopRiding(); // Paper - Teleport API
 
         // SPIGOT-5509: Wakeup, similar to riding
-        if (this.isSleeping()) {
+        if (this.isSleeping() && getHandle().checkBedExists()) { // Kiterino - wake up only if bed exists
             this.wakeup(false);
         }
 
