From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 15 Jun 2024 03:26:19 +0300
Subject: [PATCH] Material rerouting


diff --git a/src/main/java/org/bukkit/craftbukkit/legacy/MaterialRerouting.java b/src/main/java/org/bukkit/craftbukkit/legacy/MaterialRerouting.java
index e5c30847297e056782084d81fb9300f98d4a8f75..8f465bc9f03bc5fab80674bc47552f8325067c1d 100644
--- a/src/main/java/org/bukkit/craftbukkit/legacy/MaterialRerouting.java
+++ b/src/main/java/org/bukkit/craftbukkit/legacy/MaterialRerouting.java
@@ -736,4 +736,27 @@ public class MaterialRerouting {
         return itemStack.getBlockData(MaterialRerouting.transformToBlockType(material));
     }
     // Purpur end
+
+    // Kiterino start - Material rerouting
+    @RerouteStatic("org/bukkit/inventory/ItemStack")
+    public static boolean isType(ItemStack item, Material type) {
+        return ItemStack.isType(item, type);
+    }
+
+    public static Material getMaterial(me.sosedik.kiterino.modifier.item.ItemContextBox contextBox) {
+        return contextBox.getMaterial();
+    }
+
+    public static Material getNewMaterial(me.sosedik.kiterino.modifier.item.ItemContextBox contextBox) {
+        return contextBox.getNewMaterial();
+    }
+
+    public static void setNewMaterial(me.sosedik.kiterino.modifier.item.ItemContextBox contextBox, Material type) {
+        contextBox.setNewMaterial(type);
+    }
+
+    public static void setNewMaterial(me.sosedik.kiterino.modifier.item.ItemContextBox contextBox, Material type, boolean refreshMeta) {
+        contextBox.setNewMaterial(type, refreshMeta);
+    }
+    // Kiterino end - Material rerouting
 }
diff --git a/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java b/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java
index 0fac826b9367a821c6801190997592219cb47f73..85d7e19b8614e9e7d472c196f997bfbcdd1b3741 100644
--- a/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java
+++ b/src/test/java/org/bukkit/craftbukkit/legacy/MaterialReroutingTest.java
@@ -59,6 +59,11 @@ public class MaterialReroutingTest extends AbstractTestingBase {
                 // Paper start - types that cannot be translated to ItemType/BlockType
                 .filter(entry -> !entry.getName().equals("com/destroystokyo/paper/MaterialSetTag.class"))
                 // Paper end - types that cannot be translated to ItemType/BlockType
+                // Kiterino start - Material rerouting
+                .filter(entry -> !entry.getName().equals("me/sosedik/kiterino/util/IKiterinoBootstrapMaterialInjector.class"))
+                .filter(entry -> !entry.getName().equals("me/sosedik/kiterino/util/KiterinoBootstrapMaterialInjector.class"))
+                .filter(entry -> !entry.getName().equals("me/sosedik/kiterino/world/item/KiterinoItem.class"))
+                // Kiterino end - Material rerouting
                 .map(entry -> {
                     try {
                         return MaterialReroutingTest.jarFile.getInputStream(entry);
