From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 18:16:41 +0300
Subject: [PATCH] All mobs have default attack damage


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 503f4d499d28ee30a4d64c55a980e006860b6a83..929f17b4e34ff3217a381075853e8e232dfd3bf9 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -342,7 +342,7 @@ public abstract class LivingEntity extends Entity {
     }
 
     public static AttributeSupplier.Builder createLivingAttributes() {
-        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS);
+        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS).add(Attributes.ATTACK_DAMAGE); // Kiterino - add ATTACK_DAMAGE
     }
     public boolean shouldSendAttribute(Attribute attribute) { return true; } // Purpur
 
@@ -2985,7 +2985,12 @@ public abstract class LivingEntity extends Entity {
 
     public boolean doHurtTarget(Entity target) {
         this.setLastHurtMob(target);
-        return false;
+        // Kiterino start - make all mobs able to attack
+        boolean flag = target.hurt(DamageSource.mobAttack(this), (float) ((int) getAttribute(Attributes.ATTACK_DAMAGE).getValue()));
+        if (flag)
+            doEnchantDamageEffects(this, target);
+        return flag;
+        // Kiterino end
     }
 
     @Override
