From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 12 Jul 2021 19:26:22 +0300
Subject: [PATCH] Add ability to get active container id


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 5fc40c8ee24eb4aa2ed81ebf1b6777c49239dadb..2061aacb44d57b7f06a213e9c47a7d596fe44058 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -884,4 +884,16 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     public org.bukkit.entity.Entity copy(Location location) {
         throw new UnsupportedOperationException("Cannot copy human entities");
     }
+
+    // Kiterino start
+    @Override
+    public int getActiveContainerId() {
+        return getHandle().containerMenu.containerId;
+    }
+
+    @Override
+    public int getActiveContainerStateId() {
+        return getHandle().containerMenu.getStateId();
+    }
+    // Kiterino end
 }
