From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 12 Jul 2021 19:26:22 +0300
Subject: [PATCH] Add ability to get active container id


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 670a957f596ee97cce92d65e20420253ad97f7f5..1f894c8515df8f96fc02ef5175bb05d4066f7c8b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -867,4 +867,16 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         boolean success = this.getHandle().level.addFreshEntity(fireworks, SpawnReason.CUSTOM);
         return success ? (Firework) fireworks.getBukkitEntity() : null;
     }
+
+    // Kiterino start
+    @Override
+    public int getActiveContainerId() {
+        return getHandle().containerMenu.containerId;
+    }
+
+    @Override
+    public int getActiveContainerStateId() {
+        return getHandle().containerMenu.getStateId();
+    }
+    // Kiterino end
 }
