From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 12 Jul 2021 19:26:22 +0300
Subject: [PATCH] Add ability to get active container id


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 477ccdc6d9ac86678841435fedf73258b4792a67..ebd6221f4b57558ce7a550358591e75c9044f0dd 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -868,4 +868,16 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         boolean success = this.getHandle().level.addFreshEntity(fireworks, SpawnReason.CUSTOM);
         return success ? (Firework) fireworks.getBukkitEntity() : null;
     }
+
+    // Kiterino start
+    @Override
+    public int getActiveContainerId() {
+        return getHandle().containerMenu.containerId;
+    }
+
+    @Override
+    public int getActiveContainerStateId() {
+        return getHandle().containerMenu.getStateId();
+    }
+    // Kiterino end
 }
