From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 30 Jan 2021 18:55:38 +0200
Subject: [PATCH] Allow modifying vanilla block drops


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 81f5ee5154ea24961b1a88cca505a5a025c430fb..44bb816a12996438da862508666c3d4f7a87709a 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -409,6 +409,21 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
 
         if (!event.isCancelled()) {
+            // Kiterino start
+            if (event.changeDrops()) {
+                for (Item bukkit : list) {
+                    if (bukkit.isValid()) {
+                        bukkit.remove();
+                    }
+                }
+                org.bukkit.World world = block.getWorld();
+                Location loc = block.getLocation();
+                for (org.bukkit.inventory.ItemStack drop : event.getDrops()) {
+                    world.dropItemNaturally(loc, drop);
+                }
+                return;
+            }
+            // Kiterino end
             // Paper start
             for (Item bukkit : list) {
                 if (!bukkit.isValid()) {
