From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 20 Oct 2022 15:55:19 +0300
Subject: [PATCH] Update goal target when changing target


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
index c8f0f61c921f12361df6661c1f33223bf4d0127e..4dea916bbbde65fd451f1c958f8caabe88710edf 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -146,4 +146,10 @@ public abstract class TargetGoal extends Goal {
         this.unseenMemoryTicks = time;
         return this;
     }
+
+    // Kiterino start - Update goal target when changing target
+    public void setTargetMob(@Nullable LivingEntity targetMob) {
+        this.targetMob = targetMob;
+    }
+    // Kiterino end - Update goal target when changing target
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index 778a9d3f8bfe5dba59e1e655e4eeb8822678b8cf..f3f0411b69641a4e5524ef9b9e188f7a237e1719 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -29,6 +29,17 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob, io.pape
         } else if (target instanceof CraftLivingEntity) {
             entity.setTarget(((CraftLivingEntity) target).getHandle(), null, false);
         }
+        // Kiterino start - Update goal target when changing target
+        entity.targetSelector.getAvailableGoals().forEach(goal -> {
+            if (goal.getGoal() instanceof net.minecraft.world.entity.ai.goal.target.TargetGoal targetGoal) {
+                if (target == null) {
+                    targetGoal.setTargetMob(null);
+                } else if (target instanceof CraftLivingEntity craftLivingEntity) {
+                    targetGoal.setTargetMob(craftLivingEntity.getHandle());
+                }
+            }
+        });
+        // Kiterino end - Update goal target when changing target
     }
 
     @Override
