From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 31 Jul 2021 22:21:04 +0300
Subject: [PATCH] Packet Additions


diff --git a/src/main/java/com/github/weepingmc/packet/CraftPacketStepBuilder.java b/src/main/java/com/github/weepingmc/packet/CraftPacketStepBuilder.java
index 34c1a862d7dd8b710dd253047b34790bc20a8c5a..e78beb67a0c87862bf496d1e995b71a68e8e1412 100644
--- a/src/main/java/com/github/weepingmc/packet/CraftPacketStepBuilder.java
+++ b/src/main/java/com/github/weepingmc/packet/CraftPacketStepBuilder.java
@@ -102,6 +102,25 @@ public class CraftPacketStepBuilder implements PacketStepBuilder {
 		return this;
 	}
 
+	// Kiterino start
+	@Override
+	public @NotNull PacketStepBuilder sendPlayerProfile(@Nonnull PlayerProfile playerProfile, @Nonnull ProfileAction profileAction, @Nullable net.kyori.adventure.text.Component displayName, @Nullable org.bukkit.GameMode gameMode, boolean listed) {
+		GameProfile gameProfile = ((CraftPlayerProfile) playerProfile).getGameProfile();
+		var playerUpdate = new ClientboundPlayerInfoUpdatePacket.Entry(
+				gameProfile.getId(),
+				gameProfile,
+				listed,
+				0,
+				gameMode == null ? GameType.DEFAULT_MODE : GameType.byId(gameMode.getValue()),
+				displayName == null ? null : io.papermc.paper.adventure.PaperAdventure.asVanilla(displayName),
+				null
+		);
+		var clientboundPlayerInfoPacket = new ClientboundPlayerInfoUpdatePacket(from(profileAction), Collections.singletonList(playerUpdate));
+		initial.setNext(of(clientboundPlayerInfoPacket));
+		return this;
+	}
+	// Kiterino end
+
 	@Override
 	@Nonnull
 	public PacketStepBuilder spawnPlayer(int entityId, @Nonnull UUID uuid, @Nonnull Location location) {
