From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 12 Jul 2021 19:26:22 +0300
Subject: [PATCH] Add ability to get active container id


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 849ad51602983b4c8f7956337643c682c307d781..cd7ba5cf4e862220ba2fa459c405595b52891c47 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -853,4 +853,16 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         boolean success = this.getHandle().level.addFreshEntity(fireworks, SpawnReason.CUSTOM);
         return success ? (Firework) fireworks.getBukkitEntity() : null;
     }
+
+    // Kiterino start
+    @Override
+    public int getActiveContainerId() {
+        return getHandle().containerMenu.containerId;
+    }
+
+    @Override
+    public int getActiveContainerStateId() {
+        return getHandle().containerMenu.getStateId();
+    }
+    // Kiterino end
 }
