From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 19 Jun 2024 01:40:23 +0300
Subject: [PATCH] Temporary solution for respawn events parity until Paper PR
 8705 is merged


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index e4978abcc79ee51fa95442b82d9079c8ee1e60d6..1db55b8ea73f6eab8dd38bfc5104d7f520bd533f 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -843,6 +843,7 @@ public abstract class PlayerList {
         }
         // Paper start - Add PlayerPostRespawnEvent
         boolean isBedSpawn = false;
+        boolean isAnchorSpawn = false; // Kiterino - Parity for respawn events
         boolean isRespawn = false;
         // Paper end - Add PlayerPostRespawnEvent
 
@@ -906,6 +907,7 @@ public abstract class PlayerList {
             BlockState iblockdata = worldserver.getBlockState(blockposition);
 
             if (iblockdata.is(Blocks.RESPAWN_ANCHOR)) {
+                isAnchorSpawn = true; // Kiterino - Parity for respawn events
                 entityplayer1.connection.send(new ClientboundSoundPacket(SoundEvents.RESPAWN_ANCHOR_DEPLETE, SoundSource.BLOCKS, (double) blockposition.getX(), (double) blockposition.getY(), (double) blockposition.getZ(), 1.0F, 1.0F, worldserver.getRandom().nextLong()));
             }
             // Paper start - Add PlayerPostRespawnEvent
@@ -937,7 +939,13 @@ public abstract class PlayerList {
 
         // Paper start - Add PlayerPostRespawnEvent
         if (isRespawn) {
-            cserver.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerPostRespawnEvent(entityplayer.getBukkitEntity(), location, isBedSpawn));
+            // Kiterino start - Parity for respawn events
+            var flags = com.google.common.collect.ImmutableSet.<org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag>builder();
+            if (reason == RespawnReason.END_PORTAL) {
+                flags.add(PlayerRespawnEvent.RespawnFlag.END_PORTAL);
+            }
+            cserver.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerPostRespawnEvent(entityplayer.getBukkitEntity(), location, isBedSpawn && !isAnchorSpawn, isAnchorSpawn, reason, flags));
+            // Kiterino end - Parity for respawn events
         }
         // Paper end - Add PlayerPostRespawnEvent
 
