From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 14 Dec 2021 14:36:03 +0100
Subject: [PATCH] Add player to NotePlayEvent


diff --git a/src/main/java/net/minecraft/world/level/block/NoteBlock.java b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
index ac88cb949808752a340637ffcb9d7256b7d09478..abb61fdcf4139c11e37efab1c270405d19c507f7 100644
--- a/src/main/java/net/minecraft/world/level/block/NoteBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
@@ -40,6 +40,8 @@ public class NoteBlock extends Block {
     public static final IntegerProperty NOTE = BlockStateProperties.NOTE;
     public static final int NOTE_VOLUME = 3;
 
+    private @Nullable Player lastPlayedBy = null; // Kiterino
+
     @Override
     public MapCodec<NoteBlock> codec() {
         return NoteBlock.CODEC;
@@ -141,7 +143,8 @@ public class NoteBlock extends Block {
     public boolean triggerEvent(BlockState state, Level world, BlockPos pos, int type, int data) {
         NoteBlockInstrument blockpropertyinstrument = (NoteBlockInstrument) state.getValue(NoteBlock.INSTRUMENT);
         // Paper start - move NotePlayEvent call to fix instrument/note changes
-        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, blockpropertyinstrument, state.getValue(NOTE));
+        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, blockpropertyinstrument, state.getValue(NOTE), lastPlayedBy); // Kiterino - Add player param
+        lastPlayedBy = null; // Kiterino
         if (event.isCancelled()) return false;
         // Paper end - move NotePlayEvent call to fix instrument/note changes
         float f;
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 45b5c7c8c9e231ed02f32ff96fc609ee8e8ac2be..9fcf4976a1fad9c2517ebd06279d6973c1d620ac 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1558,7 +1558,12 @@ public class CraftEventFactory {
     }
 
     public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note) {
-        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note));
+        // Kiterino start
+        return callNotePlayEvent(world, pos, instrument, note, null);
+    }
+    public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note, net.minecraft.world.entity.player.Player player) { // Kiterino - Add player param
+        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note), player instanceof ServerPlayer serverPlayer ? serverPlayer.getBukkitEntity() : null); // Kiterino
+        // Kiterino end
         world.getCraftServer().getPluginManager().callEvent(event);
         return event;
     }
