From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 14 Dec 2021 14:36:03 +0100
Subject: [PATCH] Add player to NotePlayEvent


diff --git a/src/main/java/net/minecraft/world/level/block/NoteBlock.java b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
index c11752564ea48960232844ee735779aa95d82c12..f0ed3e751b12baa98c5289a426ec13f648914df6 100644
--- a/src/main/java/net/minecraft/world/level/block/NoteBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
@@ -39,6 +39,8 @@ public class NoteBlock extends Block {
     public static final IntegerProperty NOTE = BlockStateProperties.NOTE;
     public static final int NOTE_VOLUME = 3;
 
+    private @Nullable Player lastPlayedBy = null; // Kiterino
+
     public NoteBlock(BlockBehaviour.Properties settings) {
         super(settings);
         this.registerDefaultState((BlockState) ((BlockState) ((BlockState) ((BlockState) this.stateDefinition.any()).setValue(NoteBlock.INSTRUMENT, NoteBlockInstrument.HARP)).setValue(NoteBlock.NOTE, 0)).setValue(NoteBlock.POWERED, false));
@@ -138,7 +140,8 @@ public class NoteBlock extends Block {
     public boolean triggerEvent(BlockState state, Level world, BlockPos pos, int type, int data) {
         NoteBlockInstrument blockpropertyinstrument = (NoteBlockInstrument) state.getValue(NoteBlock.INSTRUMENT);
         // Paper start - move NotePlayEvent call to fix instrument/note changes
-        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, blockpropertyinstrument, state.getValue(NOTE));
+        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, blockpropertyinstrument, state.getValue(NOTE), lastPlayedBy); // Kiterino - Add player param
+        lastPlayedBy = null; // Kiterino
         if (event.isCancelled()) return false;
         // Paper end
         float f;
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index cda19cd8d9e1beda1553337e9989c8839b75040e..2cf38f6aba34116df539d47fb337809da76126ce 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1511,7 +1511,12 @@ public class CraftEventFactory {
     }
 
     public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note) {
-        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note));
+        // Kiterino start
+        return callNotePlayEvent(world, pos, instrument, note, null);
+    }
+    public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note, net.minecraft.world.entity.player.Player player) { // Kiterino - Add player param
+        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note), player instanceof ServerPlayer serverPlayer ? serverPlayer.getBukkitEntity() : null); // Kiterino
+        // Kiterino end
         world.getCraftServer().getPluginManager().callEvent(event);
         return event;
     }
