From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 14 Dec 2021 14:36:03 +0100
Subject: [PATCH] Add player to NotePlayEvent


diff --git a/src/main/java/net/minecraft/world/level/block/NoteBlock.java b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
index 6343cd0c33cafb30225cfae17ea1cf15859073b1..6e9ff01266fcb3936169782e9c4063a50ce41769 100644
--- a/src/main/java/net/minecraft/world/level/block/NoteBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
@@ -38,6 +38,8 @@ public class NoteBlock extends Block {
     public static final IntegerProperty NOTE = BlockStateProperties.NOTE;
     public static final int NOTE_VOLUME = 3;
 
+    private @Nullable Player lastPlayedBy = null; // Kiterino
+
     public NoteBlock(BlockBehaviour.Properties settings) {
         super(settings);
         this.registerDefaultState((BlockState) ((BlockState) ((BlockState) ((BlockState) this.stateDefinition.any()).setValue(NoteBlock.INSTRUMENT, NoteBlockInstrument.HARP)).setValue(NoteBlock.NOTE, 0)).setValue(NoteBlock.POWERED, false));
@@ -134,7 +136,8 @@ public class NoteBlock extends Block {
     public boolean triggerEvent(BlockState state, Level world, BlockPos pos, int type, int data) {
         NoteBlockInstrument blockpropertyinstrument = (NoteBlockInstrument) state.getValue(NoteBlock.INSTRUMENT);
         // Paper start - move NotePlayEvent call to fix instrument/note changes
-        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, blockpropertyinstrument, state.getValue(NOTE));
+        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, blockpropertyinstrument, state.getValue(NOTE), lastPlayedBy); // Kiterino - Add player param
+        lastPlayedBy = null; // Kiterino
         if (event.isCancelled()) return false;
         // Paper end
         float f;
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 2b42b33e71fa8b68fe8e2f9e60e74602ad4d4dd7..6e76890a81a524fc821c9d1c8412eb5ef9788615 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1520,7 +1520,12 @@ public class CraftEventFactory {
     }
 
     public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note) {
-        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note));
+        // Kiterino start
+        return callNotePlayEvent(world, pos, instrument, note, null);
+    }
+    public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note, net.minecraft.world.entity.player.Player player) { // Kiterino - Add player param
+        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note), player instanceof ServerPlayer serverPlayer ? serverPlayer.getBukkitEntity() : null); // Kiterino
+        // Kiterino end
         world.getCraftServer().getPluginManager().callEvent(event);
         return event;
     }
