From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 14:48:33 +0300
Subject: [PATCH] Allow sleeping anytime and anywhere


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 540ce36dc73e81f8a400f2147ce90c3a7460849e..43fac3f94978032f33fa711ec17e4c9ad6467991 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -670,8 +670,19 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 if (!event.isCancelled()) {
                     this.setDayTime(this.getDayTime() + event.getSkipAmount());
                 }
+
+                // Kiterino start
+                if (!event.isCancelled()) {
+                    this.wakeUpAllPlayers();
+                }
+                // CraftBukkit end
+                if (this.getGameRules().getBoolean(GameRules.RULE_WEATHER_CYCLE)) {
+                    this.resetWeatherCycle();
+                }
+                // Kiterino end
             }
 
+            /* Kiterino start - moved up
             if (!event.isCancelled()) {
                 this.wakeUpAllPlayers();
             }
@@ -679,6 +690,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
             if (this.getGameRules().getBoolean(GameRules.RULE_WEATHER_CYCLE) && this.isRaining()) {
                 this.resetWeatherCycle();
             }
+            */ // Kiterino end
         }
 
         this.updateSkyBrightness();
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 12637019da44019475042e4cf88d3e0ebd0fd068..9bab7ed78d089d02f9b8bcd607cc7d357f67babc 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1322,6 +1322,7 @@ public class ServerPlayer extends Player {
                 return Either.left(Player.BedSleepingProblem.OBSTRUCTED);
             } else {
                 this.setRespawnPosition(this.level.dimension(), blockposition, this.getYRot(), false, true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.BED); // Paper - PlayerSetSpawnEvent
+                if (true) return Either.right(Unit.INSTANCE); // Kiterino - allow sleeping anytime
                 if (this.level.isDay()) {
                     return Either.left(Player.BedSleepingProblem.NOT_POSSIBLE_NOW);
                 } else {
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 64ca2c92c4908bbb0272d1c22870af68aa9daad2..a6817872574afede2757151f45200946b9a5e58e 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -294,7 +294,7 @@ public abstract class Player extends LivingEntity {
                 this.sleepCounter = 100;
             }
 
-            if (!this.level.isClientSide && this.level.isDay()) {
+            if (false && !this.level.isClientSide && this.level.isDay()) { // Kiterino - allow sleeping anytime and anywhere
                 this.stopSleepInBed(false, true);
             }
         } else if (this.sleepCounter > 0) {
