From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Jun 2022 00:12:29 +0300
Subject: [PATCH] Add PlayerLoadsProjectileEvent


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index d19b1e127a1018a798eed181d8060a781b589f90..0c9d859f7cd57352e034dc6c089639c636a3d110 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -2320,7 +2320,7 @@ public abstract class Player extends LivingEntity {
             ItemStack itemstack1 = ProjectileWeaponItem.getHeldProjectile(this, predicate);
 
             if (!itemstack1.isEmpty()) {
-                return itemstack1;
+                return getProjectile(stack, itemstack1); // Kiterino - PlayerLoadProjectileEvent
             } else {
                 predicate = ((ProjectileWeaponItem) stack.getItem()).getAllSupportedProjectiles().and(item -> tryReadyArrow(stack, item)); // Paper - PlayerReadyArrowEvent
 
@@ -2328,15 +2328,33 @@ public abstract class Player extends LivingEntity {
                     ItemStack itemstack2 = this.inventory.getItem(i);
 
                     if (predicate.test(itemstack2)) {
-                        return itemstack2;
+                        return getProjectile(stack, itemstack2); // Kiterino - PlayerLoadProjectileEvent
                     }
                 }
 
-                return this.abilities.instabuild ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+                return getProjectile(stack, this.abilities.instabuild ? new ItemStack(Items.ARROW) : ItemStack.EMPTY); // Kiterino - PlayerLoadProjectileEvent
             }
         }
     }
 
+    // Kiterino start - PlayerLoadProjectileEvent
+    private ItemStack getProjectile(ItemStack weapon, ItemStack projectile) {
+        if (!(getBukkitEntity() instanceof org.bukkit.entity.Player player)) {
+            return projectile;
+        }
+        var event = new me.sosedik.kiterino.event.player.PlayerLoadsProjectileEvent(player, weapon.asBukkitMirror(), projectile.isEmpty() ? null : projectile.asBukkitMirror());
+        event.callEvent();
+        if (!event.isFiringAllowed()) {
+            return ItemStack.EMPTY;
+        }
+        var item = event.getProjectile();
+        if (item == null) {
+            return new ItemStack(Items.ARROW);
+        }
+        return item instanceof org.bukkit.craftbukkit.inventory.CraftItemStack craftItem ? craftItem.handle : org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(item);
+    }
+    // Kiterino end - PlayerLoadProjectileEvent
+
     @Override
     public Vec3 getRopeHoldPosition(float delta) {
         double d0 = 0.22D * (this.getMainArm() == HumanoidArm.RIGHT ? -1.0D : 1.0D);
