From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 21 Jun 2024 18:27:31 +0300
Subject: [PATCH] Allow jumping in water as Drowned


diff --git a/src/main/java/net/minecraft/world/entity/monster/Drowned.java b/src/main/java/net/minecraft/world/entity/monster/Drowned.java
index a28a4cd9544ffa58bd03ac901a761c797044629a..af95874f8b6756aca21e2f6549749cdd524ac60c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Drowned.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Drowned.java
@@ -92,6 +92,17 @@ public class Drowned extends Zombie implements RangedAttackMob {
     }
     // Purpur end
 
+    // Kiterino start - Allow jumping in water as Drowned
+    @Override
+    public boolean onSpacebar() {
+        if (isInWater()) {
+            jumpFromGround();
+            return true;
+        }
+        return false;
+    }
+    // Kiterino end - Allow jumping in water as Drowned
+
     @Override
     public void initAttributes() {
         this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.drownedMaxHealth);
