From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 14 Dec 2021 14:36:03 +0100
Subject: [PATCH] Add player to NotePlayEvent


diff --git a/src/main/java/net/minecraft/world/level/block/NoteBlock.java b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
index 443bf39734ba2b0a7cdf0362d415ff176b2e8c85..da181d65a146abd757e6c5dfb430550dc9342911 100644
--- a/src/main/java/net/minecraft/world/level/block/NoteBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
@@ -29,6 +29,7 @@ public class NoteBlock extends Block {
     public static final EnumProperty<NoteBlockInstrument> INSTRUMENT = BlockStateProperties.NOTEBLOCK_INSTRUMENT;
     public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
     public static final IntegerProperty NOTE = BlockStateProperties.NOTE;
+    private @Nullable Player lastPlayedBy = null; // Kiterino
 
     public NoteBlock(BlockBehaviour.Properties settings) {
         super(settings);
@@ -102,7 +103,8 @@ public class NoteBlock extends Block {
     @Override
     public boolean triggerEvent(BlockState state, Level world, BlockPos pos, int type, int data) {
         // Paper start - move NotePlayEvent call to fix instrument/note changes
-        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, state.getValue(INSTRUMENT), state.getValue(NOTE));
+        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(world, pos, state.getValue(INSTRUMENT), state.getValue(NOTE), lastPlayedBy); // Kiterino - Add player param
+        lastPlayedBy = null; // Kiterino
         if (event.isCancelled()) return false;
         int k = event.getNote().getId();
         float f = (float) Math.pow(2.0D, (double) (k - 12) / 12.0D);
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 44bb816a12996438da862508666c3d4f7a87709a..f28039de0b836998af9c3b572e65172bab85c813 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1423,7 +1423,12 @@ public class CraftEventFactory {
     }
 
     public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note) {
-        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note));
+        // Kiterino start
+        return callNotePlayEvent(world, pos, instrument, note, null);
+    }
+
+    public static NotePlayEvent callNotePlayEvent(Level world, BlockPos pos, NoteBlockInstrument instrument, int note, net.minecraft.world.entity.player.Player player) { // Kiterino - Add player param
+        NotePlayEvent event = new NotePlayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.Instrument.getByType((byte) instrument.ordinal()), new org.bukkit.Note(note), player instanceof ServerPlayer serverPlayer ? serverPlayer.getBukkitEntity() : null); // Kiterino
         world.getCraftServer().getPluginManager().callEvent(event);
         return event;
     }
