From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 20 Feb 2021 02:05:32 +0200
Subject: [PATCH] Don't edit signs upon placing


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index fa71d98b4496147ea3a30b99cee48057cf237013..c77dc078acb439fe56af537e1aa1e531e0449589 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -487,7 +487,7 @@ public final class ItemStack {
                     // SPIGOT-4678
                     if ((this.item instanceof SignItem || this.item instanceof HangingSignItem) && SignItem.openSign != null) { // Paper - trigger the hanging sign text editor now
                         try {
-                            entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign));
+                            //entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign)); // Kiterino
                         } finally {
                             SignItem.openSign = null;
                         }
