From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 20 Feb 2021 02:05:32 +0200
Subject: [PATCH] Don't edit signs upon placing


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 9e78c913da090d351d3dbaff3eff7d19fb18c8e4..5e773c76045358b4b97df3aad6b232b54761845f 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -496,7 +496,7 @@ public final class ItemStack {
                         try {
                             if (world.getBlockEntity(SignItem.openSign) instanceof SignBlockEntity tileentitysign) {
                                 if (world.getBlockState(SignItem.openSign).getBlock() instanceof SignBlock blocksign) {
-                                    blocksign.openTextEdit(entityhuman, tileentitysign, true);
+                                    //blocksign.openTextEdit(entityhuman, tileentitysign, true); // Kiterino
                                 }
                             }
                         } finally {
