From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 20 Feb 2021 02:05:32 +0200
Subject: [PATCH] Don't edit signs upon placing


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index e2b341ecddb946e9c1e546b9706d86dd030db865..f243af1ab02558f58f1184c5bbbc8697f762ede5 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -478,7 +478,7 @@ public final class ItemStack {
                     // SPIGOT-4678
                     if (this.item instanceof SignItem && SignItem.openSign != null) {
                         try {
-                            entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign));
+                            //entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign)); // Kiterino
                         } finally {
                             SignItem.openSign = null;
                         }
