From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 10 Sep 2020 19:28:30 +0300
Subject: [PATCH] Reset fire ticks when having fire resistance


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index cc1358e0aa6df446630ec663e2309e6ec4642e6d..9eba99aa50ed7a499425d47b96061f58563a62e3 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -945,6 +945,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 } else if (mobeffect.getDuration() % 600 == 0) {
                     this.onEffectUpdated(mobeffect, false, (Entity) null);
                 }
+                if (mobeffect.getEffect() == MobEffects.FIRE_RESISTANCE && getRemainingFireTicks() > 0) setRemainingFireTicks(0); // Kiterino
             }
         } catch (ConcurrentModificationException concurrentmodificationexception) {
             ;
