From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 10 Sep 2020 19:28:30 +0300
Subject: [PATCH] Reset fire ticks when having fire resistance


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 35f30dd1bd078ffbf99e8ad3a6b4912d4d3ec69b..8d1963bc2fc951eb6b8f05126cbbe56f48960be0 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -944,6 +944,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 } else if (mobeffect.getDuration() % 600 == 0) {
                     this.onEffectUpdated(mobeffect, false, (Entity) null);
                 }
+                if (mobeffect.getEffect() == MobEffects.FIRE_RESISTANCE && getRemainingFireTicks() > 0) setRemainingFireTicks(0); // Kiterino
             }
         } catch (ConcurrentModificationException concurrentmodificationexception) {
             ;
