From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 20 Apr 2022 15:01:32 +0300
Subject: [PATCH] Expose pre-collision move vector to
 VehicleBlockCollisionEvent


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index e76846a8e4e7315e6aef008302311c765a412e43..a5ce6b9ae6056d08ba58b3884e92b47a7cb667e4 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1139,7 +1139,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                     }
 
                     if (!bl.getType().isAir()) {
-                        VehicleBlockCollisionEvent event = new VehicleBlockCollisionEvent(vehicle, bl);
+                        VehicleBlockCollisionEvent event = new VehicleBlockCollisionEvent(vehicle, bl, org.bukkit.craftbukkit.util.CraftVector.toBukkit(moveVector)); // Kiterino - Expose pre-collision move vector
                         this.level.getCraftServer().getPluginManager().callEvent(event);
                     }
                 }
