From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 1 Jul 2020 13:21:09 +0300
Subject: [PATCH] Allow changing bed occupying state


diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftBed.java b/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftBed.java
index c892c0e37bd4f9b208147e2df16fc9ed024eacf7..a4e68c4fd051331a9d8cf2477a4f0a7a059b9869 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftBed.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftBed.java
@@ -22,4 +22,11 @@ public abstract class CraftBed extends CraftBlockData implements Bed {
     public boolean isOccupied() {
         return get(CraftBed.OCCUPIED);
     }
+
+    // Kiterino start
+    @Override
+    public void setOccupied(boolean occupied) {
+        set(OCCUPIED, occupied);
+    }
+    // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/impl/CraftBed.java b/src/main/java/org/bukkit/craftbukkit/block/impl/CraftBed.java
index 66a183fa0af6978dd220a3dde27255ad45e47137..80435f5ba3abdc3193e088dc47de06e083c4e455 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/impl/CraftBed.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/impl/CraftBed.java
@@ -33,6 +33,13 @@ public final class CraftBed extends org.bukkit.craftbukkit.block.data.CraftBlock
         return get(CraftBed.OCCUPIED);
     }
 
+    // Kiterino start
+    @Override
+    public void setOccupied(boolean occupied) {
+        set(OCCUPIED, occupied);
+    }
+    // Kiterino end
+
     // org.bukkit.craftbukkit.block.data.CraftDirectional
 
     private static final net.minecraft.world.level.block.state.properties.EnumProperty<?> FACING = getEnum(net.minecraft.world.level.block.BedBlock.class, "facing");
