From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 21 Jun 2024 18:27:31 +0300
Subject: [PATCH] Allow jumping in water as Drowned


diff --git a/src/main/java/net/minecraft/world/entity/monster/Drowned.java b/src/main/java/net/minecraft/world/entity/monster/Drowned.java
index 76db5a2d27ab435fdfd1e0ec0c77ef5012e128d4..7903e712f05d1263201e07cf79ea4e6d3fa76c43 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Drowned.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Drowned.java
@@ -92,6 +92,17 @@ public class Drowned extends Zombie implements RangedAttackMob {
     }
     // Purpur end - Ridables
 
+    // Kiterino start - Allow jumping in water as Drowned
+    @Override
+    public boolean onSpacebar() {
+	    if (isInWater()) {
+		    jumpFromGround();
+		    return true;
+	    }
+	    return false;
+    }
+	// Kiterino end - Allow jumping in water as Drowned
+
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
