From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 20 Feb 2021 02:05:32 +0200
Subject: [PATCH] Don't edit signs upon placing


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 7222ff43de1b570458f2513fe1a44d97a09d0799..2e05b1a4c5be46c4184bba1027d8d43babf92220 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -477,7 +477,7 @@ public final class ItemStack {
                     // SPIGOT-4678
                     if (this.item instanceof SignItem && SignItem.openSign != null) {
                         try {
-                            entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign));
+                            //entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign)); // Kiterino
                         } finally {
                             SignItem.openSign = null;
                         }
