From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 08:46:08 +0100
Subject: [PATCH] Drop stored experience API for furnaces


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftFurnace.java b/src/main/java/org/bukkit/craftbukkit/block/CraftFurnace.java
index f49bb90e6c30dd928b352c867819b687e4557893..e41a34f282444cc4fde8aeda3517cb02b14fd0bc 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftFurnace.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftFurnace.java
@@ -126,4 +126,12 @@ public abstract class CraftFurnace<T extends AbstractFurnaceBlockEntity> extends
         });
     }
     // Paper end
+
+    // Kiterino start
+    @Override
+    public void dropStoredExperience() {
+        var furnace = getTileEntity();
+        furnace.getRecipesToAwardAndPopExperience(world.getHandle(), net.minecraft.world.phys.Vec3.atCenterOf(furnace.getBlockPos()));
+    }
+    // Kiterino end
 }
