From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 20 Apr 2022 15:01:32 +0300
Subject: [PATCH] Expose pre-collision move vector to
 VehicleBlockCollisionEvent


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 9b9dfa0998200042a5d56cc843f89962f8a60edf..249a918c1aff73725f35fdb348ef8959049ae480 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1136,7 +1136,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                     }
 
                     if (!bl.getType().isAir()) {
-                        VehicleBlockCollisionEvent event = new VehicleBlockCollisionEvent(vehicle, bl);
+                        VehicleBlockCollisionEvent event = new VehicleBlockCollisionEvent(vehicle, bl, org.bukkit.craftbukkit.util.CraftVector.toBukkit(moveVector)); // Kiterino - Expose pre-collision move vector
                         this.level.getCraftServer().getPluginManager().callEvent(event);
                     }
                 }
