From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 17:04:08 +0300
Subject: [PATCH] Campfires should fall with gravity


diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index ba023c86a77d4a3210ed5d4680ea396706167e93..7f160180752a8d7f68ce949217a5665f6dd495ca 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -67,6 +67,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
                 return;
             }
         }
+        if (campfire.checkGravity(world, pos, state)) return;
         // Kiterino end
 
         for (int i = 0; i < campfire.items.size(); ++i) {
@@ -121,6 +122,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     }
 
     public static void cooldownTick(Level world, BlockPos pos, BlockState state, CampfireBlockEntity campfire) {
+        if (campfire.checkGravity(world, pos, state)) return; // Kiterino
         boolean flag = false;
 
         for (int i = 0; i < campfire.items.size(); ++i) {
@@ -136,6 +138,16 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
     }
 
+    // Kiterino start
+    private boolean checkGravity(Level world, BlockPos pos, BlockState state) {
+        if (net.minecraft.world.level.block.FallingBlock.isFree(world.getBlockState(pos.below())) && pos.getY() >= world.getMinBuildHeight()) {
+            world.addFreshEntity(new net.minecraft.world.entity.item.FallingBlockEntity(world, pos.getX() + 0.5D, pos.getY(), pos.getZ() + 0.5D, state));
+            return true;
+        }
+        return false;
+    }
+    // Kiterino end
+
     public static void particleTick(Level world, BlockPos pos, BlockState state, CampfireBlockEntity campfire) {
         RandomSource randomsource = world.random;
         int i;
