From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 17:04:08 +0300
Subject: [PATCH] Campfires should fall with gravity


diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index 36f33df5a2c29600c0d4c5da2301fb45748a8f45..aaf4f7ca60e98e054ce9273978a5abc3a8efc1ff 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -65,6 +65,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
                 return;
             }
         }
+        if (campfire.checkGravity(world, pos, state)) return;
         // Kiterino end
 
         for (int i = 0; i < campfire.items.size(); ++i) {
@@ -119,6 +120,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     }
 
     public static void cooldownTick(Level world, BlockPos pos, BlockState state, CampfireBlockEntity campfire) {
+        if (campfire.checkGravity(world, pos, state)) return; // Kiterino
         boolean flag = false;
 
         for (int i = 0; i < campfire.items.size(); ++i) {
@@ -134,6 +136,16 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
     }
 
+    // Kiterino start
+    private boolean checkGravity(Level world, BlockPos pos, BlockState state) {
+        if (net.minecraft.world.level.block.FallingBlock.isFree(world.getBlockState(pos.below())) && pos.getY() >= world.getMinBuildHeight()) {
+            world.addFreshEntity(new net.minecraft.world.entity.item.FallingBlockEntity(world, pos.getX() + 0.5D, pos.getY(), pos.getZ() + 0.5D, state));
+            return true;
+        }
+        return false;
+    }
+    // Kiterino end
+
     public static void particleTick(Level world, BlockPos pos, BlockState state, CampfireBlockEntity campfire) {
         RandomSource randomsource = world.random;
         int i;
