From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 07:39:00 +0100
Subject: [PATCH] Make silverfishes more annoying


diff --git a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
index dbc27afbccb027963ae3f035c268835adedbb933..27c76a937fae3c52e62c0ba2ea688d1ea5fbc75e 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
@@ -104,7 +104,7 @@ public class Silverfish extends Monster {
 
     @Override
     protected Entity.MovementEmission getMovementEmission() {
-        return Entity.MovementEmission.EVENTS;
+        return Entity.MovementEmission.ALL; // Kiterino - EVENT -> ALL
     }
 
     @Override
