From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 07:39:00 +0100
Subject: [PATCH] Make silverfishes more annoying


diff --git a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
index 931412a5ab315d4080a9f5209d3e85d78642f4c2..06c9c8c8f90529a918f1e7a0ca353dca8613ba06 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
@@ -105,7 +105,7 @@ public class Silverfish extends Monster {
 
     @Override
     protected Entity.MovementEmission getMovementEmission() {
-        return Entity.MovementEmission.EVENTS;
+        return Entity.MovementEmission.ALL; // Kiterino - EVENT -> ALL
     }
 
     @Override
