From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 27 Jun 2024 06:36:56 +0300
Subject: [PATCH] More resource pack API


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index 05e16103af3fd276f0196ddf1a2e5b729b025c34..4bde7dddfa6d62b407e9e7128c999c49234b8b7f 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -100,7 +100,7 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     public final int entityBroadcastRangePercentage;
     public final String textFilteringConfig;
     public final int textFilteringVersion;
-    public final Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo;
+    public Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo; // Kiterino - remove final - More resource pack API
     public final DataPackConfig initialDataPackConfiguration;
     public final Settings<DedicatedServerProperties>.MutableValue<Integer> playerIdleTimeout;
     public final Settings<DedicatedServerProperties>.MutableValue<Boolean> whiteList;
@@ -205,7 +205,7 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
         return null;
     }
 
-    private static Optional<MinecraftServer.ServerResourcePackInfo> getServerPackInfo(String id, String url, String sha1, @Nullable String hash, boolean required, String prompt) {
+    public static Optional<MinecraftServer.ServerResourcePackInfo> getServerPackInfo(String id, String url, String sha1, @Nullable String hash, boolean required, String prompt) { // Kiterino - private -> public - More resource pack API
         if (url.isEmpty()) {
             return Optional.empty();
         } else {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6031dfd15bacea2eb48c4191901f7aa7a1042bae..01e32c33b515f2e0c46ec5bce884f199de507a46 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3373,4 +3373,20 @@ public final class CraftServer implements Server {
         return getServer().lagging;
     }
     // Purpur end
+
+    // Kiterino start - More resource pack API
+    @Override
+    public ResourcePack createResourcePack(UUID uuid, String url, String hash, boolean required, net.kyori.adventure.text.Component prompt) {
+        return new CraftResourcePack(new MinecraftServer.ServerResourcePackInfo(uuid, url, hash, required, io.papermc.paper.adventure.PaperAdventure.asVanilla(prompt)));
+    }
+
+    @Override
+    public void setServerResourcePack(ResourcePack resourcePack) {
+        if (resourcePack == null) {
+            getServer().getProperties().serverResourcePackInfo = Optional.empty();
+            return;
+        }
+        getServer().getProperties().serverResourcePackInfo = Optional.of(((CraftResourcePack) resourcePack).getHandle());
+    }
+    // Kiterino end - More resource pack API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/packs/CraftResourcePack.java b/src/main/java/org/bukkit/craftbukkit/packs/CraftResourcePack.java
index 9313d17f01198f190e885579f8f6103187499d44..c915a6000ecad1dae3052f70c2b53afd158cc551 100644
--- a/src/main/java/org/bukkit/craftbukkit/packs/CraftResourcePack.java
+++ b/src/main/java/org/bukkit/craftbukkit/packs/CraftResourcePack.java
@@ -42,4 +42,10 @@ public class CraftResourcePack implements ResourcePack {
     public String toString() {
         return "CraftResourcePack{id=" + this.getId() + ",url=" + this.getUrl() + ",hash=" + this.getHash() + ",prompt=" + this.getPrompt() + ",required=" + this.isRequired() + "}";
     }
+
+    // Kiterino start - More resource pack API
+    public MinecraftServer.ServerResourcePackInfo getHandle() {
+        return this.handle;
+    }
+    // Kiterino end - More resource pack API
 }
