From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 04:13:35 +0100
Subject: [PATCH] Better Impaling


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 599094f231787db475882c38721b724a2e3bf41c..48849af29cbb3cc637ceb16f68e9f03711b9f4f3 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1704,6 +1704,12 @@ public abstract class Mob extends LivingEntity implements Targeting {
 
         if (target instanceof LivingEntity) {
             f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
+            // Kiterino start
+            if (((LivingEntity) target).getMobType() != net.minecraft.world.entity.MobType.WATER && target.isInWaterRainOrBubble()) {
+                int impalingLevel = EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.IMPALING, getMainHandItem());
+                if (impalingLevel > 0) f += net.minecraft.world.item.enchantment.Enchantments.IMPALING.getDamageBonus(impalingLevel, net.minecraft.world.entity.MobType.WATER);
+            }
+            // Kiterino end
             f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 1429b99a787f62d5bb3e9c0b8b7b525f8abb6ae8..abf04da0b8672e6d2ed9e2952d7488411f906e14 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1301,6 +1301,12 @@ public abstract class Player extends LivingEntity {
 
                 if (target instanceof LivingEntity) {
                     f1 = EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
+                    // Kiterino start
+                    if (((LivingEntity) target).getMobType() != MobType.WATER && target.isInWaterRainOrBubble()) {
+                        int impalingLevel = EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.IMPALING, getMainHandItem());
+                        if (impalingLevel > 0) f += net.minecraft.world.item.enchantment.Enchantments.IMPALING.getDamageBonus(impalingLevel, MobType.WATER);
+                    }
+                    // Kiterino end
                 } else {
                     f1 = EnchantmentHelper.getDamageBonus(this.getMainHandItem(), MobType.UNDEFINED);
                 }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
index dde841cc09ba4a3575a462b03537887551d47ba5..a615d2873184ab698f5ce20b1860a5af95911ea1 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -136,6 +136,12 @@ public class ThrownTrident extends AbstractArrow {
             LivingEntity entityliving = (LivingEntity) entity;
 
             f += EnchantmentHelper.getDamageBonus(this.tridentItem, entityliving.getMobType());
+            // Kiterino start
+            if (entityliving.getMobType() != net.minecraft.world.entity.MobType.WATER && entityliving.isInWaterRainOrBubble()) {
+                int impalingLevel = EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.IMPALING, tridentItem);
+                if (impalingLevel > 0) f += net.minecraft.world.item.enchantment.Enchantments.IMPALING.getDamageBonus(impalingLevel, net.minecraft.world.entity.MobType.WATER);
+            }
+            // Kiterino end
         }
 
         Entity entity1 = this.getOwner();
