From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 07:39:00 +0100
Subject: [PATCH] Make silverfishes more annoying


diff --git a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
index 406ce5b56f73253b034b096438c7506b63007532..a9f743cc42a910eb128be751fc9edcb784f9757b 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
@@ -107,7 +107,7 @@ public class Silverfish extends Monster {
 
     @Override
     protected Entity.MovementEmission getMovementEmission() {
-        return Entity.MovementEmission.EVENTS;
+        return Entity.MovementEmission.ALL; // Kiterino - EVENT -> ALL
     }
 
     @Override
