From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 10 Sep 2020 19:28:30 +0300
Subject: [PATCH] Reset fire ticks when having fire resistance


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index f48bd42b91772f3cbf474d63d6b5e19b159a0f1f..be32546f61934631902a742bb39b40c601b269bb 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -919,6 +919,7 @@ public abstract class LivingEntity extends Entity {
                 } else if (mobeffect.getDuration() % 600 == 0) {
                     this.onEffectUpdated(mobeffect, false, (Entity) null);
                 }
+                if (mobeffect.getEffect() == MobEffects.FIRE_RESISTANCE && getRemainingFireTicks() > 0) setRemainingFireTicks(0); // Kiterino
             }
         } catch (ConcurrentModificationException concurrentmodificationexception) {
             ;
