From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 12 Dec 2021 23:00:31 +0100
Subject: [PATCH] Fix MC-98160 and MC-105103


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index 231c71939982ba4ce9305bc8eb6174ed45102c9f..22be22265081ba8972cd27089b9f05c4f4953beb 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -899,7 +899,7 @@ public class Boat extends Entity implements VariantHolder<Boat.Type> {
         if (!this.isPassenger()) {
             if (onGround) {
                 if (this.fallDistance > 3.0F) {
-                    if (this.status != Boat.Status.ON_LAND) {
+                    if (this.status != Boat.Status.ON_LAND && this.status != Status.IN_AIR) { // Kiterino - Fix MC-98160 and MC-105103
                         this.resetFallDistance();
                         return;
                     }
