From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 8 Jul 2021 22:46:50 +0300
Subject: [PATCH] Elder Guardian appearance API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 83ffb008ee43bdfce80286522ca687fe217ce6d3..1dcaeaf7faa82e2808182666ff15547bf48f6fa2 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2985,5 +2985,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         Location loc = getLocation();
         getHandle().forceSleep(loc.getX(), loc.getY(), loc.getZ());
     }
+
+    @Override
+    public void showElderGuardian(boolean playSound) {
+        getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.GUARDIAN_ELDER_EFFECT, playSound ? 1F : 0F));
+    }
     // Kiterino end
 }
