From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 12 Sep 2024 18:07:16 +0300
Subject: [PATCH] Bats ignore invisible players

Invisible by API, not via potion effects

diff --git a/src/main/java/me/sosedik/kiterino/KiterinoWorldConfig.java b/src/main/java/me/sosedik/kiterino/KiterinoWorldConfig.java
index 0e430c0915012eb55ad9f656565499a9071b2064..2979eb09eee76d5b54e118bf5ce7d39ccc6d5eb3 100644
--- a/src/main/java/me/sosedik/kiterino/KiterinoWorldConfig.java
+++ b/src/main/java/me/sosedik/kiterino/KiterinoWorldConfig.java
@@ -16,4 +16,9 @@ public final class KiterinoWorldConfig extends WorldConfig {
         KiterinoConfig.readConfig(KiterinoWorldConfig.class, this);
     }
 
+    public boolean batsIgnoreInvisiblePlayers = false;
+    private void batsSettings() {
+        batsIgnoreInvisiblePlayers = getBoolean("gameplay-mechanics.bat.ignore-invisible-players", batsIgnoreInvisiblePlayers, "Make resting bats ignore players that are marked invisible by API");
+    }
+
 }
diff --git a/src/main/java/net/minecraft/world/entity/ambient/Bat.java b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
index 2d4fe19da8d53d47df36399531e78da2c961aecd..29a6a34e64c5f3497789353a5b1473f19b790961 100644
--- a/src/main/java/net/minecraft/world/entity/ambient/Bat.java
+++ b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
@@ -39,7 +39,7 @@ public class Bat extends AmbientCreature {
     public static final float TICKS_PER_FLAP = 10.0F;
     private static final EntityDataAccessor<Byte> DATA_ID_FLAGS = SynchedEntityData.defineId(Bat.class, EntityDataSerializers.BYTE);
     private static final int FLAG_RESTING = 1;
-    private static final TargetingConditions BAT_RESTING_TARGETING = TargetingConditions.forNonCombat().range(4.0D);
+    private static final TargetingConditions BAT_RESTING_TARGETING = TargetingConditions.forNonCombat().range(4.0D).selector((entity, level) -> !level.kiterinoConfig.batsIgnoreInvisiblePlayers || !entity.isInvisible()); // Kiterino - Bats ignore invisible players
     public final AnimationState flyAnimationState = new AnimationState();
     public final AnimationState restAnimationState = new AnimationState();
     @Nullable
