From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 20 Feb 2021 02:05:32 +0200
Subject: [PATCH] Don't edit signs upon placing


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 5c84b0f8478d20dc3c1593118addeabea28c8e73..bf89dc41f2a1845ae646d4c27bf3f67135fbfb9c 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -469,7 +469,7 @@ public final class ItemStack {
                     // SPIGOT-4678
                     if (this.item instanceof SignItem && SignItem.openSign != null) {
                         try {
-                            entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign));
+                            //entityhuman.openTextEdit((SignBlockEntity) world.getBlockEntity(SignItem.openSign)); // Kiterino
                         } finally {
                             SignItem.openSign = null;
                         }
