From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 09:58:34 +0300
Subject: [PATCH] Allow grass spread upon coarse dirt


diff --git a/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java b/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
index 61783f17b655cbb6430d22fb3a81931ab3ea130c..3757781a1f7ef49fb10a19097eec6605bd64b00c 100644
--- a/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
@@ -58,6 +58,11 @@ public abstract class SpreadingSnowyDirtBlock extends SnowyDirtBlock {
                     if (world.getBlockState(blockposition1).is(Blocks.DIRT) && SpreadingSnowyDirtBlock.canPropagate(iblockdata1, world, blockposition1)) {
                         org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(world, pos, blockposition1, (BlockState) iblockdata1.setValue(SpreadingSnowyDirtBlock.SNOWY, world.getBlockState(blockposition1.above()).is(Blocks.SNOW))); // CraftBukkit
                     }
+                    // Kiterino start
+                    else if (world.getBlockState(blockposition1).is(Blocks.COARSE_DIRT) && SpreadingSnowyDirtBlock.canPropagate(iblockdata1, (LevelReader) world, blockposition1)) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(world, pos, blockposition1, Blocks.DIRT.defaultBlockState());
+                    }
+                    // Kiterino end
                 }
             }
 
