From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 14:43:49 +0100
Subject: [PATCH] Shulker bullets despawn on owner death


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 1c50c870e59c35a39c83a0f136ee5d3c70268763..284369b2538079fb031a89e753cfe45f748b657c 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -222,6 +222,15 @@ public class ShulkerBullet extends Projectile {
 
         if (!this.level().isClientSide) {
             if (this.finalTarget == null && this.targetId != null) {
+                // Kiterino start
+                if (!isInvulnerable()) {
+                    var owner = getOwner();
+                    if (owner == null || !owner.isAlive()) {
+                        discard();
+                        return;
+                    }
+                }
+                // Kiterino end
                 this.finalTarget = ((ServerLevel) this.level()).getEntity(this.targetId);
                 if (this.finalTarget == null) {
                     this.targetId = null;
