From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 14:36:07 +0100
Subject: [PATCH] Reverse note block playing


diff --git a/src/main/java/net/minecraft/world/level/block/NoteBlock.java b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
index 6a970e293a31962faaab545f368ae02c4b20e42d..cc5c42261c25741d33e163de57a83648331ccc0c 100644
--- a/src/main/java/net/minecraft/world/level/block/NoteBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
@@ -124,7 +124,15 @@ public class NoteBlock extends Block {
         } else if (world.isClientSide) {
             return InteractionResult.SUCCESS;
         } else {
-            if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) state = (BlockState) state.cycle(NoteBlock.NOTE); // Paper - prevent noteblock note from updating
+            // Kiterino start
+            if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) {
+                if (player.isShiftKeyDown()) {
+                    state = state.cycleBackwards(NOTE);
+                } else {
+                    state = (BlockState) state.cycle(NoteBlock.NOTE); // Paper - prevent noteblock note from updating
+                }
+            }
+            // Kiterino end
             world.setBlock(pos, state, 3);
             this.playNote(player, state, world, pos);
             player.awardStat(Stats.TUNE_NOTEBLOCK);
diff --git a/src/main/java/net/minecraft/world/level/block/state/StateHolder.java b/src/main/java/net/minecraft/world/level/block/state/StateHolder.java
index 5f285d190186a2ff5a61d05070593e1d633dd79a..12c5625366f5845b996d8e75d8a9a018e76ec2d7 100644
--- a/src/main/java/net/minecraft/world/level/block/state/StateHolder.java
+++ b/src/main/java/net/minecraft/world/level/block/state/StateHolder.java
@@ -68,6 +68,22 @@ public abstract class StateHolder<O, S> {
         return iterator.next();
     }
 
+    // Kiterino start
+    public <T extends Comparable<T>> S cycleBackwards(Property<T> property) {
+        return this.setValue(property, findPreviousInCollection(property.getPossibleValues(), this.getValue(property)));
+    }
+
+    protected static <T> T findPreviousInCollection(Collection<T> values, T value) {
+        T previous = null;
+        for (T next : values) {
+            if (next.equals(value) && previous != null)
+                return previous;
+            previous = next;
+        }
+        return previous == null ? value : previous;
+    }
+    // Kiterino end
+
     @Override
     public String toString() {
         StringBuilder stringBuilder = new StringBuilder();
