From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 12 Sep 2024 23:52:50 +0300
Subject: [PATCH] Panda force scared state


diff --git a/src/main/java/net/minecraft/world/entity/animal/Panda.java b/src/main/java/net/minecraft/world/entity/animal/Panda.java
index 1a0ffbe2f835cbe37ef670d01daf08fb2f48b4fa..72b63fbfdd593093a0e3c756e994a971da49c226 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Panda.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Panda.java
@@ -111,6 +111,7 @@ public class Panda extends Animal {
 
         return (itemstack.is(Blocks.BAMBOO.asItem()) || itemstack.is(Blocks.CAKE.asItem())) && entityitem.isAlive() && !entityitem.hasPickUpDelay();
     };
+    public net.kyori.adventure.util.TriState forceScared = net.kyori.adventure.util.TriState.NOT_SET; // Kiterino - Panda force scared state
 
     public Panda(EntityType<? extends Panda> type, Level world) {
         super(type, world);
@@ -462,6 +463,11 @@ public class Panda extends Animal {
     }
 
     public boolean isScared() {
+        // Kiterino start - Panda force scared state
+        if (this.forceScared != net.kyori.adventure.util.TriState.NOT_SET) {
+            return this.forceScared.toBooleanOrElse(false);
+        }
+        // Kiterino end - Panda force scared state
         return this.isWorried() && this.level().isThundering();
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java
index 01d104d91de9e1319d27e39d3f474318c7809486..a3aba105d6135dea9b1a529f9f8084a00d8c3325 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPanda.java
@@ -128,6 +128,13 @@ public class CraftPanda extends CraftAnimals implements Panda {
         return this.getHandle().isScared();
     }
 
+    // Kiterino start - Panda force scared state
+    @Override
+    public void setScared(net.kyori.adventure.util.TriState scared) {
+        this.getHandle().forceScared = scared;
+    }
+    // Kiterino end - Panda force scared state
+
     @Override
     public int getUnhappyTicks() {
         return this.getHandle().getUnhappyCounter();
