From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 14:43:49 +0100
Subject: [PATCH] Shulker bullets despawn on owner death


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 113ab29ba7adbb40f61eb0cfbda23ada7ef1305d..79574d1420dec94465185271def85269500d51d9 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -218,6 +218,15 @@ public class ShulkerBullet extends Projectile {
         Vec3 vec3d;
 
         if (!this.level.isClientSide) {
+            // Kiterino start
+            if (!isInvulnerable()) {
+                var owner = getOwner();
+                if (owner == null || !owner.isAlive()) {
+                    discard();
+                    return;
+                }
+            }
+            // Kiterino end
             if (this.finalTarget == null && this.targetId != null) {
                 this.finalTarget = ((ServerLevel) this.level).getEntity(this.targetId);
                 if (this.finalTarget == null) {
