From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 14:43:49 +0100
Subject: [PATCH] Shulker bullets despawn on owner death


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 93faa6df5d50d4b3037b5796de8b9ccfc98548d0..5bff04af15e9dda1393f6588c3de1530bf165f09 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -217,6 +217,15 @@ public class ShulkerBullet extends Projectile {
         Vec3 vec3d;
 
         if (!this.level.isClientSide) {
+            // Kiterino start
+            if (!isInvulnerable()) {
+                var owner = getOwner();
+                if (owner == null || !owner.isAlive()) {
+                    discard();
+                    return;
+                }
+            }
+            // Kiterino end
             if (this.finalTarget == null && this.targetId != null) {
                 this.finalTarget = ((ServerLevel) this.level).getEntity(this.targetId);
                 if (this.finalTarget == null) {
