From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 29 Jun 2020 17:41:30 +0300
Subject: [PATCH] Campfires store potion effects


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
index fee09e6ff72cf1da389d5811dd005642cd50a5b4..bf88aa71178d9744e9421977f4d0ac16cb51f7ed 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
@@ -112,6 +112,7 @@ public class ThrownPotion extends ThrowableItemProjectile implements ItemSupplie
                 } else {
                     this.applySplash(list, hitResult.getType() == HitResult.Type.ENTITY ? ((EntityHitResult) hitResult).getEntity() : null);
                 }
+                splashCampfires(list); // Kiterino
             }
 
             int i = potionregistry.hasInstantEffects() ? 2007 : 2002;
@@ -121,6 +122,29 @@ public class ThrownPotion extends ThrowableItemProjectile implements ItemSupplie
         }
     }
 
+    // Kiterino start
+    private void splashCampfires(List<MobEffectInstance> list) {
+        if (list.isEmpty()) return;
+        var aabb = getBoundingBox().inflate(4.0D, 2.0D, 4.0D);
+        for (int x = (int) aabb.minX; x <= aabb.maxX; x++) {
+            for (int z = (int) aabb.minZ; z <= aabb.maxZ; z++) {
+                for (int y = (int) aabb.minY; y <= aabb.maxY; y++) {
+                    var pos = new BlockPos(x, y, z);
+                    var te = level.getBlockEntity(pos);
+                    if (te instanceof net.minecraft.world.level.block.entity.CampfireBlockEntity campfire) {
+                        if (!campfire.getBlockState().getValue(CampfireBlock.LIT)) continue;
+                        campfire.effect = list.get(0).getEffect();
+                        for (var i = 0; i < 6 ; i++) {
+                            var position = te.getBlockPos();
+                            ((net.minecraft.server.level.ServerLevel) level).sendParticles(null, net.minecraft.core.particles.ParticleTypes.HEART, position.getX() + 0.5, position.getY() + 1D, position.getZ() + 0.5, 1, 0.5, 0.5, 0.5, 0, true);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // Kiterino end
+
     private static final Predicate<net.minecraft.world.entity.LivingEntity> APPLY_WATER_GET_ENTITIES_PREDICATE = ThrownPotion.WATER_SENSITIVE.or(Axolotl.class::isInstance); // Paper
     private void applyWater() {
         AABB axisalignedbb = this.getBoundingBox().inflate(4.0D, 2.0D, 4.0D);
diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index 122b3eaabb02b207bfb9d0b03905e0702dc8aead..34fe81e24edfef1ac6422ad5d1eeb0854580ec12 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -40,6 +40,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     public final int[] cookingTime;
     private final RecipeManager.CachedCheck<Container, CampfireCookingRecipe> quickCheck;
     public final boolean[] stopCooking; // Paper
+    public @org.jetbrains.annotations.Nullable net.minecraft.world.effect.MobEffect effect; // Kiterino
 
     public CampfireBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.CAMPFIRE, pos, state);
@@ -64,6 +65,14 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
                 setChanged(world, pos, state);
                 return;
             }
+            if (campfire.effect != null) {
+                boolean signalBoost = state.getValue(CampfireBlock.SIGNAL_FIRE);
+                int duration = signalBoost ? 180 : 80;
+                int amp = signalBoost ? 1 : 0;
+                int range = signalBoost ? 10 : 5;
+                var effect = new net.minecraft.world.effect.MobEffectInstance(campfire.effect, duration, amp, true, true);
+                world.getEntitiesOfClass(net.minecraft.world.entity.player.Player.class, new net.minecraft.world.phys.AABB(pos).inflate(range)).forEach(player -> player.addEffect(effect));
+            }
         }
         if (campfire.checkGravity(world, pos, state)) return;
         // Kiterino end
@@ -118,6 +127,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     }
 
     public static void cooldownTick(Level world, BlockPos pos, BlockState state, CampfireBlockEntity campfire) {
+        if (campfire.effect != null) campfire.effect = null; // Kiterino - reset effect if unlit
         if (campfire.checkGravity(world, pos, state)) return; // Kiterino
         boolean flag = false;
 
@@ -203,6 +213,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
             System.arraycopy(cookingState, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, abyte.length));
         }
         // Paper end
+        if (nbt.contains("Kiterino.Effect")) effect = net.minecraft.core.Registry.MOB_EFFECT.getOptional(net.minecraft.resources.ResourceLocation.tryParse(nbt.getString("Kiterino.Effect"))).orElse(net.minecraft.world.effect.MobEffects.REGENERATION); // Kiterino
     }
 
     @Override
@@ -218,6 +229,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
         }
         nbt.putByteArray("Paper.StopCooking", cookingState);
         // Paper end
+        if (effect != null) nbt.putString("Kiterino.Effect", net.minecraft.core.Registry.MOB_EFFECT.getKey(effect).getPath()); // Kiterino
     }
 
     @Override
