From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 28 Sep 2024 01:03:13 +0300
Subject: [PATCH] Anger nearby piglins API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 4951b127b0f740929d92964a8f0522c3c943d511..eac50b965b0a64cfae8f78382bf16df46a3cf5ca 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3698,4 +3698,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket(0, stateId, slot, CraftItemStack.asNMSCopy(item)));
     }
     // Kiterino end - Send ItemStack API
+
+    // Kiterino start - Anger nearby piglins API
+    @Override
+    public void angerNearbyPiglins(boolean blockOpen) {
+        net.minecraft.world.entity.monster.piglin.PiglinAi.angerNearbyPiglins(this.getHandle(), blockOpen);
+    }
+    // Kiterino end - Anger nearby piglins API
 }
