From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 13:13:03 +0100
Subject: [PATCH] Use unused ghast sound


diff --git a/src/main/java/net/minecraft/world/entity/monster/Ghast.java b/src/main/java/net/minecraft/world/entity/monster/Ghast.java
index 225a4e549c2cbf64beaba52d26b196af5b868433..151a57544d6e96f9e1cc0cc883aef4b11127c533 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Ghast.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Ghast.java
@@ -164,6 +164,7 @@ public class Ghast extends FlyingMob implements Enemy {
 
     @Override
     protected SoundEvent getAmbientSound() {
+        if (!level.dimensionType().ultraWarm() && random.nextInt(8) == 0) return SoundEvents.GHAST_SCREAM; // Kiterino
         return SoundEvents.GHAST_AMBIENT;
     }
 
