From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 19 Feb 2021 22:48:24 +0200
Subject: [PATCH] Flammable cobwebs


diff --git a/src/main/java/net/minecraft/world/level/block/Blocks.java b/src/main/java/net/minecraft/world/level/block/Blocks.java
index 7e95a8dbf6f3c0202bf72655ad8eccb6027b84c5..e614946f11b48e11e3c0d34fa7ad43a1167a14cf 100644
--- a/src/main/java/net/minecraft/world/level/block/Blocks.java
+++ b/src/main/java/net/minecraft/world/level/block/Blocks.java
@@ -771,6 +771,7 @@ public class Blocks {
                 .requiresCorrectToolForDrops()
                 .strength(4.0F)
                 .pushReaction(PushReaction.DESTROY)
+                .ignitedByLava() // Kiterino - Flammable cobwebs
         )
     );
     public static final Block SHORT_GRASS = register(
diff --git a/src/main/java/net/minecraft/world/level/block/FireBlock.java b/src/main/java/net/minecraft/world/level/block/FireBlock.java
index 7f05e58358024d303eab9ab4fbc1bb299760ad1e..84c004b38273a1a0c7c4d5b99b7996a29399a4eb 100644
--- a/src/main/java/net/minecraft/world/level/block/FireBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FireBlock.java
@@ -566,5 +566,6 @@ public class FireBlock extends BaseFireBlock {
         blockfire.setFlammable(Blocks.SMALL_DRIPLEAF, 60, 100);
         blockfire.setFlammable(Blocks.HANGING_ROOTS, 30, 60);
         blockfire.setFlammable(Blocks.GLOW_LICHEN, 15, 100);
+        blockfire.setFlammable(Blocks.COBWEB, 60, 100); // Kiterino - Flammable cobwebs
     }
 }
