From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 13 Dec 2021 07:39:00 +0100
Subject: [PATCH] Make silverfishes more annoying


diff --git a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
index 7f8d1a048cbe42782be78b60d7cc5f22a6879539..74ee60f339774db2ed6e9b7162dff309c13b75b2 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Silverfish.java
@@ -104,7 +104,7 @@ public class Silverfish extends Monster {
 
     @Override
     protected Entity.MovementEmission getMovementEmission() {
-        return Entity.MovementEmission.EVENTS;
+        return Entity.MovementEmission.ALL; // Kiterino - EVENT -> ALL
     }
 
     @Override
