From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 28 Sep 2024 01:17:04 +0300
Subject: [PATCH] Open ender chest API


diff --git a/src/main/java/net/minecraft/world/level/block/EnderChestBlock.java b/src/main/java/net/minecraft/world/level/block/EnderChestBlock.java
index bfe79431dc5707677671df5c0787817c6e14a676..b020227337f31edc4f9209f4e87391172548e415 100644
--- a/src/main/java/net/minecraft/world/level/block/EnderChestBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/EnderChestBlock.java
@@ -43,7 +43,7 @@ public class EnderChestBlock extends AbstractChestBlock<EnderChestBlockEntity> i
     public static final EnumProperty<Direction> FACING = HorizontalDirectionalBlock.FACING;
     public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;
     protected static final VoxelShape SHAPE = Block.box(1.0, 0.0, 1.0, 15.0, 14.0, 15.0);
-    private static final Component CONTAINER_TITLE = Component.translatable("container.enderchest");
+    public static final Component CONTAINER_TITLE = Component.translatable("container.enderchest"); // Kiterino - private -> public - Open ender chest API
 
     @Override
     public MapCodec<EnderChestBlock> codec() {
@@ -100,7 +100,7 @@ public class EnderChestBlock extends AbstractChestBlock<EnderChestBlockEntity> i
     }
 
     // Purpur start
-    private ChestMenu getEnderChestSixRows(int syncId, net.minecraft.world.entity.player.Inventory inventory, Player player, PlayerEnderChestContainer playerEnderChestContainer) {
+    public static ChestMenu getEnderChestSixRows(int syncId, net.minecraft.world.entity.player.Inventory inventory, Player player, PlayerEnderChestContainer playerEnderChestContainer) { // Kiterino - private -> public static - Open ender chest API
         if (org.purpurmc.purpur.PurpurConfig.enderChestPermissionRows) {
             org.bukkit.craftbukkit.entity.CraftHumanEntity bukkitPlayer = player.getBukkitEntity();
             if (bukkitPlayer.hasPermission("purpur.enderchest.rows.six")) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 415e95c7665ecf159651814cb96d6eb0e4b12931..dcd6cc10d722e5d32fbe10aa53a77e4abb3b5130 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3702,4 +3702,19 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         net.minecraft.world.entity.monster.piglin.PiglinAi.angerNearbyPiglins(this.getHandle().serverLevel(), this.getHandle(), blockOpen);
     }
     // Kiterino end - Anger nearby piglins API
+
+    // Kiterino start - Open ender chest API
+    /**
+     * Opens ender chest for the player
+     *
+     * @param enderChest ender chest
+     */
+    @Override
+    public boolean openEnderChest(@NotNull org.bukkit.block.EnderChest enderChest) {
+        net.minecraft.world.inventory.PlayerEnderChestContainer playerEnderChestContainer = this.getHandle().getEnderChestInventory();
+        net.minecraft.world.level.block.entity.EnderChestBlockEntity blockEntity = ((org.bukkit.craftbukkit.block.CraftEnderChest) enderChest).getTileEntity();
+        playerEnderChestContainer.setActiveChest(blockEntity);
+        return this.getHandle().openMenu(new net.minecraft.world.SimpleMenuProvider((i, inventory, playerx) -> org.purpurmc.purpur.PurpurConfig.enderChestSixRows ? net.minecraft.world.level.block.EnderChestBlock.getEnderChestSixRows(i, inventory, this.getHandle(), playerEnderChestContainer) : net.minecraft.world.inventory.ChestMenu.threeRows(i, inventory, playerEnderChestContainer), net.minecraft.world.level.block.EnderChestBlock.CONTAINER_TITLE)).isPresent();
+    }
+    // Kiterino end - Open ender chest API
 }
