From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 13 Aug 2020 17:35:52 +0300
Subject: [PATCH] KItem RecipeChoice


diff --git a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
index f47eab4c31925f51de4a6bc8be730281cb3388fc..5e8f5f020165c46d7bfd1c333c7060b142e6e5cd 100644
--- a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
+++ b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
@@ -40,6 +40,7 @@ public final class Ingredient implements Predicate<ItemStack> {
     private IntList stackingIds;
     public boolean exact; // CraftBukkit
     public Predicate<org.bukkit.inventory.ItemStack> predicate;
+    public boolean kItem; // Kiterino
 
     public Ingredient(Stream<? extends Ingredient.Value> entries) {
         this.values = (Ingredient.Value[]) entries.toArray((i) -> {
@@ -81,6 +82,25 @@ public final class Ingredient implements Predicate<ItemStack> {
             for (int j = 0; j < i; ++j) {
                 ItemStack itemstack1 = aitemstack[j];
 
+                // Kiterino start
+                if (kItem) {
+                    var compound = itemstack1.getTag();
+                    String namespace = org.bukkit.Bukkit.getName();
+                    if (compound != null && compound.contains(namespace)) {
+                        compound = compound.getCompound(namespace);
+                        if (compound.contains("Id")) {
+                            var compound1 = itemstack.getTag();
+                            if (compound1 == null) continue;
+                            if (!compound1.contains(namespace)) continue;
+                            compound1 = compound1.getCompound(namespace);
+                            if (!compound1.contains("Id")) continue;
+                            if (!compound.getString("Id").equals(compound1.getString("Id"))) continue;
+                            return true;
+                        }
+                    }
+                }
+                // Kiterino end
+
                 // CraftBukkit start
                 if (this.exact) {
                     if (itemstack1.getItem() == itemstack.getItem() && ItemStack.isSameItemSameTags(itemstack, itemstack1)) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
index 553ecc9b5631ffc0848a798bb3295f16d1722c1f..92f69578b349fc00f9d224101b5380933e481b1e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
@@ -31,6 +31,9 @@ public interface CraftRecipe extends Recipe {
             stack = new Ingredient(((RecipeChoice.ExactChoice) bukkit).getChoices().stream().map((mat) -> new net.minecraft.world.item.crafting.Ingredient.ItemValue(CraftItemStack.asNMSCopy(mat))));
             stack.exact = true;
             stack.predicate = ((RecipeChoice.ExactChoice) bukkit).getPredicate(); // Purpur
+        } else if (bukkit instanceof me.sosedik.kiterino.inventory.ItemChoice itemChoice) {
+            stack = new Ingredient(itemChoice.getChoices().stream().map(mat -> new net.minecraft.world.item.crafting.Ingredient.ItemValue(CraftItemStack.asNMSCopy(mat))));
+            stack.kItem = true;
         } else {
             throw new IllegalArgumentException("Unknown recipe stack instance " + bukkit);
         }
@@ -50,6 +53,16 @@ public interface CraftRecipe extends Recipe {
             return null;
         }
 
+        // Kiterino start
+        if (list.kItem) {
+            List<org.bukkit.inventory.ItemStack> choices = new ArrayList<>(list.itemStacks.length);
+            for (net.minecraft.world.item.ItemStack i : list.itemStacks) {
+                choices.add(CraftItemStack.asBukkitCopy(i));
+            }
+            return new me.sosedik.kiterino.inventory.ItemChoice(choices);
+        }
+        // Kiterino end
+
         if (list.exact) {
             List<org.bukkit.inventory.ItemStack> choices = new ArrayList<>(list.itemStacks.length);
             for (net.minecraft.world.item.ItemStack i : list.itemStacks) {
