From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 8 Jun 2022 13:03:53 +0300
Subject: [PATCH] Support recipe grid for AItems


diff --git a/src/main/java/net/minecraft/world/entity/player/Inventory.java b/src/main/java/net/minecraft/world/entity/player/Inventory.java
index 302ca7391109c10e81a7745504b3c530bc3be6ad..874f83b9eb774d7618d1e8fb1d6584094d30ba2b 100644
--- a/src/main/java/net/minecraft/world/entity/player/Inventory.java
+++ b/src/main/java/net/minecraft/world/entity/player/Inventory.java
@@ -199,7 +199,7 @@ public class Inventory implements Container, Nameable {
         for (int i = 0; i < this.items.size(); ++i) {
             ItemStack itemstack1 = (ItemStack) this.items.get(i);
 
-            if (!((ItemStack) this.items.get(i)).isEmpty() && ItemStack.isSameItemSameTags(stack, (ItemStack) this.items.get(i)) && !((ItemStack) this.items.get(i)).isDamaged() && !itemstack1.isEnchanted() && !itemstack1.hasCustomHoverName()) {
+            if (!itemstack1.isEmpty() && itemstack1.getItem() == stack.getItem() && ItemStack.recipesTagsMatch(itemstack1, stack)) { // Kiterino - less limited items in the crafting grid
                 return i;
             }
         }
diff --git a/src/main/java/net/minecraft/world/entity/player/StackedContents.java b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
index 842b921799111789b37a34b76644c9217bc85794..1b0bb36df3e662125b29a2da2820017599b73c90 100644
--- a/src/main/java/net/minecraft/world/entity/player/StackedContents.java
+++ b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
@@ -22,7 +22,7 @@ public class StackedContents {
     public final Int2IntMap contents = new Int2IntOpenHashMap();
 
     public void accountSimpleStack(ItemStack stack) {
-        if (!stack.isDamaged() && !stack.isEnchanted() && !stack.hasCustomHoverName()) {
+        if (/*!stack.isDamaged() && !stack.isEnchanted() && */!stack.hasCustomHoverName()) { // Kiterino - allow damaged and enchanted items
             this.accountStack(stack);
         }
 
@@ -54,16 +54,21 @@ public class StackedContents {
 
             @Override
             public boolean equals(@Nullable ItemStack a, @Nullable ItemStack b) {
-                if (a == null || b == null) {
-                    return false;
-                }
-                return ItemStack.tagMatches(a, b);
+                return ItemStack.recipesTagsMatch(a, b); // Kiterino - less strict check
             }
         });
         private final it.unimi.dsi.fastutil.ints.Int2ObjectMap<net.minecraft.world.item.ItemStack> idToItemstack = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>();
 
         @Override
         public int getId(ItemStack value) {
+            // Kiterino start
+            if (!this.itemstackToId.containsKey(value)) {
+                final ItemStack copy = value.copy();
+                copy.setTag(ItemStack.getStripped(copy.getTag()));
+                if (!copy.hasTag()) return getStackingIndex(copy);
+                value = copy;
+            }
+            // Kiterino end
             if (!this.itemstackToId.containsKey(value)) {
                 final int id = this.idCounter.incrementAndGet();
                 final ItemStack copy = value.copy();
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index b380fc87b9093e8596ae3f02964523152dae4f0e..d1142c32ac195edf074be903b1f2d77ceee38e24 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -768,6 +768,54 @@ public final class ItemStack {
         return stack.is(otherStack.getItem()) && ItemStack.tagMatches(stack, otherStack);
     }
 
+    // Kiterino start
+    public static boolean recipesTagsMatch(@javax.annotation.Nullable ItemStack comparing, @javax.annotation.Nullable ItemStack expected) {
+        if (comparing == null || expected == null) return false;
+        if (comparing.isEmpty()) return expected.isEmpty();
+        if (expected.isEmpty()) return false;
+
+        var compund1 = getStripped(expected.getTag());
+        if (compund1 == null) { // Vanilla tagless item
+            return comparing.is(expected.getItem());
+        }
+
+        String namespace = org.bukkit.Bukkit.getName();
+        if (compund1.contains(namespace, Tag.TAG_COMPOUND)) {
+            compund1 = compund1.getCompound(namespace);
+            if (compund1.contains("Id", Tag.TAG_STRING)) { // Custom kItem
+                var compound2 = getStripped(comparing.getTag());
+                if (compound2 == null) return false;
+                if (!compound2.contains(namespace, Tag.TAG_COMPOUND)) return false;
+
+                compound2 = compound2.getCompound(namespace);
+                return compund1.getString("Id").equals(compound2.getString("Id"));
+            }
+        }
+
+        return comparing.is(expected.getItem());
+    }
+
+    public static @javax.annotation.Nullable CompoundTag getStripped(@javax.annotation.Nullable CompoundTag nbt) {
+        if (nbt == null) return null;
+        nbt = nbt.copy();
+        nbt.remove(ItemStack.TAG_ENCH);
+        nbt.remove(ItemStack.TAG_DAMAGE);
+        // Only keep custom item ID
+        String namespace = org.bukkit.Bukkit.getName();
+        if (nbt.contains(namespace, Tag.TAG_COMPOUND)) {
+            var knbt = nbt.getCompound(namespace);
+            nbt.remove(namespace);
+            if (knbt.contains("Id", Tag.TAG_STRING)) {
+                String id = knbt.getString("Id");
+                knbt = new CompoundTag();
+                knbt.putString("Id", id);
+                nbt.put(namespace, knbt);
+            }
+        }
+        return nbt.isEmpty() ? null : nbt;
+    }
+    // Kiterino end
+
     public String getDescriptionId() {
         return this.getItem().getDescriptionId(this);
     }
diff --git a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
index 07218093a085293f87d8e562f100688580bdd343..ed643bd24c1cf3baf74e19f6ccff3581badba503 100644
--- a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
+++ b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
@@ -54,7 +54,7 @@ public final class Ingredient implements Predicate<ItemStack> {
                 return recipeitemstack_provider.getItems().stream();
                 // PaperPR start
             }).distinct().peek(stack -> {
-                stack.isExactRecipeIngredient = this.exact;
+                stack.isExactRecipeIngredient = this.exact || this.kItem; // Kiterino
             }).toArray((i) -> {
                 // PaperPR end
                 return new ItemStack[i];
