From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 14 Jul 2021 17:03:20 +0300
Subject: [PATCH] Max stack size API


diff --git a/src/main/java/net/minecraft/world/item/Item.java b/src/main/java/net/minecraft/world/item/Item.java
index c072e3e9742f733aad0d13833fb700c42ca1c356..4dd799d6fb1969f7bdfb8112431515b7477d7c9d 100644
--- a/src/main/java/net/minecraft/world/item/Item.java
+++ b/src/main/java/net/minecraft/world/item/Item.java
@@ -60,7 +60,7 @@ public class Item implements FeatureElement, ItemLike {
     public static final int MAX_BAR_WIDTH = 13;
     private final Holder.Reference<Item> builtInRegistryHolder = BuiltInRegistries.ITEM.createIntrusiveHolder(this);
     public final Rarity rarity;
-    private final int maxStackSize;
+    private int maxStackSize; // Kiterino - remove final
     private final int maxDamage;
     private final boolean isFireResistant;
     @Nullable
@@ -156,6 +156,12 @@ public class Item implements FeatureElement, ItemLike {
         return this.maxStackSize;
     }
 
+    // Kiterino start
+    public void setMaxStackSize(int maxStackSize) {
+        this.maxStackSize = maxStackSize;
+    }
+    // Kiterino end
+
     public final int getMaxDamage() {
         return this.maxDamage;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f1e73c507f556fd54013c902779c5b63b9ab5048..578b50459c86a8ba0208556c0c94800460224e3b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3363,4 +3363,12 @@ public final class CraftServer implements Server {
         return getServer().lagging;
     }
     // Purpur end
+
+    // Kiterino start
+    @Override
+    public void setMaxStackSize(org.bukkit.Material material, int maxStackSize) {
+        Server.super.setMaxStackSize(material, maxStackSize);
+        CraftItemStack.asNMSCopy(new ItemStack(material)).getItem().setMaxStackSize(maxStackSize);
+    }
+    // Kiterino end
 }
