From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 14 Jul 2021 17:03:20 +0300
Subject: [PATCH] Max stack size API


diff --git a/src/main/java/net/minecraft/world/item/Item.java b/src/main/java/net/minecraft/world/item/Item.java
index dd2f6c67533cc3344a171fe3ae9b3704796d8c50..e8a9447ac5560c925df0a3e7cfaaee15bdd5df52 100644
--- a/src/main/java/net/minecraft/world/item/Item.java
+++ b/src/main/java/net/minecraft/world/item/Item.java
@@ -60,7 +60,7 @@ public class Item implements FeatureElement, ItemLike {
     public static final int MAX_BAR_WIDTH = 13;
     private final Holder.Reference<Item> builtInRegistryHolder = BuiltInRegistries.ITEM.createIntrusiveHolder(this);
     public final Rarity rarity;
-    private final int maxStackSize;
+    private int maxStackSize; // Kiterino - remove final
     private final int maxDamage;
     private final boolean isFireResistant;
     @Nullable
@@ -156,6 +156,12 @@ public class Item implements FeatureElement, ItemLike {
         return this.maxStackSize;
     }
 
+    // Kiterino start
+    public void setMaxStackSize(int maxStackSize) {
+        this.maxStackSize = maxStackSize;
+    }
+    // Kiterino end
+
     public final int getMaxDamage() {
         return this.maxDamage;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1ca42de0efd77cf98772ccce0b6a216385c20233..68ace95578411c8ac772f10469f405641a335430 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3229,4 +3229,12 @@ public final class CraftServer implements Server {
         return getServer().lagging;
     }
     // Purpur end
+
+    // Kiterino start
+    @Override
+    public void setMaxStackSize(org.bukkit.Material material, int maxStackSize) {
+        Server.super.setMaxStackSize(material, maxStackSize);
+        CraftItemStack.asCraftCopy(new ItemStack(material)).handle.getItem().setMaxStackSize(maxStackSize);
+    }
+    // Kiterino end
 }
