From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 18 Dec 2021 05:06:48 +0100
Subject: [PATCH] Add block place API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 68846615a058ce40a4ad8fd73cd974be862897c0..56b7f13e9e4b3d218338971035f8efd869691f98 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3153,4 +3153,37 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().connection.send(packet);
     }
     // Purpur end
+
+    // Kiterino start
+    @Override
+    public boolean placeBlock(org.bukkit.inventory.EquipmentSlot hand, Location location, org.bukkit.block.BlockFace blockFace, boolean consume) {
+        return placeBlock(hand, getInventory().getItem(hand), location, blockFace, consume);
+    }
+
+    @Override
+    public boolean placeBlock(ItemStack block, Location location, org.bukkit.block.BlockFace blockFace, boolean consume) {
+        return placeBlock(org.bukkit.inventory.EquipmentSlot.HAND, block, location, blockFace, consume);
+    }
+
+    private boolean placeBlock(org.bukkit.inventory.EquipmentSlot hand, ItemStack block, Location location, org.bukkit.block.BlockFace blockFace, boolean consume) {
+        if (block == null) return false;
+        if (!block.getType().isBlock()) return false;
+
+        var nmsItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(block);
+        var pos = new BlockPos(location.getX(), location.getY(), location.getZ());
+        var nmsPlayer = getHandle();
+
+        var mop = new net.minecraft.world.phys.BlockHitResult(new Vec3(location.getX(), location.getY(), location.getZ()), com.google.common.base.Enums.getIfPresent(net.minecraft.core.Direction.class, blockFace.toString()).or(net.minecraft.core.Direction.UP), pos, false);
+
+        var interactionHand = hand == org.bukkit.inventory.EquipmentSlot.OFF_HAND ? net.minecraft.world.InteractionHand.OFF_HAND : net.minecraft.world.InteractionHand.MAIN_HAND;
+        var result = nmsItem.useOn(new net.minecraft.world.item.context.BlockPlaceContext(nmsPlayer, interactionHand, nmsItem, mop), interactionHand);
+        if (result != net.minecraft.world.InteractionResult.CONSUME) return false;
+
+        swingHand(hand);
+        getWorld().playSound(location, location.getBlock().getBlockData().getSoundGroup().getPlaceSound(), 1F, 1F);
+        if (consume && getGameMode() != GameMode.CREATIVE) block.subtract();
+
+        return true;
+    }
+    // Kiterino end
 }
