From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 7 Jun 2022 14:48:33 +0300
Subject: [PATCH] Add PlayerStartUsingItemEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index cb0b2f5876e69ecf886a39ea245f38a92318dc9a..9f53aa88626508b24773ec2285e30e6aaf648fdd 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3938,7 +3938,18 @@ public abstract class LivingEntity extends Entity {
         if (!itemstack.isEmpty() && !this.isUsingItem() || forceUpdate) { // Paper use override flag
             this.useItem = itemstack;
             // Paper start - lag compensate eating
-            this.useItemRemaining = this.totalEatTimeTicks = itemstack.getUseDuration();
+            // Kiterino start - custom item use duration
+            int useDuration = itemstack.getUseDuration();
+            if (this instanceof ServerPlayer player) {
+                var event = new me.sosedik.kiterino.event.player.PlayerStartUsingItemEvent(player.getBukkitEntity(), itemstack.asBukkitMirror(), useDuration);
+                if (!event.callEvent()) {
+                    stopUsingItem();
+                    return;
+                }
+                useDuration = event.getUseDuration();
+            }
+            this.useItemRemaining = this.totalEatTimeTicks = useDuration;
+            // Kiterino end
             this.eatStartTime = System.nanoTime();
             // Paper end
             if (!this.level.isClientSide) {
