From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 17 Dec 2024 11:10:01 +0200
Subject: [PATCH] Allow forcing arrows to fall


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index e136738ed53a488ad0aa67a04237ac6243fe712c..088d8e60e23d94141fa687166136effb721eed96 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -368,7 +368,7 @@ public abstract class AbstractArrow extends Projectile {
         return this.isInGround() && this.level().noCollision((new AABB(this.position(), this.position())).inflate(0.06D));
     }
 
-    private void startFalling() {
+    public void startFalling() { // Kiterino - private -> public - Allow forcing arrows to fall
         this.setInGround(false);
         Vec3 vec3d = this.getDeltaMovement();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftAbstractArrow.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftAbstractArrow.java
index d0c30fd12aa9866900fe72b97d10c257479cf010..0fd4501cf2e6fa48abaa5796d1bc7021c3298988 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftAbstractArrow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftAbstractArrow.java
@@ -185,4 +185,11 @@ public class CraftAbstractArrow extends AbstractProjectile implements AbstractAr
         this.getHandle().projectileSource = shooter;
     }
     // Paper end - Fix PickupStatus getting reset
+
+    // Kiterino start - Allow forcing arrows to fall
+    @Override
+    public void startFalling() {
+        this.getHandle().startFalling();
+    }
+    // Kiterino end - Allow forcing arrows to fall
 }
