From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 14 Jul 2021 17:03:20 +0300
Subject: [PATCH] Max stack size API


diff --git a/src/main/java/net/minecraft/world/item/Item.java b/src/main/java/net/minecraft/world/item/Item.java
index 253d3c1b83d1e8f5d1afbdb5fb7007196ed94436..6cf1762cafa628dd88c89c5df7b400758e054401 100644
--- a/src/main/java/net/minecraft/world/item/Item.java
+++ b/src/main/java/net/minecraft/world/item/Item.java
@@ -59,7 +59,7 @@ public class Item implements FeatureElement, ItemLike {
     public static final int MAX_BAR_WIDTH = 13;
     private final Holder.Reference<Item> builtInRegistryHolder = BuiltInRegistries.ITEM.createIntrusiveHolder(this);
     public final Rarity rarity;
-    private final int maxStackSize;
+    private int maxStackSize; // Kiterino - remove final
     private final int maxDamage;
     private final boolean isFireResistant;
     @Nullable
@@ -155,6 +155,12 @@ public class Item implements FeatureElement, ItemLike {
         return this.maxStackSize;
     }
 
+    // Kiterino start
+    public void setMaxStackSize(int maxStackSize) {
+        this.maxStackSize = maxStackSize;
+    }
+    // Kiterino end
+
     public final int getMaxDamage() {
         return this.maxDamage;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index bdb25cba758572f673eb25d66edb92e02c674966..46740bbf609c7f761ec9299a562669e8aaff9c90 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3025,4 +3025,12 @@ public final class CraftServer implements Server {
         return getServer().lagging;
     }
     // Purpur end
+
+    // Kiterino start
+    @Override
+    public void setMaxStackSize(org.bukkit.Material material, int maxStackSize) {
+        Server.super.setMaxStackSize(material, maxStackSize);
+        CraftItemStack.asCraftCopy(new ItemStack(material)).handle.getItem().setMaxStackSize(maxStackSize);
+    }
+    // Kiterino end
 }
