From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 7 Jun 2022 15:31:47 +0300
Subject: [PATCH] Vanilla edible items


diff --git a/src/main/java/net/minecraft/world/item/Item.java b/src/main/java/net/minecraft/world/item/Item.java
index 4dd799d6fb1969f7bdfb8112431515b7477d7c9d..b353a5d669021b53fc5fd02c8209fdddc29b4534 100644
--- a/src/main/java/net/minecraft/world/item/Item.java
+++ b/src/main/java/net/minecraft/world/item/Item.java
@@ -68,7 +68,7 @@ public class Item implements FeatureElement, ItemLike {
     @Nullable
     private String descriptionId;
     @Nullable
-    private final FoodProperties foodProperties;
+    private FoodProperties foodProperties; // Kiterino - remove final
     private final FeatureFlagSet requiredFeatures;
 
     public static int getId(Item item) {
@@ -354,6 +354,12 @@ public class Item implements FeatureElement, ItemLike {
         return this.foodProperties;
     }
 
+    // Kiterino start
+    public void setFoodProperties(@Nullable FoodProperties foodProperties) {
+        this.foodProperties = foodProperties;
+    }
+    // Kiterino end
+
     public SoundEvent getDrinkingSound() {
         return SoundEvents.GENERIC_DRINK;
     }
diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index 35b3e2658a376769fa3b2127e68406b3285a09e5..b3b9956c8816372125d29dbb10fe72d1614fbbb6 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -514,12 +514,23 @@ public class PurpurConfig {
         );
 
         properties.getKeys(false).forEach(foodKey -> {
-            FoodProperties food = Foods.ALL_PROPERTIES.get(foodKey);
-            if (food == null) {
-                PurpurConfig.log(Level.SEVERE, "Invalid food property: " + foodKey);
-                return;
+            // Kiterino start
+            FoodProperties foodCheck = Foods.ALL_PROPERTIES.get(foodKey);
+            FoodProperties food;
+            org.bukkit.Material material;
+            if (foodCheck == null) {
+                material = org.bukkit.Material.getMaterial(foodKey.toUpperCase());
+                if (material == null) {
+                    PurpurConfig.log(Level.SEVERE, "Invalid food property: " + foodKey);
+                    return;
+                }
+                food = new FoodProperties.Builder().build();
+            } else {
+                food = foodCheck;
+                material = null;
             }
-            FoodProperties foodDefaults = Foods.DEFAULT_PROPERTIES.get(foodKey);
+            FoodProperties foodDefaults = material == null ? Foods.DEFAULT_PROPERTIES.get(foodKey) : food.copy();
+            // Kiterino end
             food.setNutrition(properties.getInt(foodKey + ".nutrition", foodDefaults.getNutrition()));
             food.setSaturationModifier((float) properties.getDouble(foodKey + ".saturation-modifier", foodDefaults.getSaturationModifier()));
             food.setIsMeat(properties.getBoolean(foodKey + ".is-meat", foodDefaults.isMeat()));
@@ -564,6 +575,7 @@ public class PurpurConfig {
                     food.getEffects().add(Pair.of(new MobEffectInstance(effect, duration, amplifier, ambient, visible, showIcon), chance));
                 });
             }
+            if (material != null) org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftCopy(new org.bukkit.inventory.ItemStack(material)).handle.getItem().setFoodProperties(food); // Kiterino
         });
     }
 
