From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 14 Oct 2022 09:41:48 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 6f08a94e9373d01880af0285939fea2250498b8d..555a1b394f8c85d4a32e27a06e488dee0fbaf7f1 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3007,6 +3007,20 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         }
     }
 
+    // Kiterino start
+    public boolean setRider(Player rider) {
+        if (!new org.spigotmc.event.entity.EntityMountEvent(rider.getBukkitEntity(), this.getBukkitEntity()).callEvent()) return false;
+        if (!rider.startRiding(this, true)) return false;
+        onMount(rider);
+        this.rider = rider;
+        return true;
+    }
+
+    public boolean hasRider() {
+        return getRider() != null;
+    }
+    // Kiterino end
+
     // Paper start
     protected boolean removePassenger(Entity entity) { return removePassenger(entity, false);}
     protected boolean removePassenger(Entity entity, boolean suppressCancellation) { // CraftBukkit
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index af461265593c9c36706e99c9462089f830518061..91c463b2b84f25025fde47e0142b1f7b3e177e9d 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1810,7 +1810,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
     }
 
     public InteractionResult tryRide(Player player, InteractionHand hand, InteractionResult result) {
-        if (!isRidable()) {
+        if (!hasRider()) { // Kiterino
             return result;
         }
         if (hand != InteractionHand.MAIN_HAND) {
@@ -1838,7 +1838,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
                 return InteractionResult.PASS;
             }
         }
-        if (!player.getBukkitEntity().hasPermission("allow.ride." + net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(getType()).getPath())) {
+        if (isRidable() && !player.getBukkitEntity().hasPermission("allow.ride." + net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(getType()).getPath())) { // Kiterino
             player.sendMiniMessage(org.purpurmc.purpur.PurpurConfig.cannotRideMob);
             return InteractionResult.PASS;
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 8973c8a3bad120e55269bf1b7b810284ad0fe14c..31690b6c446c138fb8160703f54923b6539e86ee 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1534,6 +1534,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return rider != null ? (org.bukkit.entity.Player) rider.getBukkitEntity() : null;
     }
 
+    // Kiterino start
+    @Override
+    public boolean setRider(org.bukkit.entity.Player rider, boolean dismount) {
+        var currentRider = getRider();
+        if (rider == currentRider) return true;
+        if (currentRider != null) {
+            if (!dismount) return false;
+            currentRider.leaveVehicle();
+        }
+        return getHandle().setRider(((CraftPlayer) rider).getHandle());
+    }
+    // Kiterino end
+
     @Override
     public boolean hasRider() {
         return getHandle().getRider() != null;
