From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 17 Oct 2022 15:33:07 +0300
Subject: [PATCH] Trigger PlayerInventorySlotChangeEvent upon changing player's
 cursor


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index ce939f09dff1084acadc03d76e5fbb758e578960..8a30dc5427350d1eb24917f22eec6c7f2ea1dbd0 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -397,7 +397,7 @@ public class ServerPlayer extends Player {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
                             return;
                         }
-                        io.papermc.paper.event.player.PlayerInventorySlotChangeEvent event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asBukkitCopy(stack));
+                        io.papermc.paper.event.player.PlayerInventorySlotChangeEvent event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asCraftMirror(stack)); // Kiterino - mirror to allow live changes
                         event.callEvent();
                         if (event.shouldTriggerAdvancements()) {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index fe46261f5c9b95c4eda189989a3005be0e35742e..3173aba0002a50b7d1961c878f64f7d7df6bcc05 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -119,11 +119,21 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
 
     @Override
     public void setItemOnCursor(ItemStack item) {
+        var oldStack = getItemOnCursor(); // Kiterino
         net.minecraft.world.item.ItemStack stack = CraftItemStack.asNMSCopy(item);
         this.getHandle().containerMenu.setCarried(stack);
         if (this instanceof CraftPlayer) {
             this.getHandle().containerMenu.broadcastCarriedItem(); // Send set slot for cursor
         }
+        // Kiterino start
+        if (this instanceof CraftPlayer player) {
+            var event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(player, -1, oldStack, CraftItemStack.asCraftMirror(stack));
+            event.callEvent();
+            if (event.shouldTriggerAdvancements()) {
+                net.minecraft.advancements.CriteriaTriggers.INVENTORY_CHANGED.trigger(player.getHandle(), player.getHandle().getInventory(), stack);
+            }
+        }
+        // Kiterino end
     }
 
     // Paper start
