From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 17 Oct 2022 15:33:07 +0300
Subject: [PATCH] Trigger PlayerInventorySlotChangeEvent upon changing player's
 cursor


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 234832a5b76451c0e84b292188dc6a34311b74f0..dc8a894c39a56087b1d06f490e662db15122dd1b 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -404,7 +404,7 @@ public class ServerPlayer extends Player {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
                             return;
                         }
-                        io.papermc.paper.event.player.PlayerInventorySlotChangeEvent event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asBukkitCopy(stack));
+                        io.papermc.paper.event.player.PlayerInventorySlotChangeEvent event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asCraftMirror(stack)); // Kiterino - mirror to allow live changes
                         event.callEvent();
                         if (event.shouldTriggerAdvancements()) {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index d4035d470ffead2076c1b5edf49fd344b89ceadc..983609f694c86944ddfd782611cc4dfebf6f3218 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -119,11 +119,21 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
 
     @Override
     public void setItemOnCursor(ItemStack item) {
+        var oldStack = getItemOnCursor(); // Kiterino
         net.minecraft.world.item.ItemStack stack = CraftItemStack.asNMSCopy(item);
         this.getHandle().containerMenu.setCarried(stack);
         if (this instanceof CraftPlayer) {
             this.getHandle().containerMenu.broadcastCarriedItem(); // Send set slot for cursor
         }
+        // Kiterino start
+        if (this instanceof CraftPlayer player) {
+            var event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(player, -1, oldStack, CraftItemStack.asCraftMirror(stack));
+            event.callEvent();
+            if (event.shouldTriggerAdvancements()) {
+                net.minecraft.advancements.CriteriaTriggers.INVENTORY_CHANGED.trigger(player.getHandle(), player.getHandle().getInventory(), stack);
+            }
+        }
+        // Kiterino end
     }
 
     // Paper start
