From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 14 Oct 2022 09:41:48 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f1ca71f96861f2d02b28d01add661579bef5e87e..a351cfb7c5381b5fe9628942c90fe04f70afb62c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2835,6 +2835,20 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return true; // CraftBukkit
     }
 
+    // Kiterino start
+    public boolean setRider(Player rider) {
+        if (!new org.spigotmc.event.entity.EntityMountEvent(rider.getBukkitEntity(), this.getBukkitEntity()).callEvent()) return false;
+        if (!rider.startRiding(this, true)) return false;
+        onMount(rider);
+        this.rider = rider;
+        return true;
+    }
+
+    public boolean hasRider() {
+        return getRider() != null;
+    }
+    // Kiterino end
+
     // Paper start
     protected boolean removePassenger(Entity entity) { return removePassenger(entity, false);}
     protected boolean removePassenger(Entity entity, boolean suppressCancellation) { // CraftBukkit
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index ff48c65b07f139ce678f4c0bedb05f6347638ccb..fe977c8bb92fb652febb951cdd9f862d7335d005 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1792,7 +1792,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
     }
 
     public InteractionResult tryRide(Player player, InteractionHand hand, InteractionResult result) {
-        if (!isRidable()) {
+        if (!hasRider()) { // Kiterino
             return result;
         }
         if (hand != InteractionHand.MAIN_HAND) {
@@ -1820,7 +1820,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
                 return InteractionResult.PASS;
             }
         }
-        if (!player.getBukkitEntity().hasPermission("allow.ride." + net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(getType()).getPath())) {
+        if (isRidable() && !player.getBukkitEntity().hasPermission("allow.ride." + net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(getType()).getPath())) { // Kiterino
             player.sendMiniMessage(org.purpurmc.purpur.PurpurConfig.cannotRideMob);
             return InteractionResult.PASS;
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 9f1017e2e51bd6e0c9a6a59663e1acbf17675f3b..916cac2bb36f3c017f082a13ca63d57c4f343449 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1459,6 +1459,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return rider != null ? (org.bukkit.entity.Player) rider.getBukkitEntity() : null;
     }
 
+    // Kiterino start
+    @Override
+    public boolean setRider(org.bukkit.entity.Player rider, boolean dismount) {
+        var currentRider = getRider();
+        if (rider == currentRider) return true;
+        if (currentRider != null) {
+            if (!dismount) return false;
+            currentRider.leaveVehicle();
+        }
+        return getHandle().setRider(((CraftPlayer) rider).getHandle());
+    }
+    // Kiterino end
+
     @Override
     public boolean hasRider() {
         return getHandle().getRider() != null;
