From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 20 Oct 2022 15:55:19 +0300
Subject: [PATCH] Update goal target when changing target


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
index 606357b6e8aa363123b28c3d7188348ff6e9192e..406bd807361759638513336b3d7322f0ad07d56e 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -146,4 +146,10 @@ public abstract class TargetGoal extends Goal {
         this.unseenMemoryTicks = time;
         return this;
     }
+
+    // Kiterino start
+    public void setTargetMob(@Nullable LivingEntity targetMob) {
+        this.targetMob = targetMob;
+    }
+    // Kiterino end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index 5f6a1fb323782816f706ccf7df922cb203238302..cb2341b1b2d8405e3530fa1317dad24edb16bfcf 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -30,6 +30,17 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
         } else if (target instanceof CraftLivingEntity) {
             entity.setTarget(((CraftLivingEntity) target).getHandle(), null, false);
         }
+        // Kiterino start
+        entity.targetSelector.getRunningGoals().forEach(goal -> {
+            if (goal.getGoal() instanceof net.minecraft.world.entity.ai.goal.target.TargetGoal targetGoal) {
+                if (target == null) {
+                    targetGoal.setTargetMob(null);
+                } else if (target instanceof CraftLivingEntity craftLivingEntity) {
+                    targetGoal.setTargetMob(craftLivingEntity.getHandle());
+                }
+            }
+        });
+        // Kiterino end
     }
 
     @Override
