From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 14 Oct 2022 09:41:48 +0300
Subject: [PATCH] Allow setting entity's rider


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 8eec827d70d8e42029b72c56ded2dc594d25076f..5dd2f56ef1de54637add0e29325c996b8dc8212d 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2861,6 +2861,20 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return true; // CraftBukkit
     }
 
+    // Kiterino start
+    public boolean setRider(Player rider) {
+        if (!new org.spigotmc.event.entity.EntityMountEvent(rider.getBukkitEntity(), this.getBukkitEntity()).callEvent()) return false;
+        if (!rider.startRiding(this, true)) return false;
+        onMount(rider);
+        this.rider = rider;
+        return true;
+    }
+
+    public boolean hasRider() {
+        return getRider() != null;
+    }
+    // Kiterino end
+
     // Paper start
     protected boolean removePassenger(Entity entity) { return removePassenger(entity, false);}
     protected boolean removePassenger(Entity entity, boolean suppressCancellation) { // CraftBukkit
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 1cc1a2aa075818504ab3e6325183a9c5f3e33657..71e2e1d9c7f2b0c7e546df5098aef338d9e430ee 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1772,7 +1772,7 @@ public abstract class Mob extends LivingEntity {
     }
 
     public InteractionResult tryRide(Player player, InteractionHand hand, InteractionResult result) {
-        if (!isRidable()) {
+        if (!hasRider()) { // Kiterino
             return result;
         }
         if (hand != InteractionHand.MAIN_HAND) {
@@ -1800,7 +1800,7 @@ public abstract class Mob extends LivingEntity {
                 return InteractionResult.PASS;
             }
         }
-        if (!player.getBukkitEntity().hasPermission("allow.ride." + net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(getType()).getPath())) {
+        if (isRidable() && !player.getBukkitEntity().hasPermission("allow.ride." + net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(getType()).getPath())) { // Kiterino
             player.sendMiniMessage(org.purpurmc.purpur.PurpurConfig.cannotRideMob);
             return InteractionResult.PASS;
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index d1c16a80c2cc4f0e96f3a7f04e5bc69657206f97..39559d5b37b4cf3f5404bdf68d66f5755c8ff3f3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1421,6 +1421,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return rider != null ? (org.bukkit.entity.Player) rider.getBukkitEntity() : null;
     }
 
+    // Kiterino start
+    @Override
+    public boolean setRider(org.bukkit.entity.Player rider, boolean dismount) {
+        var currentRider = getRider();
+        if (rider == currentRider) return true;
+        if (currentRider != null) {
+            if (!dismount) return false;
+            currentRider.leaveVehicle();
+        }
+        return getHandle().setRider(((CraftPlayer) rider).getHandle());
+    }
+    // Kiterino end
+
     @Override
     public boolean hasRider() {
         return getHandle().getRider() != null;
