From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 17 Oct 2022 15:33:07 +0300
Subject: [PATCH] Trigger PlayerInventorySlotChangeEvent upon changing player's
 cursor


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 2d0aa701dfc1fb1db24f9b3496e4be3779248834..3dd475614d50b342ba106280eade730bfb5f90b3 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -350,7 +350,7 @@ public class ServerPlayer extends Player {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
                             return;
                         }
-                        io.papermc.paper.event.player.PlayerInventorySlotChangeEvent event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asBukkitCopy(stack));
+                        io.papermc.paper.event.player.PlayerInventorySlotChangeEvent event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asCraftMirror(stack)); // Kiterino - mirror to allow live changes
                         event.callEvent();
                         if (event.shouldTriggerAdvancements()) {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 1f894c8515df8f96fc02ef5175bb05d4066f7c8b..c47f48d65f110712fda0a77fbbe39949f228c2e9 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -118,11 +118,21 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
 
     @Override
     public void setItemOnCursor(ItemStack item) {
+        var oldStack = getItemOnCursor(); // Kiterino
         net.minecraft.world.item.ItemStack stack = CraftItemStack.asNMSCopy(item);
         this.getHandle().containerMenu.setCarried(stack);
         if (this instanceof CraftPlayer) {
             this.getHandle().containerMenu.broadcastCarriedItem(); // Send set slot for cursor
         }
+        // Kiterino start
+        if (this instanceof CraftPlayer player) {
+            var event = new io.papermc.paper.event.player.PlayerInventorySlotChangeEvent(player, -1, oldStack, CraftItemStack.asCraftMirror(stack));
+            event.callEvent();
+            if (event.shouldTriggerAdvancements()) {
+                net.minecraft.advancements.CriteriaTriggers.INVENTORY_CHANGED.trigger(player.getHandle(), player.getHandle().getInventory(), stack);
+            }
+        }
+        // Kiterino end
     }
 
     // Paper start
