From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 26 Dec 2021 19:08:38 +0200
Subject: [PATCH] Deny stacking potions in brewing stand


diff --git a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
index d5acd56f39af64f2f72d418da48fa7fffd680243..07575b415b6a7d0e29a0f1c9b90547df7c76d877 100644
--- a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -768,6 +768,11 @@ public abstract class AbstractContainerMenu {
         return this.moveItemStackTo(stack, startIndex, endIndex, fromLast, false);
     }
     protected boolean moveItemStackTo(ItemStack stack, int startIndex, int endIndex, boolean fromLast, boolean isCheck) {
+        // Kiterino start
+        return this.moveItemStackTo(stack, startIndex, endIndex, fromLast, isCheck, false);
+    }
+    protected boolean moveItemStackTo(ItemStack stack, int startIndex, int endIndex, boolean fromLast, boolean isCheck, boolean ignoreStacks) {
+        // Kiterino end
         if (isCheck) {
             stack = stack.copy();
         }
@@ -782,7 +787,7 @@ public abstract class AbstractContainerMenu {
         Slot slot;
         ItemStack itemstack1;
 
-        if (stack.isStackable()) {
+        if (stack.isStackable() && !ignoreStacks) { // Kiterino
             while (!stack.isEmpty()) {
                 if (fromLast) {
                     if (k < startIndex) {
diff --git a/src/main/java/net/minecraft/world/inventory/BrewingStandMenu.java b/src/main/java/net/minecraft/world/inventory/BrewingStandMenu.java
index 13720c2df8ac37d020d4a785e48c45877edf74b9..c6dbd248488d4a61fb9bcb9ef7afe87f7961265f 100644
--- a/src/main/java/net/minecraft/world/inventory/BrewingStandMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/BrewingStandMenu.java
@@ -93,8 +93,8 @@ public class BrewingStandMenu extends AbstractContainerMenu {
                     if (!this.moveItemStackTo(itemstack1, 3, 4, false)) {
                         return ItemStack.EMPTY;
                     }
-                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(itemstack) && itemstack.getCount() == 1) {
-                    if (!this.moveItemStackTo(itemstack1, 0, 3, false)) {
+                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(itemstack)/* && itemstack.getCount() == 1*/) {
+                    if (!this.moveItemStackTo(itemstack1, 0, 3, false, false, true)) { // Kiterino - Move only one potion
                         return ItemStack.EMPTY;
                     }
                 } else if (index >= 5 && index < 32) {
