From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 24 Oct 2022 18:51:55 +0300
Subject: [PATCH] Get destroy speed from BlockData


diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
index 27344a44b776311801b03034730fd6710190e11a..6567d220d35ef45495ecd765e9190fa769cd33ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
@@ -682,4 +682,28 @@ public class CraftBlockData implements BlockData {
         return this.state.isRandomlyTicking();
     }
     // Paper end
+
+    // Kiterino start
+    @Override
+    public float getDestroySpeed(org.bukkit.inventory.ItemStack itemStack, boolean considerEnchants) {
+        net.minecraft.world.item.ItemStack nmsItemStack;
+        if (itemStack instanceof org.bukkit.craftbukkit.inventory.CraftItemStack craftItemStack) {
+            nmsItemStack = craftItemStack.handle;
+            if (nmsItemStack == null) {
+                nmsItemStack = net.minecraft.world.item.ItemStack.EMPTY;
+            }
+        } else {
+            nmsItemStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(itemStack);
+        }
+        float speed = nmsItemStack.getDestroySpeed(state);
+        if (speed > 1.0F && considerEnchants) {
+            int enchantLevel = net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.BLOCK_EFFICIENCY, nmsItemStack);
+            if (enchantLevel > 0) {
+                speed += enchantLevel * enchantLevel + 1;
+            }
+        }
+        return speed;
+    }
+    // Kiterino end
+
 }
