From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 2 Dec 2022 09:28:31 +0200
Subject: [PATCH] Super dumb liquid ticks limiter


diff --git a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
index 05d561289a9f6a23b6c181661bd10784e6e5e99c..774df73be7a62150f736a26d1c8bb1fe11f83d89 100644
--- a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
@@ -807,6 +807,14 @@ public abstract class FlowingFluid extends Fluid {
                 world.updateNeighborsAt(pos, iblockdata.getBlock());
             }
         }
+        // Kiterino start - Super dumb liquid physics limiter to reduce the load
+        else {
+            if (org.bukkit.Bukkit.isLagging() || org.bukkit.Bukkit.getCurrentTick() % 20 > 12) {
+                world.scheduleTick(pos, this, 9);
+                return;
+            }
+        }
+        // Kiterino end
 
         this.spread(world, pos, state);
     }
