From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 17 Oct 2022 15:33:07 +0300
Subject: [PATCH] Trigger PlayerInventorySlotChangeEvent upon changing player's
 cursor


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 757e097125c869c24fe89f9d05be587838d5cc17..6b41c5fdc9aa54527579fa0ad2946cb5ecc26de8 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -333,7 +333,7 @@ public class ServerPlayer extends Player {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
                             return;
                         }
-                        com.destroystokyo.paper.event.player.PlayerInventorySlotChangeEvent event = new com.destroystokyo.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asBukkitCopy(stack));
+                        com.destroystokyo.paper.event.player.PlayerInventorySlotChangeEvent event = new com.destroystokyo.paper.event.player.PlayerInventorySlotChangeEvent(ServerPlayer.this.getBukkitEntity(), slotId, CraftItemStack.asBukkitCopy(oldStack), CraftItemStack.asCraftMirror(stack)); // Kiterino - mirror to allow live changes
                         event.callEvent();
                         if (event.isTriggerAdvancements()) {
                             CriteriaTriggers.INVENTORY_CHANGED.trigger(ServerPlayer.this, ServerPlayer.this.getInventory(), stack);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 1b925ae42608bceddfe9c536146c211d3f37a8c2..683b56014e0992eb57894465543e4677c7615df7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -118,11 +118,21 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
 
     @Override
     public void setItemOnCursor(ItemStack item) {
+        var oldStack = getItemOnCursor(); // Kiterino
         net.minecraft.world.item.ItemStack stack = CraftItemStack.asNMSCopy(item);
         this.getHandle().containerMenu.setCarried(stack);
         if (this instanceof CraftPlayer) {
             this.getHandle().containerMenu.broadcastCarriedItem(); // Send set slot for cursor
         }
+        // Kiterino start
+        if (this instanceof CraftPlayer player) {
+            var event = new com.destroystokyo.paper.event.player.PlayerInventorySlotChangeEvent(player, -1, oldStack, CraftItemStack.asCraftMirror(stack));
+            event.callEvent();
+            if (event.isTriggerAdvancements()) {
+                net.minecraft.advancements.CriteriaTriggers.INVENTORY_CHANGED.trigger(player.getHandle(), player.getHandle().getInventory(), stack);
+            }
+        }
+        // Kiterino end
     }
 
     // Paper start
