From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 2 Dec 2022 15:15:41 +0200
Subject: [PATCH] Detecting Powered Rails


diff --git a/src/main/java/net/minecraft/world/level/block/DetectorRailBlock.java b/src/main/java/net/minecraft/world/level/block/DetectorRailBlock.java
index 129f0a73eec1d557fbd3034b1434d0dbd722b8db..2fd6083008d6d4f5dd76c7abfdfbf8110281c2a6 100644
--- a/src/main/java/net/minecraft/world/level/block/DetectorRailBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/DetectorRailBlock.java
@@ -181,7 +181,7 @@ public class DetectorRailBlock extends BaseRailBlock {
         return 0;
     }
 
-    private <T extends AbstractMinecart> List<T> getInteractingMinecartOfType(Level world, BlockPos pos, Class<T> entityClass, Predicate<Entity> entityPredicate) {
+    public <T extends AbstractMinecart> List<T> getInteractingMinecartOfType(Level world, BlockPos pos, Class<T> entityClass, Predicate<Entity> entityPredicate) { // Kiterino - private -> public
         // Kiterino start - directional detector rails
         BlockState belowState = world.getBlockState(pos.below());
         if (belowState.is(Blocks.MAGENTA_GLAZED_TERRACOTTA)) {
diff --git a/src/main/java/net/minecraft/world/level/block/PoweredRailBlock.java b/src/main/java/net/minecraft/world/level/block/PoweredRailBlock.java
index e03125281767845564c48c98c3e6b6bbd269ade1..c6ee5f8c084c2e03a544231871c19aa3af93ebeb 100644
--- a/src/main/java/net/minecraft/world/level/block/PoweredRailBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PoweredRailBlock.java
@@ -265,4 +265,60 @@ public class PoweredRailBlock extends BaseRailBlock {
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(PoweredRailBlock.SHAPE, PoweredRailBlock.POWERED, PoweredRailBlock.WATERLOGGED);
     }
+
+    // Kiterino start - copypaste from DetectorRailBlock
+    @Override
+    public boolean isSignalSource(BlockState state) {
+        return true;
+    }
+
+    @Override
+    public int getSignal(BlockState state, net.minecraft.world.level.BlockGetter world, BlockPos pos, net.minecraft.core.Direction direction) {
+        if (!(world instanceof Level level)) return 0;
+        BlockState relativeBlockState = world.getBlockState(pos.relative(direction.getOpposite()));
+        if (!relativeBlockState.is(Blocks.REPEATER)) return 0;
+        if (relativeBlockState.getValue(RepeaterBlock.FACING) != direction) return 0;
+        return ((DetectorRailBlock) Blocks.DETECTOR_RAIL).getInteractingMinecartOfType(level, pos, net.minecraft.world.entity.vehicle.AbstractMinecart.class, e -> true).isEmpty() ? 0 : 15;
+    }
+
+    @Override
+    public boolean hasAnalogOutputSignal(BlockState state) {
+        return true;
+    }
+
+    @Override
+    public int getAnalogOutputSignal(BlockState state, Level world, BlockPos pos) {
+        return Blocks.DETECTOR_RAIL.getAnalogOutputSignal(Blocks.DETECTOR_RAIL.defaultBlockState().setValue(DetectorRailBlock.POWERED, true), world, pos);
+    }
+
+    @Override
+    public void entityInside(BlockState state, Level world, BlockPos pos, net.minecraft.world.entity.Entity entity) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, pos)).callEvent()) { return; } // Paper
+        if (!world.isClientSide) {
+            if (entity instanceof net.minecraft.world.entity.vehicle.AbstractMinecart) {
+                outputSignal(pos, world);
+            }
+            world.scheduleTick(pos, this, 20);
+            world.updateNeighbourForOutputSignal(pos, this);
+        }
+    }
+
+    @Override
+    public void tick(BlockState state, net.minecraft.server.level.ServerLevel world, BlockPos pos, net.minecraft.util.RandomSource random) {
+        world.updateNeighbourForOutputSignal(pos, this);
+        outputSignal(pos, world);
+    }
+
+    private void outputSignal(BlockPos pos, Level world) {
+        for (net.minecraft.core.Direction direction : net.minecraft.core.Direction.Plane.HORIZONTAL) {
+            BlockPos relative = pos.relative(direction);
+            BlockState relativeState = world.getBlockStateIfLoaded(relative);
+            if (relativeState == null) continue;
+            if (relativeState.getBlock() != Blocks.REPEATER) continue;
+            if (relativeState.getValue(RepeaterBlock.FACING) != direction.getOpposite()) continue;
+
+            relativeState.neighborChanged(world, relative, this, pos, false);
+        }
+    }
+    // Kiterino end
 }
