From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 22 Oct 2022 21:34:57 +0300
Subject: [PATCH] Fix entity serialization


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f2d28e5da1bb1ed9aec5299277f1fee4c938521d..41af2d10417a36cf4fb5d781806eae01ed451d01 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2192,15 +2192,6 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         }
     }
 
-    // Paper start - Entity serialization api
-    public boolean serializeEntity(CompoundTag compound) {
-        List<Entity> pass = new java.util.ArrayList<>(this.getPassengers());
-        this.passengers = ImmutableList.of();
-        boolean result = save(compound);
-        this.passengers = ImmutableList.copyOf(pass);
-        return result;
-    }
-    // Paper end
     public boolean save(CompoundTag nbt) {
         return this.isPassenger() ? false : this.saveAsPassenger(nbt);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index d506d6f2682e558bc92cf02b474d89dd5fa8a300..fff41e09d0006d66552cdd9466ae04a998f898e7 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -467,8 +467,13 @@ public final class CraftMagicNumbers implements UnsafeValues {
         Preconditions.checkNotNull(entity, "null cannot be serialized");
         Preconditions.checkArgument(entity instanceof org.bukkit.craftbukkit.entity.CraftEntity, "only CraftEntities can be serialized");
 
+        net.minecraft.world.entity.Entity nmsEntity = ((org.bukkit.craftbukkit.entity.CraftEntity) entity).getHandle();
         CompoundTag compound = new CompoundTag();
-        ((org.bukkit.craftbukkit.entity.CraftEntity) entity).getHandle().serializeEntity(compound);
+        compound.putString("id", net.minecraft.world.entity.EntityType.getKey(nmsEntity.getType()).toString());
+        List<net.minecraft.world.entity.Entity> pass = new java.util.ArrayList<>(nmsEntity.getPassengers());
+        nmsEntity.passengers = com.google.common.collect.ImmutableList.of();
+        nmsEntity.saveWithoutId(compound);
+        nmsEntity.passengers = com.google.common.collect.ImmutableList.copyOf(pass);
         return serializeNbtToBytes(compound);
     }
 
