From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 2 Dec 2022 23:05:04 +0200
Subject: [PATCH] Increase manual minecart speed


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index 4acce4c2f8f7c2eac88189914912c0f31d050bcb..be86320227a2c2a2af5e1abf481248c2326d5c86 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -694,8 +694,8 @@ public abstract class AbstractMinecart extends VehicleEntity {
             double d9 = vec3d2.horizontalDistanceSqr();
             double d10 = this.getDeltaMovement().horizontalDistanceSqr();
 
-            if (d9 > 1.0E-4D && d10 < 0.01D) {
-                this.setDeltaMovement(this.getDeltaMovement().add(vec3d2.x * 0.1D, 0.0D, vec3d2.z * 0.1D));
+            if (d9 > 1.0E-4D && d10 < 0.04D) { // Kiterino - increase max manual speed
+                this.setDeltaMovement(this.getDeltaMovement().add(vec3d2.x * 0.12D, 0.0D, vec3d2.z * 0.12D)); // Kiterino - slightly increase speed buildup
                 flag1 = false;
             }
         }
