From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 4 Dec 2022 21:57:07 +0200
Subject: [PATCH] Allow placing rails on more surfaces


diff --git a/src/main/java/net/minecraft/world/level/block/BaseRailBlock.java b/src/main/java/net/minecraft/world/level/block/BaseRailBlock.java
index 547d09c3a1cd0c5401aacb78886896365ba192e7..8f623c870558e09eb483853fa6d8b43691a74230 100644
--- a/src/main/java/net/minecraft/world/level/block/BaseRailBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BaseRailBlock.java
@@ -50,7 +50,7 @@ public abstract class BaseRailBlock extends Block implements SimpleWaterloggedBl
 
     @Override
     public boolean canSurvive(BlockState state, LevelReader world, BlockPos pos) {
-        return canSupportRigidBlock(world, pos.below());
+        return canSupportCenter(world, pos.below(), Direction.UP); // Kiterino
     }
 
     @Override
@@ -85,6 +85,7 @@ public abstract class BaseRailBlock extends Block implements SimpleWaterloggedBl
     }
 
     private static boolean shouldBeRemoved(BlockPos pos, Level world, RailShape shape) {
+        if (canSupportCenter(world, pos.below(), Direction.UP)) return false; // Kiterino
         if (!canSupportRigidBlock(world, pos.below())) {
             return true;
         } else {
