From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 5 Dec 2022 11:17:51 +0200
Subject: [PATCH] Lower minecarts rail sticking height


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index f5f09bcfe85b7cf3bc1ae101d26ba234e82db5b9..43394bb43874af9b36eb18e8b4a9ee65673a5a4c 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -1018,12 +1018,19 @@ public abstract class AbstractMinecart extends Entity {
         int k = Mth.floor(z);
 
         if (this.level().getBlockState(new BlockPos(i, j - 1, k)).is(BlockTags.RAILS)) {
-            --j;
+            // Kiterino start
+            if (getY() - i < 0.2) {
+                BlockState belowState = this.level().getBlockState(new BlockPos(i, j - 1, k));
+                if (belowState.getBlock() instanceof BaseRailBlock railBlock && belowState.getValue(railBlock.getShapeProperty()).isAscending()) {
+                    --j;
+                }
+            }
+            // Kiterino end
         }
 
         BlockState iblockdata = this.level().getBlockState(new BlockPos(i, j, k));
 
-        if (BaseRailBlock.isRail(iblockdata)) {
+        if (iblockdata.is(BlockTags.RAILS) && iblockdata.getBlock() instanceof BaseRailBlock railBlock && (iblockdata.getValue(railBlock.getShapeProperty()).isAscending() || getY() < getBlockY() + 0.1)) { // Kiterino
             RailShape blockpropertytrackposition = (RailShape) iblockdata.getValue(((BaseRailBlock) iblockdata.getBlock()).getShapeProperty());
 
             y = (double) j;
