From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 4 Dec 2022 20:08:59 +0200
Subject: [PATCH] Collision based landing pos


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 808425e184f4c48ef07560b2a8bae5c72021fb8f..d70d030cd52ab9ea6b8cda15643906b6fe5e7cee 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1454,6 +1454,14 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     /** @deprecated */
     @Deprecated
     public BlockPos getOnPosLegacy() {
+        // Kiterino start
+        for (VoxelShape shape : this.level.getBlockCollisions(this, this.getBoundingBox().move(0, 0.2D, 0))) {
+            Vec3 boundingBoxShape = shape.bounds().getCenter();
+            if (boundingBoxShape.y() <= this.getY()) {
+                return new BlockPos((int) boundingBoxShape.x(), (int) boundingBoxShape.y(), (int) boundingBoxShape.z());
+            }
+        }
+        // Kiterino end
         return this.getOnPos(0.2F);
     }
 
