From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 4 Dec 2022 20:08:59 +0200
Subject: [PATCH] Collision based landing pos


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 428225f3addc32cb254de7b2483db29a7dd54d14..92119df5971f4dd1d715515ad5a7f1584c13c8d6 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1402,6 +1402,14 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     /** @deprecated */
     @Deprecated
     public BlockPos getOnPosLegacy() {
+        // Kiterino start
+        for (VoxelShape shape : this.level.getBlockCollisions(this, this.getBoundingBox().move(0, 0.2D, 0))) {
+            Vec3 boundingBoxShape = shape.bounds().getCenter();
+            if (boundingBoxShape.y() <= this.getY()) {
+                return new BlockPos((int) boundingBoxShape.x(), (int) boundingBoxShape.y(), (int) boundingBoxShape.z());
+            }
+        }
+        // Kiterino end
         return this.getOnPos(0.2F);
     }
 
