From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 4 Dec 2022 20:39:25 +0200
Subject: [PATCH] Recalculate melee target location when lost


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
index d59e0044db03ba30d5bab9a8c126b597b7659015..aa531e4bc2ad11c8eb2f5566e794e7311e37937a 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
@@ -56,9 +56,16 @@ public class MeleeAttackGoal extends Goal {
         } else if (!livingEntity.isAlive()) {
             return false;
         } else {
-            return !this.followingTargetEvenIfNotSeen
-                ? !this.mob.getNavigation().isDone()
-                : this.mob.isWithinRestriction(livingEntity.blockPosition())
+            if (!this.followingTargetEvenIfNotSeen) {
+                boolean idling = this.mob.getNavigation().isDone();
+                if (!idling) return true;
+                if (this.mob.distanceTo(this.mob.getTarget()) < 10) {
+                    this.path = this.mob.getNavigation().createPath(this.mob.getTarget(), 0);
+                    this.mob.getNavigation().moveTo(this.path, this.speedModifier);
+                }
+                return false;
+            }
+            return this.mob.isWithinRestriction(livingEntity.blockPosition())
                     && (!(livingEntity instanceof Player) || !livingEntity.isSpectator() && !((Player)livingEntity).isCreative());
         }
     }
