From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 4 Dec 2022 20:08:59 +0200
Subject: [PATCH] Collision based landing pos


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 2c8cea3dddc8b60ab2fb32a7ece36f1398a424b4..5098766735c059de2d74aa92c9559a534a132ad6 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1443,6 +1443,14 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     /** @deprecated */
     @Deprecated
     public BlockPos getOnPosLegacy() {
+        // Kiterino start
+        for (VoxelShape shape : this.level.getBlockCollisions(this, this.getBoundingBox().move(0, 0.2D, 0))) {
+            Vec3 boundingBoxShape = shape.bounds().getCenter();
+            if (boundingBoxShape.y() <= this.getY()) {
+                return new BlockPos((int) boundingBoxShape.x(), (int) boundingBoxShape.y(), (int) boundingBoxShape.z());
+            }
+        }
+        // Kiterino end
         return this.getOnPos(0.2F);
     }
 
