From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 7 Dec 2022 10:39:11 +0200
Subject: [PATCH] Allow explosions underwater (/fluids)


diff --git a/src/main/java/net/minecraft/world/level/ExplosionDamageCalculator.java b/src/main/java/net/minecraft/world/level/ExplosionDamageCalculator.java
index 5b93c038331c1750260a42726f5bfb97998d93a9..b2373a2f20c3e26af2161e284fd943a616b8f7dc 100644
--- a/src/main/java/net/minecraft/world/level/ExplosionDamageCalculator.java
+++ b/src/main/java/net/minecraft/world/level/ExplosionDamageCalculator.java
@@ -9,9 +9,14 @@ import net.minecraft.world.phys.Vec3;
 
 public class ExplosionDamageCalculator {
     public Optional<Float> getBlockExplosionResistance(Explosion explosion, BlockGetter world, BlockPos pos, BlockState blockState, FluidState fluidState) {
-        return blockState.isAir() && fluidState.isEmpty()
+        // Kiterino start - Allow explosions underwater (/fluids)
+        if (blockState.isAir()) return Optional.empty();
+        if (!fluidState.isEmpty()) return Optional.empty();
+        return Optional.of(blockState.getBlock().getExplosionResistance());
+        /*return blockState.isAir() && fluidState.isEmpty()
             ? Optional.empty()
             : Optional.of(Math.max(blockState.getBlock().getExplosionResistance(), fluidState.getExplosionResistance()));
+         */ // Kiterino end - Allow explosions underwater (/fluids)
     }
 
     public boolean shouldBlockExplode(Explosion explosion, BlockGetter world, BlockPos pos, BlockState state, float power) {
