From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 9 Dec 2022 22:21:40 +0200
Subject: [PATCH] Fallback to Dimension if World UUID is unknown


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 96a6a9cf557ec3d461edcbf84df2e2af3f8d4914..8c359f70c268a8f3069ba21a9da469cf86e709f9 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -212,7 +212,10 @@ public abstract class PlayerList {
             if (bWorld != null) {
                 resourcekey = ((CraftWorld) bWorld).getHandle().dimension();
             } else {
-                resourcekey = Level.OVERWORLD;
+                // Kiterino start
+                DataResult<ResourceKey<Level>> dataResult = Level.RESOURCE_KEY_CODEC.parse(new Dynamic<>(NbtOps.INSTANCE, nbttagcompound.get("Dimension")));
+                resourcekey = dataResult.result().orElse(Level.OVERWORLD);
+                // Kiterino end
             }
         } else if (nbttagcompound != null) {
             // Vanilla migration support
