From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 4 Dec 2022 20:08:59 +0200
Subject: [PATCH] Collision based landing pos


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c9e59d23f24fb7761100f3c244cd6ba2691c4586..921931786c57e5e832d4b153980b24dc997dd806 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1347,6 +1347,14 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     /** @deprecated */
     @Deprecated
     public BlockPos getOnPosLegacy() {
+        // Kiterino start
+        for (VoxelShape shape : this.level.getBlockCollisions(this, this.getBoundingBox().move(0, 0.2D, 0))) {
+            Vec3 boundingBoxShape = shape.bounds().getCenter();
+            if (boundingBoxShape.y() <= this.getY()) {
+                return new BlockPos(boundingBoxShape);
+            }
+        }
+        // Kiterino end
         return this.getOnPos(0.2F);
     }
 
