From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 4 Dec 2022 20:39:25 +0200
Subject: [PATCH] Recalculate melee target location when lost


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
index e6cf828a267f694d3f23e578b62c8f8146f22afb..9dcfb9906b0bc6c43cd55a93991ae423a6207e36 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
@@ -60,7 +60,13 @@ public class MeleeAttackGoal extends Goal {
         } else if (!livingEntity.isAlive()) {
             return false;
         } else if (!this.followingTargetEvenIfNotSeen) {
-            return !this.mob.getNavigation().isDone();
+            boolean idling = this.mob.getNavigation().isDone();
+            if (!idling) return true;
+            if (this.mob.distanceTo(this.mob.getTarget()) < 10) {
+                this.path = this.mob.getNavigation().createPath(this.mob.getTarget(), 0);
+                this.mob.getNavigation().moveTo(this.path, this.speedModifier);
+            }
+            return false;
         } else if (!this.mob.isWithinRestriction(livingEntity.blockPosition())) {
             return false;
         } else {
