From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 5 Dec 2022 18:09:34 +0200
Subject: [PATCH] Render Nether maps at dynamic height


diff --git a/src/main/java/net/minecraft/world/item/EmptyMapItem.java b/src/main/java/net/minecraft/world/item/EmptyMapItem.java
index e92bb09dd3218d5a13e6251bddd0812acadda2be..6788a32c71ead670409d5712dd93885bd2e15b44 100644
--- a/src/main/java/net/minecraft/world/item/EmptyMapItem.java
+++ b/src/main/java/net/minecraft/world/item/EmptyMapItem.java
@@ -25,6 +25,10 @@ public class EmptyMapItem extends ComplexItem {
             user.awardStat(Stats.ITEM_USED.get(this));
             user.level.playSound((Player)null, user, SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT, user.getSoundSource(), 1.0F, 1.0F);
             ItemStack itemStack2 = MapItem.create(world, user.getBlockX(), user.getBlockZ(), (byte)0, true, false);
+            // Kiterino start
+            net.minecraft.nbt.CompoundTag tag = itemStack2.getTag();
+            if (tag != null) tag.putInt("MapYLevel", user.getBlockY());
+            // Kiterino end
             if (itemStack.isEmpty()) {
                 return InteractionResultHolder.consume(itemStack2);
             } else {
diff --git a/src/main/java/net/minecraft/world/item/MapItem.java b/src/main/java/net/minecraft/world/item/MapItem.java
index 45a7e01288f780cf8a812d8e0ae12c4fb79d79e1..05711954023846dc8bf2e269612b8100423ebad9 100644
--- a/src/main/java/net/minecraft/world/item/MapItem.java
+++ b/src/main/java/net/minecraft/world/item/MapItem.java
@@ -109,7 +109,7 @@ public class MapItem extends ComplexItem {
             int i1 = Mth.floor(entity.getZ() - (double) k) / i + 64;
             int j1 = 128 / i;
 
-            if (world.dimensionType().hasCeiling()) {
+            if (false && world.dimensionType().hasCeiling()) { // Kiterino
                 j1 /= 2;
             }
 
@@ -140,7 +140,7 @@ public class MapItem extends ComplexItem {
                                 int k3 = 0;
                                 double d1 = 0.0D;
 
-                                if (world.dimensionType().hasCeiling()) {
+                                if (false && world.dimensionType().hasCeiling()) { // Kiterino
                                     int l3 = k2 + l2 * 231871;
 
                                     l3 = l3 * l3 * 31287121 + l3 * 11;
@@ -157,7 +157,7 @@ public class MapItem extends ComplexItem {
 
                                     for (int i4 = 0; i4 < i; ++i4) {
                                         for (int j4 = 0; j4 < i; ++j4) {
-                                            int k4 = chunk.getHeight(Heightmap.Types.WORLD_SURFACE, i4 + i3, j4 + j3) + 1;
+                                            int k4 = world.dimensionType().hasCeiling() ? getRenderHeight((Player) entity, state) : chunk.getHeight(Heightmap.Types.WORLD_SURFACE, i4 + i3, j4 + j3) + 1; // Kiterino
                                             BlockState iblockdata;
 
                                             if (k4 > world.getMinBuildHeight() + 1) {
@@ -231,6 +231,19 @@ public class MapItem extends ComplexItem {
         }
     }
 
+    // Kiterino start
+    public int getRenderHeight(Player player, MapItemSavedData state) {
+        for (int slot = 0; slot < player.getInventory().getContainerSize(); slot++) {
+            ItemStack item = player.getInventory().getItem(slot);
+            if (item.getItem() instanceof MapItem && MapItem.getSavedData(item, player.getLevel()) == state) {
+                CompoundTag tag = item.getTag();
+                return (tag == null || !tag.contains("MapYLevel")) ? 100 : tag.getInt("MapYLevel");
+            }
+        }
+        return 100;
+    }
+    // Kiterino end
+
     private BlockState getCorrectStateForFluidBlock(Level world, BlockState state, BlockPos pos) {
         FluidState fluid = state.getFluidState();
 
