From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Jan 2023 00:50:12 +0200
Subject: [PATCH] Make boats buoyancy under flowing water to workaround
 MC-91206


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index 92795003b3f7a817d497eb01ac61efe9a5b27fa8..db0e5ba73938c4db98723c964123b17b5dd5d113 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -425,6 +425,15 @@ public class Boat extends VehicleEntity implements VariantHolder<Boat.Type> {
             }
         }
 
+        // Kiterino start - Workaround MC-91206; make boats buoyancy under flowing water
+        if (isUnderWater() && hasControllingPassenger()) {
+            FluidState fluidAbove = this.level().getFluidState(blockPosition().above());
+            if (fluidAbove.is(FluidTags.WATER) && !fluidAbove.isSource() && fluidAbove.getAmount() != 8) {
+                setDeltaMovement(getDeltaMovement().add(0, 0.5, 0));
+            }
+        }
+        // Kiterino end
+
     }
 
     private void tickBubbleColumn() {
@@ -670,7 +679,7 @@ public class Boat extends VehicleEntity implements VariantHolder<Boat.Type> {
 
                     if (fluid.is(FluidTags.WATER) && d0 < (double) ((float) blockposition_mutableblockposition.getY() + fluid.getHeight(this.level(), blockposition_mutableblockposition))) {
                         if (!fluid.isSource()) {
-                            return fluid.getAmount() == 8 ? Boat.Status.UNDER_FLOWING_WATER : Boat.Status.IN_WATER; // Kiterino - Do not sink boats in flowing water
+                            return Boat.Status.UNDER_FLOWING_WATER; // Kiterino - Do not sink boats in flowing water
                         }
 
                         flag = true;
