From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Jan 2023 01:04:02 +0200
Subject: [PATCH] Fix MC-2025


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 1ff2ee4d9064d20a469781a0f05fe9fd6ad592c8..50d6a838b223e9c0f678731d64d2483c616b642f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2578,6 +2578,19 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
                 nbttagcompound.putUUID("Kiterino.SpawningPlayerUUID", spawningPlayerUuid);
             }
             // Kiterino end
+
+            // Kiterino start - Fix MC-2025
+            // See: https://redd.it/8pgd4q
+            AABB boundingBox = getBoundingBox();
+            var boundingBoxList = new ListTag();
+            boundingBoxList.add(DoubleTag.valueOf(boundingBox.minX));
+            boundingBoxList.add(DoubleTag.valueOf(boundingBox.minY));
+            boundingBoxList.add(DoubleTag.valueOf(boundingBox.minZ));
+            boundingBoxList.add(DoubleTag.valueOf(boundingBox.maxX));
+            boundingBoxList.add(DoubleTag.valueOf(boundingBox.maxY));
+            boundingBoxList.add(DoubleTag.valueOf(boundingBox.maxZ));
+            nbttagcompound.put("Kiterino.BoundingBox", boundingBoxList);
+            // Kiterino end
             return nbttagcompound;
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Saving entity NBT");
@@ -2655,6 +2668,21 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
                         this.reapplyPosition();
                     }
 
+                    // Kiterino start - Fix MC-2025
+                    // See: https://redd.it/8pgd4q
+                    if (nbt.contains("Kiterino.BoundingBox", net.minecraft.nbt.Tag.TAG_LIST)) {
+                        ListTag boundingBoxList = nbt.getList("Kiterino.BoundingBox", net.minecraft.nbt.Tag.TAG_DOUBLE);
+                        setBoundingBox(new AABB(
+                                boundingBoxList.getDouble(0),
+                                boundingBoxList.getDouble(1),
+                                boundingBoxList.getDouble(2),
+                                boundingBoxList.getDouble(3),
+                                boundingBoxList.getDouble(4),
+                                boundingBoxList.getDouble(5)
+                        ));
+                    }
+                    // Kiterino end
+
                 } else {
                     throw new IllegalStateException("Entity has invalid rotation");
                 }
