From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Jan 2023 00:40:34 +0200
Subject: [PATCH] Expand Pose API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 056714a3844faa0ba50ee0913ea532fd875612f2..3ba543466b971ca94018fb066df52b650a8e790a 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -418,6 +418,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean collidingWithWorldBorder; // Paper
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    public boolean fixedPose = false; // Kiterino
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -770,6 +771,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public void onClientRemoval() {}
 
     public void setPose(net.minecraft.world.entity.Pose pose) {
+        // Kiterino start
+        if (fixedPose) {
+            return;
+        }
+        // Kiterino end
         // CraftBukkit start
         if (pose == this.getPose()) {
             return;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 123a3395a69c7093f3a9ef023cfb5b325b236850..646aed0ee6b8f6ad17414d1eb7644c5741ca9298 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1511,5 +1511,18 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public UUID getSpawningPlayerUuid() {
         return getHandle().spawningPlayerUuid;
     }
+
+    @Override
+    public void setPose(Pose pose, boolean fixed) {
+        Preconditions.checkNotNull(pose, "Pose cannot be null");
+        getHandle().fixedPose = false;
+        getHandle().setPose(net.minecraft.world.entity.Pose.values()[pose.ordinal()]);
+        getHandle().fixedPose = fixed;
+    }
+
+    @Override
+    public boolean hasFixedPose() {
+        return getHandle().fixedPose;
+    }
     // Kiterino end
 }
