From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Jan 2023 01:34:53 +0200
Subject: [PATCH] Fix MC-145311 - consider entities in water cauldrons being
 wet


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 119ff585e8b970d37a3531095e2b50d1a8391ac1..dc44fcf0f45874cf27db5527bcb3b41a28877ee8 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1891,6 +1891,18 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return this.isInWaterOrBubble() || this.isInLava();
     }
 
+    // Kiterino start
+    public boolean isInCauldronFilledWithWater() {
+        BlockState state = getFeetBlockState();
+        if (!state.is(Blocks.WATER_CAULDRON)) {
+            return false;
+        }
+        BlockPos pos = blockPosition();
+        double contentHeight = (6.0D + (double) state.getValue(net.minecraft.world.level.block.LayeredCauldronBlock.LEVEL) * 3.0D) / 16.0D; // Copypaste from LayeredCauldronBlock#getContentHeight
+        return getY() < pos.getY() + contentHeight && getBoundingBox().maxY > pos.getY() + 0.25D; // Copypaste from AbstractCauldronBlock#isEntityInsideContent
+    }
+    // Kiterino end
+
     public boolean isUnderWater() {
         return this.wasEyeInWater && this.isInWater();
     }
@@ -1925,7 +1937,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             }
         }
 
-        if (this.updateFluidHeightAndDoFluidPushing(FluidTags.WATER, 0.014D)) {
+        if (this.updateFluidHeightAndDoFluidPushing(FluidTags.WATER, 0.014D) || isInCauldronFilledWithWater()) { // Kiterino
             if (!this.wasTouchingWater && !this.firstTick) {
                 this.doWaterSplashEffect();
             }
