From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 28 Dec 2022 18:24:53 +0200
Subject: [PATCH] Add PlayerArmSwingFromServerEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 661ead674eebf2284579823a02e0bef178acc94f..cfc1f8e485c49e224365e76ee0cd0ecb10f26beb 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2402,6 +2402,7 @@ public abstract class LivingEntity extends Entity {
                 ServerChunkCache chunkproviderserver = ((ServerLevel) this.level).getChunkSource();
 
                 if (fromServerPlayer) {
+                    if (this instanceof ServerPlayer serverPlayer && new me.sosedik.kiterino.event.player.PlayerArmSwingFromServerEvent(serverPlayer.getBukkitEntity(), hand == InteractionHand.MAIN_HAND ? org.bukkit.inventory.EquipmentSlot.HAND : org.bukkit.inventory.EquipmentSlot.OFF_HAND).callEvent()) // Kiterino
                     chunkproviderserver.broadcastAndSend(this, packetplayoutanimation);
                 } else {
                     chunkproviderserver.broadcast(this, packetplayoutanimation);
