From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Jan 2023 01:34:53 +0200
Subject: [PATCH] Fix MC-145311 - consider entities in water cauldrons being
 wet


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index a5c87b6ae57959a087e6183fb4797520f1e8874b..16b2c60949bef8fd099b8df30d9bae1512228c02 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1804,6 +1804,18 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return this.isInWater() || this.isInBubbleColumn();
     }
 
+    // Kiterino start
+    public boolean isInCauldronFilledWithWater() {
+        BlockState state = getFeetBlockState();
+        if (!state.is(Blocks.WATER_CAULDRON)) {
+            return false;
+        }
+        BlockPos pos = blockPosition();
+        double contentHeight = (6.0D + (double) state.getValue(net.minecraft.world.level.block.LayeredCauldronBlock.LEVEL) * 3.0D) / 16.0D; // Copypaste from LayeredCauldronBlock#getContentHeight
+        return getY() < pos.getY() + contentHeight && getBoundingBox().maxY > pos.getY() + 0.25D; // Copypaste from AbstractCauldronBlock#isEntityInsideContent
+    }
+    // Kiterino end
+
     public boolean isUnderWater() {
         return this.wasEyeInWater && this.isInWater();
     }
@@ -1838,7 +1850,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             }
         }
 
-        if (this.updateFluidHeightAndDoFluidPushing(FluidTags.WATER, 0.014D)) {
+        if (this.updateFluidHeightAndDoFluidPushing(FluidTags.WATER, 0.014D) || isInCauldronFilledWithWater()) { // Kiterino
             if (!this.wasTouchingWater && !this.firstTick) {
                 this.doWaterSplashEffect();
             }
