From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 10 Jan 2023 06:05:12 +0200
Subject: [PATCH] Don't eject players from boat underwater automatically


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index 641170407c8c487b6faf66497cc9f2b08090c958..d87d8cce6c330ae04b1aa7a69264b54a187cbbef 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -357,7 +357,15 @@ public class Boat extends Entity implements VariantHolder<Boat.Type> {
         }
 
         if (!this.level.isClientSide && this.outOfControlTicks >= 60.0F) {
-            this.ejectPassengers();
+            // Kiterino start - Don't eject players
+            //this.ejectPassengers();
+            for (int i = this.passengers.size() - 1; i >= 0; --i) {
+                Entity passenger = this.passengers.get(i);
+                if (!(passenger instanceof Player)) {
+                    passenger.stopRiding();
+                }
+            }
+            // Kiterino end
         }
 
         if (this.getHurtTime() > 0) {
