From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 15 Jan 2023 08:08:54 +0200
Subject: [PATCH] Call BlockDestroyEvent for breakNaturally


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 4d999e8b798506267633517e41acbb07d4313b7f..73944f321cc3d86f647020ad56f28c3dbc7e4960 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -501,9 +501,23 @@ public class CraftBlock implements Block {
         net.minecraft.world.item.ItemStack nmsItem = CraftItemStack.asNMSCopy(item);
         boolean result = false;
 
+        // Kiterino start
+        boolean dropResources = true;
+        int expDrop = dropExperience ? block.getExpDrop(iblockdata, this.world.getMinecraftWorld(), this.position, nmsItem, true) : 0;
+        if (block != Blocks.AIR) {
+            var blockDestroyEvent = new com.destroystokyo.paper.event.block.BlockDestroyEvent(this, iblockdata.getFluidState().createLegacyBlock().createCraftBlockData(), iblockdata.createCraftBlockData(), expDrop, true);
+            blockDestroyEvent.setPlayEffect(triggerEffect);
+            if (!blockDestroyEvent.callEvent()) return false;
+
+            triggerEffect = blockDestroyEvent.playEffect();
+            dropResources = blockDestroyEvent.willDrop();
+            expDrop = blockDestroyEvent.getExpToDrop();
+        }
+        // Kiterino end
+
         // Modelled off EntityHuman#hasBlock
         if (block != Blocks.AIR && (item == null || !iblockdata.requiresCorrectToolForDrops() || nmsItem.isCorrectToolForDrops(iblockdata))) {
-            net.minecraft.world.level.block.Block.dropResources(iblockdata, this.world.getMinecraftWorld(), this.position, this.world.getBlockEntity(this.position), null, nmsItem, false); // Paper
+            if (dropResources) net.minecraft.world.level.block.Block.dropResources(iblockdata, this.world.getMinecraftWorld(), this.position, this.world.getBlockEntity(this.position), null, nmsItem, false); // Paper // Kiterino
             // Paper start - improve Block#breanNaturally
             if (triggerEffect) {
                 if (iblockdata.getBlock() instanceof net.minecraft.world.level.block.BaseFireBlock) {
@@ -512,7 +526,7 @@ public class CraftBlock implements Block {
                     this.world.levelEvent(net.minecraft.world.level.block.LevelEvent.PARTICLES_DESTROY_BLOCK, this.position, net.minecraft.world.level.block.Block.getId(iblockdata));
                 }
             }
-            if (dropExperience) block.popExperience(this.world.getMinecraftWorld(), this.position, block.getExpDrop(iblockdata, this.world.getMinecraftWorld(), this.position, nmsItem, true));
+            if (expDrop > 0) block.popExperience(this.world.getMinecraftWorld(), this.position, expDrop); // Kiterino
             // Paper end
             result = true;
         }
