From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 12 Jan 2023 02:20:10 +0200
Subject: [PATCH] Always allow to enter the boat


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index 39a6620ae6ec46f5ce77e8022bb2e0f0d71f2fc2..6c729980e4bd21e5c413203eb3d04bdca27cbd01 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -915,7 +915,7 @@ public class Boat extends Entity implements VariantHolder<Boat.Type> {
 
     @Override
     public InteractionResult interact(Player player, InteractionHand hand) {
-        return player.isSecondaryUseActive() ? InteractionResult.PASS : (this.outOfControlTicks < 60.0F ? (!this.level().isClientSide ? (player.startRiding(this) ? InteractionResult.CONSUME : InteractionResult.PASS) : InteractionResult.SUCCESS) : InteractionResult.PASS);
+        return player.isSecondaryUseActive() ? InteractionResult.PASS : (true || this.outOfControlTicks < 60.0F ? (!this.level().isClientSide ? (player.startRiding(this) ? InteractionResult.CONSUME : InteractionResult.PASS) : InteractionResult.SUCCESS) : InteractionResult.PASS); // Kiterino - always allow to enter the boat
     }
 
     @Override
@@ -1011,7 +1011,7 @@ public class Boat extends Entity implements VariantHolder<Boat.Type> {
 
     @Override
     protected boolean canAddPassenger(Entity passenger) {
-        return this.getPassengers().size() < this.getMaxPassengers() && !this.isEyeInFluid(FluidTags.WATER);
+        return this.getPassengers().size() < this.getMaxPassengers()/* && !this.isEyeInFluid(FluidTags.WATER)*/; // Kiterino - always allow to enter the boat
     }
 
     protected int getMaxPassengers() {
