From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 21 Jan 2023 20:26:44 +0200
Subject: [PATCH] More FishHook API


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index 91497f5e6c07fcf1b05eca6846c51e1a15ed3bc0..d4d959ed82258baa894475142c5b8374613826fb 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -64,7 +64,7 @@ public class FishingHook extends Projectile {
     private int life;
     private int nibble;
     public int timeUntilLured;
-    private int timeUntilHooked;
+    public int timeUntilHooked; // Kiterino
     private float fishAngle;
     private boolean openWater;
     @Nullable
@@ -410,6 +410,16 @@ public class FishingHook extends Projectile {
                     this.fishAngle = Mth.nextFloat(this.random, this.minLureAngle, this.maxLureAngle);
                     this.timeUntilHooked = Mth.nextInt(this.random, this.minLureTime, this.maxLureTime);
                     // CraftBukkit end
+                    // Kiterino start
+                    if (this.getPlayerOwner() != null) {
+                        var playerFishEvent = new PlayerFishEvent((Player) this.getPlayerOwner().getBukkitEntity(), null, (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.LURED);
+                        if (!playerFishEvent.callEvent()) {
+                            this.fishAngle = 0F;
+                            this.timeUntilHooked = 0;
+                            return;
+                        }
+                    }
+                    // Kiterino end
                 }
             } else {
                 // CraftBukkit start - logic to modify fishing wait time
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
index 116f23a3ee9fa409d7bc34c3769b94ed2bb07183..8554a2fef80b610b0caa0a1ad482188cd61415eb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
@@ -207,4 +207,16 @@ public class CraftFishHook extends CraftProjectile implements FishHook {
         this.getHandle().timeUntilLured = ticks;
     }
     // Paper end
+
+    // Kiterino start
+    @Override
+    public int getTimeUntilHooked() {
+        return getHandle().timeUntilHooked;
+    }
+
+    @Override
+    public void setTimeUntilHooked(int ticks) {
+        getHandle().timeUntilHooked = ticks;
+    }
+    // Kiterino end
 }
