From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 Feb 2023 03:42:48 +0200
Subject: [PATCH] Chest isBlocked API


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java b/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java
index 09dc268cf5a6ba675ed198e03f252c944f4f7117..ff8c472b459669249d260f782970c78131848125 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftChest.java
@@ -94,4 +94,11 @@ public class CraftChest extends CraftLootable<ChestBlockEntity> implements Chest
         return getTileEntity().openersCounter.opened;
     }
     // Paper end - More Lidded Block API
+
+    // Kiterino start
+    @Override
+    public boolean isBlocked() {
+        return ChestBlock.isChestBlockedAt(getWorldHandle(), getPosition());
+    }
+    // Kiterino end
 }
