From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sun, 15 Jan 2023 08:08:54 +0200
Subject: [PATCH] Call BlockDestroyEvent for breakNaturally


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index b6c4f3d34603929a56073444be5f25fda15ff0c7..5c6f6dce5241f175afc3ece917253848878bcbee 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -526,9 +526,21 @@ public class CraftBlock implements Block {
         net.minecraft.world.item.ItemStack nmsItem = CraftItemStack.asNMSCopy(item);
         boolean result = false;
 
+        // Kiterino start
+        boolean dropResources = true;
+        if (block != Blocks.AIR) {
+            var blockDestroyEvent = new com.destroystokyo.paper.event.block.BlockDestroyEvent(this, iblockdata.getFluidState().createLegacyBlock().createCraftBlockData(), true);
+            blockDestroyEvent.setPlayEffect(triggerEffect);
+            if (!blockDestroyEvent.callEvent()) return false;
+
+            triggerEffect = blockDestroyEvent.playEffect();
+            dropResources = blockDestroyEvent.willDrop();
+        }
+        // Kiterino end
+
         // Modelled off EntityHuman#hasBlock
         if (block != Blocks.AIR && (item == null || !iblockdata.requiresCorrectToolForDrops() || nmsItem.isCorrectToolForDrops(iblockdata))) {
-            net.minecraft.world.level.block.Block.dropResources(iblockdata, this.world.getMinecraftWorld(), position, this.world.getBlockEntity(position), null, nmsItem);
+            if (dropResources) net.minecraft.world.level.block.Block.dropResources(iblockdata, this.world.getMinecraftWorld(), position, this.world.getBlockEntity(position), null, nmsItem); // Kiterino
             // Paper start - improve Block#breanNaturally
             if (triggerEffect) {
                 if (iblockdata.getBlock() instanceof net.minecraft.world.level.block.BaseFireBlock) {
