From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 20 Feb 2023 15:31:35 +0200
Subject: [PATCH] Don't dismount ArmorStands from entities unless forced


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index ffe99cff6a7085ba25c9c5569f1444cee4216e31..61058ad493f5638b1bbc969df667ab9195ff20bc 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3805,6 +3805,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
     @Override
     public void stopRiding(boolean suppressCancellation) {
+        if (!suppressCancellation && isAlive() && valid && this instanceof net.minecraft.world.entity.decoration.ArmorStand) return; // Kiterino - don't dismount ArmorStands from entities unless forced
         // Paper end - Force entity dismount during teleportation
         Entity entity = this.getVehicle();
 
