From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 20 Feb 2023 15:31:35 +0200
Subject: [PATCH] Don't dismount ArmorStands from entities unless forced


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index e1db75df41dd451cc6a225fdf0285f078982e29d..6445959785f156ae35b5084b8fc3dd81b1a70dfe 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3775,6 +3775,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
     @Override
     public void stopRiding(boolean suppressCancellation) {
+        if (!suppressCancellation && isAlive() && valid && this instanceof net.minecraft.world.entity.decoration.ArmorStand) return; // Kiterino - don't dismount ArmorStands from entities unless forced
         // Paper end
         Entity entity = this.getVehicle();
 
