From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 Feb 2023 04:21:42 +0200
Subject: [PATCH] Additional drop method


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 207cd21508fc3c7890633ed432c177ccf4488de6..d194292af7c05fb0fd46dcd7b7cae06ba6e93714 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -887,5 +887,14 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     public int getActiveContainerStateId() {
         return getHandle().containerMenu.getStateId();
     }
+
+    @Override
+    public org.bukkit.entity.Item drop(ItemStack item) {
+        if (!(getHandle() instanceof ServerPlayer player)) {
+            return dropItemNaturally(item);
+        }
+        var drop = player.drop(CraftItemStack.asNMSCopy(item), false, true);
+        return drop != null && drop.getBukkitEntity() instanceof org.bukkit.entity.Item droppedItem ? droppedItem : null;
+    }
     // Kiterino end
 }
