From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 21 Jan 2023 20:26:44 +0200
Subject: [PATCH] More FishHook API


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index 15fb5ee374b19366ebb23181896fb943e95819f0..23e09f11451245e521513cd05fc9a16f001d0fef 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -63,7 +63,7 @@ public class FishingHook extends Projectile {
     private int life;
     private int nibble;
     public int timeUntilLured;
-    private int timeUntilHooked;
+    public int timeUntilHooked; // Kiterino
     private float fishAngle;
     private boolean openWater;
     @Nullable
@@ -401,6 +401,16 @@ public class FishingHook extends Projectile {
                 if (this.timeUntilLured <= 0) {
                     this.fishAngle = Mth.nextFloat(this.random, 0.0F, 360.0F);
                     this.timeUntilHooked = Mth.nextInt(this.random, 20, 80);
+                    // Kiterino start
+                    if (this.getPlayerOwner() != null) {
+                        var playerFishEvent = new PlayerFishEvent((Player) this.getPlayerOwner().getBukkitEntity(), null, (FishHook) this.getBukkitEntity(), PlayerFishEvent.State.LURED);
+                        if (!playerFishEvent.callEvent()) {
+                            this.fishAngle = 0F;
+                            this.timeUntilHooked = 0;
+                            return;
+                        }
+                    }
+                    // Kiterino end
                 }
             } else {
                 // CraftBukkit start - logic to modify fishing wait time
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
index 5e0c2c5094e1578162d1a50d50701fbd25e6d961..5f9afbc58225aeda1d540553b9f64fcdf9d3bf0c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
@@ -130,4 +130,16 @@ public class CraftFishHook extends CraftProjectile implements FishHook {
         this.getHandle().timeUntilLured = ticks;
     }
     // Paper end
+
+    // Kiterino start
+    @Override
+    public int getTimeUntilHooked() {
+        return getHandle().timeUntilHooked;
+    }
+
+    @Override
+    public void setTimeUntilHooked(int ticks) {
+        getHandle().timeUntilHooked = ticks;
+    }
+    // Kiterino end
 }
