From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 20 Feb 2023 15:31:35 +0200
Subject: [PATCH] Don't dismount ArmorStands from entities unless forced


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index dbfe567f024f455d37a1c49651fbd0a422f256fd..d51efd9147672f9ecaed08c7839dc1eadfc7d350 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3730,6 +3730,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
     @Override
     public void stopRiding(boolean suppressCancellation) {
+        if (!suppressCancellation && isAlive() && valid && this instanceof net.minecraft.world.entity.decoration.ArmorStand) return; // Kiterino - don't dismount ArmorStands from entities unless forced
         // Paper end
         Entity entity = this.getVehicle();
 
