From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 20 Feb 2023 15:31:35 +0200
Subject: [PATCH] Don't dismount ArmorStands from entities unless forced


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 691f88d2e58c0d88a7f74ab4e54a6cbcc1777077..8f75ee2af19c949bf845e580db644523cc3ba192 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3740,6 +3740,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
     @Override
     public void stopRiding(boolean suppressCancellation) {
+        if (!suppressCancellation && isAlive() && valid && this instanceof net.minecraft.world.entity.decoration.ArmorStand) return; // Kiterino - don't dismount ArmorStands from entities unless forced
         // Paper end
         Entity entity = this.getVehicle();
 
