From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 3 May 2023 21:31:27 +0300
Subject: [PATCH] Don't count Purpur's riding towards distance statistics


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index e3ea46e1df6ff274aac1fc6d3bf9e99fe1f9aa91..638c49cb0ba738b3df049a84381d8eb8d12c3a7c 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1796,6 +1796,11 @@ public abstract class Player extends LivingEntity {
 
             if (i > 0) {
                 Entity entity = this.getVehicle();
+                // Kiterino start
+                if (entity != null && entity.hasRider()) {
+                    return;
+                }
+                // Kiterino end
 
                 if (entity instanceof AbstractMinecart) {
                     this.awardStat(Stats.MINECART_ONE_CM, i);
