From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 22 Jun 2023 09:10:01 +0300
Subject: [PATCH] Additional Sign API


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
index ec96e8020c2042d939ab836464c547838447d708..81b3dea626febc1be3d93df227676802891f573d 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
@@ -225,4 +225,16 @@ public class CraftSign<T extends SignBlockEntity> extends CraftBlockEntityState<
     private static String revertComponent(Component component) {
         return CraftChatMessage.fromComponent(component);
     }
+
+    // Kiterino start
+    @Override
+    public java.util.UUID getWhoMayEditUuid() {
+        return getTileEntity().getPlayerWhoMayEdit();
+    }
+
+    @Override
+    public void setWhoMayEditUuid(@NotNull java.util.UUID uuid) {
+        getTileEntity().setAllowedPlayerEditor(uuid);
+    }
+    // Kiterino end
 }
