From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 3 May 2023 21:27:39 +0300
Subject: [PATCH] Require wearing elytra for elytra distance statistics


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 01c1363a9a9c2c21eac1a9fee6b487552cdeea70..096fc4d0c5ccd3e09625c2ccd31d1662e9660671 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1767,8 +1767,13 @@ public abstract class Player extends LivingEntity {
                     }
                 }
             } else if (this.isFallFlying()) {
-                i = Math.round((float) Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
-                this.awardStat(Stats.AVIATE_ONE_CM, i);
+                // Kiterino start
+                ItemStack chestItem = getItemBySlot(EquipmentSlot.CHEST);
+                if (chestItem.is(Items.ELYTRA) && ElytraItem.isFlyEnabled(chestItem)) {
+                    i = Math.round((float) Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
+                    this.awardStat(Stats.AVIATE_ONE_CM, i);
+                }
+                // Kiterino end
             } else {
                 i = Math.round((float) Math.sqrt(dx * dx + dz * dz) * 100.0F);
                 if (i > 25) {
