From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 27 Jun 2023 16:52:18 +0300
Subject: [PATCH] Expose checkBedExists


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index a2d1c22e778aa4c3c44c3a1a604c1656cc86f0da..7f002ded11c00f981dca0403a6823074fba338c4 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1146,4 +1146,11 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().setShouldBurnInDay(shouldBurnInDay);
     }
     // Purpur end
+
+    // Kiterino start
+    @Override
+    public boolean bedExists() {
+        return getHandle().checkBedExists();
+    }
+    // Kiterino end
 }
