From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 27 Jun 2023 20:28:13 +0300
Subject: [PATCH] Add missing EntityBlockChange


diff --git a/src/main/java/net/minecraft/world/item/HoeItem.java b/src/main/java/net/minecraft/world/item/HoeItem.java
index 18cbcfcd331f7b7a112383311e3e2b9984857028..d71d53ab6ce15e4e0977ea2f52f46e50d289a192 100644
--- a/src/main/java/net/minecraft/world/item/HoeItem.java
+++ b/src/main/java/net/minecraft/world/item/HoeItem.java
@@ -61,6 +61,11 @@ public class HoeItem extends DiggerItem {
             // Purpur end
             if (predicate.test(context)) {
                 Player player = context.getPlayer();
+                // Kiterino start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, blockPos, tillable.into().defaultBlockState())) {
+                    return InteractionResult.PASS;
+                }
+                // Kiterino end
                 if (!TILLABLES.containsKey(clickedBlock)) level.playSound(null, blockPos, SoundEvents.HOE_TILL, SoundSource.BLOCKS, 1.0F, 1.0F); // Purpur - force sound
                 if (!level.isClientSide) {
                     consumer.accept(context);
