From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 5 Jul 2023 19:06:27 +0300
Subject: [PATCH] Fixup recipe grid with ASItems


diff --git a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
index a65d07cb3d2dae85225921498d1721dc3c6a6bf3..3aad6a85f43548d97af520918aea896e250d9144 100644
--- a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
+++ b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -153,6 +153,11 @@ public class ServerPlaceRecipe<C extends Container> implements PlaceRecipe<Integ
         if (i != -1) {
             ItemStack itemStack = this.inventory.getItem(i);
             if (!itemStack.isEmpty()) {
+                // Kiterino start
+                if (!slot.getItem().isEmpty() && !ItemStack.isSameItemSameTags(stack, itemStack)) {
+                    return;
+                }
+                // Kiterino end
                 if (itemStack.getCount() > 1) {
                     this.inventory.removeItem(i, 1);
                 } else {
