From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 5 Jul 2023 19:06:27 +0300
Subject: [PATCH] Fixup recipe grid with ASItems


diff --git a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
index c8e9e85728d6b60bb829e69c015ab3c5172afd54..2406eae5d71b18c6faa680713d9ed9590618b5e3 100644
--- a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
+++ b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -134,6 +134,11 @@ public class ServerPlaceRecipe<C extends Container> implements PlaceRecipe<Integ
         if (i != -1) {
             ItemStack itemStack = this.inventory.getItem(i);
             if (!itemStack.isEmpty()) {
+                // Kiterino start
+                if (!slot.getItem().isEmpty() && !ItemStack.isSameItemSameTags(stack, itemStack)) {
+                    return;
+                }
+                // Kiterino end
                 if (itemStack.getCount() > 1) {
                     this.inventory.removeItem(i, 1);
                 } else {
