From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 21 Oct 2023 00:20:09 +0300
Subject: [PATCH] Allow checking for Halloween season


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 50a3680d5290e101db34640ce4cc40571799aa02..101cb3b8212197bfb4dcc2f891f40cf2d2cc1e47 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2633,4 +2633,15 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return result == net.minecraft.world.InteractionResult.CONSUME;
     }
     // Paper end
+
+    // Kiterino start
+    /**
+     * Checks whether it's currently Halloween season
+     *
+     * @return whether it's currently Halloween season
+     */
+    public boolean isHalloweenSeason() {
+        return net.minecraft.world.entity.ambient.Bat.isHalloweenSeason(getHandle());
+    }
+    // Kiterino end
 }
