From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 Nov 2023 15:44:55 +0200
Subject: [PATCH] Teleport with passengers by default


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index d27aaca91f091b4308f4888ebe459bb33fe1f86f..b2a338989aee332fb2c73a5a89cf100fbcfd3c7f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -324,7 +324,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     @Override
     public boolean teleport(Location location, TeleportCause cause) {
         // Paper start - Teleport passenger API
-        return teleport(location, cause, new io.papermc.paper.entity.TeleportFlag[0]);
+        return teleport(location, cause, io.papermc.paper.entity.TeleportFlag.EntityState.RETAIN_PASSENGERS); // Kiterino - retain passengers by default
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index e9af03a8520f5f85aae619750fa3fb215e52ae4e..a673342608e943dabccd3afdc9f8ef8ad4463ebb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1265,7 +1265,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     @Override
     public boolean teleport(Location location, PlayerTeleportEvent.TeleportCause cause) {
         // Paper start - Teleport API
-        return this.teleport(location, cause, new io.papermc.paper.entity.TeleportFlag[0]);
+        return this.teleport(location, cause, io.papermc.paper.entity.TeleportFlag.EntityState.RETAIN_PASSENGERS); // Kiterino - retain passengers by default
     }
 
     @Override
