From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 Nov 2023 15:45:58 +0200
Subject: [PATCH] Remove collision from leaves on server-side


diff --git a/src/main/java/net/minecraft/world/level/block/Blocks.java b/src/main/java/net/minecraft/world/level/block/Blocks.java
index d72112f4f864a35421a395adda9cdeed3cd23b05..7615332a8746b05a7d48dd5504e9af5949327fa1 100644
--- a/src/main/java/net/minecraft/world/level/block/Blocks.java
+++ b/src/main/java/net/minecraft/world/level/block/Blocks.java
@@ -656,6 +656,7 @@ public class Blocks {
                 .ignitedByLava()
                 .pushReaction(PushReaction.DESTROY)
                 .isRedstoneConductor(Blocks::never)
+                .noCollission() // Kiterino - No collision for leaves
         )
     );
     public static final Block DARK_OAK_LEAVES = register("dark_oak_leaves", leaves(SoundType.GRASS));
@@ -674,6 +675,7 @@ public class Blocks {
                 .ignitedByLava()
                 .pushReaction(PushReaction.DESTROY)
                 .isRedstoneConductor(Blocks::never)
+                .noCollission() // Kiterino - No collision for leaves
         )
     );
     public static final Block AZALEA_LEAVES = register("azalea_leaves", leaves(SoundType.AZALEA_LEAVES));
@@ -7720,6 +7722,7 @@ public class Blocks {
                 .ignitedByLava()
                 .pushReaction(PushReaction.DESTROY)
                 .isRedstoneConductor(Blocks::never)
+                .noCollission() // Kiterino - No collision for leaves
         );
     }
 
