From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 8 Nov 2023 03:56:05 +0200
Subject: [PATCH] Set resource pack API


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index cda4544ae96a4fcb5c6c4483df67a59f1b53fd27..06a953b34bf59f32981b89c42dcdb1b55c3ebaae 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -101,7 +101,7 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     public final boolean hideOnlinePlayers;
     public final int entityBroadcastRangePercentage;
     public final String textFilteringConfig;
-    public final Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo;
+    public Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo; // Kiterino - remive final
     public final DataPackConfig initialDataPackConfiguration;
     public final Settings<DedicatedServerProperties>.MutableValue<Integer> playerIdleTimeout;
     public final Settings<DedicatedServerProperties>.MutableValue<Boolean> whiteList;
@@ -200,7 +200,7 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
         return null;
     }
 
-    private static Optional<MinecraftServer.ServerResourcePackInfo> getServerPackInfo(String url, String sha1, @Nullable String hash, boolean required, String prompt) {
+    public static Optional<MinecraftServer.ServerResourcePackInfo> getServerPackInfo(String url, String sha1, @Nullable String hash, boolean required, String prompt) { // Kiterino - private -> public
         if (url.isEmpty()) {
             return Optional.empty();
         } else {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6ccd715fc46eef9a84b3b03cbced4beda70fd684..38bb48f19883282bd8e346e2fcc442b7ce38282a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3314,5 +3314,14 @@ public final class CraftServer implements Server {
         Server.super.setMaxStackSize(material, maxStackSize);
         CraftItemStack.asNMSCopy(new ItemStack(material)).getItem().setMaxStackSize(maxStackSize);
     }
+
+    @Override
+    public void setResourcePack(String url, String hash, boolean required, net.kyori.adventure.text.Component prompt) {
+        if (url == null || hash == null || url.isEmpty()) {
+            getServer().getProperties().serverResourcePackInfo = Optional.empty();
+            return;
+        }
+        getServer().getProperties().serverResourcePackInfo = DedicatedServerProperties.getServerPackInfo(url, hash, null, required, net.kyori.adventure.text.serializer.gson.GsonComponentSerializer.gson().serialize(prompt));
+    }
     // Kiterino end
 }
