From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 8 Nov 2023 03:56:05 +0200
Subject: [PATCH] Set resource pack API


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index d53e0b67d847dfec2f4b118b5ca3f0ed1dc29ad6..eaf62ea70ed58a7f57442ef07661efc89f97bfbf 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -100,7 +100,7 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     public final boolean hideOnlinePlayers;
     public final int entityBroadcastRangePercentage;
     public final String textFilteringConfig;
-    public final Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo;
+    public Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo; // Kiterino - remive final
     public final DataPackConfig initialDataPackConfiguration;
     public final Settings<DedicatedServerProperties>.MutableValue<Integer> playerIdleTimeout;
     public final Settings<DedicatedServerProperties>.MutableValue<Boolean> whiteList;
@@ -199,7 +199,13 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
         return null;
     }
 
-    private static Optional<MinecraftServer.ServerResourcePackInfo> getServerPackInfo(String id, String url, String sha1, @Nullable String hash, boolean required, String prompt) {
+    // Kiterino start
+    public String getDefaultResourcePackId() {
+        return get("resource-pack-id", "");
+    }
+    // Kiterino end
+
+    public static Optional<MinecraftServer.ServerResourcePackInfo> getServerPackInfo(String id, String url, String sha1, @Nullable String hash, boolean required, String prompt) { // Kiterino - private -> public
         if (url.isEmpty()) {
             return Optional.empty();
         } else {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f575304b06c267b7b3a155e5092d5ef1a72d3e6d..9e8df3a0cb0b05d62bb41d47a7f02b21cd8597f4 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3351,5 +3351,14 @@ public final class CraftServer implements Server {
         Server.super.setMaxStackSize(material, maxStackSize);
         CraftItemStack.asNMSCopy(new ItemStack(material)).getItem().setMaxStackSize(maxStackSize);
     }
+
+    @Override
+    public void setResourcePack(String url, String hash, boolean required, net.kyori.adventure.text.Component prompt) {
+        if (url == null || hash == null || url.isEmpty()) {
+            getServer().getProperties().serverResourcePackInfo = Optional.empty();
+            return;
+        }
+        getServer().getProperties().serverResourcePackInfo = DedicatedServerProperties.getServerPackInfo(getServer().getProperties().getDefaultResourcePackId(), url, hash, null, required, net.kyori.adventure.text.serializer.gson.GsonComponentSerializer.gson().serialize(prompt));
+    }
     // Kiterino end
 }
