From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Mon, 6 Nov 2023 00:36:45 +0200
Subject: [PATCH] Check for FarmBlock instead of FARMLAND


diff --git a/src/main/java/net/minecraft/world/entity/animal/Rabbit.java b/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
index b3d563285c45ef8e4468c6633d5d8818d1018bd8..9b9176ccd9dd47e9d40f03a25b0b478fbd0e7a4f 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Rabbit.java
@@ -705,7 +705,7 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
         protected boolean isValidTarget(LevelReader world, BlockPos pos) {
             BlockState iblockdata = world.getBlockState(pos);
 
-            if (iblockdata.is(Blocks.FARMLAND) && this.wantsToRaid && !this.canRaid) {
+            if (/*iblockdata.is(Blocks.FARMLAND)*/ iblockdata.getBlock() instanceof net.minecraft.world.level.block.FarmBlock && this.wantsToRaid && !this.canRaid) { // Kiterino
                 iblockdata = world.getBlockState(pos.above());
                 if (iblockdata.getBlock() instanceof CarrotBlock && ((CarrotBlock) iblockdata.getBlock()).isMaxAge(iblockdata)) {
                     this.canRaid = true;
diff --git a/src/main/java/net/minecraft/world/level/block/AttachedStemBlock.java b/src/main/java/net/minecraft/world/level/block/AttachedStemBlock.java
index e98f566aaba3ff7e3058fbd53bcac467c2c63182..6ca67618ed6a009f00bdaf67c7a96c1c7cbe4b68 100644
--- a/src/main/java/net/minecraft/world/level/block/AttachedStemBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/AttachedStemBlock.java
@@ -43,7 +43,7 @@ public class AttachedStemBlock extends BushBlock {
 
     @Override
     protected boolean mayPlaceOn(BlockState floor, BlockGetter world, BlockPos pos) {
-        return floor.is(Blocks.FARMLAND);
+        return /*floor.is(Blocks.FARMLAND)*/ floor.getBlock() instanceof FarmBlock; // Kiterino
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/level/block/BushBlock.java b/src/main/java/net/minecraft/world/level/block/BushBlock.java
index ca906b0250e5332f7ececf1419ca6d2c1d385adc..bd302e91aba3783eb1d64d9796272da883bd68e5 100644
--- a/src/main/java/net/minecraft/world/level/block/BushBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BushBlock.java
@@ -17,7 +17,7 @@ public class BushBlock extends Block {
     }
 
     protected boolean mayPlaceOn(BlockState floor, BlockGetter world, BlockPos pos) {
-        return floor.is(BlockTags.DIRT) || floor.is(Blocks.FARMLAND);
+        return floor.is(BlockTags.DIRT) || /*floor.is(Blocks.FARMLAND)*/ floor.getBlock() instanceof FarmBlock; // Kiterino
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/level/block/CropBlock.java b/src/main/java/net/minecraft/world/level/block/CropBlock.java
index 61d051f42d8c3d1f039b97fdc7a61b54f57ee86a..12da74e0d66eca7c454037f9ddce31dd3dff9440 100644
--- a/src/main/java/net/minecraft/world/level/block/CropBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CropBlock.java
@@ -40,7 +40,7 @@ public class CropBlock extends BushBlock implements BonemealableBlock {
 
     @Override
     protected boolean mayPlaceOn(BlockState floor, BlockGetter world, BlockPos pos) {
-        return floor.is(Blocks.FARMLAND);
+        return /*floor.is(Blocks.FARMLAND)*/ floor.getBlock() instanceof FarmBlock; // Kiterino
     }
 
     protected IntegerProperty getAgeProperty() {
@@ -125,7 +125,7 @@ public class CropBlock extends BushBlock implements BonemealableBlock {
                 float f1 = 0.0F;
                 BlockState iblockdata = world.getBlockState(blockposition1.offset(i, 0, j));
 
-                if (iblockdata.is(Blocks.FARMLAND)) {
+                if (/*iblockdata.is(Blocks.FARMLAND)*/iblockdata.getBlock() instanceof FarmBlock) { // Kiterino
                     f1 = 1.0F;
                     if ((Integer) iblockdata.getValue(FarmBlock.MOISTURE) > 0) {
                         f1 = 3.0F;
diff --git a/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java b/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
index 1641453fd17938bc8fcdba7f89a0515ebeef20c5..c06b58bb42f770bb4a38c2aae106f80b6acb4f98 100644
--- a/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
@@ -76,7 +76,7 @@ public class PitcherCropBlock extends DoublePlantBlock implements BonemealableBl
 
     @Override
     protected boolean mayPlaceOn(BlockState floor, BlockGetter world, BlockPos pos) {
-        return floor.is(Blocks.FARMLAND);
+        return /*floor.is(Blocks.FARMLAND)*/ floor.getBlock() instanceof FarmBlock; // Kiterino
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/level/block/StemBlock.java b/src/main/java/net/minecraft/world/level/block/StemBlock.java
index 0d9be3b28ba4957c3f4da5455283fca903836c91..3eed7ee5d05ef0f46e46365f0df67060c92533cb 100644
--- a/src/main/java/net/minecraft/world/level/block/StemBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/StemBlock.java
@@ -45,7 +45,7 @@ public class StemBlock extends BushBlock implements BonemealableBlock {
 
     @Override
     protected boolean mayPlaceOn(BlockState floor, BlockGetter world, BlockPos pos) {
-        return floor.is(Blocks.FARMLAND);
+        return /*floor.is(Blocks.FARMLAND)*/ floor.getBlock() instanceof FarmBlock; // Kiterino
     }
 
     @Override
@@ -64,7 +64,7 @@ public class StemBlock extends BushBlock implements BonemealableBlock {
                     BlockPos blockposition1 = pos.relative(enumdirection);
                     BlockState iblockdata1 = world.getBlockState(blockposition1.below());
 
-                    if (world.getBlockState(blockposition1).isAir() && (iblockdata1.is(Blocks.FARMLAND) || iblockdata1.is(BlockTags.DIRT))) {
+                    if (world.getBlockState(blockposition1).isAir() && (/*iblockdata1.is(Blocks.FARMLAND)*/ iblockdata1.getBlock() instanceof FarmBlock || iblockdata1.is(BlockTags.DIRT))) { // Kiterino
                         // CraftBukkit start
                         if (!CraftEventFactory.handleBlockGrowEvent(world, blockposition1, this.fruit.defaultBlockState())) {
                             return;
