From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 17 Nov 2023 04:07:09 +0200
Subject: [PATCH] Manually send hand swing for bonemeal interactions


diff --git a/src/main/java/net/minecraft/world/item/BoneMealItem.java b/src/main/java/net/minecraft/world/item/BoneMealItem.java
index 2cc68515addfc96cdf296a22f9d9896a8072108e..a7615fc6c62df9baf7892b2e668bb3419e214c54 100644
--- a/src/main/java/net/minecraft/world/item/BoneMealItem.java
+++ b/src/main/java/net/minecraft/world/item/BoneMealItem.java
@@ -52,6 +52,7 @@ public class BoneMealItem extends Item {
         if (BoneMealItem.growCrop(itemactioncontext.getItemInHand(), world, blockposition)) {
             if (!world.isClientSide) {
                 if (itemactioncontext.getPlayer() != null) itemactioncontext.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
+                if (itemactioncontext.getPlayer() != null) itemactioncontext.getPlayer().swing(itemactioncontext.getHand(), true); // Kiterino
                 if (showParticles) world.levelEvent(1505, blockposition, 0); // Paper - BoneMeal API
             }
 
@@ -63,6 +64,7 @@ public class BoneMealItem extends Item {
             if (flag && BoneMealItem.growWaterPlant(itemactioncontext.getItemInHand(), world, blockposition1, itemactioncontext.getClickedFace())) {
                 if (!world.isClientSide) {
                     if (itemactioncontext.getPlayer() != null) itemactioncontext.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
+                    if (itemactioncontext.getPlayer() != null) itemactioncontext.getPlayer().swing(itemactioncontext.getHand(), true); // Kiterino
                     if (showParticles) world.levelEvent(1505, blockposition1, 0); // Paper - BoneMeal API
                 }
 
