--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -724,6 +_,27 @@
         pipeline.addLast("hackfix", new ChannelOutboundHandlerAdapter() {
             @Override
             public void write(ChannelHandlerContext channelHandlerContext, Object object, ChannelPromise channelPromise) throws Exception {
+                // Kiterino start - Item Modifiers API
+                if (object instanceof Packet<?> packet) {
+                    net.minecraft.server.level.ServerPlayer player = getPlayer();
+                    object = me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.processPacket(player == null ? null : player.getBukkitEntity(), packet);
+                    // Kiterino start - Don't send custom mob effects - Data-driven mob effects
+                    if (packet instanceof net.minecraft.network.protocol.game.ClientboundUpdateMobEffectPacket mobEffectPacket) {
+                        java.util.Optional<net.minecraft.resources.ResourceKey<net.minecraft.world.effect.MobEffect>> key = mobEffectPacket.getEffect().unwrapKey();
+                        if (key.isPresent() && !org.bukkit.NamespacedKey.MINECRAFT.equals(key.get().location().getNamespace())) {
+                            packet.onPacketDispatchFinish(player, null);
+                            return;
+                        }
+                    } else if (packet instanceof net.minecraft.network.protocol.game.ClientboundRemoveMobEffectPacket mobEffectPacket) {
+                        java.util.Optional<net.minecraft.resources.ResourceKey<net.minecraft.world.effect.MobEffect>> key = mobEffectPacket.effect().unwrapKey();
+                        if (key.isPresent() && !org.bukkit.NamespacedKey.MINECRAFT.equals(key.get().location().getNamespace())) {
+                            packet.onPacketDispatchFinish(player, null);
+                            return;
+                        }
+                    }
+                    // Kiterino end - Don't send custom mob effects - Data-driven mob effects
+                }
+                // Kiterino end - Item Modifiers API
                 super.write(channelHandlerContext, object, channelPromise);
             }
         }).addLast("packet_handler", this);
