--- a/src/main/java/org/bukkit/inventory/CookingRecipe.java
+++ b/src/main/java/org/bukkit/inventory/CookingRecipe.java
@@ -20,6 +_,7 @@
     private int cookingTime;
     private String group = "";
     private CookingBookCategory category = CookingBookCategory.MISC;
+    private boolean special = false; // Kiterino - Extra recipes API
 
     /**
      * Create a cooking recipe to craft the specified ItemStack.
@@ -198,4 +_,27 @@
         Preconditions.checkArgument(category != null, "category cannot be null");
         this.category = category;
     }
+
+    // Kiterino start - Extra recipes API
+    /**
+     * Checks whether this recipe is special.
+     * <br>Special recipes do not get sent to the client
+     * and bypass the limited crafting game rule.
+     *
+     * @return whether this recipe is special
+     */
+    public boolean isSpecial() {
+        return this.special;
+    }
+
+    /**
+     * Sets whether this recipe is special
+     *
+     * @param special whether this recipe is special
+     * @see #isSpecial()
+     */
+    public void setSpecial(boolean special) {
+        this.special = special;
+    }
+    // Kiterino end - Extra recipes API
 }
